---
title: "Nanopore validation paper: Tamarin pipeline, no downsampling + 10x genotyping cutoff (ds0x_g10x)"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_knit$set(echo = TRUE, root.dir = "/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_g10x_toShare/")
```

# 1 Overview

This document outlines the analysis pipeline used for the tamarin subset
of the the Nanopore validation paper. Here, we're comparing **Illumina
vs. Nanopore** genotype results for **20 tamarin blood samples** at **77
SNP loci (1,540 genotypes total)**.

The scripts included here include those created by myself (R. Voyt) in
addition to those created by McKinney, Pascal, and Seeb (2020) as part
of their GTscore pipeline for the analysis of Genotyping-in-Thousands
sequencing data. I've included details on the GTscore scripts that I've
included here; information on the full GTscore pipeline be found on the
[GTscore GitHub page](https://github.com/gjmckinney/GTscore).

# 2 Packages

```{r, message=FALSE}
library(kableExtra, quietly = T)
library(tidyverse, quietly = T)
```

# 3 Data

## 3.1 Sample metadata

Our sample dataset is composed of **20 tamarin blood samples**,
including 10 saddleback tamarins (*Leontocebus weddelli*; n = 6
females, n = 4 males) and 10 emperor tamarins (*Saguinus imperator*; n
= 5 females, n = 5 males).

```{r, echo=FALSE}
md <- read.csv("metadata_paper1_nanoporeValidation_ds0x_g10x.csv")
```

```{r, echo=FALSE}
# For html
md %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 3.2 Loci

Our loci dataset is composed of **77 dual-species SNPs**, including SNPs
used for the identification of individuals (INDID; n = 68), sex (SEXID;
n = 3), and species (SPECIESID; n = 6). All 77 loci are informative for
both tamarin species.

```{r, echo=FALSE}
loci <- read.table("primerProbeFile_ds0x_g10x.txt", header = T)

loci_bySet <- data.frame(
  lociSet = c("INDID", "SEXID", "SPECIESID", "TOTAL"),
  count = c(
    sum(str_detect(loci$Locus, "INDID")),
    sum(str_detect(loci$Locus, "SEXID")),
    sum(str_detect(loci$Locus, "SPECIESID")),
    nrow(loci)
  )
)
```

```{r, echo=FALSE}
# For html
loci_bySet %>%
  kbl() %>%
  kable_styling(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato")
```

# 4 GTscore: Count allele reads

The first step in our analyses is to obtain allele read counts for each
sample at each locus for both Illumina and Nanopore datasets. Here I'm
using the AmpliconReadCounter.pl script from the GTscore pipeline, which can be found in the /scripts folder. An
explanation of the script is in Section 4.1 below - I've pulled most of
the explanatory text directly from the original GTscore documentation,
with some modifications for clarity.

## 4.1 Info on AmpliconReadCounter.pl

The read counter is written in perl (AmpliconReadCounter.pl), but can be
called from R. Running it does the following:

1.  Identifies each unique sequence, then counts the number of times
    each unique sequence occurs within an individual
2.  Aligns each unique sequence with forward primer and probe sequences
    (the "probe" here is a 16-nt sequence overlapping the SNP); if the
    sequence doesn't align, then it is excluded as an off-target
    sequence and reports by individual and by locus are given.

**Note:** By default, all primers are trimmed to the length of the
shortest primer to increase speed. Optionally the full length primer can
be used for the primer but this may significantly increase run time
depending on variation in primer lengths across loci.

**Input flags for this script are:**

-   --p a tab delimited file containing primer/probe information for
    each locus
-   --files a text file containing a list of .fastq sequence files to
    count reads from

**Optional flags:**

-   \--prefix optional prefix for output file names

-   \--inDir option to specify directory containing sequence data

-   \--outDir option to specify output directory (**Note:** This option
    is not listed in the original AmpliconReadCounter documentation, but
    is an option in the script itself)

-   \--inputType; options are fq or fastqgz (defaults to fastqgz)

-   \--useFullPrimer uses the full primer for counting reads rather than
    the trimmed primer

-   \--alleleOrder order of alleles output in locusTable file; options
    are original (matches primer-probe file order) or alphabetical
    (default)

-   \--printMatched outputs matched reads for each individual (**Note:**
    RV modified lines 325-329 to automatically create a "matchedReads"
    subdirectory within the specified "outDir" if this option is chosen)

-   \--printDiscarded outputs discarded reads for each individual
    (**Note:** RV modified lines 310-314 to automatically create an
    "unmatchedReads" subdirectory within the specified "outDir" if this
    option is chosen)

## 4.2 Run AmpliconReadCounter.pl

```{r, eval=FALSE}
# illumina dataset
system2("perl",
        args="./scripts/AmpliconReadCounter_modified.pl -p primerProbeFile_ds0x_g10x.txt --files sampleFiles_ill_ds0x_g10x.txt --inDir ./fastqFiles_ill/ --outDir . --prefix ill_ds0x_g0x_")

# nanopore dataset
system2("perl",
        args="./scripts/AmpliconReadCounter_modified.pl -p primerProbeFile_ds0x_g10x.txt --files sampleFiles_ont_ds0x_g10x.txt --inDir ./fastqFiles_ont/ --outDir . --prefix ont_ds0x_g0x_")
```

We can view the first few lines of the allele reads files below:

```{r, echo=FALSE}
# For html
alleleReads_ill <- read.delim("ill_ds0x_g0x_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  merge(., md[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  select(-sampleID) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

alleleReads_ont <- read.delim("ont_ds0x_g0x_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  merge(., md[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  select(-sampleID) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# tables
head(alleleReads_ill) %>%
  kbl(caption = "Illumina read counts") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

head(alleleReads_ont) %>%
  kbl(caption = "Nanopore read counts") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 5 Recode loci \<10x coverage

Genotyping by sequencing generally requires a minimum of 10x coverage to be considered reliable. However, the GTscore genotyping script doesn't include this cutoff, and will provide genotypes with as little as one read per allele for heterozygous calls and 3 reads (vs. 0) for a homozygous call.

As such, I'm adding a filtering step here to include a 10x cutoff, accomplished by recoding all loci with \<10 reads to "0" so that loci with lower than 10x coverage will not receive genotypes.

```{r, eval=FALSE}
# Read in allele counts for the full dataset
readCounts_original_ill <- read.table("ill_ds0x_g0x_AlleleReads_singleSNPs.txt")

readCounts_original_ont <- read.table("ont_ds0x_g0x_AlleleReads_singleSNPs.txt")

# Create read counts copy for 10x coverage
readCounts10x_ill <- readCounts_original_ill

readCounts10x_ont <- readCounts_original_ont

# Set up a function to sum the read counts per allele for each locus, using package gsubfn
repl <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# Then apply the function to readCounts to sum each set of allele reads for each locus
readCounts_sum10x_ill <- replace(readCounts10x_ill, TRUE, lapply(readCounts10x_ill, repl)) %>%
  mutate(across(everything(),as.numeric))

readCounts_sum10x_ont <- replace(readCounts10x_ont, TRUE, lapply(readCounts10x_ont, repl)) %>%
  mutate(across(everything(),as.numeric))

# Recode <10x loci with "0"
readCounts10x_ill[readCounts_sum10x_ill < 10] <- "0,0"

readCounts10x_ont[readCounts_sum10x_ont < 10] <- "0,0"

# Export new 10x AlleleReads_singleSNPs files
write.table(readCounts10x_ill, "ill_ds0x_g10x_AlleleReads_singleSNPs.txt", quote = FALSE, sep = "\t")

write.table(readCounts10x_ont, "ont_ds0x_g10x_AlleleReads_singleSNPs.txt", quote = FALSE, sep = "\t")
```

# 6 GTscore: Genotyping

Now that we have the 10x-filtered allele reads for Illumina and Nanopore
datasets, we can plug them into the GTscore *PolyGen* genotyping
function, which is included in the GTscore R scripts (imported in
Section 6.2 below). I've included an explanation of the function (pulled
from [McKinney et al.,
2018](https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12763)) in
Section 6.1.

## 5.1 Info on PolyGen genotyper

The *PolyGen* genotyper (originally described in [McKinney et al.,
2018](https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12763)) can
be used for both diploid and polyploid loci. It works by calling
genotypes based on allele dosage, which is inferred using a
maximum-likelihood algorithm that performs equations in the following
steps:

1.  Calculates relative dosage for each possible genotype
2.  Calculates the chance that a read will be sampled from a given
    allele $p(a)$ given a particular underlying genotype as a function
    of the relative dosage of the allele in the genotype as well as the
    sequencing error rate:
    -   $$
        p(a) = d_{r}(a) \times (1 - \varepsilon) + (1 - d_{r}(a)) \times \varepsilon
        $$

    -   Where $d_{r}(a)$ is the relative dosage of allele a and
        $\varepsilon$ is the sequencing error rate
3.  Calculates the overall log-likelihood of genotype $L(g)$ by summing
    the relative dosage log-likelihoods for each allele:
    -   $$
        L(g) = \sum_{a}^n ln(p(a)) \times c_{a}
        $$
    -   Where $p(a)$ is the chance that a read will be sampled from
        allele a, $c_{a}$ is the count of observations of allele a, and
        there are $n$ alleles
4.  Compares the two most likely genotypes via a likelihood ratio test
    with one degree of freedom; the most likely genotype is assigned if
    the likelihood ratio test is significant at $\alpha = 0.05$;
    otherwise no genotype is assigned

## 5.2 Import GTscore scripts

```{r, message=FALSE, warning=FALSE}
source("./scripts/GTscore_modified.R") # NOTE- I added an option for "\\[ATGC\\]"="N" (lines 616 and 617) to replace brackets in probe
```

The *PolyGen* R function takes the following as inputs:

-   locusTable - a text file containing a table of all loci names,
    ploidy, and alleles; this file is automatically created through the
    AmpliconReadCounter.pl script

-   readCounts - a text file containing a table of allele reads for all
    samples at all loci; this file is also automatically created via the
    AmpliconReadCounter.pl script

-   p_thresh - corresponds to significance level $\alpha$ in the
    equations above; default is 0.05

-   epsilon - corresponds to sequencing error rate $\varepsilon$ in the
    equations above; default is 0.01

## 5.3 Genotype samples

Below I'm using the *PolyGen* function to call genotypes for both
Illumina and Nanopore datasets. Note that the default sequencing error
rate is 0.01, which I've chosen not to alter here. Note also that
*PolyGen* requires both the locus table and allele reads files output by AmpliconReadCounter in Section
4.2.

```{r, eval=FALSE}
# load locus table
locusTable <- read.delim("ill_ds0x_g0x_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

# load allele reads
alleleReads_ill <- read.delim("ill_ds0x_g10x_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  merge(., md[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  select(-sampleID) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

alleleReads_ont <- read.delim("ont_ds0x_g10x_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  merge(., md[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  select(-sampleID) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# genotype: illumina
polyGenResults_ill <- polyGen(locusTable, alleleReads_ill)

# genotype: nanopore
polyGenResults_ont <- polyGen(locusTable, alleleReads_ont)
```

```{r, eval=FALSE, echo=FALSE}
# export genotypes
write.table(polyGenResults_ill, "ill_ds0x_g10x_polyGenResults.txt", quote = FALSE, sep = "\t")

write.table(polyGenResults_ont, "ont_ds0x_g10x_polyGenResults.txt", quote = FALSE, sep = "\t")
```

# 7 Genotype consistency

Now that we have our genotypes called from our downsampled allele reads,
we can compare the genotypes called for each sample at each locus
between Illumina and Nanopore datasets.

## 7.1 Import genotypes

First a bit of data reformatting to make comparisons easier -- we can
see the genotypes and allele reads for each sample at each locus for Illumina vs. Nanopore here:

```{r, echo=FALSE}
# illumina
genos_ill <- read.table("ill_ds0x_g10x_polyGenResults.txt", header = T, check.names = F)

genos_ill_l <- genos_ill %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ill")

alleleReads_ill_l <- alleleReads_ill %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_ill")
  
# nanopore
genos_ont <- read.table("ont_ds0x_g10x_polyGenResults.txt", header = T, check.names = F)

genos_ont_l <- genos_ont %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ont")

alleleReads_ont_l <- alleleReads_ont %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_ont")

# comparison df
compareGenos <- genos_ill_l %>%
  merge(., genos_ont_l, by = c("sampleID", "locus")) %>%
  mutate(
    genoMatch = case_when(
      genos_ill == "0" | genos_ont == "0" ~ NA,
      genos_ill == genos_ont ~ "yes",
      .default = "no"
    )
  ) %>%
  merge(., alleleReads_ill_l, by = c("locus", "sampleID")) %>%
  merge(., alleleReads_ont_l, by = c("locus", "sampleID"))
```

```{r, echo=FALSE}
# For html
compareGenos %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 7.2 Compare genotypes

Comparisons show that out of the full set of 1,540 genotypes from the 20 samples & 77 loci, **96% of the genotypes match** between Illumina and Nanopore datasets.

```{r, echo=FALSE}
compareGenos_table <- data.frame(
  genosMismatch = sum(compareGenos$genoMatch == "no"),
  genosMatch = sum(compareGenos$genoMatch == "yes"),
  genosNA = sum(is.na(compareGenos$genoMatch)),
  totalGenos = nrow(compareGenos)
) %>%
  mutate(
    propMismatch = round(genosMismatch/totalGenos, 2),
    propMatch = round(genosMatch/totalGenos, 2)
  ) %>%
  relocate(c(propMismatch, propMatch))
```

```{r, echo=FALSE}
# For html
compareGenos_table %>%
  kbl() %>%
  kable_styling(full_width = F, position = "center") %>%
  row_spec(0, color = "black", background = "tomato") %>%
  row_spec(0:1, align = "c")
```

On a **per-sample** basis, the proportion of genotypes matching between Illumina and Nanopore sequencing runs ranges from 0.74 to 1, with a median of 0.97.

```{r, echo=FALSE}
compareGenos_bySample <- compareGenos %>%
  group_by(sampleID) %>%
  summarise(
    propMatch = round((sum(genoMatch == "yes")/77), 2)
  )

compareGenos_bySample_table <- data.frame(
  med_propMatch = round(median(compareGenos_bySample$propMatch), 2),
  min_propMatch = min(compareGenos_bySample$propMatch),
  max_propMatch = max(compareGenos_bySample$propMatch)
)
```

```{r, echo=FALSE, message=FALSE}
# for html
compareGenos_bySample_table %>%
  kbl() %>%
  kable_styling(full_width = F, position = "center") %>%
  row_spec(0, color = "black", background = "tomato") %>%
  row_spec(0:1, align = "c")

compareGenos_bySample %>%
  ggplot(aes(x = propMatch, y = factor(0))) +
  geom_boxplot(color = "#0099f8") +
  theme_bw() +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(
    x = "Per-sample proportion of matching genotypes") +
  geom_dotplot(binaxis = "x", stackdir = "center", dotsize = 0.3)
```

And on a **per-locus** basis, the proportion of genotypes matching between Illumina and Nanopore sequencing runs ranges from 0.74 to 1, with a median of 0.97.

```{r, echo=FALSE}
compareGenos_byLocus <- compareGenos %>%
  group_by(locus) %>%
  summarise(
    propMatch = round((sum(genoMatch == "yes")/20), 2)
  )

compareGenos_byLocus_table <- data.frame(
  med_propMatch = round(median(compareGenos_byLocus$propMatch), 2),
  min_propMatch = min(compareGenos_byLocus$propMatch),
  max_propMatch = max(compareGenos_byLocus$propMatch)
)
```

```{r, echo=FALSE, message=FALSE}
# for html
compareGenos_byLocus_table %>%
  kbl() %>%
  kable_styling(full_width = F, position = "center") %>%
  row_spec(0, color = "black", background = "tomato") %>%
  row_spec(0:1, align = "c")

compareGenos_byLocus %>%
  ggplot(aes(x = propMatch, y = factor(0))) +
  geom_boxplot(color = "#0099f8") +
  theme_bw() +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(
    x = "Per-locus proportion of matching genotypes") +
  geom_dotplot(binaxis = "x", stackdir = "center", dotsize = 0.3)
```