---
title: "tamGeneticsc_paper4_chimerism"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: 
  rmdformats::downcute:
    downcute_theme: "chaos"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
```

```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto
}

pre[class] {
  max-height: 100px;
}
```

# 1 Overview

This set of scripts is for calculating chimerism percentages for each
tamarin invididual in a twin pair. In doing so, I'm following the protocols outlined in Vynck et al. 2023, which includes chimerism calculations for NGS SNP data as well as additional calculations to account for amplification bias of particular alleles.

Vynck, Matthijs, Friedel Nollet, Lode Sibbens, and Helena Devos. 2023. “Bias Reduction Improves Accuracy and Informativity of High-Throughput Sequencing Chimerism Assays.” Clinica Chimica Acta 547 (June): 117452. https://doi.org/10.1016/j.cca.2023.117452.

# 2 Packages

```{r}
library(gsubfn)
library(janitor)
library(kableExtra)
library(rlist)
library(tidyverse)

source("./project_scripts/ravo_ch2Scripts.R")
source("./project_scripts/ravo_gtScripts.R")

library(paletteer) # https://r-graph-gallery.com/color-palette-finder.html
my_colors_light <- paletteer::paletteer_d("ghibli::PonyoLight")
my_colors_med <- paletteer::paletteer_d("ghibli::PonyoMedium")
my_colors_peru <- paletteer::paletteer_d("MetBrewer::Peru1")
my_colors_klein <- paletteer::paletteer_d("MoMAColors::Klein")
my_colors_cont <- paletteer::paletteer_c("ggthemes::Sunset-Sunrise Diverging", n = 1)
my_colors <- paletteer::paletteer_d("rcartocolor::Antique")
```

# 3 Data

## 3.1 Metadata

```{r}
md <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv") %>%
  mutate(captureYear = str_sub(captureDate, 1, 4)) %>%
  relocate(captureYear, .after = captureDate)

sampleRef <- read.csv("./project_data/master_sampleInfo_v2.csv") %>%
  arrange(sampleID)
```

## 3.2 capData

Capture data includes all records of individual captures from 2009-2019.

Use the latest version of capData_byIndiv (from
tamGenetics_paper3_dataOrganization), then 1) filter to first capture
for each animalID 2) filter again to animalIDs in tamRun5

**note** exclude those in capData w/"UNK" animalID; did a lot of digging
in ch2_dataOrganization.Rmd and most likely these entries need to be
ditched

```{r}
capData_2009to2023 <- read.csv("./DISSERTATION/ch2_demography/00_data/01_demoData_clean/captureData_byIndividual_v6.csv") %>%
  # ditch UNK animalIDs
  filter(animalID != "UNK") %>%
  mutate(
    captureYear = str_sub(captureDate, 1, 4),
    captureYear = as.numeric(captureYear)
  ) %>%
  relocate(captureYear, .after = captureDate)

capData_2009to2019 <- capData_2009to2023 %>%
  filter(rowID <= 613)

capData_2009to2023_firstEntry <- capData_2009to2023[match(unique(capData_2009to2023$animalID), capData_2009to2023$animalID),] %>%
  select(rowID, captureDate, captureYear, animalID, ageClass, ageClass_classifier, animalName1, animalName2, groupName, species, sex, notes_MD, notes_RV)

capData_2009to2019_firstEntry <- capData_2009to2023_firstEntry %>%
  filter(rowID <= 613)
```

## 3.5 parityAssignments

Using "parity2" -- meaning either left or right (vs avg) meets threshold

```{r}
parityAssignments <- get_parityStatus(capData_file = capData_2009to2019) %>%
  select(animalID, captureDate, parity2) %>%
  dplyr::rename("parity" = "parity2") %>%
  mutate(
    captureYear = str_sub(captureDate, 1, 4)
  ) %>%
  #select(-captureDate) %>%
  distinct() %>%
  merge(., sampleRef[, c("animalID", "birthYear_est", "natalGroup")], by = "animalID", all.x = T) %>%
  merge(., capData_2009to2019[, c("animalID", "captureDate", "groupName")], by = c("animalID", "captureDate")) %>%
  dplyr::rename("capGroup" = "groupName")

parousOnly <- parityAssignments %>%
  filter(parity == "parous") %>%
  arrange(animalID, captureYear)

parityAssigments_firstParous <- parousOnly[match(unique(parousOnly$animalID), parousOnly$animalID),] %>%
  filter(parity == "parous") %>%
  dplyr::rename("parityYear" = "captureYear") %>%
  mutate(natalCap_match = natalGroup == capGroup)
```

## 3.3 Loci

List of all loci in version 3 of the primer pool

```{r}
lociRef <- read.csv("./project_data/master_lociInfo.csv")
```

## 3.4 Genos

```{r}
genos10x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_10x.txt", header = T, na.strings = "0") %>%
  `rownames<-`(sub('[_][^_]+$', '', rownames(.)))
```

## 3.5 Allele reads

```{r}
readCounts <- read.table("./05_tamRun5/03_run5GTscore/fullSet_AlleleReads_singleSNPs_10x.txt") %>%
  `rownames<-`(sub('[_][^_]+$', '', rownames(.))) %>%
  # replace 0,0 with NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))
```

# 4 Filter genos

## 4.1 Remove dups

Identify duplicate animalIDs (previously identified in metadataReconciliation) and keep the one with the highest genoSuccess

```{r}
# get genoSuccess
locusTable <- read.table("./05_tamRun5/03_run5GTscore/fullSet_LocusTable_singleSNPs.txt", header = T) %>%
  mutate(Locus_ID = (sub('[_][^_]+$', '', Locus_ID)))

genoSuccess <- get_genoSuccess(genos10x, locusTable, md, "bySample", sampleID_colName = "sampleID", exclude_nonTargetSp = "no")

# hair
hairDups <- md %>%
  filter(sampleType == "hair") %>%
  get_dupes(animalID) %>%
  merge(., genoSuccess, by = "sampleID") %>%
  relocate(genoSuccess) %>%
  arrange(animalID, genoSuccess)

toRemove_hairDups <- hairDups %>%
  group_by(animalID) %>%
  slice_min(genoSuccess) %>%
  ungroup() %>%
  select(sampleID) %>% pull()

# blood
bloodDups <- md %>%
  filter(sampleType == "blood") %>%
  get_dupes(animalID) %>%
  merge(., genoSuccess, by = "sampleID") %>%
  relocate(genoSuccess) %>%
  arrange(animalID, genoSuccess)

toRemove_bloodDups <- bloodDups %>%
  group_by(animalID) %>%
  slice_min(genoSuccess) %>%
  ungroup() %>%
  select(sampleID) %>% pull()

# clean!
genos_temp <- genos10x %>%
  select(-all_of(c(toRemove_hairDups, toRemove_bloodDups)))
```

## 4.2 Remove 0% genoSuccess

Subset to INDID and SEXID loci, then remove any samples or loci with 0% genoSuccess

```{r}
# 250 samples, 194 loci
genos_temp2 <- genos_temp %>%
  # just want indid & sex loci
  filter(!str_detect(rownames(.), "SPECIES")) %>%
  # remove samples w/all NA
  select(., where(function(x) !all(is.na(x)))) %>%
  # remove loci w/all NA
  filter(if_any(everything(), ~!is.na(.)))
```

## 4.3 Create clean versions

```{r}
# genos (250 samples, 194 loci)
genos_clean <- genos_temp2

# geno success
genoSuccess_clean <- get_genoSuccess(genos_clean, locusTable, md, "bySample", sampleID_colName = "sampleID", exclude_nonTargetSp = "no") %>%
  merge(., sampleRef[, c("sampleID", "animalID", "sampleType")], by = "sampleID", all.x = T)

# allele reads
readCounts_clean <- readCounts %>%
  select(colnames(genos_clean)) %>%
  filter(rownames(.) %in% rownames(genos_clean))
 
# md
md_clean <- md %>%
  filter(sampleID %in% colnames(genos_clean))

md_hair_clean <- md_clean %>%
  filter(sampleType == "hair")
md_blood_clean <- md_clean %>%
  filter(sampleType == "blood")

hairSamples_clean <- md_hair_clean$sampleID
bloodSamples_clean <- md_blood_clean$sampleID
```

# 5 Confirm twin pairs

## 5.1 Potential twins - capData

First get a list of likely twin pairs; identified as indivs who were captured as juveniles in the same group in the same year

**NOTE** there are two records in which we have THREE juvs in a group in one year, both LWED:
-   Royals 2015 - 161, 162, 163
-   Ghosts 2018 - 222, 223, 224

Capture records don't say anything to indicate whether these are triplets or twin pair + 1 or what; trapping reports indicate they seem to be around the same age

```{r}
twinList <- capData_2009to2019 %>%
  filter(ageClass == "juvenile") %>%
  select(c("species", "captureDate", "groupName", "animalID")) %>%
  group_by(species, captureDate, groupName) %>%
  summarise(twinPairs = toString(sort(unique(animalID)))) %>%
  filter(str_detect(twinPairs, ",")) %>%
  separate_wider_delim(twinPairs, delim = ",", names = c("twin1", "twin2", "twin3"), too_few = "align_start") %>%
  as.data.frame() %>%
  select(-captureDate) %>%
  distinct() %>%
  arrange(groupName) #%>%
  #mutate(twinID = row_number()) %>%
  #relocate(twinID)

twinList %>%
  group_by(species) %>%
  summarise(count = n()) # 19 lwed, 18 simp
```

## 5.2 Twins with necessary samples

```{r}
twinPairs_run5 <- twinList %>%
  # recode triplets based on ch2 findings
  mutate(
    twin1 = case_when(
      !is.na(twin3) & groupName == "Royals" ~ "161",
      !is.na(twin3) & groupName == "Ghosts" ~ "223",
      .default = twin1
    ),
    twin2 = case_when(
      !is.na(twin3) & groupName == "Royals" ~ "162",
      !is.na(twin3) & groupName == "Ghosts" ~ "224",
      .default = twin2
    )
  ) %>%
  select(-twin3) %>%
  arrange(groupName) %>%
  mutate(twinID = row_number()) %>%
  relocate(twinID) %>%
  # add hair sample ID & species (need samples for both)
  mutate(twin1 = as.character(twin1),
         twin2 = as.character(twin2)) %>%
  mutate_if(is.character, str_trim) %>%
  merge(., md_hair_clean[, c("animalID", "sampleID")], by.x = "twin1", by.y = "animalID") %>%
  rename("sampleID_hair_twin1" = "sampleID",
         "animalID_twin1" = "twin1") %>%
  merge(., md_hair_clean[, c("animalID", "sampleID")], by.x = "twin2", by.y = "animalID") %>%
  rename("sampleID_hair_twin2" = "sampleID",
         "animalID_twin2" = "twin2") %>%
  # add blood sample ID
  merge(., md_blood_clean[, c("animalID", "sampleID")], by.x = "animalID_twin1", by.y = "animalID", all.x = T) %>%
  rename("sampleID_blood_twin1" = "sampleID") %>%
  merge(., md_blood_clean[, c("animalID", "sampleID")], by.x = "animalID_twin2", by.y = "animalID", all.x = T) %>%
  rename("sampleID_blood_twin2" = "sampleID") %>%
  # remove twin pair only if both blood samples are missing
  filter_at(vars(sampleID_blood_twin1, sampleID_blood_twin2), any_vars(!is.na(.))) %>%
  select(c("twinID", "species", "animalID_twin1", "animalID_twin2", "sampleID_hair_twin1", "sampleID_hair_twin2", "sampleID_blood_twin1", "sampleID_blood_twin2")) %>%
  arrange(twinID)

# 12 twin pairs total; 4 lwed, 8 simp
twinPairs_run5 %>%
  group_by(species) %>%
  summarise(count = n())

twinPairs_run5_lwed <- twinPairs_run5 %>%
  filter(species == "LWED")

twinPairs_run5_simp <- twinPairs_run5 %>%
  filter(species == "SIMP")

# Create long-format dataframe; gives each individual + their hair/blood sample IDs
twinPairs_twin1 <- twinPairs_run5 %>%
  select(c("twinID", "species", contains("twin1"))) %>%
  dplyr::rename("animalID" = "animalID_twin1",
         "sampleID_hair" = "sampleID_hair_twin1",
         "sampleID_blood" = "sampleID_blood_twin1")

twinPairs_twin2 <- twinPairs_run5 %>%
  select(c("twinID", "species", contains("twin2"))) %>%
  dplyr::rename("animalID" = "animalID_twin2",
         "sampleID_hair" = "sampleID_hair_twin2",
         "sampleID_blood" = "sampleID_blood_twin2")

twinPairs_run5_l <- rbind(twinPairs_twin1, twinPairs_twin2)

# Twin sample ID lists by sample type
twinPairs_hair_sampleList <- data.frame(
  sampleID_hair = c(twinPairs_run5$sampleID_hair_twin1, twinPairs_run5$sampleID_hair_twin2)
) %>%
  pull()

twinPairs_blood_sampleList <- data.frame(
  sampleID_blood = c(twinPairs_run5$sampleID_blood_twin1, twinPairs_run5$sampleID_blood_twin2)
) %>%
  na.omit() %>%
  pull()
```

## 5.3 Maternity checks

### CKMRsim

**I did this already in ch2_popGen.Rmd; transfer it over here if time allows, but for now, LOD cutoff values are as follows:

lambda_star = 6 for both for fs vs. u...apparently I didn't do this for hs vs. u (**need to do that!!**)

```{r}
#remotes::install_github("eriqande/CKMRsim")
library(CKMRsim)
```

```{r}

```

### FRANz

**1) Add blood genos from potential twin pair indivs to FRANz genos**

```{r}
# originals
lwed_genos_franz <- read.csv("./DISSERTATION/ch2_demography/00_data/lwed_genos_franz.csv")
simp_genos_franz <- read.csv("./DISSERTATION/ch2_demography/00_data/simp_genos_franz.csv")

# blood genos from twin pairs
genos_clean_blood_temp <- genos_clean %>%
  select(all_of(twinPairs_blood_sampleList)) %>%
  t() %>%
  as.data.frame()

# get into franz format
genos_clean_blood <- names(genos_clean_blood_temp) %>%
  map_dfc(~ genos_clean_blood_temp %>%
            select(all_of(.x)) %>%
            separate(.x,
                     into = paste0(.x, c("a", "b")),
                     sep = ",")
          ) %>%
  # add prior info
  rownames_to_column("sampleID") %>%
  merge(., sampleRef[, c("sampleID", "sampleID_franz", "animalID", "species", "sex", "birthYear_est", "deathYear_est")], by = "sampleID", all.x = T) %>%
  relocate(c(species, sampleID, sampleID_franz, birthYear_est, deathYear_est, sex)) %>%
  dplyr::rename("birthYear" = "birthYear_est",
                "deathYear" = "deathYear_est") %>%
  
  # replace allele letters with numbers (required by FRANz)
  mutate_at(vars(starts_with(c("INDID", "LWED", "SIMP"))), ~ str_replace(., "A", "4")) %>%
  mutate_at(vars(starts_with(c("INDID", "LWED", "SIMP"))), ~ str_replace(., "T", "7")) %>%
  mutate_at(vars(starts_with(c("INDID", "LWED", "SIMP"))), ~ str_replace(., "C", "3")) %>%
  mutate_at(vars(starts_with(c("INDID", "LWED", "SIMP"))), ~ str_replace(., "G", "6")) %>%
  arrange(sampleID) %>%
  select(-animalID, -sampleID)

# species subsets; select only loci in current genos_franz
genos_clean_blood_lwed <- genos_clean_blood %>%
  filter(species == "LWED") %>%
  select(-species) %>%
  select(all_of(colnames(lwed_genos_franz)))

genos_clean_blood_simp <- genos_clean_blood %>%
  filter(species == "SIMP") %>%
  select(-species) %>%
  select(all_of(colnames(simp_genos_franz)))

# add to franz genos
lwed_genos_franz_plusBlood <- rbind(lwed_genos_franz, genos_clean_blood_lwed)

simp_genos_franz_plusBlood <- rbind(simp_genos_franz, genos_clean_blood_simp)
```

```{r}
# export
write.csv(lwed_genos_franz_plusBlood, "./DISSERTATION/ch3_chimerism/00_data/lwed_genos_franz_plusBlood.csv", row.names = F)
write.csv(simp_genos_franz_plusBlood, "./DISSERTATION/ch3_chimerism/00_data/simp_genos_franz_plusBlood.csv", row.names = F)
```

**2) csv to dat**

```{r}
system2("perl",
        args="./project_scripts/franz/csv_modified.pl --in ./DISSERTATION/ch3_chimerism/00_data/lwed_genos_franz_plusBlood.csv --alleles_per_col 1 --has_header --missing_allele 'NA' --birth_col 1 --death_col 2 --sex_col 3 --data_col 4 > ./DISSERTATION/ch3_chimerism/00_data/lwed_genos_franz_plusBlood.dat")

system2("perl",
        args="./project_scripts/franz/csv_modified.pl --in ./DISSERTATION/ch3_chimerism/00_data/simp_genos_franz_plusBlood.csv --alleles_per_col 1 --has_header --missing_allele 'NA' --birth_col 1 --death_col 2 --sex_col 3 --data_col 4 > ./DISSERTATION/ch3_chimerism/00_data/simp_genos_franz_plusBlood.dat")
```

**3) Run FRANz**

Can't include lwed/simp_poPed_forRun2.dat created in ch2_demography b/c doesn't have same number of indivs

```{bash}
# lwed
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/DISSERTATION/ch3_chimerism/02_results/franz/lwed

FRANz --femrepro 2:30 --malerepro 2:30 --Nf 86 --Nm 96 --fullsibtest --fullsibparental --updatefreqs --poutformat 2 ./../../../00_data/lwed_genos_franz_plusBlood.dat

# simp
cd ./../simp

FRANz --femrepro 2:30 --malerepro 2:30 --Nf 57 --Nm 64 --fullsibtest --fullsibparental --updatefreqs --poutformat 2 ./../../../00_data/simp_genos_franz_plusBlood.dat
```

### Load results & reformat

**1) Define better colnames**

```{r}
colnames_parentage <- c("offspring",
                        "lociTyped_offspring",
                        "parent1",
                        "lociTyped_parent1",
                        "parent2",
                        "lociTyped_parent2",
                        "LOD",
                        "posterior",
                        "commonLociTyped",
                        "mismatches",
                        "n_f",
                        "n_m",
                        "pairLOD_parent1",
                        "pairLOD_parent2",
                        "posterior_parent1",
                        "posterior_parent2",
                        "parentage_mlPedigree")
```

**2) Import results and rename cols**

```{r}
lwed_po_file <- read_csv("./DISSERTATION/ch3_chimerism/02_results/franz/lwed/parentage.csv") %>%
  separate('Posterior Parent 2', c("pp2", "mlPed"), sep = ",") %>%
  `colnames<-`(colnames_parentage)

simp_po_file <- read_csv("./DISSERTATION/ch3_chimerism/02_results/franz/simp/parentage.csv") %>%
  separate('Posterior Parent 2', c("pp2", "mlPed"), sep = ",") %>%
  `colnames<-`(colnames_parentage)
```

**3) Add metadata**

```{r}
# LWED
lwed_po_checks <- lwed_po_file %>%
  # remove NA assignments
  filter(!is.na(parent1)) %>%
  select(offspring, parent1, parent2, LOD, posterior) %>%
  
  pivot_longer(-c(offspring, LOD, posterior),
               names_to = "parentID",
               values_to = "parent") %>%
  select(-parentID) %>%
  na.omit() %>%
  
  # add birthYear and sex
  merge(., lwed_genos_franz_plusBlood[, c("sampleID_franz", "birthYear", "sex")], by.x = "parent", by.y = "sampleID_franz", all.x = T) %>%
  merge(., lwed_genos_franz_plusBlood[, c("sampleID_franz", "birthYear", "sex")], by.x = "offspring", by.y = "sampleID_franz", all.x = T, suffixes = c("_par", "_off")) %>%
  
  # add groups
  mutate(
    animalID_par = gsub(".*_([0-9]+)$", "\\1", parent),
    animalID_off = gsub(".*_([0-9]+)$", "\\1", offspring)
    ) %>%
  
  mutate(
    temp = as.numeric(birthYear_off) - 1
    ) %>%
  merge(., capData_2009to2019[, c("animalID", "captureYear", "groupName")], by.x = c("animalID_par", "temp"), by.y = c("animalID", "captureYear"), all.x = T) %>%
  select(-temp) %>%
  
  merge(., capData_2009to2019[, c("animalID", "captureYear", "groupName")], by.x = c("animalID_off", "birthYear_off"), by.y = c("animalID", "captureYear"), all.x = T, suffixes = c("_par", "_off")) %>%
  
  # same groupName?
  mutate(
    groupMatch = groupName_par == groupName_off
  ) %>%
  
  # add parity for F
  merge(., parityAssigments_firstParous[, c("animalID", "parityYear")], by.x = "animalID_par", by.y = "animalID", all.x = T) %>%
  mutate(
    diff_birth.parity = as.numeric(birthYear_off) - as.numeric(parityYear)
  ) %>%
  
  # check birthYears when available
  mutate(
    diff_birthOff.birthPar = as.numeric(birthYear_off) - as.numeric(birthYear_par)
  ) %>%
  
  relocate(parent, offspring, LOD, posterior, animalID_par, animalID_off, sex_par, sex_off, birthYear_par, birthYear_off, diff_birthOff.birthPar, parityYear, diff_birth.parity, groupName_par, groupName_off, groupMatch) %>%
  arrange(offspring, parent) %>%
  distinct()

# SIMP
simp_po_checks <- simp_po_file %>%
  # remove NA assignments
  filter(!is.na(parent1)) %>%
  select(offspring, parent1, parent2, LOD, posterior) %>%
  
  pivot_longer(-c(offspring, LOD, posterior),
               names_to = "parentID",
               values_to = "parent") %>%
  select(-parentID) %>%
  na.omit() %>%
  
  # add birthYear and sex
  merge(., simp_genos_franz_plusBlood[, c("sampleID_franz", "birthYear", "sex")], by.x = "parent", by.y = "sampleID_franz", all.x = T) %>%
  merge(., simp_genos_franz_plusBlood[, c("sampleID_franz", "birthYear", "sex")], by.x = "offspring", by.y = "sampleID_franz", all.x = T, suffixes = c("_par", "_off")) %>%
  
  # add groups
  mutate(
    animalID_par = gsub(".*_([0-9]+)$", "\\1", parent),
    animalID_off = gsub(".*_([0-9]+)$", "\\1", offspring)
    ) %>%
  
  mutate(
    temp = as.numeric(birthYear_off) - 1
    ) %>%
  merge(., capData_2009to2019[, c("animalID", "captureYear", "groupName")], by.x = c("animalID_par", "temp"), by.y = c("animalID", "captureYear"), all.x = T) %>%
  select(-temp) %>%
  
  merge(., capData_2009to2019[, c("animalID", "captureYear", "groupName")], by.x = c("animalID_off", "birthYear_off"), by.y = c("animalID", "captureYear"), all.x = T, suffixes = c("_par", "_off")) %>%
  
  # same groupName?
  mutate(
    groupMatch = groupName_par == groupName_off
  ) %>%
  
  # add parity for F
  merge(., parityAssigments_firstParous[, c("animalID", "parityYear")], by.x = "animalID_par", by.y = "animalID", all.x = T) %>%
  mutate(
    diff_birth.parity = as.numeric(birthYear_off) - as.numeric(parityYear)
  ) %>%
  
  # check birthYears when available
  mutate(
    diff_birthOff.birthPar = as.numeric(birthYear_off) - as.numeric(birthYear_par)
  ) %>%
  
  relocate(parent, offspring, LOD, posterior, animalID_par, animalID_off, sex_par, sex_off, birthYear_par, birthYear_off, diff_birthOff.birthPar, parityYear, diff_birth.parity, groupName_par, groupName_off, groupMatch) %>%
  arrange(offspring, parent) %>%
  distinct()
```

### PO checks

#### LWED

**Maternity**

```{r}
# a bit of filtering
lwed_maternity <- lwed_po_checks %>%
  filter(animalID_off %in% twinPairs_run5_l$animalID) %>%
  # add sampleType
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "offspring", by.y = "sampleID_franz", all.x = T) %>%
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "parent", by.y = "sampleID_franz", all.x = T, suffixes = c("_off", "_par")) %>%
  relocate(parent, offspring) %>%
  relocate(sampleType_off, .after = animalID_off) %>%
  relocate(sampleType_par, .after = animalID_par) %>%
  arrange(animalID_off, sampleType_off) %>%
  # need to ditch parents w/blood samples
  filter(sampleType_par != "blood") %>%
  # just maternity for now (can mess around w/paternity checks later)
  filter(sex_par == "F") %>%
  # subset to dam/sire with highest posterior for each offspring
  group_by(offspring, sex_par) %>%
  slice_max(posterior) %>%
  ungroup() %>%
  distinct() %>%
  # LOD filter
  filter(LOD >= 2.9)

# check each twin's maternity
lwed_maternity.twinPairs <- twinPairs_run5_lwed %>%
  select(twinID, animalID_twin1, animalID_twin2) %>%
  merge(., lwed_maternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin1", by.y = "animalID_off", all.x = T) %>%
  merge(., lwed_maternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin2", by.y = "animalID_off", all.x = T, suffixes = c(".twin1", ".twin2")) %>%
  relocate(twinID, animalID_twin1, animalID_twin2, sampleType_off.twin1, sampleType_off.twin2, animalID_par.twin1, animalID_par.twin2) %>%
  mutate(parMatch_mat = animalID_par.twin1 == animalID_par.twin2)
```

**Paternity**

```{r}
lwed_paternity <- lwed_po_checks %>%
  filter(animalID_off %in% twinPairs_run5_l$animalID) %>%
  # add sampleType
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "offspring", by.y = "sampleID_franz", all.x = T) %>%
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "parent", by.y = "sampleID_franz", all.x = T, suffixes = c("_off", "_par")) %>%
  relocate(parent, offspring) %>%
  relocate(sampleType_off, .after = animalID_off) %>%
  relocate(sampleType_par, .after = animalID_par) %>%
  arrange(animalID_off, sampleType_off) %>%
  # need to ditch parents w/blood samples
  filter(sampleType_par != "blood") %>%
  # just paternity
  filter(sex_par == "M") %>%
  # subset to dam/sire with highest posterior for each offspring
  group_by(offspring, sex_par) %>%
  slice_max(posterior) %>%
  ungroup() %>%
  distinct() %>%
  # LOD filter
  filter(LOD >= 2.9)

lwed_paternity.twinPairs <- twinPairs_run5_lwed %>%
  select(twinID, animalID_twin1, animalID_twin2) %>%
  merge(., lwed_paternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin1", by.y = "animalID_off", all.x = T) %>%
  merge(., lwed_paternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin2", by.y = "animalID_off", all.x = T, suffixes = c(".twin1", ".twin2")) %>%
  relocate(twinID, animalID_twin1, animalID_twin2, sampleType_off.twin1, sampleType_off.twin2, animalID_par.twin1, animalID_par.twin2) %>%
  mutate(parMatch_pat = animalID_par.twin1 == animalID_par.twin2)
```

#### SIMP

**Maternity**

```{r}
simp_maternity <- simp_po_checks %>%
  filter(animalID_off %in% twinPairs_run5_l$animalID) %>%
  # add sampleType
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "offspring", by.y = "sampleID_franz", all.x = T) %>%
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "parent", by.y = "sampleID_franz", all.x = T, suffixes = c("_off", "_par")) %>%
  relocate(parent, offspring) %>%
  relocate(sampleType_off, .after = animalID_off) %>%
  relocate(sampleType_par, .after = animalID_par) %>%
  arrange(animalID_off, sampleType_off) %>%
  # just maternity for now (can mess around w/paternity checks later)
  filter(sex_par == "F") %>%
  # need to ditch parents w/blood samples
  filter(sampleType_par != "blood") %>%
  # subset to dam/sire with highest posterior for each offspring
  group_by(offspring, sex_par) %>%
  slice_max(posterior) %>%
  ungroup() %>%
  distinct() %>%
  # LOD filter
  filter(LOD >= 2.7)

simp_maternity.twinPairs <- twinPairs_run5_simp %>%
  select(twinID, animalID_twin1, animalID_twin2) %>%
  merge(., simp_maternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin1", by.y = "animalID_off", all.x = T) %>%
  merge(., simp_maternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin2", by.y = "animalID_off", all.x = T, suffixes = c(".twin1", ".twin2")) %>%
  relocate(twinID, animalID_twin1, animalID_twin2, sampleType_off.twin1, sampleType_off.twin2, animalID_par.twin1, animalID_par.twin2) %>%
  mutate(parMatch_mat = animalID_par.twin1 == animalID_par.twin2)
```

**Paternity**

```{r}
simp_paternity <- simp_po_checks %>%
  filter(animalID_off %in% twinPairs_run5_l$animalID) %>%
  # add sampleType
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "offspring", by.y = "sampleID_franz", all.x = T) %>%
  merge(., sampleRef[, c("sampleID_franz", "sampleType")], by.x = "parent", by.y = "sampleID_franz", all.x = T, suffixes = c("_off", "_par")) %>%
  relocate(parent, offspring) %>%
  relocate(sampleType_off, .after = animalID_off) %>%
  relocate(sampleType_par, .after = animalID_par) %>%
  arrange(animalID_off, sampleType_off) %>%
  # just paternity
  filter(sex_par == "M") %>%
  # need to ditch parents w/blood samples
  filter(sampleType_par != "blood") %>%
  # subset to dam/sire with highest posterior for each offspring
  group_by(offspring, sex_par) %>%
  slice_max(posterior) %>%
  ungroup() %>%
  distinct() %>%
  # LOD filter
  filter(LOD >= 2.7)

simp_paternity.twinPairs <- twinPairs_run5_simp %>%
  select(twinID, animalID_twin1, animalID_twin2) %>%
  merge(., simp_paternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin1", by.y = "animalID_off", all.x = T) %>%
  merge(., simp_paternity[, c("animalID_off", "animalID_par", "sampleType_off")], by.x = "animalID_twin2", by.y = "animalID_off", all.x = T, suffixes = c(".twin1", ".twin2")) %>%
  relocate(twinID, animalID_twin1, animalID_twin2, sampleType_off.twin1, sampleType_off.twin2, animalID_par.twin1, animalID_par.twin2) %>%
  mutate(parMatch_pat = animalID_par.twin1 == animalID_par.twin2)
```

### Summarize results

Something odd is up with SIMP animalIDs 33/34...see below for details

```{r}
lwed_blood_matCheck.sum <- lwed_maternity.twinPairs %>%
  select(twinID, animalID_twin1, animalID_twin2, animalID_par.twin1, animalID_par.twin2, parMatch_mat) %>%
  distinct()
lwed_blood_patCheck.sum <- lwed_paternity.twinPairs %>%
  select(twinID, animalID_twin1, animalID_twin2, animalID_par.twin1, animalID_par.twin2, parMatch_pat) %>%
  distinct()

simp_blood_matCheck.sum <- simp_maternity.twinPairs %>%
  select(twinID, animalID_twin1, animalID_twin2, animalID_par.twin1, animalID_par.twin2, parMatch_mat) %>%
  distinct()
simp_blood_patCheck.sum <- simp_paternity.twinPairs %>%
  select(twinID, animalID_twin1, animalID_twin2, animalID_par.twin1, animalID_par.twin2, parMatch_pat) %>%
  distinct()

#############

lwed_blood_matCheck.sum
lwed_blood_patCheck.sum
simp_blood_matCheck.sum
simp_blood_patCheck.sum
```

**What's happening with animalID 33/34?**

There are a couple different twinIDs where one sample type from one individual gets assigned animalID 34 while their other sample type + the other twin get assigned animalID 33.

Based on capData (and these same matChecks), animalID 33 is the mother of animalID 34. Furthermore, parity vs. birthYear checks indicate that 34 wasn't parous until after both these twin pairs were born. This suggests that 33 is the true mother of each of these twin pairs, but something odd seems to be going on with 34...

```{r}
# 2017 = parous; 2016 = nulliparous
parityAssignments %>%
  filter(animalID == 34)
```

Looking at paternity assignments, we see that twinID 19 (34/80) is potentially a multi-paternity litter, where 34 blood sample was assigned the same father as 80, but 34 hair sample was assigned a different father.

Who is animalID 56? Maybe some kind of inbreeding thing happening? --looking at cap records makes things more confusing. From this:
-   twins born in 2012 in Ji
-   33 is the only one in Ji
-   74 is in IC
-   56 is in DangerousPeople in 2011, WhiteEmps 2012 & up

Interestingly, 74 is also the father assigned to 182 blood (no pat for 183), who is in the other twin pair with the 33/34 issue

**maternity/paternity overview for problem samples**
twinID 19 - 34/80 - F/F
34 blood =    33    74
34 hair =     33    56
80 blood =    33    74
80 hair =     --    --

twinID 21 - 167/168 - M/F
167 blood =   33
167 hair =    33
168 blood =   33
168 hair =    34

twinID 22 - 182/183 - F/F
182 blood =   33    --
182 hair =    --    --
183 blood =   34    74
183 hair =    33    --

```{r}
temp <- simp_maternity.twinPairs %>%
  filter(twinID %in% c(21, 22))

simp_maternity

# look at paternity
simp_blood_patCheck.sum %>%
  filter(twinID == 19)
View(simp_paternity.twinPairs %>% filter(twinID == 19))

# who is animalID 56?
capData_2009to2019 %>%
  filter(animalID %in% c(167, 168)) %>%
  #filter(animalID %in% c(34, 80, 33, 56, 74)) %>%
  select(captureDate, captureYear, groupName, animalID, ageClass, sex)
```

### twinConfirmed_maternity

Despite the bit of trickiness with 33/34, based on what I've found I think it's safe to confirm 33 as the mother of both twinIDs 21 (167/168) and 22 (182/183)

```{r}
twinConfirmed_mat.lwed <- lwed_maternity.twinPairs %>%
  filter(parMatch_mat == TRUE) %>%
  select(twinID, animalID_twin1, animalID_twin2, animalID_par.twin1, animalID_par.twin2) %>%
  distinct() %>%
  mutate(animalID_mat = animalID_par.twin1) %>%
  select(-all_of(contains("par")))

twinConfirmed_mat.simp <- simp_maternity.twinPairs %>%
  filter(parMatch_mat == TRUE) %>%
  select(twinID, animalID_twin1, animalID_twin2, animalID_par.twin1, animalID_par.twin2) %>%
  distinct() %>%
  mutate(animalID_mat = animalID_par.twin1) %>%
  select(-all_of(contains("par")))
```

### toRemove_maternity

```{r}
# twinID 7
toRemove_mat.lwed <- lwed_maternity.twinPairs %>%
  filter(parMatch_mat == FALSE)

# none
toRemove_mat.simp <- simp_maternity.twinPairs %>%
  filter(parMatch_mat == FALSE) %>%
  filter(!twinID %in% twinConfirmed_mat.simp$twinID)
```

## 5.4 AGHmatrix checks

copy/pasted from ch2_popGen.Rmd

### Prep genos

Remove dups

```{r}
bloodTest_genos_temp <- read.table("./05_tamRun5/06_run5BloodTest/polygenResults_bloodTest_0x.txt", sep = "\t", na.strings = "0")

locusTable_bloodTest <- read.delim("./05_tamRun5/06_run5BloodTest/bloodTest_LocusTable_singleSNPs.txt")

readCounts_bloodTest <- read.delim("./05_tamRun5/06_run5BloodTest/bloodTest_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

bloodTest_genoSuccess <- get_genoSuccess(bloodTest_genos_temp, locusTable_bloodTest, md, "bySample", sampleID_colName = "sampleID", exclude_nonTargetSp = "no")

# blood
bloodDups <- md %>%
  filter(sampleType == "blood") %>%
  get_dupes(animalID) %>%
  merge(., bloodTest_genoSuccess, by = "sampleID") %>%
  relocate(genoSuccess) %>%
  arrange(animalID, genoSuccess)

toRemove_bloodDups <- bloodDups %>%
  group_by(animalID) %>%
  slice_min(genoSuccess) %>%
  ungroup() %>%
  select(sampleID) %>% pull()

# clean!
bloodTest_genos <- bloodTest_genos_temp %>%
  select(-all_of(toRemove_bloodDups))

bloodTest_readCounts <- readCounts_bloodTest %>%
  select(-all_of(toRemove_bloodDups))
```

### AGHmatrix

Tutorial: https://cran.r-project.org/web/packages/AGHmatrix/vignettes/Tutorial_AGHmatrix.html

```{r}
#install.packages("AGHmatrix")
library(AGHmatrix)

#data("snp.pine")
#str(snp.pine)
```

### Format genos

Formatting genos as autotetraploid; must be coded according to dosage level: 0, 1, 2, 3, or 4. 

```{r}
bloodTest_genos_forGenind <- bloodTest_genos %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID")
# convert to genind to format genos as 0,1,2,3,4
bloodTest_genos_genind <- adegenet::df2genind(
  X = bloodTest_genos_forGenind[,c(2:155)],
  pop = bloodTest_genos_forGenind$pop,
  sep = ",",
  ind.names = bloodTest_genos_forGenind$sampleID,
  NA.char = "NA",
  ploidy = 4,
  type = "codom")
# keep only numeric geno for first allele (can determine allele dosage for both alleles this way)
bloodTest_genos_agh <- bloodTest_genos_genind$tab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("locus") %>%
  mutate(locus = gsub("\\..*", "", locus)) %>%
  mutate(temp = rep(c(1,2), nrow(.)/2)) %>%
  relocate(temp) %>%
  filter(temp == 1) %>%
  select(-temp) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame()

# species subsets + indid only
bloodTest_genos_agh_lwed <- bloodTest_genos_agh %>%
  filter(rownames(.) %in% md[md$species == "LWED", "sampleID"]) %>%
  select(!contains(c("SPECIES", "SEX"))) %>%
  as.matrix()
str(bloodTest_genos_agh_lwed)

bloodTest_genos_agh_simp <- bloodTest_genos_agh %>%
  filter(rownames(.) %in% md[md$species == "SIMP", "sampleID"]) %>%
  select(!contains(c("SPECIES", "SEX"))) %>%
  as.matrix()
```

**Also try w/allele ratios** jk nope I need to think about that more...

```{r}
bloodTest_alleleRatios <- bloodTest_readCounts %>%
  rownames_to_column("locus") %>%
  pivot_longer(-locus,
               names_to = "sampleID",
               values_to = "readCounts") %>%
  mutate(readRatio = get_readRatios(readCounts))
```

### Create Gmatrix

Here I'm using the VanRaden (2008) method for estimating pairwise relatedness using biallelic SNPs, which was extended by Ashraf et al. (2016) for autopolyploids. The VR method considers multiple dosage (i.e., quantifies the presence of an allele based on the TOTAL number of copies across all chromosome sets) vs. pseudo-diploid dosage (simplifies allele copies to either homozygous or heterozygous state).

With the VR method, r_VR (relatedness coefficient) is equal to the off-diagonal elements of the genomic relationship matrix (Gmatrix).

VanRaden method = marker-based additive relationship matrix

From Ashraden et al. (2016):
"Consider a matrix comprised of allele frequency  estimates  from  GBS  data with  SNPs  in  rows  and   samples in columns, X_ij being the allele frequency at SNP i in family j, and M is the matrix centered by SNP mean frequencies,  M = {M_ij} = {X_ij −  ̄X_i. Missing SNP data is substituted by the mean SNP frequency. This amounts to a mean-imputation for missing genotypes. Then the G-matrix is:

G = M'M / K

```{r}
gMat_vanRaden_lwed <- Gmatrix(SNPmatrix = bloodTest_genos_agh_lwed,
                              method = "VanRaden",
                              ploidy = 4,
                              maf = 0.03,
                              thresh.missing = 0.5)
gMat_vanRaden_simp <- Gmatrix(SNPmatrix = bloodTest_genos_agh_simp,
                              method = "VanRaden",
                              ploidy = 4,
                              maf = 0.03,
                              thresh.missing = 0.5)

# want standard error... somehow
# i'm not doing the below correctly...>>worry about this later
stats_vanRaden_lwed <- gMat_vanRaden_lwed %>%
  as.data.frame() %>%
  rownames_to_column("indiv1") %>%
  pivot_longer(-indiv1,
               names_to = "indiv2",
               values_to = "est.r") %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "indiv1", by.y = "sampleID", all.x = T) %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "indiv2", by.y = "sampleID", all.x = T, suffixes = c("1", "2")) #%>%
  
  #mutate(RMSE.PO = sqrt(mean(est.r - 0.5)^2))
  
test <- stats_vanRaden_lwed %>%
  filter(animalID1 %in% c(161, 162, 163, 51, 53, 58, 60)) %>%
  filter(animalID2 %in% c(161, 162, 163, 51, 53, 58, 60)) %>%
  mutate(temp = animalID1 == animalID2) %>%
  filter(temp != T)
```

**Data-frame version**

```{r}
lwed_gmat_df <- gMat_vanRaden_lwed %>%
  as.data.frame() %>%
  rownames_to_column("indiv1") %>%
  pivot_longer(-indiv1,
               names_to = "indiv2",
               values_to = "est.r") %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "indiv1", by.y = "sampleID", all.x = T) %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "indiv2", by.y = "sampleID", all.x = T, suffixes = c("1", "2")) %>%
  relocate(indiv1) %>%
  # ditch self comparisons
  mutate(temp = indiv1 == indiv2) %>%
  filter(temp == FALSE) %>%
  select(-temp)

simp_gmat_df <- gMat_vanRaden_simp %>%
  as.data.frame() %>%
  rownames_to_column("indiv1") %>%
  pivot_longer(-indiv1,
               names_to = "indiv2",
               values_to = "est.r") %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "indiv1", by.y = "sampleID", all.x = T) %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "indiv2", by.y = "sampleID", all.x = T, suffixes = c("1", "2")) %>%
  relocate(indiv1) %>%
  # ditch self comparisons
  mutate(temp = indiv1 == indiv2) %>%
  filter(temp == FALSE) %>%
  select(-temp) 
```

### r-value distribution

To have something to compare to, first I want to get the r-values for blood samples of KNOWN twin pairs.

For confidently confirmed twin pairs: 

```{r}
lwed_rval_twinConfirmed <- lwed_gmat_df %>%
  merge(., twinConfirmed_mat.lwed[, c("animalID_twin1", "animalID_twin2", "twinID")], by.x = c("animalID1", "animalID2"), by.y = c("animalID_twin1", "animalID_twin2")) %>%
  distinct()

simp_rval_twinConfirmed <- simp_gmat_df %>%
  merge(., twinConfirmed_mat.simp[, c("animalID_twin1", "animalID_twin2", "twinID")], by.x = c("animalID1", "animalID2"), by.y = c("animalID_twin1", "animalID_twin2")) %>%
  distinct()


# avg r-val for confidently assigned twin pairs = 0.8241373 +/- SD 0.1018159
lwed_rval_twinConfirmed %>%
  mutate(species = "LWED") %>%
  rbind(simp_rval_twinConfirmed %>% mutate(species = "SIMP")) %>%
  filter(!twinID %in% c(21, 22)) %>%
  summarise(mean.r = mean(est.r),
            sd.r = sd(est.r))
# "potentially" confirmed twin pairs: 167/168 = 0.871, 182/183 = 0.280
lwed_rval_twinConfirmed %>%
  mutate(species = "LWED") %>%
  rbind(simp_rval_twinConfirmed %>% mutate(species = "SIMP")) %>%
  filter(twinID %in% c(21, 22))
```

Now let's check for addt'l confirmations--

```{r}
# LWED - booo no new confirms
lwed_gmat_twinCheck <- lwed_gmat_df %>%
  merge(., twinPairs_run5_l[, c("animalID", "twinID")], by.x = "animalID1", by.y = "animalID") %>%
  merge(., twinPairs_run5_l[, c("animalID", "twinID")], by.x = "animalID2", by.y = "animalID") %>%
  mutate(match = twinID.x == twinID.y) %>%
  filter(match == TRUE) %>%
  dplyr::rename("twinID" = "twinID.x") %>%
  select(-twinID.y) %>%
  mutate(alreadyConf = case_when(
    twinID %in% twinConfirmed_mat.lwed$twinID ~ T,
    .default = F
  ))

# SIMP - 1 new (twinID 5; 207/208)
simp_gmat_twinCheck <- simp_gmat_df %>%
  merge(., twinPairs_run5_l[, c("animalID", "twinID")], by.x = "animalID1", by.y = "animalID") %>%
  merge(., twinPairs_run5_l[, c("animalID", "twinID")], by.x = "animalID2", by.y = "animalID") %>%
  mutate(match = twinID.x == twinID.y) %>%
  filter(match == TRUE) %>%
  dplyr::rename("twinID" = "twinID.x") %>%
  select(-twinID.y) %>%
  mutate(alreadyConf = case_when(
    twinID %in% twinConfirmed_mat.simp$twinID ~ T,
    .default = F
  ))
```

### twinConfirmed

```{r}
twinConfirmed_rVal.lwed <- unique(lwed_gmat_twinCheck$twinID)
twinConfirmed_rVal.simp <- unique(simp_gmat_twinCheck$twinID)
```

## 5.5 Confirmed twin pairs

Retained 7 twin pairs, all but one emp individual has corresponding blood sample

-   7 twin pairs; lwed = 2, simp = 5
-   13 indivs w/blood samples; lwed = 4, simp = 9

```{r}
twinConfirmed_list <- unique(c(twinConfirmed_mat.lwed$twinID,
                       twinConfirmed_mat.simp$twinID,
                       twinConfirmed_rVal.lwed,
                       twinConfirmed_rVal.simp))

twinPairs_confirmed <- twinPairs_run5 %>%
  filter(twinID %in% twinConfirmed_list)

# 7 twin pairs total; 2 lwed, 5 simp
twinPairs_confirmed %>%
  group_by(species) %>%
  summarise(count = n())

twinPairs_confirmed_lwed <- twinPairs_confirmed %>%
  filter(species == "LWED")

twinPairs_confirmed_simp <- twinPairs_confirmed %>%
  filter(species == "SIMP")

# Create long-format dataframe; gives each individual + their hair/blood sample IDs
twinPairs_confirmed_twin1 <- twinPairs_confirmed %>%
  select(c("twinID", "species", contains("twin1"))) %>%
  dplyr::rename("animalID" = "animalID_twin1",
         "sampleID_hair" = "sampleID_hair_twin1",
         "sampleID_blood" = "sampleID_blood_twin1")

twinPairs_confirmed_twin2 <- twinPairs_confirmed %>%
  select(c("twinID", "species", contains("twin2"))) %>%
  dplyr::rename("animalID" = "animalID_twin2",
         "sampleID_hair" = "sampleID_hair_twin2",
         "sampleID_blood" = "sampleID_blood_twin2")

twinPairs_confirmed_l <- rbind(twinPairs_confirmed_twin1, twinPairs_confirmed_twin2) %>%
  arrange(twinID, animalID) %>%
  group_by(twinID) %>%
  mutate(twinID2 = str_c("twin", row_number())) %>%
  ungroup()

# Twin sample ID lists by sample type
twinPairs_confirmed_hair_sampleList <- data.frame(
  sampleID_hair = c(twinPairs_confirmed$sampleID_hair_twin1, twinPairs_confirmed$sampleID_hair_twin2)
) %>%
  pull()

twinPairs_confirmed_blood_sampleList <- data.frame(
  sampleID_blood = c(twinPairs_confirmed$sampleID_blood_twin1, twinPairs_confirmed$sampleID_blood_twin2)
) %>%
  na.omit() %>%
  pull()
```

# 6 GTscore 10x chimerism estimates

## 6.1 Data

### Hair genos & allele reads

Import genotypes for hair samples (using GTscore 10x cutoff) + allele reads for hair samples (unfiltered)

```{r}
hairGenos <- genos_clean %>%
  # filter to hair samples
  select(any_of(hairSamples_clean)) %>%
  t() %>%
  as.data.frame()

hairReads <- readCounts_clean %>%
  select(any_of(hairSamples_clean)) %>%
  t() %>%
  as.data.frame()
```

### Blood allele reads

Import allele reads for blood samples (unfiltered)

```{r}
bloodReads <- readCounts_clean %>%
  select(any_of(bloodSamples_clean)) %>%
  t() %>%
  as.data.frame()
```

## 6.2 Pull twinPair hair & blood data

### Hair genos

```{r}
# Filter hairGenos to twin pairs
hairGenos_allTwins <- hairGenos %>%
  rownames_to_column("sampleID") %>%
  filter(sampleID %in% twinPairs_confirmed_l$sampleID_hair) %>%
  merge(., twinPairs_confirmed_l[, c("sampleID_hair", "twinID", "animalID")], by.x = "sampleID", by.y = "sampleID_hair") %>%
  select(-sampleID) %>%
  relocate(twinID) %>%
  relocate(animalID) %>%
  mutate(
    animalID = as.character(animalID), # only necessary if ID is numeric
    animalID = ifelse(row_number() > 1, paste0(animalID, ".", letters[row_number()]), animalID),
    .by = animalID) %>%
  # Move animalID to rowname
  column_to_rownames("animalID")

# Split into groups by twin pair (twinID)
hairGenos_twinPairs <- split(hairGenos_allTwins, f = hairGenos_allTwins$twinID) %>%
  # Remove twinID column
  lapply(., function(x) x[!(names(x) %in% c("twinID"))]) %>%
  # Filter to loci where both twins have genos
  lapply(., function(x) x[, colSums(is.na(x)) == 0, drop = FALSE]) %>%
  # Filter to loci with NON-MATCHING genos
#  lapply(., function(x) x[, vapply(x, function(x) length(unique(x)) > 1, logical(1L)), drop = FALSE]) %>%
  # Remove empty dataframes
  discard(., ~length(as.matrix(.))==0) %>%
  # Transpose all dataframes
  lapply(., t) %>%
  lapply(., as.data.frame) %>%
  lapply(., rownames_to_column, "locus")

hairGenos_twinPairs$`5`
```

### Blood reads

```{r}
bloodReads_allTwins <- twinPairs_confirmed_l %>%
  select(-species, -sampleID_hair) %>%
  merge(., bloodReads, by.x = "sampleID_blood", by.y = "row.names", all.x = T) %>%
  select(-sampleID_blood) %>%
  relocate(c(animalID, twinID)) %>%
  # filter to twinIDs present in hairGenos_twinPairs
  filter(twinID %in% names(hairGenos_twinPairs)) %>%
  mutate(
    animalID = as.character(animalID), # only necessary if ID is numeric
    animalID = ifelse(row_number() > 1, paste0(animalID, ".", letters[row_number()]), animalID),
    .by = animalID) %>%
  column_to_rownames("animalID")

bloodReads_twinPairs <- split(bloodReads_allTwins, f = bloodReads_allTwins$twinID) %>%
  # Remove twinID column
  lapply(., function(x) x[!(names(x) %in% c("twinID"))]) %>%
  # Transpose all dataframes
  lapply(., t) %>%
  lapply(., as.data.frame) %>%
  lapply(., rownames_to_column, "locus")
```

## 6.3 Merge hair & blood data

```{r}
hairBlood_twinPairs <- map2(hairGenos_twinPairs, bloodReads_twinPairs, ~left_join(.x, .y, by = "locus")) %>%
  lapply(., setNames, c("locus", "hairGeno_twin1", "hairGeno_twin2", "bloodReads_twin1", "bloodReads_twin2")) %>%
  map2(., names(.), ~ mutate(.x, twinID = .y)) %>%
  lapply(., relocate, twinID) %>%
  bind_rows(.) %>%
  # add twin animalIDs
  merge(., twinPairs_confirmed[, c("twinID", "animalID_twin1", "animalID_twin2")], by = "twinID") %>%
  relocate(c(twinID, animalID_twin1, animalID_twin2))
```

## 6.4 Initial chimerism estimates

To calculate initial chimerism estimates, I'm using the following
formulas from Vynck et al. (2023):

|                |                                                    |                           |
|------------------------|------------------------|------------------------|
| Chimerism type | Biallelic constellation (donor/recipient genotype) | Formula for estimating HC |
| type0          | AA/BB                                              | VAF [B/(A+B)]             |
| type1          | BB/AA                                              | 1 - VAF                   |
| type10         | AA/AB                                              | 2 x VAF                   |
| type11         | BB/AB                                              | 2 x (1 - VAF)             |
| type20         | AB/AA                                              | 1 - 2 x VAF               |
| type21         | AB/BB                                              | 2 x VAF - 1               |

### Functions

```{r}
# Identify alleles
allele1 <- function(x) str_sub(x, 1, 1)
allele2 <- function(x) str_sub(x, -1, -1)

# Define chimType
get_chimType <- function(locus, donorGeno, recipGeno, lociRef) {
  
  lociInfo <- lociRef %>%
    select(locus, allele1, allele2) %>%
    mutate(
      a1hom = str_c(allele1, allele1, sep = ","),
      a2hom = str_c(allele2, allele2, sep = ","),
      het = str_c(allele1, allele2, sep = ",")
    )
  
  # Bind and merge with lociInfo
  combined_geno <- data.frame(locus, donorGeno, recipGeno) %>%
    merge(., lociInfo, by = "locus")
  
  results <- combined_geno %>%
    mutate(
      chimType = case_when(
        donorGeno == a1hom & recipGeno == a2hom ~ "type00", # 1a
        donorGeno == a2hom & recipGeno == a1hom ~ "type01", # 1b
        donorGeno == a1hom & recipGeno == het ~ "type10",   # 1c
        donorGeno == a2hom & recipGeno == het ~ "type11",   # 1d
        donorGeno == het & recipGeno == a1hom ~ "type20",   # 2a
        donorGeno == het & recipGeno == a2hom ~ "type21",   # 2b
        donorGeno == a1hom & recipGeno == a1hom ~ "type30", # 3a
        donorGeno == a2hom & recipGeno == a2hom ~ "type30", # 3b
        donorGeno == het & recipGeno == het ~ "type31",     # 3c
        TRUE ~ NA_character_  # Include default case to avoid warnings
      )
    )
  
  return(results$chimType)
}

# Find blood read counts
bloodReads_allele1 <- function(x) as.numeric(sub(",.*", "", x))
bloodReads_allele2 <- function(x) as.numeric(sub(".*,", "", x))

# Calculate variant allele frequency
vaf <- function(recipBloodReads) {
  round(bloodReads_allele2(recipBloodReads)/(bloodReads_allele1(recipBloodReads) + bloodReads_allele2(recipBloodReads)), 2)
}

# Calculate chimProp (hc)
calc_chimProp <- function(recipChimType, recipBloodReads) {
  case_when(
    recipChimType == "type00" ~ vaf(recipBloodReads),
    recipChimType == "type01" ~ 1 - vaf(recipBloodReads),
    recipChimType == "type10" ~ 2 * vaf(recipBloodReads),
    recipChimType == "type11" ~ 2 * (1 - vaf(recipBloodReads)),
    recipChimType == "type20" ~ 1 - (2 * vaf(recipBloodReads)),
    recipChimType == "type21" ~ (2 * vaf(recipBloodReads)) - 1,
    .default = NA
  )
}
```

### Calculations

```{r}
# Identify chimType & calculate VAF
twinChimerism <- hairBlood_twinPairs %>%
  rowwise() %>%
  mutate(
    chimType_twin1 = get_chimType(
      locus = locus,
      donorGeno = hairGeno_twin2,
      recipGeno = hairGeno_twin1,
      lociRef = lociRef
    ),
    chimType_twin2 = get_chimType(
      locus = locus,
      donorGeno = hairGeno_twin1,
      recipGeno = hairGeno_twin2,
      lociRef = lociRef
    )
  ) %>%
  ungroup() %>%  # Ensure ungrouped after rowwise operations
  mutate(
    vaf_twin1 = vaf(recipBloodReads = bloodReads_twin1),
    chimProp_twin1 = calc_chimProp(recipChimType = chimType_twin1, recipBloodReads = bloodReads_twin1),
    vaf_twin2 = vaf(recipBloodReads = bloodReads_twin2),
    chimProp_twin2 = calc_chimProp(recipChimType = chimType_twin2, recipBloodReads = bloodReads_twin2)
    ) %>%
  arrange(as.numeric(twinID))

twinChimerism %>%
  kable(caption = "GTscore 10x: Estimated % twin chimerism for twin pairs by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Reformat for individual twins vs. pairs
twinChimerism_twin1 <- twinChimerism %>%
  select(c("twinID", "locus", contains("twin1")))

twinChimerism_twin2 <- twinChimerism %>%
  select(c("twinID", "locus", contains("twin2")))
colnames(twinChimerism_twin2) <- colnames(twinChimerism_twin1)

twinChimerism_allTwins <- rbind(twinChimerism_twin1, twinChimerism_twin2) %>%
  rename(
    "animalID" = "animalID_twin1",
    "hairGenos" = "hairGeno_twin1",
    "bloodReads" = "bloodReads_twin1",
    "chimType" = "chimType_twin1",
    "vaf" = "vaf_twin1",
    "chimProp" = "chimProp_twin1"
  ) %>%
#  na.omit() %>%
  merge(., md_hair_clean[, c("animalID", "species")], by = "animalID") %>%
  relocate(c(twinID, animalID, species)) %>%
  arrange(as.numeric(twinID), locus)

twinChimerism_allTwins %>%
  kable(caption = "GTscore 10x: Estimated % twin chimerism for individuals by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Subset by species
twinChimerism_lwed <- twinChimerism_allTwins %>%
  filter(species == "LWED")

twinChimerism_simp <- twinChimerism_allTwins %>%
  filter(species == "SIMP")
```

## 6.5 Bias correction

Following Vynck et al. (2023), marker bias is calculated as the average deviation from 50% allele frequency in reference samples with allele frequencies between 40% and 60%. They performed this calculation for "(i) markers in reference samples with a VAF between 40% and 60% and (ii) type-III markers in follow-up samples with a VAF between 40% and 60%, that is where both the donor and the recipient are heterozygous. The average deviation was calculated in the combined set (i) and (ii), as well as separately". 

As such, I'm doing the following:

1.  Calculate VAF for all hair reads
    1.  Subset data to only values between 0.4 and 0.6
    2.  Calculate deviation from 0.5 for each value
    3.  Calculate median deviation from 0.5 for each locus
2.  Calculate VAF for twinPair blood reads where BOTH twins are AB/AB
    1.  Subset data to only values between 0.4 and 0.6
    2.  Calculate deviation from 0.5 for each value
    3.  Calculate median deviation from 0.5 for each locus
3.  Calculate chimProp w/bias corrections
    1.  chimProp_c1 - correct based on hair sample bias
    2.  chimProp_c2 - correct based on blood sample bias
    3.  chimProp_c3 - correct based on hair AND blood sample bias


Note -- I originally used "median" here vs mean, but I feel like when they say "average" they really do mean the "mean"?? Gonna change it to that instead.

### Functions

```{r}
calc_chimProp_c <- function(recipChimType, HC, b) {
  case_when(
        recipChimType == "type00" ~ round((HC-2*HC*b)/(-4*b*HC+2*b+1), 2),
        recipChimType == "type01" ~ round((2*b*HC+HC)/(4*b*HC-2*b+1), 2),
        recipChimType == "type10" ~ round((HC-2*b*HC)/(-2*b*HC+2*b+1), 2),
        recipChimType == "type11" ~ round((2*b*HC+HC)/(2*b*(HC-1)+1), 2),
        recipChimType == "type20" ~ round((2*b+HC)/(2*b*HC+1), 2),
        recipChimType == "type21" ~ round((2*b-HC)/(2*b*HC-1), 2)
  )
}
```

### Calculations

```{r}
# Calculate VAF for all hair reads
vaf_hairReads <- replace(hairReads, TRUE, lapply(hairReads, vaf)) %>%
  mutate(across(everything(),as.numeric)) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("locus")

# Calculate VAF for AB/AB type-III blood reads
vaf_bloodReads <- twinChimerism_allTwins %>%
  select(c("animalID", "locus", "chimType", "vaf")) %>%
  merge(., md_blood_clean[, c("sampleID", "animalID")], by = "animalID") %>%
  select(-animalID) %>%
  arrange(sampleID) %>%
  mutate(
    vaf = case_when(
      chimType != "type31" ~ NA,
      .default = vaf
      )
    ) %>%
  select(-chimType) %>%
  pivot_wider(., names_from = sampleID, values_from = vaf)

vaf_bloodReads_sampleList <- vaf_bloodReads %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  select(sampleID) %>%
  pull()

# Combine VAF from hair & blood reads
vaf_allReads <- merge(vaf_hairReads, vaf_bloodReads, by = "locus", all.x = T) %>%
  column_to_rownames("locus")

# Subset to VAF between 0.4 and 0.6
vaf_allReads[vaf_allReads > 0.6] <- NA
vaf_allReads[vaf_allReads < 0.4] <- NA

# Get median deviation from 0.5 per locus
## all hair samples
## type31 blood samples
## combination of both

vafDev <- (vaf_allReads - 0.5) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    b1 = mean(c_across(all_of(hairSamples_clean)), na.rm = TRUE),
    b2 = mean(c_across(all_of(vaf_bloodReads_sampleList)), na.rm = TRUE),
    b3 = mean(c_across(where(is.numeric)), na.rm = TRUE)
    ) %>%
  select(c("locus", "b1", "b2", "b3"))

# Calculate chimerism corrected for b1, b2, and b3 bias values
twinChimerism_allTwins_corrected <- merge(twinChimerism_allTwins, vafDev, by = "locus", all.x = T) %>%
  mutate(
    chimProp_c1 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b1),
    chimProp_c2 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b2),
    chimProp_c3 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b3)
  ) %>%
  filter(
    chimType != "type30",
    chimType != "type31"
  ) %>%
  filter(
    !is.na(bloodReads)
  ) %>%
  arrange(as.numeric(twinID))

twinChimerism_allTwins_corrected %>%
  kable(caption = "GTscore 10x: Bias-corrected estimated % twin chimerism for individuals by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Subset by species
twinChimerism_lwed_corrected <- twinChimerism_allTwins_corrected %>%
  filter(species == "LWED")

twinChimerism_simp_corrected <- twinChimerism_allTwins_corrected %>%
  filter(species == "SIMP")
```

### Analysis

```{r}
biasMetrics <- twinChimerism_allTwins_corrected %>%
  group_by(locus) %>%
  summarise(
    n = n(),
    meanBias = mean(b3)
    ) %>%
  as.data.frame() %>%
  arrange(meanBias) %>%
  mutate(locus = factor(locus, levels = locus))

# bias estimate % range
biasMetrics %>%
  summarise(
    min = min(meanBias)*100,
    max = max(meanBias)*100
  )

# little plot
biasMetrics %>%
  ggplot(aes(x = meanBias, y = locus)) +
  geom_bar(stat = "identity") +
  scale_x_continuous(limits = c(-0.04, 0.04)) +
  theme_Publication_rv()


# diff b/t pre- and post-correct chimProp estimates?




# bias in hair vs. blood samples
# lollll don't use this
twinChimerism_allTwins_corrected %>%
  ggplot(aes(x = b1, y = b2)) +
  geom_point() +
  geom_smooth(method = "lm", 
              formula = y ~ x) +
  #stat_regline_equation() +
  stat_cor(aes(label = ..rr.label..)) +
  theme_Publication_rv()
```

Diff b/t chimProp pre- and post- bias correction?

```{r}
# use non-parametric test
shapiro.test(twinChimerism_allTwins_corrected$chimProp) # p-value = 0.01191
shapiro.test(twinChimerism_allTwins_corrected$chimProp_c3) # p-value = 0.006705

# no significant diff
wilcox.test(twinChimerism_allTwins_corrected$chimProp, twinChimerism_allTwins_corrected$chimProp_c3, paired = T, alternative = "two.sided") # p-value = 0.6561
```

## 6.6 Summaries

```{r}
twinChimerism_summary <- twinChimerism_allTwins_corrected %>%
  group_by(species, twinID, animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = mean(chimProp, na.rm = T),
    mean_chimProp_c1 = mean(chimProp_c1, na.rm = T),
    mean_chimProp_c2 = mean(chimProp_c2, na.rm = T),
    mean_chimProp_c3 = mean(chimProp_c3, na.rm = T)
  ) %>%
  as.data.frame()

twinChimerism_lwed_summary <- twinChimerism_lwed_corrected %>%
  group_by(animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = mean(chimProp, na.rm = T),
    mean_chimProp_c1 = mean(chimProp_c1, na.rm = T),
    mean_chimProp_c2 = mean(chimProp_c2, na.rm = T),
    mean_chimProp_c3 = mean(chimProp_c3, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(species = "LWED") %>%
  relocate(species)

twinChimerism_lwed_summary %>%
  kable(caption = "GTscore 10x: LWED avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

twinChimerism_simp_summary <- twinChimerism_simp_corrected %>%
  group_by(animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = mean(chimProp, na.rm = T),
    mean_chimProp_c1 = mean(chimProp_c1, na.rm = T),
    mean_chimProp_c2 = mean(chimProp_c2, na.rm = T),
    mean_chimProp_c3 = mean(chimProp_c3, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(species = "SIMP") %>%
  relocate(species)

twinChimerism_simp_summary %>%
  kable(caption = "GTscore 10x: SIMP avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")
```

# 7 Analysis

## Average number of loci per indiv? Overall avg chimerism?

```{r}
# avg loci per indiv
rbind(twinChimerism_lwed_summary,
      twinChimerism_simp_summary) %>%
  group_by(species) %>%
  summarise(
    max_n.loci = max(n_loci),
    min_n.loci = min(n_loci),
    mean_n.loci = mean(n_loci),
    sd_n.loci = sd(n_loci))

# avg chimerism % per indiv - both species
twinChimerism_allTwins_corrected %>%
  group_by(animalID) %>%
  summarise(
    mean_chimProp_c3 = mean(chimProp_c3),
    sd_chimProp_c3 = sd(chimProp_c3)
  ) %>%
  as.data.frame() %>%
  summarise(
    mean = mean(mean_chimProp_c3)
  )

mean_chimProp_overall <- mean(twinChimerism_allTwins_corrected$chimProp_c3, na.rm = T) # 0.4765462

# avg chimerism % per indiv - both species
twinChimerism_allTwins_corrected %>%
  group_by(species, animalID) %>%
  summarise(
    mean_chimProp_c3 = mean(chimProp_c3),
    sd_chimProp_c3 = sd(chimProp_c3)
  ) %>%
  as.data.frame() %>%
  group_by(species) %>%
  summarise(
    min = min(mean_chimProp_c3),
    max = max(mean_chimProp_c3),
    mean = mean(mean_chimProp_c3),
    sd = sd(mean_chimProp_c3)
  )

# now w/twinIDs 5, 19, and 22 ditched
twinChimerism_allTwins_corrected %>%
  filter(!twinID %in% c(5, 22, 19)) %>%
  group_by(species, animalID) %>%
  summarise(
    mean_chimProp_c3 = mean(chimProp_c3),
    sd_chimProp_c3 = sd(chimProp_c3)
  ) %>%
  as.data.frame() %>%
  group_by(species) %>%
  summarise(
    min = min(mean_chimProp_c3),
    max = max(mean_chimProp_c3),
    mean = mean(mean_chimProp_c3),
    sd = sd(mean_chimProp_c3)
  )

# min/max for mean chimerism per indiv
twinChimerism_allTwins_corrected %>%
  group_by(animalID) %>%
  summarise(
    mean_chimProp_c3 = mean(chimProp_c3),
    sd_chimProp_c3 = sd(chimProp_c3)
  ) %>%
  as.data.frame() %>%
  summarise(min = min(mean_chimProp_c3)*100,
            max = max(mean_chimProp_c3)*100)

```

## chimDiffs

### b/t twins

Set up comparison dataframe

```{r}
# set up comparison dataframe
chimProp_twinCompare <- twinChimerism_allTwins_corrected %>%
  select(twinID, animalID, locus, chimProp_c3) %>%
  merge(., twinPairs_confirmed_l[, c("animalID", "twinID2")], by = "animalID") %>%
  pivot_wider(id_cols = c(twinID, locus),
              names_from = twinID2,
              values_from = chimProp_c3) %>%
  na.omit() %>%
  merge(twinPairs_confirmed[, c("twinID", "animalID_twin1", "animalID_twin2")], by = "twinID")
```

Since I'm interested in the diffs b/t twin1 and twin2 values, I need to check if the differences themselves are normally distributed-- they're not, so I should use the non-parametric Wilcoxen signed-rank test

```{r}
# use univariate shapiro-wilks test - shows that diffs are not normally dist
chimProp_twinCompare %>%
  mutate(diff = twin1 - twin2) %>%
  rstatix::shapiro_test(diff) # p = 0.00289 >> not normally distributed

# wilcox test
chimProp_twinCompare %>%
  pivot_longer(cols = c(twin1, twin2), names_to = "group", values_to = "value") %>%
  group_by(twinID) %>%
  rstatix::pairwise_wilcox_test(
    value ~ group,
    paired = T,
    p.adjust.method = "fdr"
    ) #%>%
  #ungroup() %>%
 # mutate(p.adjusted_globally = p.adjust(p, method = "fdr"))

head(chimProp_twinCompare)
```

### by sex?

```{r}
chimProp_twinCompare_l <- twinChimerism_allTwins_corrected %>%
  merge(., sampleRef[, c("animalID", "sex")], by = "animalID") %>%
  distinct() %>%
  select(species, sex, twinID, animalID, locus, chimProp_c3)

table(chimProp_twinCompare_l$sex)

chimProp_twinCompare_l %>%
  group_by(species) %>%
  rstatix::wilcox_test(
    chimProp_c3 ~ sex,
    paired = F,
    p.adjust.method = "fdr"
    )
```


### Does % chimerism in twins add to 1?

```{r}
temp1 <- chimProp_twinCompare %>%
  select(twinID, twin1, animalID_twin1) %>%
  group_by(twinID, animalID_twin1) %>%
  summarise(
    mean_twin1 = mean(twin1)
  )
temp2 <- chimProp_twinCompare %>%
  select(twinID, twin2, animalID_twin2) %>%
  group_by(twinID, animalID_twin2) %>%
  summarise(
    mean_twin2 = mean(twin2)
  )

temp3 <- merge(temp1, temp2, by = "twinID") %>%
  # ditch poor performing pairs
  filter(!twinID %in% c(5, 19, 22)) %>%
  mutate(sum = mean_twin1 + mean_twin2,
         diff = mean_twin1 - mean_twin2) 

chimProp_twinCompare  %>%
  # ditch poor performing pairs
  filter(!twinID %in% c(5, 19, 22)) %>%
  mutate(sum = twin1 + twin2) %>%
  group_by(twinID) %>%
  rstatix::wilcox_test(
    sum ~ 1,
    paired = F,
    p.adjust.method = "holm"
    )

temp <- chimProp_twinCompare  %>%
  # ditch poor performing pairs
  #filter(!twinID %in% c(5, 19, 22)) %>%
  mutate(sum = twin1 + twin2,
         diff = 1 - sum)

chimProp_twinCompare  %>%
  # ditch poor performing pairs
  filter(!twinID %in% c(5, 19, 22)) %>%
  mutate(sum = twin1 + twin2,
         diff = 1 - sum) %>%
  ggplot(aes(x = twinID, y = diff)) +
  geom_boxplot() +
  geom_jitter()
```

## Time diffs in sampling?

chimProp can change over time; checking here to see if individuals were at least sampled at the same time -- some yes, some no

```{r}
twinComp_samplingDates <- twinPairs_confirmed %>%
  select(twinID, animalID_twin1, animalID_twin2) %>%
  merge(., md_blood_clean[, c("animalID", "captureDate")], by.x = "animalID_twin1", by.y = "animalID") %>%
  merge(., md_blood_clean[, c("animalID", "captureDate")], by.x = "animalID_twin2", by.y = "animalID", suffixes = c("_1", "_2")) %>%
  distinct() %>%
  relocate(twinID, animalID_twin1, animalID_twin2) %>%
  mutate(dateDiff = as.Date(captureDate_1) - as.Date(captureDate_2)) %>%
  # check age at sampling for each twin
  merge(., sampleRef[, c("animalID", "birthYear_est")], by.x = "animalID_twin1", by.y = "animalID") %>%
  merge(., sampleRef[, c("animalID", "birthYear_est")], by.x = "animalID_twin2", by.y = "animalID") %>%
  distinct() %>%
  relocate(twinID, animalID_twin1) %>%
  arrange(twinID)

sampleRef %>%
  filter(animalID %in% c(56, 33, 34)) %>%
  distinct()
```


## Type-1 vs. type-2 estimates

Chim estimates diff b/t type-1 and type-2 markers?

```{r}
chimProp_type1.2_l <- twinChimerism_allTwins_corrected %>%
  # ditch twins w/poor genoSuccess
  filter(!twinID %in% c(5, 22, 19)) %>%
  # 00, 01, 10, 11 = type 1, 20, 21 = type2
  mutate(markerType = case_when(
    chimType %in% c('type00', 'type01', 'type10', 'type11') ~ "type1",
    chimType %in% c('type20', 'type21') ~ "type2",
    .default = NA
  )) %>%
  group_by(animalID, markerType) %>%
  summarise(
    chimProp = mean(chimProp_c3)
  )

rstatix::pairwise_wilcox_test(chimProp_type1.2_l, chimProp ~ markerType, pairwise = T)

chimProp_type1.2 <- chimProp_type1.2_l %>%
  pivot_wider(id_cols = animalID,
              names_from = markerType,
              values_from = chimProp)

wilcox.test(chimProp_type1.2$type1, chimProp_type1.2$type2, paired = T, alternative = "two.sided")
```


# X FM chimerism

find female blood samples

```{r}
bloodReads_fmChim <- bloodReads %>%
  select(contains("SEXID")) %>%
  rownames_to_column("sampleID") %>%
  merge(., sampleRef[, c("sampleID", "sampleType", "species", "sex", "animalID")], by = "sampleID") %>%
  filter(sampleType == "blood") %>%
  filter(sex == "F")
  
```


Ok so twinID 37 (animalIDs 136, 137) have genos that, based on sexKey, likely result from allelic dropout since the "TT" geno supposedly shouldn't be possible here--
SEXID_200 136 = TT, 137 = CT (sexKey: CT, CC)
SEXID_222 136 = GT, 137 = TT (sexKey: GT, GG)

similar issue with twinID 5

Filter to twin pair 21 & 27 = mixed sex

```{r}
sexChecks_blood <- assignSex(genos10x, md, "sampleID", exclude_nonTargetSp = "yes") %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by = "sampleID")
View(sexChecks_blood %>% filter(animalID %in% c(136, 137)))
View(genos10x %>% select(tamRun5.088, tamRun5.173))

temp <- genos_clean %>%
  select(tamRun5.351, tamRun5.367)
temp2 <- genos10x %>%
  select(tamRun5.351, tamRun5.367)

twinChimerism_simp_summary

test <- twinChimerism_allTwins_corrected %>%
  filter(str_detect(locus, "SEXID"))

test %>%
  filter(twinID %in% c(21, 27)) %>%
  group_by(twinID, animalID) %>%
  summarise(
    mean = mean(chimProp_c3)
  )
```


# 7 Tables & figures

## Table 3.1

```{r}
data_table3.1 <- read.csv("./DISSERTATION/ch3_chimerism/00_data/googleSheets_table3.1.csv") %>%
  select(-c(n_females, n_males, Results, Fragment.analysis, Other, studyYear, X, X.1), -starts_with("Presence")) %>%
  filter(Species != "unk") %>%
  mutate(XX.XY.based = case_match(
    XX.XY.based,
    "X" ~ "yes",
    NA ~ "no"
  ))

length(unique(data_table3.1$Citation)) # 15 unique papers

table3.1 <- data_table3.1 %>%
  gt() %>%
  cols_label(
    Species = md("**Species**"),
    Sample.size = md("**Sample size**"),
    Methods = md("**Method**"),
    XX.XY.based = md("**Mixed-sex twins only?**"),
    Citation = md("**Reference**")
  ) %>%
  tab_style(
    style = list(
      cell_text(style = "italic")
    ),
    locations = cells_body(
      columns = Species
    )
  ) %>%
  cols_align(
    align = "left",
    columns = everything()
  ) %>%
  tab_footnote(
    footnote = "In cases where multiple sample types were used in a study, we report only the sample size for blood samples.",
    locations = cells_column_labels(columns = Sample.size)
  ) %>%
  tab_footnote(
    footnote = "FISH: fluorescence in situ hybridization; qPCR: quantitative PCR; qfPCR: quantitative fluorescence PCR; STR: short tandem repeats; VNTR: variable number tandem repeats; WGS: whole genome sequencing",
    locations = cells_column_labels(columns = Methods)
  ) %>%
  opt_footnote_marks(marks= "standard")
table3.1

table3.1 %>%
  gtsave("./DISSERTATION/ch3_chimerism/02_results/00_tablesFigures/table3.1_calliChimerism.png")
```

## Table 3.2

```{r}
data_table3.2 <- read.csv("./DISSERTATION/ch3_chimerism/00_data/googleSheets_table3.2.csv") %>%
  filter(methodType == "Molecular") %>%
  select(Method, Markers, DNA.quantity, Sensitivity, Precision, mySource) %>%
  filter(Method != "RFLP") %>%
  select(-Precision)

table3.2 <- data_table3.2 %>%
  gt() %>%
  cols_label(
    Method = md("**Method**"),
    Markers = md("**Markers**"),
    DNA.quantity = md("**DNA quantity**"),
    Sensitivity = md("**Sensitivity**"),
    #Precision = md("**Precision**"),
    mySource = md("**Reference**")
  ) %>%
  cols_align(
    align = "left",
    columns = everything()
  ) %>%
  tab_footnote(
    footnote = "dPCR: digital PCR; ddPCR: digital droplet PCR; qPCR: quantitative PCR; qfPCR: quantitative fluorescence PCR; STR: short tandem repeats; VNTR: variable number tandem repeats; WGS: whole genome sequencing",
    locations = cells_column_labels(columns = Method)
  ) %>%
  tab_footnote(
    footnote = "1. Blouin et al. 2021, 2. Picard et al. 2023, 3. Tozzo et al. 2021",
    locations = cells_column_labels(columns = mySource)
  ) %>%
  opt_footnote_marks(marks= "standard")
table3.2

table3.2 %>%
  gtsave("./DISSERTATION/ch3_chimerism/02_results/00_tablesFigures/table3.2_chimerismMethods.png")
```


## Table 3.5

```{r}
chimSums <- twinPairs_confirmed %>%
  select(twinID, animalID_twin1, animalID_twin2) %>%
  merge(., twinChimerism_summary[, c("animalID", "mean_chimProp_c3")], by.x = "animalID_twin1", by.y = "animalID") %>%
  merge(., twinChimerism_summary[, c("animalID", "mean_chimProp_c3")], by.x = "animalID_twin2", by.y = "animalID", suffixes = c("_1", "_2")) %>%
  mutate(chimSum = mean_chimProp_c3_1 + mean_chimProp_c3_2)

# data for table
data_table3.5_marker.typeCounts <- twinChimerism_allTwins_corrected %>%
  # 00, 01, 10, 11 = type 1, 20, 21 = type2
  mutate(broadType = case_when(
    chimType %in% c('type00', 'type01', 'type10', 'type11') ~ "type1",
    chimType %in% c('type20', 'type21') ~ "type2",
    .default = NA
  )) %>%
  group_by(animalID, broadType) %>%
  summarise(
    n_loci = n()
  ) %>%
  as.data.frame() %>%
  select(animalID, broadType, n_loci) %>%
  pivot_wider(names_from = broadType,
              values_from = n_loci) %>%
  rowwise() %>%
  mutate(total = sum(type1, type2, na.rm = T)) %>%
  merge(., twinPairs_confirmed_l[, c("animalID", "twinID", "species")], by = "animalID", all = T) %>%
  merge(., sampleRef[, c("animalID", "sex")], by = "animalID") %>%
  distinct() %>%
  relocate(species, twinID) %>%
  arrange(species, as.numeric(twinID)) %>%
  merge(., twinChimerism_summary[, c("animalID", "mean_chimProp_c3")], by = "animalID", all.x = T) %>%
  dplyr::rename("chimProp" = "mean_chimProp_c3") %>%
  # add chimSums
  merge(., chimSums[, c("twinID", "chimSum")], by = "twinID", all.x = T) %>%
  # make new animalID (anonymous)
  merge(., twinPairs_confirmed_l[, c("animalID", "twinID2")], by = "animalID", all.x = T) %>%
  mutate(twinID2 = gsub("twin", "", twinID2)) %>%
  mutate(temp = str_c(twinID, ".", twinID2, "_", sex),
         animalID = temp) %>%
  select(-sex, -twinID2, -temp) %>%
  relocate(species, twinID, animalID)
data_table3.5_marker.typeCounts

# gt table
table3.5_marker.typeCounts <- data_table3.5_marker.typeCounts %>%
  mutate(species = case_when(
    species == "LWED" ~ "Saddleback tamarins",
    .default = "Emperor tamarins"
  )) %>%
  mutate(species = factor(species, levels = c("Saddleback tamarins", "Emperor tamarins"))) %>%  ## keep the original order
  arrange(species, twinID) %>%   
  mutate(twinID = ifelse(row_number() == 1, as.character(twinID), ""),
         .by = c(species, twinID)) %>% 
  mutate(species = ifelse(row_number() == 1, as.character(species), ""),
         .by = species) %>%
  #mutate(chimSum = ifelse(row_number() == 1, as.character(chimSum), ""),
  #       .by = chimSum) %>%
  select(species, twinID, everything()) %>% 
  gt() %>%
  sub_missing(missing_text = '-') %>%
  #tab_spanner(label = "Marker type",
  #            columns = c("type1", "type2")) %>%
  cols_label(
    species = md("**Species**"),
    twinID = md("**TwinID**"),
    animalID = md("**IndivID**"),
    type1 = md("**Type-1 markers**"),
    type2 = md("**Type-2 markers**"),
    total = md("**Total markers**"),
    chimProp = md("**Chimerism**"),
    chimSum = md("**Sum twin chimerism**"),
  ) %>%
  fmt_percent(
    columns = c("chimProp", "chimSum"),
    rows = everything(),
    decimals = 2,
    scale_values = T
  ) %>%
  cols_align(., align = "center", columns = 2:8) %>%
  tab_footnote(
    footnote = "Individual 27.2_F had no available blood sample for testing.",
    locations = cells_body(columns = animalID, rows = 14)
  ) %>%
  opt_footnote_marks(marks= "standard")
table3.5_marker.typeCounts

table3.5_marker.typeCounts %>%
  gtsave("./DISSERTATION/ch3_chimerism/02_results/00_tablesFigures/table3.5_marker.typeCounts.png")
```

## Plots

### n_loci vs. chimProp

```{r}
library(ggpubr)
twinID_levels <- unique(twinChimerism_allTwins_corrected$twinID)

fig_nLoci.chimProp <- twinChimerism_allTwins_corrected %>%
  mutate(species = case_match(
    species,
    "LWED" ~ "Saddlebacks",
    "SIMP" ~ "Emperors"
  ),
  species = factor(species, levels = c("Saddlebacks", "Emperors"))
  ) %>%
  group_by(species, twinID, animalID) %>%
  mutate(twinID = factor(twinID, levels = twinID_levels)) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = mean(chimProp_c3) * 100,
    sd_chimProp = sd(chimProp_c3) * 100
  ) %>%
  ggplot(aes(x = jitter(n_loci, 5), y = mean_chimProp, color = twinID, shape = twinID)) +
  geom_pointrange(aes(
    ymax = mean_chimProp + sd_chimProp,
    ymin = mean_chimProp - sd_chimProp
    ),
    #position = position_jitter(width = 0.3),
    size = 1,
    stroke = 1.5) +
  scale_shape_manual(values = c(1, 3, 5, 4, 0, 8, 2)) +
  scale_color_manual(values = my_colors) +
  scale_x_continuous(limits = c(0, 50)) +
  scale_y_continuous(limits = c(-10, 110), breaks = c(0, 25, 50, 75, 100)) +
  theme_Publication_rv() +
  labs(x = "Number of informative markers",
       y = "Mean % chimerism",
       color = "TwinID",
       shape = "TwinID") +
  facet_grid(rows = vars(species)) +
  theme(legend.position = "right")
fig_nLoci.chimProp
```

### genoSuccess vs. diff from popMean chimerism

```{r}
# mean for both saddlebacks and emperor tamarins
mean_chimProp_overall

figData_genoPerformance.diff50p <- twinChimerism_allTwins_corrected %>%
  group_by(species, twinID, animalID) %>%
  mutate(twinID = factor(twinID, levels = twinID_levels)) %>%
  summarise(
    n_loci = n(),
    mean_chimProp_c3 = round(mean(chimProp_c3, na.rm = T), 2),
    sd_chimProp_c3 = round(sd(chimProp_c3), 2)
  ) %>%
  as.data.frame() %>%
  mutate(diff = abs(mean_chimProp_overall - mean_chimProp_c3)) %>%
  merge(., twinPairs_confirmed[, c("animalID_twin1", "animalID_twin2")], by.x = "animalID", by.y = "animalID_twin1", all.x = T) %>%
  mutate(diff = abs(mean_chimProp_overall - mean_chimProp_c3)) %>%
  merge(., twinPairs_confirmed[, c("animalID_twin1", "animalID_twin2")], by.x = "animalID", by.y = "animalID_twin2", all.x = T) %>%
  mutate(animalID_twin = coalesce(animalID_twin1, animalID_twin2),
         .keep = "unused") %>%
  merge(., genoSuccess_clean[, c("animalID", "totalGenos", "sampleType")], by = "animalID") %>%
  filter(sampleType == "hair") %>%
  merge(., genoSuccess_clean[, c("animalID", "totalGenos", "sampleType")], by.x = c("animalID_twin", "sampleType"), by.y = c("animalID", "sampleType"), suffixes = c(".1", ".2")) %>%
  mutate(genoSuccess.1 = cut(totalGenos.1, breaks = c(0, 10, 20, 30, 40, 50, 100, 194)),
         genoSuccess.2 = cut(totalGenos.2, breaks = c(0, 10, 20, 30, 40, 50, 100, 194))) %>%
  mutate(gsCheck = case_when(
    totalGenos.1 > totalGenos.2 ~ genoSuccess.2,
    totalGenos.1 < totalGenos.2 ~ genoSuccess.1,
    .default = genoSuccess.1
  ))

# or... using actual geno counts
figData_genoPerformance.diff50p %>%
  mutate(minGenos = case_when(
    totalGenos.1 > totalGenos.2 ~ totalGenos.2,
    totalGenos.1 < totalGenos.2 ~ totalGenos.1,
    .default = totalGenos.1
  )) %>%
  ggplot(aes(x = as.numeric(minGenos), y = diff, color = twinID, shape = twinID)) +
  geom_point(size = 3, stroke = 1.5) +
  scale_shape_manual(values = c(1, 3, 5, 4, 0, 8, 2)) +
  scale_color_manual(values = my_colors) +
  theme_Publication() +
  theme(legend.position = "right") +
  labs(x = "Number of loci genotyped",
       y = "Difference from mean chimerism proportion",
       color = "TwinID",
       shape = "TwinID") +
  scale_x_continuous(limits = c(0, 150), breaks = seq(0, 150, by = 25))




# or again..
figData_genoPerformance.diff50p %>%
  select(twinID, mean_chimProp_c3, totalGenos.1, totalGenos.2) %>%
  pivot_longer(-c(twinID, mean_chimProp_c3),
               names_to = "whichTwin",
               values_to = "totalGenos") %>%
  ggplot(aes(x = totalGenos, y = mean_chimProp_c3, color = twinID)) +
  geom_point()



figData_genoPerformance.diff50p %>%
  ggplot(aes(x = ))


  mutate(
    genoSuccess.1 = round(genoSuccess.1, 2),
    genoSuccess.2 = round(genoSuccess.2, 2),
    gsCheck = case_when(
      genoSuccess.1 < 0.1 ~ "gs<10%",
      genoSuccess.2 < 0.1 ~ "gs<10%",
      genoSuccess.1 >= 0.1 & genoSuccess.1 < 0.15 ~ "10<gs<15%",
      genoSuccess.2 >= 0.1 & genoSuccess.2 < 0.15 ~ "10<gs<15%",
      genoSuccess.1 >= 0.15 & genoSuccess.1 < 0.2 ~ "15<gs<20%",
      genoSuccess.2 >= 0.15 & genoSuccess.2 < 0.2 ~ "15<gs<20%",
      genoSuccess.1 >= 0.2 & genoSuccess.1 < 0.3 ~ "20<gs<30%",
      genoSuccess.2 >= 0.2 & genoSuccess.2 < 0.3 ~ "20<gs<30%",
      genoSuccess.1 >= 0.3 & genoSuccess.1 < 0.4 ~ "30<gs<40%",
      genoSuccess.2 >= 0.3 & genoSuccess.2 < 0.4 ~ "30<gs<40%",
      genoSuccess.1 >= 0.4 & genoSuccess.1 < 0.5 ~ "40<gs<50%",
      genoSuccess.2 >= 0.4 & genoSuccess.2 < 0.5 ~ "40<gs<50%",
      .default = "NA"
      )
    )
  
  
  
  
  mutate(
    genoSuccess.1 = round(genoSuccess.1, 2),
    genoSuccess.2 = round(genoSuccess.2, 2),
    gsCheck = case_when(
      genoSuccess.1 < 0.1 ~ "gs<10%",
      genoSuccess.2 < 0.1 ~ "gs<10%",
      genoSuccess.1 >= 0.1 & genoSuccess.1 < 0.15 ~ "10<gs<15%",
      genoSuccess.2 >= 0.1 & genoSuccess.2 < 0.15 ~ "10<gs<15%",
      genoSuccess.1 >= 0.15 & genoSuccess.1 < 0.2 ~ "15<gs<20%",
      genoSuccess.2 >= 0.15 & genoSuccess.2 < 0.2 ~ "15<gs<20%",
      genoSuccess.1 >= 0.2 & genoSuccess.1 < 0.3 ~ "20<gs<30%",
      genoSuccess.2 >= 0.2 & genoSuccess.2 < 0.3 ~ "20<gs<30%",
      genoSuccess.1 >= 0.3 & genoSuccess.1 < 0.4 ~ "30<gs<40%",
      genoSuccess.2 >= 0.3 & genoSuccess.2 < 0.4 ~ "30<gs<40%",
      genoSuccess.1 >= 0.4 & genoSuccess.1 < 0.5 ~ "40<gs<50%",
      genoSuccess.2 >= 0.4 & genoSuccess.2 < 0.5 ~ "40<gs<50%",
      .default = "NA"
      )
    )

figPlot_genoPerformance.diff50p <- figData_genoPerformance.diff50p %>%
  mutate(twinID = factor(twinID, levels = twinID_levels),
         gsCheck = factor(gsCheck, levels = c("gs<10%", "10<gs<15%", "15<gs<20%", "20<gs<30%", "30<gs<40%", "40<gs<50%", "NA"))) %>%
  ggplot(aes(x = gsCheck, y = diff, color = twinID, shape = twinID)) +
  geom_point(size = 3, stroke = 1.5) +
  scale_shape_manual(values = c(1, 3, 5, 4, 0, 8, 2)) +
  scale_color_manual(values = my_colors) +
  theme_bw() +
  labs(x = "Genotype success cutoff",
       y = "0.50 - mean chimerism proportion",
       color = "TwinID",
       shape = "TwinID")
figPlot_genoPerformance.diff50p
```

### Metrics minus poor genoSucess

Descriptive stats after ditching twin pairs w/poor genoSuccess:

```{r}
twinChimerism_allTwins_corrected %>%
  filter(!twinID %in% c(5, 22, 19)) %>%
  group_by(species, animalID) %>%
  summarise(
    temp = mean(chimProp_c3)
  ) %>%
  as.data.frame() %>%
  group_by(species) %>%
  summarise(
    n_indiv = n(),
    mean_chimProp_c3 = mean(temp, na.rm = T),
    sd_chimProp_c3 = sd(temp),
    min_chimProp_c3 = min(temp),
    max_chimProp_c3 = max(temp)
  )


  as.data.frame() %>%
  summarise(mean = mean(mean_chimProp_c3),
            sd = sd(mean_chimProp_c3))
```




**1) Bias correction is important to implement**

```{r}
library(cowplot)
figData_mean.typeI.II.uc_vs_mean.typeI.c <- twinChimerism_allTwins_corrected %>%
  mutate(chimProp_c3 = case_when(
    chimType %in% c("type20", "type21") ~ NA,
    .default = chimProp_c3
  )) %>%
  group_by(animalID) %>%
  summarise(
    mean.typeI.II.uc = mean(chimProp, na.rm = T),
    mean.typeI.c = mean(chimProp_c3, na.rm = T)
  ) %>%
  mutate(
    diff = mean.typeI.II.uc - mean.typeI.c
  )

figPlot_mean.typeI.II.uc_vs_mean.typeI.c <- figData_mean.typeI.II.uc_vs_mean.typeI.c %>%
  ggplot(aes(x = mean.typeI.c, y = diff)) +
  geom_point() +
  ylim(-0.5, 0.5)
figPlot_mean.typeI.II.uc_vs_mean.typeI.c

# now uncorrected
figData_mean.typeI.II.c_vs_mean.typeI.c <- twinChimerism_allTwins_corrected %>%
  mutate(typeI_chimProp_c3 = case_when(
    chimType %in% c("type20", "type21") ~ NA,
    .default = chimProp_c3
  )) %>%
  group_by(animalID) %>%
  summarise(
    mean.typeI.II.c = mean(chimProp_c3, na.rm = T),
    mean.typeI.c = mean(typeI_chimProp_c3, na.rm = T)
  ) %>%
  mutate(
    diff = mean.typeI.II.c - mean.typeI.c
  )

figPlot_mean.typeI.II.c_vs_mean.typeI.c <- figData_mean.typeI.II.c_vs_mean.typeI.c %>%
  ggplot(aes(x = mean.typeI.c, y = diff)) +
  geom_point() +
  ylim(-0.5, 0.5)

figPlot_mean.typeI.II.uc_vs_mean.typeI.c
figPlot_mean.typeI.II.c_vs_mean.typeI.c
```

**2) overall chimProp per sample**


```{r}
twinChimerism_allTwins_corrected %>%
  # 00, 01, 10, 11 = type 1, 20, 21 = type2
  mutate(markerType = case_when(
    chimType %in% c('type00', 'type01', 'type10', 'type11') ~ "type1",
    chimType %in% c('type20', 'type21') ~ "type2",
    .default = NA
  )) %>%
ggplot(aes(x = chimType, y = chimProp)) +
  geom_violin() + 
  geom_point(aes(col = chimType), position = "jitter") +
  stat_summary(fun=mean,  geom="point", color="black") +
  theme_bw()





  group_by()
ggplot(aes(x = chimType, y = chimProp)) +
  geom_violin() + 
  geom_point(aes(col = chimType), position = "jitter") +
  stat_summary(fun=mean,  geom="point", color="black") +
  theme_bw()


```


scrap plot stuff:
```{r}
install.packages("ggstatsplot", dependencies = T)
library(ggstatsplot)

ggbetwe





library(pheatmap)
# Assuming `chimProp_matrix` is a matrix of individuals (rows) by chimProp types (columns)

temp <- twinChimerism_allTwins_corrected %>%
  select(animalID, locus, chimType, chimProp) %>%
  pivot_wider(id_cols = c(animalID, locus),
              names_from = chimType,
              values_from = chimProp) %>%
  mutate(type1ab = coalesce(type00, type01),
         .keep = "unused")
  arrange(animalID)
pheatmap(chimProp_matrix, cluster_rows = TRUE, cluster_cols = TRUE,
         color = colorRampPalette(c("white", "blue"))(50))
```




# XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

old stuff below

# 5 GTscore 0x chimerism estimates

**Still need to double check that I changed everything correctly after copy/pasting the 10x scripts for the 0x tests, but it seems like 0x-cutoff didn't really change much for final chimerism estimates**

## 5.0 Data

### Hair genos & allele reads

Import 0x-cutoff genotypes for hair samples (allele reads file is the same as above)

```{r}
hairGenos_0x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  # replace 0 with NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # adjust locus name formatting
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  # filter to hair samples
  select(any_of(hairSamples)) %>%
  t() %>%
  as.data.frame()
```




### Blood allele reads

Blood allele reads file is the same as above

## 4.1 Pull twinPair hair & blood data

### Hair genos

```{r}
# Filter hairGenos to twin pairs
hairGenos_allTwins_0x <- hairGenos_0x %>%
  rownames_to_column("sampleID") %>%
  filter(sampleID %in% twinPairs_run5_l$sampleID_hair) %>%
  merge(., twinPairs_run5_l[, c("sampleID_hair", "twinID", "animalID")], by.x = "sampleID", by.y = "sampleID_hair") %>%
  select(-sampleID) %>%
  relocate(twinID) %>%
  # Move animalID to rowname
  column_to_rownames("animalID")

# Split into groups by twin pair (twinID)
hairGenos_twinPairs_0x <- split(hairGenos_allTwins_0x, f = hairGenos_allTwins_0x$twinID) %>%
  # Remove twinID column
  lapply(., function(x) x[!(names(x) %in% c("twinID"))]) %>%
  # Filter to loci where both twins have genos
  lapply(., function(x) x[, colSums(is.na(x)) == 0, drop = FALSE]) %>%
  # Filter to loci with NON-MATCHING genos
#  lapply(., function(x) x[, vapply(x, function(x) length(unique(x)) > 1, logical(1L)), drop = FALSE]) %>%
  # Remove empty dataframes
  discard(., ~length(as.matrix(.))==0) %>%
  # Transpose all dataframes
  lapply(., t) %>%
  lapply(., as.data.frame) %>%
  lapply(., rownames_to_column, "locus")
```

### Blood reads

```{r}
bloodReads_allTwins_0x <- twinPairs_run5_l %>%
  select(-species, -sampleID_hair) %>%
  merge(., bloodReads, by.x = "sampleID_blood", by.y = "row.names", all.x = T) %>%
  select(-sampleID_blood) %>%
  relocate(c(animalID, twinID)) %>%
  # filter to twinIDs present in hairGenos_twinPairs
  filter(twinID %in% names(hairGenos_twinPairs_0x)) %>%
  column_to_rownames("animalID")

bloodReads_twinPairs_0x <- split(bloodReads_allTwins_0x, f = bloodReads_allTwins_0x$twinID) %>%
  # Remove twinID column
  lapply(., function(x) x[!(names(x) %in% c("twinID"))]) %>%
  # Transpose all dataframes
  lapply(., t) %>%
  lapply(., as.data.frame) %>%
  lapply(., rownames_to_column, "locus")
```

## 4.2 Merge hair & blood data

```{r}
hairBlood_twinPairs_0x <- map2(hairGenos_twinPairs_0x, bloodReads_twinPairs_0x, ~left_join(.x, .y, by = "locus")) %>%
  lapply(., setNames, c("locus", "hairGeno_twin1", "hairGeno_twin2", "bloodReads_twin1", "bloodReads_twin2")) %>%
  map2(., names(.), ~ mutate(.x, twinID = .y)) %>%
  lapply(., relocate, twinID) %>%
  bind_rows(.) %>%
  # add twin animalIDs
  merge(., twinPairs_run5[, c("twinID", "animalID_twin1", "animalID_twin2")], by = "twinID") %>%
  relocate(c(twinID, animalID_twin1, animalID_twin2))
```

## 4.3 Initial chimerism estimates

To calculate initial chimerism estimates, I'm using the following
formulas from Vynck et al. (2023):

|                |                                                    |                           |
|------------------------|------------------------|------------------------|
| Chimerism type | Biallelic constellation (donor/recipient genotype) | Formula for estimating HC |
| type0          | AA/BB                                              | VAF [B/(A+B)]             |
| type1          | BB/AA                                              | 1 - VAF                   |
| type10         | AA/AB                                              | 2 x VAF                   |
| type11         | BB/AB                                              | 2 x (1 - VAF)             |
| type20         | AB/AA                                              | 1 - 2 x VAF               |
| type21         | AB/BB                                              | 2 x VAF - 1               |

### Functions

```{r}
# Identify alleles
allele1 <- function(x) str_sub(x, 1, 1)
allele2 <- function(x) str_sub(x, -1, -1)

# Define chimType
calc_chimType <- function(donorGeno, recipGeno) 
  case_when(
    # Donor/Recipient AA/BB
    allele1(donorGeno) != allele1(recipGeno) &
      allele1(donorGeno) != allele2(recipGeno) &
      allele1(donorGeno) < allele1(recipGeno) ~ "type00",
    # D/R BB/AA
    allele1(donorGeno) != allele1(recipGeno) &
      allele1(donorGeno) != allele2(recipGeno) &
      allele1(donorGeno) > allele1(recipGeno) ~ "type01",
    # D/R AA/AB
    allele1(donorGeno) == allele2(donorGeno) &
      allele1(donorGeno) == allele1(recipGeno) &
      allele2(donorGeno) != allele2(recipGeno) ~ "type10",
    # D/R BB/AB
    allele1(donorGeno) == allele2(donorGeno) &
      allele1(donorGeno) != allele1(recipGeno) &
      allele2(donorGeno) == allele2(recipGeno) ~ "type11",
    # D/R AB/AA
    allele1(donorGeno) != allele2(donorGeno) &
      allele1(donorGeno) == allele1(recipGeno) &
      allele2(donorGeno) != allele2(recipGeno) ~ "type20",
    # D/R AB/BB
    allele1(donorGeno) != allele2(donorGeno) &
      allele1(donorGeno) != allele1(recipGeno) &
      allele2(donorGeno) == allele2(recipGeno) ~ "type21",
    # D/R AA/AA or BB/BB
    allele1(donorGeno) == allele2(donorGeno) &
      allele1(donorGeno) == allele1(recipGeno) &
      allele2(donorGeno) == allele2(recipGeno) ~ "type30",
    # D/R AB/AB
    allele1(donorGeno) != allele2(donorGeno) &
      allele1(donorGeno) == allele1(recipGeno) &
      allele2(donorGeno) == allele2(recipGeno) ~ "type31"
    )

# Find blood read counts
bloodReads_allele1 <- function(x) as.numeric(sub(",.*", "", x))
bloodReads_allele2 <- function(x) as.numeric(sub(".*,", "", x))

# Calculate variant allele frequency
vaf <- function(recipBloodReads)
  round(bloodReads_allele2(recipBloodReads)/(bloodReads_allele1(recipBloodReads) + bloodReads_allele2(recipBloodReads)), 2)

# Calculate chimProp (hc)
calc_chimProp <- function(recipChimType, recipBloodReads) 
  case_when(
    recipChimType == "type00" ~ vaf(recipBloodReads),
    recipChimType == "type01" ~ 1 - vaf(recipBloodReads),
    recipChimType == "type10" ~ 2 * vaf(recipBloodReads),
    recipChimType == "type11" ~ 2 * (1 - vaf(recipBloodReads)),
    recipChimType == "type20" ~ 1 - 2 * vaf(recipBloodReads),
    recipChimType == "type21" ~ 2 * vaf(recipBloodReads) - 1
  )
```

### Calculations

```{r}
# Identify chimType & calculate VAF
twinChimerism_0x <- hairBlood_twinPairs_0x %>%
  # twin1 as recipient, twin2 as donor
  mutate(
    chimType_twin1 = calc_chimType(donorGeno = hairGeno_twin2, recipGeno = hairGeno_twin1),
    vaf_twin1 = vaf(recipBloodReads = bloodReads_twin1),
    chimProp_twin1 = calc_chimProp(recipChimType = chimType_twin1, recipBloodReads = bloodReads_twin1)
    ) %>%
  # twin1 as donor, twin2 as recipient
  mutate(
    chimType_twin2 = calc_chimType(donorGeno = hairGeno_twin1, recipGeno = hairGeno_twin2),
    vaf_twin2 = vaf(recipBloodReads = bloodReads_twin2),
    chimProp_twin2 = calc_chimProp(recipChimType = chimType_twin2, recipBloodReads = bloodReads_twin2)
  ) %>%
  arrange(as.numeric(twinID))

twinChimerism_0x %>%
  kable(caption = "GTscore 0x: Estimated % twin chimerism for twin pairs by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Reformat for individual twins vs. pairs
twinChimerism_twin1_0x <- twinChimerism_0x %>%
  select(c("twinID", "locus", contains("twin1")))

twinChimerism_twin2_0x <- twinChimerism_0x %>%
  select(c("twinID", "locus", contains("twin2")))
colnames(twinChimerism_twin2_0x) <- colnames(twinChimerism_twin1_0x)

twinChimerism_allTwins_0x <- rbind(twinChimerism_twin1_0x, twinChimerism_twin2_0x) %>%
  rename(
    "animalID" = "animalID_twin1",
    "hairGenos" = "hairGeno_twin1",
    "bloodReads" = "bloodReads_twin1",
    "chimType" = "chimType_twin1",
    "vaf" = "vaf_twin1",
    "chimProp" = "chimProp_twin1"
  ) %>%
#  na.omit() %>%
  merge(., md_hairSamples[, c("animalID", "species")], by = "animalID") %>%
  relocate(c(twinID, animalID, species)) %>%
  arrange(as.numeric(twinID))

twinChimerism_allTwins_0x %>%
  kable(caption = "GTscore 0x: Estimated % twin chimerism for individuals by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Subset by species
twinChimerism_lwed_0x <- twinChimerism_allTwins_0x %>%
  filter(species == "LWED")

twinChimerism_simp_0x <- twinChimerism_allTwins_0x %>%
  filter(species == "SIMP")
```

## 4.4 Bias correction

Following Vynck et al. (2023), marker bias is calculated as the average deviation from 50% allele frequency in reference samples with allele frequencies between 40% and 60%. They performed this calculation for "(i) markers in reference samples with a VAF between 40% and 60% and (ii) type-III markers in follow-up samples with a VAF between 40% and 60%, that is where both the donor and the recipient are heterozygous. The average deviation was calculated in the combined set (i) and (ii), as well as separately". 

As such, I'm doing the following:

1.  Calculate VAF for all hair reads
    1.  Subset data to only values between 0.4 and 0.6
    2.  Calculate deviation from 0.5 for each value
    3.  Calculate median deviation from 0.5 for each locus
2.  Calculate VAF for twinPair blood reads where BOTH twins are AB/AB
    1.  Subset data to only values between 0.4 and 0.6
    2.  Calculate deviation from 0.5 for each value
    3.  Calculate median deviation from 0.5 for each locus
3.  Calculate chimProp w/bias corrections
    1.  chimProp_c1 - correct based on hair sample bias
    2.  chimProp_c2 - correct based on blood sample bias
    3.  chimProp_c3 - correct based on hair AND blood sample bias

### Functions

```{r}
calc_chimProp_c <- function(recipChimType, HC, b)
  case_when(
        recipChimType == "type00" ~ round((HC-2*HC*b)/(-4*b*HC+2*b+1), 2),
        recipChimType == "type01" ~ round((2*b*HC+HC)/(4*b*HC-2*b+1), 2),
        recipChimType == "type10" ~ round((HC-2*b*HC)/(-2*b*HC+2*b+1), 2),
        recipChimType == "type11" ~ round((2*b*HC+HC)/(2*b*(HC-1)+1), 2),
        recipChimType == "type20" ~ round((2*b+HC)/(2*b*HC+1), 2),
        recipChimType == "type21" ~ round((2*b-HC)/(2*b*HC-1), 2)
  )
```

### Calculations

```{r}
# Calculate VAF for all hair reads
vaf_hairReads <- replace(hairReads, TRUE, lapply(hairReads, vaf)) %>%
  mutate(across(everything(),as.numeric)) %>%
  rownames_to_column("locus")

# Calculate VAF for AB/AB type-III blood reads
vaf_bloodReads_0x <- twinChimerism_allTwins_0x %>%
  select(c("animalID", "locus", "chimType", "vaf")) %>%
  merge(., md_bloodSamples[, c("sampleID", "animalID")], by = "animalID") %>%
  select(-animalID) %>%
  arrange(sampleID) %>%
  mutate(
    vaf = case_when(
      chimType != "type31" ~ NA,
      .default = vaf
      )
    ) %>%
  select(-chimType) %>%
  pivot_wider(., names_from = sampleID, values_from = vaf)

vaf_bloodReads_sampleList_0x <- vaf_bloodReads_0x %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  select(sampleID) %>%
  pull()

# Combine VAF from hair & blood reads
vaf_allReads_0x <- merge(vaf_hairReads, vaf_bloodReads_0x, by = "locus", all.x = T) %>%
  column_to_rownames("locus")

# Subset to VAF between 0.4 and 0.6
vaf_allReads_0x[vaf_allReads_0x > 0.6] <- NA
vaf_allReads_0x[vaf_allReads_0x < 0.4] <- NA

# Get median deviation from 0.5 per locus
## all hair samples
## type31 blood samples
## combination of both

vafDev_0x <- (vaf_allReads_0x - 0.5) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    b1 = median(c_across(hairSamples), na.rm = TRUE),
    b2 = median(c_across(vaf_bloodReads_sampleList_0x), na.rm = TRUE),
    b3 = median(c_across(where(is.numeric)), na.rm = TRUE)
    ) %>%
  select(c("locus", "b1", "b2", "b3"))

# Calculate chimerism corrected for b1, b2, and b3 bias values
twinChimerism_allTwins_corrected_0x <- merge(twinChimerism_allTwins_0x, vafDev_0x, by = "locus", all.x = T) %>%
  mutate(
    chimProp_c1 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b1),
    chimProp_c2 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b2),
    chimProp_c3 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b3)
  ) %>%
  filter(
    chimType != "type30",
    chimType != "type31"
  ) %>%
  filter(
    !is.na(bloodReads)
  ) %>%
  arrange(as.numeric(twinID))

twinChimerism_allTwins_corrected_0x %>%
  kable(caption = "GTscore 0x: Bias-corrected estimated % twin chimerism for individuals by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Subset by species
twinChimerism_lwed_corrected_0x <- twinChimerism_allTwins_corrected_0x %>%
  filter(species == "LWED")

twinChimerism_simp_corrected_0x <- twinChimerism_allTwins_corrected_0x %>%
  filter(species == "SIMP")
```

## 4.3 Summaries

```{r}
twinChimerism_lwed_summary_0x <- twinChimerism_lwed_corrected_0x %>%
  group_by(animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = round(mean(chimProp), 2),
    mean_chimProp_c1 = round(mean(chimProp_c1), 2),
    mean_chimProp_c2 = round(mean(chimProp_c2), 2),
    mean_chimProp_c3 = round(mean(chimProp_c3), 2)
  ) %>%
  as.data.frame() %>%
  mutate(species = "LWED") %>%
  relocate(species)

twinChimerism_lwed_summary_0x %>%
  kable(caption = "GTscore 0x: LWED avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

twinChimerism_simp_summary_0x <- twinChimerism_simp_corrected_0x %>%
  group_by(animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = round(mean(chimProp), 2),
    mean_chimProp_c1 = round(mean(chimProp_c1), 2),
    mean_chimProp_c2 = round(mean(chimProp_c2), 2),
    mean_chimProp_c3 = round(mean(chimProp_c3), 2)
  ) %>%
  as.data.frame() %>%
  mutate(species = "SIMP") %>%
  relocate(species)

twinChimerism_simp_summary_0x %>%
  kable(caption = "GTscore 10x: SIMP avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")
```



# 6 GTseq chimerism estimates

## 6.0 Data

### Hair genos

```{r, warning=FALSE}
hairGenos_gtseq <- read_csv("./05_tamRun5/05_run5GTseq/genos/compiledGenos_gtseq_tamRun5.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  select(-Raw.Reads, -On.Target.Reads, -X.On.Target, -X.GT, -IFI) %>%
  # filter to hair samples
  filter(Sample %in% hairSamples) %>%
  column_to_rownames("Sample") %>%
  t() %>%
  as.data.frame() %>%
  # add commas so we can reuse the same functions
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ str_c(substr(., 1, 1), ",", substr(., 2, 2)))) %>%
  # replace 0 with NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                )) %>%
  t() %>%
  as.data.frame()
```

### Allele reads

#### Create allele reads file

GTseq doesn't automatically output allele reads in an easily usable
format - instead, we need to extract this data from the ".genos" files
as follows:

1.  Convert .genos files to csv files
2.  Import as one dataframe
3.  Subset relevant columns and reformat

**Convert .genos to .csv in bash**

```{bash, eval = FALSE}
# first copied all .genos files to genos_csv
# then cd to genos_csv and run the following:

for f in *.genos; do
    g="${f%.genos}.csv"
    mv "${f}" "${g}"
done
```

**Import all csv files as one dataframe & reformat**

```{r}
fileNames <- dir("./05_tamRun5/05_run5GTseq/genos/genos_csv", full.names = T)

readCounts_gtseq.list <- lapply(fileNames, function(file.name) {
  df <- read.csv(file.name, skip = 1, header = F)
  df$file.name <- file.name
  return(df)
  })

readCounts_gtseq <- rlist::list.rbind(readCounts_gtseq.list) %>%
  rename("locus" = "V1",
         "allele1_counts" = "V2",
         "allele2_counts" = "V3",
         "a1.a2_ratio" = "V4",
         "geno" = "V5",
         "genoClass" = "V6",
         "a1_correctionValue" = "V7",
         "a2_corectionValue" = "V8",
         "onTargetReads" = "V9",
         "locus_otProp.readTotal" = "V10",
         "locus_otProp.otTotal" = "V11",
         "sampleID" = "file.name") %>%
  mutate(
    readCounts = str_c(sub(".*=", "", allele1_counts), ",", sub(".*=", "", allele2_counts)),
    sampleID = sub("./05_tamRun5/05_run5GTseq/genos/genos_csv/", "", sampleID),
    sampleID = sub(".csv", "", sampleID),
    sampleID = gsub("-", "\\.", sampleID)
    ) %>%
  select(c("sampleID", "locus", "readCounts")) %>%
  pivot_wider(names_from = sampleID, values_from = readCounts) %>%
  column_to_rownames("locus") %>%
  # replace 0,0 with NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))
```

#### Hair reads

```{r}
hairReads_gtseq <- readCounts_gtseq %>%
  select(any_of(hairSamples)) %>%
  as.data.frame()
```

#### Blood reads

```{r}
bloodReads_gtseq <- readCounts_gtseq %>%
  select(any_of(bloodSamples)) %>%
  t() %>%
  as.data.frame()
```

## 6.1 Pull twinPair hair & blood data

### Hair genos

```{r}
# Filter hairGenos to twin pairs
hairGenos_allTwins_gtseq <- hairGenos_gtseq %>%
  rownames_to_column("sampleID") %>%
  filter(sampleID %in% twinPairs_run5_l$sampleID_hair) %>%
  merge(., twinPairs_run5_l[, c("sampleID_hair", "twinID", "animalID")], by.x = "sampleID", by.y = "sampleID_hair") %>%
  select(-sampleID) %>%
  relocate(twinID) %>%
  # Move animalID to rowname
  column_to_rownames("animalID")

# Split into groups by twin pair (twinID)
hairGenos_twinPairs_gtseq <- split(hairGenos_allTwins_gtseq, f = hairGenos_allTwins_gtseq$twinID) %>%
  # Remove twinID column
  lapply(., function(x) x[!(names(x) %in% c("twinID"))]) %>%
  # Filter to loci where both twins have genos
  lapply(., function(x) x[, colSums(is.na(x)) == 0, drop = FALSE]) %>%
  # Filter to loci with NON-MATCHING genos
#  lapply(., function(x) x[, vapply(x, function(x) length(unique(x)) > 1, logical(1L)), drop = FALSE]) %>%
  # Remove empty dataframes
  discard(., ~length(as.matrix(.))==0) %>%
  # Transpose all dataframes
  lapply(., t) %>%
  lapply(., as.data.frame) %>%
  lapply(., rownames_to_column, "locus")
```

### Blood reads

```{r}
bloodReads_allTwins_gtseq <- twinPairs_run5_l %>%
  select(-species, -sampleID_hair) %>%
  merge(., bloodReads_gtseq, by.x = "sampleID_blood", by.y = "row.names", all.x = T) %>%
  select(-sampleID_blood) %>%
  relocate(c(animalID, twinID)) %>%
  # filter to twinIDs present in hairGenos_twinPairs
  filter(twinID %in% names(hairGenos_twinPairs)) %>%
  column_to_rownames("animalID")

bloodReads_twinPairs_gtseq <- split(bloodReads_allTwins_gtseq, f = bloodReads_allTwins_gtseq$twinID) %>%
  # Remove twinID column
  lapply(., function(x) x[!(names(x) %in% c("twinID"))]) %>%
  # Transpose all dataframes
  lapply(., t) %>%
  lapply(., as.data.frame) %>%
  lapply(., rownames_to_column, "locus")
```

## 6.2 Merge hair & blood data

```{r}
hairBlood_twinPairs_gtseq <- map2(hairGenos_twinPairs_gtseq, bloodReads_twinPairs_gtseq, ~left_join(.x, .y, by = "locus")) %>%
  lapply(., setNames, c("locus", "hairGeno_twin1", "hairGeno_twin2", "bloodReads_twin1", "bloodReads_twin2")) %>%
  map2(., names(.), ~ mutate(.x, twinID = .y)) %>%
  lapply(., relocate, twinID) %>%
  bind_rows(.) %>%
  # add twin animalIDs
  merge(., twinPairs_run5[, c("twinID", "animalID_twin1", "animalID_twin2")], by = "twinID") %>%
  relocate(c(twinID, animalID_twin1, animalID_twin2))
```

## 6.3 Initial chimerism estimates

Initial chimerism estimates are performed as above (section 4.3)

### Calculations

```{r}
# Identify chimType & calculate VAF
twinChimerism_gtseq <- hairBlood_twinPairs_gtseq %>%
  # twin1 as recipient, twin2 as donor
  mutate(
    chimType_twin1 = calc_chimType(donorGeno = hairGeno_twin2, recipGeno = hairGeno_twin1),
    vaf_twin1 = vaf(recipBloodReads = bloodReads_twin1),
    chimProp_twin1 = calc_chimProp(recipChimType = chimType_twin1, recipBloodReads = bloodReads_twin1)
    ) %>%
  # twin1 as donor, twin2 as recipient
  mutate(
    chimType_twin2 = calc_chimType(donorGeno = hairGeno_twin1, recipGeno = hairGeno_twin2),
    vaf_twin2 = vaf(recipBloodReads = bloodReads_twin2),
    chimProp_twin2 = calc_chimProp(recipChimType = chimType_twin2, recipBloodReads = bloodReads_twin2)
  ) %>%
  arrange(as.numeric(twinID))

twinChimerism_gtseq %>%
  kable(caption = "GTseq 10x: Estimated % twin chimerism for twin pairs by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Reformat for individual twins vs. pairs
twinChimerism_twin1_gtseq <- twinChimerism_gtseq %>%
  select(c("twinID", "locus", contains("twin1")))

twinChimerism_twin2_gtseq <- twinChimerism_gtseq %>%
  select(c("twinID", "locus", contains("twin2")))
colnames(twinChimerism_twin2_gtseq) <- colnames(twinChimerism_twin1_gtseq)

twinChimerism_allTwins_gtseq <- rbind(twinChimerism_twin1_gtseq, twinChimerism_twin2_gtseq) %>%
  rename(
    "animalID" = "animalID_twin1",
    "hairGenos" = "hairGeno_twin1",
    "bloodReads" = "bloodReads_twin1",
    "chimType" = "chimType_twin1",
    "vaf" = "vaf_twin1",
    "chimProp" = "chimProp_twin1"
  ) %>%
#  na.omit() %>%
  merge(., md_hairSamples[, c("animalID", "species")], by = "animalID") %>%
  relocate(c(twinID, animalID, species)) %>%
  arrange(as.numeric(twinID))

twinChimerism_allTwins_gtseq %>%
  kable(caption = "GTseq 10x: Estimated % twin chimerism for individuals by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Subset by species
twinChimerism_lwed_gtseq <- twinChimerism_allTwins_gtseq %>%
  filter(species == "LWED")

twinChimerism_simp_gtseq <- twinChimerism_allTwins_gtseq %>%
  filter(species == "SIMP")
```

## 6.4 Bias correction

Bias correction is performed as above (section 4.4).

### Calculations

```{r}
# Calculate VAF for all hair reads
vaf_hairReads_gtseq <- replace(hairReads_gtseq, TRUE, lapply(hairReads_gtseq, vaf)) %>%
  mutate(across(everything(),as.numeric)) %>%
  rownames_to_column("locus")

# Calculate VAF for AB/AB type-III blood reads
vaf_bloodReads_gtseq <- twinChimerism_allTwins_gtseq %>%
  select(c("animalID", "locus", "chimType", "vaf")) %>%
  merge(., md_bloodSamples[, c("sampleID", "animalID")], by = "animalID") %>%
  select(-animalID) %>%
  arrange(sampleID) %>%
  mutate(
    vaf = case_when(
      chimType != "type31" ~ NA,
      .default = vaf
      )
    ) %>%
  select(-chimType) %>%
  pivot_wider(., names_from = sampleID, values_from = vaf)

vaf_bloodReads_sampleList_gtseq <- vaf_bloodReads %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  select(sampleID) %>%
  pull()

# Combine VAF from hair & blood reads
vaf_allReads_gtseq <- merge(vaf_hairReads_gtseq, vaf_bloodReads_gtseq, by = "locus", all.x = T) %>%
  column_to_rownames("locus")

# Subset to VAF between 0.4 and 0.6
vaf_allReads_gtseq[vaf_allReads_gtseq > 0.6] <- NA
vaf_allReads_gtseq[vaf_allReads_gtseq < 0.4] <- NA

# Get median deviation from 0.5 per locus
## all hair samples
## type31 blood samples
## combination of both

vafDev_gtseq <- (vaf_allReads_gtseq - 0.5) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    b1 = median(c_across(hairSamples), na.rm = TRUE),
    b2 = median(c_across(vaf_bloodReads_sampleList_gtseq), na.rm = TRUE),
    b3 = median(c_across(where(is.numeric)), na.rm = TRUE)
    ) %>%
  select(c("locus", "b1", "b2", "b3"))

# Calculate chimerism corrected for b1, b2, and b3 bias values
twinChimerism_allTwins_corrected_gtseq <- merge(twinChimerism_allTwins_gtseq, vafDev_gtseq, by = "locus", all.x = T) %>%
  mutate(
    chimProp_c1 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b1),
    chimProp_c2 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b2),
    chimProp_c3 = calc_chimProp_c(recipChimType = chimType, HC = chimProp, b = b3)
  ) %>%
  filter(
    chimType != "type30",
    chimType != "type31"
  ) %>%
  filter(
    !is.na(bloodReads)
  ) %>%
  arrange(as.numeric(twinID))

twinChimerism_allTwins_corrected_gtseq %>%
  kable(caption = "GTseq 10x: Bias-corrected estimated % twin chimerism for individuals by locus") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# Subset by species
twinChimerism_lwed_corrected_gtseq <- twinChimerism_allTwins_corrected_gtseq %>%
  filter(species == "LWED")

twinChimerism_simp_corrected_gtseq <- twinChimerism_allTwins_corrected_gtseq %>%
  filter(species == "SIMP")
```

## 6.3 Summaries

```{r}
twinChimerism_lwed_summary_gtseq <- twinChimerism_lwed_corrected_gtseq %>%
  group_by(animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = round(mean(chimProp), 2),
    mean_chimProp_c1 = round(mean(chimProp_c1), 2),
    mean_chimProp_c2 = round(mean(chimProp_c2), 2),
    mean_chimProp_c3 = round(mean(chimProp_c3), 2)
  ) %>%
  as.data.frame() %>%
  mutate(species = "LWED") %>%
  relocate(species)

twinChimerism_lwed_summary_gtseq %>%
  kable(caption = "GTseq 10x: LWED avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

twinChimerism_simp_summary_gtseq <- twinChimerism_simp_corrected_gtseq %>%
  group_by(animalID) %>%
  summarise(
    n_loci = n(),
    mean_chimProp = round(mean(chimProp), 2),
    mean_chimProp_c1 = round(mean(chimProp_c1), 2),
    mean_chimProp_c2 = round(mean(chimProp_c2), 2),
    mean_chimProp_c3 = round(mean(chimProp_c3), 2)
  ) %>%
  as.data.frame() %>%
  mutate(species = "SIMP") %>%
  relocate(species)

twinChimerism_simp_summary_gtseq %>%
  kable(caption = "GTseq 10x: SIMP avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")
```

# 6 GTscore vs. GTseq

## 6.1 Summary comparison

```{r}
# LWED
gtscore.seq_lwed_summary <- merge(twinChimerism_lwed_summary, twinChimerism_lwed_summary_gtseq, by = c("species", "animalID"), all = T, suffixes = c("_gtscore", "_gtseq")) %>%
  select(order(colnames(.))) %>%
  relocate(c(species, animalID, n_loci_gtscore, n_loci_gtseq, mean_chimProp_gtscore, mean_chimProp_gtseq))

gtscore.seq_lwed_summary %>%
  kable(caption = "GTscore vs GTseq: LWED avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")

# SIMP
gtscore.seq_simp_summary <- merge(twinChimerism_simp_summary, twinChimerism_simp_summary_gtseq, by = c("species", "animalID"), all = T, suffixes = c("_gtscore", "_gtseq")) %>%
  select(order(colnames(.))) %>%
  relocate(c(species, animalID, n_loci_gtscore, n_loci_gtseq, mean_chimProp_gtscore, mean_chimProp_gtseq))

gtscore.seq_simp_summary %>%
  kable(caption = "GTscore vs GTseq: SIMP avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")
```

## 6.2 Mismatches

Did any of the included samples have mismatched genotypes for species or sex compared to metadata?

### Import mismatches

```{r}
# Species mismatches
mismatchSpecies_gtscore10x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/genos_species_mismatch_gtscore10x.csv")

mismatchSpecies_gtscore30x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/genos_species_mismatch_gtscore30x.csv")

mismatchSpecies_gtseq <- read.csv("./05_tamRun5/05_run5GTseq/summaryFiles/genos_species_mismatch_gtseq.csv")

# Sex mismatches (hair samples only)
mismatchSex_gtscore10x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/genos_sex_mismatchAll_gtscore10x.csv")

mismatchSex_gtscore30x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/genos_sex_mismatchAll_gtscore30x.csv")

mismatchSex_gtseq <- read.csv("./05_tamRun5/05_run5GTseq/summaryFiles/genos_sex_mismatchAll_gtseq.csv")
```

### Note mismatches

#### LWED

animalID 189 had mismatched species genotypes, and animalID 190 had mismatched sex genotypes - they're from the same twin pair. Probably a good idea to remove them from TABA final results

```{r}
gtscore.seq_lwed_summary_mm <- gtscore.seq_lwed_summary %>%
  mutate(
    mmSp_gtscore10x = case_when(
      animalID %in% mismatchSpecies_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSp_gtscore30x = case_when(
      animalID %in% mismatchSpecies_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSp_gtseq = case_when(
      animalID %in% mismatchSpecies_gtseq$animalID ~ "yes",
      .default = "no")
    ) %>%
  relocate(c(mmSp_gtscore10x, mmSp_gtscore30x, mmSp_gtseq), .after = animalID) %>%
  mutate(
    mmSex_gtscore10x = case_when(
      animalID %in% mismatchSex_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSex_gtscore30x = case_when(
      animalID %in% mismatchSex_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSex_gtseq = case_when(
      animalID %in% mismatchSex_gtseq$animalID ~ "yes",
      .default = "no")
    ) %>%
  relocate(c(mmSex_gtscore10x, mmSex_gtscore30x, mmSex_gtseq), .after = mmSp_gtseq)

gtscore.seq_lwed_summary_mm %>%
  kable(caption = "GTscore vs GTseq: Species/sex mismatches in LWED avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")
```

#### SIMP

Species and sex mismatches were found in animalID 211 & 212 (which belong to the same twin pair)

Just sex mismatches were found in animalID 167 & 168 (same twin pair) as well as animalID 55 (twin is animalID 93, which didn't have a blood sample)

Probably also good to remove these individuals from TABA final results.

```{r}
gtscore.seq_simp_summary_mm <- gtscore.seq_simp_summary %>%
  mutate(
    mmSp_gtscore10x = case_when(
      animalID %in% mismatchSpecies_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSp_gtscore30x = case_when(
      animalID %in% mismatchSpecies_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSp_gtseq = case_when(
      animalID %in% mismatchSpecies_gtseq$animalID ~ "yes",
      .default = "no")
    ) %>%
  relocate(c(mmSp_gtscore10x, mmSp_gtscore30x, mmSp_gtseq), .after = animalID) %>%
  mutate(
    mmSex_gtscore10x = case_when(
      animalID %in% mismatchSex_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSex_gtscore30x = case_when(
      animalID %in% mismatchSex_gtscore10x$animalID ~ "yes",
      .default = "no"),
    mmSex_gtseq = case_when(
      animalID %in% mismatchSex_gtseq$animalID ~ "yes",
      .default = "no")
    ) %>%
  relocate(c(mmSex_gtscore10x, mmSex_gtscore30x, mmSex_gtseq), .after = mmSp_gtseq)

gtscore.seq_simp_summary_mm %>%
  kable(caption = "GTscore vs GTseq: Species/sex mismatches in LWED avg. estimated % chimerism") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"), height = "300px")
```

# Both species summary

Remove samples with mismatches and/or < 3 loci

```{r}
gtscore.seq_lwed_summary_final <- gtscore.seq_lwed_summary_mm %>%
  filter(mmSp_gtscore10x == "no") %>%
  filter(mmSex_gtscore10x == "no") %>%
  filter(n_loci_gtscore >=3) %>%
  merge(., twinPairs_run5_l[, c("animalID", "twinID")], by = "animalID")

gtscore.seq_simp_summary_final <- gtscore.seq_simp_summary_mm %>%
  filter(mmSp_gtscore10x == "no") %>%
  filter(mmSex_gtscore10x == "no") %>%
  filter(n_loci_gtscore >=3) %>%
  merge(., twinPairs_run5_l[, c("animalID", "twinID")], by = "animalID")
```

```{r}
gtscore.seq_lwedSIMP_summary_final <- rbind(gtscore.seq_lwed_summary_final, gtscore.seq_simp_summary_final)

mean(gtscore.seq_lwedSIMP_summary_final$n_loci_gtscore)

min(gtscore.seq_lwed_summary_final$mean_chimProp_gtscore)
max(gtscore.seq_lwed_summary_final$mean_chimProp_gtscore)

min(gtscore.seq_simp_summary_final$mean_chimProp_gtscore)
max(gtscore.seq_simp_summary_final$mean_chimProp_gtscore)
```




# SCRAPS

## Check ckmrSim r coefficients

```{r}
lwed_logls_fs <- read.csv("./DISSERTATION/ch2_demography/02_results/lwed/ckmr_logls_fs.u.csv")
simp_logls_fs <- read.csv("./DISSERTATION/ch2_demography/02_results/simp/ckmr_logls_fs.u.csv")
logls_fs <- rbind(lwed_logls_fs, simp_logls_fs) %>%
  select(sib2, sib1, logl_ratio, num_loc) %>%
  dplyr::rename("logl_fs" = "logl_ratio",
                "numLoci_fs" = "num_loc") %>%
  na.omit()

lwed_logls_hs <- read.csv("./DISSERTATION/ch2_demography/02_results/lwed/ckmr_logls_hs.u.csv")
simp_logls_hs <- read.csv("./DISSERTATION/ch2_demography/02_results/simp/ckmr_logls_hs.u.csv")
logls_hs <- rbind(lwed_logls_hs, simp_logls_hs) %>%
  select(sib2, sib1, logl_ratio, num_loc) %>%
  dplyr::rename("logl_hs" = "logl_ratio",
                "numLoci_hs" = "num_loc") %>%
  na.omit()

temp <- data.frame(
  species = c("LWED", "LWED", "LWED", "LWED"),
  groupName = c("Royals", "Royals", "Ghosts", "Ghosts"),
  twin1 = c(161, 162, 222, 223),
  twin2 = c(163, 163, 224, 224)
)

test <- get_dupSamples(genos_clean) %>%
  filter(proportionCommon > 0.3) %>%
  merge(., sampleRef[, c("sampleID", "animalID", "sampleType")], by.x = "Sample1", by.y = "sampleID") %>%
  merge(., sampleRef[, c("sampleID", "animalID", "sampleType")], by.x = "Sample2", by.y = "sampleID", suffixes = c("1", "2"))

twinList_ckmrCheck <- twinList %>%
  mutate(twin1 = as.numeric(twin1),
         twin2 = as.numeric(twin2)) %>%
  select(-twin3) %>%
  rbind(., temp) %>%
  merge(., logls_fs, by.x = c("twin1", "twin2"), by.y = c("sib2", "sib1"), all.x = T) %>%
  merge(., logls_hs, by.x = c("twin1", "twin2"), by.y = c("sib2", "sib1"), all.x = T) %>%
  merge(., test[, c("animalID1", "animalID2", "proportionMatch", "commonGenotypes", "sampleType1", "sampleType2")], by.x = c("twin1", "twin2"), by.y = c("animalID1", "animalID2"), all.x = T)
```

## Address triplet situation

For reference, we have two cases of groups w/3 juvs:
-   Royals 2015 - 161, 162, 163
-   Ghosts 2018 - 222, 223, 224

Trap report info:
-   Royals: "The group is now made up of 2 adult females, 2 adult males, and 3 juvenile males. Due to how unusual it is to find so many juveniles in a group as well as looking at the morphometrics of both adult females, we suspect that both of these females in the group are reproducing."
-   Ghosts: no details given, just mention that there were three untagged juvs in the group when captured

Hair samples for 222 and 224 are 0% genoSuccess, so we don't need to worry about them

```{r}
triplet_genoSuccess <- genoSuccess %>%
  merge(., sampleRef[, c("sampleID", "animalID", "sampleType")], by = "sampleID") %>%
  filter(animalID %in% c(161, 162, 163, 222, 223, 224)) %>%
  arrange(animalID)
```

**Duptest**

dupTest results suggest that **161/162** and **223/224** should be considered twin pairs, while 163 and 222 seem quite closely related (so maybe half-sibs?)

```{r}
triplet_dupTest <- genos_clean %>%
  select(any_of(triplet_genoSuccess[triplet_genoSuccess$sampleType == "blood", "sampleID"])) %>%
  get_dupSamples(.) %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "Sample1", by.y = "sampleID") %>%
  merge(., sampleRef[, c("sampleID", "animalID")], by.x = "Sample2", by.y = "sampleID", suffixes = c("1", "2"))

# 161/162 propMatch = 0.92; next highest is 0.72
triplet_dupTest %>%
  filter(animalID1 %in% c(161, 162, 163)) %>%
  filter(animalID2 %in% c(161, 162, 163)) %>%
  arrange(desc(proportionMatch)) %>%
  relocate(c(animalID1, animalID2)) %>%
  select(-c(Sample1, Sample2))

# 223/224 propMatch = 0.95; next highest is 0.86
triplet_dupTest %>%
  filter(animalID1 %in% c(222, 223, 224)) %>%
  filter(animalID2 %in% c(222, 223, 224)) %>%
  arrange(desc(proportionMatch)) %>%
  relocate(c(animalID1, animalID2)) %>%
  select(-c(Sample1, Sample2))
```

## Twin pair list

```{r}
twinPairs_run5 <- twinList %>%
  # recode triplets based on duptest above
  mutate(
    twin1 = case_when(
      !is.na(twin3) & groupName == "Royals" ~ "161",
      !is.na(twin3) & groupName == "Ghosts" ~ "223",
      .default = twin1
    ),
    twin2 = case_when(
      !is.na(twin3) & groupName == "Royals" ~ "162",
      !is.na(twin3) & groupName == "Ghosts" ~ "224",
      .default = twin2
    )
  ) %>%
  select(-twin3) %>%
  arrange(groupName) %>%
  mutate(twinID = row_number()) %>%
  relocate(twinID) %>%
  # add hair sample ID & species (need samples for both)
  mutate(twin1 = as.character(twin1),
         twin2 = as.character(twin2)) %>%
  mutate_if(is.character, str_trim) %>%
  merge(., md_hair_clean[, c("animalID", "sampleID")], by.x = "twin1", by.y = "animalID") %>%
  rename("sampleID_hair_twin1" = "sampleID",
         "animalID_twin1" = "twin1") %>%
  merge(., md_hair_clean[, c("animalID", "sampleID")], by.x = "twin2", by.y = "animalID") %>%
  rename("sampleID_hair_twin2" = "sampleID",
         "animalID_twin2" = "twin2") %>%
  # add blood sample ID
  merge(., md_blood_clean[, c("animalID", "sampleID")], by.x = "animalID_twin1", by.y = "animalID", all.x = T) %>%
  rename("sampleID_blood_twin1" = "sampleID") %>%
  merge(., md_blood_clean[, c("animalID", "sampleID")], by.x = "animalID_twin2", by.y = "animalID", all.x = T) %>%
  rename("sampleID_blood_twin2" = "sampleID") %>%
  # remove twin pair only if both blood samples are missing
  filter_at(vars(sampleID_blood_twin1, sampleID_blood_twin2), any_vars(!is.na(.))) %>%
  select(c("twinID", "species", "animalID_twin1", "animalID_twin2", "sampleID_hair_twin1", "sampleID_hair_twin2", "sampleID_blood_twin1", "sampleID_blood_twin2")) %>%
  arrange(twinID)

twinPairs_run5_lwed <- twinPairs_run5 %>%
  filter(species == "LWED")

twinPairs_run5_simp <- twinPairs_run5 %>%
  filter(species == "SIMP")

# Create long-format dataframe; gives each individual + their hair/blood sample IDs
twinPairs_twin1 <- twinPairs_run5 %>%
  select(c("twinID", "species", contains("twin1"))) %>%
  dplyr::rename("animalID" = "animalID_twin1",
         "sampleID_hair" = "sampleID_hair_twin1",
         "sampleID_blood" = "sampleID_blood_twin1")

twinPairs_twin2 <- twinPairs_run5 %>%
  select(c("twinID", "species", contains("twin2"))) %>%
  dplyr::rename("animalID" = "animalID_twin2",
         "sampleID_hair" = "sampleID_hair_twin2",
         "sampleID_blood" = "sampleID_blood_twin2")

twinPairs_run5_l <- rbind(twinPairs_twin1, twinPairs_twin2)

# Twin sample ID lists by sample type
twinPairs_hair_sampleList <- data.frame(
  sampleID_hair = c(twinPairs_run5$sampleID_hair_twin1, twinPairs_run5$sampleID_hair_twin2)
) %>%
  pull()

twinPairs_blood_sampleList <- data.frame(
  sampleID_blood = c(twinPairs_run5$sampleID_blood_twin1, twinPairs_run5$sampleID_blood_twin2)
) %>%
  pull()
```

## 6.6 Trueness

this stuff needs another look

```{r}
# avg HC est b/t type2.uc_vs_type1.c
figData_mean.type2.uc_vs_mean.type1.c <- twinChimerism_allTwins_corrected %>%
  mutate(
    chimProp = case_when(
      !chimType %in% c("type20", "type21") ~ NA,
      .default = chimProp
    ),
    chimProp_c3 = case_when(
      chimType %in% c("type20", "type21") ~ NA,
      .default = chimProp_c3
      )
    ) %>%
  group_by(animalID) %>%
  summarise(
    mean.type2.uc = mean(chimProp, na.rm = T),
    mean.type1.c = mean(chimProp_c3, na.rm = T)
  ) %>%
  mutate(
    diff = mean.type2.uc - mean.type1.c
  )

figData_mean.type2.uc_vs_mean.type1.c %>%
  ggplot(aes(x = mean.type1.c, y = diff)) +
  geom_point() +
  ylim(-0.5, 0.5)

# avg HC est b/t type2.c_vs_type1.c
figData_mean.type2.c_vs_mean.type1.c <- twinChimerism_allTwins_corrected %>%
  mutate(
    chimProp = case_when(
      !chimType %in% c("type20", "type21") ~ NA,
      .default = chimProp
    ),
    chimProp_c3 = case_when(
      chimType %in% c("type20", "type21") ~ NA,
      .default = chimProp_c3
      )
    ) %>%
  group_by(animalID) %>%
  summarise(
    mean.type2.c = mean(chimProp, na.rm = T),
    mean.type1.c = mean(chimProp_c3, na.rm = T)
  ) %>%
  mutate(
    diff = mean.type2.c - mean.type1.c
  )

figData_mean.type2.c_vs_mean.type1.c %>%
  ggplot(aes(x = mean.type1.c, y = diff)) +
  geom_point() +
  ylim(-0.5, 0.5)


twinChimerism_allTwins_corrected %>%
  merge(., sampleRef[, c("animalID", "sex")], by = "animalID") %>%
  distinct() %>%
  group_by(sex) %>%
  summarise(mean = mean(chimProp))
```