---
title: "chapter1_panelDev"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

This document is to outline all of the data and analyses for dissertation Chapter 1 on SNP panel development.

# 2 Packages

```{r, message=FALSE}
library(ggthemes)
library(gsubfn)
library(huxtable)
library(janitor)
library(seqinr)
library(tidyverse)

source("project_scripts/ggplot_theme_Publication-2.R") # for pretty ggplot themes
```

if there's issues w/installing seqinr, try installing packages "progress", "cpp11", and "Rcpp"

ggplot_theme_Publication-2.R is from: https://medium.com/analytics-vidhya/ggplot2-themes-for-publication-ready-plots-including-dark-themes-9cd65cc5a7e3 

# 3 Data

## 3.1 seqName_to_shortName

```{r}
# seqName to shortName file
seqNames_to_shortNames <- read.csv("./project_data/seqNames_to_shortNames.csv")
```

## 3.2 primerProbeFiles

```{r}
# laLab primerProbeFile - should contain all loci that could be rescued
pp_laLab <- read.csv("./02_laLab_run1/03_laLab_run1GTscore/primerProbeFile_fullSet.txt", header = T, sep = "\t")

pp_tamRun5 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)
```

# 4 SNP panel prep

## 4.1 Overview

In total, Sam identified 402 informative loci, including n = 219 INDID, n = 66 LWED, n = 66 SIMP, n = 20 SPECIES, and n = 31 SEXID loci. I used sequences surrounding these loci as input for primer design in BatchPrimer3, which successfully designed primers for 362 loci. 

bp3 first set designed primers for n = 329 loci
bp3 second set designed primers for n = 33 loci
TOTAL = 362
-   INDID       152 + 33 = 185
-   LWED        66
-   SIMP        66
-   SEXID       25
-   SPECIES     20

n = 302 passed in silico optimization


Link to [Tamarin Genetics Lab Notebook 2 - Primer Design](https://quip.com/O64RAGBSzS5r/Tamarin-Genetics-Lab-Notebook-2-Primer-Design)

## 4.2 Panel prep

### 1) multifasta_1

Sam sent an initial set of **n_loci = 357** (multifasta_1) to be used as input for bp3 - breakdown of counts for loci sets below:

-   INDID       174
-   LWED         66
-   SIMP         66
-   SEXID        31
-   SPECIESID    20
  
```{r}
multifasta_1 <- read.fasta(file = "./primers/concatenated_species_ind_sex_14.01.2022.fasta") %>%
  names(.) %>%
  as.data.frame() %>%
  dplyr::rename("original" = ".") %>%
  mutate(set = sub('\\_.*', '', original)) %>%
  filter(set %in% c("INDID", "LWED", "SEXID", "SIMP", "SPECIESID"))

multifasta_1 %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  rbind(c("total", nrow(multifasta_1)))
```

### 2) bp3Output_1

BatchPrimer3 was able to successfully design primers for **n = 329** out of the 357 loci from multifasta_1, including the following:

```{r}
bp3Output_1 <- read.csv("./primers/primerSet1_bp3Output_formattingUpdated_20Jan2022.csv") %>%
  mutate(set = sub('\\_.*', '', Seq.ID))

bp3Output_1 %>%
  select(Seq.ID, set) %>%
  distinct() %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  janitor::adorn_totals()
```
Loci that failed primer design (n = 28) include the following counts per lociSet:

```{r}
multifasta_1 %>%
  filter(!original %in% bp3Output_1$Seq.ID) %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  janitor::adorn_totals()
```

### 3) fastpcrOutput_1a

BatchPrimer3 was set to return three potential primer pairs per locus, which we subsequently uploaded to FastPCR to assign quality scores to each primer pair. FastPCR quality scores range from zero to 100 and are based on in silico amplification efficiency, with higher scores assigned to primers with wider ranges of executable temperatures and thus a lower likelihood of adverse effects associated with variation in reaction buffers, polymerase types, or annealing temperatures.

The output from FastPCR is below:

```{r}
fastpcrOutput_1a <- read.csv("./primers/primerSet1_fastpcrAnalysis_20Jan2022.csv")
```

### 4) fastpcrOutput_1b

Because FastPCR assigns quality scores to forward and reverse primers individually, I created a ranking system for the primer *pairs* based on the following:

1)    "qualType", in which I assigned pairs as "type1" when both forward and reverse primers had quality scores >= 75, "type2" when both forward and reverse primers had quality scores >= 50, but one or both scored < 75, and "type3" when both forward and reverse primers had quality scores < 50
2)    "avgQual", which represented the average of forward and reverse primer quality scores for each pair

I used these variables to choose primer pairs for each locus that maximized both "qualType" and "avgQual", resulting in fastpcrOutput2.

```{r}
fastpcrOutput_1b <- read.csv("./primers/primerSet2_forInSilico_20Jan2022.csv")
```

I have a set of scripts outlining how I did this originally, but the script below is way simpler and easier to follow (and gives the same results) --

```{r}
temp <- fastpcrOutput_1a %>%
  mutate(
    bp3ID = str_c(bp3Index, ".", bp3Rank)
  ) %>%
  dplyr::rename("quality" = "Primer_Quality...") %>%
  select(bp3ID, primerFR, quality) %>%
  pivot_wider(names_from = primerFR,
              values_from = quality) %>%
  separate(bp3ID, into = c("bp3Index", "bp3Rank")) %>%
  mutate(
    avgQuality = (forward + reverse)/2,
    qualType = case_when(
      forward >= 75 & reverse >= 75 ~ "type1",
      forward >= 75 & reverse >= 50 ~ "type2",
      forward >= 50 & reverse >= 75 ~ "type2",
      forward >= 50 & reverse >= 50 ~ "type2",
      .default = "type3"
    )
  ) %>%
  group_by(bp3Index) %>%
  # subset to highest qualType per bp3Index
  slice_min(order_by = factor(qualType, levels = c("type1", "type2", "type3"))) %>%
  # subset to highest avgQuality w/in highest qualType
  slice_max(order_by = avgQuality, n = 1) %>%
  distinct(bp3Index, .keep_all = T) %>%
  arrange(as.numeric(bp3Index)) %>%
  mutate(
    bp3id = str_c(bp3Index, ".", bp3Rank)
  )

fastpcrOutput_1b <- fastpcrOutput_1a %>%
  mutate(
    bp3id = str_c(bp3Index, ".", bp3Rank)
  ) %>%
  filter(bp3id %in% temp$bp3id)
```

### 5) mfeOutput_1

**the metrics for this section are based on the first round of MFE only; there were additional rounds where primers for failed loci were swapped out and tried again**

I sent fastpcrOutput_1b to Sam to run MFE in silico tests. Of the n = 329 loci, **n = 235** passed and n = 94 failed in silico tests.

```{r}
mfeOutput_1 <- read.csv("./primers/primerSet2_passFail.csv") %>%
  select(-X) %>%
  filter(!is.na(inSilico)) %>%
  distinct(bp3Index, .keep_all = T) %>%
  mutate(set = case_when(
    str_detect(Seq.ID, "INDID") ~ "INDID",
    str_detect(Seq.ID, "LWED") ~ "LWED",
    str_detect(Seq.ID, "SIMP") ~ "SIMP",
    str_detect(Seq.ID, "SEXID") ~ "SEXID",
    str_detect(Seq.ID, "SPECIESID") ~ "SPECIES",
    .default = NA
  ))

sum(mfeOutput_1$inSilico == "PASS") # 235
sum(mfeOutput_1$inSilico == "FAIL") # 94
```

Info on loci that failed in-silico PCR is in multifasta_2 section (need info from multifasta_2 to figure out sets and whatnot since MFE tests went through a few rounds - that means that while 94 loci failed here, replacement primers could have passed for some)

### 6) multifasta_2

To replace primers that failed in silico optimization, Sam identified an additional 45 INDID loci for primer design. These were combined into a multi-FASTA file with the loci that had already passed in silico PCR for a total of **n = 320** total, with the following loci-set breakdown:

-   INDID       169 (124 previous + 45 additional)
-   LWED        62
-   SIMP        53
-   SEXID       18
-   SPECIESID   18
-   total       320

**NOTE** that there were some discrepancies between Sam's numbers and mine (see quip doc for details) -- the "rvEdits" fasta file was udpated to address those issues. **HOWEVER** it looks like I did something dumb in those edits b/c SEXID_gatk...815956-816448 (which I apparently thought was missing, and therefore added) is now in there twice

```{r}
multifasta_2 <- read.fasta(file = "./primers/concatenated_ALL_19.02.2022_rvEdits.fasta") %>%
  names(.) %>%
  as.data.frame() %>%
  dplyr::rename("original" = ".") %>%
  mutate(set = sub('\\_.*', '', original)) %>%
  filter(set %in% c("INDID", "LWED", "SEXID", "SIMP", "SPECIESID"))

multifasta_2 %>%
  get_dupes(original) # SEXID_gatk...815956-816448 in there twice - REMOVE one

multifasta_2 <- multifasta_2 %>%
  filter(original != "SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_815956-816448_chosen_consensus_modified") %>%
  rbind(., c("SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_815956-816448_chosen_consensus_modified", "SEXID"))

multifasta_2 %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals()
```

That means that the loci that failed the first round in-silico PCR include the following counts per lociSet:

```{r}
# seq names from multifasta_1 that passed first round in silico (n = 275)
temp <- multifasta_2 %>%
  filter(original %in% multifasta_1$original)

temp %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  adorn_totals()

# seq names that failed first round in silico
bp3Output_1 %>% # 329 primers designed
  select(Seq.ID, set) %>%
  distinct() %>% # has entries for both f/r - ditch dups
  filter(!Seq.ID %in% temp$original) %>% # ditch those that passed in silico
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  janitor::adorn_totals()
```

### 7) bp3Output_2

I used multifasta_2 as input for BatchPrimer3 to see if we could find primers for the additional 45 loci. **BatchPrimer3 was able to design primers for 33 out of the 45 additional loci**, yielding primers for a total of **n = 308** loci.

**NOTE** that my quip doc shows bp3 output w/primers for 309 loci; this is just due to the duplicated SEXID sequence (see above)

Counts for loci sets below:

-   INDID       157
-   LWED         62
-   SEXID        18
-   SIMP         53
-   SPECIESID    18
-   total       308

```{r}
bp3Output_2 <- read.csv("./primers/primerSet3_bp3Output_originalFile_22Feb2022.csv") %>%
  mutate(set = sub('\\_.*', '', Seq.ID))

# bp3 found primers for 308 loci
bp3Output_2 %>%
  select(Seq.ID, set) %>%
  distinct() %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals()

# primers designed for 33 out of the 45 new loci
bp3Output_2 %>%
  select(Seq.ID) %>%
  distinct() %>%
  filter(str_detect(Seq.ID, "additional")) %>%
  nrow()
```

Loci that failed primer design (n = 12) for the 45 new loci include the following counts per lociSet:

```{r}
multifasta_2 %>%
  filter(!original %in% bp3Output_2$Seq.ID) %>%
  group_by(set) %>%
  summarise(count = n())
```

### 8) fastpcrOutput_2a

bp3Output_2 for the 33 new primers was input to FastPCR to assign quality scores with the following output:

```{r}
fastpcrOutput_2a <- read.csv("./primers/primerSet3_fastpcrAnalysis_22Feb2022.csv")
```

### 9) fastpcrOutput_2b

As before, I used the quality scores assigned by FastPCR to choose the highest-scoring primer *pair* for each of the new loci, resulting in the following primer choices for the 33 new loci:

```{r}
fastpcrOutput_2b <- read.csv("./primers/primerSet4_forInSilico_22Feb2022.csv")
```

### 10) mfeOutput_2

**These numbers aren't accurate; doesn't account for replacement primers and re-testing in silico - based on my investigations though, it seems like we dropped 6 here**

I sent fastpcrOutput_2b to Sam to run MFE in silico tests. Of the n = 33 new loci, **n = 12** passed and n = 21 failed in silico tests.

```{r}
mfeOutput_2 <- read.csv("./primers/primerSet4_resultsInSilico_23Feb2022.csv") %>%
  filter(!is.na(inSilico)) %>%
  distinct(bp3Index, .keep_all = T) %>%
  mutate(lociSet = case_when(
    str_detect(Seq.ID, "INDID") ~ "INDID", # (all are INDID anyway)
    .default = NA
  ))

sum(mfeOutput_2$inSilico == "PASS") # 12
sum(mfeOutput_2$inSilico == "FAIL") # 21
```

Loci that failed (n = 21) include the following counts per lociSet:

```{r}
mfeOutput_2 %>%
  filter(inSilico == "FAIL") %>%
  group_by(lociSet) %>%
  summarise(count = n())
```

### 11) bp3error

**Quick-load**

```{r}
bp3errorTracking <- read.csv("./00_tamRun1/bp3error_investigation/bp3error_master_1July2024.csv")
```

```{r}
bp3errorTracking %>%
  filter(pf_bp3error == "fail") %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  adorn_totals()
```


**figuring shit out (not cleaned up)**

Acording to Sam's notes/thesis/etc., this is how things break down:
-   started with 302 loci
-   bp3 error for 99
-   found new variants w/in amplicon for 23 (99 - 23 = 76)
-   used MSA files from tamRun1 to find variants for another 36 (76 - 36 = 40)

According to my own investigations, we have:

           outcome count
 bp3error_noRescue    48
  bp3error_rescue1    33
  bp3error_rescue2    28
           noError   193
             Total   302


```{r}
# 99 loci w/bp3 error
bp3error_99Loci <- read.csv("./00_tamRun1/bp3error_investigation/failing_snps1.csv", header = F) %>%
  merge(., seqNames_to_shortNames[, c("seqID1", "primerName2")], by.x = "V1", by.y = "seqID1", all.x = T) %>%
  dplyr::rename("seqID" = "V1",
                "locus" = "primerName2") %>%
  mutate(locus = sub('\\..*', '', locus))

# 226 loci, including some number of bp3error loci that were part of rescue1 (Sam found other informative variant in amplicon)
pp_bp3clean <- read.table("./00_tamRun1/03_run1GTscore/primerProbeFile.txt", header = T) %>%
  mutate(
    Locus = case_when(
      Locus == "SEXID_195.3" ~ "SEXID_LWED_195.3",
      Locus == "SEXID_208.3" ~ "SEXID_LWED_208.3",
      Locus == "SEXID_211.3" ~ "SEXID_LWED_211.3",
      
      Locus == "SEXID_197.1" ~ "SEXID_SIMP_197.1",
      Locus == "SEXID_198.3" ~ "SEXID_SIMP_198.3",
      Locus == "SEXID_203.2" ~ "SEXID_SIMP_203.2",
      Locus == "SEXID_218.1" ~ "SEXID_SIMP_218.1",
      
      .default = Locus
    )
  ) %>%
  mutate(Locus = sub('\\..*', '', Locus)) %>%
  mutate(
    fileSet = "set226Loci"
  )

# 76 loci w/bp3 error that couldn't be "rescued" immediately; but Sam apparently used MSA sequences to identify other variants for some number of them
pp_bp3error <- read.table("./00_tamRun1/03_run1GTscore/bp3error_primerProbeFile.txt", header = T) %>%
  mutate(
    Locus = case_when(
      Locus == "SEXID_195.3" ~ "SEXID_LWED_195.3",
      Locus == "SEXID_208.3" ~ "SEXID_LWED_208.3",
      Locus == "SEXID_211.3" ~ "SEXID_LWED_211.3",
      
      Locus == "SEXID_197.1" ~ "SEXID_SIMP_197.1",
      Locus == "SEXID_198.3" ~ "SEXID_SIMP_198.3",
      Locus == "SEXID_203.2" ~ "SEXID_SIMP_203.2",
      Locus == "SEXID_218.1" ~ "SEXID_SIMP_218.1",
      
      .default = Locus
    )
  ) %>%
  mutate(Locus = sub('\\..*', '', Locus)) %>%
  mutate(
    fileSet = "set76Loci"
  )
```

```{r}
# 35 loci w/bp3 error that could be rescued
bp3error_rescued <- read.csv("./00_tamRun1/bp3error_investigation/failing_snps_that_have_other_variants.csv", header = F) %>%
  merge(., seqNames_to_shortNames[, c("seqID1", "primerName3")], by.x = "V1", by.y = "seqID1", all.x = T) %>%
  select(-V2) %>%
  dplyr::rename("seqName" = "V1",
                "locus" = "primerName3") %>%
  mutate(
    locus = case_when(
      locus == "SEXID_195" ~ "SEXID_LWED_195",
      locus == "SEXID_208" ~ "SEXID_LWED_208",
      locus == "SEXID_211" ~ "SEXID_LWED_211",
      
      locus == "SEXID_197" ~ "SEXID_SIMP_197",
      locus == "SEXID_198" ~ "SEXID_SIMP_198",
      locus == "SEXID_203" ~ "SEXID_SIMP_203",
      locus == "SEXID_218" ~ "SEXID_SIMP_218",
      
      .default = locus
    )
  ) %>%
  mutate(
    rescueSet = case_when(
      locus %in% pp_bp3clean$Locus ~ "rescue1",
      locus %in% pp_bp3error$Locus ~ "rescue2",
      .default = "noRescue"
    )
  )
  
table(bp3error_rescued$rescueSet) 
```

**Loci/primer tracking**

```{r}
# dropped b/t panel v1 & v2
lociRemoved_v1.v2 <- tamRun1_primerProbeFile %>%
  filter(!Locus %in% pp_laLab$Locus) %>%
  select(Locus) %>%
  mutate(
    removedAfter = "opt1"
  )

# dropped b/t panel v2 & v3
lociRemoved_v2.v3 <- pp_laLab %>%
  filter(!Locus %in% pp_tamRun5$Locus) %>%
  select(Locus) %>%
  mutate(
    removedAfter = "opt2"
  )

lociRemoved <- rbind(lociRemoved_v1.v2, lociRemoved_v2.v3) %>%
  dplyr::rename("locus" = "Locus")
```

Plus associated metadata, including whether it was a bp3error locus, whether it was rescued, etc.

if loci were flagged for bp3 error, but pass pf_opt1, this means they were "rescued"; loci flagged for bp3 error but fail pf_opt1 = dropped

```{r}
snpTracking <- multifasta_1 %>%
  rbind(., multifasta_2) %>%
  distinct() %>%
  merge(., seqNames_to_shortNames[, c("seqID3", "primerName3")], by.x = "original", by.y = "seqID3", all.x = T) %>%
  dplyr::rename("seqID" = "original",
                "locus" = "primerName3") %>%
  # redo lociSet
  mutate(
    set = sub('[_][^_]+$', '', locus)
  ) %>%
  # add bp3error info
  mutate(bp3error = case_when(
      locus %in% bp3error_99Loci$locus ~ "bp3error",
      .default = NA
      )
    ) %>%
  merge(., bp3error_rescued[, c("locus", "rescueSet")], by = "locus", all.x = T) %>%
  # ID each seqID as pass/fail at each panel prep step
  mutate(
    pf_primerDesign = case_when(
      !is.na(locus) ~ "pass",
      .default = "fail"
    ),
    pf_opt1 = case_when(
      pf_primerDesign == "fail" ~ NA,
      pf_primerDesign == "pass" & locus %in% pp_laLab$Locus ~ "pass",
      .default = "fail"
    ),
    pf_bp3error = case_when(
      !is.na(bp3error) & !is.na(rescueSet) ~ "pass",
      !is.na(bp3error) & is.na(rescueSet) & pf_opt1 == "pass" ~ "pass",
      !is.na(bp3error) & is.na(rescueSet) ~ "fail",
      .default = NA
    )
  ) %>%
  merge(., lociRemoved, by = "locus", all.x = T) %>%
  mutate(
    pf_opt2 = case_when(
      is.na(pf_opt1) ~ NA,
      removedAfter == "opt1" ~ NA,
      removedAfter == "opt2" ~ "fail",
      .default = "pass"
    )
  ) %>%
  select(seqID, locus, set, pf_primerDesign, pf_bp3error, pf_opt1, pf_opt2)

table(snpTracking$pf_opt2)



snpTracking_table <- snpTracking %>%
  filter(!is.na(bp3error)) %>%
  filter(pf_bp3error == "fail")

table(snpTracking_table$pf_opt1)

table(snpTracking_table$set)
```

export bp3error loci tracking

```{r}
bp3errorTracking <- snpTracking %>%
  filter(!is.na(bp3error))

write.csv(bp3errorTracking, "./00_tamRun1/bp3error_investigation/bp3error_master_1July2024.csv", row.names = F)
```

## 4.3 Summary

asdfasdf

# 5 library_1 tamRun1

## 5.1 Overview




## 5.2 Data

### Sample metadata

**Data**

```{r}
tamRun1_md <- read.csv("./00_tamRun1/03_run1GTscore/tamRun1_metadata_v2.csv")
```

**Metrics**

```{r}
tamRun1_sampleOverview <- tamRun1_md %>%
  mutate(
    temp = str_c(sampleType, sex, sep = "_")
  ) %>%
  group_by(species, temp) %>%
  summarise(count = n()) %>%
  mutate(
    species = case_when(
      species == "pcr1Neg" ~ "negCtrl",
      .default = species
    ),
    temp = case_when(
      is.na(temp) ~ "negCtrl",
      .default = temp
    )
  ) %>%
  pivot_wider(names_from = species,
              values_from = count) %>%
  separate(temp, into = c("sampleType", "sex"), sep = "_") %>%
  rowwise() %>%
  mutate(
    TOTAL = sum(LWED,SIMP,negCtrl, na.rm = T)
  ) %>%
  janitor::adorn_totals("row") %>%
  as.data.frame()

tamRun1_sampleOverview
```
```{r}
tamRun1_md %>%
  group_by(sampleType) %>%
  summarise(count = n()) %>%
  as.data.frame()
```
```{r}
str_c(tamRun1_md %>%
  select(animalID) %>%
  filter(animalID != "pcr1Neg") %>%
  distinct() %>%
  nrow(), " unique individuals")
```
```{r}
str_c(tamRun1_md %>%
  filter(species == "LWED") %>%
  select(animalID) %>%
  filter(animalID != "pcr1Neg") %>%
  distinct() %>%
  nrow(), " unique LWED")
```
```{r}
str_c(tamRun1_md %>%
  filter(species == "SIMP") %>%
  select(animalID) %>%
  filter(animalID != "pcr1Neg") %>%
  distinct() %>%
  nrow(), " unique SIMP")
```
```{r}
# duplicate & triplicate samples
tamRun1_md %>%
  select(animalID, sampleType, sampleName) %>%
  filter(animalID != "pcr1Neg") %>%
  get_dupes(.) %>%
  distinct() %>%
  as.data.frame()
```
### Loci metadata

I've imported the primer-probe file that contains ALL loci originally in the tamRun1 panel, including those with the BatchPrimer3 error regardless of whether they were rescued. 

**Data**

```{r}
tamRun1_primerProbeFile <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/primerProbeFile_fullSet.txt", header = T)
```

**Metrics**

```{r}
tamRun1_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "SEXID_LWED") ~ "sexSNP_lwed",
      str_detect(Locus, "SEXID_SIMP") ~ "sexSNP_simp",
      str_detect(Locus, "SEXID") ~ "sexSNP_dual",
      str_detect(Locus, "INDID") ~ "indSNP_dual",
      str_detect(Locus, "LWED") ~ "indSNP_lwed",
      str_detect(Locus, "SIMP") ~ "indSNP_simp",
      str_detect(Locus, "SPECIES") ~ "spSNP_dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
```

```{r}
tamRun1_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "LWED") ~ "lwed",
      str_detect(Locus, "SIMP") ~ "simp",
      .default = "dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
```

### Sample/loci lists

```{r}
# sample lists
tamRun1_posSamples <- tamRun1_md %>%
  filter(species %in% c("LWED", "SIMP"))

tamRun1_negSamples <- tamRun1_md %>%
  filter(!sampleID %in% tamRun1_posSamples$sampleID)

tamRun1_lwedSamples <- tamRun1_md %>%
  filter(species == "LWED")

tamRun1_simpSamples <- tamRun1_md %>%
  filter(species == "SIMP")

# loci lists
tamRun1_lwedLoci <- tamRun1_primerProbeFile %>%
  filter(!str_detect(Locus, "SIMP"))

tamRun1_simpLoci <- tamRun1_primerProbeFile %>%
  filter(!str_detect(Locus, "LWED"))
```

### Genos

```{r}
tamRun1_gtscore0x_genos <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

tamRun1_gtscore10x_genos <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")
```

### Summaries

**metrics reflect non-negative, target-species only**

```{r}
tamRun1_gtscore_locusSum <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/summaryFiles/master_locusSummary.csv")

tamRun1_gtscore_sampleSum <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/summaryFiles/master_sampleSummary.csv")
```

**metrics include other combos - fullset samples/loci, etc.**

```{r}
tamRun1_gtscore_locusSum_lwed_simpSpec <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/LWED_simpSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples") %>%
  relocate(sampleSet, .after = Locus)

tamRun1_gtscore_locusSum_simp_lwedSpec <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/SIMP_lwedSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples") %>%
  relocate(sampleSet, .after = Locus)

tamRun1_gtscore_sampleSum_fullSet <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/fullSet_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., tamRun1_md[, c("sampleID", "species")], by.x = "Sample", by.y = "sampleID") %>%
  relocate(species, .after = Sample)
```

## 5.3 Read metrics

### bySample

**all samples + sp-specific loci sets**

```{r}
tamRun1_gtscore_sampleSum %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**pos only (no negs) + sp-specific loci sets**

```{r}
tamRun1_gtscore_sampleSum %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**pos only (no negs) + all loci**

```{r}
tamRun1_gtscore_sampleSum_lociAll %>%
  filter(!str_detect(species, "Neg")) %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(otProp, otReads, primerProp, primerReads, totalReads)
```

### byLocus

#### Overview

```{r}
tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  mutate(
    lociSet = case_when(
      sampleSet == "lwedSamples" ~ "lwedSpec",
      sampleSet == "simpSamples" ~ "simpSpec",
      sampleSet == "posSamples" ~ "dual",
      .default = sampleSet
    )
  ) %>%
  relocate(sampleSet, lociSet, otProp, otReads, primerReads)
```

#### dualLoci

lwed vs. simp performance

```{r}
tamRun1_gtscore_locusSum %>%
  filter(sampleSet != "posSamples") %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP")) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### lwedLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun1_gtscore_locusSum %>%
  filter(str_detect(Locus, "LWED")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun1_gtscore_locusSum_simp_lwedSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("lwedSamples", "lwedSamples*", sampleSet)
  )
```

#### simpLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun1_gtscore_locusSum %>%
  filter(str_detect(Locus, "SIMP")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun1_gtscore_locusSum_lwed_simpSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("simpSamples", "simpSamples*", sampleSet)
  )
```

## 5.4 Geno success

### bySample

```{r}
temp1 <- tamRun1_gtscore_sampleSum %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "fullSet")

temp2 <- tamRun1_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "blood")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "blood")

temp3 <- tamRun1_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "hair")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "hair")

rbind(temp1, temp2, temp3) %>%
  relocate(sampleSet, n)
```

### byLocus

#### Overview

```{r}
tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "lwedSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "simpSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )
```

#### dualLoci



#### lwedLoci

**data**

```{r}
tamRun1_genoSuccess_gtscore0x_lwedSpec <- tamRun1_gtscore0x_genos %>%
  select(tamRun1_posSamples$sampleID) %>%
  filter(str_detect(rownames(.), "LWED")) %>%
  mutate(across(everything(), \(x) as.character(x))) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  mutate(
    
    totalGenos_lwed = rowSums(!is.na(select(., tamRun1_lwedSamples$sampleID))),
    totalGenos_simp = rowSums(!is.na(select(., tamRun1_simpSamples$sampleID))),
    
    propSuccess_lwed = round(totalGenos_lwed/nrow(tamRun1_lwedSamples), 2),
    propSuccess_simp = round(totalGenos_simp/nrow(tamRun1_simpSamples), 2)
    
  ) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    
    uniqueGenos_lwed = n_distinct(c_across(tamRun1_lwedSamples$sampleID), na.rm = T),
    uniqueGenos_simp = n_distinct(c_across(tamRun1_simpSamples$sampleID), na.rm = T)
    
  ) %>%
  relocate(contains("tamOpt"), .after = last_col()) %>%
  relocate(contains("uniqueGenos"), .after = locus)
```

**metrics**

```{r}
mean(tamRun1_genoSuccess_gtscore0x_lwedSpec$propSuccess_lwed)
```
```{r}
mean(tamRun1_genoSuccess_gtscore0x_lwedSpec$propSuccess_simp)
```

#### simpLoci

```{r}
tamRun1_genoSuccess_gtscore0x_simpSpec <- tamRun1_gtscore0x_genos %>%
  select(tamRun1_posSamples$sampleID) %>%
  filter(str_detect(rownames(.), "SIMP")) %>%
  mutate(across(everything(), \(x) as.character(x))) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  mutate(
    
    totalGenos_lwed = rowSums(!is.na(select(., tamRun1_lwedSamples$sampleID))),
    totalGenos_simp = rowSums(!is.na(select(., tamRun1_simpSamples$sampleID))),
    
    propSuccess_lwed = round(totalGenos_lwed/nrow(tamRun1_lwedSamples), 2),
    propSuccess_simp = round(totalGenos_simp/nrow(tamRun1_simpSamples), 2)
    
  ) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    
    uniqueGenos_lwed = n_distinct(c_across(tamRun1_lwedSamples$sampleID), na.rm = T),
    uniqueGenos_simp = n_distinct(c_across(tamRun1_simpSamples$sampleID), na.rm = T)
    
  ) %>%
  relocate(contains("tamOpt"), .after = last_col()) %>%
  relocate(contains("uniqueGenos"), .after = locus)
```

**metrics**

```{r}
mean(tamRun1_genoSuccess_gtscore0x_simpSpec$propSuccess_lwed)
```

```{r}
mean(tamRun1_genoSuccess_gtscore0x_simpSpec$propSuccess_simp)
```

## 5.5 Panel optimization summary

Removed...

-   41 loci due to bp3 error
-   9 variants not producing genotypes (based on GTseq analysis)



```{r}
tamRun1_lociFail_bp3error <- bp3errorTracking %>%
  filter(pf_bp3error == "fail") %>%
  select(locus)

tamRun1_lociFail_zeroReads <- data.frame(
  locus = c("INDID_23",
            "INDID_51",
            "INDID_58",
            "INDID_96",
            "LWED_287",
            "SPECIES_3",
            "SPECIES_6",
            "SPECIES_7",
            "SPECIES_10")
)

rbind(tamRun1_lociFail_zeroReads, tamRun1_lociFail_bp3error) %>%
  nrow()

View(tamRun1_gtscore_locusSum %>%
       arrange(GenotypeRate_0x))

temp <- tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  select(Locus, sampleSet, AvgReadDepth, GenotypeRate_0x, GenotypeRate_10x, allFreqs_0x, allFreqs_10x) %>%
  dplyr::rename("locus" = "Locus") %>%
  mutate(
    AvgReadDepth = round(AvgReadDepth, 2),
    GenotypeRate_0x = round(GenotypeRate_0x, 2),
    GenotypeRate_10x = round(GenotypeRate_10x, 2)
  ) %>%
  merge(., lociRemoved, by = "locus", all.x = T)
```

# 6 library_2 - tamRun3 metrics

## 6.1 Overview

asdf

## 6.2 Data

### Sample metadata

this is md_tamRun3_v3; see sandbox GTscore_tamRun5_plus_tamRun3.Rmd for details on samples vs. metadata

**Data**

```{r}
tamRun3_md <- read.csv("./03_tamRun3/03_run3GTscore/tamRun3_metadata_abcat_5June2023.csv") %>%
  filter(!str_detect(sampleID, "cat")) %>%
  filter(!str_detect(sampleID, "3b"))
```

**Metrics**

```{r}
tamRun3_sampleOverview <- tamRun3_md %>%
  mutate(
    temp = str_c(sampleType, sex, sep = "_")
  ) %>%
  group_by(species, temp) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = species,
              values_from = count) %>%
  separate(temp, into = c("sampleType", "sex"), sep = "_") %>%
  rowwise() %>%
  mutate(
    TOTAL = sum(LWED,SIMP,pcr1Neg,xtnNeg, na.rm = T)
  ) %>%
  janitor::adorn_totals("row") %>%
  as.data.frame()

tamRun3_sampleOverview
```

```{r}
tamRun3_md %>%
  group_by(sampleType) %>%
  summarise(count = n()) %>%
  as.data.frame()
```

```{r}
str_c(tamRun3_md %>%
  select(animalID) %>%
  filter(!str_detect(animalID, "Neg")) %>%
  distinct() %>%
  nrow(), " unique individuals")
```
```{r}
str_c(tamRun3_md %>%
  filter(species == "LWED") %>%
  select(animalID) %>%
  filter(!str_detect(animalID, "Neg")) %>%
  distinct() %>%
  nrow(), " unique LWED")
```
```{r}
str_c(tamRun3_md %>%
  filter(species == "SIMP") %>%
  select(animalID) %>%
  filter(!str_detect(animalID, "Neg")) %>%
  distinct() %>%
  nrow(), " unique SIMP")
```
### Loci metadata

**Data**

```{r}
tamRun3_primerProbeFile <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt", header = T)
```

**Metrics**

```{r}
tamRun3_lociOverview <- tamRun3_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "SEXID_LWED") ~ "sexSNP_lwed",
      str_detect(Locus, "SEXID_SIMP") ~ "sexSNP_simp",
      str_detect(Locus, "SEXID") ~ "sexSNP_dual",
      str_detect(Locus, "INDID") ~ "indSNP_dual",
      str_detect(Locus, "LWED") ~ "indSNP_lwed",
      str_detect(Locus, "SIMP") ~ "indSNP_simp",
      str_detect(Locus, "SPECIES") ~ "spSNP_dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
  
tamRun3_lociOverview
```

### Sample/loci lists

```{r}
# sample lists
tamRun3_posSamples <- tamRun3_md %>%
  filter(species %in% c("LWED", "SIMP"))

tamRun3_negSamples <- tamRun3_md %>%
  filter(!sampleID %in% tamRun3_posSamples$sampleID)

tamRun3_lwedSamples <- tamRun3_md %>%
  filter(species == "LWED")

tamRun3_simpSamples <- tamRun3_md %>%
  filter(species == "SIMP")

# loci lists
tamRun3_lwedLoci <- tamRun3_primerProbeFile %>%
  filter(!str_detect(Locus, "SIMP"))

tamRun3_simpLoci <- tamRun3_primerProbeFile %>%
  filter(!str_detect(Locus, "LWED"))
```

### Genos

using tamRun3_a

```{r}
tamRun3_gtscore0x_genos <- read.table("./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

tamRun3_gtscore10x_genos <- read.table("./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")
```

### Summaries

**metrics reflect non-negative, target-species only**

```{r}
tamRun3_gtscore_locusSum <- read.csv("./03_tamRun3/03_run3GTscore/summaryFiles/master_locusSummary_abcat.csv") %>%
  filter(str_detect(sampleSet, "_a"))

tamRun3_gtscore_sampleSum <- read.csv("./03_tamRun3/03_run3GTscore/summaryFiles/master_sampleSummary_abcat.csv") %>%
  filter(!str_detect(sampleID, "3b")) %>%
  filter(!str_detect(sampleID, "cat_"))
```

**metrics include other combos - fullset samples/loci, etc.**

```{r}
tamRun3_gtscore_locusSum_lwed_simpSpec <- read.csv("./03_tamRun3/03_run3GTscore/LWED_simpSpec_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples") %>%
  relocate(sampleSet, .after = Locus)

tamRun3_gtscore_locusSum_simp_lwedSpec <- read.csv("./03_tamRun3/03_run3GTscore/SIMP_lwedSpec_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples") %>%
  relocate(sampleSet, .after = Locus)

tamRun3_gtscore_sampleSum_fullSet <- read.csv("./03_tamRun3/03_run3GTscore/fullSet_a_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., tamRun3_md[, c("sampleID", "species")], by.x = "Sample", by.y = "sampleID") %>%
  relocate(species, .after = Sample)
```

## 6.3 Read metrics

### bySample

**with negs**

```{r}
tamRun3_gtscore_sampleSum %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**without negs**

```{r}
tamRun3_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

### byLocus

#### Overview

```{r}
tamRun3_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### dualLoci

lwed vs. simp vs. both performance

```{r}
tamRun3_gtscore_locusSum %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP")) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### lwedLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun3_gtscore_locusSum %>%
  filter(str_detect(Locus, "LWED")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun3_gtscore_locusSum_simp_lwedSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("lwedSamples_a", "lwedSamples*", sampleSet)
  )
```

#### simpLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun3_gtscore_locusSum %>%
  filter(str_detect(Locus, "SIMP")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun3_gtscore_locusSum_lwed_simpSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("simpSamples_a", "simpSamples*", sampleSet)
  )
```

## 6.4 Geno success

### bySample

```{r}
# all samples
temp1 <- tamRun3_gtscore_sampleSum %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "fullSet")

temp2 <- tamRun3_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "blood")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "blood")

temp3 <- tamRun3_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "hair")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "hair")

rbind(temp1, temp2, temp3) %>%
  relocate(sampleSet, n)
```

### byLocus

```{r}
tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "lwedSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "simpSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )
```



# 7 library_3 tamRun4

## 7.1 Overview



## 7.2 Data

### Sample metadata

**Data**

```{r}
tamRun4_md <- read.csv("./04_tamRun4/tamRun4_metadata_illONT.csv")
```

**Metrics**

```{r}
tamRun4_sampleOverview <- tamRun4_md %>%
  mutate(
    temp = str_c(sampleType, sex, sep = "_")
  ) %>%
  mutate(
    species = case_when(
      str_detect(sampleType, "Neg") ~ "negCtrl",
      .default = species
    ),
    temp = case_when(
      str_detect(sampleType, "Neg") ~ "negCtrl",
      .default = temp
    )
  ) %>%
  group_by(species, temp) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = species,
              values_from = count) %>%
  separate(temp, into = c("sampleType", "sex"), sep = "_") %>%
  rowwise() %>%
  mutate(
    TOTAL = sum(LWED,SIMP,negCtrl, na.rm = T)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals("row")

tamRun4_sampleOverview
```
### Loci metadata

**Data**

```{r}
tamRun4_primerProbeFile <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)
```

**Metrics**

```{r}
tamRun4_lociOverview <- tamRun4_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "SEXID_LWED") ~ "sexSNP_lwed",
      str_detect(Locus, "SEXID_SIMP") ~ "sexSNP_simp",
      str_detect(Locus, "SEXID") ~ "sexSNP_dual",
      str_detect(Locus, "INDID") ~ "indSNP_dual",
      str_detect(Locus, "LWED") ~ "indSNP_lwed",
      str_detect(Locus, "SIMP") ~ "indSNP_simp",
      str_detect(Locus, "SPECIES") ~ "spSNP_dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
  
tamRun4_lociOverview
```

### Sample/loci lists

```{r}
# sample lists
tamRun4_posSamples <- tamRun4_md %>%
  filter(species %in% c("LWED", "SIMP"))

tamRun4_negSamples <- tamRun4_md %>%
  filter(!sampleID_ill %in% tamRun4_posSamples$sampleID_ill)

tamRun4_lwedSamples <- tamRun4_md %>%
  filter(species == "LWED")

tamRun4_simpSamples <- tamRun4_md %>%
  filter(species == "SIMP")

# loci lists
tamRun4_lwedLoci <- tamRun4_primerProbeFile %>%
  filter(!str_detect(Locus, "SIMP"))

tamRun4_simpLoci <- tamRun4_primerProbeFile %>%
  filter(!str_detect(Locus, "LWED"))
```

### Allele reads

```{r}
# illumina
tamRun4_readCounts_ill_gtscore <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

# nanopore
tamRun4_readCounts_ont_gtscore <- read.delim("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")
```

### Genos

```{r, warning=FALSE}
# illumina
tamRun4_genos_ill_gtscore0x <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

tamRun4_genos_ill_gtscore10x <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

tamRun4_genos_ill_gtseq <- 
  # need read_csv vs. read.csv here
  read_csv("./04_tamRun4/00_illumina/04_run4GTseq/genos/ill_tamRun4_compiledGenos_gtseq.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  select(-Raw.Reads, -On.Target.Reads, -X.On.Target, -X.GT, -IFI) %>%
  column_to_rownames("Sample") %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), ~ str_c(str_sub(., 1, 1), ",", str_sub(., 2, 2)))) %>%
  # recode 0,0 as NA
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))

# nanopore
tamRun4_genos_ont_gtscore0x <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

tamRun4_genos_ont_gtscore10x <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")
```

### Summaries

**metrics reflect non-negative, target-species only**

```{r}
# illumina
tamRun4_ill_gtscore_locusSum <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/summaryFiles/ill_master_locusSummary.csv")

tamRun4_ill_gtscore_sampleSum <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/summaryFiles/ill_master_sampleSummary.csv")

# nanopore
tamRun4_ont_gtscore_locusSum <- read.csv("./04_tamRun4/01_nanopore/03_run4GTscore/summaryFiles/ont_master_locusSummary.csv")

tamRun4_ont_gtscore_sampleSum <- read.csv("./04_tamRun4/01_nanopore/03_run4GTscore/summaryFiles/ont_master_sampleSummary.csv")
```

**metrics include other combos - fullset samples/loci, etc.**

```{r}
# illumina
tamRun4_ill_gtscore_locusSum_lwed_simpSpec <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_lwed_simpSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples") %>%
  relocate(sampleSet, .after = Locus)

tamRun4_ill_gtscore_locusSum_simp_lwedSpec <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_simp_lwedSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples") %>%
  relocate(sampleSet, .after = Locus)

tamRun4_ill_gtscore_locusSum_fullSet <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_GTscore_locusSummary.txt", sep = "\t")

tamRun4_ill_gtscore_sampleSum_fullSet <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_GTscore_individualSummary.txt", sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "species")], by.x = "Sample", by.y = "sampleID_ill") %>%
  relocate(species, .after = Sample)

# nanopore
```

## 7.3 Read metrics

### bySample

#### lociSets

**all samples + sp-specific loci sets**

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**pos only (no negs) + sp-specific loci sets**

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```
**pos only (no negs) + all loci**

```{r}
tamRun4_ill_gtscore_sampleSum_lociAll %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(otProp, otReads, primerProp, primerReads, totalReads)
```

#### sampleTypes

**performance by sampleType only**

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(sampleType) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(sampleType, otProp, otReads, primerProp, primerReads, totalReads)
```

**performance by sampleType + species sets**

```{r, message=FALSE}
tamRun4_ill_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(sampleType, lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(sampleType, lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

### byLocus

#### Overview

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### dualLoci

lwed vs. simp vs. both performance

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP")) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### lwedLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(str_detect(Locus, "LWED")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun4_ill_gtscore_locusSum_simp_lwedSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("lwedSamples", "lwedSamples*", sampleSet)
  )
```

#### simpLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(str_detect(Locus, "SIMP")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun4_ill_gtscore_locusSum_lwed_simpSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("simpSamples", "simpSamples*", sampleSet)
  )
```

## 7.4 Geno success

### overall

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  summarise(mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
            mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2))
```

### by_sampleType

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  group_by(sampleType) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2)
  )
```

### figures

make a figure like Burgess et al., 2022 fig. 5a/b

```{r}
# gtscore0x
tamRun4_ill_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  ggplot(aes(x = sampleType, y = GenotypeRate_0x, shape = sampleType, color = sampleType)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(0.2)) +
  scale_shape_manual(values = c(1, 17, 19)) +
  scale_colour_Publication() +
  theme_Publication()
```

```{r}
# gtscore10x
tamRun4_ill_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  ggplot(aes(x = sampleType, y = GenotypeRate_10x, shape = sampleType, color = sampleType)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(0.2)) +
  scale_shape_manual(values = c(1, 17, 19)) +
  scale_colour_Publication() +
  theme_Publication()
```

# 8 GTseq vs. GTscore

## 8.1 Data

### Allele reads

```{r}
# gtscore
tamRun4_readCounts_ill_gtscore_l <- tamRun4_readCounts_ill_gtscore %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID_ill",
               values_to = "readCounts_gr") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill") %>%
  arrange(sampleID_ill, locus)

# gtseq
tamRun4_readCounts_ill_gtseq <- read.csv("./04_tamRun4/00_illumina/04_run4GTseq/gtseq_readCounts.csv") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  dplyr::rename("sampleID_ill" = "Sample")

tamRun4_readCounts_ill_gtseq_l <- tamRun4_readCounts_ill_gtseq %>%
  select(-c(Raw.Reads, On.Target.Reads, X.On.Target, X.GT, IFI)) %>%
  pivot_longer(!sampleID_ill,
               names_to = "locus",
               values_to = "readCounts_gs") %>%
  mutate(readCounts_gs = gsub("/", ",", readCounts_gs)) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill") %>%
  arrange(sampleID_ill, locus)
```

### Genos

```{r}
# gtscore
tamRun4_ill_gtscore0x_genos_l <- tamRun4_genos_ill_gtscore0x %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), na_if, "0")) %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID_ill",
               values_to = "genos_gr") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill") %>%
  arrange(sampleID_ill, locus)

# gtseq
tamRun4_genos_ill_gtseq <- read.csv("./04_tamRun4/00_illumina/04_run4GTseq/gtseq_compiledGenos.csv") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  dplyr::rename("sampleID_ill" = "Sample")

tamRun4_genos_ill_gtseq_l <- tamRun4_genos_ill_gtseq %>%
  select(-c(Raw.Reads, On.Target.Reads, X.On.Target, X.GT, IFI)) %>%
  pivot_longer(!sampleID_ill,
               names_to = "locus",
               values_to = "genos_gs") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill") %>%
  arrange(sampleID_ill, locus)
```

## 8.2 General metrics

```{r}
# total reads
sum(tamRun4_ill_gtscore_sampleSum_fullSet$Total.Reads) # 31663938
sum(tamRun4_readCounts_ill_gtseq$Raw.Reads) # 31663938

# total on-target reads
sum(tamRun4_ill_gtscore_sampleSum_fullSet$Primer.Probe.Reads) # 10727250
sum(tamRun4_readCounts_ill_gtseq$On.Target.Reads) # 10786814

seqScore_otReads <- tamRun4_ill_gtscore_sampleSum_fullSet %>%
  select(Sample, Primer.Probe.Reads) %>%
  merge(., tamRun4_readCounts_ill_gtseq[, c("sampleID_ill", "On.Target.Reads")], by.x = "Sample", by.y = "sampleID_ill") %>%
  dplyr::rename("otReads_gr" = "Primer.Probe.Reads",
                "otReads_gs" = "On.Target.Reads") %>%
  mutate(
    diff_gr.gs = otReads_gr - otReads_gs,
    diffProp_gr.gs = diff_gr.gs/otReads_gr
  )

mean(seqScore_otReads$diffProp_gr.gs)*100
min(seqScore_otReads$diffProp_gr.gs)*100
max(seqScore_otReads$diffProp_gr.gs)*100
```

## 8.3 Geno success


# 9 Geno concordance

## 9.1 Baseline genos

First want to get some baseline genotypes to compare against - to do this, going to use genos from hair samples that match across runs, specifically runs 3 and 5

#### data

```{r}
# tamRun5 data
tamRun5_md <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv")

tamRun5_hairSamples <- tamRun5_md %>%
  filter(sampleType == "hair") %>%
  filter(animalID %in% tamRun4_md$animalID)

tamRun5_readCounts_hair <- read.delim("./05_tamRun5/03_run5GTscore/fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  select(tamRun5_hairSamples$sampleID) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_run5.gr") %>%
  merge(., tamRun5_md[, c("sampleID", "animalID")], by = "sampleID")

tamRun5_genos0x_hair <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  select(tamRun5_hairSamples$sampleID) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_run5.gr") %>%
  merge(., tamRun5_md[, c("sampleID", "animalID")], by = "sampleID")
  

# tamRun4 data
## gtscore
tamRun4_hairSamples <- tamRun4_md %>%
  filter(sampleType == "hair")

tamRun4_ill_gtscore_readCounts_hair <- tamRun4_readCounts_ill_gtscore %>%
  select(tamRun4_hairSamples$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID_ill",
               values_to = "readCounts_run4.gr") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill")

tamRun4_ill_gtscore_genos0x_hair <- tamRun4_genos_ill_gtscore0x %>%
  select(tamRun4_hairSamples$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID_ill",
               values_to = "genos_run4.gr") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill")
```

#### functions

```{r}
get_readSum <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))
```


#### compare genos

```{r}
genoCompare_baseGenos <- tamRun4_ill_gtscore_genos0x_hair %>%
  merge(., tamRun5_genos0x_hair, by = c("animalID", "locus"), all.x = T) %>%
  #mutate(across(everything(), na_if, "0")) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  mutate(
    genoMatch = genos_run4.gr == genos_run5.gr
  ) %>%
  merge(., tamRun4_ill_gtscore_readCounts_hair, by = c("animalID", "locus"), all.x = T) %>%
  merge(., tamRun5_readCounts_hair, by = c("animalID", "locus"), all.x = T) %>%
  select(!starts_with("sampleID")) %>%
  merge(., tamRun4_primerProbeFile[, c("Locus", "Allele1", "Allele2")], by.x = "locus", by.y = "Locus") %>%
  relocate(animalID, locus, Allele1, Allele2) %>%
  mutate(
    readSum_run4.gr = get_readSum(readCounts_run4.gr),
    readSum_run5.gr = get_readSum(readCounts_run5.gr)
  )

ggplot(genoCompare_baseGenos, aes(x = as.numeric(readSum_run4.gr), y = as.numeric(readSum_run5.gr), color = genoMatch)) +
  geom_point(alpha = 0.5) +
  theme_Publication()




temp <- genoCompare_baseGenos %>%
  filter(readSum_run4.gr >= 30) %>%
  filter(readSum_run5.gr >= 30) %>%
  group_by(animalID) %>%
  summarise(
    genoMismatch = sum(genoMatch == FALSE, na.rm = T),
    genoMatch = sum(genoMatch == TRUE, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(
    genoTotal = genoMismatch + genoMatch,
    propMismatch = genoMismatch/genoTotal
  )

temp2 <- genoCompare_baseGenos %>%
  filter(readSum_run4.gr >= 30) %>%
  filter(readSum_run5.gr >= 30) %>%
  group_by(locus) %>%
  summarise(
    genoMismatch = sum(genoMatch == FALSE, na.rm = T),
    genoMatch = sum(genoMatch == TRUE, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(
    genoTotal = genoMismatch + genoMatch,
    propMismatch = genoMismatch/genoTotal
  )
```


**START HERE** for the code above, adjust so that 1) readCount cols come after locus, 2) add the possible alleles for each locus so that 3) can produce gtseq genos (will need to write a small function for this to turn gtseq ratios into genotypes)






# XXXXXXXXXXXXXXXXXXXXXX
## Functions

### GTseq genos

There are a few different versions of genotyping scripts for GTseq, with each basing their genotype assignments on slightly different allele ratios. I've outlined the details of these discrepancies below and have created separate functions for each (easier than running the full GTseq script).

GTseq_Genotyper versions:

-   specs_v2 = GitHub/GTseq-Pipeline/GTseq_Genotyper_v2.pl
-   specs_v3a = GitHub/GTseq-Pipeline/GTseq_Genotyper_v3.pl
-   specs_3b = GitHub/GTseek_utils/GTseq_Genotyper_v3.pl
-   specs_3c = (my files)/GTseek_Git/GTseq_Genotyper_v3.pl

    -   Nate Campbell sent this version to me via email
    
**NOTE** that versions 2, 3a, and 3b all use **allele1Counts/allele2Counts** to assign genotypes, whereas version 3c uses allele2 percentage to assign genotypes via **allele2Counts/(allele1Counts + allele2Counts)**

| callType | specs_v2  | specs_v3a | specs_v3b | specs_v3b   |
| -------- | --------- | --------- | --------- | ----------- |
| A1HOM    | [10, inf) | [10, inf) | [10, inf) | [0%, 10%]   |
| A2HOM    | [0, 0.1]  | [0, 0.1]  | [0, 0.1]  | [90%, 100%] |
| HET      | (0.2, 5]  | (0.2, 2]  | (0.5, 2]  | [30%, 70%]  |


**function to get allele ratios for gtseq geno assignments**

```{r}
get_gtseqRatio <- function(readCounts) {
  # Function to process a single read count
  process_read_count <- function(rc) {
    # Check if the read count is NA
    if (is.na(rc)) {
      return(NA)
    }
    
    # Split the input string into two numeric values
    counts <- as.numeric(unlist(strsplit(rc, ",")))
    
    # Replace 0 counts with 0.1
    counts[counts == 0] <- 0.1
    
    # Sum the read counts
    readSum <- sum(counts)
    
    # Check the sum and compute the ratio if appropriate
    if (readSum < 10) {
      return(NA)
    } else {
      return(counts[1] / counts[2])
    }
  }
  
  # Apply the processing function to each element in readCounts
  sapply(readCounts, process_read_count)
}
```

**function to assign gtseq genos**

NOTE that the different versions of the GTseq_Genotyper script assign HET using different allele ratios - I've submitted an issue to the GitHub page, but in the meantime, the differences are outlined below and I've created separate genotyping functions to reflect the ratios in each script.

specs_v2 - GTseq-Pipeline/GTseq_Genotyper_v2.pl (lines 120-124)
* (ratio >= 10) >> A1HOM
* (ratio <= 0.1) >> A2HOM
* (ratio <= 0.2) >> NA
* (ratio <= 5) >> HET

specs_v3a - GTseq-Pipeline/GTseq_Genotyper_v3.pl (lines 201-210)
* (ratio >= 10) >> A1HOM
* (ratio < 10) &&  (ratio > 5) >> NA
* (ratio <= 0.1) >> A2HOM
* (ratio <= 0.2) >> NA
* (ratio <= 2) >> HET

specs_v3b - GTseek_utils/GTseq_Genotyper_v3.pl (lines 201-210)
* (ratio >= 10) >> A1HOM
* (ratio < 10) &&  (ratio > 5) >> NA
* (ratio <= 0.1) >> A2HOM
* (ratio <= 0.5) >> NA
* (ratio <= 2) >> HET

| callType | specs_v2  | specs_v3a | specs_v3b |
| -------- | --------- | --------- | --------- |
| A1HOM    | [10, inf) | [10, inf) | [10, inf) |
| A2HOM    | [0, 0.1]  | [0, 0.1]  | [0, 0.1]  |
| HET      | (0.2, 5]  | (0.2, 2]  | (0.5, 2]  |

```{r}
## based on GTseq_Genotyper_v2.pl
get_gtseqGenos_v2 <- function(ratio, Allele1, Allele2) {
  
  # round ratio to 3 places
  ratio <- round(ratio, 3)
  
  result <- case_when(
    is.na(ratio) ~ NA_character_,
    ratio >= 10 ~ str_c(Allele1, Allele1, sep = ","),
    ratio <= 0.1 ~ str_c(Allele2, Allele2, sep = ","),
    ratio <= 0.2 ~ NA_character_,
    ratio <= 5 ~ str_c(Allele1, Allele2, sep = ","),
    TRUE ~ NA_character_  # Default case
  )
  return(result)
}

## based on GTseq-Pipeline/GTseq_Genotyper_v3.pl (the one I used for running gtseq pipeline)
get_gtseqGenos_v3a <- function(ratio, Allele1, Allele2) {
  
  # round ratio to 3 places
  ratio <- round(ratio, 3)
  
  result <- case_when(
    is.na(ratio) ~ NA_character_,
    ratio >= 10 ~ str_c(Allele1, Allele1, sep = ","),
    (ratio < 10 & ratio > 5) ~ NA_character_,
    ratio <= 0.1 ~ str_c(Allele2, Allele2, sep = ","),
    ratio <= 0.2 ~ NA_character_,
    ratio <= 2 ~ str_c(Allele1, Allele2, sep = ","),
    TRUE ~ NA_character_  # Default case
  )
  return(result)
}

## based on GTseek_utils/GTseq_Genotyper_v3.pl
get_gtseqGenos_v3b <- function(ratio, Allele1, Allele2) {
  
  # round ratio to 3 places
  ratio <- round(ratio, 3)
  
  result <- case_when(
    is.na(ratio) ~ NA_character_,
    ratio >= 10 ~ str_c(Allele1, Allele1, sep = ","),
    (ratio < 10 & ratio > 5) ~ NA_character_,
    ratio <= 0.1 ~ str_c(Allele2, Allele2, sep = ","),
    ratio <= 0.5 ~ NA_character_,
    ratio <= 2 ~ str_c(Allele1, Allele2, sep = ","),
    TRUE ~ NA_character_  # Default case
  )
  return(result)
}

## based on GTseq_Genotyper_v3.pl emailed to me by Nate on 18-July-2024
get_gtseqGenos_v3c <- function(a1Counts, a2Counts, Allele1, Allele2) {
  
  # calculate allele2 percentage
  per_a2 <- case_when(
    as.numeric(a1Counts) + as.numeric(a2Counts) >= 10 ~ (as.numeric(a2Counts)/(as.numeric(a1Counts) + as.numeric(a2Counts)))*100,
    .default = NA
  )
  
  result <- case_when(
    is.na(per_a2) ~ NA_character_,
    per_a2 <= 10 ~ str_c(Allele1, Allele1, sep = ","),
    per_a2 >= 90 ~ str_c(Allele2, Allele2, sep = ","),
    (per_a2 >=30 & per_a2 <= 70) ~ str_c(Allele1, Allele2, sep = ","),
    .default = NA_character_
  )
  return(result)
}
```

**function to assign geno call type**

```{r}
get_callType_gr <- function(a1Counts, a2Counts, geno, Allele1, Allele2) {
  
  temp <- case_when(
    as.numeric(a1Counts) + as.numeric(a2Counts) == 0 ~ "lowReads",
    .default = NA
  )
  
  result <- case_when(
    geno == str_c(Allele1, Allele1, sep = ",") ~ "a1hom",
    geno == str_c(Allele2, Allele2, sep = ",") ~ "a2hom",
    geno == str_c(Allele1, Allele2, sep = ",") ~ "het",
    temp == "lowReads" ~ "lowReads",
    .default = "noCall"
  )
}

get_callType_gs <- function(a1Counts, a2Counts, geno, Allele1, Allele2) {
  
  temp <- case_when(
    is.na(a1Counts) ~ "lowReads",
    as.numeric(a1Counts) + as.numeric(a2Counts) < 10 ~ "lowReads",
    .default = NA
  )
  
  result <- case_when(
    geno == str_c(Allele1, Allele1, sep = ",") ~ "a1hom",
    geno == str_c(Allele2, Allele2, sep = ",") ~ "a2hom",
    geno == str_c(Allele1, Allele2, sep = ",") ~ "het",
    temp == "lowReads" ~ "lowReads",
    .default = "noCall"
  )
}
```

## Compare

```{r}
compare_gtseq.score <- 
  # allele reads
  tamRun4_ill_gtscore_alleleReads_l %>%
  merge(., tamRun4_ill_gtseq_alleleReads_l, by = c("sampleID_ill", "animalID", "locus")) %>%
  
  separate(readCounts_gr, into = c("a1Counts_gr", "a2Counts_gr"), sep = ",", remove = F) %>%
  separate(readCounts_gs, into = c("a1Counts_gs", "a2Counts_gs"), sep = ",", remove = F) %>%
  
  # allele key
  merge(., tamRun4_primerProbeFile[, c("Locus", "Allele1", "Allele2")], by.x = "locus", by.y = "Locus") %>%
  
  # allele ratios for gtscore & gtseq allele reads
  mutate(
    ratio_gr = get_gtseqRatio(readCounts_gr),
    ratio_gs = get_gtseqRatio(readCounts_gs)
  ) %>%
  
  # genos assigned w/gtscore & gtseq pipelines
  merge(., tamRun4_ill_gtscore0x_genos_l, by = c("sampleID_ill", "animalID", "locus")) %>%
  merge(., tamRun4_ill_gtseq_genos_l, by = c("sampleID_ill", "animalID", "locus")) %>%
  
  # genos assigned w/my version of gtseq functions
  mutate(
    genos_v2 = get_gtseqGenos_v2(ratio_gs, Allele1, Allele2),
    genos_v3a = get_gtseqGenos_v3a(ratio_gs, Allele1, Allele2),
    genos_v3b = get_gtseqGenos_v3b(ratio_gs, Allele1, Allele2),
    genos_v3c = get_gtseqGenos_v3c(a1Counts_gs, a2Counts_gs, Allele1, Allele2)
  ) %>%
  
  # TEST - success; matchCompare does indeed work
#  mutate(
#    genos_v2 = case_when(
#      sampleID_ill == "tamRun4.002" & locus == "INDID_190" ~ "B,B",
#      .default = genos_v2
#    )
#  ) %>%
  
  # matchCompare across pipeline combos
  mutate(
    genoMatch_gs.v2 = case_when(
      is.na(genos_gs) & !is.na(genos_v2) ~ "missing_gs",
      !is.na(genos_gs) & is.na(genos_v2) ~ "missing_v2",
      genos_gs == genos_v2 ~ "TRUE",
      genos_gs != genos_v2 ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_gs.v3a = case_when(
      is.na(genos_gs) & !is.na(genos_v3a) ~ "missing_gs",
      !is.na(genos_gs) & is.na(genos_v3a) ~ "missing_v3a",
      genos_gs == genos_v3a ~ "TRUE",
      genos_gs != genos_v3a ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_gs.v3b = case_when(
      is.na(genos_gs) & !is.na(genos_v3b) ~ "missing_gs",
      !is.na(genos_gs) & is.na(genos_v3b) ~ "missing_v3b",
      genos_gs == genos_v3b ~ "TRUE",
      genos_gs != genos_v3b ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_v2.v3a = case_when(
      is.na(genos_v2) & !is.na(genos_v3a) ~ "missing_v2",
      !is.na(genos_v2) & is.na(genos_v3a) ~ "missing_v3a",
      genos_v2 == genos_v3a ~ "TRUE",
      genos_v2 != genos_v3a ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_v2.v3b = case_when(
      is.na(genos_v2) & !is.na(genos_v3b) ~ "missing_v2",
      !is.na(genos_v2) & is.na(genos_v3b) ~ "missing_v3b",
      genos_v2 == genos_v3b ~ "TRUE",
      genos_v2 != genos_v3b ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_gr.v2 = case_when(
      is.na(genos_gr) & !is.na(genos_v2) ~ "missing_gr",
      !is.na(genos_gr) & is.na(genos_v2) ~ "missing_v2",
      genos_gr == genos_v2 ~ "TRUE",
      genos_gr != genos_v2 ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_gr.v3a = case_when(
      is.na(genos_gr) & !is.na(genos_v3a) ~ "missing_gr",
      !is.na(genos_gr) & is.na(genos_v3a) ~ "missing_v3a",
      genos_gr == genos_v3a ~ "TRUE",
      genos_gr != genos_v3a ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_gr.v3b = case_when(
      is.na(genos_gr) & !is.na(genos_v3b) ~ "missing_gr",
      !is.na(genos_gr) & is.na(genos_v3b) ~ "missing_v3b",
      genos_gr == genos_v3b ~ "TRUE",
      genos_gr != genos_v3b ~ "FALSE"
    )
  ) %>%
  mutate(
    genoMatch_gr.v3c = case_when(
      is.na(genos_gr) & !is.na(genos_v3c) ~ "missing_gr",
      !is.na(genos_gr) & is.na(genos_v3c) ~ "missing_v3c",
      genos_gr == genos_v3c ~ "TRUE",
      genos_gr != genos_v3c ~ "FALSE"
    )
  ) %>%
  mutate(
    ratio_gr = round(ratio_gr, 3),
    ratio_gs = round(ratio_gs, 3)
    ) %>%
  
  # assign callTypes for each geno called
  mutate(
    callType_gr = get_callType_gr(a1Counts_gr, a2Counts_gr, genos_gr, Allele1, Allele2),
    callType_gs = get_callType_gs(a1Counts_gs, a2Counts_gs, genos_gs, Allele1, Allele2),
    callType_v2 = get_callType_gs(a1Counts_gs, a2Counts_gs, genos_v2, Allele1, Allele2),
    callType_v3a = get_callType_gs(a1Counts_gs, a2Counts_gs, genos_v3a, Allele1, Allele2),
    callType_v3b = get_callType_gs(a1Counts_gs, a2Counts_gs, genos_v3b, Allele1, Allele2),
    callType_v3c = get_callType_gs(a1Counts_gs, a2Counts_gs, genos_v3c, Allele1, Allele2)
  )


table(compare_gtseq.score$genoMatch_gs.v3a)
table(compare_gtseq.score$genoMatch_gs.v3b)
table(compare_gtseq.score$genoMatch_gr.v2)
table(compare_gtseq.score$genoMatch_gr.v3a)
table(compare_gtseq.score$genoMatch_gr.v3b)
table(compare_gtseq.score$genoMatch_gr.v3c)

# can see that some loci have more "noCall" than others
temp <- compare_gtseq.score %>%
  filter(callType_v3c == "noCall") %>%
  group_by(locus) %>%
  summarise(
    count_noCall = n()
  ) %>%
  arrange(desc(count_noCall))
```

## Results

```{r}
# on-target readSum gtscore
sum(as.numeric(compare_gtseq.score$a1Counts_gr), na.rm = T) + sum(as.numeric(compare_gtseq.score$a2Counts_gr), na.rm = T)

sum(as.numeric(compare_gtseq.score$a1Counts_gs), na.rm = T) + sum(as.numeric(compare_gtseq.score$a2Counts_gs), na.rm = T)

10727250 - 10786814
```


## plots

```{r}
ggplot(compare_gtseq.score, aes(x = as.numeric(a1Counts_gr), y = as.numeric(a2Counts_gr), color = callType_gr, shape = callType_gr)) +
  geom_point() +
  labs(title = "gtscore0x",
       x = "allele1Counts",
       y = "allele2Counts") +
  theme_Publication()

ggplot(compare_gtseq.score, aes(x = as.numeric(a1Counts_gs), y = as.numeric(a2Counts_gs), color = callType_v2, shape = callType_v2)) +
  geom_point() +
  labs(title = "gtseq_v2",
       x = "allele1Counts",
       y = "allele2Counts") +
  theme_Publication()

ggplot(compare_gtseq.score, aes(x = as.numeric(a1Counts_gs), y = as.numeric(a2Counts_gs), color = callType_v3a, shape = callType_v3a)) +
  geom_point() +
  labs(title = "gtseq_v3a",
       x = "allele1Counts",
       y = "allele2Counts") +
  theme_Publication()

ggplot(compare_gtseq.score, aes(x = as.numeric(a1Counts_gs), y = as.numeric(a2Counts_gs), color = callType_v3b, shape = callType_v3b)) +
  geom_point() +
  labs(title = "gtseq_v3b",
       x = "allele1Counts",
       y = "allele2Counts") +
  theme_Publication()

ggplot(compare_gtseq.score, aes(x = as.numeric(a1Counts_gs), y = as.numeric(a2Counts_gs), color = callType_v3c, shape = callType_v3c)) +
  geom_point() +
  labs(title = "gtseq_v3c",
       x = "allele1Counts",
       y = "allele2Counts") +
  theme_Publication()
```




### baseline genotypes

Combine results from hair samples sequenced across multiple runs to obtain baseline genotypes

#### data

```{r}
# tamRun5 data
tamRun5_md <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv")

tamRun5_hairSamples <- tamRun5_md %>%
  filter(sampleType == "hair") %>%
  filter(animalID %in% tamRun4_md$animalID)

tamRun5_alleleReads_hair <- read.delim("./05_tamRun5/03_run5GTscore/fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  select(tamRun5_hairSamples$sampleID) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_run5.gr") %>%
  merge(., tamRun5_md[, c("sampleID", "animalID")], by = "sampleID")

tamRun5_genos0x_hair <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  select(tamRun5_hairSamples$sampleID) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_run5.gr") %>%
  merge(., tamRun5_md[, c("sampleID", "animalID")], by = "sampleID")
  

# tamRun4 data
## gtscore
tamRun4_hairSamples <- tamRun4_md %>%
  filter(sampleType == "hair")

tamRun4_ill_gtscore_alleleReads_hair <- tamRun4_readCounts_ill_gtscore %>%
  select(tamRun4_hairSamples$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID_ill",
               values_to = "readCounts_run4.gr") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill")

tamRun4_ill_gtscore_genos0x_hair <- tamRun4_genos_ill_gtscore0x %>%
  select(tamRun4_hairSamples$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID_ill",
               values_to = "genos_run4.gr") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID")], by = "sampleID_ill")


```


#### compare genos

```{r}
genoCompare_baseGenos <- tamRun4_ill_gtscore_genos0x_hair %>%
  merge(., tamRun5_genos0x_hair, by = c("animalID", "locus"), all.x = T) %>%
  #mutate(across(everything(), na_if, "0")) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  mutate(
    genoMatch = genos_run4.gr == genos_run5.gr
  ) %>%
  merge(., tamRun4_ill_gtscore_alleleReads_hair, by = c("animalID", "locus"), all.x = T) %>%
  merge(., tamRun5_alleleReads_hair, by = c("animalID", "locus"), all.x = T) %>%
  select(!starts_with("sampleID")) %>%
  mutate(
    gtRatio_run4 = round(get_gtseqRatio(readCounts_run4.gr), 2),
    gtRatio_run5 = round(get_gtseqRatio(readCounts_run5.gr), 2)
  ) %>%
  merge(., tamRun4_primerProbeFile[, c("Locus", "Allele1", "Allele2")], by.x = "locus", by.y = "Locus") %>%
  relocate(animalID, locus, Allele1, Allele2)
```


**START HERE** for the code above, adjust so that 1) readCount cols come after locus, 2) add the possible alleles for each locus so that 3) can produce gtseq genos (will need to write a small function for this to turn gtseq ratios into genotypes)


### allele readSums

```{r, warning=FALSE}
# sum allele reads
get_readSum <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# calculate variant allele frequency (w/allele 2 as variant allele)
get_vaf <- function(x) gsubfn("(\\d+),(\\d+)", ~ round(as.numeric(y)/(as.numeric(x) + as.numeric(y)), 2), paste(x))

readCompare_ill_gtscore <- tamRun4_readCounts_ill_gtscore %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "sampleID_unique")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  filter(!str_detect(sampleID_unique, "Neg")) %>%
  select(-sampleID) %>%
  pivot_longer(cols = -sampleID_unique,
               names_to = "locus",
               values_to = "reads") %>%
  separate(sampleID_unique, into = c("animalID", "sampleType")) %>%
  pivot_wider(names_from = sampleType,
              values_from = reads) %>%
  mutate(across(everything(), na_if, "0,0")) %>%
  
  # calculate readSums
  mutate(
    readSum_blood = as.numeric(get_readSum(blood)),
    readSum_fecal = as.numeric(get_readSum(fecal)),
    readSum_hair = as.numeric(get_readSum(hair))
  ) %>%
  
  # calculate a1/a2 ratio (same as calculated in gtseq pipeline)
  mutate(
    vaf_blood = as.numeric(get_vaf(blood)),
    vaf_fecal = as.numeric(get_vaf(fecal)),
    vaf_hair = as.numeric(get_vaf(hair))
  ) %>%
  dplyr::rename("reads_blood" = "blood",
                "reads_fecal"= "fecal",
                "reads_hair" = "hair")
```

### homHet key

```{r}
pp_panelv3 <- read.csv("./project_data/primerProbeFile_panelv3_fullSet.txt", header = T, sep = "\t")

homoHet_key <- pp_panelv3 %>%
  select(Locus, Allele1, Allele2) %>%
  mutate(
    hom1 = str_c(Allele1, Allele1, sep = ","),
    hom2 = str_c(Allele2, Allele2, sep = ","),
    het = str_c(Allele1, Allele2, sep = ",")
  ) %>%
  select(!contains("Allele")) %>%
  dplyr::rename("locus" = "Locus") %>%
  pivot_longer(!locus,
               names_to = "callType",
               values_to = "geno")
```

### gtscore0x

210 fecal sample seems like it has the wrong metadata

```{r}
genoCompare_ill_gtscore0x <- tamRun4_genos_ill_gtscore0x %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "sampleID_unique", "species")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  filter(!str_detect(sampleID_unique, "Neg")) %>%
  select(-sampleID) %>%
  pivot_longer(cols = -c(sampleID_unique, species),
               names_to = "locus",
               values_to = "geno") %>%
  separate(sampleID_unique, into = c("animalID", "sampleType")) %>%
  # remove genos from spSpec loci for non-target sp
#  mutate(
#    geno = case_when(
#      str_detect(species, "LWED") & str_detect(locus, "SIMP") ~ "0",
#      str_detect(species, "SIMP") & str_detect(locus, "LWED") ~ "0",
#      .default = geno
#    )
#  ) %>%
  pivot_wider(names_from = sampleType,
              values_from = geno) %>%
  mutate(across(everything(), na_if, "0")) %>%
  mutate(
    bfMatch = blood == fecal,
    bhMatch = blood == hair,
    fhMatch = fecal == hair,
    totalGenos = (rowSums(!is.na(select(., c(blood, fecal, hair)))))
  ) %>%
  rowwise() %>%
  mutate(
    uniqueGenos = n_distinct(c_across(blood:hair), na.rm = T),
    seqType = "gtscore0x"
  ) %>%
  merge(., readCompare_ill_gtscore, by = c("animalID", "locus"), all.x = T) %>%
  merge(., homoHet_key, by.x = c("locus", "blood"), by.y = c("locus", "geno"), all.x = T) %>%
  merge(., homoHet_key, by.x = c("locus", "fecal"), by.y = c("locus", "geno"), suffixes = c("_blood", "_fecal"), all.x = T) %>%
  merge(., homoHet_key, by.x = c("locus", "hair"), by.y = c("locus", "geno"), all.x = T) %>%
  dplyr::rename("callType_hair" = "callType") %>%
  mutate(
    bf_misType = case_when(
      is.na(bfMatch) ~ NA,
      bfMatch == TRUE ~ NA,
      callType_blood == "het" & callType_fecal %in% c("hom1", "hom2") ~ "type1",
      callType_blood %in% c("hom1", "hom2") & callType_fecal == "het" ~ "type1",
      .default = "type2"
    )
  ) %>%
  relocate(bf_misType, .after = bfMatch) %>%
  mutate(
    bh_misType = case_when(
      is.na(bhMatch) ~ NA,
      bhMatch == TRUE ~ NA,
      callType_blood == "het" & callType_hair %in% c("hom1", "hom2") ~ "type1",
      callType_blood %in% c("hom1", "hom2") & callType_hair == "het" ~ "type1",
      .default = "type2"
    )
  ) %>%
  relocate(bh_misType, .after = bhMatch) %>%
  mutate(
    fh_misType = case_when(
      is.na(fhMatch) ~ NA,
      fhMatch == TRUE ~ NA,
      callType_fecal == "het" & callType_hair %in% c("hom1", "hom2") ~ "type1",
      callType_fecal %in% c("hom1", "hom2") & callType_hair == "het" ~ "type1",
      .default = "type2"
    )
  ) %>%
  relocate(fh_misType, .after = fhMatch) %>%
  relocate(locus, animalID, species, blood, fecal, hair)


temp_0x <- genoCompare_ill_gtscore0x %>%
  group_by(animalID) %>%
  summarise(
    type2Mismatch_bf = sum(bf_misType == "type2", na.rm = T),
    type2Mismatch_bh = sum(bh_misType == "type2", na.rm = T),
    type2Mismatch_fh = sum(fh_misType == "type2", na.rm = T),
    
    mismatch_bf = sum(bfMatch == FALSE, na.rm = T),
    mismatch_bh = sum(bhMatch == FALSE, na.rm = T),
    mismatch_fh = sum(fhMatch == FALSE, na.rm = T),
    
    common_bf = sum(!is.na(bfMatch)),
    common_bh = sum(!is.na(bhMatch)),
    common_fh = sum(!is.na(fhMatch))
  )



# general overview
temp_0x <- genoCompare_ill_gtscore0x %>%
  na.omit()

genoCompare_ill_gtscore0x_overview <- data.frame(
  totalGenos = nrow(temp_0x),
  bfMatch = sum(temp_0x$bfMatch == TRUE),
  bhMatch = sum(temp_0x$bhMatch == TRUE),
  fhMatch = sum(temp_0x$fhMatch == TRUE)
  ) %>%
  mutate(
    prop_bfMatch = round(bfMatch/totalGenos, 2),
    prop_bhMatch = round(bhMatch/totalGenos, 2),
    prop_fhMatch = round(fhMatch/totalGenos, 2)
  )

genoCompare_ill_gtscore0x_overview_byIndiv <- temp_0x %>%
  group_by(animalID) %>%
  summarise(
    totalGenos = n(),
    prop_bfMatch = round(sum(bfMatch == TRUE)/n(), 2),
    prop_bhMatch = round(sum(bhMatch == TRUE)/n(), 2),
    prop_fhMatch = round(sum(fhMatch == TRUE)/n(), 2)
  )

genoCompare_ill_gtscore0x_overview_byLociSet <- temp_0x %>%
  mutate(
    lociSet = case_when(
      str_detect(locus, "INDID") ~ "ind_dual",
      str_detect(locus, "SEXID_LWED") ~ "sex_lwed",
      str_detect(locus, "SEXID_SIMP") ~ "sex_simp",
      str_detect(locus, "SEXID") ~ "sex_dual",
      str_detect(locus, "LWED") ~ "ind_lwed",
      str_detect(locus, "SIMP") ~ "ind_simp",
      str_detect(locus, "SPECIES") ~ "sp"
    )
  ) %>%
  group_by(animalID, lociSet) %>%
  summarise(
    totalLoci = n(),
    prop_bfMatch = round(sum(bfMatch == TRUE)/n(), 2),
    prop_bhMatch = round(sum(bhMatch == TRUE)/n(), 2),
    prop_fhMatch = round(sum(fhMatch == TRUE)/n(), 2)
  )

# what about by locus?
genoCompare_ill_gtscore0x_overview_byLocus <- temp_0x %>%
  group_by(locus) %>%
  summarise(
    totalGenos = n(),
    prop_bfMatch = round(sum(bfMatch == TRUE)/n(), 2),
    prop_bhMatch = round(sum(bhMatch == TRUE)/n(), 2),
    prop_fhMatch = round(sum(fhMatch == TRUE)/n(), 2)
  ) %>%
  arrange(desc(totalGenos))
```

### gtscore10x

```{r}
genoCompare_ill_gtscore10x <- tamRun4_genos_ill_gtscore10x %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "sampleID_unique", "species")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  filter(!str_detect(sampleID_unique, "Neg")) %>%
  select(-sampleID) %>%
  pivot_longer(cols = -c(sampleID_unique, species),
               names_to = "locus",
               values_to = "geno") %>%
  separate(sampleID_unique, into = c("animalID", "sampleType")) %>%
  # remove genos from spSpec loci for non-target sp
#  mutate(
#    geno = case_when(
#      str_detect(species, "LWED") & str_detect(locus, "SIMP") ~ "0",
#      str_detect(species, "SIMP") & str_detect(locus, "LWED") ~ "0",
#      .default = geno
#    )
#  ) %>%
  pivot_wider(names_from = sampleType,
              values_from = geno) %>%
  mutate(across(everything(), na_if, "0")) %>%
  mutate(
    bfMatch = blood == fecal,
    bhMatch = blood == hair,
    fhMatch = fecal == hair,
    totalGenos = (rowSums(!is.na(select(., c(blood, fecal, hair)))))
  ) %>%
  rowwise() %>%
  mutate(
    uniqueGenos = n_distinct(c_across(blood:hair), na.rm = T),
    seqType = "gtscore10x"
  ) %>%
  merge(., readCompare_ill_gtscore, by = c("animalID", "locus"), all.x = T) %>%
  merge(., homoHet_key, by.x = c("locus", "blood"), by.y = c("locus", "geno"), all.x = T) %>%
  merge(., homoHet_key, by.x = c("locus", "fecal"), by.y = c("locus", "geno"), suffixes = c("_blood", "_fecal"), all.x = T) %>%
  merge(., homoHet_key, by.x = c("locus", "hair"), by.y = c("locus", "geno"), all.x = T) %>%
  dplyr::rename("callType_hair" = "callType") %>%
  mutate(
    bf_misType = case_when(
      is.na(bfMatch) ~ NA,
      bfMatch == TRUE ~ NA,
      callType_blood == "het" & callType_fecal %in% c("hom1", "hom2") ~ "type1",
      callType_blood %in% c("hom1", "hom2") & callType_fecal == "het" ~ "type1",
      .default = "type2"
    )
  ) %>%
  relocate(bf_misType, .after = bfMatch) %>%
  mutate(
    bh_misType = case_when(
      is.na(bhMatch) ~ NA,
      bhMatch == TRUE ~ NA,
      callType_blood == "het" & callType_hair %in% c("hom1", "hom2") ~ "type1",
      callType_blood %in% c("hom1", "hom2") & callType_hair == "het" ~ "type1",
      .default = "type2"
    )
  ) %>%
  relocate(bh_misType, .after = bhMatch) %>%
  mutate(
    fh_misType = case_when(
      is.na(fhMatch) ~ NA,
      fhMatch == TRUE ~ NA,
      callType_fecal == "het" & callType_hair %in% c("hom1", "hom2") ~ "type1",
      callType_fecal %in% c("hom1", "hom2") & callType_hair == "het" ~ "type1",
      .default = "type2"
    )
  ) %>%
  relocate(fh_misType, .after = fhMatch) %>%
  relocate(locus, animalID, species, blood, fecal, hair)


temp_10x <- genoCompare_ill_gtscore10x %>%
  group_by(animalID) %>%
  summarise(
    type2Mismatch_bf = sum(bf_misType == "type2", na.rm = T),
    type2Mismatch_bh = sum(bh_misType == "type2", na.rm = T),
    type2Mismatch_fh = sum(fh_misType == "type2", na.rm = T),
    
    mismatch_bf = sum(bfMatch == FALSE, na.rm = T),
    mismatch_bh = sum(bhMatch == FALSE, na.rm = T),
    mismatch_fh = sum(fhMatch == FALSE, na.rm = T),
    
    common_bf = sum(!is.na(bfMatch)),
    common_bh = sum(!is.na(bhMatch)),
    common_fh = sum(!is.na(fhMatch))
  )

# general overview
temp_10x <- genoCompare_ill_gtscore10x %>%
  na.omit()

genoCompare_ill_gtscore10x_overview <- data.frame(
  totalGenos = nrow(temp_10x),
  bfMatch = sum(temp_10x$bfMatch == TRUE),
  bhMatch = sum(temp_10x$bhMatch == TRUE),
  fhMatch = sum(temp_10x$fhMatch == TRUE)
  ) %>%
  mutate(
    prop_bfMatch = round(bfMatch/totalGenos, 2),
    prop_bhMatch = round(bhMatch/totalGenos, 2),
    prop_fhMatch = round(fhMatch/totalGenos, 2)
  ) %>%
  mutate(
    prop_bfMismatch = round((1 - bfMatch/totalGenos), 2),
    prop_bhMismatch = round((1 - bhMatch/totalGenos), 2),
    prop_fhMismatch = round((1 - fhMatch/totalGenos), 2)
  )

# overview by animalID
genoCompare_ill_gtscore10x_overview_byIndiv <- temp %>%
  group_by(animalID) %>%
  summarise(
    totalGenos = n(),
    prop_bfMatch = round(sum(bfMatch == TRUE)/n(), 2),
    prop_bhMatch = round(sum(bhMatch == TRUE)/n(), 2),
    prop_fhMatch = round(sum(fhMatch == TRUE)/n(), 2)
  )

# add in lociSet
genoCompare_ill_gtscore10x_overview_byLociSet <- temp %>%
  mutate(
    lociSet = case_when(
      str_detect(locus, "INDID") ~ "ind_dual",
      str_detect(locus, "SEXID_LWED") ~ "sex_lwed",
      str_detect(locus, "SEXID_SIMP") ~ "sex_simp",
      str_detect(locus, "SEXID") ~ "sex_dual",
      str_detect(locus, "LWED") ~ "ind_lwed",
      str_detect(locus, "SIMP") ~ "ind_simp",
      str_detect(locus, "SPECIES") ~ "sp"
    )
  ) %>%
  group_by(animalID, lociSet) %>%
  summarise(
    totalGenos = n(),
    prop_bfMatch = round(sum(bfMatch == TRUE)/n(), 2),
    prop_bhMatch = round(sum(bhMatch == TRUE)/n(), 2),
    prop_fhMatch = round(sum(fhMatch == TRUE)/n(), 2)
  )

# what about by locus?
genoCompare_ill_gtscore10x_overview_byLocus <- temp %>%
  group_by(locus) %>%
  summarise(
    totalGenos = n(),
    prop_bfMatch = round(sum(bfMatch == TRUE)/n(), 2),
    prop_bhMatch = round(sum(bhMatch == TRUE)/n(), 2),
    prop_fhMatch = round(sum(fhMatch == TRUE)/n(), 2)
  ) %>%
  arrange(desc(totalGenos))
```

**figures**

```{r}

```


### gtscore0x vs. 10x

looking at differences in bfh consistency b/t 0x and 10x gtscore calls

```{r}
genoCompare_ill_gtscore0x.10x_overview_byIndiv <- genoCompare_ill_gtscore0x_overview_byIndiv %>%
  merge(., genoCompare_ill_gtscore10x_overview_byIndiv, by = "animalID", suffixes = c("_0x", "_10x")) %>%
  relocate(animalID, contains("total"), contains("prop_bf"), contains("prop_bh"), contains("prop_fh")) %>%
  mutate(
    diff_totalGenos = totalGenos_0x - totalGenos_10x,
    diff_prop_bfMatch = prop_bfMatch_0x - prop_bfMatch_10x,
    diff_prop_bhMatch = prop_bhMatch_0x - prop_bhMatch_10x,
    diff_prop_fhMatch = prop_fhMatch_0x - prop_fhMatch_10x
  ) %>%
  relocate(animalID, contains("total"), contains("diff"))
```

### Allele reads

how do the allele reads look for these genos?

**ADD GENOS TO DF BELOW, SEE IF MATCHING CORRELATES W/COVERAGE**

```{r, warning=FALSE}
# sum allele reads
get_readSum <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# calculate variant allele frequency (w/allele 2 as variant allele)
get_vaf <- function(x) gsubfn("(\\d+),(\\d+)", ~ round(as.numeric(y)/(as.numeric(x) + as.numeric(y)), 2), paste(x))

readCompare_ill_gtscore <- tamRun4_readCounts_ill_gtscore %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., tamRun4_md[, c("sampleID_ill", "sampleID_unique")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  filter(!str_detect(sampleID_unique, "Neg")) %>%
  select(-sampleID) %>%
  pivot_longer(cols = -sampleID_unique,
               names_to = "locus",
               values_to = "reads") %>%
  separate(sampleID_unique, into = c("animalID", "sampleType")) %>%
  pivot_wider(names_from = sampleType,
              values_from = reads) %>%
  mutate(across(everything(), na_if, "0,0")) %>%
  
  # calculate readSums
  mutate(
    readSum_blood = as.numeric(get_readSum(blood)),
    readSum_fecal = as.numeric(get_readSum(fecal)),
    readSum_hair = as.numeric(get_readSum(hair))
  ) %>%
  
  # calculate a1/a2 ratio (same as calculated in gtseq pipeline)
  mutate(
    vaf_blood = as.numeric(get_vaf(blood)),
    vaf_fecal = as.numeric(get_vaf(fecal)),
    vaf_hair = as.numeric(get_vaf(hair))
  ) %>%
  dplyr::rename("reads_blood" = "blood",
                "reads_fecal"= "fecal",
                "reads_hair" = "hair")
```

**add genos**

```{r}
genoReadCompare_ill_gtscore <- rbind(genoCompare_ill_gtscore0x, genoCompare_ill_gtscore10x) %>%
  merge(., readCompare_ill_gtscore, by = c("animalID", "locus"), all.x = T)

genoReadCompare_ill_gtscore0x_geno100 <- genoReadCompare_ill_gtscore %>%
  filter(seqType == "gtscore0x") %>%
  filter(totalGenos == 3)

genoReadCompare_ill_gtscore10x_geno100 <- genoReadCompare_ill_gtscore %>%
  filter(seqType == "gtscore10x") %>%
  filter(totalGenos == 3)
```

### figures

```{r}
genoReadCompare_ill_gtscore0x_geno100 %>%
  ggplot(aes(x = readSum_fecal, y = readSum_hair, shape = fhMatch, color = fhMatch)) +
  geom_point(alpha = 0.5) +
  scale_colour_Publication() +
  theme_Publication()
```

### new approach










# SCRAPS

**NEEDS UPDATING** bp3 error shit has me confused - n = 302 is right, but snp location and such may not be

After doing a few more in silico tests in which we replaced the primers that failed in silico with their bp3 alternatives, we end up with **n = 302** primer-pairs that passed in silico and were cleared for primer ordering:

-   INDID       151 (122 original + 29 additional)
-   LWED        62
-   SIMP        53
-   SEXID       18 (11 general, 3 LWED, 4 SIMP)
-   SPECIES     18
-   total       302

```{r}
panel_v1 <- read.csv("./primers/isPCR5_results.csv") %>%
  filter(isPCR5_final == "PASS") %>%
  arrange(as.numeric(uniqueIndex)) %>%
  mutate(
    set = case_when(
      set == "IND_ID" ~ "INDID",
      set == "IND_ID_add" ~ "INDID_add",
      set == "LWED_IND_ID" ~ "LWED",
      set == "SEX_ID" ~ "SEXID",
      set == "SIMP_IND_ID" ~ "SIMP",
      set == "SPECIES_ID" ~ "SPECIES"
    )
  )

# make sure this matches primerSet_v1 - success
primerSet_v1 <- read.csv("./primers/03_lociChoices/primerSet_v1.csv") %>%
  mutate(uniqueIndex = str_c(bp3Index, ".", bp3Rank))

primerSet_v1 %>%
  select(lociSet, primerName2) %>%
  distinct() %>%
  group_by(lociSet) %>%
  summarise(count = n())

primerSet_v1 %>%
  filter(!uniqueIndex %in% panel_v1$uniqueIndex)

panel_v1 %>%
  filter(!uniqueIndex %in% primerSet_v1$uniqueIndex)

# side quest -- primerSet_v1 has NAs for chrom number and such for some loci; fill in using panel_v1
primerSet_v1_updated <- primerSet_v1 %>%
  dplyr::rename("snpLocation" = "SNP",
                "seqStart" = "start", 
                "seqEnd" = "stop",
                "seqLength" = "length") %>%
  merge(., panel_v1[, c("uniqueIndex", "chrom_num", "chrom_name", "snpLocation", "seqStart", "seqEnd", "seqLength")], by = "uniqueIndex") %>%
  mutate(
    chrom_num = coalesce(chrom_num.x, chrom_num.y),
    chrom_name = coalesce(chrom_name.x, chrom_name.y),
    snpLocation = coalesce(snpLocation.x, snpLocation.y),
    seqStart = coalesce(seqStart.x, seqStart.y),
    seqEnd = coalesce(seqEnd.x, seqEnd.y),
    seqLength = coalesce(seqLength.x, seqLength.y),
    .keep = "unused"
  ) %>%
  relocate(c(chrom_num, chrom_name, snpLocation, seqStart, seqEnd, seqLength), .after = bp3Rank)

# also has NAs for some primer sequences -- fill in below
primerSet_v1_updated2
```

### panel_v2

#### samTable

```{r}
tamRun1_pp_fullSet <- read.table("./00_tamRun1/03_run1GTscore/primerProbeFile.txt", header = T) %>%
  mutate(
    Locus = case_when(
      Locus == "SEXID_195.3" ~ "SEXID_LWED_195.3",
      Locus == "SEXID_208.3" ~ "SEXID_LWED_208.3",
      Locus == "SEXID_211.3" ~ "SEXID_LWED_211.3",
      
      Locus == "SEXID_197.1" ~ "SEXID_SIMP_197.1",
      Locus == "SEXID_198.3" ~ "SEXID_SIMP_198.3",
      Locus == "SEXID_203.2" ~ "SEXID_SIMP_203.2",
      Locus == "SEXID_218.1" ~ "SEXID_SIMP_218.1",
      
      .default = Locus
    )
  )
  
tamRun1_pp_bp3error <- read.table("./00_tamRun1/03_run1GTscore/bp3error_primerProbeFile.txt", header = T)

samTable <- read.csv("./00_tamRun1/03_run1GTscore/snpPanel_samTable.csv", na.strings = "") %>%
  dplyr::rename("primerF" = "Forward.primer",
                "primerR" = "Reverse.primer") %>%
  mutate(
    primerF = toupper(primerF),
    primerR = toupper(primerR)
    ) %>%
  merge(., primerSet_v1_updated[, c("seqPrimer", "uniqueIndex")], by.x = "primerF", by.y = "seqPrimer", all.x = T) %>%
  merge(., tamRun1_pp_fullSet[, c("Primer", "Locus")], by.x = "primerF", by.y = "Primer", all.x = T) %>%
  merge(., tamRun1_pp_bp3error[, c("Primer", "Locus")], by.x = "primerF", by.y = "Primer", all.x = T) %>%
  relocate(primerF, .before = primerR) %>%
  mutate(Locus = coalesce(Locus.x, Locus.y), .keep = "unused") %>%
  separate(Locus, into = c("temp1", "temp2", "temp3"), sep = "_") %>%
  mutate(
    temp1 = case_when(
      temp2 %in% c("LWED", "SIMP") ~ str_c(temp1, temp2, sep = "_"),
      .default = temp1
    ),
    temp2 = case_when(
      temp2 %in% c("LWED", "SIMP") ~ temp3,
      .default = temp2
    )
  ) %>%
  select(-temp3) %>%
  mutate(Set = temp1) %>%
  select(-temp1) %>%
  mutate(uniqueIndex = temp2) %>%
  select(-temp2) %>%
  relocate(uniqueIndex) %>%
  arrange(as.numeric(uniqueIndex))
```


```{r}
tamRun5_pp_fullSet <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)
```



### primerSet_v2

n_loci = 221

```{r}
primerSet_v2 <- read.csv("./primers/03_lociChoices/primerSet_v2.csv") %>%
  mutate(uniqueIndex = str_c(bp3Index, ".", bp3Rank))

temp <- primerSet_v2 %>%
  filter(!uniqueIndex %in% samTable$uniqueIndex)

table(primerSet_v2$lociSet)
```

### primerSet_v3

n_loci = 208

```{r}
primerSet_v3 <- read.csv("./primers/03_lociChoices/primerSet_v3.csv")

nrow(primerSet_v3)/2

table(primerSet_v3$lociSet)
```



## 7.1 Read metrics

Read metrics are based on the full set of samples + the full set of loci - so including negatives as well as species-specific loci

Total PE reads = 31663938
Primer-probe reads = 10727250 (0.34 of total PE reads)
Primer only reads = 3705585 (0.12 of total PE reads)
Completely off-target reads = 17231103 (0.54 of total PE reads)
On-target reads = 14432835 (0.46 of total PE reads)

```{r}
# data


tamRun4_indivSum_negs <- tamRun4_indivSum_fullSet %>%
  filter(rownames(.) %in% tamRun4_md_negs$sampleID_ill)

tamRun4_indivSum_lwed <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_lwed_GTscore_individualSummary.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = gsub("-", "\\.", sampleID)) %>%
  filter(!sampleID %in% tamRun4_md_negs$sampleID_ill) %>%
  column_to_rownames("sampleID")

tamRun4_indivSum_simp <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_simp_GTscore_individualSummary.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = gsub("-", "\\.", sampleID)) %>%
  filter(!sampleID %in% tamRun4_md_negs$sampleID_ill) %>%
  column_to_rownames("sampleID")

# FULLSET metrics
sum(tamRun4_indivSum_fullSet$Total.Reads) # 31663938
sum(tamRun4_indivSum_fullSet$Primer.Probe.Reads) # 10727250
sum(tamRun4_indivSum_fullSet$Primer.Only.Reads) # 3705585
sum(tamRun4_indivSum_fullSet$Off.target.Reads) # 17231103

# GROUP-SPECIFIC metrics
# primer-probe reads
sum(tamRun4_indivSum_negs$Primer.Probe.Reads) # 6365
sum(tamRun4_indivSum_lwed$Primer.Probe.Reads) # 4715980
sum(tamRun4_indivSum_simp$Primer.Probe.Reads) # 4452993

sum(tamRun4_indivSum_negs$Primer.Probe.Reads) +
  sum(tamRun4_indivSum_lwed$Primer.Probe.Reads) +
  sum(tamRun4_indivSum_simp$Primer.Probe.Reads) # 9175338

# primer only reads
sum(tamRun4_indivSum_negs$Primer.Only.Reads) # 279127
sum(tamRun4_indivSum_lwed$Primer.Only.Reads) # 1391804
sum(tamRun4_indivSum_simp$Primer.Only.Reads) # 1442793

sum(tamRun4_indivSum_negs$Primer.Only.Reads) +
  sum(tamRun4_indivSum_lwed$Primer.Only.Reads) +
  sum(tamRun4_indivSum_simp$Primer.Only.Reads) # 3113724

# completely off-target reads (no primer, no probe)
sum(tamRun4_indivSum_negs$Off.target.Reads) # 494852
sum(tamRun4_indivSum_lwed$Off.target.Reads) # 9671716
sum(tamRun4_indivSum_simp$Off.target.Reads) # 9208308

sum(tamRun4_indivSum_negs$Off.target.Reads) +
  sum(tamRun4_indivSum_lwed$Off.target.Reads) +
  sum(tamRun4_indivSum_simp$Off.target.Reads) # 19374876
```

## 7.2 Geno success

(w/loci for other species removed)

### gtscore0x

```{r}
genoSuccess_bySample_gtscore0x <- tamRun4_genos_ill_gtscore0x %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), na_if, "0")) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(
    totalGenos = case_when(
      sampleID %in% sampleList_lwed$sampleID_ill ~ rowSums(!is.na(select(., lociList_lwed$Locus))),
      sampleID %in% sampleList_simp$sampleID_ill ~ rowSums(!is.na(select(., lociList_simp$Locus))),
      .default = rowSums(!is.na(select(., -sampleID)))
    )
  ) %>%
  select(sampleID, totalGenos) %>%
  mutate(
    possibleGenos = case_when(
      sampleID %in% sampleList_lwed$sampleID_ill ~ nrow(lociList_lwed),
      sampleID %in% sampleList_simp$sampleID_ill ~ nrow(lociList_simp),
      .default = nrow(tamRun4_primerProbeFile)
    ),
    propSuccess = round(totalGenos/possibleGenos, 2)
  ) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "species", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(c(sampleID, species, sampleType, propSuccess))

# geno success per animalID across sampleTypes
genoSuccess_bySample_gtscore0x_byID <- genoSuccess_bySample_gtscore0x %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID", "captureDate")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  select(-sampleID) %>%
  relocate(animalID) %>%
  arrange(animalID)

genoSuccess_bySample_gtscore0x_byID %>%
  filter(propSuccess < 0.5) %>%
  select(animalID) %>%
  distinct()
```

```{r}
# summary of avg genoSuccess by species/sampleType
genoSuccess_bySample_gtscore0x %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  group_by(species, sampleType) %>%
  summarise(
    mean_genoSuccess = round(mean(propSuccess), 2)
  ) %>%
  as.data.frame() %>%
  mutate(
    possibleGenos = case_when(
      species == "LWED" ~ nrow(lociList_lwed),
      species == "SIMP" ~ nrow(lociList_simp),
      .default = nrow(tamRun4_primerProbeFile)
    )
  )

# species combined, per sample type:
genoSuccess_bySample_gtscore0x %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  group_by(sampleType) %>%
  summarise(
    mean_genoSuccess = round(mean(propSuccess), 2)
  )
```

### gtscore10x

```{r}
genoSuccess_bySample_gtscore10x <- tamRun4_genos_ill_gtscore10x %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), na_if, "0")) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(
    totalGenos = case_when(
      sampleID %in% sampleList_lwed$sampleID_ill ~ rowSums(!is.na(select(., lociList_lwed$Locus))),
      sampleID %in% sampleList_simp$sampleID_ill ~ rowSums(!is.na(select(., lociList_simp$Locus))),
      .default = rowSums(!is.na(select(., -sampleID)))
    )
  ) %>%
  select(sampleID, totalGenos) %>%
  mutate(
    possibleGenos = case_when(
      sampleID %in% sampleList_lwed$sampleID_ill ~ nrow(lociList_lwed),
      sampleID %in% sampleList_simp$sampleID_ill ~ nrow(lociList_simp),
      .default = nrow(tamRun4_primerProbeFile)
    ),
    propSuccess = round(totalGenos/possibleGenos, 2)
  ) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "species", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(c(sampleID, species, sampleType, propSuccess))

# geno success per animalID across sampleTypes
genoSuccess_bySample_gtscore10x_byID <- genoSuccess_bySample_gtscore10x %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "animalID", "captureDate")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  select(-sampleID) %>%
  relocate(animalID) %>%
  arrange(animalID)

genoSuccess_bySample_gtscore10x_byID %>%
  filter(propSuccess < 0.5) %>%
  select(animalID) %>%
  distinct()
```

```{r}
# summary of avg genoSuccess by species/sampleType
genoSuccess_bySample_gtscore10x %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  group_by(species, sampleType) %>%
  summarise(
    mean_genoSuccess = round(mean(propSuccess), 2)
  ) %>%
  as.data.frame() %>%
  mutate(
    possibleGenos = case_when(
      species == "LWED" ~ nrow(lociList_lwed),
      species == "SIMP" ~ nrow(lociList_simp),
      .default = nrow(tamRun4_primerProbeFile)
    )
  )

# species combined, per sample type:
genoSuccess_bySample_gtscore10x %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  group_by(sampleType) %>%
  summarise(
    mean_genoSuccess = round(mean(propSuccess), 2)
  )
```
