---
title: "GTscorePipeline_COMBINED"
output: html_document
date: '2022-06-07'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,fig.path = "tools/")
```

# 1 Packages

```{r load packages, include=FALSE}
source("./project_scripts/GTscore_modified.R") # NOTE- added N=ATGC to script for later analyses (not included in original)
library(gsubfn)
library(phylotools)
library(TidyMultiqc)
library(tidyverse)
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", version = "3.16")
BiocManager::install("Biostrings")
```

GTscore also requires Perl and the packages "Algorithm::Combinatorics" and "Excel-Writer-XLSX". Perform the following in terminal to install on Linux. NOTE -- be sure to install both perl and the modules in root. See how to install perl packages [here](http://www.cpan.org/modules/INSTALL.html).

NOTE - the old way of installing perl through conda (conda install -c conda-forge perl) doesn't play nice anymore; better to install directly from perl:

```{bash, eval = FALSE}
# Install perl
curl -L http://xrl.us/installperlnix | bash

# Install cpanminus (makes installing perl modules easier)
cpan App::cpanminus

# Install necessary perl modules
sudo cpan Algorithm::Combinatorics
sudo cpan Excel::Writer::XLSX
```

I also recently had issues with installing perl modules with "cpanm" -- installing with "cpan" instead solved things. Generally speaking, perl continues to be a massive pain in new fun ways every time I open this up to re-run something.

# 2 Overview

Redoing GTscorePipeline_tamRun1_clean after removing loci w/bp3error


# 3 Files

## 3.1 Sample files

```{r}
sf_fullSet <- read.table("./00_tamRun1_optimization/03_run1GTscore/sampleFiles_fullSet.txt")

sf_lwed <- read.table("./00_tamRun1_optimization/03_run1GTscore/sampleFiles_LWED.txt")

sf_simp <- read.table("./00_tamRun1_optimization/03_run1GTscore/sampleFiles_SIMP.txt")

sf_pos <- read.table("./00_tamRun1_optimization/03_run1GTscore/sampleFiles_pos.txt")
```

## 5.2 Primer probe files

```{r}
bp3error_toRemove <- read.csv("./00_tamRun1_optimization/bp3error_investigation/bp3error_master_1July2024.csv") %>%
  filter(pf_bp3error == "fail")

pp_fullSet <- read.table("./00_tamRun1_optimization/03_run1GTscore/primerProbeFile_fullSet.txt", header = T) %>%
  filter(!Locus %in% bp3error_toRemove$locus)

pp_dual <- pp_fullSet %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP"))

pp_lwed <- pp_fullSet %>%
  filter(!str_detect(Locus, "SIMP"))

pp_simp <- pp_fullSet %>%
  filter(!str_detect(Locus, "LWED"))

pp_lwedSpec <- pp_fullSet %>%
  filter(str_detect(Locus, "LWED"))

pp_simpSpec <- pp_fullSet %>%
  filter(str_detect(Locus, "SIMP"))

write.table(pp_fullSet, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_fullSet.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_dual, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_dual.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_lwed, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_LWED.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_simp, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_SIMP.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_lwedSpec, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_lwedSpec.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_simpSpec, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_simpSpec.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```

# 6 Count reads for amplicons

```{r count reads for amplicons, eval=FALSE}
# all samples, all loci (includes negatives)
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_fullSet.txt --files ./00_tamRun1_optimization/03_run1GTscore/sampleFiles_fullSet.txt --inDir ./00_tamRun1_optimization/02_run1Interleaved/ --outDir ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/ --prefix fullSet_")

# pos samples (no negs), dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_dual.txt --files ./00_tamRun1_optimization/03_run1GTscore/sampleFiles_pos.txt --inDir ./00_tamRun1_optimization/02_run1Interleaved/ --outDir ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/ --prefix posDual_")

# lwed samples, lwed + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_LWED.txt --files ./00_tamRun1_optimization/03_run1GTscore/sampleFiles_LWED.txt --inDir ./00_tamRun1_optimization/02_run1Interleaved/ --outDir ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/ --prefix LWED_")

# lwed samples, simp-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_simpSpec.txt --files ./00_tamRun1_optimization/03_run1GTscore/sampleFiles_LWED.txt --inDir ./00_tamRun1_optimization/02_run1Interleaved/ --outDir ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/ --prefix LWED_simpSpec_")

# simp samples, simp + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_SIMP.txt --files ./00_tamRun1_optimization/03_run1GTscore/sampleFiles_SIMP.txt --inDir ./00_tamRun1_optimization/02_run1Interleaved/ --outDir ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/ --prefix SIMP_")

# simp samples, lwed-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/primerProbeFile_lwedSpec.txt --files ./00_tamRun1_optimization/03_run1GTscore/sampleFiles_SIMP.txt --inDir ./00_tamRun1_optimization/02_run1Interleaved/ --outDir ./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/ --prefix SIMP_lwedSpec_")
```

# 7 Genotyping

Genotyping is accomplished using the polyGen function. The genotyping algorithm is described in McKinney et al. 2018 and is a maximum likelihood algorithm capable of genotyping any number of alleles and ploidy per locus. This allows genoyping of single SNPs as well as microhaplotypes, and loci with elevated ploidy.

Two arguments are required for polyGen, the locusTable and alleleReads files output by AmpliconReadCounter.

Optional arguments for polyGen are:

* p_thresh  - threshold p-value for likelihood ratio test (default 0.05)
* epsilon   - error rate for genotyping model (default 0.01)

**NOTE** that only primer probe reads are used in genotyping!

Note also that GTscore doesn't have a coverage cutoff for genotyping - as such I'm running the genotyping function as-is plus again w/a 10x cutoff

## 7.1 0x cutoff

### Full set genos

```{r}
#load locus table and 0x allele reads file
fullSet_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

fullSet_singleSNP_alleleReads_0x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm, adjust "0" formatting
fullSet_polyGenResults_singleSNP_0x <- polyGen(fullSet_singleSNP_locusTable, fullSet_singleSNP_alleleReads_0x)

# write results
write.table(fullSet_polyGenResults_singleSNP_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

### Species subsets

#### pos samples only (no negs)

```{r Genotyping}
#load locus table and allele reads file
posDual_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

posDual_singleSNP_alleleReads_0x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
posDual_polyGenResults_singleSNP_0x <- polyGen(posDual_singleSNP_locusTable, posDual_singleSNP_alleleReads_0x)

#write results
write.table(posDual_polyGenResults_singleSNP_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

#### LWED

```{r Genotyping}
#load locus table and 10x allele reads file
LWED_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

LWED_singleSNP_alleleReads_0x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
LWED_polyGenResults_singleSNP_0x <- polyGen(LWED_singleSNP_locusTable, LWED_singleSNP_alleleReads_0x)

#write results
write.table(LWED_polyGenResults_singleSNP_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

#### SIMP

```{r}
#load locus table and 10x allele reads file
SIMP_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

SIMP_singleSNP_alleleReads_0x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
SIMP_polyGenResults_singleSNP_0x <- polyGen(SIMP_singleSNP_locusTable, SIMP_singleSNP_alleleReads_0x)

#write results
write.table(SIMP_polyGenResults_singleSNP_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

## 7.2 10x cutoff

### Recode loci \<10x coverage

Step 1: Recode loci with \<10x coverage in the full set of allele read counts

```{r}
# Read in allele counts for the full dataset
readCounts_original <- read.table("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_AlleleReads_singleSNPs.txt")

# Create read counts copy for 10x coverage
readCounts10x <- readCounts_original

# Set up a function to sum the read counts per allele for each locus, using package gsubfn
repl <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# Then apply the function to readCounts to sum each set of allele reads for each locus
readCounts_sum10x <- replace(readCounts10x, TRUE, lapply(readCounts10x, repl)) %>%
  mutate(across(everything(),as.numeric))

# Recode <10x loci with "0"
readCounts10x[readCounts_sum10x < 10] <- "0,0"
```

Step 2: Create lwed & simp subsets & create new AlleleReads_singleSNPs files

```{r}
# lwed and simp loci sets
lwed_loci <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

simp_loci <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

readCounts10x_lwed <- readCounts10x %>%
  select(lwed$sampleID) %>%
  filter(rownames(.) %in% lwed_loci$Locus_ID)

readCounts10x_simp <- readCounts10x %>%
  select(simp$sampleID) %>%
  filter(rownames(.) %in% simp_loci$Locus_ID)

readCounts10x_posDual <- readCounts10x %>%
  select(colnames(posDual_singleSNP_alleleReads_0x)) %>%
  filter(rownames(.) %in% rownames(posDual_singleSNP_alleleReads_0x))
  
# Export new 10x AlleleReads_singleSNPs files
write.table(readCounts10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_AlleleReads_singleSNPs_10x.txt", quote = FALSE, sep = "\t")

write.table(readCounts10x_posDual, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_AlleleReads_singleSNPs_10x.txt", quote = FALSE, sep = "\t")

write.table(readCounts10x_lwed, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_AlleleReads_singleSNPs_10x.txt", quote = FALSE, sep = "\t")

write.table(readCounts10x_simp, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_AlleleReads_singleSNPs_10x.txt", quote = FALSE, sep = "\t")
```

Genotyping is accomplished using the polyGen function. The genotyping algorithm is described in McKinney et al. 2018 and is a maximum likelihood algorithm capable of genotyping any number of alleles and ploidy per locus. This allows genoyping of single SNPs as well as microhaplotypes, and loci with elevated ploidy.

Two arguments are required for polyGen, the locusTable and alleleReads files output by AmpliconReadCounter.

Optional arguments for polyGen are:

-   p_thresh - threshold p-value for likelihood ratio test (default 0.05)
-   epsilon - error rate for genotyping model (default 0.01)

**NOTE** that only primer probe reads are used in genotyping!

Note also that I'm making three genotype files, one for all loci vs. all samples and two species-specific sets - this allows our genotyping metrics to be more accurate.

### Full set genos

```{r}
#load locus table and 10x allele reads file
fullSet_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

fullSet_singleSNP_alleleReads_10x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm, adjust "0" formatting
fullSet_polyGenResults_singleSNP_10x <- polyGen(fullSet_singleSNP_locusTable, fullSet_singleSNP_alleleReads_10x)

# write results
write.table(fullSet_polyGenResults_singleSNP_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

### Subsets

#### pos samples only

```{r Genotyping}
#load locus table and 10x allele reads file
posDual_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

posDual_singleSNP_alleleReads_10x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
posDual_polyGenResults_singleSNP_10x <- polyGen(posDual_singleSNP_locusTable, posDual_singleSNP_alleleReads_10x)

#write results
write.table(posDual_polyGenResults_singleSNP_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

#### LWED

```{r Genotyping}
#load locus table and 10x allele reads file
LWED_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

LWED_singleSNP_alleleReads_10x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
LWED_polyGenResults_singleSNP_10x <- polyGen(LWED_singleSNP_locusTable, LWED_singleSNP_alleleReads_10x)

#write results
write.table(LWED_polyGenResults_singleSNP_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

#### SIMP

```{r}
#load locus table and 10x allele reads file
SIMP_singleSNP_locusTable <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

SIMP_singleSNP_alleleReads_10x <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
SIMP_polyGenResults_singleSNP_10x <- polyGen(SIMP_singleSNP_locusTable, SIMP_singleSNP_alleleReads_10x)

#write results
write.table(SIMP_polyGenResults_singleSNP_10x,"./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

# 8 Data summaries

## 8.1 Locus summaries

### Summarize single SNP results for loci

The summarizeGTscore command generates summary data for each locus in table form. The summary data includes genotype rate, average read depth, minor (least frequent) allele frequency, major (most frequent) allele frequency, alleles per locus, and frequency per allele. Minor allele frequency is a common metric for filtering loci that are likely to be uninformative for population genetics; however, loci with haplotype alleles may have an allele with very low frequency but still have appreciable frequency at multiple other alleles. Because of this, the major allele frequency is included in output, as well as the observed frequencies for all alleles at a given locus.

**gtscore0x**

```{r}
fullSet_singleSNP_alleleReads <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

posDual_singleSNP_alleleReads <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

LWED_singleSNP_alleleReads <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

SIMP_singleSNP_alleleReads <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#summarize single SNP results
fullSet_singleSNP_summary_0x <- summarizeGTscore(fullSet_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_polyGenResults_singleSNP_0x)

posDual_singleSNP_summary_0x <- summarizeGTscore(posDual_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_polyGenResults_singleSNP_0x)

LWED_singleSNP_summary_0x <- summarizeGTscore(LWED_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_polyGenResults_singleSNP_0x)

SIMP_singleSNP_summary_0x <- summarizeGTscore(SIMP_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_polyGenResults_singleSNP_0x)

#write results
write.table(fullSet_singleSNP_summary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(posDual_singleSNP_summary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(LWED_singleSNP_summary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(SIMP_singleSNP_summary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

**gtscore10x**

```{r}
#summarize single SNP results
fullSet_singleSNP_summary_10x <- summarizeGTscore(fullSet_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_polyGenResults_singleSNP_10x)

posDual_singleSNP_summary_10x <- summarizeGTscore(posDual_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_polyGenResults_singleSNP_10x)

LWED_singleSNP_summary_10x <- summarizeGTscore(LWED_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_polyGenResults_singleSNP_10x)

SIMP_singleSNP_summary_10x <- summarizeGTscore(SIMP_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_polyGenResults_singleSNP_10x)

#write results
write.table(fullSet_singleSNP_summary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(posDual_singleSNP_summary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(LWED_singleSNP_summary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(SIMP_singleSNP_summary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

### Master locus summary

Generate locus summary files with single SNP summaries + GTscore locus summaries created earlier

```{r}
# per-locus read summaries
ls_posDual <- read.table("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(
    sampleSet = "posSamples"
  )

ls_lwed <- read.table("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(
    sampleSet = "lwedSamples"
  )

ls_simp <- read.table("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(
    sampleSet = "simpSamples"
  )

# gtscore0x per-locus depth/geno/etc summaries
ss_posDual_0x <- posDual_singleSNP_summary_0x %>%
  mutate(
    sampleSet = "posSamples"
  )

ss_lwed_0x <- LWED_singleSNP_summary_0x %>%
  mutate(
    sampleSet = "lwedSamples"
  )

ss_simp_0x <- SIMP_singleSNP_summary_0x %>%
  mutate(
    sampleSet = "simpSamples"
  )

# gtscore10x per-locus depth/geno/etc summaries
ss_posDual_10x <- posDual_singleSNP_summary_10x %>%
  mutate(
    sampleSet = "posSamples"
  )

ss_lwed_10x <- LWED_singleSNP_summary_10x %>%
  mutate(
    sampleSet = "lwedSamples"
  )

ss_simp_10x <- SIMP_singleSNP_summary_10x %>%
  mutate(
    sampleSet = "simpSamples"
  )

# master summary
ls_combo <- rbind(ls_posDual, ls_lwed, ls_simp)
ss_combo_0x <- rbind(ss_posDual_0x, ss_lwed_0x, ss_simp_0x) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))
ss_combo_10x <- rbind(ss_posDual_10x, ss_lwed_10x, ss_simp_10x) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

master_locusSummary <- ls_combo %>%
  merge(., ss_combo_0x, by.x = c("Locus", "sampleSet"), by.y = c("Locus_ID", "sampleSet")) %>%
  merge(., ss_combo_10x[, c("Locus_ID", "sampleSet", "GenotypeRate", "minAF", "majAF", "allFreqs", "conScore")], by.x = c("Locus", "sampleSet"), by.y = c("Locus_ID", "sampleSet"), suffixes = c("_0x", "_10x")) %>%
  mutate(
    metricUse = case_when(
      str_detect(Locus, "LWED") & sampleSet == "lwedSamples" ~ "for_panelMetrics",
      str_detect(Locus, "SIMP") & sampleSet == "simpSamples" ~ "for_panelMetrics",
      sampleSet == "posSamples" ~ "for_panelMetrics",
      .default = "for_spMetrics"
    )
  )

write.csv(master_locusSummary, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/summaryFiles/master_locusSummary.csv", row.names = F)
```

### Plots

plot genotype rate

```{r locus genotype rate, warning=FALSE}
# gtscore0x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() +
  geom_histogram(aes(x = GenotypeRate_0x), binwidth = 0.03) +
  xlim(-0.01,1.01) +
  labs(title = "GTscore 0x: Locus genotype rate",
       x = "Genotype rate",
       y="count") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))

# gtscore10x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() +
  geom_histogram(aes(x = GenotypeRate_10x), binwidth = 0.03) +
  xlim(-0.01,1.01) +
  labs(title = "GTscore 10x: Locus genotype rate",
       x = "Genotype rate",
       y="count") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))
```

plot average read depth for single SNP data

```{r locus read depth}
# gtscore0x/10x (avg read depth is sampe for both)
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_histogram(aes(x = AvgReadDepth), binwidth = 1)+
  labs(title = "Average Read Depth per SNP",
       x = "Average Read Depth",
       y = "Count") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5), 
        plot.subtitle=element_text(hjust = 0.5))
```

plot genotype rate relative to average depth

```{r locus genotype rate vs read depth}
# gtscore0x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_point(aes(x = AvgReadDepth,
                 y = GenotypeRate_0x)) +
  ylim(0, 1) +
  labs(title = "0x Genotype Rate vs Average Depth per SNP",
       x = "Average Depth",
       y = "Genotype Rate") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))

# gtscore10x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_point(aes(x = AvgReadDepth,
                 y = GenotypeRate_10x)) +
  ylim(0, 1) +
  labs(title = "10x Genotype Rate vs Average Depth per SNP",
       x = "Average Depth",
       y = "Genotype Rate") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))
```

plot distribution of minor allele frequency

```{r histogram of MAF, warning=FALSE}
# gtscore0x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_histogram(aes(x = minAF_0x),
                 binwidth = 0.01)+
  labs(title = "0x Minor Allele Frequency Single SNP",
       x = "Minor Allele Frequency",
       y = "Count") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        plot.subtitle = element_text(hjust = 0.5))

# gtscore10x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_histogram(aes(x = minAF_10x),
                 binwidth = 0.01)+
  labs(title = "10x Minor Allele Frequency Single SNP",
       x = "Minor Allele Frequency",
       y = "Count") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        plot.subtitle = element_text(hjust = 0.5))
```

plot distribution of major allele frequency

```{r histogram of MajAF, warning=FALSE}
# gtscore0x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_histogram(aes(x = majAF_0x),
                 binwidth = 0.01)+
  labs(title = "0x Major Allele Frequency Single SNP",
       x = "Major Allele Frequency",
       y = "Count") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        plot.subtitle = element_text(hjust = 0.5))

# gtscore10x
master_locusSummary %>%
  filter(metricUse == "for_panelMetrics") %>%
  ggplot() + 
  geom_histogram(aes(x = majAF_10x),
                 binwidth = 0.01)+
  labs(title = "10x Major Allele Frequency Single SNP",
       x = "Major Allele Frequency",
       y = "Count") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5), 
        plot.subtitle = element_text(hjust = 0.5))
```

## 8.2 Sample summaries

### Summarize single SNP summaries for samples

**gtscore 0x**

```{r}
fullSet_singleSNP_sampleSummary_0x <- summarizeSamples(fullSet_polyGenResults_singleSNP_0x, fullSet_singleSNP_alleleReads_0x)

posDual_singleSNP_sampleSummary_0x <- summarizeSamples(posDual_polyGenResults_singleSNP_0x, posDual_singleSNP_alleleReads_0x)

LWED_singleSNP_sampleSummary_0x <- summarizeSamples(LWED_polyGenResults_singleSNP_0x, LWED_singleSNP_alleleReads_0x)

SIMP_singleSNP_sampleSummary_0x <- summarizeSamples(SIMP_polyGenResults_singleSNP_0x, SIMP_singleSNP_alleleReads_0x)

write.csv(fullSet_singleSNP_sampleSummary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_singleSNP_sampleSummary_0x.csv", row.names = F)

write.csv(posDual_singleSNP_sampleSummary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_singleSNP_sampleSummary_0x.csv", row.names = F)

write.csv(LWED_singleSNP_sampleSummary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_singleSNP_sampleSummary_0x.csv", row.names = F)

write.csv(SIMP_singleSNP_sampleSummary_0x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_singleSNP_sampleSummary_0x.csv", row.names = F)
```

**gtscore 10x**

```{r}
fullSet_singleSNP_sampleSummary_10x <- summarizeSamples(fullSet_polyGenResults_singleSNP_10x, fullSet_singleSNP_alleleReads_10x)

posDual_singleSNP_sampleSummary_10x <- summarizeSamples(posDual_polyGenResults_singleSNP_10x, posDual_singleSNP_alleleReads_10x)

LWED_singleSNP_sampleSummary_10x <- summarizeSamples(LWED_polyGenResults_singleSNP_10x, LWED_singleSNP_alleleReads_10x)

SIMP_singleSNP_sampleSummary_10x <- summarizeSamples(SIMP_polyGenResults_singleSNP_10x, SIMP_singleSNP_alleleReads_10x)

write.csv(fullSet_singleSNP_sampleSummary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_singleSNP_sampleSummary_10x.csv", row.names = F)

write.csv(posDual_singleSNP_sampleSummary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/posDual_singleSNP_sampleSummary_10x.csv", row.names = F)

write.csv(LWED_singleSNP_sampleSummary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_singleSNP_sampleSummary_10x.csv", row.names = F)

write.csv(SIMP_singleSNP_sampleSummary_10x, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_singleSNP_sampleSummary_10x.csv", row.names = F)
```

### Master sample summary

Sample metrics for genotypes and such are based on species sample/loci sets -

-   LWED sample metrics are based on dual loci + lwed-specific loci
-   SIMP sample metrics are based on dual loci + simp-specific loci
-   negCtrl sample metrics are based on all loci

```{r}
# import individual summaries from AmpliconReadCounter.pl output
is_neg <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/fullSet_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  filter(!Sample %in% lwed$sampleID) %>%
  filter(!Sample %in% simp$sampleID)

is_lwed <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/LWED_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))

is_simp <- read.delim("./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/SIMP_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))

# gtscore0x sample summary (from above)
sampleSum_fullSet_0x <- fullSet_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "fullSet")

sampleSum_neg_0x <- sampleSum_fullSet_0x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)

sampleSum_lwed_0x <- LWED_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "lwedSet")

sampleSum_simp_0x <- SIMP_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "simpSet")

# gtscore10x sample summary (from above)
sampleSum_fullSet_10x <- fullSet_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "fullSet")

sampleSum_neg_10x <- sampleSum_fullSet_10x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)

sampleSum_lwed_10x <- LWED_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "lwedSet")

sampleSum_simp_10x <- SIMP_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "simpSet")

# combos
indSum_combo <- rbind(is_neg, is_lwed, is_simp) %>%
  arrange(Sample)

sampleSum_combo_0x <- sampleSum_neg_0x %>%
  rbind(sampleSum_lwed_0x, sampleSum_simp_0x) %>%
  arrange(sample)

sampleSum_combo_10x <- sampleSum_neg_10x %>%
  rbind(sampleSum_lwed_10x, sampleSum_simp_10x) %>%
  arrange(sample) %>%
  select(-lociSet)

# master summary
master_sampleSummary <- indSum_combo %>%
  merge(., sampleSum_combo_0x, by.x = "Sample", by.y = "sample") %>%
  merge(., sampleSum_combo_10x, by.x = "Sample", by.y = "sample", suffixes = c("_0x", "_10x")) %>%
  relocate(lociSet, .after = Sample) %>%
  dplyr::rename("sampleID" = "Sample") %>%
  # add metadata
  merge(., md[, c("sampleID", "sampleID_md")], by = "sampleID") %>%
  relocate(sampleID_md, .after = sampleID)

write.csv(master_sampleSummary, "./00_tamRun1_optimization/04_run1GTscore_bp3errorRemoved/summaryFiles/master_sampleSummary.csv", row.names = F)
```

### Plots

plot histogram of genotype rate

```{r histogram of genotype rate, warning=FALSE}
# gtscore0x
master_sampleSummary %>%
  ggplot() +
  geom_histogram(aes(x = GenotypeRate_0x), binwidth = 0.01) +
  xlim(-0.01, 1.01) +
  labs(title = "0x Sample Genotype Rate",
       x = "Genotype Rate",
       y = "Count") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))

# gtscore10x
master_sampleSummary %>%
  ggplot() +
  geom_histogram(aes(x = GenotypeRate_10x), binwidth = 0.01) +
  xlim(-0.01, 1.01) +
  labs(title = "10x Sample Genotype Rate",
       x = "Genotype Rate",
       y = "Count") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))
```

plot histogram of heterozygosity

```{r histogram of heterozygosity, warning=FALSE}
# gtscore0x
master_sampleSummary %>%
  ggplot() +
  geom_histogram(aes(x = Heterozygosity_0x), binwidth = 0.03) + 
  xlim(-0.01, 1.01)+
  labs(title = "0x Sample Heterozygosity",
       x = "Heterozygosity",
       y = "Count")+
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle=element_text(hjust = 0.5))

# gtscore10x
master_sampleSummary %>%
  ggplot() +
  geom_histogram(aes(x = Heterozygosity_10x), binwidth = 0.03) + 
  xlim(-0.01, 1.01)+
  labs(title = "10x Sample Heterozygosity",
       x = "Heterozygosity",
       y = "Count")+
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle=element_text(hjust = 0.5))
```

plot genotype rate vs primer probe reads

```{r genotype rate vs primer probe reads}
#dashed line added at 90% genotype rate, this is not a strict threshold, just a goal to aim for

# gtscore0x
master_sampleSummary %>%
  separate(sampleID_md, into = c("id", "species", "animalID", "sampleType"), sep = "_") %>%
  ggplot(aes(x = Primer.Probe.Reads,
             y = GenotypeRate_0x,
             color = sampleType)) +
  geom_point(stat = "identity") +
  geom_smooth(method = "loess") +
  labs(title = "0x Genotype Rate vs Total Reads per Sample",
       x = "Primer Probe Reads",
       y = "Genotype Rate")+
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  geom_hline(yintercept = 0.9, lty = "dashed")

# gtscore10x
master_sampleSummary %>%
  separate(sampleID_md, into = c("id", "species", "animalID", "sampleType"), sep = "_") %>%
  ggplot(aes(x = Primer.Probe.Reads,
             y = GenotypeRate_10x,
             color = sampleType)) +
  geom_point(stat = "identity") +
  geom_smooth(method = "loess") +
  labs(title = "10x Genotype Rate vs Total Reads per Sample",
       x = "Primer Probe Reads",
       y = "Genotype Rate")+
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  geom_hline(yintercept = 0.9, lty = "dashed")
```

Samples with unusually high heterozygosity may be contaminated or have elevated ploidy. [[Given chimerism, however, this isn't super helpful for blood samples]]

plot heterozygosity vs primer probe reads

```{r heterozygosity vs primer probe reads}
# gtscore0x
master_sampleSummary %>%
  separate(sampleID_md, into = c("id", "species", "animalID", "sampleType"), sep = "_") %>%
  ggplot() +
  geom_point(aes(x = Primer.Probe.Reads,
                 y = Heterozygosity_0x,
                 color = sampleType))+
  labs(title = "0x Heterozygosity vs Total Reads per Sample",
       x = "Primer Probe Reads",
       y = "Heterozygosity")+
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))+
  geom_hline(yintercept = 0.3, lty = "dashed")

# gtscore10x
master_sampleSummary %>%
  separate(sampleID_md, into = c("id", "species", "animalID", "sampleType"), sep = "_") %>%
  ggplot() +
  geom_point(aes(x = Primer.Probe.Reads,
                 y = Heterozygosity_10x,
                 color = sampleType))+
  labs(title = "10x Heterozygosity vs Total Reads per Sample",
       x = "Primer Probe Reads",
       y = "Heterozygosity")+
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5))+
  geom_hline(yintercept = 0.3, lty = "dashed")
```
