---
title: "paper1_nanoporeValidation"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/")
```

NOTE TO SELF (18 april 2024) -- I'm in the process of re-organizing this file to reflect all of the analyses I did for paper1, including all of the initial stuff with gtscore vs. gtseq, qc filters, etc. Some of that info is in the archived version of this document (esp the gtseq stuff which I deleted from here before I decided I should actually keep it here). I decided on keeping sections 1-4; just need to put stuff where it needs to go. BUT - the clean rmd to share with collaborators for paper1 is good to go, and ultimately this doc can wait bc I have more pressing matters to attend to... like my dissertation.

# 1 Overview

This set of scripts is for tamGenetics_paper1, which focuses on validating the tamGenetics SNP panel on the Nanopore sequencing platform by comparing to Illumina sequencing results. Results from each sequencing platform are from the tamRun4 sample set, which includes paired blood, fecal, and hair samples for 30 individuals (n = 15 LWED, n = 15 SIMP) + 6 negative controls sequenced at 208 loci (including SPECIESID, SEXID, and INDID loci sets). The present set of analyses focuses only on **blood samples** + the **140 dual-species loci (n = 123 INDID, n = 5 SEXID, n = 12 SPECIESID)**.

Analyses include the following:

-   An overview of total read counts for Illumina vs. Nanopore results
-   An overview of genotype success for Illumina vs. Nanopore results
-   Identification of the optimal subset of samples/loci (using Illumina data) to include to compare Illumina vs. Nanopore genotypes, including subsets where:
    -   Each locus within the subset is genotyped for 100% of the sample subset
    -   Each locus within the subset is genotyped for 90% of the sample subset
-   Genotype consistency between Illumina and Nanopore results for the identified sample/loci subsets
-   Quality score comparison between Illumina and Nanopore results
-   Comparisons of genotype consistency after filtering for different quality scores for Illumina vs. Nanopore results
-   Downsampling tests, where coverage per sample per locus is set at 6x, 10x, 15x, or 30x, and subsequent comparisons of genotype success and consistency for Illumina vs. Nanopore results

# 2 Packages

```{r, message=FALSE}
library(gsubfn, quietly = T)
library(kableExtra, quietly = T)
library(purrr, quietly = T)
library(readxl, quietly = T)
library(tidyverse, quietly = T)

source("./GTscore_sourceScripts/GTscore_modified.R")
```

# 3 Data

## 3.1 Metadata

Our sample dataset is composed of **30 tamarin blood samples**,
including 15 saddleback tamarins (*Leontocebus weddelli*; n = 8
females, n = 7 males) and 15 emperor tamarins (*Saguinus imperator*; n
= 8 females, n = 7 males).

```{r, echo=FALSE}
# Metadata w/both Illumina and Nanopore sampleIDs/sampleFiles
md_tamRun4 <- read.csv("./04_tamRun4/illONT_tamRun4_metadata.csv")

# Metadata subsets by sample type
md_tamRun4_blood <- md_tamRun4 %>%
  filter(sampleType == "blood")

md_tamRun4_fecal <- md_tamRun4 %>%
  filter(sampleType == "fecal")

md_tamRun4_hair <- md_tamRun4 %>%
  filter(sampleType == "hair")
```

```{r, echo=FALSE}
# For html
md_tamRun4_blood %>%
  dplyr::rename("animalID_sampleType" = "sampleID_unique") %>%
  select(c(animalID_sampleType, sampleID_ill, sampleID_ont, species, sex)) %>%
  arrange(sampleID_ill) %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 3.2 Loci

Our loci dataset is composed of **140 dual-species SNPs**, including SNPs
used for the identification of individuals (INDID; n = 123), sex (SEXID;
n = 5), and species (SPECIESID; n = 12). All 140 loci are informative for
both tamarin species.

```{r, echo=FALSE}
primerList.v3 <- read.csv("./primers/03_lociChoices/tamGenetics_primerList_v3.csv")

primerList.v3_lociDual <- primerList.v3 %>%
  select(locus) %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  arrange(locus)

primerList.v3_lociDual_bySet <- data.frame(
  lociSet = c("INDID", "SEXID", "SPECIESID", "TOTAL"),
  count = c(
    sum(str_detect(primerList.v3_lociDual$locus, "INDID")),
    sum(str_detect(primerList.v3_lociDual$locus, "SEXID")),
    sum(str_detect(primerList.v3_lociDual$locus, "SPECIESID")),
    nrow(primerList.v3_lociDual)
  )
)
```

```{r, echo=FALSE}
# For html
primerList.v3_lociDual_bySet %>%
  kbl() %>%
  kable_styling(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato")
```

## 3.3 Allele reads

The allele reads files imported here are outputs from the GTscore AmpliconReadCounter.pl script used in GTscorePipeline_tamRun4_ill.Rmd and GTscorePipeline_tamRun4_ont.Rmd for Illumina and Nanopore datasets, respectively. I've included both "qcUnfiltered" and "qcFiltered" versions, where "qcUnfiltered" allele reads are derived from the original fastq files and "qcFiltered" allele reads are derived from fastq files that were subjected to a q30 quality filter.

Below I'm reformatting the files slightly to 1) revert loci names back to their original names (GTscore appends the SNP position to the locus name), and 2) assign a common sampleID to corresponding samples between Illumina and Nanopore datasets.

### gtscore

#### qcUnfiltered

**Illumina**

```{r, echo=FALSE}
alleleReads_ill_gtscore_samplesAll_lociDual <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  select(-sampleID) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()
```

**Nanopore**

```{r}
alleleReads_ont_gtscore_samplesAll_lociDual <- read.delim("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()
```

#### qcFiltered

Sequences filtered to q30

**Illumina**

```{r, echo=FALSE}
alleleReads_ill_qcFiltered_gtscore_samplesAll_lociDual <- read.table("./04_tamRun4/04_run4QualityFilterTests/03_run4Analyses_qcFiltered/illONT_qcFiltered_fullSet_AlleleReads_singleSNPs.txt", header = T) %>%
  select(starts_with("tamRun4")) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  separate("sampleID", into = c("sampleID", "qc"), sep = "_") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  mutate(sampleID = str_c(sampleID_unique, qc, sep = "_")) %>%
  select(-sampleID_unique, -qc) %>%
  arrange(sampleID) %>%
  column_to_rownames("sampleID") %>%
  t() %>%
  as.data.frame()
```

**Nanopore**

```{r}
alleleReads_ont_qcFiltered_gtscore_samplesAll_lociDual <- read.table("./04_tamRun4/04_run4QualityFilterTests/03_run4Analyses_qcFiltered/illONT_qcFiltered_fullSet_AlleleReads_singleSNPs.txt", header = T) %>%
  select(starts_with("barcode")) %>%
  # reformat loci names
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # update sampleID
  rownames_to_column("sampleID") %>%
  separate("sampleID", into = c("sampleID", "qc"), sep = "_") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  mutate(sampleID = str_c(sampleID_unique, qc, sep = "_")) %>%
  select(-sampleID_unique, -qc) %>%
  arrange(sampleID) %>%
  column_to_rownames("sampleID") %>%
  t() %>%
  as.data.frame()
```

### gtseq

#### qcFiltered

```{r}

```

#### qcUnfiltered

```{r}

```

## 3.4 Genotypes

### gtscore0x

#### qcUnfiltered

**Illumina**

```{r}
genos_ill_gtscore0x_samplesAll_lociDual <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # switch sampleID to animalID_sampleType
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill") %>%
  select(-sampleID_ill) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# subsets by sample type
genos_ill_gtscore0x_samplesBlood_lociDual <- genos_ill_gtscore0x_samplesAll_lociDual %>%
  select(contains("blood"))

genos_ill_gtscore0x_samplesFecal_lociDual <- genos_ill_gtscore0x_samplesAll_lociDual %>%
  select(contains("fecal"))

genos_ill_gtscore0x_samplesHair_lociDual <- genos_ill_gtscore0x_samplesAll_lociDual %>%
  select(contains("hair"))
```

**Nanopore**

```{r}
# full set
genos_ont_gtscore0x_samplesAll_lociDual <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # switch sampleID to animalID_sampleType
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont") %>%
  select(-sampleID_ont) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# sample-type subsets
genos_ont_gtscore0x_samplesBlood_lociDual <- genos_ont_gtscore0x_samplesAll_lociDual %>%
  select(contains("blood"))

genos_ont_gtscore0x_samplesFecal_lociDual <- genos_ont_gtscore0x_samplesAll_lociDual %>%
  select(contains("fecal"))

genos_ont_gtscore0x_samplesHair_lociDual <- genos_ont_gtscore0x_samplesAll_lociDual %>%
  select(contains("hair"))
```

#### qcFiltered

**Illumina**

```{r}

```

**Nanopore**

```{r}

```

### gtscore10x

#### qcUnfiltered

**Illumina**

```{r}
genos_ill_gtscore10x_samplesAll_lociDual <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # switch sampleID to animalID_sampleType
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill") %>%
  select(-sampleID_ill) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# subsets by sample type
genos_ill_gtscore10x_samplesBlood_lociDual <- genos_ill_gtscore10x_samplesAll_lociDual %>%
  select(contains("blood"))

genos_ill_gtscore10x_samplesFecal_lociDual <- genos_ill_gtscore10x_samplesAll_lociDual %>%
  select(contains("fecal"))

genos_ill_gtscore10x_samplesHair_lociDual <- genos_ill_gtscore10x_samplesAll_lociDual %>%
  select(contains("hair"))
```

**Nanopore**

```{r}
# full set
genos_ont_gtscore10x_samplesAll_lociDual <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # switch sampleID to animalID_sampleType
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont") %>%
  select(-sampleID_ont) %>%
  arrange(sampleID_unique) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# sample-type subsets
genos_ont_gtscore10x_samplesBlood_lociDual <- genos_ont_gtscore10x_samplesAll_lociDual %>%
  select(contains("blood"))

genos_ont_gtscore10x_samplesFecal_lociDual <- genos_ont_gtscore10x_samplesAll_lociDual %>%
  select(contains("fecal"))

genos_ont_gtscore10x_samplesHair_lociDual <- genos_ont_gtscore10x_samplesAll_lociDual %>%
  select(contains("hair"))
```

#### qcFiltered

**Illumina**

```{r}

```

**Nanopore**

```{r}

```

### gtseq

Note that gtseq automatically includes a 10x coverage cutoff as part of its genotyping algorithm.

#### qcUnfiltered

**Illumina**

```{r}

```

**Nanopore**

```{r}

```

#### qcFiltered

**Illumina**

```{r}

```

**Nanopore**

```{r}

```

## 3.5 Whole-genome sequenced samples

Animal IDs that have whole-genome sequences available:

```{r, echo=FALSE, warning=FALSE, message=FALSE}
wholeGenomeSeqs <- read_excel("./04_tamRun4/DNASubmissionForm_MEW.xlsx",
                              range = "A17:R44") %>%
  as.data.frame() %>%
  dplyr::rename("animalID" = "Animal ID") %>%
  select(animalID)
```

```{r, echo=FALSE}
# For html
wholeGenomeSeqs %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 3 Data prep

## 3.1 Long-format allele reads & sampleType subsets

### gtscore

#### qcUnfiltered

**Illumina**

```{r, echo=FALSE}
# all samples
alleleReads_ill_gtscore_samplesAll_lociDual_l <- alleleReads_ill_gtscore_samplesAll_lociDual %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
  separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
  select(-readCounts_a1, -readCounts_a2) %>%
  mutate(
    seqPlatform = "illumina",
    animalID = sub('[_][^_]+$', '', sampleID),
    sampleType = sub('^[^_]*[_]?', '', sampleID)
    )

# all samples, no negs
alleleReads_ill_gtscore_samplesNoNegs_lociDual_l <- alleleReads_ill_gtscore_samplesAll_lociDual_l %>%
  filter(!str_detect(sampleID, "Neg"))

# subsets by sample type
alleleReads_ill_gtscore_samplesBlood_lociDual_l <- alleleReads_ill_gtscore_samplesNoNegs_lociDual_l %>%
  filter(sampleType == "blood")

alleleReads_ill_gtscore_samplesFecal_lociDual_l <- alleleReads_ill_gtscore_samplesNoNegs_lociDual_l %>%
  filter(sampleType == "fecal")

alleleReads_ill_gtscore_samplesHair_lociDual_l <- alleleReads_ill_gtscore_samplesNoNegs_lociDual_l %>%
  filter(sampleType == "hair")
```

**Nanopore**

```{r, echo=FALSE}
# all samples
alleleReads_ont_gtscore_samplesAll_lociDual_l <- alleleReads_ont_gtscore_samplesAll_lociDual %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
  separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
  select(-readCounts_a1, -readCounts_a2) %>%
  mutate(
    seqPlatform = "nanopore",
    animalID = sub('[_][^_]+$', '', sampleID),
    sampleType = sub('^[^_]*[_]?', '', sampleID)
    )

## all samples, no negs
alleleReads_ont_gtscore_samplesNoNegs_lociDual_l <- alleleReads_ont_gtscore_samplesAll_lociDual_l %>%
  filter(!str_detect(sampleID, "Neg"))

## subsets by sample type
alleleReads_ont_gtscore_samplesBlood_lociDual_l <- alleleReads_ont_gtscore_samplesNoNegs_lociDual_l %>%
  filter(sampleType == "blood")

alleleReads_ont_gtscore_samplesFecal_lociDual_l <- alleleReads_ont_gtscore_samplesNoNegs_lociDual_l %>%
  filter(sampleType == "fecal")

alleleReads_ont_gtscore_samplesHair_lociDual_l <- alleleReads_ont_gtscore_samplesNoNegs_lociDual_l %>%
  filter(sampleType == "hair")
```

#### qcFiltered

**Illumina**

```{r, echo=FALSE}

```

**Nanopore**

```{r, echo=FALSE}

```

### gtseq

#### qcUnfiltered

**Illumina**

```{r}

```

**Nanopore**

```{r}

```

#### qcFiltered

**Illumina**

```{r}

```

**Nanopore**

```{r}

```

## 3.2 Long-format genotypes & sample subsets

### gtscore0x

#### qcUnfiltered

```{r}
genos_ill_gtscore0x_samplesAll_lociDual_l <- genos_ill_gtscore0x_samplesAll_lociDual %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ill")

genos_ont_gtscore0x_samplesAll_lociDual_l <- genos_ont_gtscore0x_samplesAll_lociDual %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ont")
```

#### qcFiltered

### gtscore10x

#### qcUnfiltered

```{r}
genos_ill_gtscore10x_samplesAll_lociDual_l <- genos_ill_gtscore10x_samplesAll_lociDual %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ill")

genos_ont_gtscore10x_samplesAll_lociDual_l <- genos_ont_gtscore10x_samplesAll_lociDual %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ont")
```

#### qcFiltered

### gtseq

#### qcUnfiltered

#### qcFiltered

# 4 Overview: Illumina vs. ONT

## 4.1 Coverage

### gtscore

#### qcUnfiltered

On-target reads were substantially higher for Illumina vs. Nanopore, including total reads, mean reads per sample for each sample type, and mean reads per locus per sample for each sample type.

```{r, echo=FALSE}
# total reads for the run
table_total_reads <- alleleReads_ill_gtscore_samplesAll_lociDual_l %>%
  rbind(., alleleReads_ont_gtscore_samplesAll_lociDual_l) %>%
  group_by(seqPlatform) %>%
  summarise(totalReads = sum(readCounts_sum)) %>%
  pivot_wider(names_from = seqPlatform,
              values_from = totalReads)

# mean total reads per sample for each sample type
table_mean_reads_perSample <- alleleReads_ill_gtscore_samplesNoNegs_lociDual_l %>%
  rbind(., alleleReads_ont_gtscore_samplesNoNegs_lociDual_l) %>%
  group_by(seqPlatform, sampleType, sampleID) %>%
  summarise(sum_reads_perSample = sum(readCounts_sum, na.rm = T)) %>%
  as.data.frame() %>%
  group_by(seqPlatform, sampleType) %>%
  summarise(mean_reads_perSample = ceiling(mean(sum_reads_perSample))) %>%
  as.data.frame() %>%
  pivot_wider(names_from = seqPlatform,
              values_from = mean_reads_perSample)

# mean reads per locus per sample for each sample type
table_mean_reads_perLocus <- alleleReads_ill_gtscore_samplesNoNegs_lociDual_l %>%
  rbind(., alleleReads_ont_gtscore_samplesNoNegs_lociDual_l) %>%
  group_by(seqPlatform, sampleType) %>%
  summarise(mean_reads_perLocus = ceiling(mean(readCounts_sum, na.rm = T))) %>%
  as.data.frame() %>%
  pivot_wider(names_from = seqPlatform,
              values_from = mean_reads_perLocus)

```

```{r, echo=FALSE}
# for html
table_total_reads %>%
  mutate(illumina = prettyNum(illumina, big.mark = ","),
         nanopore = prettyNum(nanopore, big.mark = ",")) %>%
  kbl(caption = "Total on-target read counts for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

table_mean_reads_perSample %>%
  mutate(illumina = prettyNum(illumina, big.mark = ","),
         nanopore = prettyNum(nanopore, big.mark = ",")) %>%
  kbl(caption = "Mean on-target reads per sample across sample types for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

table_mean_reads_perLocus %>%
  mutate(illumina = prettyNum(illumina, big.mark = ","),
         nanopore = prettyNum(nanopore, big.mark = ",")) %>%
  kbl(caption = "Mean on-target reads per locus, per sample for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### qcFiltered

```{r}

```

### gtseq

#### qcUnfiltered

```{r}

```

#### qcFiltered

```{r}

```

## 4.2 Geno success


## 4.3 Geno consistency

### gtscore0x

```{r}
# dataframe
compareGenos_gtscore0x_samplesAll <- genos_ill_gtscore0x_samplesAll_lociDual_l %>%
  merge(., genos_ont_gtscore0x_samplesAll_lociDual_l, by = c("sampleID", "locus")) %>%
  mutate(
    genoMatch = case_when(
      is.na(genos_ill) | is.na(genos_ont) ~ NA,
      genos_ill == genos_ont ~ "yes",
      .default = "no"
    )
  ) %>%
  # ditch negs
  filter(!str_detect(sampleID, "Neg"))

compareGenos_gtscore0x_samplesBlood <- compareGenos_gtscore0x_samplesAll %>%
  filter(str_detect(sampleID, "blood"))

# summary tables
compareGenos_gtscore0x_samplesAll_table <- data.frame(
  genosMismatch = sum(compareGenos_gtscore0x_samplesAll$genoMatch == "no", na.rm = T),
  genosMatch = sum(compareGenos_gtscore0x_samplesAll$genoMatch == "yes", na.rm = T),
  genosNA = sum(is.na(compareGenos_gtscore0x_samplesAll$genoMatch)),
  totalGenos = nrow(compareGenos_gtscore0x_samplesAll)
) %>%
  mutate(
    genosCommon = genosMismatch + genosMatch,
    propMismatch = round(genosMismatch/genosCommon, 2),
    propMatch = round(genosMatch/genosCommon, 2)
  ) %>%
  relocate(c(propMismatch, propMatch, genosMismatch, genosMatch, genosNA, genosCommon, totalGenos))

compareGenos_gtscore0x_samplesBlood_table <- data.frame(
  genosMismatch = sum(compareGenos_gtscore0x_samplesBlood$genoMatch == "no", na.rm = T),
  genosMatch = sum(compareGenos_gtscore0x_samplesBlood$genoMatch == "yes", na.rm = T),
  genosNA = sum(is.na(compareGenos_gtscore0x_samplesBlood$genoMatch)),
  totalGenos = nrow(compareGenos_gtscore0x_samplesBlood)
) %>%
  mutate(
    genosCommon = genosMismatch + genosMatch,
    propMismatch = round(genosMismatch/genosCommon, 2),
    propMatch = round(genosMatch/genosCommon, 2)
  ) %>%
  relocate(c(propMismatch, propMatch, genosMismatch, genosMatch, genosNA, genosCommon, totalGenos))
```

### gtscore10x

```{r}
# dataframe
compareGenos_gtscore10x_samplesAll <- genos_ill_gtscore10x_samplesAll_lociDual_l %>%
  merge(., genos_ont_gtscore10x_samplesAll_lociDual_l, by = c("sampleID", "locus")) %>%
  mutate(
    genoMatch = case_when(
      is.na(genos_ill) | is.na(genos_ont) ~ NA,
      genos_ill == genos_ont ~ "yes",
      .default = "no"
    )
  ) %>%
  # ditch negs
  filter(!str_detect(sampleID, "Neg"))

compareGenos_gtscore10x_samplesBlood <- compareGenos_gtscore10x_samplesAll %>%
  filter(str_detect(sampleID, "blood"))

# summary tables
compareGenos_gtscore10x_samplesAll_table <- data.frame(
  genosMismatch = sum(compareGenos_gtscore10x_samplesAll$genoMatch == "no", na.rm = T),
  genosMatch = sum(compareGenos_gtscore10x_samplesAll$genoMatch == "yes", na.rm = T),
  genosNA = sum(is.na(compareGenos_gtscore10x_samplesAll$genoMatch)),
  totalGenos = nrow(compareGenos_gtscore10x_samplesAll)
) %>%
  mutate(
    genosCommon = genosMismatch + genosMatch,
    propMismatch = round(genosMismatch/genosCommon, 2),
    propMatch = round(genosMatch/genosCommon, 2)
  ) %>%
  relocate(c(propMismatch, propMatch, genosMismatch, genosMatch, genosNA, genosCommon, totalGenos))

compareGenos_gtscore10x_samplesBlood_table <- data.frame(
  genosMismatch = sum(compareGenos_gtscore10x_samplesBlood$genoMatch == "no", na.rm = T),
  genosMatch = sum(compareGenos_gtscore10x_samplesBlood$genoMatch == "yes", na.rm = T),
  genosNA = sum(is.na(compareGenos_gtscore10x_samplesBlood$genoMatch)),
  totalGenos = nrow(compareGenos_gtscore10x_samplesBlood)
) %>%
  mutate(
    genosCommon = genosMismatch + genosMatch,
    propMismatch = round(genosMismatch/genosCommon, 2),
    propMatch = round(genosMatch/genosCommon, 2)
  ) %>%
  relocate(c(propMismatch, propMatch, genosMismatch, genosMatch, genosNA, genosCommon, totalGenos))
```




# 5 Preliminary analyses

## 5.1 GTscore vs. GTseq

## 5.2 QC filtering

## 5.3 Conclusions

Moving forward with GTscore, no qc filtering

# 6 Downsampling

## 6.1 Downsampling function

I created the function below to downsample sequencing results using the allele read counts file from the GTscore pipeline as the input. In doing so, we can take advantage of the mapping and read counting functions already built into the GTscore pipeline and avoid more time-intensive downsampling methods that require going back into the fastq files.

This function works as follows:

1.  Using the allele reads file, the function generates a list of elements of length *"n_iterations"* (specified by the user), where each element is a randomized string of a's and b's for each sample at each locus.
    1.  The number of times that each letter occurs in each string corresponds to the read counts for that sample at that allele (e.g., read counts of "3,2" would be transformed to abbaa, or some other random combination of 3 a's and 2 b's).
2.  Next, the function selects the first *"n_reads"* characters from each element in the list, where *"n_reads"* (specified by the user) is the desired downsampled coverage per sample per locus.
3.  The function then re-counts the allele reads based on the number of a's (allele 1 read counts) and b's (allele 2 read counts) selected in each element and averages them (rounded to the nearest integer) to provide the final downsampled read counts. Any sample/locus read counts that are below the desired coverage are recoded to "0,0"
4.  Finally, the function returns the downsampled reads as a new allele reads text file, which can then be input back into GTscore for genotyping.

**Quick-load function:**

```{r}
# load in fxn + necessary packages
source("./my_sourceScripts/downsample_alleleReads.R")
```

**View function:**

```{r, eval=FALSE}
###############################
### DOWNSAMPLE ALLELE READS ###
###############################

# Input file = UNFILTERED allele reads file from GTscore
## rownames = loci
## colnames = samples
## allele read values = comma-separated (e.g., 4,23)

library(stringi)

downsample_alleleReads <- function(alleleReads_file, n_reads, n_iterations) {
  # reformat alleleReads file
  ds_alleleReads <- alleleReads_file %>%
    rownames_to_column("locus") %>%
    pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
    # obtain separate read counts for allele 1, allele 2, and their sum
    separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
    # create "a/b" character strings based on allele counts (separate for each allele at first)
    rowwise() %>%
    mutate(
      vec_a1 = paste(replicate(readCounts_a1, "a"), collapse = ""),
      vec_a2 = paste(replicate(readCounts_a2, "b"), collapse = "")
    ) %>%
    # randomly combine "a" and "b" allele strings "n_iterations" times & combine as a list
    mutate(
      vec_a1a2 = list(replicate(n_iterations, stri_rand_shuffle(str_c(vec_a1, vec_a2))))
    ) %>%
    # downsample by extracting the first "n_reads" characters from each string in the list
    mutate(
      ds_vec_a1a2 = list(substr(vec_a1a2, 1, n_reads))
    ) %>%
    # count the number of a's & b's in each string in the list,
    # then calculate the mean of those counts, rounding to nearest integer
    mutate(
      dsCounts_a1 = as.integer(round(mean(unlist(list(str_count(ds_vec_a1a2, "a")))))),
      dsCounts_a2 = as.integer(round(mean(unlist(list(str_count(ds_vec_a1a2, "b"))))))
    ) %>%
    # calculate the sum for the downsampled read counts,
    # recode as NA if sum < n_reads,
    # and combine the rest in "x,x" format
    mutate(
      ds_readCounts_sum = dsCounts_a1 + dsCounts_a2,
      ds_readCounts = case_when(
        ds_readCounts_sum < n_reads ~ "0,0",
        .default = str_c(dsCounts_a1, dsCounts_a2, sep = ",")
      )
    ) %>%
    # create new alleleReads dataframe,
    # can be saved as a txt file or used as input for GTscore genotyping function
    select(c(sampleID, locus, ds_readCounts)) %>%
    pivot_wider(id_cols = locus,
                names_from = sampleID,
                values_from = ds_readCounts) %>%
    column_to_rownames("locus")
  
  return(ds_alleleReads)
}
```

## 6.2 Run function

### Illumina

```{r, eval=FALSE}
# downsample allele reads
ds6x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 6, 100)
ds10x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 10, 100)
ds15x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 15, 100)
ds20x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 20, 100)
ds25x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 25, 100)
ds30x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 30, 100)
ds50x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 50, 100)
ds100x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 100, 100)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled allele reads
write.table(ds6x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds10x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds15x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds20x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds25x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds30x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds50x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds100x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

### Nanopore

```{r, eval=FALSE}
# downsample allele reads
ds6x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 6, 100)
ds10x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 10, 100)
ds15x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 15, 100)
ds20x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 20, 100)
ds25x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 25, 100)
ds30x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 30, 100)
ds50x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 50, 100)
ds100x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 100, 100)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled allele reads
write.table(ds6x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds10x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds15x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds20x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds25x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds30x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds50x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds100x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

## 5.3 Genotyping

Now that we have the downsampled allele reads for Illumina and Nanopore, we can plug them into the GTscore PolyGen genotyping function. Note that this function also requires the locus table, imported below.

### Illumina

```{r, eval=FALSE}
# load locus table
fullSet_singleSNP_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

# generate singleSNP genotypes using the polyGen algorithm
ds6x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds6x_alleleReads_ill)
ds10x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds10x_alleleReads_ill)
ds15x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds15x_alleleReads_ill)
ds20x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ill)
ds25x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds25x_alleleReads_ill)
ds30x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds30x_alleleReads_ill)
ds50x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds50x_alleleReads_ill)
ds100x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds100x_alleleReads_ill)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled genotype results
write.table(ds6x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds10x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds15x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds20x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds25x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds30x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds50x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds100x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

### Nanopore

```{r, eval=FALSE}
# load locus table
fullSet_singleSNP_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

# generate singleSNP genotypes using the polyGen algorithm
ds6x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds6x_alleleReads_ont)
ds10x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds10x_alleleReads_ont)
ds15x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds15x_alleleReads_ont)
ds20x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ont)
ds25x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds25x_alleleReads_ont)
ds30x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds30x_alleleReads_ont)
ds50x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds50x_alleleReads_ont)
ds100x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds100x_alleleReads_ont)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled genotype results
write.table(ds6x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds10x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds15x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds20x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds25x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds30x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds50x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds100x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

## 5.4 Downsampled geno data

### Genotype files

#### Illumina

```{r}
genos_ill_gtscore0x_samplesBlood_lociDual_ds0x <- genos_ill_gtscore0x_samplesBlood_lociDual

genos_ill_gtscore10x_samplesBlood_lociDual_ds0x <- genos_ill_gtscore10x_samplesBlood_lociDual

genos_ill_gtscore_samplesBlood_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds20x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds25x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))
```

#### Nanopore

```{r}
genos_ont_gtscore0x_samplesBlood_lociDual_ds0x <- genos_ont_gtscore0x_samplesBlood_lociDual

genos_ont_gtscore10x_samplesBlood_lociDual_ds0x <- genos_ont_gtscore10x_samplesBlood_lociDual

genos_ont_gtscore_samplesBlood_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds20x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds25x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))
```

### Long dataframes

#### Illumina

```{r}
genos_ill_gtscore0x_samplesBlood_lociDual_ds0x_l <- genos_ill_gtscore0x_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds0x.0x")

genos_ill_gtscore10x_samplesBlood_lociDual_ds0x_l <- genos_ill_gtscore10x_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds10x.0x")

genos_ill_gtscore_samplesBlood_lociDual_ds6x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds6x")

genos_ill_gtscore_samplesBlood_lociDual_ds10x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds10x")

genos_ill_gtscore_samplesBlood_lociDual_ds15x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds15x")

genos_ill_gtscore_samplesBlood_lociDual_ds20x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds20x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds20x")

genos_ill_gtscore_samplesBlood_lociDual_ds25x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds25x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds25x")

genos_ill_gtscore_samplesBlood_lociDual_ds30x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds30x")

genos_ill_gtscore_samplesBlood_lociDual_ds50x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds50x")

genos_ill_gtscore_samplesBlood_lociDual_ds100x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds100x")
```

#### Nanopore

```{r}
genos_ont_gtscore0x_samplesBlood_lociDual_ds0x_l <- genos_ont_gtscore0x_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds0x.0x")

genos_ont_gtscore10x_samplesBlood_lociDual_ds0x_l <- genos_ont_gtscore10x_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds10x.0x")

genos_ont_gtscore_samplesBlood_lociDual_ds6x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds6x")

genos_ont_gtscore_samplesBlood_lociDual_ds10x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds10x")

genos_ont_gtscore_samplesBlood_lociDual_ds15x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds15x")

genos_ont_gtscore_samplesBlood_lociDual_ds20x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds20x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds20x")

genos_ont_gtscore_samplesBlood_lociDual_ds25x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds25x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds25x")

genos_ont_gtscore_samplesBlood_lociDual_ds30x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds30x")

genos_ont_gtscore_samplesBlood_lociDual_ds50x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds50x")

genos_ont_gtscore_samplesBlood_lociDual_ds100x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds100x")
```

### Comparison dataframe

The dataframe below combines the genotypes from each of the downsampled (and original) genotype results from both Illumina and ONT.

```{r}
genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos <- purrr::reduce(
  list(
    genos_ill_gtscore0x_samplesBlood_lociDual_ds0x_l,
    genos_ill_gtscore10x_samplesBlood_lociDual_ds0x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds6x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds10x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds15x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds20x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds25x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds30x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds50x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds100x_l,
    
    genos_ont_gtscore0x_samplesBlood_lociDual_ds0x_l,
    genos_ont_gtscore10x_samplesBlood_lociDual_ds0x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds6x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds10x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds15x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds20x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds25x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds30x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds50x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds100x_l
    ),
  dplyr::left_join, by = c("sampleID", "locus")) %>%
  relocate(sampleID) %>%
  arrange(sampleID)
```

# 8 Geno consistency - optimized (blood samples only)

## 8.1 Optimize sample/loci combos

Given the difference in genotype success between Illumina and Nanopore results, prior to comparing genotype *consistency* we need to subset the samples and loci within the Illumina and Nanopore datasets so that we can 1) have the highest number of samples and loci possible while 2) ensuring that there is no missing data (i.e., we need 100% genotype success for our chosen combination of samples and loci).

### Functions

To identify the optimal combination of samples and loci between **both** Illumina and Nanopore datasets, I've created two functions. Each function has a **version a** and a **version b**, where **version a** optimizes samples/loci based on a single dataset, and **version b** optimizes samples/loci based on two datasets.

#### Function 1

**Function 1a:** For a given dataframe of genotypes, this function returns the number of loci genotyped at the desired proportion of n samples ("propSamples") for subsets of 2 to 30 samples

```{r}
# number of loci genotyped at propSamples (0 to 1.0) of samples from 2 to 30 samples
optimize_samplesLoci <- function(df, propSamples) {
  # Initialize an empty list to store results
  result_list <- list()

  for (n in 2:30) {
    # create matrix where NAs are 0 and non-NAs are 1
    is_na_vector <- ifelse(is.na(df), 0, 1)

    crossprod_vector <- crossprod(is_na_vector)
    col_sums <- colSums(crossprod_vector)

    x_df <- as.data.frame(df)  # to get meaningful colnames
    
    # use colSums vector to select n columns;
    # will rank all columns and give the n first;
    # gives n columns w/the max number of non-NA rows
    res <- x_df[, rank(-col_sums, ties.method = "first") <= n]

    # Store the result in the list
    result_list[[as.character(n)]] <- c(n, nrow(res[which(rowMeans(!is.na(res)) >= propSamples), ]))
  }

  # Convert the list to a dataframe
  result_df <- do.call(rbind, result_list) %>%
    as.data.frame()

  # Rename columns
  colnames(result_df) <- c("samples", "loci")

  return(result_df)
}
```

**Function 1b:** This function does the same as Function 1a, but allows for optimization based on two genotype files. This means that the specific samples included in any given "n" of "propSamples" are the **same** samples in both the Illumina and Nanopore datasets.

**NOTE:** The two genotype files must be formatted such that the rows and columns are in the exact same order. This may require some reformatting since illumina and nanopore samples were numbered differently.

```{r}
optimize_samplesLoci_illONT <- function(df1, df2, propSamples) {
  # Initialize an empty list to store results
  result_list <- list()
  
  # convert genos to binary matrix (NAs = 0, non-NAs = 1)
  matrix1 <- ifelse(is.na(df1), 0, 1)
  matrix2 <- ifelse(is.na(df2), 0, 1)
    
  # multiply matrices
  matrix_prod <- matrix1 * matrix2
  
  # create new combo df (turn 0 back to NA)
  df3 <- as.data.frame(matrix_prod) %>%
    mutate(across(everything(), ~na_if(., 0)))

  # identify sample/loci counts w/100% geno success
  for (n_samples in 2:30) {
  
    crossprod_vector <- crossprod(matrix_prod)
    col_sums <- colSums(crossprod_vector)

    x_df <- as.data.frame(df3)  # to get meaningful colnames
    
    # use colSums vector to select n columns;
    # will rank all columns and give the n first;
    # gives n columns w/the max number of non-NA rows
    res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

    # Store the result in the list
    result_list[[as.character(n_samples)]] <- c(n_samples, nrow(res[which(rowMeans(!is.na(res)) >= propSamples), ]))
  }

  # Convert the list to a dataframe
  result_df <- do.call(rbind, result_list) %>%
    as.data.frame()

  # Rename columns
  colnames(result_df) <- c("samples", "loci")

  return(result_df)
}
```

#### Function 2

**Function 2a:** For a given dataframe of genotypes, this function returns an optimized set of genotypes for n_samples, where the number of samples and the proportion of those n samples successfully genotyped per locus (propSamples) is chosen based on the results of function 1 above

```{r}
optimized_genos <- function(df, n_samples, propSamples) {
  matrix <- ifelse(is.na(df), 0, 1)
  crossprod_vector <- crossprod(matrix)
  col_sums <- colSums(crossprod_vector)

  x_df <- as.data.frame(df)
  res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

  result <- res[which(rowMeans(!is.na(res)) >= propSamples), ]
  return(result)
}
```

**Function 2b:**

This function returns a list of two dataframes; access via x[[a]] and x[[2]].

**NOTE:** The two genotype files must be formatted such that the rows and columns are in the exact same order. This may require some reformatting since illumina and nanopore samples were numbered differently.

```{r}
optimized_genos_illONT <- function(df1, df2, n_samples, propSamples) {
  # create combo geno matrix
  matrix1 <- ifelse(is.na(df1), 0, 1)
  matrix2 <- ifelse(is.na(df2), 0, 1)
  matrix_prod <- matrix1 * matrix2
  
  # obtain crossproducts & col sums
  crossprod_vector <- crossprod(matrix_prod)
  col_sums <- colSums(crossprod_vector)

  x_df <- as.data.frame(matrix_prod) %>%
    mutate(across(everything(), ~na_if(., 0)))
  
  res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

  result <- res[which(rowMeans(!is.na(res)) >= propSamples), ]
  
  # subset each geno file to optimized sample/loci
  optGenos_df1 <- df1 %>%
    filter(row.names(.) %in% rownames(result)) %>%
    select(colnames(result))
  
  optGenos_df2 <- df2 %>%
    filter(row.names(.) %in% rownames(result)) %>%
    select(colnames(result))
  
  return(list(optGenos_df1, optGenos_df2))
}
```

### Run functions

#### Identify optimal sample/loci numbers

Below I'm using **function 1b** to identify the number of loci w/100% genotype success in both Illumina and Nanopore datasets for sample subsets of 2 to 30 samples across various read depth cutoffs.

```{r, echo=FALSE}
opt100_illONT_gtscore_samplesBlood_lociDual_dsTests <-
  optimize_samplesLoci_illONT(
    genos_ill_gtscore0x_samplesBlood_lociDual_ds0x,
    genos_ont_gtscore0x_samplesBlood_lociDual_ds0x,
    1.0
  ) %>%
  dplyr::rename("ds0x.0x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore10x_samplesBlood_lociDual_ds0x,
          genos_ont_gtscore10x_samplesBlood_lociDual_ds0x,
          1.0
          )
        ) %>%
  dplyr::rename("ds10x.0x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds6x,
          genos_ont_gtscore_samplesBlood_lociDual_ds6x,
          1.0
          )
        ) %>%
  dplyr::rename("ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds10x,
          genos_ont_gtscore_samplesBlood_lociDual_ds10x,
          1.0
          )
        ) %>%
  dplyr::rename("ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds15x,
          genos_ont_gtscore_samplesBlood_lociDual_ds15x,
          1.0
          )
        ) %>%
  dplyr::rename("ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds20x,
          genos_ont_gtscore_samplesBlood_lociDual_ds20x,
          1.0
          )
        ) %>%
  dplyr::rename("ds20x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds25x,
          genos_ont_gtscore_samplesBlood_lociDual_ds25x,
          1.0
          )
        ) %>%
  dplyr::rename("ds25x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds30x,
          genos_ont_gtscore_samplesBlood_lociDual_ds30x,
          1.0
          )
        ) %>%
  dplyr::rename("ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds50x,
          genos_ont_gtscore_samplesBlood_lociDual_ds50x,
          1.0
          )
        ) %>%
  dplyr::rename("ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds100x,
          genos_ont_gtscore_samplesBlood_lociDual_ds100x,
          1.0
          )
        ) %>%
  dplyr::rename("ds100x" = "loci")
```

```{r, echo=FALSE}
# for html
opt100_illONT_gtscore_samplesBlood_lociDual_dsTests %>%
  kbl(caption = "Number of loci w/100% geno success in both Illumina & Nanopore datasets across sample subsets & coverage cutoffs") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "500px", fixed_thead = list(enabled = T))
```

#### Subset optimal samples/loci

Based on the results above, I'm opting to go with 10 samples for all read depths along with their respective loci that have 100% genotype success in both Illumina and Nanopore datasets. Below, I'm using **function 2b** to create the new sample/loci subsets for both Illumina and Nanopore datasets at each read depth cutoff.

```{r}
# ds0x.0x
optGenos100_illONT_gtscore0x_samplesBlood_lociDual_ds0x <- optimized_genos_illONT(
  genos_ill_gtscore0x_samplesBlood_lociDual_ds0x,
  genos_ont_gtscore0x_samplesBlood_lociDual_ds0x,
  10,
  1.0
)

optGenos100_ill_gtscore0x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore0x_samplesBlood_lociDual_ds0x[[1]]

optGenos100_ont_gtscore0x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore0x_samplesBlood_lociDual_ds0x[[2]]

# ds10x.0x
optGenos100_illONT_gtscore10x_samplesBlood_lociDual_ds0x <- optimized_genos_illONT(
  genos_ill_gtscore10x_samplesBlood_lociDual_ds0x,
  genos_ont_gtscore10x_samplesBlood_lociDual_ds0x,
  10,
  1.0
)

optGenos100_ill_gtscore10x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore10x_samplesBlood_lociDual_ds0x[[1]]

optGenos100_ont_gtscore10x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore10x_samplesBlood_lociDual_ds0x[[2]]

# ds6x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds6x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds6x,
  genos_ont_gtscore_samplesBlood_lociDual_ds6x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds6x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds6x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds6x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds6x[[2]]

# ds10x 
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds10x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds10x,
  genos_ont_gtscore_samplesBlood_lociDual_ds10x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds10x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds10x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds10x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds10x[[2]]

# ds15x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds15x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds15x,
  genos_ont_gtscore_samplesBlood_lociDual_ds15x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds15x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds15x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds15x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds15x[[2]]

# ds20x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds20x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds20x,
  genos_ont_gtscore_samplesBlood_lociDual_ds20x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds20x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds20x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds20x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds20x[[2]]

# ds25x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds25x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds25x,
  genos_ont_gtscore_samplesBlood_lociDual_ds25x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds25x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds25x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds25x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds25x[[2]]

# ds30x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds30x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds30x,
  genos_ont_gtscore_samplesBlood_lociDual_ds30x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds30x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds30x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds30x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds30x[[2]]

# ds50x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds50x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds50x,
  genos_ont_gtscore_samplesBlood_lociDual_ds50x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds50x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds50x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds50x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds50x[[2]]

# ds100x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds100x,
  genos_ont_gtscore_samplesBlood_lociDual_ds100x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds100x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x[[2]]
```

## 8.2 Geno consistency

### Data

100% geno success

```{r}
# ds0x.0x
genoCompare_illONT_gtscore0x_samplesBlood_lociDual_opt100_ds0x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore0x_samplesBlood_lociDual_ds0x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore0x_samplesBlood_lociDual_ds0x)) %>%
  select(c(sampleID, locus, ill_ds0x.0x, ont_ds0x.0x)) %>%
  mutate(
    match = case_when(
      ill_ds0x.0x == ont_ds0x.0x ~ "yes",
      .default = "no"
    )
  )

# ds10x.0x
genoCompare_illONT_gtscore10x_samplesBlood_lociDual_opt100_ds0x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore10x_samplesBlood_lociDual_ds0x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore10x_samplesBlood_lociDual_ds0x)) %>%
  select(c(sampleID, locus, ill_ds10x.0x, ont_ds10x.0x)) %>%
  mutate(
    match = case_when(
      ill_ds10x.0x == ont_ds10x.0x ~ "yes",
      .default = "no"
    )
  )

# ds6x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds6x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds6x)) %>%
  select(c(sampleID, locus, ill_ds6x, ont_ds6x)) %>%
  mutate(
    match = case_when(
      ill_ds6x == ont_ds6x ~ "yes",
      .default = "no"
    )
  )

# ds10x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds10x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds10x)) %>%
  select(c(sampleID, locus, ill_ds10x, ont_ds10x)) %>%
  mutate(
    match = case_when(
      ill_ds10x == ont_ds10x ~ "yes",
      .default = "no"
    )
  )

# ds15x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds15x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds15x)) %>%
  select(c(sampleID, locus, ill_ds15x, ont_ds15x)) %>%
  mutate(
    match = case_when(
      ill_ds15x == ont_ds15x ~ "yes",
      .default = "no"
    )
  )

# ds20x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds20x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds20x)) %>%
  select(c(sampleID, locus, ill_ds20x, ont_ds20x)) %>%
  mutate(
    match = case_when(
      ill_ds20x == ont_ds20x ~ "yes",
      .default = "no"
    )
  )

# ds25x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds25x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds25x)) %>%
  select(c(sampleID, locus, ill_ds25x, ont_ds25x)) %>%
  mutate(
    match = case_when(
      ill_ds25x == ont_ds25x ~ "yes",
      .default = "no"
    )
  )

# ds30x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds30x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds30x)) %>%
  select(c(sampleID, locus, ill_ds30x, ont_ds30x)) %>%
  mutate(
    match = case_when(
      ill_ds30x == ont_ds30x ~ "yes",
      .default = "no"
    )
  )

# ds50x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds50x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds50x)) %>%
  select(c(sampleID, locus, ill_ds50x, ont_ds50x)) %>%
  mutate(
    match = case_when(
      ill_ds50x == ont_ds50x ~ "yes",
      .default = "no"
    )
  )

# ds100x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x)) %>%
  select(c(sampleID, locus, ill_ds100x, ont_ds100x)) %>%
  mutate(
    match = case_when(
      ill_ds100x == ont_ds100x ~ "yes",
      .default = "no"
    )
  )
```

### Results

```{r}
# ds0x.0x
summary_genoConsistency_ds0x.0x <- genoCompare_illONT_gtscore0x_samplesBlood_lociDual_opt100_ds0x %>%
  {table(.$match)} %>%
  as.data.frame() %>%
  pivot_wider(names_from = Var1, values_from = Freq) %>%
  dplyr::rename("genosMatch" = "yes",
                "genosMismatch" = "no") %>%
  mutate(
    totalGenos = genosMatch + genosMismatch,
    totalSamples = length(unique(genoCompare_illONT_gtscore0x_samplesBlood_lociDual_opt100_ds0x$sampleID)),
    totalLoci = length(unique(genoCompare_illONT_gtscore0x_samplesBlood_lociDual_opt100_ds0x$locus)),
    propMismatch = round(genosMismatch/totalGenos, 2),
    dataset = "ds0x.0x"
  ) %>%
  relocate(c(dataset, totalSamples, totalLoci, propMismatch, genosMismatch, genosMatch))

# ds10x.0x
summary_genoConsistency_ds10x.0x <- genoCompare_illONT_gtscore10x_samplesBlood_lociDual_opt100_ds0x %>%
  {table(.$match)} %>%
  as.data.frame() %>%
  pivot_wider(names_from = Var1, values_from = Freq) %>%
  dplyr::rename("genosMatch" = "yes",
                "genosMismatch" = "no") %>%
  mutate(
    totalGenos = genosMatch + genosMismatch,
    totalSamples = length(unique(genoCompare_illONT_gtscore10x_samplesBlood_lociDual_opt100_ds0x$sampleID)),
    totalLoci = length(unique(genoCompare_illONT_gtscore10x_samplesBlood_lociDual_opt100_ds0x$locus)),
    propMismatch = round(genosMismatch/totalGenos, 2),
    dataset = "ds10x.0x"
  ) %>%
  relocate(c(dataset, totalSamples, totalLoci, propMismatch, genosMismatch, genosMatch))

# ds6x

# ds10x

# ds15x

# ds20x

# ds25x

# ds30x

# ds50x
summary_genoConsistency_ds50x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x %>%
  {table(.$match)} %>%
  as.data.frame() %>%
  pivot_wider(names_from = Var1, values_from = Freq) %>%
  dplyr::rename("genosMatch" = "yes",
                "genosMismatch" = "no") %>%
  mutate(
    totalGenos = genosMatch + genosMismatch,
    totalSamples = length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$sampleID)),
    totalLoci = length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$locus)),
    propMismatch = round(genosMismatch/totalGenos, 2),
    dataset = "ds50x"
  ) %>%
  relocate(c(dataset, totalSamples, totalLoci, propMismatch, genosMismatch, genosMatch))

# ds100x
summary_genoConsistency_ds100x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x %>%
  {table(.$match)} %>%
  as.data.frame() %>%
  pivot_wider(names_from = Var1, values_from = Freq) %>%
  dplyr::rename("genosMatch" = "yes",
                "genosMismatch" = "no") %>%
  mutate(
    totalGenos = genosMatch + genosMismatch,
    totalSamples = length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$sampleID)),
    totalLoci = length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$locus)),
    propMismatch = round(genosMismatch/totalGenos, 2),
    dataset = "ds100x"
  ) %>%
  relocate(c(dataset, totalSamples, totalLoci, propMismatch, genosMismatch, genosMatch))
```

### Summary

```{r}

```


# X Data for "toShare" reports

## ds0x.0x

Going to up the sample count for the "toShare" version to have 20 samples + 93 loci

```{r}
# ds0x.0x
optGenos100_illONT_gtscore0x_samplesBlood_lociDual_ds0x <- optimized_genos_illONT(
  genos_ill_gtscore0x_samplesBlood_lociDual_ds0x,
  genos_ont_gtscore0x_samplesBlood_lociDual_ds0x,
  20,
  1.0
)

optGenos100_ill_gtscore0x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore0x_samplesBlood_lociDual_ds0x[[1]]

optGenos100_ont_gtscore0x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore0x_samplesBlood_lociDual_ds0x[[2]]
```

Metadata

```{r}
md_toShare_ds0x <- md_tamRun4_blood %>%
  filter(sampleID_unique %in% colnames(optGenos100_ill_gtscore0x_samplesBlood_lociDual_ds0x)) %>%
  select(sampleID_unique, sampleID_ill, sampleID_ont, species, sex)

write.csv(md_toShare_ds0x, "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_toShare/metadata_paper1_nanoporeValidation_ds0x.csv", row.names = F)
```

Sample files

```{r}
# originals
sampleFiles_ill_original <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_sampleFiles.txt")

sampleFiles_ont_original <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_sampleFiles.txt")

# subset to files in ds0x
sampleFiles_ill_ds0x <- sampleFiles_ill_original %>%
  merge(., md_tamRun4[, c("sampleFile_ill", "sampleID_unique")], by.x = "V1", by.y = "sampleFile_ill") %>%
  filter(sampleID_unique %in% md_toShare_ds0x$sampleID_unique) %>%
  select(-sampleID_unique)

sampleFiles_ont_ds0x <- sampleFiles_ont_original %>%
  merge(., md_tamRun4[, c("sampleFile_ont", "sampleID_unique")], by.x = "V1", by.y = "sampleFile_ont") %>%
  filter(sampleID_unique %in% md_toShare_ds0x$sampleID_unique) %>%
  select(-sampleID_unique)
 
# export
write.table(sampleFiles_ill_ds0x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_toShare/sampleFiles_ill_ds0x.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(sampleFiles_ont_ds0x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_toShare/sampleFiles_ont_ds0x.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

Primer probe file

```{r}
pp_original <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)

pp_ds0x.0x <- pp_original %>%
  filter(Locus %in% rownames(optGenos100_ill_gtscore0x_samplesBlood_lociDual_ds0x))

write.table(pp_ds0x.0x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_toShare/primerProbeFile_ds0x.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```

## ds0x_g10x

20 samples + 77 loci

```{r}
# ds0x.0x
optGenos100_illONT_gtscore10x_samplesBlood_lociDual_ds0x <- optimized_genos_illONT(
  genos_ill_gtscore10x_samplesBlood_lociDual_ds0x,
  genos_ont_gtscore10x_samplesBlood_lociDual_ds0x,
  20,
  1.0
)

optGenos100_ill_gtscore10x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore10x_samplesBlood_lociDual_ds0x[[1]]

optGenos100_ont_gtscore10x_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore10x_samplesBlood_lociDual_ds0x[[2]]
```

Metadata

```{r}
md_toShare_ds0x_g10x <- md_tamRun4_blood %>%
  filter(sampleID_unique %in% colnames(optGenos100_ill_gtscore10x_samplesBlood_lociDual_ds0x)) %>%
  select(sampleID_unique, sampleID_ill, sampleID_ont, species, sex)

write.csv(md_toShare_ds0x_g10x, "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_g10x_toShare/metadata_paper1_nanoporeValidation_ds0x_g10x.csv", row.names = F)
```

Sample files

```{r}
# originals
sampleFiles_ill_original <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_sampleFiles.txt")

sampleFiles_ont_original <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_sampleFiles.txt")

# subset to files in ds0x
sampleFiles_ill_ds0x_g10x <- sampleFiles_ill_original %>%
  merge(., md_tamRun4[, c("sampleFile_ill", "sampleID_unique")], by.x = "V1", by.y = "sampleFile_ill") %>%
  filter(sampleID_unique %in% md_toShare_ds0x_g10x$sampleID_unique) %>%
  select(-sampleID_unique)

sampleFiles_ont_ds0x_g10x <- sampleFiles_ont_original %>%
  merge(., md_tamRun4[, c("sampleFile_ont", "sampleID_unique")], by.x = "V1", by.y = "sampleFile_ont") %>%
  filter(sampleID_unique %in% md_toShare_ds0x_g10x$sampleID_unique) %>%
  select(-sampleID_unique)
 
# export
write.table(sampleFiles_ill_ds0x_g10x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_g10x_toShare/sampleFiles_ill_ds0x_g10x.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(sampleFiles_ont_ds0x_g10x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_g10x_toShare/sampleFiles_ont_ds0x_g10x.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

Primer probe file

```{r}
pp_original <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)

pp_ds0x_g10x <- pp_original %>%
  filter(Locus %in% rownames(optGenos100_ill_gtscore10x_samplesBlood_lociDual_ds0x))

write.table(pp_ds0x_g10x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds0x_g10x_toShare/primerProbeFile_ds0x_g10x.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```


## ds100x

```{r}
# ds0x.0x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds100x,
  genos_ont_gtscore_samplesBlood_lociDual_ds100x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds100x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x[[2]]
```

Metadata

```{r}
md_toShare_ds100x <- md_tamRun4_blood %>%
  filter(sampleID_unique %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x)) %>%
  select(sampleID_unique, sampleID_ill, sampleID_ont, species, sex)

write.csv(md_toShare_ds100x, "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds100x_toShare/metadata_paper1_nanoporeValidation_ds100x.csv", row.names = F)
```

Sample files

```{r}
# originals
sampleFiles_ill_original <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_sampleFiles.txt")

sampleFiles_ont_original <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_sampleFiles.txt")

# subset to files in ds0x
sampleFiles_ill_ds100x <- sampleFiles_ill_original %>%
  merge(., md_tamRun4[, c("sampleFile_ill", "sampleID_unique")], by.x = "V1", by.y = "sampleFile_ill") %>%
  filter(sampleID_unique %in% md_toShare_ds100x$sampleID_unique) %>%
  select(-sampleID_unique)

sampleFiles_ont_ds100x <- sampleFiles_ont_original %>%
  merge(., md_tamRun4[, c("sampleFile_ont", "sampleID_unique")], by.x = "V1", by.y = "sampleFile_ont") %>%
  filter(sampleID_unique %in% md_toShare_ds100x$sampleID_unique) %>%
  select(-sampleID_unique)
 
# export
write.table(sampleFiles_ill_ds100x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds100x_toShare/sampleFiles_ill_ds100x.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(sampleFiles_ont_ds100x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds100x_toShare/sampleFiles_ont_ds100x.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

Primer probe file

```{r}
pp_original <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)

pp_ds100x <- pp_original %>%
  filter(Locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x))

write.table(pp_ds100x, file = "./paper1_nanoporeValidation/paper1_nanoporeValidation_ds100x_toShare/primerProbeFile_ds100x.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```



# XXXXXXXXXX
# 4 Analyses

## 4.1 Illumina vs. ONT overview

Without making any alterations to the GTscore and GTseq results for Illumina and Nanopore datasets, how do the two compare?

**Note** that the comparisons below are based on **dual-species loci only** (n = 140 loci).

### Coverage

The number of reads per locus for each sample is consistently higher for Illumina vs. Nanopore sequencing results, with an average of 528 

#### gtscore

Illumina samples consistently have higher coverage relative to Nanopore across all samples, with an average of 528 (Illumina) vs. 72 (ONT) reads per locus across all samples and 1144 (Illumina) vs. 148 (ONT) reads per locus across blood samples. Mean and median reads per locus across sample subsets can be viewed in the table and plots below.

```{r, echo=FALSE}
# summary table
alleleReads_illONT_gtscore_samplesAll_lociDual_summaryTable <- data.frame(
  sampleType = c("allSamples", "blood", "fecal", "hair"),
  mean_ill = c(
    mean(alleleReads_ill_gtscore_samplesAll_lociDual_l$readCounts_sum),
    mean(alleleReads_ill_gtscore_samplesBlood_lociDual_l$readCounts_sum),
    mean(alleleReads_ill_gtscore_samplesFecal_lociDual_l$readCounts_sum),
    mean(alleleReads_ill_gtscore_samplesHair_lociDual_l$readCounts_sum)
  ),
  mean_ont = c(
    mean(alleleReads_ont_gtscore_samplesAll_lociDual_l$readCounts_sum),
    mean(alleleReads_ont_gtscore_samplesBlood_lociDual_l$readCounts_sum),
    mean(alleleReads_ont_gtscore_samplesFecal_lociDual_l$readCounts_sum),
    mean(alleleReads_ont_gtscore_samplesHair_lociDual_l$readCounts_sum)
  ),
  med_ill = c(
    median(alleleReads_ill_gtscore_samplesAll_lociDual_l$readCounts_sum),
    median(alleleReads_ill_gtscore_samplesBlood_lociDual_l$readCounts_sum),
    median(alleleReads_ill_gtscore_samplesFecal_lociDual_l$readCounts_sum),
    median(alleleReads_ill_gtscore_samplesHair_lociDual_l$readCounts_sum)
  ),
  med_ont = c(
    median(alleleReads_ont_gtscore_samplesAll_lociDual_l$readCounts_sum),
    median(alleleReads_ont_gtscore_samplesBlood_lociDual_l$readCounts_sum),
    median(alleleReads_ont_gtscore_samplesFecal_lociDual_l$readCounts_sum),
    median(alleleReads_ont_gtscore_samplesHair_lociDual_l$readCounts_sum)
  )
) %>%
  mutate(
    mean_ill = round(mean_ill, 2),
    mean_ont = round(mean_ont, 2)
  )
```

```{r, echo=FALSE}
# for html
alleleReads_illONT_gtscore_samplesAll_lociDual_summaryTable %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "tomato") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

```{r, echo=FALSE}
# plot comparing coverage per locus across sample types for illumina vs. nanopore
rbind(alleleReads_ill_gtscore_samplesAll_lociDual_l,
      alleleReads_ont_gtscore_samplesAll_lociDual_l) %>%
  separate("sampleID", into = c("sampleID", "sampleType"), sep = "_") %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  ggplot(aes(x = sampleID, y = readCounts_sum, fill = seqPlatform)) +
  geom_boxplot() +
  geom_hline(
    data = . %>%
      group_by(sampleType, seqPlatform) %>%
      summarise(line = mean(readCounts_sum)),
    mapping = aes(yintercept = line)
    ) +
#  geom_hline(yintercept = 528, col = "black") +
#  geom_hline(yintercept = 72, col = "black") +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  facet_grid(sampleType~., scales = "free") +
  labs(
    title = "Dual-species loci coverage across sample types for Illumina vs. Nanopore",
    x = "animalID",
    y = "reads per locus")
```

```{r, include=FALSE}
# separate plots by sample type (keeping just in case)
## blood samples
rbind(alleleReads_ill_gtscore_samplesBlood_lociDual_l,
      alleleReads_ont_gtscore_samplesBlood_lociDual_l) %>%
  ggplot(aes(x = sampleID, y = readCounts_sum, fill = seqPlatform)) +
  geom_boxplot() +
  geom_hline(yintercept = 1144, col = "black") +
  geom_hline(yintercept = 148, col = "black") +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_y_continuous(
    breaks = c(0, 148, 1144, 2500, 5000, 7500, 10000),
    labels = c(0, "mean(ont)=148", "mean(ill)=1144", 2500, 5000, 7500, 10000)
  ) +
  labs(
    title = "Dual-species loci coverage across blood samples for Illumina vs. Nanopore",
    y = "Reads per locus")

## fecal samples
rbind(alleleReads_ill_gtscore_samplesFecal_lociDual_l,
      alleleReads_ont_gtscore_samplesFecal_lociDual_l) %>%
  ggplot(aes(x = sampleID, y = readCounts_sum, fill = seqPlatform)) +
  geom_boxplot() +
  geom_hline(yintercept = 447, col = "black") +
  geom_hline(yintercept = 66, col = "black") +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_y_continuous(
    breaks = c(0, 66, 447, 2500, 5000, 7500, 10000),
    labels = c(0, "mean(ont)=66", "mean(ill)=148", 2500, 5000, 7500, 10000)
  ) +
  labs(
    title = "Dual-species loci coverage across fecal samples for Illumina vs. Nanopore",
    y = "Reads per locus")

## hair samples
rbind(alleleReads_ill_gtscore_samplesHair_lociDual_l,
      alleleReads_ont_gtscore_samplesHair_lociDual_l) %>%
  ggplot(aes(x = sampleID, y = readCounts_sum, fill = seqPlatform)) +
  geom_boxplot() +
  geom_hline(yintercept = 97, col = "black") +
  geom_hline(yintercept = 17, col = "black") +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_y_continuous(
    breaks = c(0, 17, 97, 2500, 5000, 7500, 10000),
    labels = c(0, "mean(ont)=17", "mean(ill)=97", 2500, 5000, 7500, 10000)
  ) +
  labs(
    title = "Dual-species loci coverage across hair samples for Illumina vs. Nanopore",
    y = "Reads per locus")
```

#### gtseq



### Geno success

#### gtscore0x

#### gtscore10x

#### gtseq

### Geno consistency

#### gtscore

#### gtseq

## 4.2 GTscore vs. GTseq

## 4.3 QC filtering

## 4.4 Downsampling 







# 5 Downsampling

Prior to examining genotype consistency between Illumina and Nanopore runs, we first need to downsample the sequencing results so that the read counts for each locus in each sample are consistent. This allows us to **1)** eliminate coverage as a potential variable responsible for any differences in genotypes, and **2)** determine whether genotype consistency between Illumina and Nanopore results differs across read depths.

## 5.1 Downsampling function

I created the function below to downsample sequencing results using the allele read counts file from the GTscore pipeline as the input. In doing so, we can take advantage of the mapping and read counting functions already built into the GTscore pipeline and avoid more time-intensive downsampling methods that require going back into the fastq files.

This function works as follows:

1.  Using the allele reads file, the function generates a list of elements of length *"n_iterations"* (specified by the user), where each element is a randomized string of a's and b's for each sample at each locus.
    1.  The number of times that each letter occurs in each string corresponds to the read counts for that sample at that allele (e.g., read counts of "3,2" would be transformed to abbaa, or some other random combination of 3 a's and 2 b's).
2.  Next, the function selects the first *"n_reads"* characters from each element in the list, where *"n_reads"* (specified by the user) is the desired downsampled coverage per sample per locus.
3.  The function then re-counts the allele reads based on the number of a's (allele 1 read counts) and b's (allele 2 read counts) selected in each element and averages them (rounded to the nearest integer) to provide the final downsampled read counts. Any sample/locus read counts that are below the desired coverage are recoded to "0,0"
4.  Finally, the function returns the downsampled reads as a new allele reads text file, which can then be input back into GTscore for genotyping.

**Quick-load function:**

```{r}
# load in fxn + necessary packages
source("./scripts/downsample_alleleReads.R")
```

**View function:**

```{r, eval=FALSE}
###############################
### DOWNSAMPLE ALLELE READS ###
###############################

# Input file = UNFILTERED allele reads file from GTscore
## rownames = loci
## colnames = samples
## allele read values = comma-separated (e.g., 4,23)

library(stringi)

downsample_alleleReads <- function(alleleReads_file, n_reads, n_iterations) {
  # reformat alleleReads file
  ds_alleleReads <- alleleReads_file %>%
    rownames_to_column("locus") %>%
    pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
    # obtain separate read counts for allele 1, allele 2, and their sum
    separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
    # create "a/b" character strings based on allele counts (separate for each allele at first)
    rowwise() %>%
    mutate(
      vec_a1 = paste(replicate(readCounts_a1, "a"), collapse = ""),
      vec_a2 = paste(replicate(readCounts_a2, "b"), collapse = "")
    ) %>%
    # randomly combine "a" and "b" allele strings "n_iterations" times & combine as a list
    mutate(
      vec_a1a2 = list(replicate(n_iterations, stri_rand_shuffle(str_c(vec_a1, vec_a2))))
    ) %>%
    # downsample by extracting the first "n_reads" characters from each string in the list
    mutate(
      ds_vec_a1a2 = list(substr(vec_a1a2, 1, n_reads))
    ) %>%
    # count the number of a's & b's in each string in the list,
    # then calculate the mean of those counts, rounding to nearest integer
    mutate(
      dsCounts_a1 = as.integer(round(mean(unlist(list(str_count(ds_vec_a1a2, "a")))))),
      dsCounts_a2 = as.integer(round(mean(unlist(list(str_count(ds_vec_a1a2, "b"))))))
    ) %>%
    # calculate the sum for the downsampled read counts,
    # recode as NA if sum < n_reads,
    # and combine the rest in "x,x" format
    mutate(
      ds_readCounts_sum = dsCounts_a1 + dsCounts_a2,
      ds_readCounts = case_when(
        ds_readCounts_sum < n_reads ~ "0,0",
        .default = str_c(dsCounts_a1, dsCounts_a2, sep = ",")
      )
    ) %>%
    # create new alleleReads dataframe,
    # can be saved as a txt file or used as input for GTscore genotyping function
    select(c(sampleID, locus, ds_readCounts)) %>%
    pivot_wider(id_cols = locus,
                names_from = sampleID,
                values_from = ds_readCounts) %>%
    column_to_rownames("locus")
  
  return(ds_alleleReads)
}
```

## 5.2 Run function

### Illumina

```{r, eval=FALSE}
# downsample allele reads
ds6x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 6, 100)
ds10x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 10, 100)
ds15x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 15, 100)
ds20x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 20, 100)
ds25x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 25, 100)
ds30x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 30, 100)
ds50x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 50, 100)
ds100x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll_lociDual, 100, 100)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled allele reads
write.table(ds6x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds10x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds15x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds20x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds25x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds30x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds50x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds100x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

### Nanopore

```{r, eval=FALSE}
# downsample allele reads
ds6x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 6, 100)
ds10x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 10, 100)
ds15x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 15, 100)
ds20x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 20, 100)
ds25x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 25, 100)
ds30x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 30, 100)
ds50x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 50, 100)
ds100x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll_lociDual, 100, 100)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled allele reads
write.table(ds6x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds10x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds15x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds20x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds25x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds30x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds50x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(ds100x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

## 5.3 Genotyping

Now that we have the downsampled allele reads for Illumina and Nanopore, we can plug them into the GTscore PolyGen genotyping function. Note that this function also requires the locus table, imported below.

### Illumina

```{r, eval=FALSE}
# load locus table
fullSet_singleSNP_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

# generate singleSNP genotypes using the polyGen algorithm
ds6x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds6x_alleleReads_ill)
ds10x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds10x_alleleReads_ill)
ds15x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds15x_alleleReads_ill)
ds20x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ill)
ds25x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds25x_alleleReads_ill)
ds30x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds30x_alleleReads_ill)
ds50x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds50x_alleleReads_ill)
ds100x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds100x_alleleReads_ill)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled genotype results
write.table(ds6x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds10x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds15x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds20x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds25x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds30x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds50x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds100x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

### Nanopore

```{r, eval=FALSE}
# load locus table
fullSet_singleSNP_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

# generate singleSNP genotypes using the polyGen algorithm
ds6x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds6x_alleleReads_ont)
ds10x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds10x_alleleReads_ont)
ds15x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds15x_alleleReads_ont)
ds20x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ont)
ds25x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds25x_alleleReads_ont)
ds30x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds30x_alleleReads_ont)
ds50x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds50x_alleleReads_ont)
ds100x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds100x_alleleReads_ont)
```

```{r, eval=FALSE, echo=FALSE}
# export downsampled genotype results
write.table(ds6x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds10x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds15x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds20x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds25x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds30x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds50x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
write.table(ds100x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

## 5.4 Downsampled geno data

### Genotype files

#### Illumina

```{r}
genos_ill_gtscore_samplesBlood_lociDual_ds0x <- genos_ill_gtscore_samplesBlood_lociDual

genos_ill_gtscore_samplesBlood_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds20x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds25x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))
```

#### Nanopore

```{r}
genos_ont_gtscore_samplesBlood_lociDual_ds0x <- genos_ont_gtscore_samplesBlood_lociDual

genos_ont_gtscore_samplesBlood_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds20x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds25x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(contains("blood")) %>%
  # filter to dual-species loci
  filter(row.names(.) %in% primerList.v3_lociDual$locus) %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  # make sure rownames are in alpha-order
  select(order(colnames(.)))
```

### Long dataframes

#### Illumina

```{r}
genos_ill_gtscore_samplesBlood_lociDual_ds0x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds0x")

genos_ill_gtscore_samplesBlood_lociDual_ds6x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds6x")

genos_ill_gtscore_samplesBlood_lociDual_ds10x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds10x")

genos_ill_gtscore_samplesBlood_lociDual_ds15x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds15x")

genos_ill_gtscore_samplesBlood_lociDual_ds20x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds20x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds20x")

genos_ill_gtscore_samplesBlood_lociDual_ds25x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds25x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds25x")

genos_ill_gtscore_samplesBlood_lociDual_ds30x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds30x")

genos_ill_gtscore_samplesBlood_lociDual_ds50x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds50x")

genos_ill_gtscore_samplesBlood_lociDual_ds100x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds100x")
```

#### Nanopore

```{r}
genos_ont_gtscore_samplesBlood_lociDual_ds0x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds0x")

genos_ont_gtscore_samplesBlood_lociDual_ds6x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds6x")

genos_ont_gtscore_samplesBlood_lociDual_ds10x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds10x")

genos_ont_gtscore_samplesBlood_lociDual_ds15x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds15x")

genos_ont_gtscore_samplesBlood_lociDual_ds20x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds20x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds20x")

genos_ont_gtscore_samplesBlood_lociDual_ds25x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds25x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds25x")

genos_ont_gtscore_samplesBlood_lociDual_ds30x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds30x")

genos_ont_gtscore_samplesBlood_lociDual_ds50x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds50x")

genos_ont_gtscore_samplesBlood_lociDual_ds100x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds100x")
```

### Comparison dataframe

The dataframe below combines the genotypes from each of the downsampled (and original) genotype results from both Illumina and ONT.

```{r}
genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos <- purrr::reduce(
  list(
    genos_ill_gtscore_samplesBlood_lociDual_ds0x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds6x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds10x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds15x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds20x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds25x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds30x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds50x_l,
    genos_ill_gtscore_samplesBlood_lociDual_ds100x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds0x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds6x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds10x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds15x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds20x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds25x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds30x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds50x_l,
    genos_ont_gtscore_samplesBlood_lociDual_ds100x_l
    ),
  dplyr::left_join, by = c("sampleID", "locus")) %>%
  relocate(sampleID) %>%
  arrange(sampleID)
```

# 6 Geno success (blood only)

For the set of 30 blood samples, Illumina gives higher genotype success relative to Nanopore regardless of coverage, with the difference in average genotype success between Illumina and Nanopore ranging from 0.06 to 0.40. The magnitude of difference between Illumina and Nanopore genotype success increases with increasing read depth cutoffs, which is to be expected given the higher coverage in Illumina vs. Nanopore samples.  

**Table**

```{r, echo=FALSE}
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests <- 
  genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  select(-locus) %>%
  group_by(sampleID) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  as.data.frame() %>%
  pivot_longer(!sampleID,
               names_to = "depth",
               values_to = "totalGenos") %>%
  mutate(
    propSuccess = totalGenos/140
  ) %>%
  separate(., depth, into = c("seqPlatform", "depth"), sep = "_") %>%
  mutate(
    seqPlatform = case_when(
      seqPlatform == "ill" ~ "Illumina",
      seqPlatform == "ont" ~ "Nanopore"
    )
  )

# table
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests_summaryTable <- genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests %>%
  group_by(seqPlatform, depth) %>%
  summarise(
    meanSuccess = round(mean(propSuccess), 2)
  ) %>%
  as.data.frame() %>%
  pivot_wider(names_from = seqPlatform,
              values_from = meanSuccess) %>%
  arrange(factor(depth, c("ds0x", "ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x"))) %>%
  mutate(
    meanSuccess_diff = Illumina - Nanopore
  )
```

```{r, echo=FALSE}
# for html
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests_summaryTable %>%
  kbl(caption = "Mean geno success per sample for Illumina vs. Nanopore across coverage depths") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

**Plot**

```{r, echo=FALSE}
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests %>%
  ggplot(aes(x = factor(depth, c("ds0x", "ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x")), y = propSuccess, fill = seqPlatform)) +
  geom_boxplot() +
  theme_linedraw() +
  labs(
    title = "Blood sample genotype success across coverage depths for Illumina vs. Nanopore",
    x = "Downsampled sequencing depth per sample per locus",
    y = "Proportion dual-species loci genotyped per sample")
```

# 7 Optimize sample/loci combos

Given the difference in genotype success between Illumina and Nanopore results, prior to comparing genotype *consistency* we need to subset the samples and loci within the Illumina and Nanopore datasets so that we can 1) have the highest number of samples and loci possible while 2) ensuring that there is no missing data (i.e., we need 100% genotype success for our chosen combination of samples and loci).

## 7.1 Functions

To identify the optimal combination of samples and loci between **both** Illumina and Nanopore datasets, I've created two functions. Each function has a **version a** and a **version b**, where **version a** optimizes samples/loci based on a single dataset, and **version b** optimizes samples/loci based on two datasets.

### Function 1

**Function 1a:** For a given dataframe of genotypes, this function returns the number of loci genotyped at the desired proportion of n samples ("propSamples") for subsets of 2 to 30 samples

```{r}
# number of loci genotyped at propSamples (0 to 1.0) of samples from 2 to 30 samples
optimize_samplesLoci <- function(df, propSamples) {
  # Initialize an empty list to store results
  result_list <- list()

  for (n in 2:30) {
    # create matrix where NAs are 0 and non-NAs are 1
    is_na_vector <- ifelse(is.na(df), 0, 1)

    crossprod_vector <- crossprod(is_na_vector)
    col_sums <- colSums(crossprod_vector)

    x_df <- as.data.frame(df)  # to get meaningful colnames
    
    # use colSums vector to select n columns;
    # will rank all columns and give the n first;
    # gives n columns w/the max number of non-NA rows
    res <- x_df[, rank(-col_sums, ties.method = "first") <= n]

    # Store the result in the list
    result_list[[as.character(n)]] <- c(n, nrow(res[which(rowMeans(!is.na(res)) >= propSamples), ]))
  }

  # Convert the list to a dataframe
  result_df <- do.call(rbind, result_list) %>%
    as.data.frame()

  # Rename columns
  colnames(result_df) <- c("samples", "loci")

  return(result_df)
}
```

**Function 1b:** This function does the same as Function 1a, but allows for optimization based on two genotype files. This means that the specific samples included in any given "n" of "propSamples" are the **same** samples in both the Illumina and Nanopore datasets.

**NOTE:** The two genotype files must be formatted such that the rows and columns are in the exact same order. This may require some reformatting since illumina and nanopore samples were numbered differently.

```{r}
optimize_samplesLoci_illONT <- function(df1, df2, propSamples) {
  # Initialize an empty list to store results
  result_list <- list()
  
  # convert genos to binary matrix (NAs = 0, non-NAs = 1)
  matrix1 <- ifelse(is.na(df1), 0, 1)
  matrix2 <- ifelse(is.na(df2), 0, 1)
    
  # multiply matrices
  matrix_prod <- matrix1 * matrix2
  
  # create new combo df (turn 0 back to NA)
  df3 <- as.data.frame(matrix_prod) %>%
    mutate(across(everything(), ~na_if(., 0)))

  # identify sample/loci counts w/100% geno success
  for (n_samples in 2:30) {
  
    crossprod_vector <- crossprod(matrix_prod)
    col_sums <- colSums(crossprod_vector)

    x_df <- as.data.frame(df3)  # to get meaningful colnames
    
    # use colSums vector to select n columns;
    # will rank all columns and give the n first;
    # gives n columns w/the max number of non-NA rows
    res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

    # Store the result in the list
    result_list[[as.character(n_samples)]] <- c(n_samples, nrow(res[which(rowMeans(!is.na(res)) >= propSamples), ]))
  }

  # Convert the list to a dataframe
  result_df <- do.call(rbind, result_list) %>%
    as.data.frame()

  # Rename columns
  colnames(result_df) <- c("samples", "loci")

  return(result_df)
}
```

### Function 2

**Function 2a:** For a given dataframe of genotypes, this function returns an optimized set of genotypes for n_samples, where the number of samples and the proportion of those n samples successfully genotyped per locus (propSamples) is chosen based on the results of function 1 above

```{r}
optimized_genos <- function(df, n_samples, propSamples) {
  matrix <- ifelse(is.na(df), 0, 1)
  crossprod_vector <- crossprod(matrix)
  col_sums <- colSums(crossprod_vector)

  x_df <- as.data.frame(df)
  res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

  result <- res[which(rowMeans(!is.na(res)) >= propSamples), ]
  return(result)
}
```

**Function 2b:**

This function returns a list of two dataframes; access via x[[a]] and x[[2]].

**NOTE:** The two genotype files must be formatted such that the rows and columns are in the exact same order. This may require some reformatting since illumina and nanopore samples were numbered differently.

```{r}
optimized_genos_illONT <- function(df1, df2, n_samples, propSamples) {
  # create combo geno matrix
  matrix1 <- ifelse(is.na(df1), 0, 1)
  matrix2 <- ifelse(is.na(df2), 0, 1)
  matrix_prod <- matrix1 * matrix2
  
  # obtain crossproducts & col sums
  crossprod_vector <- crossprod(matrix_prod)
  col_sums <- colSums(crossprod_vector)

  x_df <- as.data.frame(matrix_prod) %>%
    mutate(across(everything(), ~na_if(., 0)))
  
  res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

  result <- res[which(rowMeans(!is.na(res)) >= propSamples), ]
  
  # subset each geno file to optimized sample/loci
  optGenos_df1 <- df1 %>%
    filter(row.names(.) %in% rownames(result)) %>%
    select(colnames(result))
  
  optGenos_df2 <- df2 %>%
    filter(row.names(.) %in% rownames(result)) %>%
    select(colnames(result))
  
  return(list(optGenos_df1, optGenos_df2))
}
```

## 7.2 Run functions

### Identify optimal sample/loci numbers

Below I'm using **function 1b** to identify the number of loci w/100% genotype success in both Illumina and Nanopore datasets for sample subsets of 2 to 30 samples across various read depth cutoffs.

```{r, echo=FALSE}
opt100_illONT_gtscore_samplesBlood_lociDual_dsTests <-
  optimize_samplesLoci_illONT(
    genos_ill_gtscore_samplesBlood_lociDual_ds0x,
    genos_ont_gtscore_samplesBlood_lociDual_ds0x,
    1.0
  ) %>%
  dplyr::rename("ds0x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds6x,
          genos_ont_gtscore_samplesBlood_lociDual_ds6x,
          1.0
          )
        ) %>%
  dplyr::rename("ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds10x,
          genos_ont_gtscore_samplesBlood_lociDual_ds10x,
          1.0
          )
        ) %>%
  dplyr::rename("ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds15x,
          genos_ont_gtscore_samplesBlood_lociDual_ds15x,
          1.0
          )
        ) %>%
  dplyr::rename("ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds20x,
          genos_ont_gtscore_samplesBlood_lociDual_ds20x,
          1.0
          )
        ) %>%
  dplyr::rename("ds20x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds25x,
          genos_ont_gtscore_samplesBlood_lociDual_ds25x,
          1.0
          )
        ) %>%
  dplyr::rename("ds25x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds30x,
          genos_ont_gtscore_samplesBlood_lociDual_ds30x,
          1.0
          )
        ) %>%
  dplyr::rename("ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds50x,
          genos_ont_gtscore_samplesBlood_lociDual_ds50x,
          1.0
          )
        ) %>%
  dplyr::rename("ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds100x,
          genos_ont_gtscore_samplesBlood_lociDual_ds100x,
          1.0
          )
        ) %>%
  dplyr::rename("ds100x" = "loci")
```

```{r, echo=FALSE}
# for html
opt100_illONT_gtscore_samplesBlood_lociDual_dsTests %>%
  kbl(caption = "Number of loci w/100% geno success in both Illumina & Nanopore datasets across sample subsets & coverage cutoffs") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "500px", fixed_thead = list(enabled = T))
```

### Subset optimal samples/loci

Based on the results above, I'm opting to go with 10 samples for all read depths along with their respective loci that have 100% genotype success in both Illumina and Nanopore datasets. Below, I'm using **function 2b** to create the new sample/loci subsets for both Illumina and Nanopore datasets at each read depth cutoff.

```{r}
# ds0x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds0x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds0x,
  genos_ont_gtscore_samplesBlood_lociDual_ds0x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds0x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds0x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds0x[[2]]

# ds6x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds6x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds6x,
  genos_ont_gtscore_samplesBlood_lociDual_ds6x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds6x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds6x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds6x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds6x[[2]]

# ds10x 
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds10x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds10x,
  genos_ont_gtscore_samplesBlood_lociDual_ds10x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds10x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds10x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds10x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds10x[[2]]

# ds15x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds15x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds15x,
  genos_ont_gtscore_samplesBlood_lociDual_ds15x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds15x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds15x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds15x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds15x[[2]]

# ds20x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds20x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds20x,
  genos_ont_gtscore_samplesBlood_lociDual_ds20x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds20x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds20x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds20x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds20x[[2]]

# ds25x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds25x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds25x,
  genos_ont_gtscore_samplesBlood_lociDual_ds25x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds25x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds25x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds25x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds25x[[2]]

# ds30x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds30x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds30x,
  genos_ont_gtscore_samplesBlood_lociDual_ds30x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds30x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds30x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds30x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds30x[[2]]

# ds50x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds50x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds50x,
  genos_ont_gtscore_samplesBlood_lociDual_ds50x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds50x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds50x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds50x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds50x[[2]]

# ds100x
optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds100x,
  genos_ont_gtscore_samplesBlood_lociDual_ds100x,
  10,
  1.0
)

optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x[[1]]

optGenos100_ont_gtscore_samplesBlood_lociDual_ds100x <- optGenos100_illONT_gtscore_samplesBlood_lociDual_ds100x[[2]]
```

## 7.3 Subset data

```{r}
# ds6x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds6x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds6x)) %>%
  select(c(sampleID, locus, ill_ds6x, ont_ds6x)) %>%
  mutate(
    match = case_when(
      ill_ds6x == ont_ds6x ~ "yes",
      .default = "no"
    )
  )

# ds10x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds10x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds10x)) %>%
  select(c(sampleID, locus, ill_ds10x, ont_ds10x)) %>%
  mutate(
    match = case_when(
      ill_ds10x == ont_ds10x ~ "yes",
      .default = "no"
    )
  )

# ds15x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds15x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds15x)) %>%
  select(c(sampleID, locus, ill_ds15x, ont_ds15x)) %>%
  mutate(
    match = case_when(
      ill_ds15x == ont_ds15x ~ "yes",
      .default = "no"
    )
  )

# ds20x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds20x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds20x)) %>%
  select(c(sampleID, locus, ill_ds20x, ont_ds20x)) %>%
  mutate(
    match = case_when(
      ill_ds20x == ont_ds20x ~ "yes",
      .default = "no"
    )
  )

# ds25x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds25x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds25x)) %>%
  select(c(sampleID, locus, ill_ds25x, ont_ds25x)) %>%
  mutate(
    match = case_when(
      ill_ds25x == ont_ds25x ~ "yes",
      .default = "no"
    )
  )

# ds30x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds30x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds30x)) %>%
  select(c(sampleID, locus, ill_ds30x, ont_ds30x)) %>%
  mutate(
    match = case_when(
      ill_ds30x == ont_ds30x ~ "yes",
      .default = "no"
    )
  )

# ds50x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds50x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds50x)) %>%
  select(c(sampleID, locus, ill_ds50x, ont_ds50x)) %>%
  mutate(
    match = case_when(
      ill_ds50x == ont_ds50x ~ "yes",
      .default = "no"
    )
  )

# ds100x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_samplesBlood_lociDual_ds100x)) %>%
  select(c(sampleID, locus, ill_ds100x, ont_ds100x)) %>%
  mutate(
    match = case_when(
      ill_ds100x == ont_ds100x ~ "yes",
      .default = "no"
    )
  )
```

## 7.4 Sample/loci lists

```{r}
sampleList_ds10x <- unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$sampleID)

lociList_ds10x <- unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$locus)

# create file key file
fileKey_ds10x <- md_tamRun4 %>%
  filter(sampleID_unique %in% sampleList_ds10x) %>%
  select(c(sampleID_unique, sampleFile_ill, sampleFile_ont)) %>%
  dplyr::rename(animalID_sampleType = sampleID_unique)

write.csv(fileKey_ds10x, "./paper1_nanoporeValidation/fastqFiles_ds10x/illONT_fileKey.csv", row.names = F)

# create loci list file
primerProbeFile_ds10x <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T) %>%
  filter(Locus %in% lociList_ds10x) %>%
  dplyr::rename(
    locus = Locus,
    ploidy = Ploidy,
    snpPos = SNPpos,
    allele1 = Allele1,
    allele2 = Allele2,
    probe1 = Probe1,
    probe2 = Probe2,
    forwardPrimer = Primer)

write.csv(primerProbeFile_ds10x, "./paper1_nanoporeValidation/files_toShare_ds10x/primerProbeFile_ds10x.csv", row.names = F)

test <- read.csv("primerProbeFile_ds10x.csv")

write.table(test, file = "primerProbeFile_ontValidation_toShare.txt", sep = "\t", row.names = F, col.names = F, quote = F)

# geno file
illONT_genoFile_ds10x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x

write.csv(illONT_genoFile_ds10x, "./paper1_nanoporeValidation/files_toShare_ds10x/genos_illONT_ds10x_opt100.csv", row.names = F)
```

##### Summary

```{r}
genoCompare_illONT_samplesBlood_opt100_dsTests <- 
  data.frame(
    coverage = c("ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x"),
    totalMismatch = c(
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$match == "no")
    ),
    totalMatch = c(
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$match == "yes")
    ),
    totalSamples = c(
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$sampleID))
    ),
    totalLoci = c(
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$locus))
    )
  ) %>%
  mutate(
    totalGenos = totalMismatch + totalMatch,
    propMatch = round(totalMatch/totalGenos, 3)
  ) %>%
  select(c(coverage, totalSamples, totalLoci, totalGenos, totalMismatch, totalMatch, propMatch))
```




# 8 Geno consistency

```{r}
# illumina
optGenos100_ill_gtscore_samplesBlood_lociDual_dsAll <- 


# ds6x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds6x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds6x)) %>%
  select(c(sampleID, locus, ill_ds6x, ont_ds6x)) %>%
  mutate(
    match = case_when(
      ill_ds6x == ont_ds6x ~ "yes",
      .default = "no"
    )
  )

# ds10x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds10x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds10x)) %>%
  select(c(sampleID, locus, ill_ds10x, ont_ds10x)) %>%
  mutate(
    match = case_when(
      ill_ds10x == ont_ds10x ~ "yes",
      .default = "no"
    )
  )

# ds15x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds15x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds15x)) %>%
  select(c(sampleID, locus, ill_ds15x, ont_ds15x)) %>%
  mutate(
    match = case_when(
      ill_ds15x == ont_ds15x ~ "yes",
      .default = "no"
    )
  )

# ds20x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds20x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds20x)) %>%
  select(c(sampleID, locus, ill_ds20x, ont_ds20x)) %>%
  mutate(
    match = case_when(
      ill_ds20x == ont_ds20x ~ "yes",
      .default = "no"
    )
  )

# ds25x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds25x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds25x)) %>%
  select(c(sampleID, locus, ill_ds25x, ont_ds25x)) %>%
  mutate(
    match = case_when(
      ill_ds25x == ont_ds25x ~ "yes",
      .default = "no"
    )
  )

# ds30x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds30x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds30x)) %>%
  select(c(sampleID, locus, ill_ds30x, ont_ds30x)) %>%
  mutate(
    match = case_when(
      ill_ds30x == ont_ds30x ~ "yes",
      .default = "no"
    )
  )

# ds50x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds50x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds50x)) %>%
  select(c(sampleID, locus, ill_ds50x, ont_ds50x)) %>%
  mutate(
    match = case_when(
      ill_ds50x == ont_ds50x ~ "yes",
      .default = "no"
    )
  )

# ds100x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds100x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds100x)) %>%
  select(c(sampleID, locus, ill_ds100x, ont_ds100x)) %>%
  mutate(
    match = case_when(
      ill_ds100x == ont_ds100x ~ "yes",
      .default = "no"
    )
  )
```





### gtscore

#### Illumina

##### qcUnfiltered

Original files + sample subsets

```{r, echo=FALSE}
# All samples
genos_ill_gtscore_samplesAll_lociDual <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

# Sample-type subsets
genos_ill_gtscore_samplesBlood_lociDual <- genos_ill_gtscore_samplesAll_lociDual %>%
  select(c(md_tamRun4_blood$sampleID_ill))

genos_ill_gtscore_samplesFecal_lociDual <- genos_ill_gtscore_samplesAll_lociDual %>%
  select(c(md_tamRun4_fecal$sampleID_ill))

genos_ill_gtscore_samplesHair_lociDual <- genos_ill_gtscore_samplesAll_lociDual %>%
  select(c(md_tamRun4_hair$sampleID_ill))
```

Long format genos

```{r, echo=FALSE}
genos_ill_gtscore_samplesAll_lociDual_l <- genos_ill_gtscore_samplesAll_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_qc0") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)
```

##### qcFiltered

Original files + sample subsets

```{r, echo=FALSE}
# All samples
genos_ill_qcFiltered_gtscore_samplesAll_lociDual <- read.table("./04_tamRun4/04_run4QualityFilterTests/03_run4Analyses_qcFiltered/illONT_qcFiltered_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  select(starts_with("tamRun4")) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))
```

Long format genos

```{r, echo=FALSE}
genos_ill_qcFiltered_gtscore_samplesAll_lociDual_l <- genos_ill_qcFiltered_gtscore_samplesAll_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_qc30") %>%
  separate(., sampleID, into = c("sampleID", "qc"), sep = "_") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID, -qc)
```

#### Nanopore

##### qcUnfiltered

Original files + sample subsets

```{r, echo=FALSE}
# All samples
genos_ont_gtscore_samplesAll_lociDual <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

# Sample-type subsets
genos_ont_gtscore_samplesBlood_lociDual <- genos_ont_gtscore_samplesAll_lociDual %>%
  select(c(md_tamRun4_blood$sampleID_ont))

genos_ont_gtscore_samplesFecal_lociDual <- genos_ont_gtscore_samplesAll_lociDual %>%
  select(c(md_tamRun4_fecal$sampleID_ont))

genos_ont_gtscore_samplesHair_lociDual <- genos_ont_gtscore_samplesAll_lociDual %>%
  select(c(md_tamRun4_hair$sampleID_ont))
```

```{r, echo=FALSE}
# For html
genos_ill_gtscore_samplesAll_lociDual %>%
  kbl(caption = "Illumina GTscore 10x genotypes") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

genos_ont_gtscore_samplesAll_lociDual %>%
  kbl(caption = "Nanopore GTscore 10x genotypes") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

Long format

```{r, echo=FALSE}
genos_ont_gtscore_samplesAll_lociDual_l <- genos_ont_gtscore_samplesAll_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_qc0") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)
```

##### qcFiltered

Original files + sample subsets

```{r, echo=FALSE}
# All samples
genos_ont_qcFiltered_gtscore_samplesAll_lociDual <- read.table("./04_tamRun4/04_run4QualityFilterTests/03_run4Analyses_qcFiltered/illONT_qcFiltered_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  select(starts_with("barcode")) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))
```

Long format

```{r, echo=FALSE}
genos_ont_qcFiltered_gtscore_samplesAll_lociDual_l <- genos_ont_qcFiltered_gtscore_samplesAll_lociDual %>%
  rename_with(~ str_remove(., "_interleaved"), everything()) %>%
  rownames_to_column("locus") %>%
  pivot_longer(.,
               cols = starts_with("barcode"),
               names_to = c("sampleID", ".value"),
               names_sep = "_") %>%
  relocate(sampleID) %>%
  # remove any rows w/NA geno for all filters
  filter(if_any(c(qc11, qc13, qc15, qc20), complete.cases)) %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  select(-sampleID) %>%
  relocate(sampleID_unique) %>%
  dplyr::rename(
    "ont_qc11" = "qc11",
    "ont_qc13" = "qc13",
    "ont_qc15" = "qc15",
    "ont_qc20" = "qc20"
  )
```

### gtseq

Original files + sample subsets

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# All samples
genos_ill_gtseq_samplesAll_lociDual <- 
  # need read_csv vs. read.csv here
  read_csv("./04_tamRun4/00_illumina/04_run4GTseq/genos/ill_tamRun4_compiledGenos_gtseq.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  column_to_rownames("Sample") %>%
  select(c(primerList.v3_lociDual$locus)) %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), ~ str_c(str_sub(., 1, 1), ",", str_sub(., 2, 2)))) %>%
  # recode 0,0 as NA
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))

genos_ont_gtseq_samplesAll_lociDual <- 
  # need read_csv vs. read.csv here
  read_csv("./04_tamRun4/01_nanopore/04_run4GTseq/genos/ont_tamRun4_compiledGenos_gtseq.csv") %>%
  as.data.frame() %>%
  mutate(Sample = str_sub(Sample, 1, 9)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  column_to_rownames("Sample") %>%
  select(c(primerList.v3_lociDual$locus)) %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), ~ str_c(str_sub(., 1, 1), ",", str_sub(., 2, 2)))) %>%
  # recode 0,0 as NA
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))

# Blood samples
genos_ill_gtseq_samplesBlood_lociDual <- genos_ill_gtseq_samplesAll_lociDual %>%
  select(c(md_tamRun4_blood$sampleID_ill))

genos_ont_gtseq_samplesBlood_lociDual <- genos_ont_gtseq_samplesAll_lociDual %>%
  select(c(md_tamRun4_blood$sampleID_ont))

# Fecal samples
genos_ill_gtseq_samplesFecal_lociDual <- genos_ill_gtseq_samplesAll_lociDual %>%
  select(c(md_tamRun4_fecal$sampleID_ill))

genos_ont_gtseq_samplesFecal_lociDual <- genos_ont_gtseq_samplesAll_lociDual %>%
  select(c(md_tamRun4_fecal$sampleID_ont))

# Hair samples
genos_ill_gtseq_samplesHair_lociDual <- genos_ill_gtseq_samplesAll_lociDual %>%
  select(c(md_tamRun4_hair$sampleID_ill))

genos_ont_gtseq_samplesHair_lociDual <- genos_ont_gtseq_samplesAll_lociDual %>%
  select(c(md_tamRun4_hair$sampleID_ont))
```

```{r, echo=FALSE}
# For html
genos_ill_gtseq_samplesAll_lociDual %>%
  kbl(caption = "Illumina GTseq 10x genotypes") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

genos_ont_gtseq_samplesAll_lociDual %>%
  kbl(caption = "Nanopore GTseq 10x genotypes") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 4 Blood samples

## 4.1 Geno success

### By sample

#### gtscore.10x

```{r, echo=FALSE}
# Geno success by sample
genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual <- genos_ill_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina"
  )

genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual <- genos_ont_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual <- rbind(genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual, genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_uniqueID <- genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill")

genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_uniqueID <- genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont")

genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_uniqueID <- merge(genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_uniqueID, genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_uniqueID, by = "sampleID_unique") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(sampleID_unique, sampleID_ill, sampleID_ont, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_uniqueID %>%
  kbl(caption = "GTscore.10x blood sample genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### gtseq

```{r, echo=FALSE}
# Illumina
genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual <- genos_ill_gtseq_samplesBlood_lociDual %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual <- genos_ont_gtseq_samplesBlood_lociDual %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual <- rbind(genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual, genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual_uniqueID <- genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill") %>%
  arrange(sampleID_unique)

genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual_uniqueID <- genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont") %>%
  arrange(sampleID_unique)

genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual_uniqueID <- merge(genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual_uniqueID, genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual_uniqueID, by = "sampleID_unique") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(sampleID_unique, sampleID_ill, sampleID_ont, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual_uniqueID %>%
  kbl(caption = "GTseq blood sample genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Summary

```{r, echo=FALSE}
summary_genoSuccess_bySample_samplesBlood_lociDual <- data.frame(
  pipeline = c("gtscore", "gtscore", "gtseq", "gtseq"),
  seqPlatform = rep(c("Illumina", "Nanopore"), 2),
  sampleType = rep("blood", 4),
  min_genoSuccess = c(
    min(genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual$propSuccess),
    min(genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual$propSuccess),
    min(genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual$propSuccess),
    min(genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual$propSuccess)
  ),
  max_genoSuccess = c(
    max(genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual$propSuccess),
    max(genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual$propSuccess),
    max(genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual$propSuccess),
    max(genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual$propSuccess)
  ),
  mean_genoSuccess = c(
    mean(genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual$propSuccess),
    mean(genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual$propSuccess),
    mean(genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual$propSuccess),
    mean(genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual$propSuccess)
  )
) %>%
  mutate(
    mean_genoSuccess = round(mean_genoSuccess, 2),
    min_genoSuccess = round(min_genoSuccess, 2),
    max_genoSuccess = round(max_genoSuccess, 2)
    )
```

```{r, echo=FALSE}
# For html
summary_genoSuccess_bySample_samplesBlood_lociDual %>%
  kbl(caption = "Avg genotype success for blood samples for Illumina vs. Nanopore in GTscore vs. GTseq pipelines") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Plots

```{r, echo=FALSE}
# Distribution barplots - geno success by sample
## GTscore
p1 <- ggplot(genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 140, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTscore (10x): Number of blood samples per total loci genotyped") +
  theme_bw() +
  labs(x = "Number of blood samples", y = "Total loci genotyped")

## GTseq
p2 <- ggplot(genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 140, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTseq: Number of blood samples per total loci genotyped") +
  theme_bw() +
  labs(x = "Number of blood samples", y = "Total loci genotyped")

p1
p2
```

```{r, echo=FALSE}
# Dumbbell plot - geno success by sample
## GTscore
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_uniqueID_l <- rbind(genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_uniqueID[, c("sampleID_unique", "totalGenos", "propSuccess", "seqPlatform")],
        genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_uniqueID[, c("sampleID_unique", "totalGenos", "propSuccess", "seqPlatform")]                                                                    )

genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_uniqueID_l %>%
  arrange(sampleID_unique) %>%
  mutate(paired = rep(1:(n()/2), each = 2),
         sampleID_unique = factor(sampleID_unique)) %>%
  ggplot(aes(x = propSuccess, y = reorder(sampleID_unique, propSuccess))) +
  geom_line(aes(group = paired)) +
  geom_point(aes(color = seqPlatform)) +
  theme_linedraw() +
  labs(
    title = "GTscore sample genotype success for Illumina vs. Nanopore runs",
    y = "sampleID",
    x = "proportion dual-species loci genotyped")

## GTseq
genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual_uniqueID_l <- rbind(genoSuccess_bySample_ill_gtseq_samplesBlood_lociDual_uniqueID[, c("sampleID_unique", "totalGenos", "propSuccess", "seqPlatform")],
        genoSuccess_bySample_ont_gtseq_samplesBlood_lociDual_uniqueID[, c("sampleID_unique", "totalGenos", "propSuccess", "seqPlatform")]                                                                    )

genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual_uniqueID_l %>%
  arrange(sampleID_unique) %>%
  mutate(paired = rep(1:(n()/2), each = 2),
         sampleID_unique = factor(sampleID_unique)) %>%
  ggplot(aes(x = propSuccess, y = reorder(sampleID_unique, propSuccess))) +
  geom_line(aes(group = paired)) +
  geom_point(aes(color = seqPlatform)) +
  theme_linedraw() +
  labs(
    title = "GTseq sample genotype success for Illumina vs. Nanopore runs",
    y = "sampleID",
    x = "proportion dual-species loci genotyped")
```

```{r, echo=FALSE, warning=FALSE}
# Compare across platforms & pipeline
genoSuccess_bySample_illONT_gtscoreSeq_samplesBlood_lociDual_uniqueID_l <-
  genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_uniqueID_l %>%
  mutate(seqPlatform = str_c(seqPlatform, "_gtscore")) %>%
  rbind(.,
        genoSuccess_bySample_illONT_gtseq_samplesBlood_lociDual_uniqueID_l) %>%
  mutate(seqPlatform = case_when(
    seqPlatform == "Illumina" ~ "Illumina_gtseq",
    seqPlatform == "Nanopore" ~ "Nanopore_gtseq",
    .default = seqPlatform
  ))

genoSuccess_bySample_illONT_gtscoreSeq_samplesBlood_lociDual_uniqueID_l %>%
  arrange(sampleID_unique) %>%
  mutate(paired = rep(1:(n()/4), each = 4),
         sampleID_unique = factor(sampleID_unique)) %>%
  ggplot(aes(x = propSuccess, y = reorder(sampleID_unique, propSuccess))) +
  geom_point(aes(color = seqPlatform)) +
  theme_linedraw() +
  labs(
    title = "Sample genotype success across platforms & pipelines",
    y = "sampleID",
    x = "Proportion dual-species loci genotyped")

ggplot(
  genoSuccess_bySample_illONT_gtscoreSeq_samplesBlood_lociDual_uniqueID_l,
  aes(x = reorder(sampleID_unique, propSuccess), y = propSuccess, fill = seqPlatform)) +
  geom_bar(stat="identity", position = position_dodge(width = 0), width = 2) +
  coord_flip() +
  theme_linedraw() +
  labs(
    title = "Sample geno success across platforms & pipelines",
    x = "sampleID",
    y = "Proportion of dual-species loci genotyped"
  )
```

### By locus

#### gtscore.10x

```{r, echo=FALSE}
# Illumina
genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual <- genos_ill_gtscore_samplesBlood_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual <- genos_ont_gtscore_samplesBlood_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual <- rbind(genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual, genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual_uniqueID <- genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual_uniqueID <- genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual_uniqueID <- merge(genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual_uniqueID, genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual_uniqueID, by = "locus") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(locus, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual_uniqueID %>%
  kbl(caption = "GTscore.10x loci genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### gtseq

```{r, echo=FALSE}
# Illumina
genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual <- genos_ill_gtseq_samplesBlood_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual <- genos_ont_gtseq_samplesBlood_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual <- rbind(genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual, genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual_uniqueID <- genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual_uniqueID <- genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual_uniqueID <- merge(genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual_uniqueID, genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual_uniqueID, by = "locus") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(locus, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual_uniqueID %>%
  kbl(caption = "GTseq loci genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Summary

```{r, echo=FALSE}
summary_genoSuccess_byLocus_samplesBlood_lociDual <- data.frame(
  pipeline = c("gtscore", "gtscore", "gtseq", "gtseq"),
  seqPlatform = rep(c("Illumina", "Nanopore"), 2),
  sampleType = rep("blood", 4),
  min_genoSuccess = c(
    min(genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual$propSuccess),
    min(genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual$propSuccess),
    min(genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual$propSuccess),
    min(genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual$propSuccess)
  ),
  max_genoSuccess = c(
    max(genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual$propSuccess),
    max(genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual$propSuccess),
    max(genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual$propSuccess),
    max(genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual$propSuccess)
  ),
  mean_genoSuccess = c(
    mean(genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual$propSuccess)
  )
) %>%
  mutate(
    mean_genoSuccess = round(mean_genoSuccess, 2),
    min_genoSuccess = round(min_genoSuccess, 2),
    max_genoSuccess = round(max_genoSuccess, 2)
    )
```

```{r, echo=FALSE}
# For html
summary_genoSuccess_byLocus_samplesBlood_lociDual %>%
  kbl(caption = "Overview of genotype success by locus for Illumina vs. Nanopore blood samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Plots

```{r, echo=FALSE}
p3 <- ggplot(genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 30, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTscore (10x): Number of loci per total blood samples genotyped") +
  theme_bw() +
  labs(x = "Number of loci", y = "Total blood samples genotyped")

# GTseq
p4 <- ggplot(genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 30, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTseq: Number of loci per total blood samples genotyped") +
  theme_bw() +
  labs(x = "Number of loci", y = "Total blood samples genotyped")

p3
p4
```

```{r, echo=FALSE}
# Sample comparison dumbbell plot
## GTscore
genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual_uniqueID_l <- rbind(genoSuccess_byLocus_ill_gtscore_samplesBlood_lociDual_uniqueID,
        genoSuccess_byLocus_ont_gtscore_samplesBlood_lociDual_uniqueID                                                                    )

genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual_uniqueID_l %>%
  arrange(locus) %>%
  mutate(paired = rep(1:(n()/2), each = 2),
         locus = factor(locus)) %>%
  ggplot(aes(x = propSuccess, y = reorder(locus, propSuccess))) +
  geom_line(aes(group = paired)) +
  geom_point(aes(color = seqPlatform)) +
  theme_linedraw() +
  labs(
    title = "GTscore locus genotype success in Illumina vs. Nanopore runs",
    y = "locus",
    x = "proportion samples genotyped")

## GTseq
genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual_uniqueID_l <- rbind(genoSuccess_byLocus_ill_gtseq_samplesBlood_lociDual_uniqueID,
        genoSuccess_byLocus_ont_gtseq_samplesBlood_lociDual_uniqueID                                                                    )

genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual_uniqueID_l %>%
  arrange(locus) %>%
  mutate(paired = rep(1:(n()/2), each = 2),
         locus = factor(locus)) %>%
  ggplot(aes(x = propSuccess, y = reorder(locus, propSuccess))) +
  geom_line(aes(group = paired)) +
  geom_point(aes(color = seqPlatform)) +
  theme_linedraw() +
  labs(
    title = "GTseq locus genotype success for Illumina vs. Nanopore runs",
    y = "locus",
    x = "proportion samples genotyped")
```

```{r, echo=FALSE, warning=FALSE}
# Compare across platforms & pipeline
genoSuccess_byLocus_illONT_gtscoreSeq_samplesBlood_lociDual_uniqueID_l <-
  genoSuccess_byLocus_illONT_gtscore_samplesBlood_lociDual_uniqueID_l %>%
  mutate(seqPlatform = str_c(seqPlatform, "_gtscore")) %>%
  rbind(.,
        genoSuccess_byLocus_illONT_gtseq_samplesBlood_lociDual_uniqueID_l) %>%
  mutate(seqPlatform = case_when(
    seqPlatform == "Illumina" ~ "Illumina_gtseq",
    seqPlatform == "Nanopore" ~ "Nanopore_gtseq",
    .default = seqPlatform
  ))

genoSuccess_byLocus_illONT_gtscoreSeq_samplesBlood_lociDual_uniqueID_l %>%
  arrange(locus) %>%
  mutate(paired = rep(1:(n()/4), each = 4),
         locus = factor(locus)) %>%
  ggplot(aes(x = propSuccess, y = reorder(locus, propSuccess))) +
  geom_point(aes(color = seqPlatform)) +
  theme_linedraw() +
  labs(
    title = "Locus geno success across platforms & pipelines",
    y = "sampleID",
    x = "proportion samples genotyped")

ggplot(
  genoSuccess_byLocus_illONT_gtscoreSeq_samplesBlood_lociDual_uniqueID_l,
  aes(x = reorder(locus, propSuccess), y = propSuccess, fill = seqPlatform)) +
  geom_bar(stat="identity", position = position_dodge(width = 0), width = 2) +
  coord_flip() +
  theme_linedraw() +
  labs(
    title = "Locus geno success across platforms & pipelines",
    x = "sampleID",
    y = "Proportion of samples genotyped"
  )
```

## 4.2 Identify optimal sample/loci combos

### Functions

#### Functions 1a & 1b

**Function 1a:** For a given dataframe of genotypes, this function returns the number of loci genotyped at the desired proportion of n samples ("propSamples") for subsets of 2 to 30 samples

```{r}
# number of loci genotyped at propSamples (0 to 1.0) of samples from 2 to 30 samples
optimize_samplesLoci <- function(df, propSamples) {
  # Initialize an empty list to store results
  result_list <- list()

  for (n in 2:30) {
    # create matrix where NAs are 0 and non-NAs are 1
    is_na_vector <- ifelse(is.na(df), 0, 1)

    crossprod_vector <- crossprod(is_na_vector)
    col_sums <- colSums(crossprod_vector)

    x_df <- as.data.frame(df)  # to get meaningful colnames
    
    # use colSums vector to select n columns;
    # will rank all columns and give the n first;
    # gives n columns w/the max number of non-NA rows
    res <- x_df[, rank(-col_sums, ties.method = "first") <= n]

    # Store the result in the list
    result_list[[as.character(n)]] <- c(n, nrow(res[which(rowMeans(!is.na(res)) >= propSamples), ]))
  }

  # Convert the list to a dataframe
  result_df <- do.call(rbind, result_list) %>%
    as.data.frame()

  # Rename columns
  colnames(result_df) <- c("samples", "loci")

  return(result_df)
}
```

**Function 1b:** This function does the same as Function 1a, but allows for optimization based on two genotype files.

**NOTE:** The two genotype files must be formatted such that the rows and columns are in the exact same order. This may require some reformatting since illumina and nanopore samples were numbered differently.

```{r}
optimize_samplesLoci_illONT <- function(df1, df2, propSamples) {
  # Initialize an empty list to store results
  result_list <- list()
  
  # convert genos to binary matrix (NAs = 0, non-NAs = 1)
  matrix1 <- ifelse(is.na(df1), 0, 1)
  matrix2 <- ifelse(is.na(df2), 0, 1)
    
  # multiply matrices
  matrix_prod <- matrix1 * matrix2
  
  # create new combo df (turn 0 back to NA)
  df3 <- as.data.frame(matrix_prod) %>%
    mutate(across(everything(), ~na_if(., 0)))

  # identify sample/loci counts w/100% geno success
  for (n_samples in 2:30) {
  
    crossprod_vector <- crossprod(matrix_prod)
    col_sums <- colSums(crossprod_vector)

    x_df <- as.data.frame(df3)  # to get meaningful colnames
    
    # use colSums vector to select n columns;
    # will rank all columns and give the n first;
    # gives n columns w/the max number of non-NA rows
    res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

    # Store the result in the list
    result_list[[as.character(n_samples)]] <- c(n_samples, nrow(res[which(rowMeans(!is.na(res)) >= propSamples), ]))
  }

  # Convert the list to a dataframe
  result_df <- do.call(rbind, result_list) %>%
    as.data.frame()

  # Rename columns
  colnames(result_df) <- c("samples", "loci")

  return(result_df)
}
```

#### Functions 2a & b

**Function 2a:** For a given dataframe of genotypes, this function returns an optimized set of genotypes for n_samples, where the number of samples and the proportion of those n samples successfully genotyped per locus (propSamples) is chosen based on the results of function 1 above

```{r}
optimized_genos <- function(df, n_samples, propSamples) {
  matrix <- ifelse(is.na(df), 0, 1)
  crossprod_vector <- crossprod(matrix)
  col_sums <- colSums(crossprod_vector)

  x_df <- as.data.frame(df)
  res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

  result <- res[which(rowMeans(!is.na(res)) >= propSamples), ]
  return(result)
}
```

**Function 2b:**

This function returns a list of two dataframes; access via x[[a]] and x[[2]].

**NOTE:** The two genotype files must be formatted such that the rows and columns are in the exact same order. This may require some reformatting since illumina and nanopore samples were numbered differently.

```{r}
optimized_genos_illONT <- function(df1, df2, n_samples, propSamples) {
  # create combo geno matrix
  matrix1 <- ifelse(is.na(df1), 0, 1)
  matrix2 <- ifelse(is.na(df2), 0, 1)
  matrix_prod <- matrix1 * matrix2
  
  # obtain crossproducts & col sums
  crossprod_vector <- crossprod(matrix_prod)
  col_sums <- colSums(crossprod_vector)

  x_df <- as.data.frame(matrix_prod) %>%
    mutate(across(everything(), ~na_if(., 0)))
  
  res <- x_df[, rank(-col_sums, ties.method = "first") <= n_samples]

  result <- res[which(rowMeans(!is.na(res)) >= propSamples), ]
  
  # subset each geno file to optimized sample/loci
  optGenos_df1 <- df1 %>%
    filter(row.names(.) %in% rownames(result)) %>%
    select(colnames(result))
  
  optGenos_df2 <- df2 %>%
    filter(row.names(.) %in% rownames(result)) %>%
    select(colnames(result))
  
  return(list(optGenos_df1, optGenos_df2))
}
```

### Identify optimal sample/loci combos

#### Fun opt function

```{r}
# reformat genos
genos_ill_gtscore_samplesBlood_lociDual_sampleID_unique <- genos_ill_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID_ill) %>%
  arrange(as.numeric(gsub("_blood", "", sampleID_unique))) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

genos_ont_gtscore_samplesBlood_lociDual_sampleID_unique <- genos_ont_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID_ont) %>%
  arrange(as.numeric(gsub("_blood", "", sampleID_unique))) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# run optimization function
opt100_samplesLoci_illONT_gtscore_blood_ds0x <- optimize_samplesLoci_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_sampleID_unique,
  genos_ont_gtscore_samplesBlood_lociDual_sampleID_unique,
  1.0
)
```

#### Summaries

Based on the optimization results, I would choose the following sample/loci subsets for blood samples based on Illumina results:

GTscore.10x - 100% of sample subset genotyped: 20 samples + 122 loci GTscore.10x - 90% of sample subset genotyped: 30 samples + 103 loci GTseq - 100% of sample subset genotyped: 10 samples + 56 loci GTseq - 90% of sample subset genotyped: 10 samples + 94 loci

```{r, echo=FALSE}
# loci genotyped at 100% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_blood_summary <- merge(opt100_samplesLoci_ill_gtscore_blood, opt100_samplesLoci_ont_gtscore_blood, by = "samples") %>%
  merge(., opt100_samplesLoci_ill_gtseq_blood, by = "samples") %>%
  merge(., opt100_samplesLoci_ont_gtseq_blood, by = "samples")

# loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_blood_summary <- merge(opt90_samplesLoci_ill_gtscore_blood, opt90_samplesLoci_ont_gtscore_blood, by = "samples") %>%
  merge(., opt90_samplesLoci_ill_gtseq_blood, by = "samples") %>%
  merge(., opt90_samplesLoci_ont_gtseq_blood, by = "samples")
```

```{r, echo=FALSE}
# For html
opt100_samplesLoci_blood_summary %>%
  kbl(caption = "Sample/loci subsets for which loci are genotyped at 100% of samples in subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

opt90_samplesLoci_blood_summary %>%
  kbl(caption = "Sample/loci subsets for which loci are genotyped at 90% of samples in subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Dataframes

#### 100% geno success

##### gtscore

```{r, echo=FALSE}
optGenos100_ill_gtscore_blood <- optimized_genos(genos_ill_gtscore_samplesBlood_lociDual, 20, 1.0)

optGenos100_ont_gtscore_blood <- optimized_genos(genos_ont_gtscore_samplesBlood_lociDual, 20, 1.0)
```

```{r, echo=FALSE}
# For html
optGenos100_ill_gtscore_blood %>%
  kbl(caption = "Illumina-GTscore optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos100_ont_gtscore_blood %>%
  kbl(caption = "Nanopore-GTscore optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

##### gtseq

```{r, echo=FALSE}
optGenos100_ill_gtseq_blood <- optimized_genos(genos_ill_gtseq_samplesBlood_lociDual, 10, 1.0)

optGenos100_ont_gtseq_blood <- optimized_genos(genos_ont_gtseq_samplesBlood_lociDual, 10, 1.0)
```

```{r, echo=FALSE}
# For html
optGenos100_ill_gtseq_blood %>%
  kbl(caption = "Illumina-GTseq optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos100_ont_gtseq_blood %>%
  kbl(caption = "Nanopore-GTseq optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### 90% geno success

##### gtscore

```{r, echo=FALSE}
optGenos90_ill_gtscore_blood <- optimized_genos(genos_ill_gtscore_samplesBlood_lociDual, 30, 0.9)

optGenos90_ont_gtscore_blood <- optimized_genos(genos_ont_gtscore_samplesBlood_lociDual, 30, 0.9)
```

```{r, echo=FALSE}
# For html
optGenos90_ill_gtscore_blood %>%
  kbl(caption = "Illumina-GTscore optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos90_ont_gtscore_blood %>%
  kbl(caption = "Nanopore-GTscore optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

##### gtseq

```{r, echo=FALSE}
optGenos90_ill_gtseq_blood <- optimized_genos(genos_ill_gtseq_samplesBlood_lociDual, 10, 0.9)

optGenos90_ont_gtseq_blood <- optimized_genos(genos_ont_gtseq_samplesBlood_lociDual, 10, 0.9)
```

```{r, echo=FALSE}
# For html
optGenos90_ill_gtseq_blood %>%
  kbl(caption = "Illumina-GTseq optimized sample/loci subset w/90% genotype success across blood samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos90_ont_gtseq_blood %>%
  kbl(caption = "Nanopore-GTseq optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Sample/loci lists

Including which samples have whole-genome sequences available

```{r, echo=FALSE}
sampleList_optGenos100_gtscore_samplesBlood <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtscore_blood)
) %>%
  merge(., md_tamRun4_blood[, c("sampleID_ill", "sampleID_ont", "species", "animalID")], by = "sampleID_ill") %>%
  mutate(
    wholeGenomeSeq = case_when(
      animalID %in% wholeGenomeSeqs$animalID ~ "yes",
      .default = "no"
    )
  )

sampleList_optGenos100_gtseq_samplesBlood <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtseq_blood)
) %>%
  merge(., md_tamRun4_blood[, c("sampleID_ill", "sampleID_ont", "species", "animalID")], by = "sampleID_ill") %>%
  mutate(
    wholeGenomeSeq = case_when(
      animalID %in% wholeGenomeSeqs$animalID ~ "yes",
      .default = "no"
    )
  )
```

```{r, echo=FALSE}
# For html
sampleList_optGenos100_gtscore_samplesBlood %>%
  kbl(caption = "GTscore optimized sample list for samples 100% genotyped for loci subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

sampleList_optGenos100_gtseq_samplesBlood %>%
  kbl(caption = "GTsceq optimized sample list for samples 100% genotyped for loci subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 4.3 Genotype consistency

### 100% geno success

#### gtscore

Subset: 20 blood samples + 122 loci w/100% genotype success (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos100_ill_gtscore_blood_l <- optGenos100_ill_gtscore_blood %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtscore_samplesBlood_lociDual_l <- genos_ont_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare100_illONT_optGenos_gtscore_blood <- merge(optGenos100_ill_gtscore_blood_l, genos_ont_gtscore_samplesBlood_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

#### gtseq

Subset: 10 samples + 56 loci w/100% genotypes (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos100_ill_gtseq_blood_l <- optGenos100_ill_gtseq_blood %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtseq_samplesBlood_lociDual_l <- genos_ont_gtseq_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare100_illONT_optGenos_gtseq_blood <- merge(optGenos100_ill_gtseq_blood_l, genos_ont_gtseq_samplesBlood_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

### 90% geno success

#### gtscore

Subset: 30 samples + 103 loci w/genotypes for at least 90% of samples (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos90_ill_gtscore_blood_l <- optGenos90_ill_gtscore_blood %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtscore_samplesBlood_lociDual_l <- genos_ont_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare90_illONT_optGenos_gtscore_blood <- merge(optGenos90_ill_gtscore_blood_l, genos_ont_gtscore_samplesBlood_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ill_geno) ~ "NA",
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

#### gtseq

Subset: 10 samples + 94 loci w/genotypes for at least 90% of samples (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos90_ill_gtseq_blood_l <- optGenos90_ill_gtseq_blood %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtseq_samplesBlood_lociDual_l <- genos_ont_gtseq_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare90_illONT_optGenos_gtseq_blood <- merge(optGenos90_ill_gtseq_blood_l, genos_ont_gtseq_samplesBlood_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

### Summary

```{r, echo=FALSE}
genoCompare_illONT_blood_summary <- data.frame(
  match = c("illONT_NA", "illONT_mismatch", "illONT_match"),
  gtscore100 = table(genoCompare100_illONT_optGenos_gtscore_blood$match),
  gtscore90 = table(genoCompare90_illONT_optGenos_gtscore_blood$match),
  gtseq100 = table(genoCompare100_illONT_optGenos_gtseq_blood$match),
  gtseq90 = table(genoCompare90_illONT_optGenos_gtseq_blood$match)
) %>%
  select(-ends_with("Var1")) %>%
  column_to_rownames("match") %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    pipeline = c("gtscore100", "gtscore90", "gtseq100", "gtseq90"),
    samples = c("20", "30", "10", "10"),
    loci = c("122", "103", "56", "94")
  ) %>%
  relocate(c(samples, loci)) %>%
  mutate(
    propNA = round(illONT_NA/(illONT_NA + illONT_match + illONT_mismatch), 2),
    propMatch = round(illONT_match/(illONT_match + illONT_mismatch), 2),
    propMismatch = round(illONT_mismatch/(illONT_match + illONT_mismatch), 2)
  ) %>%
  rowwise() %>%
  mutate(
    genosTotal = illONT_NA + illONT_match + illONT_mismatch,
    genosCommon = illONT_match + illONT_mismatch
  ) %>%
  select(c(pipeline, samples, loci, genosTotal, illONT_NA, propNA, genosCommon, illONT_mismatch, propMismatch, illONT_match, propMatch))
```

```{r, echo=FALSE}
# For html
genoCompare_illONT_blood_summary %>%
  kbl(caption = "Summary of blood sample genotype consistency for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 5 Fecal samples

## 5.1 Geno success

### By sample

#### gtscore.10x

```{r, echo=FALSE}
# Geno success by sample
genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual <- genos_ill_gtscore_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina"
  )

genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual <- genos_ont_gtscore_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_bySample_illONT_gtscore_samplesFecal_lociDual <- rbind(genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual, genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual_uniqueID <- genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill")

genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual_uniqueID <- genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont")

genoSuccess_bySample_illONT_gtscore_samplesFecal_lociDual_uniqueID <- merge(genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual_uniqueID, genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual_uniqueID, by = "sampleID_unique") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(sampleID_unique, sampleID_ill, sampleID_ont, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_illONT_gtscore_samplesFecal_lociDual_uniqueID %>%
  kbl(caption = "GTscore.10x fecal sample genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### gtseq

```{r, echo=FALSE}
# Illumina
genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual <- genos_ill_gtseq_samplesFecal_lociDual %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual <- genos_ont_gtseq_samplesFecal_lociDual %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_bySample_illONT_gtseq_samplesFecal_lociDual <- rbind(genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual, genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual_uniqueID <- genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill")

genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual_uniqueID <- genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont")

genoSuccess_bySample_illONT_gtseq_samplesFecal_lociDual_uniqueID <- merge(genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual_uniqueID, genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual_uniqueID, by = "sampleID_unique") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(sampleID_unique, sampleID_ill, sampleID_ont, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_bySample_illONT_gtseq_samplesFecal_lociDual_uniqueID %>%
  kbl(caption = "GTseq fecal sample genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Summary

```{r, echo=FALSE}
summary_genoSuccess_bySample_samplesFecal_lociDual <- data.frame(
  pipeline = c("gtscore", "gtscore", "gtseq", "gtseq"),
  seqPlatform = rep(c("Illumina", "Nanopore"), 2),
  sampleType = rep("fecal", 4),
  min_genoSuccess = c(
    min(genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual$propSuccess),
    min(genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual$propSuccess),
    min(genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual$propSuccess),
    min(genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual$propSuccess)
  ),
  max_genoSuccess = c(
    max(genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual$propSuccess),
    max(genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual$propSuccess),
    max(genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual$propSuccess),
    max(genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual$propSuccess)
  ),
  mean_genoSuccess = c(
    mean(genoSuccess_bySample_ill_gtscore_samplesFecal_lociDual$propSuccess),
    mean(genoSuccess_bySample_ont_gtscore_samplesFecal_lociDual$propSuccess),
    mean(genoSuccess_bySample_ill_gtseq_samplesFecal_lociDual$propSuccess),
    mean(genoSuccess_bySample_ont_gtseq_samplesFecal_lociDual$propSuccess)
  )
) %>%
  mutate(
    mean_genoSuccess = round(mean_genoSuccess, 2),
    min_genoSuccess = round(min_genoSuccess, 2),
    max_genoSuccess = round(max_genoSuccess, 2)
    )
```

```{r, echo=FALSE}
# For html
summary_genoSuccess_bySample_samplesFecal_lociDual %>%
  kbl(caption = "Avg genotype success for fecal samples for Illumina vs. Nanopore in GTscore vs. GTseq pipelines") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Plots

```{r, echo=FALSE}
p5 <- ggplot(genoSuccess_bySample_illONT_gtscore_samplesFecal_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 140, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTscore (10x): Number of fecal samples per total loci genotyped") +
  theme_bw() +
  labs(x = "Number of fecal samples", y = "Total loci genotyped")

# GTseq
p6 <- ggplot(genoSuccess_bySample_illONT_gtseq_samplesFecal_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 140, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTseq: Number of fecal samples per total loci genotyped") +
  theme_bw() +
  labs(x = "Number of fecal samples", y = "Total loci genotyped")

p5
p6
```

### By locus

#### gtscore.10x

```{r, echo=FALSE}
# Illumina
genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual <- genos_ill_gtscore_samplesFecal_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual <- genos_ont_gtscore_samplesFecal_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_byLocus_illONT_gtscore_samplesFecal_lociDual <- rbind(genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual, genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual_uniqueID <- genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual_uniqueID <- genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_illONT_gtscore_samplesFecal_lociDual_uniqueID <- merge(genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual_uniqueID, genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual_uniqueID, by = "locus") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(locus, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_byLocus_illONT_gtscore_samplesFecal_lociDual_uniqueID %>%
  kbl(caption = "GTscore.10x loci genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### gtseq

```{r, echo=FALSE}
# Illumina
genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual <- genos_ill_gtseq_samplesFecal_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual <- genos_ont_gtseq_samplesFecal_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_byLocus_illONT_gtseq_samplesFecal_lociDual <- rbind(genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual, genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual_uniqueID <- genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual_uniqueID <- genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_illONT_gtseq_samplesFecal_lociDual_uniqueID <- merge(genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual_uniqueID, genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual_uniqueID, by = "locus") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(locus, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_byLocus_illONT_gtseq_samplesFecal_lociDual_uniqueID %>%
  kbl(caption = "GTseq loci genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Summary

```{r, echo=FALSE}
summary_genoSuccess_byLocus_samplesFecal_lociDual <- data.frame(
  pipeline = c("gtscore", "gtscore", "gtseq", "gtseq"),
  seqPlatform = rep(c("Illumina", "Nanopore"), 2),
  sampleType = rep("fecal", 4),
  min_genoSuccess = c(
    min(genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual$propSuccess),
    min(genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual$propSuccess),
    min(genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual$propSuccess),
    min(genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual$propSuccess)
  ),
  max_genoSuccess = c(
    max(genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual$propSuccess),
    max(genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual$propSuccess),
    max(genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual$propSuccess),
    max(genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual$propSuccess)
  ),
  mean_genoSuccess = c(
    mean(genoSuccess_byLocus_ill_gtscore_samplesFecal_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ont_gtscore_samplesFecal_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ill_gtseq_samplesFecal_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ont_gtseq_samplesFecal_lociDual$propSuccess)
  )
) %>%
  mutate(
    mean_genoSuccess = round(mean_genoSuccess, 2),
    min_genoSuccess = round(min_genoSuccess, 2),
    max_genoSuccess = round(max_genoSuccess, 2)
    )
```

```{r, echo=FALSE}
# For html
summary_genoSuccess_byLocus_samplesFecal_lociDual %>%
  kbl(caption = "Overview of genotype success by locus for Illumina vs. Nanopore fecal samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Plots

```{r, echo=FALSE}
p7 <- ggplot(genoSuccess_byLocus_illONT_gtscore_samplesFecal_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 30, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTscore (10x): Number of loci per total fecal samples genotyped") +
  theme_bw() +
  labs(x = "Number of loci", y = "Total fecal samples genotyped")

# GTseq
p8 <- ggplot(genoSuccess_byLocus_illONT_gtseq_samplesFecal_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 30, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTseq: Number of loci per total fecal samples genotyped") +
  theme_bw() +
  labs(x = "Number of loci", y = "Total fecal samples genotyped")

p7
p8
```

## 5.2 Identify optimal sample/loci combos

### Identify optimal sample/loci combos

```{r, echo=FALSE}
# gtscore
## loci genotyped at 100% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_ill_gtscore_fecal <- optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.gtscore" = "loci")

opt100_samplesLoci_ont_gtscore_fecal <- optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.gtscore" = "loci")

## loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_ill_gtscore_fecal <- optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.gtscore" = "loci")

opt90_samplesLoci_ont_gtscore_fecal <- optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.gtscore" = "loci")

# gtseq
## loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_ill_gtseq_fecal <- optimize_samplesLoci(genos_ill_gtseq_samplesFecal_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.gtseq" = "loci")

opt100_samplesLoci_ont_gtseq_fecal <- optimize_samplesLoci(genos_ont_gtseq_samplesFecal_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.gtseq" = "loci")

## loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_ill_gtseq_fecal <- optimize_samplesLoci(genos_ill_gtseq_samplesFecal_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.gtseq" = "loci")

opt90_samplesLoci_ont_gtseq_fecal <- optimize_samplesLoci(genos_ont_gtseq_samplesFecal_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.gtseq" = "loci")
```

#### Summaries

Based on the optimization results, I would choose the following sample/loci subsets for fecal samples based on Illumina results:

GTscore.10x - 100% of sample subset genotyped: 20 samples + 90 loci GTscore.10x - 90% of sample subset genotyped: 20 samples + 109 loci GTseq - 100% of sample subset genotyped: 10 samples + 86 loci GTseq - 90% of sample subset genotyped: 10 samples + 115 loci

```{r, echo=FALSE}
# loci genotyped at 100% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_fecal_summary <- merge(opt100_samplesLoci_ill_gtscore_fecal, opt100_samplesLoci_ont_gtscore_fecal, by = "samples") %>%
  merge(., opt100_samplesLoci_ill_gtseq_fecal, by = "samples") %>%
  merge(., opt100_samplesLoci_ont_gtseq_fecal, by = "samples")

# loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_fecal_summary <- merge(opt90_samplesLoci_ill_gtscore_fecal, opt90_samplesLoci_ont_gtscore_fecal, by = "samples") %>%
  merge(., opt90_samplesLoci_ill_gtseq_fecal, by = "samples") %>%
  merge(., opt90_samplesLoci_ont_gtseq_fecal, by = "samples")
```

```{r, echo=FALSE}
# For html
opt100_samplesLoci_fecal_summary %>%
  kbl(caption = "Sample/loci subsets for which loci are genotyped at 100% of samples in subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

opt90_samplesLoci_fecal_summary %>%
  kbl(caption = "Sample/loci subsets for which loci are genotyped at 90% of samples in subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Dataframes

#### 100% geno success

##### gtscore

```{r, echo=FALSE}
optGenos100_ill_gtscore_fecal <- optimized_genos(genos_ill_gtscore_samplesFecal_lociDual, 20, 1.0)

optGenos100_ont_gtscore_fecal <- optimized_genos(genos_ont_gtscore_samplesFecal_lociDual, 20, 1.0)
```

```{r, echo=FALSE}
# For html
optGenos100_ill_gtscore_fecal %>%
  kbl(caption = "Illumina-GTscore optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos100_ont_gtscore_fecal %>%
  kbl(caption = "Nanopore-GTscore optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

##### gtseq

```{r, echo=FALSE}
optGenos100_ill_gtseq_fecal <- optimized_genos(genos_ill_gtseq_samplesFecal_lociDual, 10, 1.0)

optGenos100_ont_gtseq_fecal <- optimized_genos(genos_ont_gtseq_samplesFecal_lociDual, 10, 1.0)
```

```{r, echo=FALSE}
# For html
optGenos100_ill_gtseq_fecal %>%
  kbl(caption = "Illumina-GTseq optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos100_ont_gtseq_fecal %>%
  kbl(caption = "Nanopore-GTseq optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### 90% geno success

##### gtscore

```{r, echo=FALSE}
optGenos90_ill_gtscore_fecal <- optimized_genos(genos_ill_gtscore_samplesFecal_lociDual, 20, 0.9)

optGenos90_ont_gtscore_fecal <- optimized_genos(genos_ont_gtscore_samplesFecal_lociDual, 20, 0.9)
```

```{r, echo=FALSE}
# For html
optGenos90_ill_gtscore_fecal %>%
  kbl(caption = "Illumina-GTscore optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos90_ont_gtscore_fecal %>%
  kbl(caption = "Nanopore-GTscore optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

##### gtseq

```{r, echo=FALSE}
optGenos90_ill_gtseq_fecal <- optimized_genos(genos_ill_gtseq_samplesFecal_lociDual, 10, 0.9)

optGenos90_ont_gtseq_fecal <- optimized_genos(genos_ont_gtseq_samplesFecal_lociDual, 10, 0.9)
```

```{r, echo=FALSE}
# For html
optGenos90_ill_gtseq_fecal %>%
  kbl(caption = "Illumina-GTseq optimized sample/loci subset w/90% genotype success across fecal samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos90_ont_gtseq_fecal %>%
  kbl(caption = "Nanopore-GTseq optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Sample/loci lists

(Including which samples have whole-genome sequences available)

```{r, echo=FALSE}
sampleList_optGenos100_gtscore_samplesFecal <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtscore_fecal)
) %>%
  merge(., md_tamRun4_fecal[, c("sampleID_ill", "sampleID_ont", "species", "animalID")], by = "sampleID_ill") %>%
  mutate(
    wholeGenomeSeq = case_when(
      animalID %in% wholeGenomeSeqs$animalID ~ "yes",
      .default = "no"
    )
  )

sampleList_optGenos100_gtseq_samplesFecal <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtseq_fecal)
) %>%
  merge(., md_tamRun4_fecal[, c("sampleID_ill", "sampleID_ont", "species", "animalID")], by = "sampleID_ill") %>%
  mutate(
    wholeGenomeSeq = case_when(
      animalID %in% wholeGenomeSeqs$animalID ~ "yes",
      .default = "no"
    )
  )
```

```{r, echo=FALSE}
# For html
sampleList_optGenos100_gtscore_samplesFecal %>%
  kbl(caption = "GTscore optimized sample list for samples 100% genotyped for loci subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

sampleList_optGenos100_gtseq_samplesFecal %>%
  kbl(caption = "GTsceq optimized sample list for samples 100% genotyped for loci subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 5.3 Genotype consistency

### 100% geno success

#### gtscore

Subset: 20 fecal samples + 90 loci w/100% genotype success (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos100_ill_gtscore_fecal_l <- optGenos100_ill_gtscore_fecal %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtscore_samplesFecal_lociDual_l <- genos_ont_gtscore_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare100_illONT_optGenos_gtscore_fecal <- merge(optGenos100_ill_gtscore_fecal_l, genos_ont_gtscore_samplesFecal_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

#### gtseq

Subset: 10 samples + 86 loci w/100% genotypes (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos100_ill_gtseq_fecal_l <- optGenos100_ill_gtseq_fecal %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtseq_samplesFecal_lociDual_l <- genos_ont_gtseq_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare100_illONT_optGenos_gtseq_fecal <- merge(optGenos100_ill_gtseq_fecal_l, genos_ont_gtseq_samplesFecal_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

### 90% geno success

#### gtscore

Subset: 20 samples + 109 loci w/genotypes for at least 90% of samples (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos90_ill_gtscore_fecal_l <- optGenos90_ill_gtscore_fecal %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtscore_samplesFecal_lociDual_l <- genos_ont_gtscore_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare90_illONT_optGenos_gtscore_fecal <- merge(optGenos90_ill_gtscore_fecal_l, genos_ont_gtscore_samplesFecal_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ill_geno) ~ "NA",
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

#### gtseq

Subset: 10 samples + 115 loci w/genotypes for at least 90% of samples (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos90_ill_gtseq_fecal_l <- optGenos90_ill_gtseq_fecal %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtseq_samplesFecal_lociDual_l <- genos_ont_gtseq_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare90_illONT_optGenos_gtseq_fecal <- merge(optGenos90_ill_gtseq_fecal_l, genos_ont_gtseq_samplesFecal_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

### Summary

```{r, echo=FALSE}
genoCompare_illONT_fecal_summary <- data.frame(
  match = c("illONT_NA", "illONT_mismatch", "illONT_match"),
  gtscore100 = table(genoCompare100_illONT_optGenos_gtscore_fecal$match),
  gtscore90 = table(genoCompare90_illONT_optGenos_gtscore_fecal$match),
  gtseq100 = table(genoCompare100_illONT_optGenos_gtseq_fecal$match),
  gtseq90 = table(genoCompare90_illONT_optGenos_gtseq_fecal$match)
) %>%
  select(-ends_with("Var1")) %>%
  column_to_rownames("match") %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    pipeline = c("gtscore100", "gtscore90", "gtseq100", "gtseq90"),
    samples = c("20", "20", "10", "10"),
    loci = c("122", "131", "56", "94")
  ) %>%
  relocate(c(samples, loci)) %>%
  mutate(
    propNA = round(illONT_NA/(illONT_NA + illONT_match + illONT_mismatch), 2),
    propMatch = round(illONT_match/(illONT_match + illONT_mismatch), 2),
    propMismatch = round(illONT_mismatch/(illONT_match + illONT_mismatch), 2)
  ) %>%
  rowwise() %>%
  mutate(
    genosTotal = illONT_NA + illONT_match + illONT_mismatch,
    genosCommon = illONT_match + illONT_mismatch
  ) %>%
  select(c(pipeline, samples, loci, genosTotal, illONT_NA, propNA, genosCommon, illONT_mismatch, propMismatch, illONT_match, propMatch))
```

```{r, echo=FALSE}
# For html
genoCompare_illONT_fecal_summary %>%
  kbl(caption = "Summary of fecal sample genotype consistency for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 6 Hair samples

## 6.1 Geno success

### By sample

#### gtscore.10x

```{r, echo=FALSE}
# Geno success by sample
genoSuccess_bySample_ill_gtscore_samplesHair_lociDual <- genos_ill_gtscore_samplesHair_lociDual %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina"
  )

genoSuccess_bySample_ont_gtscore_samplesHair_lociDual <- genos_ont_gtscore_samplesHair_lociDual %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_bySample_illONT_gtscore_samplesHair_lociDual <- rbind(genoSuccess_bySample_ill_gtscore_samplesHair_lociDual, genoSuccess_bySample_ont_gtscore_samplesHair_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_ill_gtscore_samplesHair_lociDual_uniqueID <- genoSuccess_bySample_ill_gtscore_samplesHair_lociDual %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill")

genoSuccess_bySample_ont_gtscore_samplesHair_lociDual_uniqueID <- genoSuccess_bySample_ont_gtscore_samplesHair_lociDual %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont")

genoSuccess_bySample_illONT_gtscore_samplesHair_lociDual_uniqueID <- merge(genoSuccess_bySample_ill_gtscore_samplesHair_lociDual_uniqueID, genoSuccess_bySample_ont_gtscore_samplesHair_lociDual_uniqueID, by = "sampleID_unique") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(sampleID_unique, sampleID_ill, sampleID_ont, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_bySample_illONT_gtscore_samplesHair_lociDual_uniqueID %>%
  kbl(caption = "GTscore.10x hair sample genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### gtseq

```{r, echo=FALSE}
# Illumina
genoSuccess_bySample_ill_gtseq_samplesHair_lociDual <- genos_ill_gtseq_samplesHair_lociDual %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_bySample_ont_gtseq_samplesHair_lociDual <- genos_ont_gtseq_samplesHair_lociDual %>%
  mutate(across(everything(), ~ as.character(.))) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_bySample_illONT_gtseq_samplesHair_lociDual <- rbind(genoSuccess_bySample_ill_gtseq_samplesHair_lociDual, genoSuccess_bySample_ont_gtseq_samplesHair_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_bySample_ill_gtseq_samplesHair_lociDual_uniqueID <- genoSuccess_bySample_ill_gtseq_samplesHair_lociDual %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_ill", "sampleID_unique")], by = "sampleID_ill")

genoSuccess_bySample_ont_gtseq_samplesHair_lociDual_uniqueID <- genoSuccess_bySample_ont_gtseq_samplesHair_lociDual %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_unique")], by = "sampleID_ont")

genoSuccess_bySample_illONT_gtseq_samplesHair_lociDual_uniqueID <- merge(genoSuccess_bySample_ill_gtseq_samplesHair_lociDual_uniqueID, genoSuccess_bySample_ont_gtseq_samplesHair_lociDual_uniqueID, by = "sampleID_unique") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(sampleID_unique, sampleID_ill, sampleID_ont, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_bySample_illONT_gtseq_samplesHair_lociDual_uniqueID %>%
  kbl(caption = "GTseq hair sample genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Summary

```{r, echo=FALSE}
summary_genoSuccess_bySample_samplesHair_lociDual <- data.frame(
  pipeline = c("gtscore", "gtscore", "gtseq", "gtseq"),
  seqPlatform = rep(c("Illumina", "Nanopore"), 2),
  sampleType = rep("hair", 4),
  min_genoSuccess = c(
    min(genoSuccess_bySample_ill_gtscore_samplesHair_lociDual$propSuccess),
    min(genoSuccess_bySample_ont_gtscore_samplesHair_lociDual$propSuccess),
    min(genoSuccess_bySample_ill_gtseq_samplesHair_lociDual$propSuccess),
    min(genoSuccess_bySample_ont_gtseq_samplesHair_lociDual$propSuccess)
  ),
  max_genoSuccess = c(
    max(genoSuccess_bySample_ill_gtscore_samplesHair_lociDual$propSuccess),
    max(genoSuccess_bySample_ont_gtscore_samplesHair_lociDual$propSuccess),
    max(genoSuccess_bySample_ill_gtseq_samplesHair_lociDual$propSuccess),
    max(genoSuccess_bySample_ont_gtseq_samplesHair_lociDual$propSuccess)
  ),
  mean_genoSuccess = c(
    mean(genoSuccess_bySample_ill_gtscore_samplesHair_lociDual$propSuccess),
    mean(genoSuccess_bySample_ont_gtscore_samplesHair_lociDual$propSuccess),
    mean(genoSuccess_bySample_ill_gtseq_samplesHair_lociDual$propSuccess),
    mean(genoSuccess_bySample_ont_gtseq_samplesHair_lociDual$propSuccess)
  )
) %>%
  mutate(
    mean_genoSuccess = round(mean_genoSuccess, 2),
    min_genoSuccess = round(min_genoSuccess, 2),
    max_genoSuccess = round(max_genoSuccess, 2)
    )
```

```{r, echo=FALSE}
# For html
summary_genoSuccess_bySample_samplesHair_lociDual %>%
  kbl(caption = "Avg genotype success for hair samples for Illumina vs. Nanopore in GTscore vs. GTseq pipelines") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Plots

```{r, echo=FALSE}
p9 <- ggplot(genoSuccess_bySample_illONT_gtscore_samplesHair_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 140, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTscore (10x): Number of hair samples per total loci genotyped") +
  theme_bw() +
  labs(x = "Number of hair samples", y = "Total loci genotyped")

# GTseq
p10 <- ggplot(genoSuccess_bySample_illONT_gtseq_samplesHair_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 140, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTseq: Number of hair samples per total loci genotyped") +
  theme_bw() +
  labs(x = "Number of hair samples", y = "Total loci genotyped")

p9
p10
```

### By locus

#### gtscore.10x

```{r, echo=FALSE}
# Illumina
genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual <- genos_ill_gtscore_samplesHair_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual <- genos_ont_gtscore_samplesHair_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_byLocus_illONT_gtscore_samplesHair_lociDual <- rbind(genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual, genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual_uniqueID <- genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual_uniqueID <- genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_illONT_gtscore_samplesHair_lociDual_uniqueID <- merge(genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual_uniqueID, genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual_uniqueID, by = "locus") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(locus, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_byLocus_illONT_gtscore_samplesHair_lociDual_uniqueID %>%
  kbl(caption = "GTscore.10x loci genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### gtseq

```{r, echo=FALSE}
# Illumina
genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual <- genos_ill_gtseq_samplesHair_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Illumina"
  )

# Nanopore
genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual <- genos_ont_gtseq_samplesHair_lociDual %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/30, 2),
    seqPlatform = "Nanopore"
  )

# Combined
genoSuccess_byLocus_illONT_gtseq_samplesHair_lociDual <- rbind(genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual, genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual)
```

```{r, echo=FALSE}
# For html
genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual_uniqueID <- genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual_uniqueID <- genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual %>%
  rownames_to_column("locus")

genoSuccess_byLocus_illONT_gtseq_samplesHair_lociDual_uniqueID <- merge(genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual_uniqueID, genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual_uniqueID, by = "locus") %>%
  dplyr::rename("totalGenos_ill" = "totalGenos.x",
                "totalGenos_ont" = "totalGenos.y",
                "propSuccess_ill" = "propSuccess.x",
                "propSuccess_ont" = "propSuccess.y") %>%
  select(c(locus, totalGenos_ill, totalGenos_ont, propSuccess_ill, propSuccess_ont))

genoSuccess_byLocus_illONT_gtseq_samplesHair_lociDual_uniqueID %>%
  kbl(caption = "GTseq loci genotype success for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Summary

```{r, echo=FALSE}
summary_genoSuccess_byLocus_samplesHair_lociDual <- data.frame(
  pipeline = c("gtscore", "gtscore", "gtseq", "gtseq"),
  seqPlatform = rep(c("Illumina", "Nanopore"), 2),
  sampleType = rep("hair", 4),
  min_genoSuccess = c(
    min(genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual$propSuccess),
    min(genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual$propSuccess),
    min(genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual$propSuccess),
    min(genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual$propSuccess)
  ),
  max_genoSuccess = c(
    max(genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual$propSuccess),
    max(genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual$propSuccess),
    max(genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual$propSuccess),
    max(genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual$propSuccess)
  ),
  mean_genoSuccess = c(
    mean(genoSuccess_byLocus_ill_gtscore_samplesHair_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ont_gtscore_samplesHair_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ill_gtseq_samplesHair_lociDual$propSuccess),
    mean(genoSuccess_byLocus_ont_gtseq_samplesHair_lociDual$propSuccess)
  )
) %>%
  mutate(
    mean_genoSuccess = round(mean_genoSuccess, 2),
    min_genoSuccess = round(min_genoSuccess, 2),
    max_genoSuccess = round(max_genoSuccess, 2)
    )
```

```{r, echo=FALSE}
# For html
summary_genoSuccess_byLocus_samplesHair_lociDual %>%
  kbl(caption = "Overview of genotype success by locus for Illumina vs. Nanopore hair samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### Plots

```{r, echo=FALSE}
p11 <- ggplot(genoSuccess_byLocus_illONT_gtscore_samplesHair_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 30, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTscore (10x): Number of loci per total hair samples genotyped") +
  theme_bw() +
  labs(x = "Number of loci", y = "Total hair samples genotyped")

# GTseq
p12 <- ggplot(genoSuccess_byLocus_illONT_gtseq_samplesHair_lociDual, aes(y = totalGenos)) +
  geom_bar() +
  scale_y_continuous(breaks = seq(0, 30, 5)) +
  facet_wrap(~seqPlatform) + 
  ggtitle(label = "GTseq: Number of loci per total hair samples genotyped") +
  theme_bw() +
  labs(x = "Number of loci", y = "Total hair samples genotyped")

p11
p12
```

## 6.2 Identify optimal sample/loci combos

### Identify optimal sample/loci combos

```{r, echo=FALSE}
# gtscore
## loci genotyped at 100% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_ill_gtscore_hair <- optimize_samplesLoci(genos_ill_gtscore_samplesHair_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.gtscore" = "loci")

opt100_samplesLoci_ont_gtscore_hair <- optimize_samplesLoci(genos_ont_gtscore_samplesHair_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.gtscore" = "loci")

## loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_ill_gtscore_hair <- optimize_samplesLoci(genos_ill_gtscore_samplesHair_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.gtscore" = "loci")

opt90_samplesLoci_ont_gtscore_hair <- optimize_samplesLoci(genos_ont_gtscore_samplesHair_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.gtscore" = "loci")

# gtseq
## loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_ill_gtseq_hair <- optimize_samplesLoci(genos_ill_gtseq_samplesHair_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.gtseq" = "loci")

opt100_samplesLoci_ont_gtseq_hair <- optimize_samplesLoci(genos_ont_gtseq_samplesHair_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.gtseq" = "loci")

## loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_ill_gtseq_hair <- optimize_samplesLoci(genos_ill_gtseq_samplesHair_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.gtseq" = "loci")

opt90_samplesLoci_ont_gtseq_hair <- optimize_samplesLoci(genos_ont_gtseq_samplesHair_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.gtseq" = "loci")
```

#### Summaries

Based on the optimization results, I would choose the following sample/loci subsets for hair samples based on Illumina results:

GTscore.10x - 100% of sample subset genotyped: 10 samples + 83 loci GTscore.10x - 90% of sample subset genotyped: 10 samples + 99 loci GTseq - 100% of sample subset genotyped: 5 samples + 84 loci GTseq - 90% of sample subset genotyped: 5 samples + 84 loci

```{r, echo=FALSE}
# loci genotyped at 100% of samples among subsets of 10 to 30 samples
opt100_samplesLoci_hair_summary <- merge(opt100_samplesLoci_ill_gtscore_hair, opt100_samplesLoci_ont_gtscore_hair, by = "samples") %>%
  merge(., opt100_samplesLoci_ill_gtseq_hair, by = "samples") %>%
  merge(., opt100_samplesLoci_ont_gtseq_hair, by = "samples")

# loci genotyped at 90% of samples among subsets of 10 to 30 samples
opt90_samplesLoci_hair_summary <- merge(opt90_samplesLoci_ill_gtscore_hair, opt90_samplesLoci_ont_gtscore_hair, by = "samples") %>%
  merge(., opt90_samplesLoci_ill_gtseq_hair, by = "samples") %>%
  merge(., opt90_samplesLoci_ont_gtseq_hair, by = "samples")
```

```{r, echo=FALSE}
# For html
opt100_samplesLoci_hair_summary %>%
  kbl(caption = "Sample/loci subsets for which loci are genotyped at 100% of samples in subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

opt90_samplesLoci_hair_summary %>%
  kbl(caption = "Sample/loci subsets for which loci are genotyped at 90% of samples in subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Dataframes

#### 100% geno success

##### gtscore

```{r, echo=FALSE}
optGenos100_ill_gtscore_hair <- optimized_genos(genos_ill_gtscore_samplesHair_lociDual, 10, 1.0)

optGenos100_ont_gtscore_hair <- optimized_genos(genos_ont_gtscore_samplesHair_lociDual, 10, 1.0)
```

```{r, echo=FALSE}
# For html
optGenos100_ill_gtscore_hair %>%
  kbl(caption = "Illumina-GTscore optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos100_ont_gtscore_hair %>%
  kbl(caption = "Nanopore-GTscore optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

##### gtseq

```{r, echo=FALSE}
optGenos100_ill_gtseq_hair <- optimized_genos(genos_ill_gtseq_samplesHair_lociDual, 5, 1.0)

optGenos100_ont_gtseq_hair <- optimized_genos(genos_ont_gtseq_samplesHair_lociDual, 5, 1.0)
```

```{r, echo=FALSE}
# For html
optGenos100_ill_gtseq_hair %>%
  kbl(caption = "Illumina-GTseq optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos100_ont_gtseq_hair %>%
  kbl(caption = "Nanopore-GTseq optimized sample/loci subset w/100% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

#### 90% geno success

##### gtscore

```{r, echo=FALSE}
optGenos90_ill_gtscore_hair <- optimized_genos(genos_ill_gtscore_samplesHair_lociDual, 10, 0.9)

optGenos90_ont_gtscore_hair <- optimized_genos(genos_ont_gtscore_samplesHair_lociDual, 10, 0.9)
```

```{r, echo=FALSE}
# For html
optGenos90_ill_gtscore_hair %>%
  kbl(caption = "Illumina-GTscore optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos90_ont_gtscore_hair %>%
  kbl(caption = "Nanopore-GTscore optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

##### gtseq

```{r, echo=FALSE}
optGenos90_ill_gtseq_hair <- optimized_genos(genos_ill_gtseq_samplesHair_lociDual, 5, 0.9)

optGenos90_ont_gtseq_hair <- optimized_genos(genos_ont_gtseq_samplesHair_lociDual, 5, 0.9)
```

```{r, echo=FALSE}
# For html
optGenos90_ill_gtseq_hair %>%
  kbl(caption = "Illumina-GTseq optimized sample/loci subset w/90% genotype success across hair samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

optGenos90_ont_gtseq_hair %>%
  kbl(caption = "Nanopore-GTseq optimized sample/loci subset w/90% genotype success across samples") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Sample/loci lists

(Including which samples have whole-genome sequences available)

```{r, echo=FALSE}
sampleList_optGenos100_gtscore_samplesHair <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtscore_hair)
) %>%
  merge(., md_tamRun4_hair[, c("sampleID_ill", "sampleID_ont", "species", "animalID")], by = "sampleID_ill") %>%
  mutate(
    wholeGenomeSeq = case_when(
      animalID %in% wholeGenomeSeqs$animalID ~ "yes",
      .default = "no"
    )
  )

sampleList_optGenos100_gtseq_samplesHair <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtseq_hair)
) %>%
  merge(., md_tamRun4_hair[, c("sampleID_ill", "sampleID_ont", "species", "animalID")], by = "sampleID_ill") %>%
  mutate(
    wholeGenomeSeq = case_when(
      animalID %in% wholeGenomeSeqs$animalID ~ "yes",
      .default = "no"
    )
  )
```

```{r, echo=FALSE}
# For html
sampleList_optGenos100_gtscore_samplesHair %>%
  kbl(caption = "GTscore optimized sample list for samples 100% genotyped for loci subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))

sampleList_optGenos100_gtseq_samplesHair %>%
  kbl(caption = "GTsceq optimized sample list for samples 100% genotyped for loci subset") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 6.3 Genotype consistency

### 100% geno success

#### gtscore

Subset: 10 hair samples + 83 loci w/100% genotype success (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos100_ill_gtscore_hair_l <- optGenos100_ill_gtscore_hair %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtscore_samplesHair_lociDual_l <- genos_ont_gtscore_samplesHair_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare100_illONT_optGenos_gtscore_hair <- merge(optGenos100_ill_gtscore_hair_l, genos_ont_gtscore_samplesHair_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

#### gtseq

Subset: 5 samples + 84 loci w/100% genotypes (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos100_ill_gtseq_hair_l <- optGenos100_ill_gtseq_hair %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtseq_samplesHair_lociDual_l <- genos_ont_gtseq_samplesHair_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare100_illONT_optGenos_gtseq_hair <- merge(optGenos100_ill_gtseq_hair_l, genos_ont_gtseq_samplesHair_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

### 90% geno success

#### gtscore

Subset: 10 samples + 99 loci w/genotypes for at least 90% of samples (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos90_ill_gtscore_hair_l <- optGenos90_ill_gtscore_hair %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtscore_samplesHair_lociDual_l <- genos_ont_gtscore_samplesHair_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare90_illONT_optGenos_gtscore_hair <- merge(optGenos90_ill_gtscore_hair_l, genos_ont_gtscore_samplesHair_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ill_geno) ~ "NA",
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

#### gtseq

Subset: 5 samples + 84 loci w/genotypes for at least 90% of samples (in Illumina results)

```{r, echo=FALSE}
# Convert genotypes to long-format
optGenos90_ill_gtseq_hair_l <- optGenos90_ill_gtseq_hair %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ill_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

genos_ont_gtseq_samplesHair_lociDual_l <- genos_ont_gtseq_samplesHair_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "ont_geno") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID)

# compare ill vs. ont genos
genoCompare90_illONT_optGenos_gtseq_hair <- merge(optGenos90_ill_gtseq_hair_l, genos_ont_gtseq_samplesHair_lociDual_l, by = c("sampleID_unique", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_geno) ~ "NA",
      ill_geno == ont_geno ~ "yes",
      ill_geno != ont_geno ~ "no"
      )
    )
```

### Summary

```{r, echo=FALSE}
genoCompare_illONT_hair_summary <- data.frame(
  match = c("illONT_NA", "illONT_mismatch", "illONT_match"),
  gtscore100 = table(genoCompare100_illONT_optGenos_gtscore_hair$match),
  gtscore90 = table(genoCompare90_illONT_optGenos_gtscore_hair$match),
  gtseq100 = table(genoCompare100_illONT_optGenos_gtseq_hair$match),
  gtseq90 = table(genoCompare90_illONT_optGenos_gtseq_hair$match)
) %>%
  select(-ends_with("Var1")) %>%
  column_to_rownames("match") %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    pipeline = c("gtscore100", "gtscore90", "gtseq100", "gtseq90"),
    samples = c("20", "20", "10", "10"),
    loci = c("122", "131", "56", "94")
  ) %>%
  relocate(c(samples, loci)) %>%
  mutate(
    propNA = round(illONT_NA/(illONT_NA + illONT_match + illONT_mismatch), 2),
    propMatch = round(illONT_match/(illONT_match + illONT_mismatch), 2),
    propMismatch = round(illONT_mismatch/(illONT_match + illONT_mismatch), 2)
  ) %>%
  rowwise() %>%
  mutate(
    genosTotal = illONT_NA + illONT_match + illONT_mismatch,
    genosCommon = illONT_match + illONT_mismatch
  ) %>%
  select(c(pipeline, samples, loci, genosTotal, illONT_NA, propNA, genosCommon, illONT_mismatch, propMismatch, illONT_match, propMatch))
```

```{r, echo=FALSE}
# For html
genoCompare_illONT_hair_summary %>%
  kbl(caption = "Summary of hair sample genotype consistency for Illumina vs. Nanopore") %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 7 Q-score comparisons

## 7.1 Quality comparison overview

Looking at the metrics for all samples from tamRun4 on Illumina and Nanopore, we see that median quality scores per sample were higher for Illumina vs. Nanopore for both read 1 (median Illumina: 37, ONT: 18) and read 2 (median Illumina: 26, ONT: 18).

```{r, echo=FALSE}
#######################################
### LOAD FASTQC MULTIQC REPORT TO R ###
#######################################

# illumina
ill_MultiQCfastQCpath <- file.path("/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/04_tamRun4/00_illumina/01_run4QualityChecks/multiqc_data", "multiqc_data.json")

ill_MultiQCfastQC <- TidyMultiqc::load_multiqc(ill_MultiQCfastQCpath, sections = c("general", "raw"))

# nanopore
ont_MultiQCfastQCpath <- file.path("/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/04_tamRun4/01_nanopore/01_run4QualityChecks/multiqc_data", "multiqc_data.json")

ont_MultiQCfastQC <- TidyMultiqc::load_multiqc(ont_MultiQCfastQCpath, sections = c("general", "raw"))

####################################
### OBTAIN MEDIAN QUALITY SCORES ###
####################################
# illumina
ill_df <- TidyMultiqc::load_multiqc(
  ill_MultiQCfastQCpath, 
  sections = 'plot',
  plots = "fastqc_per_sequence_quality_scores_plot")

# nanopore
ont_df <- TidyMultiqc::load_multiqc(
  ont_MultiQCfastQCpath, 
  sections = 'plot',
  plots = "fastqc_per_sequence_quality_scores_plot")

###################
### UNNEST DATA ###
###################

# illumina
ill_df_unNest <- ill_df %>%
  dplyr::mutate(
    purrr::map_dfr(plot.fastqc_per_sequence_quality_scores_plot, function(plot_df){
      hist = HistDat::HistDat(vals=plot_df$x, counts = plot_df$y)
      list(
        mean_qc = mean(hist),
        median_qc = median(hist),
        min_qc = min(hist),
        max_qc = max(hist)
      )
    }),
    plot.fastqc_per_sequence_quality_scores_plot = NULL
  )

# nanopore
ont_df_unNest <- ont_df %>%
  dplyr::mutate(
    purrr::map_dfr(plot.fastqc_per_sequence_quality_scores_plot, function(plot_df){
      hist = HistDat::HistDat(vals=plot_df$x, counts = plot_df$y)
      list(
        mean_qc = mean(hist),
        median_qc = median(hist),
        min_qc = min(hist),
        max_qc = max(hist)
      )
    }),
    plot.fastqc_per_sequence_quality_scores_plot = NULL
  )

#############################################
### REARRANGE & GET MEDIAN QUALITY SCORES ###
#############################################

# illumina
ill_seqQC_r1 <- ill_df_unNest %>%
  select(c("metadata.sample_id", "median_qc")) %>%
  filter(grepl("R1", metadata.sample_id)) %>%
  dplyr::rename("medianQC_r1" = "median_qc") %>%
  mutate(sampleNo = substr(metadata.sample_id, start = 9, stop = 11))

ill_seqQC_r2 <- ill_df_unNest %>%
  select(c("metadata.sample_id", "median_qc")) %>%
  filter(grepl("R2", metadata.sample_id)) %>%
  dplyr::rename("medianQC_r2" = "median_qc") %>%
  mutate(sampleNo = substr(metadata.sample_id, start = 9, stop = 11))

ill_seqQC <- merge(ill_seqQC_r1, ill_seqQC_r2, by = "sampleNo") %>%
  select(c("sampleNo", "medianQC_r1", "medianQC_r2")) %>%
  arrange(sampleNo) %>%
  mutate(seqPlatform = "illumina")

# nanopore
ont_seqQC_r1 <- ont_df_unNest %>%
  select(c("metadata.sample_id", "median_qc")) %>%
  filter(grepl("r1", metadata.sample_id)) %>%
  dplyr::rename("medianQC_r1" = "median_qc") %>%
  mutate(sampleID_ont = substr(metadata.sample_id, 1, 9)) %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_ill")], by = "sampleID_ont") %>%
  mutate(sampleNo = substr(sampleID_ill, start = 9, stop = 11)) %>%
  arrange(as.numeric(sampleNo))

ont_seqQC_r2 <- ont_df_unNest %>%
  select(c("metadata.sample_id", "median_qc")) %>%
  filter(grepl("r2", metadata.sample_id)) %>%
  dplyr::rename("medianQC_r2" = "median_qc") %>%
  mutate(sampleID_ont = substr(metadata.sample_id, 1, 9)) %>%
  merge(., md_tamRun4[, c("sampleID_ont", "sampleID_ill")], by = "sampleID_ont") %>%
  mutate(sampleNo = substr(sampleID_ill, start = 9, stop = 11)) %>%
  arrange(as.numeric(sampleNo))

ont_seqQC <- merge(ont_seqQC_r1, ont_seqQC_r2, by = "sampleNo") %>%
  select(c("sampleNo", "medianQC_r1", "medianQC_r2")) %>%
  arrange(sampleNo) %>%
  mutate(seqPlatform = "nanopore")

# both
illONT_seqQC_r1 <- merge(ill_seqQC_r1[, c("sampleNo", "medianQC_r1")], ont_seqQC_r1[, c("sampleNo", "medianQC_r1")], by = "sampleNo") %>%
  dplyr::rename("ill_medianQC_r1" = "medianQC_r1.x",
                "ont_medianQC_r1" = "medianQC_r1.y")

illONT_seqQC_r2 <- merge(ill_seqQC_r2[, c("sampleNo", "medianQC_r2")], ont_seqQC_r2[, c("sampleNo", "medianQC_r2")], by = "sampleNo") %>%
  dplyr::rename("ill_medianQC_r2" = "medianQC_r2.x",
                "ont_medianQC_r2" = "medianQC_r2.y")

illONT_seqQC_r1r2 <- merge(illONT_seqQC_r1, illONT_seqQC_r2, by = "sampleNo")
```

```{r, collapse=TRUE}
# Illumina read 1
median(illONT_seqQC_r1r2$ill_medianQC_r1)

# ONT read 1
median(illONT_seqQC_r1r2$ont_medianQC_r1)

# Illumina read 2
median(illONT_seqQC_r1r2$ill_medianQC_r2)

# ONT read 2
median(illONT_seqQC_r1r2$ont_medianQC_r2)
```

```{r, echo=FALSE}
#################
### VISUALIZE ###
#################

# overview of all samples
illONT_seqQC_l <- rbind(ill_seqQC, ont_seqQC) %>%
  pivot_longer(cols = c("medianQC_r1", "medianQC_r2"),
               names_to = "readNo",
               values_to = "medianQC")

ggplot(illONT_seqQC_l, aes(medianQC)) +
  geom_bar(stat = "count") +
  facet_grid(rows = vars(seqPlatform), cols = vars(readNo)) +
  labs(y = "sample count") +
  theme_linedraw() +
  labs(title = "Sample count vs. median R1/R2 QC values for Illumina vs. Nanopore")

ggplot(illONT_seqQC_l, aes(x = readNo, y = medianQC, fill = seqPlatform)) +
  geom_boxplot() +
  labs(x = "") +
  scale_y_continuous(limits = c(0, 40), breaks = seq(0, 40, 5)) +
  theme_linedraw() +
  labs(title = "Distribution of median sample R1/R2 QC values for Illumina vs. Nanopore")

# sample by sample
## read1
illONT_seqQC_r1_l <- illONT_seqQC_l %>%
  filter(str_detect(readNo, "r1"))

ggplot(illONT_seqQC_r1_l) +
  geom_segment(data = ill_seqQC_r1,
               aes(x = medianQC_r1,
                   y = sampleNo,
                   xend = ont_seqQC_r1$medianQC_r1,
                   yend = ont_seqQC_r1$sampleNo)) +
  geom_point(aes(x = medianQC, y = sampleNo, color = seqPlatform)) +
  theme_linedraw() +
  labs(title = "Median sample R1 QC in Illumina vs. Nanopore runs") +
  scale_y_discrete(guide = guide_axis(n.dodge = 2))

## read2
illONT_seqQC_r2_l <- illONT_seqQC_l %>%
  filter(str_detect(readNo, "r2"))

ggplot(illONT_seqQC_r2_l) +
  geom_segment(data = ill_seqQC_r2,
               aes(x = medianQC_r2,
                   y = sampleNo,
                   xend = ont_seqQC_r2$medianQC_r2,
                   yend = ont_seqQC_r2$sampleNo)) +
  geom_point(aes(x = medianQC, y = sampleNo, color = seqPlatform)) +
  theme_linedraw() +
  labs(title = "Median sample R2 QC in Illumina vs. Nanopore runs") +
  scale_y_discrete(guide = guide_axis(n.dodge = 2))
```

## 7.2 Q-score filters vs. geno consistency

Below I'm checking how genotype consistency between Illumina and Nanopore compares when we filter out low-quality reads from the Nanopore sequencing results. I also included a filtering step for Illumina reads to retain only \>= q30 reads. Scripts for filtering and genotyping can be found in tamRun4_qualityFilterTests.Rmd.

I filtered sequencing results based on the average quality per read, resulting in the following filtered sequence sets:

-   ill_qc0 - no reads filtered
-   ill_qc30 - filtered out reads with \< q30 average quality
-   ont_q0 - no reads filtered out
-   ont_q11 - filtered out reads with \< q11 average quality
-   ont_q13 - filtered out reads with \< q13 average quality
-   ont_q15 - filtered out reads with \< q15 average quality
-   ont_q20 - filtered out reads with \< q20 average quality

Interestingly, we see a pattern where the proportion of matching genotypes between Illumina and Nanopore results **doesn't really change, and if anything, DECREASES** with increasing quality filters for Nanopore sequences - this pattern holds despite the loss of reads with each filtering step and both with unfiltered and q30-filtered Illumina results.

```{r, echo=FALSE}
#############################
### GET TOTAL READ COUNTS ###
#############################

repl <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# illumina
readCounts_ill_gtscore_fullSet_qc0 <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_AlleleReads_singleSNPs.txt") %>%
  t() %>%
  as.data.frame() %>%
  replace(., TRUE, lapply(., repl)) %>%
  mutate(across(everything(),as.numeric)) %>%
  mutate(
    totalReads = rowSums(.)
  ) %>%
  select(totalReads) %>%
  rownames_to_column("sampleID") %>%
  mutate(
    sampleID = str_c(sampleID, "_qc0")
  )

readCounts_ill_gtscore_fullSet_qc30 <- read.table("./04_tamRun4/04_run4QualityFilterTests/03_run4Analyses_qcFiltered/illONT_qcFiltered_fullSet_AlleleReads_singleSNPs.txt") %>%
  select(starts_with("tamRun4")) %>%
  t() %>%
  as.data.frame() %>%
  replace(., TRUE, lapply(., repl)) %>%
  mutate(across(everything(),as.numeric)) %>%
  mutate(
    totalReads = rowSums(.)
  ) %>%
  select(totalReads) %>%
  rownames_to_column("sampleID")

readCounts_ill_gtscore_fullSet_sampleSum <- rbind(readCounts_ill_gtscore_fullSet_qc0, readCounts_ill_gtscore_fullSet_qc30) %>%
  separate(., sampleID, sep = "_", into = c("sampleID", "qc")) %>%
  pivot_wider(names_from = qc, values_from = totalReads) %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  select(-sampleID) %>%
  relocate(sampleID_unique)

readCounts_ill_gtscore_blood_sampleSum <- readCounts_ill_gtscore_fullSet_sampleSum %>%
  filter(str_detect(sampleID_unique, "blood")) %>%
  dplyr::rename(
    "ill_qc0" = "qc0",
    "ill_qc30" = "qc30"
  )

# nanopore
readCounts_ont_gtscore_fullSet_qc0 <- read.table("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_AlleleReads_singleSNPs.txt") %>%
  t() %>%
  as.data.frame() %>%
  replace(., TRUE, lapply(., repl)) %>%
  mutate(across(everything(),as.numeric)) %>%
  mutate(
    totalReads = rowSums(.)
  ) %>%
  select(totalReads) %>%
  rownames_to_column("sampleID") %>%
  mutate(
    sampleID = str_c(sampleID, "_qc0")
  )
  
readCounts_ont_gtscore_fullSet_qcFiltered <- read.table("./04_tamRun4/04_run4QualityFilterTests/03_run4Analyses_qcFiltered/illONT_qcFiltered_fullSet_AlleleReads_singleSNPs.txt") %>%
  select(starts_with("barcode")) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = str_remove(sampleID, "_interleaved")) %>%
  column_to_rownames("sampleID") %>%
  replace(., TRUE, lapply(., repl)) %>%
  mutate(across(everything(),as.numeric)) %>%
  mutate(
    totalReads = rowSums(.)
  ) %>%
  select(totalReads) %>%
  rownames_to_column("sampleID")

readCounts_ont_gtscore_fullSet_sampleSum <- rbind(readCounts_ont_gtscore_fullSet_qc0, readCounts_ont_gtscore_fullSet_qcFiltered) %>%
  separate(., sampleID, sep = "_", into = c("sampleID", "qc")) %>%
  pivot_wider(names_from = qc, values_from = totalReads) %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  select(-sampleID) %>%
  relocate(sampleID_unique)

readCounts_ont_gtscore_blood_sampleSum <- readCounts_ont_gtscore_fullSet_sampleSum %>%
  filter(str_detect(sampleID_unique, "blood")) %>%
  dplyr::rename(
    "ont_qc0" = "qc0",
    "ont_qc11" = "qc11",
    "ont_qc13" = "qc13",
    "ont_qc15" = "qc15",
    "ont_qc20" = "qc20",
  )

##################################################
### SUMMARIZE PER SAMPLE & PER SUM READ COUNTS ###
##################################################

readCounts_illONT_gtscore_blood_sampleSum <- merge(readCounts_ill_gtscore_blood_sampleSum, readCounts_ont_gtscore_blood_sampleSum, by = "sampleID_unique")

readCounts_illONT_gtscore_blood_runSum <- readCounts_illONT_gtscore_blood_sampleSum %>%
  summarise(across(where(is.numeric),
                   list(totalReads = ~ sum(.x, na.rm = T)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("set") %>%
  mutate(set = gsub("_totalReads", "", set)) %>%
  dplyr::rename("totalReads" = "V1")
```

```{r, echo=FALSE}
###########################
### GENO CONSISTENCY DF ###
###########################

# full dataframe
genoCompare_qcFilters <- merge(
  genos_ill_gtscore_samplesAll_lociDual_l,
  genos_ill_qcFiltered_gtscore_samplesAll_lociDual_l,
  by = c("sampleID_unique", "locus"),
  all.x = T) %>%
  merge(.,
        genos_ont_gtscore_samplesAll_lociDual_l,
        by = c("sampleID_unique", "locus"),
        all.x = T) %>%
  merge(.,
        genos_ont_qcFiltered_gtscore_samplesAll_lociDual_l,
        by = c("sampleID_unique", "locus"),
        all.x = T)

# blood samples only
genoCompare_blood_qcFilters <- genoCompare_qcFilters %>%
  filter(str_detect(sampleID_unique, "blood"))

###############################
### COMPARE B/T ALL FILTERS ###
###############################

df <- genoCompare_blood_qcFilters

cols_of_interest <- setdiff(names(df), c("sampleID_unique", "locus"))

colCombos <- combn(cols_of_interest, 2)

check <- df[,colCombos[1,]] > 0 & df[,colCombos[2,]] > 0 & df[,colCombos[1,]] == df[,colCombos[2,]]
colnames(check) = paste0(colCombos[1,], ".", colCombos[2,])

genoCompare_blood_illONT_allCombos <- cbind(df, ifelse(check == T, 1, 0))

##################################
### SUMMARIZE GENO CONSISTENCY ###
##################################

genoConsistency_summary <- genoCompare_blood_illONT_allCombos %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("combo") %>%
  separate(., combo, into = c("combo", "metric"), sep = "(?<=[0-9])_") %>%
  pivot_wider(names_from = metric, values_from = V1) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.") %>%
  mutate(
    propCommonGenos = round(totalCommonGenos/(30*140), 2),
    propMatchGenos = totalMatchGenos/totalCommonGenos
  ) %>%
  merge(., readCounts_illONT_gtscore_blood_runSum, by.x = "set1", by.y = "set") %>%
  merge(., readCounts_illONT_gtscore_blood_runSum, by.x = "set2", by.y = "set") %>%
  dplyr::rename("set1_totalReads" = "totalReads.x",
         "set2_totalReads" = "totalReads.y") %>%
  arrange(set1) %>%
  mutate(
    diff_totalReads = set1_totalReads - set2_totalReads
  ) %>%
  select(c(set1, set2, propMatchGenos, propCommonGenos, totalMatchGenos, totalCommonGenos, set1_totalReads, set2_totalReads, diff_totalReads))
```

```{r, echo=FALSE}
# for html
genoConsistency_summary %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 8 Downsampling tests

## 8.1 Coverage overview

### Data

```{r}
# originals
alleleReads_ill_gtscore_samplesAll <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

alleleReads_ont_gtscore_samplesAll <- read.delim("./04_tamRun4/01_nanopore/03_run4GTscore/ont_fullSet_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame()

# reformat
alleleReads_ill_gtscore_samplesAll_l <- alleleReads_ill_gtscore_samplesAll %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
  separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
  select(-readCounts_a1, -readCounts_a2) %>%
  mutate(seqPlatform = "illumina")

alleleReads_ont_gtscore_samplesAll_l <- alleleReads_ont_gtscore_samplesAll %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
  separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
  select(-readCounts_a1, -readCounts_a2) %>%
  mutate(seqPlatform = "nanopore")
```

### Blood samples

#### Overview

The mean coverage per sample/loci combination is as follows: \* Illumina mean = 1144 reads \* Nanopore mean = 148 reads

```{r}
# illumina
mean(alleleReads_ill_gtscore_samplesBlood_lociDual_l$readCounts_sum) # 1144
median(alleleReads_ill_gtscore_samplesBlood_lociDual_l$readCounts_sum) # 762

# nanopore
mean(alleleReads_ont_gtscore_samplesBlood_lociDual_l$readCounts_sum) # 148
median(alleleReads_ont_gtscore_samplesBlood_lociDual_l$readCounts_sum) # 52
```

#### Plots

```{r}
# allele reads for blood samples + dual-species loci
alleleReads_ill_gtscore_samplesBlood_lociDual_l <- alleleReads_ill_gtscore_samplesAll_l %>%
  filter(str_detect(sampleID, "blood")) %>%
  filter(locus %in% primerList.v3_lociDual$locus)

alleleReads_ont_gtscore_samplesBlood_lociDual_l <- alleleReads_ont_gtscore_samplesAll_l %>%
  filter(str_detect(sampleID, "blood")) %>%
  filter(locus %in% primerList.v3_lociDual$locus)


# plot comparing coverage per locus across blood samples for illumina vs. nanopore
rbind(alleleReads_ill_gtscore_samplesBlood_lociDual_l,
      alleleReads_ont_gtscore_samplesBlood_lociDual_l) %>%
  ggplot(aes(x = sampleID, y = readCounts_sum, fill = seqPlatform)) +
  geom_boxplot() +
  geom_hline(yintercept = 1144, col = "black") +
  geom_hline(yintercept = 148, col = "black") +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_y_continuous(
    breaks = c(0, 148, 1144, 2500, 5000, 7500, 10000),
    labels = c(0, "mean(ont)=148", "mean(ill)=1144", 2500, 5000, 7500, 10000)
  ) +
  labs(
    title = "Dual-species loci coverage across blood samples for Illumina vs. Nanopore",
    y = "Reads per locus")
```

## 8.2 Downsampling

### Function

Quick-load function:

```{r}
# load in fxn + necessary packages
source("./scripts/downsample_alleleReads.R")
```

I created the function below to downsample sequencing results using the allele read counts file output from the GTscore pipeline. This means we can take advantage of the mapping and read counting functions already built into the GTscore pipeline and avoid more time-intensive downsampling methods that require going back into the fastq files.

This function works by using the allele reads to generate a list of length "n_iterations" in which each element is a randomized string of a's and b's for each sample at each locus. The number of each letter in the string corresponds to the read counts of each allele (e.g., read counts of 3,2 would be transformed to abbaa, or some other random combination of 3 a's and 2 b's).

After generating the list of randomized strings of a's and b's, the function selects the first n characters of each element in the list, where n is the desired downsampled coverage per sample per locus. The function then re-counts the allele reads based on the number of a's and b's selected in each element and averages them (rounded to the nearest integer) to provide the final downsampled read counts. Any sample/locus read counts that are below the desired coverage are recoded to "0,0"

Finally, the function returns the downsampled reads as a new allele reads file, which can then be input back into GTscore for genotyping.

```{r}
downsample_alleleReads <- function(alleleReads_file, n_reads, n_iterations) {
  ds_alleleReads <- alleleReads_file %>%
    rownames_to_column("locus") %>%
    pivot_longer(!locus,
                 names_to = "sampleID",
                 values_to = "readCounts") %>%
    relocate(sampleID) %>%
    arrange(sampleID) %>%
    separate(readCounts, into = c("readCounts_a1", "readCounts_a2"), sep = ",") %>%
    mutate(
      readCounts_a1 = as.numeric(readCounts_a1),
      readCounts_a2 = as.numeric(readCounts_a2),
      readCounts_sum = readCounts_a1 + readCounts_a2
    ) %>%
    # create character strings based on allele counts, separate for each allele
    rowwise() %>%
    mutate(
      vec_a1 = paste(replicate(readCounts_a1, "a"), collapse = ""),
      vec_a2 = paste(replicate(readCounts_a2, "b"), collapse = "")
    ) %>%
    # combine allele strings & create a list (n_iterations long) of randomized character strings
    mutate(
      vec_a1a2 = list(replicate(n_iterations, stri_rand_shuffle(str_c(vec_a1, vec_a2))))
    ) %>%
    # downsample each randomized character string
    mutate(
      ds_vec_a1a2 = list(substr(vec_a1a2, 1, n_reads))
    ) %>%
    # get avg read counts for each allele, rounding to nearest integer
    mutate(
      dsCounts_a1 = as.integer(round(mean(unlist(list(str_count(ds_vec_a1a2, "a")))))),
      dsCounts_a2 = as.integer(round(mean(unlist(list(str_count(ds_vec_a1a2, "b"))))))
    ) %>%
    # sum ds read counts, recode as NA if < n_reads, combine the rest in "x,x" format
    mutate(
      ds_readCounts_sum = dsCounts_a1 + dsCounts_a2,
      ds_readCounts = case_when(
        ds_readCounts_sum < n_reads ~ "0,0",
        .default = str_c(dsCounts_a1, dsCounts_a2, sep = ",")
      )
    ) %>%
    # create downsampled allele reads file
    select(c(sampleID, locus, ds_readCounts)) %>%
    pivot_wider(id_cols = locus,
                names_from = sampleID,
                values_from = ds_readCounts) %>%
    column_to_rownames("locus")
  
  return(ds_alleleReads)
}
```

### Illumina

#### Downsample

```{r}
# downsample allele reads
ds6x_alleleReads_ill_10 <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 6, 100)

ds10x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 10, 100)

ds15x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 15, 100)

ds20x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 20, 100)

ds25x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 25, 100)

ds30x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 30, 100)

ds50x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 50, 100)

ds100x_alleleReads_ill <- downsample_alleleReads(alleleReads_ill_gtscore_samplesAll, 100, 100)

# export downsampled allele reads
write.table(ds6x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds10x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds15x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds20x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds25x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds30x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds50x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds100x_alleleReads_ill, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

#### Genotyping

##### error 0.01

```{r}
#load locus table
fullSet_singleSNP_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

#generate singleSNP genotypes using the polyGen algorithm
ds6x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds6x_alleleReads_ill)

ds10x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds10x_alleleReads_ill)

ds15x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds15x_alleleReads_ill)

ds20x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ill)

ds25x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds25x_alleleReads_ill)

ds30x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds30x_alleleReads_ill)

ds50x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds50x_alleleReads_ill)

ds100x_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds100x_alleleReads_ill)

# write results
write.table(ds6x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds10x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds15x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds20x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds25x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds30x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds50x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds100x_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

##### error 0.10

```{r}
ds20x.e10_ill_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ill, epsilon = 0.10)

write.table(ds20x.e10_ill_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x.e10_ill_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

### Nanopore

#### Downsample

```{r}
# downsample allele reads
ds6x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 6, 100)

ds10x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 10, 100)

ds15x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 15, 100)

ds20x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 20, 100)

ds25x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 25, 100)

ds30x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 30, 100)

ds50x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 50, 100)

ds100x_alleleReads_ont <- downsample_alleleReads(alleleReads_ont_gtscore_samplesAll, 100, 100)

# export downsampled allele reads
write.table(ds6x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds10x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds15x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds20x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds25x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds30x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds50x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(ds100x_alleleReads_ont, file = "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_AlleleReads_singleSNPs.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

#### Genotyping

##### error 0.01

```{r}
#load locus table
fullSet_singleSNP_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

#generate singleSNP genotypes using the polyGen algorithm
ds6x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds6x_alleleReads_ont)

ds10x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds10x_alleleReads_ont)

ds15x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds15x_alleleReads_ont)

ds20x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ont)

ds25x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds25x_alleleReads_ont)

ds30x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds30x_alleleReads_ont)

ds50x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds50x_alleleReads_ont)

ds100x_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds100x_alleleReads_ont)

# write results
write.table(ds6x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds10x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds15x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds20x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds25x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds30x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds50x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

write.table(ds100x_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

##### error 0.10 + error 0.02

```{r}
# 0.10
ds20x.e10_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ont, epsilon = 0.10)

write.table(ds20x.e10_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x.e10_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")

# 0.02
ds20x.e2_ont_polyGenResults <- polyGen(fullSet_singleSNP_locusTable, ds20x_alleleReads_ont, epsilon = 0.02)

write.table(ds20x.e2_ont_polyGenResults, "./downsamplingTests/tamRun4/01_run4GTscore/ds20x.e2_ont_fullSet_polyGenResults_singleSNP.txt", quote = FALSE, sep = "\t")
```

## 8.3 Blood samples

### 8.3.1 Data

#### Original dataframes

```{r}
# ILLUMINA
genos_ill_gtscore_samplesBlood_lociDual_ds0x <- genos_ill_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ill") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by = "sampleID_ill", all.x = T) %>%
  relocate(sampleID_unique) %>%
  select(-sampleID_ill) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame() %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds20x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds25x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ill_gtscore_samplesBlood_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

# NANOPORE
genos_ont_gtscore_samplesBlood_lociDual_ds0x <- genos_ont_gtscore_samplesBlood_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_ont") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by = "sampleID_ont", all.x = T) %>%
  relocate(sampleID_unique) %>%
  select(-sampleID_ont) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame() %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds20x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds25x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds25x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))
```

#### Long dataframes

```{r}
# ILLUMINA
genos_ill_gtscore_samplesBlood_lociDual_ds0x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds0x")

genos_ill_gtscore_samplesBlood_lociDual_ds6x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds6x")

genos_ill_gtscore_samplesBlood_lociDual_ds10x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds10x")

genos_ill_gtscore_samplesBlood_lociDual_ds15x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds15x")

genos_ill_gtscore_samplesBlood_lociDual_ds20x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds20x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds20x")

genos_ill_gtscore_samplesBlood_lociDual_ds25x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds25x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds25x")

genos_ill_gtscore_samplesBlood_lociDual_ds30x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds30x")

genos_ill_gtscore_samplesBlood_lociDual_ds50x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds50x")

genos_ill_gtscore_samplesBlood_lociDual_ds100x_l <- genos_ill_gtscore_samplesBlood_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds100x")


# NANOPORE
genos_ont_gtscore_samplesBlood_lociDual_ds0x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds0x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds0x")

genos_ont_gtscore_samplesBlood_lociDual_ds6x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds6x")

genos_ont_gtscore_samplesBlood_lociDual_ds10x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds10x")

genos_ont_gtscore_samplesBlood_lociDual_ds15x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds15x")

genos_ont_gtscore_samplesBlood_lociDual_ds20x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds20x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds20x")

genos_ont_gtscore_samplesBlood_lociDual_ds25x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds25x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds25x")

genos_ont_gtscore_samplesBlood_lociDual_ds30x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds30x")

genos_ont_gtscore_samplesBlood_lociDual_ds50x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds50x")

genos_ont_gtscore_samplesBlood_lociDual_ds100x_l <- genos_ont_gtscore_samplesBlood_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds100x")
```

#### Combo dataframe

```{r}
genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos <-
  # illumina
  genos_ill_gtscore_samplesBlood_lociDual_ds0x_l %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds6x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds10x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds15x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds20x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds25x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds30x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds50x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ill_gtscore_samplesBlood_lociDual_ds100x_l,
        by = c("sampleID", "locus")) %>%
  # nanopore
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds0x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds6x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds10x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds15x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds20x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds25x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds30x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds50x_l,
        by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds100x_l,
        by = c("sampleID", "locus"))
```

old comparison dataframe script:

```{r}
genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_ds6x <- genos_ill_gtscore_samplesBlood_lociDual_ds6x_l %>%
  na.omit() %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds6x_l,
        by = c("sampleID", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_ds6x) ~ "NA",
      ill_ds6x == ont_ds6x ~ "yes",
      ill_ds6x != ont_ds6x ~ "no"
      )
    )

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_ds10x <- genos_ill_gtscore_samplesBlood_lociDual_ds10x_l %>%
  na.omit() %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds10x_l,
        by = c("sampleID", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_ds10x) ~ "NA",
      ill_ds10x == ont_ds10x ~ "yes",
      ill_ds10x != ont_ds10x ~ "no"
      )
    )

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_ds15x <- genos_ill_gtscore_samplesBlood_lociDual_ds15x_l %>%
  na.omit() %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds15x_l,
        by = c("sampleID", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_ds15x) ~ "NA",
      ill_ds15x == ont_ds15x ~ "yes",
      ill_ds15x != ont_ds15x ~ "no"
      )
    )

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_ds30x <- genos_ill_gtscore_samplesBlood_lociDual_ds30x_l %>%
  na.omit() %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds30x_l,
        by = c("sampleID", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_ds30x) ~ "NA",
      ill_ds30x == ont_ds30x ~ "yes",
      ill_ds30x != ont_ds30x ~ "no"
      )
    )

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_ds50x <- genos_ill_gtscore_samplesBlood_lociDual_ds50x_l %>%
  na.omit() %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds50x_l,
        by = c("sampleID", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_ds50x) ~ "NA",
      ill_ds50x == ont_ds50x ~ "yes",
      ill_ds50x != ont_ds50x ~ "no"
      )
    )

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_ds0x <- genos_ill_gtscore_samplesBlood_lociDual_ds0x_l %>%
  na.omit() %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds0x_l,
        by = c("sampleID", "locus")) %>%
  mutate(
    match = case_when(
      is.na(ont_ds0x) ~ "NA",
      ill_ds0x == ont_ds0x ~ "yes",
      ill_ds0x != ont_ds0x ~ "no"
      )
    )
```

### 8.3.2 Geno success

```{r}
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests <- 
  genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  select(-locus) %>%
  group_by(sampleID) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  as.data.frame() %>%
#  column_to_rownames("sampleID") %>%
#  mutate_all(list(~ ./140))
  pivot_longer(!sampleID,
               names_to = "depth",
               values_to = "totalGenos") %>%
  mutate(
    propSuccess = totalGenos/140
  ) %>%
  separate(., depth, into = c("seqPlatform", "depth"), sep = "_") %>%
  mutate(
    seqPlatform = case_when(
      seqPlatform == "ill" ~ "Illumina",
      seqPlatform == "ont" ~ "Nanopore"
    )
  )
```

Plots

```{r}
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests %>%
  ggplot(aes(x = factor(depth, c("ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x", "ds0x")), y = propSuccess, fill = seqPlatform)) +
  geom_boxplot() +
  theme_linedraw() +
  labs(
    title = "Blood sample genotype success across coverage depths for Illumina vs. Nanopore",
    x = "Downsampled sequencing depth per sample per locus",
    y = "Proportion dual-species loci genotyped per sample")
```

OLD CODE:

By sample

```{r}
# ILLUMINA
genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds6x <- genos_ill_gtscore_samplesBlood_lociDual_ds6x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina",
    depth = "ds6x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds10x <- genos_ill_gtscore_samplesBlood_lociDual_ds10x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina",
    depth = "ds10x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds15x <- genos_ill_gtscore_samplesBlood_lociDual_ds15x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina",
    depth = "ds15x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds30x <- genos_ill_gtscore_samplesBlood_lociDual_ds30x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina",
    depth = "ds30x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds50x <- genos_ill_gtscore_samplesBlood_lociDual_ds50x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Illumina",
    depth = "ds50x"
  ) %>%
  rownames_to_column("sampleID")

# NANOPORE
genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds6x <- genos_ont_gtscore_samplesBlood_lociDual_ds6x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore",
    depth = "ds6x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds10x <- genos_ont_gtscore_samplesBlood_lociDual_ds10x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore",
    depth = "ds10x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds15x <- genos_ont_gtscore_samplesBlood_lociDual_ds15x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore",
    depth = "ds15x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds30x <- genos_ont_gtscore_samplesBlood_lociDual_ds30x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore",
    depth = "ds30x"
  ) %>%
  rownames_to_column("sampleID")

genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds50x <- genos_ont_gtscore_samplesBlood_lociDual_ds50x %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalGenos = rowSums(!is.na(.))
  ) %>%
  select(totalGenos) %>%
  mutate(
    propSuccess = round(totalGenos/140, 2),
    seqPlatform = "Nanopore",
    depth = "ds50x"
  ) %>%
  rownames_to_column("sampleID")

# Combined
## add in originals too (no downsampling)
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_ds0x <- genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill", all.x = T) %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont", all.x = T) %>%
  select(-sampleID) %>%
  mutate(
    sampleID = coalesce(sampleID_unique.x, sampleID_unique.y),
    .keep = "unused"
  ) %>%
  relocate(sampleID) %>%
  mutate(
    depth = "ds0x"
  )

genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests <- rbind(
  genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds6x,
  genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds10x,
  genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds15x,
  genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds30x,
  genoSuccess_bySample_ill_gtscore_samplesBlood_lociDual_ds50x,
  genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds6x,
  genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds10x,
  genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds15x,
  genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds30x,
  genoSuccess_bySample_ont_gtscore_samplesBlood_lociDual_ds50x,
  genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_ds0x)
```

Plots

```{r}
genoSuccess_bySample_illONT_gtscore_samplesBlood_lociDual_dsTests %>%
  ggplot(aes(x = factor(depth, c("ds6x", "ds10x", "ds15x", "ds30x", "ds50x", "ds0x")), y = propSuccess, fill = seqPlatform)) +
  geom_boxplot() +
  theme_linedraw() +
  labs(
    title = "Blood sample genotype success across coverage depths for Illumina vs. Nanopore",
    x = "Downsampled sequencing depth per sample per locus",
    y = "Proportion dual-species loci genotyped per sample")
```

### 8.3.3 Geno consistency

#### All genos

##### Run comparison

```{r}
df <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos

cols_of_interest <- setdiff(names(df), c("sampleID", "locus"))

colCombos <- combn(cols_of_interest, 2)

check_dsTests <- df[,colCombos[1,]] > 0 & df[,colCombos[2,]] > 0 & df[,colCombos[1,]] == df[,colCombos[2,]]
colnames(check_dsTests) = paste0(colCombos[1,], ".", colCombos[2,])

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_dsTests_allCombos <- cbind(df, ifelse(check_dsTests == T, 1, 0))
```

##### Summary

```{r}
# all comparisons
genoConsistency_samplesBlood_dsTests_allGenos_summary <- genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_dsTests_allCombos %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("combo") %>%
  separate(., combo, into = c("combo", "metric"), sep = "_t") %>%
  mutate(metric = str_c("t", metric)) %>%
  pivot_wider(names_from = metric, values_from = V1) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.") %>%
  mutate(
    propCommonGenos = round(totalCommonGenos/(30*140), 2),
    propMatchGenos = round(totalMatchGenos/totalCommonGenos, 4)
  ) %>%
  mutate(
    compType = str_c(set1, set2, sep = ".")
  )

# ill vs. ont only
genoConsistency_samplesBlood_dsTests_allGenos_illONTOnly_summary <- genoConsistency_samplesBlood_dsTests_allGenos_summary %>%
  filter(str_detect(compType, "ill") & str_detect(compType, "ont")) %>%
  select(-compType)

# ill vs. ont same coverage only
genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary <- genoConsistency_samplesBlood_dsTests_allGenos_summary %>%
  filter(compType %in% c("ill_ds0x.ont_ds0x",
                         "ill_ds6x.ont_ds6x",
                         "ill_ds10x.ont_ds10x",
                         "ill_ds15x.ont_ds15x",
                         "ill_ds20x.ont_ds20x",
                         "ill_ds25x.ont_ds25x",
                         "ill_ds30x.ont_ds30x",
                         "ill_ds50x.ont_ds50x",
                         "ill_ds100x.ont_ds100x")) %>%
  mutate(coverage = gsub("ill_", "", set1))
```

##### Plots

```{r}
# all genos
genoConsistency_samplesBlood_dsTests_allGenos_illONTOnly_summary %>%
  ggplot(aes(x = factor(set1, c("ill_ds6x", "ill_ds10x", "ill_ds15x", "ill_ds20x", "ill_ds25x", "ill_ds30x", "ill_ds50x", "ill_ds100x", "ill_ds0x")),
             y = propMatchGenos,
             fill = factor(set2, c("ont_ds6x", "ont_ds10x", "ont_ds15x", "ont_ds20x", "ont_ds25x", "ont_ds30x", "ont_ds50x", "ont_ds100x", "ont_ds0x")))) +
  geom_bar(position = "dodge", stat = "identity") +
  theme_linedraw() +
  guides(
    color = guide_legend(title = "ONT depth")
  ) +
  labs(
    title = "Geno consistency across coverage depths for Illumina vs. Nanopore",
    x = "Illumina depth",
    y = "Proportion matching genotypes"
    ) +
  scale_fill_discrete(name = "ONT depth")

# same coverage only
## barplot
genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary %>%
  ggplot(aes(x = factor(coverage, c("ds0x", "ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x")), y = propMatchGenos)) +
  geom_bar(stat = "identity") +
  theme_linedraw() +
  labs(
    title = "Proportion of matching genotypes across coverage depths for Illumina vs. Nanopore",
    x = "coverage"
  )

## dot plots
genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary %>%
  ggplot(aes(x = totalCommonGenos, y = propMatchGenos)) +
  geom_point() +
  facet_wrap(~factor(set1, c("ill_ds6x", "ill_ds10x", "ill_ds15x", "ill_ds20x", "ill_ds25x", "ill_ds30x", "ill_ds50x", "ill_ds100x", "ill_ds0x"))) +
#  scale_color_brewer(palette = "RdYlBu") +
#  scale_color_viridis(discrete = T) +
  scale_shape_discrete(name = "ONT depth") +
  theme_linedraw() +
  labs(
    title = "Geno consistency across coverage depths for 30 blood samples + 140 loci (up to 4200 genos) for Illumina vs. Nanopore",
    x = "Total common genos",
    y = "Proportion of matched genos (out of common)"
  ) +
#  xlim(1200, 3200) +
  ylim(0.90, 1.0) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds6x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds6x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds10x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds10x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds15x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds15x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds20x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds20x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds25x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds25x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds30x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds30x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds50x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds50x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds100x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds100x)))) +
  geom_vline(data = filter(genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary, set1=="ill_ds0x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos$ill_ds0x))))
```

##### Stats

Comparing match counts between Illumina and Nanopore results w/the same coverage show NO significant difference between coverage cutoffs.

```{r}
stats_table <- genoConsistency_samplesBlood_dsTests_allGenos_illONT_sameCoverageOnly_summary %>%
  select(c(coverage, totalMatchGenos, totalCommonGenos)) %>%
  column_to_rownames("coverage") %>%
  t() %>%
  as.data.table()

# run chisq test
chisq.test(stats_table, correct = F)

# view residuals; can show where disagreement of obs/exp counts is greatest (look for residuals w/largest abs values)
chisq.test(stats_table, correct = F)$resi
```

#### All genos - error 0.10

##### Data

```{r}
genos_ont_gtscore_samplesBlood_lociDual_ds20x.e10 <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x.e10_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds20x.e10_l <- genos_ont_gtscore_samplesBlood_lociDual_ds20x.e10 %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds20xe10")
```

##### Compare

```{r}
genoCompare_blood_illONT_errorTests_allGenos <- genos_ill_gtscore_samplesBlood_lociDual_ds20x_l %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds20x_l,
        by = c("sampleID", "locus"),
        all.x = T) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds20x.e10_l,
        by = c("sampleID", "locus"),
        all.x = T)

df <- genoCompare_blood_illONT_errorTests_allGenos

cols_of_interest <- setdiff(names(df), c("sampleID", "locus"))

colCombos <- combn(cols_of_interest, 2)

check_dsTests <- df[,colCombos[1,]] > 0 & df[,colCombos[2,]] > 0 & df[,colCombos[1,]] == df[,colCombos[2,]]
colnames(check_dsTests) = paste0(colCombos[1,], ".", colCombos[2,])

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_errorTest <- cbind(df, ifelse(check_dsTests == T, 1, 0))
```

##### Summary

```{r}
# all comparisons
genoConsistency_samplesBlood_errorTests_allGenos_summary <- genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_errorTest %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("combo") %>%
  separate(., combo, into = c("combo", "metric"), sep = "_t") %>%
  mutate(metric = str_c("t", metric)) %>%
  pivot_wider(names_from = metric, values_from = V1) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.") %>%
  mutate(
    propCommonGenos = round(totalCommonGenos/(30*140), 2),
    propMatchGenos = round(totalMatchGenos/totalCommonGenos, 4)
  ) %>%
  mutate(
    compType = str_c(set1, set2, sep = ".")
  )
```

##### testing

```{r}
test1 <- genoCompare_blood_illONT_errorTests_allGenos %>%
  na.omit() %>%
  mutate(
    match_illONT = case_when(
      ill_ds20x == ont_ds20x ~ "yes",
      .default = "no"
    ),
    match_illONTe10 = case_when(
      ill_ds20x == ont_ds20xe10 ~ "yes",
      .default = "no"
    ),
    compare = 
      case_when(
        match_illONT == "no" & match_illONTe10 == "no" ~ "bothMismatch",
        match_illONT == "yes" & match_illONTe10 == "no" ~ "e10Mismatch",
        match_illONT == "no" & match_illONTe10 == "yes" ~ "e1Mismatch",
        match_illONT == "yes" & match_illONTe10 == "yes" ~ "bothMatch",
      )
  )

table(test1$compare)
```

#### All genos - error 0.2

##### Data

```{r}
genos_ont_gtscore_samplesBlood_lociDual_ds20x.e2 <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds20x.e2_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_blood$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                )) %>%
  select(order(colnames(.)))

genos_ont_gtscore_samplesBlood_lociDual_ds20x.e2_l <- genos_ont_gtscore_samplesBlood_lociDual_ds20x.e2 %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds20xe2")
```

##### Compare

```{r}
genoCompare_blood_illONT_errorTests_allGenos <- genos_ill_gtscore_samplesBlood_lociDual_ds20x_l %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds20x_l,
        by = c("sampleID", "locus"),
        all.x = T) %>%
  merge(.,
        genos_ont_gtscore_samplesBlood_lociDual_ds20x.e2_l,
        by = c("sampleID", "locus"),
        all.x = T)

df <- genoCompare_blood_illONT_errorTests_allGenos

cols_of_interest <- setdiff(names(df), c("sampleID", "locus"))

colCombos <- combn(cols_of_interest, 2)

check_dsTests <- df[,colCombos[1,]] > 0 & df[,colCombos[2,]] > 0 & df[,colCombos[1,]] == df[,colCombos[2,]]
colnames(check_dsTests) = paste0(colCombos[1,], ".", colCombos[2,])

genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_errorTest.e2 <- cbind(df, ifelse(check_dsTests == T, 1, 0))
```

##### Summary

```{r}
# all comparisons
genoConsistency_samplesBlood_errorTests.e2_allGenos_summary <- genoCompare_illONT_gtscore_samplesBlood_lociDual_allGenos_errorTest.e2 %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("combo") %>%
  separate(., combo, into = c("combo", "metric"), sep = "_t") %>%
  mutate(metric = str_c("t", metric)) %>%
  pivot_wider(names_from = metric, values_from = V1) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.") %>%
  mutate(
    propCommonGenos = round(totalCommonGenos/(30*140), 2),
    propMatchGenos = round(totalMatchGenos/totalCommonGenos, 4)
  ) %>%
  mutate(
    compType = str_c(set1, set2, sep = ".")
  )
```

##### testing

```{r}
test1 <- genoCompare_blood_illONT_errorTests_allGenos %>%
  na.omit() %>%
  mutate(
    match_illONT = case_when(
      ill_ds20x == ont_ds20x ~ "yes",
      .default = "no"
    ),
    match_illONTe10 = case_when(
      ill_ds20x == ont_ds20xe10 ~ "yes",
      .default = "no"
    ),
    compare = 
      case_when(
        match_illONT == "no" & match_illONTe10 == "no" ~ "bothMismatch",
        match_illONT == "yes" & match_illONTe10 == "no" ~ "e10Mismatch",
        match_illONT == "no" & match_illONTe10 == "yes" ~ "e1Mismatch",
        match_illONT == "yes" & match_illONTe10 == "yes" ~ "bothMatch",
      )
  )

table(test1$compare)
```

#### opt100 genos

##### Data

```{r}
opt100_samplesLoci_illONT_gtscore_blood_dsTests <-
  optimize_samplesLoci_illONT(
    genos_ill_gtscore_samplesBlood_lociDual_ds0x,
    genos_ont_gtscore_samplesBlood_lociDual_ds0x,
    1.0
  ) %>%
  dplyr::rename("ds0x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds6x,
          genos_ont_gtscore_samplesBlood_lociDual_ds6x,
          1.0
          )
        ) %>%
  dplyr::rename("ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds10x,
          genos_ont_gtscore_samplesBlood_lociDual_ds10x,
          1.0
          )
        ) %>%
  dplyr::rename("ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds15x,
          genos_ont_gtscore_samplesBlood_lociDual_ds15x,
          1.0
          )
        ) %>%
  dplyr::rename("ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds20x,
          genos_ont_gtscore_samplesBlood_lociDual_ds20x,
          1.0
          )
        ) %>%
  dplyr::rename("ds20x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds25x,
          genos_ont_gtscore_samplesBlood_lociDual_ds25x,
          1.0
          )
        ) %>%
  dplyr::rename("ds25x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds30x,
          genos_ont_gtscore_samplesBlood_lociDual_ds30x,
          1.0
          )
        ) %>%
  dplyr::rename("ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds50x,
          genos_ont_gtscore_samplesBlood_lociDual_ds50x,
          1.0
          )
        ) %>%
  dplyr::rename("ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci_illONT(
          genos_ill_gtscore_samplesBlood_lociDual_ds100x,
          genos_ont_gtscore_samplesBlood_lociDual_ds100x,
          1.0
          )
        ) %>%
  dplyr::rename("ds100x" = "loci")
```

OLD CODE:

(doesn't combine illONT to optimize samples/loci)

```{r}
# ILLUMINA
opt100_samplesLoci_ill_gtscore_blood_dsTests <- optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds6x, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds10x, 1)) %>%
  dplyr::rename("loci100_ill.ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds15x, 1)) %>%
  dplyr::rename("loci100_ill.ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds20x, 1)) %>%
  dplyr::rename("loci100_ill.ds20x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds25x, 1)) %>%
  dplyr::rename("loci100_ill.ds25x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds30x, 1)) %>%
  dplyr::rename("loci100_ill.ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds50x, 1)) %>%
  dplyr::rename("loci100_ill.ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds100x, 1)) %>%
  dplyr::rename("loci100_ill.ds100x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual, 1)) %>%
  dplyr::rename("loci100_ill.ds0x" = "loci")

# NANOPORE
opt100_samplesLoci_ont_gtscore_blood_dsTests <- optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds6x, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds10x, 1)) %>%
  dplyr::rename("loci100_ont.ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds15x, 1)) %>%
  dplyr::rename("loci100_ont.ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds20x, 1)) %>%
  dplyr::rename("loci100_ont.ds20x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds25x, 1)) %>%
  dplyr::rename("loci100_ont.ds25x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds30x, 1)) %>%
  dplyr::rename("loci100_ont.ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds50x, 1)) %>%
  dplyr::rename("loci100_ont.ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds100x, 1)) %>%
  dplyr::rename("loci100_ont.ds100x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual, 1)) %>%
  dplyr::rename("loci100_ont.ds0x" = "loci")
```

##### Subsample to optGenos

```{r}
# ds0x
optGenos100_illONT_gtscore_blood_ds0x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds0x,
  genos_ont_gtscore_samplesBlood_lociDual_ds0x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds0x <- optGenos100_illONT_gtscore_blood_ds0x[[1]]

optGenos100_ont_gtscore_blood_ds0x <- optGenos100_illONT_gtscore_blood_ds0x[[2]]

# ds6x
optGenos100_illONT_gtscore_blood_ds6x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds6x,
  genos_ont_gtscore_samplesBlood_lociDual_ds6x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds6x <- optGenos100_illONT_gtscore_blood_ds6x[[1]]

optGenos100_ont_gtscore_blood_ds6x <- optGenos100_illONT_gtscore_blood_ds6x[[2]]

# ds10x 
optGenos100_illONT_gtscore_blood_ds10x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds10x,
  genos_ont_gtscore_samplesBlood_lociDual_ds10x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds10x <- optGenos100_illONT_gtscore_blood_ds10x[[1]]

optGenos100_ont_gtscore_blood_ds10x <- optGenos100_illONT_gtscore_blood_ds10x[[2]]

# ds15x
optGenos100_illONT_gtscore_blood_ds15x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds15x,
  genos_ont_gtscore_samplesBlood_lociDual_ds15x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds15x <- optGenos100_illONT_gtscore_blood_ds15x[[1]]

optGenos100_ont_gtscore_blood_ds15x <- optGenos100_illONT_gtscore_blood_ds15x[[2]]

# ds20x
optGenos100_illONT_gtscore_blood_ds20x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds20x,
  genos_ont_gtscore_samplesBlood_lociDual_ds20x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds20x <- optGenos100_illONT_gtscore_blood_ds20x[[1]]

optGenos100_ont_gtscore_blood_ds20x <- optGenos100_illONT_gtscore_blood_ds20x[[2]]

# ds25x
optGenos100_illONT_gtscore_blood_ds25x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds25x,
  genos_ont_gtscore_samplesBlood_lociDual_ds25x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds25x <- optGenos100_illONT_gtscore_blood_ds25x[[1]]

optGenos100_ont_gtscore_blood_ds25x <- optGenos100_illONT_gtscore_blood_ds25x[[2]]

# ds30x
optGenos100_illONT_gtscore_blood_ds30x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds30x,
  genos_ont_gtscore_samplesBlood_lociDual_ds30x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds30x <- optGenos100_illONT_gtscore_blood_ds30x[[1]]

optGenos100_ont_gtscore_blood_ds30x <- optGenos100_illONT_gtscore_blood_ds30x[[2]]

# ds50x
optGenos100_illONT_gtscore_blood_ds50x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds50x,
  genos_ont_gtscore_samplesBlood_lociDual_ds50x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds50x <- optGenos100_illONT_gtscore_blood_ds50x[[1]]

optGenos100_ont_gtscore_blood_ds50x <- optGenos100_illONT_gtscore_blood_ds50x[[2]]

# ds100x
optGenos100_illONT_gtscore_blood_ds100x <- optimized_genos_illONT(
  genos_ill_gtscore_samplesBlood_lociDual_ds100x,
  genos_ont_gtscore_samplesBlood_lociDual_ds100x,
  10,
  1.0
)
optGenos100_ill_gtscore_blood_ds100x <- optGenos100_illONT_gtscore_blood_ds100x[[1]]

optGenos100_ont_gtscore_blood_ds100x <- optGenos100_illONT_gtscore_blood_ds100x[[2]]
```

##### Compare

```{r}
# ds6x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds6x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds6x)) %>%
  select(c(sampleID, locus, ill_ds6x, ont_ds6x)) %>%
  mutate(
    match = case_when(
      ill_ds6x == ont_ds6x ~ "yes",
      .default = "no"
    )
  )

# ds10x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds10x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds10x)) %>%
  select(c(sampleID, locus, ill_ds10x, ont_ds10x)) %>%
  mutate(
    match = case_when(
      ill_ds10x == ont_ds10x ~ "yes",
      .default = "no"
    )
  )

# ds15x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds15x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds15x)) %>%
  select(c(sampleID, locus, ill_ds15x, ont_ds15x)) %>%
  mutate(
    match = case_when(
      ill_ds15x == ont_ds15x ~ "yes",
      .default = "no"
    )
  )

# ds20x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds20x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds20x)) %>%
  select(c(sampleID, locus, ill_ds20x, ont_ds20x)) %>%
  mutate(
    match = case_when(
      ill_ds20x == ont_ds20x ~ "yes",
      .default = "no"
    )
  )

# ds25x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds25x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds25x)) %>%
  select(c(sampleID, locus, ill_ds25x, ont_ds25x)) %>%
  mutate(
    match = case_when(
      ill_ds25x == ont_ds25x ~ "yes",
      .default = "no"
    )
  )

# ds30x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds30x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds30x)) %>%
  select(c(sampleID, locus, ill_ds30x, ont_ds30x)) %>%
  mutate(
    match = case_when(
      ill_ds30x == ont_ds30x ~ "yes",
      .default = "no"
    )
  )

# ds50x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds50x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds50x)) %>%
  select(c(sampleID, locus, ill_ds50x, ont_ds50x)) %>%
  mutate(
    match = case_when(
      ill_ds50x == ont_ds50x ~ "yes",
      .default = "no"
    )
  )

# ds100x
genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% colnames(optGenos100_ill_gtscore_blood_ds100x)) %>%
  filter(locus %in% rownames(optGenos100_ill_gtscore_blood_ds100x)) %>%
  select(c(sampleID, locus, ill_ds100x, ont_ds100x)) %>%
  mutate(
    match = case_when(
      ill_ds100x == ont_ds100x ~ "yes",
      .default = "no"
    )
  )
```

##### Summary

```{r}
genoCompare_illONT_samplesBlood_opt100_dsTests <- 
  data.frame(
    coverage = c("ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x"),
    totalMismatch = c(
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$match == "no"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$match == "no")
    ),
    totalMatch = c(
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$match == "yes"),
      sum(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$match == "yes")
    ),
    totalSamples = c(
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$sampleID)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$sampleID))
    ),
    totalLoci = c(
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$locus)),
      length(unique(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$locus))
    )
  ) %>%
  mutate(
    totalGenos = totalMismatch + totalMatch,
    propMatch = round(totalMatch/totalGenos, 3)
  ) %>%
  select(c(coverage, totalSamples, totalLoci, totalGenos, totalMismatch, totalMatch, propMatch))


  
  rbind(
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds6x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds10x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds15x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds20x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds25x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds30x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds50x$match),
  table(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_ds100x$match)
) %>%
  as.data.frame() %>%
  mutate(
    coverage = c("ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x")
  ) %>%
  mutate(
    totalGenos = no + yes,
    propMatch = yes/totalGenos
  ) %>%
  select(c(coverage, propMatch, totalGenos))
```

##### OLD CODE BELOW

OLD CODE:

```{r}
# ds0x
optGenos100_ill_gtscore_blood_ds0x <- optimized_genos(genos_ill_gtscore_samplesBlood_lociDual,
                                                         25, 1.0)

# ds6x
optGenos100_ill_gtscore_blood_ds20x <- optimized_genos(genos_ill_gtscore_samplesBlood_lociDual, 20, 1.0)

# optGenos based on ONT + ds30x downsampling
optGenos100_ont_gtscore_blood_ds30x <- optimized_genos(genos_ont_gtscore_samplesBlood_lociDual_ds30x, 20, 1.0)
```

##### Sample/loci lists

```{r}
# baseline = illumina + ds0x
sampleList_optGenos100_gtscore_samplesBlood_illds0x <- data.frame(
  sampleID_ill = colnames(optGenos100_ill_gtscore_blood_ds0x)
) %>%
  merge(., md_tamRun4_blood[, c("sampleID_ill", "sampleID_ont", "sampleID_unique")], by = "sampleID_ill")

lociList_optGenos100_gtscore_samplesBlood_illds0x <- optGenos100_ill_gtscore_blood_ds0x %>%
  rownames_to_column("locus") %>%
  select(locus)

# ds20x
sampleList_optGenos100_gtscore_samplesBlood_ds20x <- colnames(optGenos100_ill_gtscore_blood_ds20x)

lociList_optGenos100_gtscore_samplesBlood_ds20x <- rownames(optGenos100_ill_gtscore_blood_ds20x)

# starting w/ont_ds30x as a baseline
sampleList_optGenos100_gtscore_samplesBlood_ontds30x <- data.frame(
  sampleID_unique = colnames(optGenos100_ont_gtscore_blood_ds30x)
) %>%
  merge(., md_tamRun4_blood[, c("sampleID_ill", "sampleID_ont", "sampleID_unique")], by = "sampleID_unique")

lociList_optGenos100_gtscore_samplesBlood_ontds30x <- optGenos100_ont_gtscore_blood_ds30x %>%
  rownames_to_column("locus") %>%
  select(locus)
```

##### optGenos df

```{r}
# baseline illumina + ds20x
genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_optGenos <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_allGenos %>%
  filter(sampleID %in% sampleList_optGenos100_gtscore_samplesBlood_ds20x) %>%
  filter(locus %in% lociList_optGenos100_gtscore_samplesBlood_ds20x)
```

##### Run comparison

```{r}
df_opt100 <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_optGenos

cols_of_interest <- setdiff(names(df_opt100), c("sampleID", "locus"))

colCombos <- combn(cols_of_interest, 2)

check_dsTests <- df_opt100[,colCombos[1,]] > 0 & df_opt100[,colCombos[2,]] > 0 & df_opt100[,colCombos[1,]] == df_opt100[,colCombos[2,]]
colnames(check_dsTests) = paste0(colCombos[1,], ".", colCombos[2,])

genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos <- cbind(df_opt100, ifelse(check_dsTests == T, 1, 0))
```

##### Summary

###### Overview

```{r}
genoConsistency_samplesBlood_dsTests_opt100_summary <- genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("combo") %>%
  separate(., combo, into = c("combo", "metric"), sep = "_t") %>%
  mutate(metric = str_c("t", metric)) %>%
  pivot_wider(names_from = metric, values_from = V1) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.") %>%
  mutate(
    set1_totalGenos = case_when(
      set1 == "ill_ds6x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds6x)),
      set1 == "ill_ds10x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds10x)),
      set1 == "ill_ds15x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds15x)),
      set1 == "ill_ds20x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds20x)),
      set1 == "ill_ds25x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds25x)),
      set1 == "ill_ds30x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds30x)),
      set1 == "ill_ds50x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds50x)),
      set1 == "ill_ds100x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds100x)),
      set1 == "ill_ds0x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds0x)),
      set1 == "ont_ds6x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds6x)),
      set1 == "ont_ds10x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds10x)),
      set1 == "ont_ds15x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds15x)),
      set1 == "ont_ds20x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds20x)),
      set1 == "ont_ds30x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds30x)),
      set1 == "ont_ds50x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds50x)),
      set1 == "ont_ds100x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds100x)),
      set1 == "ont_ds0x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds0x))
    )
  ) %>%
  mutate(
    set2_totalGenos = case_when(
      set2 == "ill_ds6x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds6x)),
      set2 == "ill_ds10x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds10x)),
      set2 == "ill_ds15x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds15x)),
      set2 == "ill_ds20x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds20x)),
      set2 == "ill_ds25x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds25x)),
      set2 == "ill_ds30x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds30x)),
      set2 == "ill_ds50x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds50x)),
      set2 == "ill_ds100x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds100x)),
      set2 == "ill_ds0x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds0x)),
      set2 == "ont_ds6x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds6x)),
      set2 == "ont_ds10x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds10x)),
      set2 == "ont_ds15x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds15x)),
      set2 == "ont_ds20x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds20x)),
      set2 == "ont_ds25x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds25x)),
      set2 == "ont_ds30x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds30x)),
      set2 == "ont_ds50x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds50x)),
      set2 == "ont_ds100x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds100x)),
      set2 == "ont_ds0x" ~ sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ont_ds0x))
    )
  ) %>%
  mutate(
    propCommonGenos = round(totalCommonGenos/(nrow(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos)), 2),
    propMatchGenos = round(totalMatchGenos/totalCommonGenos, 4)
  ) %>%
  mutate(
    compType = case_when(
      str_detect(set1, "ill") & str_detect(set2, "ill") ~ "ill.ill",
      str_detect(set1, "ill") & str_detect(set2, "ont") ~ "ill.ont",
      str_detect(set1, "ont") & str_detect(set2, "ill") ~ "ill.ill",
      str_detect(set1, "ont") & str_detect(set2, "ont") ~ "ont.ont"
    )
  ) %>%
  relocate(c("set1_totalGenos", "set2_totalGenos"), .after = "set2")

genoConsistency_samplesBlood_dsTests_opt100_illONTOnly_summary <- genoConsistency_samplesBlood_dsTests_opt100_summary %>%
  filter(compType == "ill.ont") %>%
  select(-compType)

genoConsistency_samplesBlood_dsTests_opt100_sameCoverageOnly_summary <- genoConsistency_samplesBlood_dsTests_opt100_summary %>%
  mutate(compType = str_c(set1, set2, sep = ".")) %>%
  filter(compType %in% c("ill_ds0x.ont_ds0x",
                         "ill_ds6x.ont_ds6x",
                         "ill_ds10x.ont_ds10x",
                         "ill_ds15x.ont_ds15x",
                         "ill_ds20x.ont_ds20x",
                         "ill_ds25x.ont_ds25x",
                         "ill_ds30x.ont_ds30x",
                         "ill_ds50x.ont_ds50x",
                         "ill_ds100x.ont_ds100x"))
```

###### By sample

```{r}
genoConsistency_samplesBlood_dsTests_opt100_summary_bySample1 <- genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos %>%
  group_by(sampleID) %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  pivot_longer(!sampleID) %>%
  separate(., name, into = c("combo", "metric"), sep = "_t") %>%
  mutate(metric = str_c("t", metric)) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.")

genoConsistency_samplesBlood_dsTests_opt100_summary_bySample2 <- genoConsistency_samplesBlood_dsTests_opt100_summary_bySample1 %>%
  filter(metric == "totalCommonGenos") %>%
  select(-metric) %>%
  dplyr::rename("totalCommonGenos" = "value")

genoConsistency_samplesBlood_dsTests_opt100_summary_bySample3 <- genoConsistency_samplesBlood_dsTests_opt100_summary_bySample1 %>%
  filter(metric == "totalMatchGenos") %>%
  select(-metric) %>%
  dplyr::rename("totalMatchGenos" = "value")

# FULL SET
genoConsistency_samplesBlood_dsTests_opt100_summary_bySample <- merge(genoConsistency_samplesBlood_dsTests_opt100_summary_bySample2, genoConsistency_samplesBlood_dsTests_opt100_summary_bySample3, by = c("sampleID", "set1", "set2")) %>%
  mutate(
    propMatch = totalMatchGenos/totalCommonGenos
  )

# ILL-ONT ONLY
genoConsistency_samplesBlood_dsTests_opt100_summary_bySample_illONT.only <- genoConsistency_samplesBlood_dsTests_opt100_summary_bySample %>%
  filter(str_detect(set1, "ill") & str_detect(set2, "ont"))

# ILL-ONT SAME COVERAGE ONLY
genoConsistency_samplesBlood_dsTests_opt100_summary_bySample_sameCoverage.only <- genoConsistency_samplesBlood_dsTests_opt100_summary_bySample %>%
  mutate(compType = str_c(set1, set2, sep = ".")) %>%
  filter(compType %in% c("ill_ds0x.ont_ds0x",
                         "ill_ds6x.ont_ds6x",
                         "ill_ds10x.ont_ds10x",
                         "ill_ds15x.ont_ds15x",
                         "ill_ds20x.ont_ds20x",
                         "ill_ds25x.ont_ds25x",
                         "ill_ds30x.ont_ds30x",
                         "ill_ds50x.ont_ds50x",
                         "ill_ds100x.ont_ds100x")) %>%
  mutate(coverage = gsub("ill_", "", set1))
```

PLOTS

```{r}
genoConsistency_samplesBlood_dsTests_opt100_summary_bySample_sameCoverage.only %>%
  ggplot(aes(x = factor(coverage, c("ds0x", "ds6x", "ds10x", "ds15x", "ds20x", "ds25x", "ds30x", "ds50x", "ds100x")), y = propMatch)) +
  geom_boxplot() +
  labs(x = "coverage")
```

100x clearly looks best; so which samples/genos are giving mismatches?

```{r}
test <- genoCompare_illONT_gtscore_samplesBlood_lociDual_dsTests_optGenos %>%
  select(c(sampleID, locus, ill_ds100x, ont_ds100x)) %>%
  na.omit() %>%
  mutate(
    match = case_when(
      ill_ds100x == ont_ds100x ~ "yes",
      .default = "no"
    )
  )

length(unique(test$sampleID))

test2 <- test %>%
  filter(match == "no")

length(unique(test2$sampleID)) # 10
table(test2$sampleID)
table(test2$locus)

test3 <- test2 %>%
  filter(!sampleID %in% c("15_blood", "181_blood", "202_blood", "139_blood"))

table(test3$locus)
```

##### Plots

```{r}
# barplot
genoConsistency_dsTests_opt100_illONTOnly_summary %>%
  ggplot(aes(x = factor(set1, c("ill_ds6x", "ill_ds10x", "ill_ds15x", "ill_ds20x", "ill_ds25x", "ill_ds30x", "ill_ds50x", "ill_ds100x", "ill_ds0x")), y = propMatchGenos, fill = factor(set2, c("ont_ds6x", "ont_ds10x", "ont_ds15x", "ont_ds20x", "ont_ds25x", "ont_ds30x", "ont_ds50x", "ont_ds100x", "ont_ds0x")))) +
  geom_bar(position = "dodge", stat = "identity") +
  theme_linedraw() +
  guides(
    color = guide_legend(title = "ONT depth")
  ) +
  labs(
    title = "Geno consistency across coverage depths for Illumina vs. Nanopore",
    x = "Illumina depth",
    y = "Proportion matching genotypes"
    ) +
  scale_fill_discrete(name = "ONT depth")

# dot plot
genoConsistency_dsTests_opt100_illONTOnly_summary %>%
  ggplot(aes(x = totalCommonGenos, y = propMatchGenos, color = factor(set2, c("ont_ds6x", "ont_ds10x", "ont_ds15x", "ont_ds20x", "ont_ds25x", "ont_ds30x", "ont_ds50x", "ont_ds100x", "ont_ds0x")))) +
  geom_point() +
  facet_wrap(~factor(set1, c("ill_ds6x", "ill_ds10x", "ill_ds15x", "ill_ds20x", "ill_ds25x", "ill_ds30x", "ill_ds50x", "ill_ds100x", "ill_ds0x"))) +
#  scale_color_brewer(palette = "RdYlBu") +
#  scale_color_viridis(discrete = T) +
  scale_shape_discrete(name = "ONT depth") +
  theme_linedraw() +
  labs(
    title = "Geno consistency across coverage depths for 25 blood samples + 100 loci (up to 2500 genos) for Illumina vs. Nanopore",
    x = "Total common genos",
    y = "Proportion of matched genos (out of common)"
  ) +
  xlim(1000, 2500) +
  ylim(0.85, 1.0) +
  geom_vline(data = filter(genoConsistency_dsTests_opt100_illONTOnly_summary, set1=="ill_ds6x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds6x)))) +
  geom_vline(data = filter(genoConsistency_dsTests_opt100_illONTOnly_summary, set1=="ill_ds10x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds10x)))) +
  geom_vline(data = filter(genoConsistency_dsTests_opt100_illONTOnly_summary, set1=="ill_ds15x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds15x)))) +
  geom_vline(data = filter(genoConsistency_dsTests_opt100_illONTOnly_summary, set1=="ill_ds30x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds30x)))) +
  geom_vline(data = filter(genoConsistency_dsTests_opt100_illONTOnly_summary, set1=="ill_ds50x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds50x)))) +
  geom_vline(data = filter(genoConsistency_dsTests_opt100_illONTOnly_summary, set1=="ill_ds0x"), aes(xintercept = sum(!is.na(genoCompare_illONT_gtscore_samplesBlood_lociDual_opt100_dsTests_allCombos$ill_ds0x))))

# compare same coverage
genoConsistency_dsTests_opt100_illONT.sameDepthOnly_summary <- genoConsistency_dsTests_opt100_illONTOnly_summary %>%
  mutate(coverage = case_when(
    str_detect(set1, "ds6x") & str_detect(set2, "ds6x") ~ "6x",
    str_detect(set1, "ds10x") & str_detect(set2, "ds10x") ~ "10x",
    str_detect(set1, "ds15x") & str_detect(set2, "ds15x") ~ "15x",
    str_detect(set1, "ds20x") & str_detect(set2, "ds20x") ~ "20x",
    str_detect(set1, "ds25x") & str_detect(set2, "ds25x") ~ "25x",
    str_detect(set1, "ds30x") & str_detect(set2, "ds30x") ~ "30x",
    str_detect(set1, "ds50x") & str_detect(set2, "ds50x") ~ "50x",
    str_detect(set1, "ds100x") & str_detect(set2, "ds100x") ~ "100x"
  )) %>%
  na.omit() %>%
  as.data.frame()

genoConsistency_dsTests_opt100_illONT.sameDepthOnly_summary %>%
  ggplot(aes(x = coverage, y = propMatchGenos)) +
  geom_bar(stat = "identity")
```

#### opt90 genos

```{r}
# ILLUMINA
opt90_samplesLoci_ill_gtscore_blood_dsTests <- optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds6x, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds10x, 0.9)) %>%
  dplyr::rename("loci90_ill.ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds15x, 0.9)) %>%
  dplyr::rename("loci90_ill.ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds30x, 0.9)) %>%
  dplyr::rename("loci90_ill.ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual_ds50x, 0.9)) %>%
  dplyr::rename("loci90_ill.ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual, 0.9)) %>%
  dplyr::rename("loci90_ill.ds0x" = "loci")

# NANOPORE
opt90_samplesLoci_ont_gtscore_blood_dsTests <- optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds6x, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds10x, 0.9)) %>%
  dplyr::rename("loci90_ont.ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds15x, 0.9)) %>%
  dplyr::rename("loci90_ont.ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds30x, 0.9)) %>%
  dplyr::rename("loci90_ont.ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual_ds50x, 0.9)) %>%
  dplyr::rename("loci90_ont.ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual, 0.9)) %>%
  dplyr::rename("loci90_ont.ds0x" = "loci")
```

## 8.4 Fecal samples

### 8.4.1 Data

#### Original dataframes

```{r}
# ILLUMINA
genos_ill_gtscore_samplesFecal_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ill_gtscore_samplesFecal_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ill_gtscore_samplesFecal_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ill_gtscore_samplesFecal_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ill_gtscore_samplesFecal_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ill_gtscore_samplesFecal_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ill_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

# NANOPORE
genos_ont_gtscore_samplesFecal_lociDual_ds6x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds6x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ont_gtscore_samplesFecal_lociDual_ds10x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds10x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ont_gtscore_samplesFecal_lociDual_ds15x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds15x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ont_gtscore_samplesFecal_lociDual_ds30x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds30x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ont_gtscore_samplesFecal_lociDual_ds50x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds50x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))

genos_ont_gtscore_samplesFecal_lociDual_ds100x <- read.table("./downsamplingTests/tamRun4/01_run4GTscore/ds100x_ont_fullSet_polyGenResults_singleSNP.txt", header = T, check.names = F) %>%
  # filter to blood samples
  select(md_tamRun4_fecal$sampleID_unique) %>%
  # filter to dual-species loci
  rownames_to_column("locus") %>%
  filter(locus %in% primerList.v3_lociDual$locus) %>%
  column_to_rownames("locus") %>%
  # recode 0 as NA
  mutate(across(everything(), ~ as.character(.))) %>%
  mutate(across(everything(), ~ 
                  case_when(. == "0" ~ gsub("0", NA, .),
                            .default = .)
                ))
```

#### Long dataframes

```{r}
# ILLUMINA
genos_ill_gtscore_samplesFecal_lociDual_ds6x_l <- genos_ill_gtscore_samplesFecal_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds6x")

genos_ill_gtscore_samplesFecal_lociDual_ds10x_l <- genos_ill_gtscore_samplesFecal_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds10x")

genos_ill_gtscore_samplesFecal_lociDual_ds15x_l <- genos_ill_gtscore_samplesFecal_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds15x")

genos_ill_gtscore_samplesFecal_lociDual_ds30x_l <- genos_ill_gtscore_samplesFecal_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds30x")

genos_ill_gtscore_samplesFecal_lociDual_ds50x_l <- genos_ill_gtscore_samplesFecal_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds50x")

genos_ill_gtscore_samplesFecal_lociDual_ds100x_l <- genos_ill_gtscore_samplesFecal_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds100x")

genos_ill_gtscore_samplesFecal_lociDual_ds0x_l <- genos_ill_gtscore_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ill")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ill_ds0x")

# NANOPORE
genos_ont_gtscore_samplesFecal_lociDual_ds6x_l <- genos_ont_gtscore_samplesFecal_lociDual_ds6x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds6x")

genos_ont_gtscore_samplesFecal_lociDual_ds10x_l <- genos_ont_gtscore_samplesFecal_lociDual_ds10x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds10x")

genos_ont_gtscore_samplesFecal_lociDual_ds15x_l <- genos_ont_gtscore_samplesFecal_lociDual_ds15x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds15x")

genos_ont_gtscore_samplesFecal_lociDual_ds30x_l <- genos_ont_gtscore_samplesFecal_lociDual_ds30x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds30x")

genos_ont_gtscore_samplesFecal_lociDual_ds50x_l <- genos_ont_gtscore_samplesFecal_lociDual_ds50x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds50x")

genos_ont_gtscore_samplesFecal_lociDual_ds100x_l <- genos_ont_gtscore_samplesFecal_lociDual_ds100x %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds100x")

genos_ont_gtscore_samplesFecal_lociDual_ds0x_l <- genos_ont_gtscore_samplesFecal_lociDual %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun4[, c("sampleID_unique", "sampleID_ont")], by.x = "sampleID", by.y = "sampleID_ont") %>%
  relocate(sampleID_unique) %>%
  select(-sampleID) %>%
  column_to_rownames("sampleID_unique") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "ont_ds0x")
```

#### Combo dataframe

```{r}
genoCompare_illONT_gtscore_samplesFecal_lociDual_dsTests_allGenos <-
  # 0x
  genos_ill_gtscore_samplesFecal_lociDual_ds0x_l %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds0x_l, by = c("sampleID", "locus")) %>%
  # 6x
  merge(.,
        genos_ill_gtscore_samplesFecal_lociDual_ds6x_l, by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds6x_l, by = c("sampleID", "locus")) %>%
  # 10x
  merge(.,
        genos_ill_gtscore_samplesFecal_lociDual_ds10x_l, by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds10x_l, by = c("sampleID", "locus")) %>%
  # 15x
  merge(.,
        genos_ill_gtscore_samplesFecal_lociDual_ds15x_l, by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds15x_l, by = c("sampleID", "locus")) %>%
  # 30x
  merge(.,
        genos_ill_gtscore_samplesFecal_lociDual_ds30x_l, by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds30x_l, by = c("sampleID", "locus")) %>%
  # 50x
  merge(.,
        genos_ill_gtscore_samplesFecal_lociDual_ds50x_l, by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds50x_l, by = c("sampleID", "locus")) %>%
  # 100x
  merge(.,
        genos_ill_gtscore_samplesFecal_lociDual_ds100x_l, by = c("sampleID", "locus")) %>%
  merge(.,
        genos_ont_gtscore_samplesFecal_lociDual_ds100x_l, by = c("sampleID", "locus"))
```

#### TEST

```{r}
test <- genoCompare_illONT_gtscore_samplesFecal_lociDual_dsTests_allGenos %>%
  select(c(sampleID, locus, ill_ds100x, ont_ds100x)) %>%
  na.omit() %>%
  mutate(
    match = case_when(
      ill_ds100x == ont_ds100x ~ "true",
      .default = "false"
    )
  )

length(unique(test$sampleID)) # 26

test2 <- test %>%
  filter(match == "false")

table(test2$sampleID) # 6 samples, 7_fecal has 22, 91_fecal has 4

test3 <- test %>%
  filter(!sampleID %in% test2$sampleID)
```

### 8.4.2 Geno success

```{r}
genoSuccess_bySample_illONT_gtscore_samplesFecal_lociDual_dsTests <- 
  genoCompare_illONT_gtscore_samplesFecal_lociDual_dsTests_allGenos %>%
  select(-locus) %>%
  group_by(sampleID) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  as.data.frame() %>%
#  column_to_rownames("sampleID") %>%
#  mutate_all(list(~ ./140))
  pivot_longer(!sampleID,
               names_to = "depth",
               values_to = "totalGenos") %>%
  mutate(
    propSuccess = totalGenos/140
  ) %>%
  separate(., depth, into = c("seqPlatform", "depth"), sep = "_") %>%
  mutate(
    seqPlatform = case_when(
      seqPlatform == "ill" ~ "Illumina",
      seqPlatform == "ont" ~ "Nanopore"
    )
  )
```

Plots

```{r}
genoSuccess_bySample_illONT_gtscore_samplesFecal_lociDual_dsTests %>%
  ggplot(aes(x = factor(depth, c("ds6x", "ds10x", "ds15x", "ds30x", "ds50x", "ds100x", "ds0x")), y = propSuccess, fill = seqPlatform)) +
  geom_boxplot() +
  theme_linedraw() +
  labs(
    title = "Fecal sample genotype success across coverage depths for Illumina vs. Nanopore",
    x = "Downsampled sequencing depth per sample per locus",
    y = "Proportion dual-species loci genotyped per sample")
```

### 8.4.3 Geno consistency

#### All genos

##### Run comparison

```{r}
df <- genoCompare_illONT_gtscore_samplesFecal_lociDual_dsTests_allGenos

cols_of_interest <- setdiff(names(df), c("sampleID", "locus"))

colCombos <- combn(cols_of_interest, 2)

check_dsTests <- df[,colCombos[1,]] > 0 & df[,colCombos[2,]] > 0 & df[,colCombos[1,]] == df[,colCombos[2,]]
colnames(check_dsTests) = paste0(colCombos[1,], ".", colCombos[2,])

genoCompare_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_allCombos <- cbind(df, ifelse(check_dsTests == T, 1, 0))
```

##### Summary

```{r}
# all comparisons
genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_summary <- genoCompare_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_allCombos %>%
  summarise(across(where(is.numeric),
                   list(totalCommonGenos = ~ sum(!is.na(.x)),
                        totalMatchGenos = ~ sum(.x, na.rm = T)))) %>%
  as.data.frame() %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("combo") %>%
  separate(., combo, into = c("combo", "metric"), sep = "_t") %>%
  mutate(metric = str_c("t", metric)) %>%
  pivot_wider(names_from = metric, values_from = V1) %>%
  separate(., combo, into = c("set1", "set2"), sep = "\\.") %>%
  mutate(
    propCommonGenos = round(totalCommonGenos/(30*140), 2),
    propMatchGenos = round(totalMatchGenos/totalCommonGenos, 4)
  ) %>%
  mutate(
    compType = str_c(set1, set2, sep = ".")
  )

# ill vs. ont only
genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_illONTonly_summary <- genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_summary %>%
  filter(str_detect(compType, "ill") & str_detect(compType, "ont")) %>%
  select(-compType)

# ill vs. ont same coverage only
genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_sameCoverageOnly_summary <- genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_summary %>%
  filter(compType %in% c("ill_ds0x.ont_ds0x",
                         "ill_ds6x.ont_ds6x",
                         "ill_ds10x.ont_ds10x",
                         "ill_ds15x.ont_ds15x",
                         "ill_ds30x.ont_ds30x",
                         "ill_ds50x.ont_ds50x",
                         "ill_ds100x.ont_ds100x"))
```

##### Plots

```{r}
# all genos
genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_illONTonly_summary %>%
  ggplot(aes(x = factor(set1, c("ill_ds6x", "ill_ds10x", "ill_ds15x", "ill_ds30x", "ill_ds50x", "ill_ds100x", "ill_ds0x")), y = propMatchGenos, fill = factor(set2, c("ont_ds6x", "ont_ds10x", "ont_ds15x", "ont_ds30x", "ont_ds50x", "ont_ds100x", "ont_ds0x")))) +
  geom_bar(position = "dodge", stat = "identity") +
  theme_linedraw() +
  guides(
    color = guide_legend(title = "ONT depth")
  ) +
  labs(
    title = "Geno consistency across coverage depths for Illumina vs. Nanopore",
    x = "Illumina depth",
    y = "Proportion matching genotypes"
    ) +
  scale_fill_discrete(name = "ONT depth")

# same coverage only
genoConsistency_illONT_gtscore_samplesFecal_lociDual_allGenos_dsTests_sameCoverageOnly_summary %>%
  ggplot(aes(x = compType, y = propMatchGenos)) +
  geom_bar(stat = "identity")
```

#### opt100 genos

##### Data

```{r}
# ILLUMINA
opt100_samplesLoci_ill_gtscore_fecal_dsTests <- optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual_ds6x, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual_ds10x, 1)) %>%
  dplyr::rename("loci100_ill.ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual_ds15x, 1)) %>%
  dplyr::rename("loci100_ill.ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual_ds30x, 1)) %>%
  dplyr::rename("loci100_ill.ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual_ds50x, 1)) %>%
  dplyr::rename("loci100_ill.ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ill_gtscore_samplesFecal_lociDual, 1)) %>%
  dplyr::rename("loci100_ill.ds0x" = "loci")

# NANOPORE
opt100_samplesLoci_ont_gtscore_blood_dsTests <- optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual_ds6x, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.ds6x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual_ds10x, 1)) %>%
  dplyr::rename("loci100_ont.ds10x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual_ds15x, 1)) %>%
  dplyr::rename("loci100_ont.ds15x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual_ds30x, 1)) %>%
  dplyr::rename("loci100_ont.ds30x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual_ds50x, 1)) %>%
  dplyr::rename("loci100_ont.ds50x" = "loci") %>%
  merge(.,
        optimize_samplesLoci(genos_ont_gtscore_samplesFecal_lociDual, 1)) %>%
  dplyr::rename("loci100_ont.ds0x" = "loci")
```

```{=html}
<script src="https://hypothes.is/embed.js" async></script>
```
```{r, echo=FALSE}
knitr::knit_exit()
```

```{r}
library(trackdown, quietly = T)

trackdown::update_file(file = "/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/paper1_nanoporeValidation/paper1_nanoporeValidation.Rmd", hide_code = F)
```

# RESOURCES

for optimizing samples/loci to include: <https://stackoverflow.com/questions/24016612/subset-of-data-frame-columns-to-maximize-complete-observations>

alternative approach: <https://stackoverflow.com/questions/42858959/best-possible-subset-of-data-frame-without-na>

hypothes.is for html editing <https://markdly.github.io/reviewr/index.html>

The following line from <https://github.com/hypothesis/client> adds commenting functionality:

```{=html}
<script src="https://hypothes.is/embed.js" async></script>
```
# Scraps

The comparisons above are looking at all common genotypes in Illumina vs. each of the filtered Nanopore runs. We can also look to see how the proportion of matches change for genotypes that are called in all filtered Nanopore sequencing sets.

Like before though, we see the same pattern - the proportion of matching genotypes between Illumina and Nanopore results DECREASES with higher quality filters.

```{r, eval=FALSE, echo=FALSE}
genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent <- genoCompare_qcFilters %>%
  filter(str_detect(sampleID_unique, "blood")) %>%
  select(-ill_qc30) %>%
  na.omit()

genoCompare_blood_illqc0.ontAllFilters.allPresent_qcFilters_summary <- data.frame(
  illONT_qcCompare = c("qc0.qc0", "qc0.qc11", "qc0.qc13", "qc0.qc15", "qc0.qc20"),
  readCounts_ill = rep(sum(readCounts_ill_gtscore_blood_sampleSum$qc0)),
  readCounts_ont = c(
    sum(readCounts_ont_gtscore_blood_sampleSum$qc0),
    sum(readCounts_ont_gtscore_blood_sampleSum$qc11),
    sum(readCounts_ont_gtscore_blood_sampleSum$qc13),
    sum(readCounts_ont_gtscore_blood_sampleSum$qc15),
    sum(readCounts_ont_gtscore_blood_sampleSum$qc20)
  ),
  totalGenos_ill = rep(sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0)), 5),
  totalGenos_ont = c(
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc0)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc11)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc13)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc15)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc20))
    ),
  totalCommon = c(
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0) & !is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc0)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0) & !is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc11)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0) & !is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc13)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0) & !is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc15)),
    sum(!is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0) & !is.na(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc20))
  ),
  totalMatch = c(
    sum(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0 == genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc0, na.rm = T),
    sum(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0 == genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc11, na.rm = T),
    sum(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0 == genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc13, na.rm = T),
    sum(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0 == genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc15, na.rm = T),
    sum(genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ill_qc0 == genoCompare_blood_qcFilters_illqc0.ontAllFilters.allPresent$ont_qc20, na.rm = T)
  )
) %>%
  mutate(
    propMatch = round(totalMatch/totalCommon, 2)
  )
```

```{r, echo=FALSE}
# gtscore
## loci genotyped at 100% of samples among subsets of 2 to 30 samples
opt100_samplesLoci_ill_gtscore_blood <- optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.gtscore" = "loci")

opt100_samplesLoci_ont_gtscore_blood <- optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.gtscore" = "loci")

## loci genotyped at 90% of samples among subsets of 2 to 30 samples
opt90_samplesLoci_ill_gtscore_blood <- optimize_samplesLoci(genos_ill_gtscore_samplesBlood_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.gtscore" = "loci")

opt90_samplesLoci_ont_gtscore_blood <- optimize_samplesLoci(genos_ont_gtscore_samplesBlood_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.gtscore" = "loci")

# gtseq
## loci genotyped at 90% of samples among subsets of 2 to 30 samples
opt100_samplesLoci_ill_gtseq_blood <- optimize_samplesLoci(genos_ill_gtseq_samplesBlood_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ill.gtseq" = "loci")

opt100_samplesLoci_ont_gtseq_blood <- optimize_samplesLoci(genos_ont_gtseq_samplesBlood_lociDual, 1) %>%
  as.data.frame() %>%
  dplyr::rename("loci100_ont.gtseq" = "loci")

## loci genotyped at 90% of samples among subsets of 2 to 30 samples
opt90_samplesLoci_ill_gtseq_blood <- optimize_samplesLoci(genos_ill_gtseq_samplesBlood_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ill.gtseq" = "loci")

opt90_samplesLoci_ont_gtseq_blood <- optimize_samplesLoci(genos_ont_gtseq_samplesBlood_lociDual, 0.9) %>%
  as.data.frame() %>%
  dplyr::rename("loci90_ont.gtseq" = "loci")
```
