---
title: "GTscorePipeline_COMBINED"
output: html_document
date: '2022-06-07'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,fig.path = "tools/")
```

```{r setwd & load packages, include=FALSE}
library(tidyverse)
library(phylotools)
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("Biostrings")
source("GTscore.R") # NOTE- added N=ATGC to script for later analyses (not included in original)
```

GTscore also requires Perl as well as the the packages "Algorithm::Combinatorics" and "Excel-Writer-XLSX". Perform the following in terminal to install on Linux. NOTE -- be sure to install both perl and the modules in root.
how to install perl packages: http://www.cpan.org/modules/INSTALL.html
```{bash}
# Install perl
conda install -c conda-forge perl

# Install cpanminus (makes installing perl modules easier)
cpan App::cpanminus
sudo cpanm Algorithm::Combinatorics
```

## About the pipeline
This pipeline includes all steps needed to analyze MiSeq data output, including how to interleave the fasta sequences and create the necessary files for analysis in GTscore. The GTscore part of the pipeline is based on that created by Garrett McKinney (see the associated github here), and combines scripts included in his original "GTscore_pipeline.R" as well as his original "README.md" - there's a good bit of overlap between these files, but a few scripts exist in one but not the other.

In this pipeline, I've included separate analyses of shared loci and species-specific loci, where analyses of species-specific loci only include samples and amplicon reference sequences from the respective species. I've also included a separate analysis of the 76 loci for which bp3 incorrectly designed primers.

## About the data
The sequence data used in this pipeline is from tamarinGenetics optimization run #1.

## 0) Interleave fasta sequences
Prior to running this pipeline, we need to combine reads 1 & 2 for each sample into a single interleaved file (run in terminal):

**Extract fastq.gz files**

```{bash Extract sequencing results, eval = FALSE}
# Navigate to the directory containing the .fastq.gz files from the sequencing run
cd 00_run1Seqs

# Move all .gz files out of their individual folders and into the current directory & remove empty folders
mv */* ./
find -empty -delete

# Extract all files
gunzip -r .
```

```{bash Interleave fasta seqs}
cd ./../

# Interleave reads 1 & 2 for each sample and place in a new folder
for i in ./00_run1Seqs/*_R1_001.fastq; do name=$(basename $i _R1_001.fastq); reformat.sh in=00_run1Seqs/${name}_R1_001.fastq in2=00_run1Seqs/${name}_R2_001.fastq out=02_run1Interleaved/${name}_interleaved.fastq; done

# Trim filenames to make them more manageable
cd ./02_run1Interleaved/

for f in *.fastq; do g="${f%%_S*}.fastq"; mv "${f}" "${g}"; done
```


AND/OR - create interleaved files from the filtered reads:
```{bash}
# Interleave reads 1 & 2 for each sample and place in a new folder
for i in ./01_optSeqs_filtered/*_filtered_R1.fastq.gz; do name=$(basename $i _filtered_R1.fastq.gz); reformat.sh in=01_optSeqs_filtered/${name}_filtered_R1.fastq.gz in2=01_optSeqs_filtered/${name}_filtered_R2.fastq.gz out=02_optInterleaved_filtered/${name}_interleaved_filtered.fastq.gz; done

# cd to 02_optInterleaved_filtered and extract all .gz files
cd 02_optInterleaved_filtered
gunzip -r .
```

## 1) Create primer-probe & sample files
We also need to make two files to input into the GTscore pipeline:

File #1: "sampleFiles.txt" contains all the names of the interleaved fasta files created above (make this by hand; use terminal to list all file names and create a new txt file with that then create separate LWED and SIMP sample files). 

File #2: "primerProbeFile.txt" (with separate files for LWED- and SIMP-only loci) contains:
1. Locus - locus name
2. Ploidy - in our case, all are diploid so ploidy = 2
3. SNPpos - position of SNP in the amplicon, assuming that the first base = 0
4. Allele1 - one of the options for the SNP
5. Allele2 - the other option for the SNP
6. Probe1 - an 8 nt sequence overlapping the SNP; contains Allele1
7. Probe2 - same as Probe1, but contains Allele2
8. Primer - the forward primer sequence for each locus

The files needed to create the primer-probe file (plus their respective scripts, such as to obtain SNP position & references amplicon sequences) were created by Sam, with edits where needed by RV. The script for creating the final primerProbeFile.txt files is below:

```{r Create primer-probe files}
# Load files
probes226 <- read.csv("assay_info_probes_final_rvEdits.csv", header = F, sep = ",") %>% 
  dplyr::rename(c(name=V1, primer=V2, probe1=V3, probe2=V4)) #locus name, probe1&2, & primer for the 226 loci where bp3 successfully recognized SNP [] and/or are informative; RV edited to include brackets around every IUB (some bracket pairs had multiple IUB codes inside)
probeInfo <- read.csv("probe_info_ALL.csv") #locus name, allele1&2, probe1&2 for ALL loci
snpPos <- read.csv("pos_amplicon.csv", sep = ",", colClasses = "character") #SNP position for the 226 loci; read in all as character so as not to strip leading zeros in snp position
nameTranslation <- read.csv("primers_seqName_to_shortName.csv") #file to translate sequence names to shorter bp3 names (for easier readability)

# Add allele1 & allele2 info from probeInfo to probes226
pp1 <- merge(probes226, probeInfo, by = "name", all.x = T, all.y = F) %>%
  select(c("name", "allele1", "allele2", "probe1.x", "probe2.x", "primer"))

# Add SNP position from snpPos
pp2 <- merge(pp1, snpPos, by.x = "name", by.y = "SNP_name") %>%
  select(c("name", "allele1", "allele2", "probe1.x", "probe2.x", "primer", "pos_in_amplicon")) %>%
  dplyr::rename(c(Allele1=allele1, Allele2=allele2, Probe1=probe1.x, Probe2=probe2.x,
           Primer=primer, SNPpos = pos_in_amplicon))
View(pp2)

# Change locus name to short name, add ploidy column
pp3 <- merge(pp2, nameTranslation, by.x = "name", by.y = "seqID1") %>%
  select("primerName2", "Allele1", "Allele2", "Probe1", "Probe2", "SNPpos", "Primer") %>%
  dplyr::rename(Locus=primerName2) %>%
  mutate(Ploidy=2, .after=Locus) %>%
  relocate(SNPpos, .after=Ploidy)
View(pp3)
length(pp3$Locus) #226; good!

# Replace IUPAC codes with bases
pp4 <- pp3 %>%
  mutate_at(vars(Probe1, Probe2, Primer), ~ str_replace_all(., c("R"="AG",
                                                                 "Y"="CT",
                                                                 "S"="GC",
                                                                 "W"="AT",
                                                                 "K"="GT",
                                                                 "M"="AC",
                                                                 "N"="ATGC")))
View(pp4)

# Create separate LWED and SIMP primer-probe files
ppLWED <- pp4 %>%
  filter(!str_detect(Locus, "SIMP"))
View(ppLWED)

ppSIMP <- pp4 %>%
  filter(!str_detect(Locus, "LWED"))
View(ppSIMP)

# Export
write.table(pp4, "primerProbeFile.txt", sep = "\t", row.names = F, quote = F)
write.table(ppLWED, "primerProbeFile_LWED.txt", sep = "\t", row.names = F, quote = F)
write.table(ppSIMP, "primerProbeFile_SIMP.txt", sep = "\t", row.names = F, quote = F)
```

## 2) Count reads for amplicons
Once the files are created, we can proceed with the second step in the GTscore pipeline (first step is demultiplexing, but MiSeq did that for us automatically) - counting amplicon reads. We'll be doing three runs: 1) all loci + all samples, 2) LWED samples for all but SIMP-specific loci, and 3) SIMP samples for all but LWED-specific loci. 

The read counter is written in perl (AmpliconReadCounter.pl), but can be called from R. Running it does the following:

1. Identifies each unique sequence, then counts the number of times each unique sequence occurs within an individual

2. Aligns each unique sequence with primer and probe; if the sequence doesn’t align, then it is excluded as an off-target sequence and reports by individual and by locus are given.

* Note:  By default, all primers are trimmed to the length of the shortest primer to increase speed.  Optionally the full length primer can be used for the primer but this may significantly increase run timing depending on variation in primer lengths across loci.

Input flags for this script are:
* –p       a tab delimited file containing primer/probe information for each locus
* --files  a text file containing a list of .fastq sequence files to count reads from. 

Optional flags:
* --prefix          optional prefix for output file names
* --inDir           option to specify directory containing sequence data
* --inputType       fq or fastqgz (defaults to fastqgz)
* --useFullPrimer	  uses the full primer for counting reads rather than the trimmed primer
* --alleleOrder     order of alleles output in locusTable file. Options are original (matches primer-probe file order) or alphabetical (default) 
* --printMatched    outputs matched reads for each individual
* --printDiscarded  outputs discarded reads for each individual

#### Setting up directories and running AmpliconReadCounterl.pl
To run from the command line, use:
  perl AmpliconReadCounter.pl -p primerProbeFile.txt --files sampleList.txt

To run from R, use:
```{r count reads for amplicons, eval=FALSE}
# Unfiltered reads
## All
system2("perl",
        args="AmpliconReadCounter.pl -p primerProbeFile.txt --files sampleFiles.txt")

## LWED
system2("perl",
        args="AmpliconReadCounter.pl -p primerProbeFile_LWED.txt --files sampleFiles_LWED.txt --prefix LWED_")

## SIMP
system2("perl",
        args="AmpliconReadCounter.pl -p primerProbeFile_SIMP.txt --files sampleFiles_SIMP.txt --prefix SIMP_")
```

AmpliconReadCounter.pl outputs a LocusTable file and an AlleleReads file for single-SNP and haplotype data, plus two summary files.

The default names for these files are as follows, with _LWED and _SIMP appended to files associated with species-specific analyses:

* LocusTable_singleSNPs.txt - locus name, ploidy, and alleles for each SNP
* AlleleReads_singleSNPs.txt - counts for each SNP allele (rows are loci, columns are individuals)
* LocusTable_haplotypes.txt - same as above (doesn't matter for us since we're only using single SNPs)
* AlleleReads_haplotypes.txt - same as above (doesn't matter for us since we're only using single SNPs)
* GTscore_individualSummary.txt - counts by individual of total reads, off-target reads, primer-only reads, and primer probe reads
* GTscore_locusSummary.txt - counts by locus of primer reads and primer probe reads

## 3) Genotyping
Genotyping is accomplished using the polyGen function. The genotyping algorithm is described in McKinney et al. 2018 and is a maximum likelihood algorithm capable of genotyping any number of alleles and ploidy per locus. This allows genoyping of single SNPs as well as microhaplotypes, and loci with elevated ploidy.

Two arguments are required for polyGen, the locusTable and alleleReads files output by AmpliconReadCounter.

Optional arguments for polyGen are:

* p_thresh  - threshold p-value for likelihood ratio test (default 0.05)
* epsilon   - error rate for genotyping model (default 0.01)

**NOTE** that only primer probe reads are used in genotyping!

```{r Genotyping}
#load locus table and allele reads file
singleSNP_locusTable<-read.delim("LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)
singleSNP_alleleReads<-read.delim("AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE)

head(singleSNP_locusTable)
singleSNP_alleleReads[1:5,1:5]

singleSNP_locusTable_LWED<-read.delim("LWED_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)
singleSNP_alleleReads_LWED<-read.delim("LWED_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE)

head(singleSNP_locusTable_LWED)
singleSNP_alleleReads_LWED[1:5,1:5]

singleSNP_locusTable_SIMP<-read.delim("SIMP_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)
singleSNP_alleleReads_SIMP<-read.delim("SIMP_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE)

head(singleSNP_locusTable_SIMP)
singleSNP_alleleReads_SIMP[1:5,1:5]

#generate singleSNP genotypes using the polyGen algorithm
polyGenResults_singleSNP<-polyGen(singleSNP_locusTable,singleSNP_alleleReads)
polyGenResults_singleSNP_LWED<-polyGen(singleSNP_locusTable_LWED,singleSNP_alleleReads_LWED)
polyGenResults_singleSNP_SIMP<-polyGen(singleSNP_locusTable_SIMP,singleSNP_alleleReads_SIMP)

#look at first five rows and columns
polyGenResults_singleSNP[1:5,1:5]
polyGenResults_singleSNP_LWED[1:5,1:5]
polyGenResults_singleSNP_SIMP[1:5,1:5]

#write results
write.table(polyGenResults_singleSNP,"polyGenResults_singleSNP.txt",quote=FALSE,sep="\t")
write.table(polyGenResults_singleSNP_LWED,"polyGenResults_singleSNP_LWED.txt",quote=FALSE,sep="\t")
write.table(polyGenResults_singleSNP_SIMP,"polyGenResults_singleSNP_SIMP.txt",quote=FALSE,sep="\t")
```

## 4) Data summaries
### 4.1) Locus summaries
#### Summarize single SNP results for loci

The summarizeGTscore command generates summary data for each locus in table form.  The summary data includes genotype rate, average read depth, minor (least frequent) allele frequency, major (most frequent) allele frequency, alleles per locus, and frequency per allele.  Minor allele frequency is a common metric for filtering loci that are likely to be uninformative for population genetics; however, loci with haplotype alleles may have an allele with very low frequency but still have appreciable frequency at multiple other alleles. Because of this, the major allele frequency is included in output, as well as the observed frequencies for all alleles at a given locus.

```{r}
#summarize single SNP results
singleSNP_summary<-summarizeGTscore(singleSNP_alleleReads, singleSNP_locusTable, polyGenResults_singleSNP)
singleSNP_summary_LWED<-summarizeGTscore(singleSNP_alleleReads_LWED, singleSNP_locusTable_LWED, polyGenResults_singleSNP_LWED)
singleSNP_summary_SIMP<-summarizeGTscore(singleSNP_alleleReads_SIMP, singleSNP_locusTable_SIMP, polyGenResults_singleSNP_SIMP)
#view results
View(singleSNP_summary)
View(singleSNP_summary_LWED)
View(singleSNP_summary_SIMP)
#write results
write.table(singleSNP_summary,"singleSNP_summary.txt",quote=FALSE,sep="\t",row.names=FALSE)
write.table(singleSNP_summary_LWED,"singleSNP_summary_LWED.txt",quote=FALSE,sep="\t",row.names=FALSE)
write.table(singleSNP_summary_SIMP,"singleSNP_summary_SIMP.txt",quote=FALSE,sep="\t",row.names=FALSE)
```

May also be helpful to combine single SNP summaries with the GTscore locus summary created earlier:
```{r}
# Load data
ls <- read.table("GTscore_locusSummary.txt", header = T, sep = "\t")
ls_lwed <- read.table("LWED_GTscore_locusSummary.txt", header = T, sep = "\t")
ls_simp <- read.table("SIMP_GTscore_locusSummary.txt", header = T, sep = "\t")

# Separate summaries into results from shared and species-specific loci analyses
ls1 <- ls %>%
  filter(!str_detect(Locus, "LWED|SIMP"))
ls_lwed1 <- ls_lwed %>%
  filter(str_detect(Locus, "LWED"))
ls_simp1 <- ls_simp %>%
  filter(str_detect(Locus, "SIMP"))

ss <- singleSNP_summary %>%
  filter(!str_detect(Locus_ID, "LWED|SIMP"))
ss_lwed <- singleSNP_summary_LWED %>%
  filter(str_detect(Locus_ID, "LWED"))
ss_simp <- singleSNP_summary_SIMP %>%
  filter(str_detect(Locus_ID, "SIMP"))

# Recombine gtscore & single snp summary files, then merge the two
ls_recombine <- rbind(ls1, ls_lwed1, ls_simp1)
View(ls_recombine)
ss_recombine <- rbind(ss, ss_lwed, ss_simp)
ss_recombine$Locus_ID <- substr(ss_recombine$Locus_ID,0,nchar(ss_recombine$Locus_ID)-4) # change locus names to match ls
View(ss_recombine)

ls_ss <- merge(ls_recombine, ss_recombine, by.x = "Locus", by.y = "Locus_ID") 
View(ls_ss)

# Export
write.csv(ls_ss, "./summaryFiles/locusSummary_combined.csv", row.names = F)
```


#### Generate plots for single SNP results
NOTE - plots below are for full dataset only (not LWED and SIMP subsets)

plot genotype rate
```{r locus genotype rate, warning=FALSE}
ggplot()+geom_histogram(data=singleSNP_summary,aes(x=GenotypeRate),binwidth=0.03)+xlim(-0.01,1.01)+
  labs(title="Locus Genotype Rate", x="Genotype Rate", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot average read depth for single SNP data
```{r locus read depth}
ggplot()+geom_histogram(data=singleSNP_summary,aes(x=AvgReadDepth),binwidth=20)+
  labs(title="Average Read Depth per SNP", x="Average Read Depth", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot genotype rate relative to average depth
```{r locus genotype rate vs read depth}
ggplot()+geom_point(data=singleSNP_summary,aes(x=AvgReadDepth,y=GenotypeRate))+ylim(0,1)+
  labs(title="Genotype Rate vs Average Depth per SNP", x="Average Depth", y="Genotype Rate")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot distribution of minor allele frequency
```{r histogram of MAF, warning=FALSE}
ggplot()+geom_histogram(data=singleSNP_summary,aes(x=minAF),binwidth=0.01)+
  labs(title="Minor Allele Frequency Single SNP", x="Minor Allele Frequency", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot distribution of major allele frequency
```{r histogram of MajAF, warning=FALSE}
ggplot()+geom_histogram(data=singleSNP_summary,aes(x=majAF),binwidth=0.01)+
  labs(title="Major Allele Frequency Single SNP", x="Major Allele Frequency", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

### 4.2) Sample summaries
load sample summary from AmpliconReadCounter
```{r}
GTscore_individualSummary<-read.delim("GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
head(GTscore_individualSummary)

GTscore_individualSummary_LWED<-read.delim("LWED_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
head(GTscore_individualSummary_LWED)

GTscore_individualSummary_SIMP<-read.delim("SIMP_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
head(GTscore_individualSummary_SIMP)
```

summarize single SNP results for samples
```{r}
singleSNP_sampleSummary<-summarizeSamples(polyGenResults_singleSNP,singleSNP_alleleReads)
head(singleSNP_sampleSummary)

singleSNP_sampleSummary_LWED<-summarizeSamples(polyGenResults_singleSNP_LWED,singleSNP_alleleReads_LWED)
head(singleSNP_sampleSummary_LWED)

singleSNP_sampleSummary_SIMP<-summarizeSamples(polyGenResults_singleSNP_SIMP,singleSNP_alleleReads_SIMP)
head(singleSNP_sampleSummary_SIMP)
```

combine AmpliconReadCounter individual summary data with GTscore sample summary
```{r}
# make new copies
gt_indivSummary <- GTscore_individualSummary
gt_indivSummary_LWED <- GTscore_individualSummary_LWED
gt_indivSummary_SIMP <- GTscore_individualSummary_SIMP

#first adjust names from sample_genotypeRate_singleSNP to match GTscore_individualSummary (some of the GTscore functions change - to .)
gt_indivSummary$Sample <- sub('tamOpt-', 'tamOpt.', gt_indivSummary$Sample)
gt_indivSummary<-merge(gt_indivSummary,singleSNP_sampleSummary,by.x="Sample",by.y="sample")

gt_indivSummary_LWED$Sample <- sub('tamOpt-', 'tamOpt.', gt_indivSummary_LWED$Sample)
gt_indivSummary_LWED<-merge(gt_indivSummary_LWED,singleSNP_sampleSummary_LWED,by.x="Sample",by.y="sample")

gt_indivSummary_SIMP$Sample <- sub('tamOpt-', 'tamOpt.', gt_indivSummary_SIMP$Sample)
gt_indivSummary_SIMP<-merge(gt_indivSummary_SIMP,singleSNP_sampleSummary_SIMP,by.x="Sample",by.y="sample")

# Combine LWED & SIMP to form one
gt_indivSummary_combined <- rbind(gt_indivSummary_LWED, gt_indivSummary_SIMP)
View(gt_indivSummary_combined)

# Export
write.csv(gt_indivSummary, "./summaryFiles/individualSummary_FULL.csv", row.names = F)
write.csv(gt_indivSummary_LWED, "./summaryFiles/individualSummary_LWED_FULL.csv", row.names = F)
write.csv(gt_indivSummary_SIMP, "./summaryFiles/individualSummary_SIMP_FULL.csv", row.names = F)
write.csv(gt_indivSummary_combined, "./summaryFiles/individualSummary_combined_FULL.csv", row.names = F)
```

#### plots

plot histogram of genotype rate
```{r histogram of genotype rate, warning=FALSE}
ggplot()+geom_histogram(data=GTscore_individualSummary,aes(x=GenotypeRate),binwidth=0.03)+xlim(-0.01,1.01)+
  labs(title="Sample Genotype Rate", x="Genotype Rate", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot histogram of heterozygosity
```{r histogram of heterozygosity, warning=FALSE}
ggplot()+geom_histogram(data=GTscore_individualSummary,aes(x=Heterozygosity),binwidth=0.03)+xlim(-0.01,1.01)+
  labs(title="Sample Heterozygosity", x="Heterozygosity", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot genotype rate vs primer probe reads
```{r genotype rate vs primer probe reads}
#dashed line added at 90% genotype rate, this is not a strict threshold, just a goal to aim for
ggplot()+geom_point(data=GTscore_individualSummary,aes(x=Primer.Probe.Reads,y=GenotypeRate))+
  labs(title="Genotype Rate vs Total Reads per Sample", x="Primer Probe Reads", y="Genotype Rate")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))+
  geom_hline(yintercept=0.9,lty="dashed")
```


Samples with unusually high heterozygosity may be contaminated or have elevated ploidy. [[Given chimerism, however, this isn't super helpful for blood samples]]

plot heterozygosity vs primer probe reads
```{r heterozygosity vs primer probe reads}
ggplot()+geom_point(data=GTscore_individualSummary,aes(x=Primer.Probe.Reads,y=Heterozygosity))+
  labs(title="Heterozygosity vs Total Reads per Sample", x="Primer Probe Reads", y="Heterozygosity")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))+
  geom_hline(yintercept=0.3, lty="dashed")
```

## 5) Quality control
#### Identify duplicate samples

The IDduplicateSamples function does all pairwise comparisons of samples and outputs two metrics:

* proportionCommon - The proportion of loci that had genotypes for both samples in a pair 
* proportionMatch  - The proportion of loci that have identical genotypes in the sample pair

```{r identify duplicate samples,eval=FALSE}
#convert missing genotypes "0" to NA
polyGenResults_singleSNP_NA<-polyGenResults_singleSNP
polyGenResults_singleSNP_NA[polyGenResults_singleSNP_NA=="0"]<-NA

#compare all samples, for each comparison get proportion of loci that have genotypes
#in both samples (proportionCommon) and proportion of shared loci that have identical
#genotypes (proportionMatch).
#Samples with a high proportionCommon and high proportionMatch are likely duplicates
polyGenResults_dupTest<-IDduplicateSamples(polyGenResults_singleSNP_NA)

write.table(polyGenResults_dupTest,"polyGenResults_dupTest.txt",quote=FALSE,sep="\t",row.names=FALSE)

```

```{r duplicate sample results}
polyGenResults_dupTest<-read.delim("polyGenResults_dupTest.txt",stringsAsFactors=FALSE)
head(polyGenResults_dupTest)
```
polyGenResults_dupTest

Plot results of IDduplicateSamples

McKinney says that he generally sets thresholds of proportionMatch > 0.8 and proportionCommon > 0.75 to identify duplicate sample pairs, but the appropriate thresholds will depend on the marker set being used and the relatedness among samples. In this plot, he notes that it is possible that individuals at the lower range of identical genotypes (~0.85) are relatives rather than duplicated samples.

```{r plot IDduplicateSamples results}
#potential duplicates are proportionMatch>0.8 and proportionCommon>0.75
#feel free to adjust these thresholds as needed
matchThresh=0.8
commonThresh=0 #switched to 0, looks like some loci were successfully genotyped for one sample and some for the other

#plot results
ggplot()+geom_point(data=polyGenResults_dupTest,aes(x=proportionCommon,y=proportionMatch))+
  geom_segment(aes(x=commonThresh,xend=1,y=matchThresh,yend=matchThresh),lty="dashed")+
  geom_segment(aes(x=commonThresh,xend=commonThresh,y=matchThresh,yend=1),lty="dashed")
```

Identify duplicate sample pairs with thresholds set previously
```{r}
#filter to potentially duplicated samples using thresholds above
polyGenResults_dupTest %>% filter(proportionMatch>=matchThresh,proportionCommon>=commonThresh)

```

1-29 YES
2-30 YES
3-31 YES
5-27 YES
5-33 YES
6-17 NO (but in same group)
6-28 YES
6-34 YES
17-19 NO (but in same group)
17-28 NO (but in same group)

Note that the three that aren NOT actually duplicates have the lowest proportion matched (0.86-0.83)

#### Identify contaminated samples

Contaminated samples can be identified through elevated heterozygosity. NOTE- this isn't going to be helpful for us given the blood chimerism.

```{r identify contaminated samples (heterozygosity)}
#plot heterozygosity vs genotype rate per sample
#samples with unusually high heterozygosity relative to others are candidates for contamination
#the sample above the dashed line in this example is likely contaminated
ggplot()+geom_point(data=GTscore_individualSummary,aes(x=GenotypeRate,y=Heterozygosity))+
  labs(title="Heterozygosity vs Genotype Rate per Sample", x="Genotype Rate", y="Heterozygosity")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))+
  geom_hline(yintercept=0.30, lty="dashed")
```

Set heterozygosity treshold to identify likely contaminated sample
```{r}
#identify likely contaminated sample
contaminatedSample<-GTscore_individualSummary %>% filter(Heterozygosity>0.30) %>% pull(Sample)
contaminatedSample
```

Contaminated samples can also be identified by an elevated contamination score.  In this example, samples with a contamination score > 0.3 are known to be contaminated.
```{r }
#plot histogram of contamination score
ggplot()+geom_histogram(data=GTscore_individualSummary,aes(x=conScore),binwidth=0.02)+geom_vline(xintercept=0.3,lty="dashed")

```

Set contamination score treshold to identify contaminated samples
```{r}
#identify likely contaminated samples
contaminatedSamples2<-GTscore_individualSummary %>% filter(conScore>=0.3) %>% pull(Sample)
contaminatedSamples2
```


Plot genotype scatterplots for each sample
```{r}
#Scatter Plots can show evidence of contamination or elevated ploidy
plotGenotypes_sample(singleSNP_locusTable, singleSNP_alleleReads, polyGenResults_singleSNP, type='scatter', savePlot="Y", saveDir="scatterPlots_sample")
#look at plots for samples, particularly for putatively contaminated samples identified by high heterozygosity
```

```{r,echo = FALSE}
library(knitr)
```

#### Identify poor quality loci to remove

Flag SNPs with < 50% genotype rate as candidates for removal. Will need to do this with LWED and SIMP samples separately, then recombine. 
```{r}
poorQualitySNPs <- singleSNP_summary %>% filter(GenotypeRate<0.5) %>% mutate(Locus_ID=as.character(Locus_ID)) %>% pull(Locus_ID)

poorQualitySNPs_LWED <- singleSNP_summary_LWED %>% filter(GenotypeRate<0.5) %>% mutate(Locus_ID=as.character(Locus_ID)) %>% pull(Locus_ID)

poorQualitySNPs_SIMP <- singleSNP_summary_SIMP %>% filter(GenotypeRate<0.5) %>% mutate(Locus_ID=as.character(Locus_ID)) %>% pull(Locus_ID)

pq_lwed <- data.frame(name=poorQualitySNPs_LWED) %>%
  mutate(species="LWED")
pq_simp<- data.frame(name=poorQualitySNPs_SIMP) %>%
  mutate(species="SIMP")

pq_both <- rbind(pq_lwed, pq_simp)
View(pq_both)

pq_both2 <- pq_both %>%
  group_by(name) %>%
  mutate(count=n()) %>%
  distinct(name, .keep_all = T) %>%
  within(species[count == '2'] <- "both")
View(pq_both2)

# Export file
write.csv(pq_both2, "./summaryFiles/poorQualitySNPs_combined.csv", row.names = F)
```

#### plots
Ratio & Scatter Plots can show evidence of systemic issues in read counting (off-target reads, biases) or elevated ploidy

Plot genotype ratio & scatterplots for each locus
```{r}
#Allele Ratio Plots  
plotGenotypes(singleSNP_locusTable, singleSNP_alleleReads, polyGenResults_singleSNP, type='ratio', savePlot="Y", saveDir="ratioPlots_loci")
#Scatter Plots 
plotGenotypes(singleSNP_locusTable, singleSNP_alleleReads, polyGenResults_singleSNP, type='scatter', savePlot="Y", saveDir="scatterPlots_loci")
```


Plot histogram of contamination score for loci. An arbitrary score of 0.3 was set to identify problematic loci, but a lower or higher threshold may be more suitable.  Visual examination of loci with elevated scores is recommended.  In cases where there are very few heterozygous individuals, a high contamination score may be not truly reflect locus performance..
```{r, warning=FALSE}
#loci with high contamination scores should have scatterplots examined to ensure genotypes look accurate
singleSNP_summary %>% ggplot(data=.) + geom_histogram(aes(x=conScore),binwidth=0.01)+geom_vline(xintercept=0.3)
```

Plot contamination score vs minor allele frequency.  Loci with both a high contamination score and high minor allele frequency are likely to be problematic.
```{r, warning=FALSE}
singleSNP_summary %>% ggplot(data=.) + geom_point(aes(x=conScore,y=minAF))
```

Get list of loci with high contamination score
```{r}
singleSNP_summary %>% filter(conScore>0.3)
```
Look at scatterplots for loci with high contamination score. See orignal GTscore README.Rmd for examples of a paralog, locus w/likely off-target sequence, locus w/high contamination score but may be fine. 

## 6) Export genotype files (unfiltered & filtered)
When exporting files, you can either export the full results and remove any poor quality samples or loci afterwards, or you can use whitelists or blacklists to filter during export.  Both the exportGenepop and exportRubias functions have the following options: sampleWhitelist, locusWhitelist, sampleBlacklist, locusBlacklist.  The whitelists are a list of samples or loci to retain, the blacklists are lists of samples or loci to discard.

#### Genepop format

export genepop format - no filtering
```{r}
#export genepop, no filtering
exportGenepop(polyGenResults_singleSNP,singleSNP_locusTable,filename="polyGenResults_singleSNP_genepop.txt")
exportGenepop(polyGenResults_singleSNP_LWED,singleSNP_locusTable_LWED,filename="polyGenResults_singleSNP_genepop_LWED.txt")
exportGenepop(polyGenResults_singleSNP_SIMP,singleSNP_locusTable_SIMP,filename="polyGenResults_singleSNP_genepop_SIMP.txt")
```

export genepop format - poor quality loci removed
```{r}
#example with poor quality loci and contaminated sample removed
exportGenepop(polyGenResults_singleSNP,singleSNP_locusTable,
              locusBlacklist=poorQualitySNPs,
              filename="polyGenResults_singleSNP_genepop_filtered.txt")
exportGenepop(polyGenResults_singleSNP_LWED,singleSNP_locusTable_LWED,
              locusBlacklist=poorQualitySNPs_LWED,
              filename="polyGenResults_singleSNP_genepop_filtered_LWED.txt")
exportGenepop(polyGenResults_singleSNP_SIMP,singleSNP_locusTable_SIMP,
              locusBlacklist=poorQualitySNPs_SIMP,
              filename="polyGenResults_singleSNP_genepop_filtered_SIMP.txt")
```

#### Rubias format
Data output for Rubias should have metadata associated with each sample that specifies the sample type, mixture if a mixture sample, reference if a baseline sample, the reporting group, the sample collection, and individual name.

Sample metadata columns should have the following names: sample_type, repunit, collection, indiv

metadata can be created here or loaded from file

create sample metadata for mixture sample
```{r}
#load sample data file
sampleMetaData<-read.delim("sampleMetaData.csv",header=TRUE, sep = ",")
#adjust indiv names in metadata (some of the GTscore functions change - to .)
sampleMetaData$indiv <- sub('tamOpt-', 'tamOpt.', sampleMetaData$indiv)
#view
View(sampleMetaData)
```


export Rubias format with no filtering
```{r}
#export rubias format
exportRubias(polyGenResults_singleSNP,singleSNP_locusTable,sampleMetaData,filename="polyGenResults_singleSNP_rubias.txt")
exportRubias(polyGenResults_singleSNP_LWED,singleSNP_locusTable_LWED,sampleMetaData,filename="polyGenResults_singleSNP_rubias_LWED.txt")
exportRubias(polyGenResults_singleSNP_SIMP,singleSNP_locusTable_SIMP,sampleMetaData,filename="polyGenResults_singleSNP_rubias_SIMP.txt")
```

export Rubias format with poor quality loci and contaminated sample removed
```{r}
exportRubias(polyGenResults_singleSNP,singleSNP_locusTable,sampleMetaData,
             locusBlacklist=poorQualitySNPs,
             filename="polyGenResults_singleSNP_rubias_filtered.txt")
exportRubias(polyGenResults_singleSNP_LWED,singleSNP_locusTable_LWED,sampleMetaData,
             locusBlacklist=poorQualitySNPs_LWED,
             filename="polyGenResults_singleSNP_rubias_filtered_LWED.txt")
exportRubias(polyGenResults_singleSNP_SIMP,singleSNP_locusTable_SIMP,sampleMetaData,
             locusBlacklist=poorQualitySNPs_SIMP,
             filename="polyGenResults_singleSNP_rubias_filtered_SIMP.txt")
```

## 7) Locus diagnostics
Diagnostic functions are provided to help in identifying patterns of sequence variation for loci.  This can assist in identifying SNPs that were not accounted for in the initial probe design and in adjusting the in-silico probes to exclude off-target sequence from genotyping.  

There are two approaches used by GTscore: First, plotting the number of nucleotide mismatches by position in the sequence data relative to the reference sequence for each locus, and second, performing sequence alignments using the program MSA (Bodenhofer et al. 2015).  Plotting the mismatches by position is fast and easy to interpret except in the case of indels.  The MSA alignments take longer to complete but facilitate the visualization of indels.

The first step for both methods is to get the sequences that aligned to each locus, this can be done based on primer alignments or primer-probe alignments using the script matchReads.pl.

### Sequence alignment to each locus (Perl)
Input flags for this script are:
* –p    a tab delimited file containing primer/probe information for each locus
* --files  a text file containing a list of .fastq sequence files to count reads from
* --matchType the match type chosen for retaining sequences for each locus. Options are “primer” to retain sequences that match the primer for a locus or “primerProbe” to retain sequences that match both the primer and probe.  primerProbe is the default if no option is specified.

Optional flags:
* --prefix  optional prefix for output file names

You can run matchReads.pl from the command line using the following example:
perl matchReads.pl -p primerProbeFile.txt --files sampleFiles.txt --matchType primerProbe [--prefix]

Alternatively, you can also run it through R:
```{r match reads, eval=FALSE}
#primer match
system2("perl",
        args="matchReads.pl -p primerProbeFile.txt --files sampleFiles.txt --matchType primer")
system2("perl",
        args="matchReads.pl -p primerProbeFile.txt --files sampleFiles_LWED.txt --matchType primer --prefix LWED_")
system2("perl",
        args="matchReads.pl -p primerProbeFile.txt --files sampleFiles_SIMP.txt --matchType primer --prefix SIMP_")

#primer AND probe match
system2("perl",
        args="matchReads.pl -p primerProbeFile.txt --files sampleFiles.txt --matchType primerProbe")
system2("perl",
        args="matchReads.pl -p primerProbeFile.txt --files sampleFiles_LWED.txt --matchType primerProbe --prefix LWED_")
system2("perl",
        args="matchReads.pl -p primerProbeFile.txt --files sampleFiles_SIMP.txt --matchType primerProbe --prefix SIMP_")
```

The default output file names are matchReads_primerAligned.txt or matchReads_primerProbeAligned.txt depending on which match type was chosen.

### Create reference amplicon sequences file
To conduct locus diagnostics, we need both a text file and fasta file of reference amplicon sequences. The sequences are already in Sam's pos_amplicon.csv file, we just need to get it in the right format:

Text file first:
```{r create reference sequences file}
# Packages
library(Biostrings)
library(tidyverse)

# Import data
amplicon_csv <- read.csv("pos_amplicon.csv", sep = ",")
View(amplicon_csv)
translation <- read.csv("primers_seqName_to_shortName.csv")
View(translation)

# Subset to get names and amplicon sequences only & remove brackets
amplicons <- amplicon_csv %>%
  select(c("SNP_name", "amplicon")) %>%
  mutate(amplicon = str_replace_all(amplicon, "\\[|\\]", ""))
View(amplicons)

# Add short names to to amplicon seqs
amplicons_shortnames <- merge(amplicons, translation, by.x = "SNP_name", by.y = "seqID1") %>%
  select(c("primerName2", "amplicon")) %>%
  dplyr::rename(Locus=primerName2, refSeq=amplicon)
View(amplicons_shortnames)
length(amplicons_shortnames$Locus) #226 - good!

# Create LWED and SIMP subsets
amplicons_lwed <- amplicons_shortnames %>%
  filter(!str_detect(Locus, "SIMP"))

amplicons_simp <- amplicons_shortnames %>%
  filter(!str_detect(Locus, "LWED"))


# Export
write.table(amplicons_shortnames, "./ampliconRefSeqs.txt", sep= "\t", row.names = F, quote = F)
write.table(amplicons_lwed, "./ampliconRefSeqs_LWED.txt", sep= "\t", row.names = F, quote = F)
write.table(amplicons_simp, "./ampliconRefSeqs_SIMP.txt", sep= "\t", row.names = F, quote = F)
```

Then fasta file:
```{r create amplicon fasta file}
library(phylotools)

amplicon_txt <- read.table("ampliconRefSeqs.txt", header = T) %>%
  dplyr::rename(c(seq.name=Locus, seq.text=refSeq))
View(amplicon_txt)

dat2fasta(amplicon_txt, outfile = "ampliconRefSeqs.fasta")

amplicon_txt_lwed <- read.table("ampliconRefSeqs_LWED.txt", header = T) %>%
  dplyr::rename(c(seq.name=Locus, seq.text=refSeq))
View(amplicon_txt_lwed)

dat2fasta(amplicon_txt_lwed, outfile = "ampliconRefSeqs_LWED.fasta")

amplicon_txt_simp <- read.table("ampliconRefSeqs_SIMP.txt", header = T) %>%
  dplyr::rename(c(seq.name=Locus, seq.text=refSeq))
View(amplicon_txt_simp)

dat2fasta(amplicon_txt_simp, outfile = "ampliconRefSeqs_SIMP.fasta")
```

### MSA alignment (GTscore.R)
MSA alignment of the matched reads can be done directly on the output from matchReads.pl using the alignMatchedSeqs command in GTscore. Probe sequence is incorporated into the alignment to visualize how probes may be affecting read counting, and reference sequence can optionally be included to facilitate comparison of the observed sequence with the target locus. The probe sequence is obtained from the primer-probe file originally used by AmpliconReadCounter.pl; the reference sequence file has a two column format where the first column is the locus name and the second column is the reference sequence for the locus amplicon. MSA will take a very long time to run if too many sequences are included for alignment so a minimum read threshold (minReads) is included as an option in the alignment command (suggest begin with minReads=20). In addition, a maximum of 100 unique sequences (ranked by number of reads) will be aligned.  The maximum number of reads allowed for alignment can be changed using the maxAlignedSeqs option. Be sure the required perl package has been installed (Excel-Writer-XLSX).

```{r MSA alignment}
#load reference sequences in table format
referenceSeqs<-read.delim("ampliconRefSeqs.txt", header=TRUE, stringsAsFactors=FALSE )
referenceSeqs_LWED<-read.delim("ampliconRefSeqs_LWED.txt", header=TRUE, stringsAsFactors=FALSE )
referenceSeqs_SIMP<-read.delim("ampliconRefSeqs_SIMP.txt", header=TRUE, stringsAsFactors=FALSE )
View(referenceSeqs)
length(referenceSeqs$Locus)

#load primer probe file
primerProbes<-read.delim("primerProbeFile.txt", header=TRUE, stringsAsFactors=FALSE)
primerProbes_LWED<-read.delim("primerProbeFile_LWED.txt", header=TRUE, stringsAsFactors=FALSE)
primerProbes_SIMP<-read.delim("primerProbeFile_SIMP.txt", header=TRUE, stringsAsFactors=FALSE)
View(primerProbes)
length(primerProbes$Locus)

#primer aligned reads
primerMatchedReads<-read.delim("matchedReads_primerAligned.txt", header=TRUE, stringsAsFactors=FALSE)
View(primerMatchedReads)
alignMatchedSeqs(referenceSeqs=referenceSeqs,primerProbes=primerProbes,matchedReads=primerMatchedReads,minReads=20, maxAlignedSeqs=100,type="primer",saveDir="MSA_primerMatched")

primerMatchedReads_LWED<-read.delim("LWED_matchedReads_primerAligned.txt", header=TRUE, stringsAsFactors=FALSE)
View(primerMatchedReads_LWED)
alignMatchedSeqs(referenceSeqs=referenceSeqs_LWED,primerProbes=primerProbes_LWED,matchedReads=primerMatchedReads_LWED,minReads=20, maxAlignedSeqs=100,type="primer",saveDir="MSA_primerMatched_LWED")

primerMatchedReads_SIMP<-read.delim("SIMP_matchedReads_primerAligned.txt", header=TRUE, stringsAsFactors=FALSE)
View(primerMatchedReads_SIMP)
alignMatchedSeqs(referenceSeqs=referenceSeqs_SIMP,primerProbes=primerProbes_SIMP,matchedReads=primerMatchedReads_SIMP,minReads=20, maxAlignedSeqs=100,type="primer",saveDir="MSA_primerMatched_SIMP")

#primer probe aligned reads
primerProbeMatchedReads<-read.delim("matchedReads_primerProbeAligned.txt", header=TRUE, stringsAsFactors=FALSE)
alignMatchedSeqs(referenceSeqs=referenceSeqs, primerProbes=primerProbes, matchedReads=primerProbeMatchedReads,minReads=20, maxAlignedSeqs=100,type="primerProbe",saveDir="MSA_primerProbeMatched")

primerProbeMatchedReads_LWED<-read.delim("LWED_matchedReads_primerProbeAligned.txt", header=TRUE, stringsAsFactors=FALSE)
alignMatchedSeqs(referenceSeqs=referenceSeqs_LWED, primerProbes=primerProbes_LWED, matchedReads=primerProbeMatchedReads_LWED,minReads=20, maxAlignedSeqs=100,type="primerProbe",saveDir="MSA_primerProbeMatched_LWED")

primerProbeMatchedReads_SIMP<-read.delim("SIMP_matchedReads_primerProbeAligned.txt", header=TRUE, stringsAsFactors=FALSE)
alignMatchedSeqs(referenceSeqs=referenceSeqs_SIMP, primerProbes=primerProbes_SIMP, matchedReads=primerProbeMatchedReads_SIMP,minReads=20, maxAlignedSeqs=100,type="primerProbe",saveDir="MSA_primerProbeMatched_SIMP")
```

Excel files are output for each locus within the directory specified by savDir. The first row is a consensus sequence based on the sequence alignments. This is followed by the probes, reference sequence if included, and then the sequence alignments.


### Plotting mismatches by position (Perl and GTscore.R)
#### Calculate sequence mismatches by position (Perl)
Plotting the number of mismatches by position requires first comparing each matched sequence against the reference sequence and summing the number of mismatches for each nucleotide position.  This is done with the perl script seqMismatchPositions.pl.  

Input flags for this script are 
* --amplicon    a fasta file containing the reference amplicon sequence for each locus
* --matchedSeqs  a tab delimited file containing sequences that matched each locus.  This file is generated by matchReads.pl
* --matchType the match type chosen for retaining sequences for each locus.  Options are “primer” to retain sequences that match the primer for a locus or “primerProbe” to retain sequences that match both the primer and probe.  primerProbe is the default if no option is specified.

Optional flags
* --prefix  optional prefix for output file names

You can run matchReads.pl from the command line using the following example:
perl seqMismatchPositions.pl --amplicon ampliconRefSeqs.fasta --matchedSeqs matchedReads_primerAligned.txt  --matchType primer [--prefix]

Alternatively, you can also run it through R:
```{r sequence mismatches, eval=FALSE}
# Combined
system2("perl",
        args="seqMismatchPositions.pl --amplicon ampliconRefSeqs.fasta --matchedSeqs matchedReads_primerAligned.txt --matchType primer")
system2("perl",
        args="seqMismatchPositions.pl --amplicon ampliconRefSeqs.fasta --matchedSeqs matchedReads_primerProbeAligned.txt --matchType primerProbe")

# LWED
system2("perl",
        args="seqMismatchPositions.pl --amplicon ampliconRefSeqs_LWED.fasta --matchedSeqs LWED_matchedReads_primerAligned.txt --matchType primer --prefix LWED_")
system2("perl",
        args="seqMismatchPositions.pl --amplicon ampliconRefSeqs_LWED.fasta --matchedSeqs LWED_matchedReads_primerProbeAligned.txt --matchType primerProbe --prefix LWED_")

# SIMP
system2("perl",
        args="seqMismatchPositions.pl --amplicon ampliconRefSeqs_SIMP.fasta --matchedSeqs SIMP_matchedReads_primerAligned.txt --matchType primer --prefix SIMP_")
system2("perl",
        args="seqMismatchPositions.pl --amplicon ampliconRefSeqs_SIMP.fasta --matchedSeqs SIMP_matchedReads_primerProbeAligned.txt --matchType primerProbe --prefix SIMP_")
```
The default output file names are mismatchPositions_primer.txt or mismatchPositions_primerProbe.txt depending on which match type was chosen.  The output files will be used by the summarizeMismatches function in GTscore to generate plots of mismatches by position for each locus.

#### Plot of sequence mismatches by position (GTscore.R)
```{r}
#load results from seqMismatchPositions.pl
#primer matched sequences
mismatchPositionData_primer<-read.delim("mismatchPositions_primer.txt", header=TRUE, stringsAsFactors=FALSE)
mismatchPositionData_primer_LWED<-read.delim("mismatchPositions_LWED_primer.txt", header=TRUE, stringsAsFactors=FALSE)
mismatchPositionData_primer_SIMP<-read.delim("mismatchPositions_SIMP_primer.txt", header=TRUE, stringsAsFactors=FALSE)
#primer probe matched sequences
mismatchPositionData_primerProbe<-read.delim("mismatchPositions_primerProbe.txt", header=TRUE, stringsAsFactors=FALSE)
mismatchPositionData_primerProbe_LWED<-read.delim("mismatchPositions_LWED_primerProbe.txt", header=TRUE, stringsAsFactors=FALSE)
mismatchPositionData_primerProbe_SIMP<-read.delim("mismatchPositions_SIMP_primerProbe.txt", header=TRUE, stringsAsFactors=FALSE)


#generate plots of mismatches by position
#primer matched sequences
summarizeMismatches(mismatchPositionData_primer,saveDir="mismatchPositionPlots_primer")
summarizeMismatches(mismatchPositionData_primer_LWED,saveDir="mismatchPositionPlots_primer_LWED")
summarizeMismatches(mismatchPositionData_primer_SIMP,saveDir="mismatchPositionPlots_primer_SIMP")
#primer probe matched sequences
summarizeMismatches(mismatchPositionData_primerProbe,saveDir="mismatchPositionPlots_primerProbe")
summarizeMismatches(mismatchPositionData_primerProbe_LWED,saveDir="mismatchPositionPlots_primerProbe_LWED")
summarizeMismatches(mismatchPositionData_primerProbe_SIMP,saveDir="mismatchPositionPlots_primerProbe_SIMP")
```

## 8) FILTERED analysis & summaries

Need to remove:
1) underperforming loci
2) underperforming samples

First, we'll make new sampleFiles.txt and primerProbes.txt by filtering out 1) underperforming samples and 2) underperforming loci based on that analyses performed earlier. We'll first set a cutoff of 2000 total primer-probe reads per sample, which leaves us with 28 samples.

For the loci, we'll set a cut-off of 280 total primer-probe reads since we want at least 10x coverage per sample. Note that because our read counts at present include those from the samples we cut, we may miss some loci that should be cut (particularly those with just above 280 reads) - this is also why we're not using avg read depth as a cutoff at this point. We'll look at total read count per loci again after re-running the analyses.
```{r}
# Filter samples by removing those with under 2000 reads (each species set has a little over 200 loci, 200 * 10x coverage = 2000)
sampleFiles <- read.table("sampleFiles.txt", header = F)
sampleFiles_LWED <- read.table("sampleFiles_LWED.txt", header = F)
sampleFiles_SIMP <- read.table("sampleFiles_SIMP.txt", header = F)
failedSamples <- c("tamOpt-004_S4_L001_interleaved.fastq","tamOpt-007_S7_L001_interleaved.fastq","tamOpt-022_S25_L001_interleaved.fastq","tamOpt-025_S16_L001_interleaved.fastq","tamOpt-035_S33_L001_interleaved.fastq")

sampleFiles_filtered <- sampleFiles %>%
  filter(!V1 %in% failedSamples)
View(sampleFiles_filtered)

sampleFiles_filtered_LWED <- sampleFiles_LWED %>%
  filter(!V1 %in% failedSamples)
sampleFiles_filtered_SIMP <- sampleFiles_SIMP %>%
  filter(!V1 %in% failedSamples)

# Export
write.table(sampleFiles_filtered, "f_sampleFiles.txt", col.names = F, row.names = F, quote = F)
write.table(sampleFiles_filtered_LWED, "f_sampleFiles_LWED.txt", col.names = F, row.names = F, quote = F)
write.table(sampleFiles_filtered_SIMP, "f_sampleFiles_SIMP.txt", col.names = F, row.names = F, quote = F)

# Create list of loci with under 10x avg read depth
under10 <- ls_ss %>%
  filter(AvgReadDepth < 10)
View(under10)

# Create filtered primerProbe files
ppf <- pp4 %>%
  filter(!Locus %in% under10$Locus)
View(ppf)
ppfLWED <- ppf %>%
  filter(!str_detect(Locus, "SIMP"))
ppfSIMP <- ppf %>%
  filter(!str_detect(Locus, "LWED"))

write.table(ppf, "./filteredLociSamples/primerProbeFile-f.txt", sep = "\t", row.names = F, quote = F)
write.table(ppfLWED, "./filteredLociSamples/primerProbeFile_LWED-f.txt", sep = "\t", row.names = F, quote = F)
write.table(ppfSIMP, "./filteredLociSamples/primerProbeFile_SIMP-f.txt", sep = "\t", row.names = F, quote = F)
```

Now we can re-run AmmpliconReadCounter.pl
```{r count reads for amplicons, eval=FALSE}
# Unfiltered reads
## All
system2("perl",
        args="AmpliconReadCounter.pl -p f_primerProbeFile.txt --files f_sampleFiles.txt --prefix f_")

## LWED
system2("perl",
        args="AmpliconReadCounter.pl -p f_primerProbeFile_LWED.txt --files f_sampleFiles_LWED --prefix f_LWED_")

## SIMP
system2("perl",
        args="AmpliconReadCounter.pl -p f_primerProbeFile_SIMP.txt --files f_sampleFiles_SIMP.txt --prefix f_SIMP_")
```


### 8.1) Genotyping
```{r Genotyping}
#load locus table and allele reads file
singleSNP_locusTable_f<-read.delim("f_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)
singleSNP_alleleReads_f<-read.delim("f_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE)

head(singleSNP_locusTable_f)
singleSNP_alleleReads_f[1:5,1:5]

singleSNP_locusTable_LWED_f<-read.delim("f_LWED_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)
singleSNP_alleleReads_LWED_f<-read.delim("f_LWED_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE)

head(singleSNP_locusTable_LWED)
singleSNP_alleleReads_LWED[1:5,1:5]

singleSNP_locusTable_SIMP_f<-read.delim("f_SIMP_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)
singleSNP_alleleReads_SIMP_f<-read.delim("f_SIMP_AlleleReads_singleSNPs.txt",header=TRUE,row.names=1,stringsAsFactors=FALSE)

head(singleSNP_locusTable_SIMP)
singleSNP_alleleReads_SIMP[1:5,1:5]

#generate singleSNP genotypes using the polyGen algorithm
polyGenResults_singleSNP_f<-polyGen(singleSNP_locusTable_f,singleSNP_alleleReads_f)
polyGenResults_singleSNP_LWED_f<-polyGen(singleSNP_locusTable_LWED_f,singleSNP_alleleReads_LWED_f)
polyGenResults_singleSNP_SIMP_f<-polyGen(singleSNP_locusTable_SIMP_f,singleSNP_alleleReads_SIMP_f)

#look at first five rows and columns
polyGenResults_singleSNP_f[1:5,1:5]
polyGenResults_singleSNP_LWED_f[1:5,1:5]
polyGenResults_singleSNP_SIMP_f[1:5,1:5]

#write results
write.table(polyGenResults_singleSNP_f,"f_polyGenResults_singleSNP.txt",quote=FALSE,sep="\t")
write.table(polyGenResults_singleSNP_LWED_f,"f_polyGenResults_singleSNP_LWED.txt",quote=FALSE,sep="\t")
write.table(polyGenResults_singleSNP_SIMP_f,"f_polyGenResults_singleSNP_SIMP.txt",quote=FALSE,sep="\t")
```


### 8.2) Sample summaries
load sample summary from AmpliconReadCounter
```{r}
GTscore_individualSummary_f<-read.delim("f_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
head(f_GTscore_individualSummary)

GTscore_individualSummary_LWED_f<-read.delim("f_LWED_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
head(f_GTscore_individualSummary_LWED)

GTscore_individualSummary_SIMP_f<-read.delim("f_SIMP_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
head(f_GTscore_individualSummary_SIMP)
```

summarize single SNP results for samples
```{r}
singleSNP_sampleSummary_f<-summarizeSamples(f_polyGenResults_singleSNP,f_singleSNP_alleleReads)
head(f_singleSNP_sampleSummary)

singleSNP_sampleSummary_LWED_f<-summarizeSamples(f_polyGenResults_singleSNP_LWED,f_singleSNP_alleleReads_LWED)
head(f_singleSNP_sampleSummary_LWED)

singleSNP_sampleSummary_SIMP_f<-summarizeSamples(f_polyGenResults_singleSNP_SIMP,f_singleSNP_alleleReads_SIMP)
head(f_singleSNP_sampleSummary_SIMP)
```

combine AmpliconReadCounter individual summary data with GTscore sample summary
```{r}
# make new copies
gt_indivSummary_f <- GTscore_individualSummary_f
gt_indivSummary_LWED_f <- GTscore_individualSummary_LWED_f
gt_indivSummary_SIMP_f <- GTscore_individualSummary_SIMP_f

#first adjust names from sample_genotypeRate_singleSNP to match GTscore_individualSummary (some of the GTscore functions change - to .)
gt_indivSummary_f$Sample <- sub('tamOpt-', 'tamOpt.', gt_indivSummary_f$Sample)
gt_indivSummary_f<-merge(gt_indivSummary_f,singleSNP_sampleSummary_f,by.x="Sample",by.y="sample")

gt_indivSummary_LWED_f$Sample <- sub('tamOpt-', 'tamOpt.', gt_indivSummary_LWED_f$Sample)
gt_indivSummary_LWED_f<-merge(gt_indivSummary_LWED_f,singleSNP_sampleSummary_LWED_f,by.x="Sample",by.y="sample")

gt_indivSummary_SIMP_f$Sample <- sub('tamOpt-', 'tamOpt.', gt_indivSummary_SIMP_f$Sample)
gt_indivSummary_SIMP_f<-merge(gt_indivSummary_SIMP_f,singleSNP_sampleSummary_SIMP_f,by.x="Sample",by.y="sample")

# Combine LWED & SIMP to form one
gt_indivSummary_combined_f <- rbind(gt_indivSummary_LWED_f, gt_indivSummary_SIMP_f)
View(gt_indivSummary_combined_f)

# Export
write.csv(gt_indivSummary_f, "./summaryFiles/f_individualSummary_FULL.csv", row.names = F)
write.csv(gt_indivSummary_LWED_f, "./summaryFiles/f_individualSummary_LWED_FULL.csv", row.names = F)
write.csv(gt_indivSummary_SIMP_f, "./summaryFiles/f_individualSummary_SIMP_FULL.csv", row.names = F)
write.csv(gt_indivSummary_combined_f, "./summaryFiles/f_individualSummary_combined_FULL.csv", row.names = F)
```

### 8.4) Locus summaries
```{r}
#summarize single SNP results
singleSNP_summary_f<-summarizeGTscore(singleSNP_alleleReads_f, singleSNP_locusTable_f, polyGenResults_singleSNP_f)
singleSNP_summary_LWED_f<-summarizeGTscore(singleSNP_alleleReads_LWED_f, singleSNP_locusTable_LWED_f, polyGenResults_singleSNP_LWED_f)
singleSNP_summary_SIMP_f<-summarizeGTscore(singleSNP_alleleReads_SIMP_f, singleSNP_locusTable_SIMP_f, polyGenResults_singleSNP_SIMP_f)
#view results
View(singleSNP_summary_f)
View(singleSNP_summary_LWED_f)
View(singleSNP_summary_SIMP_f)
#write results
write.table(singleSNP_summary_f,"./summaryFiles/f_singleSNP_summary.txt",quote=FALSE,sep="\t",row.names=FALSE)
write.table(singleSNP_summary_LWED_f,"./summaryFiles/f_singleSNP_summary_LWED.txt",quote=FALSE,sep="\t",row.names=FALSE)
write.table(singleSNP_summary_SIMP_f,"./summaryFiles/f_singleSNP_summary_SIMP.txt",quote=FALSE,sep="\t",row.names=FALSE)
```

May also be helpful to combine single SNP summaries with the GTscore locus summary created earlier:
```{r}
# Load data
ls_f <- read.table("f_GTscore_locusSummary.txt", header = T, sep = "\t")
ls_lwed_f <- read.table("LWED-f_GTscore_locusSummary.txt", header = T, sep = "\t")
ls_simp_f <- read.table("SIMP-f_GTscore_locusSummary.txt", header = T, sep = "\t")

# Separate summaries into results from shared and species-specific loci analyses
ls1_f <- ls_f %>%
  filter(!str_detect(Locus, "LWED|SIMP"))
ls_lwed1_f <- ls_lwed_f %>%
  filter(str_detect(Locus, "LWED"))
ls_simp1_f <- ls_simp_f %>%
  filter(str_detect(Locus, "SIMP"))

ss_f <- singleSNP_summary_f %>%
  filter(!str_detect(Locus_ID, "LWED|SIMP"))
ss_lwed_f <- singleSNP_summary_LWED_f %>%
  filter(str_detect(Locus_ID, "LWED"))
ss_simp_f <- singleSNP_summary_SIMP_f %>%
  filter(str_detect(Locus_ID, "SIMP"))

# Recombine gtscore & single snp summary files, then merge the two
ls_recombine_f <- rbind(ls1_f, ls_lwed1_f, ls_simp1_f)
View(ls_recombine_f)
ss_recombine_f <- rbind(ss_f, ss_lwed_f, ss_simp_f)
ss_recombine_f$Locus_ID <- substr(ss_recombine_f$Locus_ID,0,nchar(ss_recombine_f$Locus_ID)-4) # change locus names to match ls
View(ss_recombine_f)

ls_ss_f <- merge(ls_recombine_f, ss_recombine_f, by.x = "Locus", by.y = "Locus_ID") 
View(ls_ss_f)

# Export
write.csv(ls_ss_f, "./summaryFiles/f_locusSummary_combined.csv", row.names = F)
```
