---
title: "tamRun1_gtseqPipeline"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

Running gtseq here for tamRun1 - yes Sam already did this, but want to have it here for consistency across other runs. Also for consistency, using GTseq_Genotyper_v2.pl

# 2 Packages

```{r}
library(phylotools)
library(seqinr)
library(tidyverse)
```

# 3 Pipeline

## 3.1 GTseq_HashSeqs.pl

First use the GTseq_HashSeqs.pl script on plate .fastq files. This script collects and counts unique reads within the fastq file and reduces the compute time for the GTseq_SeqTest.pl script.

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/00_tamRun1/05_run1GTseq

for i in ./../02_run1Interleaved/tamOpt*; do perl ./../../project_scripts/GTseq/GTseq_HashSeqs.pl $i > ./hashSeqs/$(basename "${i/.fastq/.hash}"); done
```

## 3.2 GTseq_SeqTest [[NOT COMPLETED]]

### Create locusInfo_seqTest file

```{r, eval = F}
locusInfo_seqTest <- read.table("./04_tamRun4/00_illumina/04_run4GTseq/primerProbe_snpPanel_v3_gtseq.txt", sep = "\t", header = F)

write.table(locusInfo_seqTest, "./04_tamRun4/00_illumina/04_run4GTseq/locusInfo_seqTest.csv", col.names = F, row.names = F, sep = ",", quote = F)
```

### Run script

**NOTE** there are some weird errors in the updated SeqTest script Nate sent me; I modified a few lines so that it would run. Also, there's a line in the script that for forward-primer counts, it changes 0 to 0.1 -- not sure what the purpose of this is?? 

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/04_tamRun4/00_illumina/04_run4GTseq

for i in ./hashSeqs/tamRun4*; do perl ./../../../project_scripts/GTseek_Git/GTseq_SeqTest_v3_rvModified.pl primerProbe_snpPanel_v3_gtseq.csv $i > ./seqTest/$(basename "${i/.hash/.seqtest.csv}"); done
```

## 3.3 GTseq_Genotyper_v2.pl

### Create locusInfo_genotyper file (csv)

```{r}
run1_primerProbeFile <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/primerProbeFile_fullSet.txt", header = T)

locusInfo_genotyper <- run1_primerProbeFile %>%
  select(c(Locus, Allele1, Allele2, Probe1, Probe2, Primer))

write.table(locusInfo_genotyper, "./00_tamRun1/05_run1GTseq/locusInfo_genotyper.csv", col.names = F, row.names = F, sep = ",", quote = F)
```

### Run script

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/00_tamRun1/05_run1GTseq

for i in ./../02_run1Interleaved/tamOpt*; do perl ./../../project_scripts/GTseq/GTseq_Genotyper_v2.pl locusInfo_genotyper.csv $i > ./genos/$(basename "${i/.fastq/.genos}"); done
```

## 3.4 Run GTseq_GenoCompile_v3.pl

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/00_tamRun1/05_run1GTseq/genos

# numeric genos
perl ./../../../project_scripts/GTseq/GTseq_GenoCompile_v3.pl N > ./../tamRun1_numericGenos.csv

# readSums
perl ./../../../project_scripts/GTseq/GTseq_GenoCompile_v3.pl C > ./../tamRun1_readSums.csv

# genos
perl ./../../../project_scripts/GTseq/GTseq_GenoCompile_v3.pl > ./../tamRun1_compiledGenos.csv
```
