---
title: "tamRun5_gtseqPipeline"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

This document outlines data analysis for tamGenetics_run5 using the "GTseq pipeline" ([GitHub link](https://github.com/GTseq/GTseq-Pipeline)) created by Nate Campbell.

I previously ran this in October 2023 with GTseq_Genotyper_v3.pl; rerunning now with GTseq_Genotyper_v2.pl to align with previous analyses/allele ratio cutoffs

# 2 Packages

```{r}
library(phylotools)
library(seqinr)
library(tidyverse)
```

# 3 Pipeline

## 3.1 GTseq_HashSeqs.pl

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/05_tamRun5/05_run5GTseq

for i in ./../02_run5Interleaved/tamRun5*; do perl ./../../project_scripts/GTseq/GTseq_HashSeqs.pl $i > ./hashSeqs/$(basename "${i/.fastq/.hash}"); done
```

## 3.2 GTseq_SeqTest

### Create locusInfo_seqTest file

Below I'm taking the version 3 primer-probe file that I used for the GTscore pipeline and reformatting it for the GTseq_SeqTest.pl script.

```{r, eval = F}
primerList_v3 <- read.csv("./primers/03_lociChoices/tamGenetics_primerList_v3.csv")

pp_fullSet <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt", header = T) %>%
  filter(Locus %in% primerList_v3$locus)

pp_gtseq_seqTest <- pp_fullSet %>%
  select(c(Locus, Primer, Probe1, Probe2))

write.table(pp_gtseq_seqTest,"./05_tamRun5/05_run5GTseq/locusInfo_seqTest.txt", quote = F, sep = "\t", col.names = F, row.names = F)
```

### Run script

Output has four columns:
  
  1. Locus
  2. Forward-primer count
  3. Probe count
  4. Both count

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/05_tamRun5/05_run5GTseq

for i in ./hashSeqs/tamRun5*; do perl GTseq_SeqTest.pl primerProbe_gtseq.txt $i > ./seqTest/$(basename "${i/.hash/.seqtest.csv}"); done
```

## 3.3 GTseq_Genotyper_v2.pl

### Create locusInfo_genotyper file (csv)

The genotyper script requires a slightly different loci info file than for the seqTest script. This loci information file needs to contain locus names, allele names, and in-silico probe sequences in .csv format.

**NOTE** I created GTseq_Genotyper_v3_v2Cutoffs_rvModified.pl bc the original v2 script wasn't playing nice; instead I just modifed the v3 version to have v2 cutoffs

```{r, eval = F}
pp_gtseq_genotyper <- pp_fullSet %>%
  select(c(Locus, Allele1, Allele2, Probe1, Probe2, Primer))

write.table(pp_gtseq_genotyper, "./05_tamRun5/05_run5GTseq/locusInfo_genotyper.csv", col.names = F, row.names = F, sep = ",", quote = F)
```

### Run script

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/05_tamRun5/05_run5GTseq

for i in ./../02_run5Interleaved/tamRun5*; do perl ./../../project_scripts/GTseq/GTseq_Genotyper_v3_v2Cutoffs_rvModified.pl locusInfo_genotyper.csv $i > ./genos/$(basename "${i/.fastq/.genos}"); done
```

Output files contain a header line with summary information followed by locus specific data. Immediately below the header are a set of locus specific metrics, with (unlabeled) columns for the following:

-   LocusName
-   Allele1_counts
-   Allele2_counts
-   A1/A2-ratio
-   Genotype
-   Genotype_Class
-   A1_correction value
-   A2_correction value
-   On_target reads
-   Locus OT_percentage
-   Locus On-target reads as percentage of total on-target reads

## 3.4 GTseq_genoCompile

**NOTE** I created a modified version of GTseq_genoCompile_v3 so that it could output individual allele reads; previously it only had options for genos and readSums

```{bash, eval = F}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/05_tamRun5/05_run5GTseq/genos

# genos
perl ./../../../project_scripts/GTseq/GTseq_GenoCompile_v3_rvModified.pl > ./../tamRun5_compiledGenos.csv

# readSums
perl ./../../../project_scripts/GTseq/GTseq_GenoCompile_v3_rvModified.pl C > ./../tamRun5_readSums.csv

# alleleReads
perl ./../../../project_scripts/GTseq/GTseq_GenoCompile_v3_rvModified.pl R > ./../tamRun5_alleleReads.csv
```