---
title: "chapter1_panelDev"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

This document is to outline all of the data and analyses for dissertation Chapter 1 on SNP panel development.

# 2 Packages

```{r, message=FALSE}
library(ggthemes)
library(gsubfn)
library(huxtable)
library(janitor)
library(seqinr)
library(tidyverse)

source("project_scripts/ggplot_theme_Publication-2.R") # for pretty ggplot themes
source("project_scripts/ravo_gtScripts.R") # for my custom functions
```

if there's issues w/installing seqinr, try installing packages "progress", "cpp11", and "Rcpp"

ggplot_theme_Publication-2.R is from: https://medium.com/analytics-vidhya/ggplot2-themes-for-publication-ready-plots-including-dark-themes-9cd65cc5a7e3 

# 3 Data

## 3.1 seqName_to_shortName

```{r}
# seqName to shortName file
seqNames_to_shortNames <- read.csv("./project_data/seqNames_to_shortNames.csv")
```

## 3.2 primerProbeFiles

```{r}
# laLab primerProbeFile - should contain all loci that could be rescued
pp_laLab <- read.csv("./02_laLab_run1/03_laLab_run1GTscore/primerProbeFile_fullSet.txt", header = T, sep = "\t")

pp_tamRun5 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)
```

# 4 SNP panel prep

## 4.1 Overview

In total, Sam identified 402 informative loci, including n = 219 INDID, n = 66 LWED, n = 66 SIMP, n = 20 SPECIES, and n = 31 SEXID loci. I used sequences surrounding these loci as input for primer design in BatchPrimer3, which successfully designed primers for 362 loci. 

bp3 first set designed primers for n = 329 loci
bp3 second set designed primers for n = 33 loci
TOTAL = 362
-   INDID       152 + 33 = 185
-   LWED        66
-   SIMP        66
-   SEXID       25
-   SPECIES     20

n = 302 passed in silico optimization


Link to [Tamarin Genetics Lab Notebook 2 - Primer Design](https://quip.com/O64RAGBSzS5r/Tamarin-Genetics-Lab-Notebook-2-Primer-Design)

## 4.2 Panel prep

### 1) multifasta_1

Sam sent an initial set of **n_loci = 357** (multifasta_1) to be used as input for bp3 - breakdown of counts for loci sets below:

-   INDID       174
-   LWED         66
-   SIMP         66
-   SEXID        31
-   SPECIESID    20
  
```{r}
multifasta_1 <- read.fasta(file = "./primers/concatenated_species_ind_sex_14.01.2022.fasta") %>%
  names(.) %>%
  as.data.frame() %>%
  dplyr::rename("original" = ".") %>%
  mutate(set = sub('\\_.*', '', original)) %>%
  filter(set %in% c("INDID", "LWED", "SEXID", "SIMP", "SPECIESID"))

multifasta_1 %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  rbind(c("total", nrow(multifasta_1)))
```

### 2) bp3Output_1

BatchPrimer3 was able to successfully design primers for **n = 329** out of the 357 loci from multifasta_1, including the following:

```{r}
bp3Output_1 <- read.csv("./primers/primerSet1_bp3Output_formattingUpdated_20Jan2022.csv") %>%
  mutate(set = sub('\\_.*', '', Seq.ID))

bp3Output_1 %>%
  select(Seq.ID, set) %>%
  distinct() %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  janitor::adorn_totals()
```
Loci that failed primer design (n = 28) include the following counts per lociSet:

```{r}
multifasta_1 %>%
  filter(!original %in% bp3Output_1$Seq.ID) %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  janitor::adorn_totals()
```

### 3) fastpcrOutput_1a

BatchPrimer3 was set to return three potential primer pairs per locus, which we subsequently uploaded to FastPCR to assign quality scores to each primer pair. FastPCR quality scores range from zero to 100 and are based on in silico amplification efficiency, with higher scores assigned to primers with wider ranges of executable temperatures and thus a lower likelihood of adverse effects associated with variation in reaction buffers, polymerase types, or annealing temperatures.

The output from FastPCR is below:

```{r}
fastpcrOutput_1a <- read.csv("./primers/primerSet1_fastpcrAnalysis_20Jan2022.csv")
```

### 4) fastpcrOutput_1b

Because FastPCR assigns quality scores to forward and reverse primers individually, I created a ranking system for the primer *pairs* based on the following:

1)    "qualType", in which I assigned pairs as "type1" when both forward and reverse primers had quality scores >= 75, "type2" when both forward and reverse primers had quality scores >= 50, but one or both scored < 75, and "type3" when both forward and reverse primers had quality scores < 50
2)    "avgQual", which represented the average of forward and reverse primer quality scores for each pair

I used these variables to choose primer pairs for each locus that maximized both "qualType" and "avgQual", resulting in fastpcrOutput2.

```{r}
fastpcrOutput_1b <- read.csv("./primers/primerSet2_forInSilico_20Jan2022.csv")
```

I have a set of scripts outlining how I did this originally, but the script below is way simpler and easier to follow (and gives the same results) --

```{r}
temp <- fastpcrOutput_1a %>%
  mutate(
    bp3ID = str_c(bp3Index, ".", bp3Rank)
  ) %>%
  dplyr::rename("quality" = "Primer_Quality...") %>%
  select(bp3ID, primerFR, quality) %>%
  pivot_wider(names_from = primerFR,
              values_from = quality) %>%
  separate(bp3ID, into = c("bp3Index", "bp3Rank")) %>%
  mutate(
    avgQuality = (forward + reverse)/2,
    qualType = case_when(
      forward >= 75 & reverse >= 75 ~ "type1",
      forward >= 75 & reverse >= 50 ~ "type2",
      forward >= 50 & reverse >= 75 ~ "type2",
      forward >= 50 & reverse >= 50 ~ "type2",
      .default = "type3"
    )
  ) %>%
  group_by(bp3Index) %>%
  # subset to highest qualType per bp3Index
  slice_min(order_by = factor(qualType, levels = c("type1", "type2", "type3"))) %>%
  # subset to highest avgQuality w/in highest qualType
  slice_max(order_by = avgQuality, n = 1) %>%
  distinct(bp3Index, .keep_all = T) %>%
  arrange(as.numeric(bp3Index)) %>%
  mutate(
    bp3id = str_c(bp3Index, ".", bp3Rank)
  )

fastpcrOutput_1b <- fastpcrOutput_1a %>%
  mutate(
    bp3id = str_c(bp3Index, ".", bp3Rank)
  ) %>%
  filter(bp3id %in% temp$bp3id)
```

### 5) mfeOutput_1

**the metrics for this section are based on the first round of MFE only; there were additional rounds where primers for failed loci were swapped out and tried again**

I sent fastpcrOutput_1b to Sam to run MFE in silico tests. Of the n = 329 loci, **n = 235** passed and n = 94 failed in silico tests.

```{r}
mfeOutput_1 <- read.csv("./primers/primerSet2_passFail.csv") %>%
  select(-X) %>%
  filter(!is.na(inSilico)) %>%
  distinct(bp3Index, .keep_all = T) %>%
  mutate(set = case_when(
    str_detect(Seq.ID, "INDID") ~ "INDID",
    str_detect(Seq.ID, "LWED") ~ "LWED",
    str_detect(Seq.ID, "SIMP") ~ "SIMP",
    str_detect(Seq.ID, "SEXID") ~ "SEXID",
    str_detect(Seq.ID, "SPECIESID") ~ "SPECIES",
    .default = NA
  ))

sum(mfeOutput_1$inSilico == "PASS") # 235
sum(mfeOutput_1$inSilico == "FAIL") # 94
```

Info on loci that failed in-silico PCR is in multifasta_2 section (need info from multifasta_2 to figure out sets and whatnot since MFE tests went through a few rounds - that means that while 94 loci failed here, replacement primers could have passed for some)

### 6) multifasta_2

To replace primers that failed in silico optimization, Sam identified an additional 45 INDID loci for primer design. These were combined into a multi-FASTA file with the loci that had already passed in silico PCR for a total of **n = 320** total, with the following loci-set breakdown:

-   INDID       169 (124 previous + 45 additional)
-   LWED        62
-   SIMP        53
-   SEXID       18
-   SPECIESID   18
-   total       320

**NOTE** that there were some discrepancies between Sam's numbers and mine (see quip doc for details) -- the "rvEdits" fasta file was udpated to address those issues. **HOWEVER** it looks like I did something dumb in those edits b/c SEXID_gatk...815956-816448 (which I apparently thought was missing, and therefore added) is now in there twice

```{r}
multifasta_2 <- read.fasta(file = "./primers/concatenated_ALL_19.02.2022_rvEdits.fasta") %>%
  names(.) %>%
  as.data.frame() %>%
  dplyr::rename("original" = ".") %>%
  mutate(set = sub('\\_.*', '', original)) %>%
  filter(set %in% c("INDID", "LWED", "SEXID", "SIMP", "SPECIESID"))

multifasta_2 %>%
  get_dupes(original) # SEXID_gatk...815956-816448 in there twice - REMOVE one

multifasta_2 <- multifasta_2 %>%
  filter(original != "SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_815956-816448_chosen_consensus_modified") %>%
  rbind(., c("SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_815956-816448_chosen_consensus_modified", "SEXID"))

multifasta_2 %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals()
```

That means that the loci that failed the first round in-silico PCR include the following counts per lociSet:

```{r}
# seq names from multifasta_1 that passed first round in silico (n = 275)
temp <- multifasta_2 %>%
  filter(original %in% multifasta_1$original)

temp %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  adorn_totals()

# seq names that failed first round in silico
bp3Output_1 %>% # 329 primers designed
  select(Seq.ID, set) %>%
  distinct() %>% # has entries for both f/r - ditch dups
  filter(!Seq.ID %in% temp$original) %>% # ditch those that passed in silico
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  janitor::adorn_totals()
```

### 7) bp3Output_2

I used multifasta_2 as input for BatchPrimer3 to see if we could find primers for the additional 45 loci. **BatchPrimer3 was able to design primers for 33 out of the 45 additional loci**, yielding primers for a total of **n = 308** loci.

**NOTE** that my quip doc shows bp3 output w/primers for 309 loci; this is just due to the duplicated SEXID sequence (see above)

Counts for loci sets below:

-   INDID       157
-   LWED         62
-   SEXID        18
-   SIMP         53
-   SPECIESID    18
-   total       308

```{r}
bp3Output_2 <- read.csv("./primers/primerSet3_bp3Output_originalFile_22Feb2022.csv") %>%
  mutate(set = sub('\\_.*', '', Seq.ID))

# bp3 found primers for 308 loci
bp3Output_2 %>%
  select(Seq.ID, set) %>%
  distinct() %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals()

# primers designed for 33 out of the 45 new loci
bp3Output_2 %>%
  select(Seq.ID) %>%
  distinct() %>%
  filter(str_detect(Seq.ID, "additional")) %>%
  nrow()
```

Loci that failed primer design (n = 12) for the 45 new loci include the following counts per lociSet:

```{r}
multifasta_2 %>%
  filter(!original %in% bp3Output_2$Seq.ID) %>%
  group_by(set) %>%
  summarise(count = n())
```

### 8) fastpcrOutput_2a

bp3Output_2 for the 33 new primers was input to FastPCR to assign quality scores with the following output:

```{r}
fastpcrOutput_2a <- read.csv("./primers/primerSet3_fastpcrAnalysis_22Feb2022.csv")
```

### 9) fastpcrOutput_2b

As before, I used the quality scores assigned by FastPCR to choose the highest-scoring primer *pair* for each of the new loci, resulting in the following primer choices for the 33 new loci:

```{r}
fastpcrOutput_2b <- read.csv("./primers/primerSet4_forInSilico_22Feb2022.csv")
```

### 10) mfeOutput_2

**These numbers aren't accurate; doesn't account for replacement primers and re-testing in silico - based on my investigations though, it seems like we dropped 6 here**

I sent fastpcrOutput_2b to Sam to run MFE in silico tests. Of the n = 33 new loci, **n = 12** passed and n = 21 failed in silico tests.

```{r}
mfeOutput_2 <- read.csv("./primers/primerSet4_resultsInSilico_23Feb2022.csv") %>%
  filter(!is.na(inSilico)) %>%
  distinct(bp3Index, .keep_all = T) %>%
  mutate(lociSet = case_when(
    str_detect(Seq.ID, "INDID") ~ "INDID", # (all are INDID anyway)
    .default = NA
  ))

sum(mfeOutput_2$inSilico == "PASS") # 12
sum(mfeOutput_2$inSilico == "FAIL") # 21
```

Loci that failed (n = 21) include the following counts per lociSet:

```{r}
mfeOutput_2 %>%
  filter(inSilico == "FAIL") %>%
  group_by(lociSet) %>%
  summarise(count = n())
```

### 11) bp3error

**Quick-load**

```{r}
bp3errorTracking <- read.csv("./00_tamRun1/bp3error_investigation/bp3error_master_1July2024.csv")
```

```{r}
bp3errorTracking %>%
  filter(pf_bp3error == "fail") %>%
  group_by(set) %>%
  summarise(count = n()) %>%
  adorn_totals()
```


**figuring shit out (not cleaned up)**

Acording to Sam's notes/thesis/etc., this is how things break down:
-   started with 302 loci
-   bp3 error for 99
-   found new variants w/in amplicon for 23 (99 - 23 = 76)
-   used MSA files from tamRun1 to find variants for another 36 (76 - 36 = 40)

According to my own investigations, we have:

           outcome count
 bp3error_noRescue    48
  bp3error_rescue1    33
  bp3error_rescue2    28
           noError   193
             Total   302


```{r}
# 99 loci w/bp3 error
bp3error_99Loci <- read.csv("./00_tamRun1/bp3error_investigation/failing_snps1.csv", header = F) %>%
  merge(., seqNames_to_shortNames[, c("seqID1", "primerName2")], by.x = "V1", by.y = "seqID1", all.x = T) %>%
  dplyr::rename("seqID" = "V1",
                "locus" = "primerName2") %>%
  mutate(locus = sub('\\..*', '', locus))

# 226 loci, including some number of bp3error loci that were part of rescue1 (Sam found other informative variant in amplicon)
pp_bp3clean <- read.table("./00_tamRun1/03_run1GTscore/primerProbeFile.txt", header = T) %>%
  mutate(
    Locus = case_when(
      Locus == "SEXID_195.3" ~ "SEXID_LWED_195.3",
      Locus == "SEXID_208.3" ~ "SEXID_LWED_208.3",
      Locus == "SEXID_211.3" ~ "SEXID_LWED_211.3",
      
      Locus == "SEXID_197.1" ~ "SEXID_SIMP_197.1",
      Locus == "SEXID_198.3" ~ "SEXID_SIMP_198.3",
      Locus == "SEXID_203.2" ~ "SEXID_SIMP_203.2",
      Locus == "SEXID_218.1" ~ "SEXID_SIMP_218.1",
      
      .default = Locus
    )
  ) %>%
  mutate(Locus = sub('\\..*', '', Locus)) %>%
  mutate(
    fileSet = "set226Loci"
  )

# 76 loci w/bp3 error that couldn't be "rescued" immediately; but Sam apparently used MSA sequences to identify other variants for some number of them
pp_bp3error <- read.table("./00_tamRun1/03_run1GTscore/bp3error_primerProbeFile.txt", header = T) %>%
  mutate(
    Locus = case_when(
      Locus == "SEXID_195.3" ~ "SEXID_LWED_195.3",
      Locus == "SEXID_208.3" ~ "SEXID_LWED_208.3",
      Locus == "SEXID_211.3" ~ "SEXID_LWED_211.3",
      
      Locus == "SEXID_197.1" ~ "SEXID_SIMP_197.1",
      Locus == "SEXID_198.3" ~ "SEXID_SIMP_198.3",
      Locus == "SEXID_203.2" ~ "SEXID_SIMP_203.2",
      Locus == "SEXID_218.1" ~ "SEXID_SIMP_218.1",
      
      .default = Locus
    )
  ) %>%
  mutate(Locus = sub('\\..*', '', Locus)) %>%
  mutate(
    fileSet = "set76Loci"
  )
```

```{r}
# 35 loci w/bp3 error that could be rescued
bp3error_rescued <- read.csv("./00_tamRun1/bp3error_investigation/failing_snps_that_have_other_variants.csv", header = F) %>%
  merge(., seqNames_to_shortNames[, c("seqID1", "primerName3")], by.x = "V1", by.y = "seqID1", all.x = T) %>%
  select(-V2) %>%
  dplyr::rename("seqName" = "V1",
                "locus" = "primerName3") %>%
  mutate(
    locus = case_when(
      locus == "SEXID_195" ~ "SEXID_LWED_195",
      locus == "SEXID_208" ~ "SEXID_LWED_208",
      locus == "SEXID_211" ~ "SEXID_LWED_211",
      
      locus == "SEXID_197" ~ "SEXID_SIMP_197",
      locus == "SEXID_198" ~ "SEXID_SIMP_198",
      locus == "SEXID_203" ~ "SEXID_SIMP_203",
      locus == "SEXID_218" ~ "SEXID_SIMP_218",
      
      .default = locus
    )
  ) %>%
  mutate(
    rescueSet = case_when(
      locus %in% pp_bp3clean$Locus ~ "rescue1",
      locus %in% pp_bp3error$Locus ~ "rescue2",
      .default = "noRescue"
    )
  )
  
table(bp3error_rescued$rescueSet) 
```

**Loci/primer tracking**

```{r}
# dropped b/t panel v1 & v2
lociRemoved_v1.v2 <- tamRun1_primerProbeFile %>%
  filter(!Locus %in% pp_laLab$Locus) %>%
  select(Locus) %>%
  mutate(
    removedAfter = "opt1"
  )

# dropped b/t panel v2 & v3
lociRemoved_v2.v3 <- pp_laLab %>%
  filter(!Locus %in% pp_tamRun5$Locus) %>%
  select(Locus) %>%
  mutate(
    removedAfter = "opt2"
  )

lociRemoved <- rbind(lociRemoved_v1.v2, lociRemoved_v2.v3) %>%
  dplyr::rename("locus" = "Locus")
```

Plus associated metadata, including whether it was a bp3error locus, whether it was rescued, etc.

if loci were flagged for bp3 error, but pass pf_opt1, this means they were "rescued"; loci flagged for bp3 error but fail pf_opt1 = dropped

```{r}
snpTracking <- multifasta_1 %>%
  rbind(., multifasta_2) %>%
  distinct() %>%
  merge(., seqNames_to_shortNames[, c("seqID3", "primerName3")], by.x = "original", by.y = "seqID3", all.x = T) %>%
  dplyr::rename("seqID" = "original",
                "locus" = "primerName3") %>%
  # redo lociSet
  mutate(
    set = sub('[_][^_]+$', '', locus)
  ) %>%
  # add bp3error info
  mutate(bp3error = case_when(
      locus %in% bp3error_99Loci$locus ~ "bp3error",
      .default = NA
      )
    ) %>%
  merge(., bp3error_rescued[, c("locus", "rescueSet")], by = "locus", all.x = T) %>%
  # ID each seqID as pass/fail at each panel prep step
  mutate(
    pf_primerDesign = case_when(
      !is.na(locus) ~ "pass",
      .default = "fail"
    ),
    pf_opt1 = case_when(
      pf_primerDesign == "fail" ~ NA,
      pf_primerDesign == "pass" & locus %in% pp_laLab$Locus ~ "pass",
      .default = "fail"
    ),
    pf_bp3error = case_when(
      !is.na(bp3error) & !is.na(rescueSet) ~ "pass",
      !is.na(bp3error) & is.na(rescueSet) & pf_opt1 == "pass" ~ "pass",
      !is.na(bp3error) & is.na(rescueSet) ~ "fail",
      .default = NA
    )
  ) %>%
  merge(., lociRemoved, by = "locus", all.x = T) %>%
  mutate(
    pf_opt2 = case_when(
      is.na(pf_opt1) ~ NA,
      removedAfter == "opt1" ~ NA,
      removedAfter == "opt2" ~ "fail",
      .default = "pass"
    )
  ) %>%
  select(seqID, locus, set, pf_primerDesign, pf_bp3error, pf_opt1, pf_opt2)

table(snpTracking$pf_opt2)



snpTracking_table <- snpTracking %>%
  filter(!is.na(bp3error)) %>%
  filter(pf_bp3error == "fail")

table(snpTracking_table$pf_opt1)

table(snpTracking_table$set)
```

export bp3error loci tracking

```{r}
bp3errorTracking <- snpTracking %>%
  filter(!is.na(bp3error))

write.csv(bp3errorTracking, "./00_tamRun1/bp3error_investigation/bp3error_master_1July2024.csv", row.names = F)
```

## 4.3 Summary

asdfasdf

# 5 library_1 tamRun1

## 5.1 Overview




## 5.2 Data

### Sample metadata

**Data**

```{r}
run1_md <- read.csv("./00_tamRun1/03_run1GTscore/tamRun1_metadata_v2.csv")
```

**Metrics**

```{r}
run1_sampleOverview <- run1_md %>%
  mutate(
    temp = str_c(sampleType, sex, sep = "_")
  ) %>%
  group_by(species, temp) %>%
  summarise(count = n()) %>%
  mutate(
    species = case_when(
      species == "pcr1Neg" ~ "negCtrl",
      .default = species
    ),
    temp = case_when(
      is.na(temp) ~ "negCtrl",
      .default = temp
    )
  ) %>%
  pivot_wider(names_from = species,
              values_from = count) %>%
  separate(temp, into = c("sampleType", "sex"), sep = "_") %>%
  rowwise() %>%
  mutate(
    TOTAL = sum(LWED,SIMP,negCtrl, na.rm = T)
  ) %>%
  janitor::adorn_totals("row") %>%
  as.data.frame()

run1_sampleOverview
```

  sampleType  sex LWED SIMP negCtrl TOTAL
1      blood    F    4    9      NA    13
2      blood    M   11    5      NA    16
3       hair    F    2   NA      NA     2
4       hair    M   NA    1      NA     1
5    negCtrl <NA>   NA   NA       1     1
6      Total    -   17   15       1    33

```{r}
run1_md %>%
  group_by(sampleType) %>%
  summarise(count = n()) %>%
  as.data.frame()
```

  sampleType count
1      blood    29
2       hair     3
3    pcr1Neg     1

```{r}
str_c(run1_md %>%
  select(animalID) %>%
  filter(animalID != "pcr1Neg") %>%
  distinct() %>%
  nrow(), " unique individuals")
```

"22 unique individuals"

```{r}
str_c(run1_md %>%
  filter(species == "LWED") %>%
  select(animalID) %>%
  filter(animalID != "pcr1Neg") %>%
  distinct() %>%
  nrow(), " unique LWED")
```

"10 unique LWED"

```{r}
str_c(run1_md %>%
  filter(species == "SIMP") %>%
  select(animalID) %>%
  filter(animalID != "pcr1Neg") %>%
  distinct() %>%
  nrow(), " unique SIMP")
```

"12 unique SIMP"

```{r}
# duplicate & triplicate samples
run1_md %>%
  select(animalID, sampleType, sampleName) %>%
  filter(animalID != "pcr1Neg") %>%
  get_dupes(.) %>%
  distinct() %>%
  as.data.frame()
```

  animalID sampleType sampleName dupe_count
1       10      blood    lm_1183          3
2      182      blood    lm_1275          3
3       27      blood    lm_1191          3
4      203      blood     lm_700          2
5      216      blood    lm_1197          2
6      223       hair     env_15          2
7       73      blood     lm_505          2

### Loci metadata

I've imported the primer-probe file that contains ALL loci originally in the tamRun1 panel, including those with the BatchPrimer3 error regardless of whether they were rescued. 

**Data**

```{r}
run1_primerProbeFile <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/primerProbeFile_fullSet.txt", header = T) %>%
  arrange(Locus)
```

**Metrics**

```{r}
run1_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "SEXID_LWED") ~ "sexSNP_lwed",
      str_detect(Locus, "SEXID_SIMP") ~ "sexSNP_simp",
      str_detect(Locus, "SEXID") ~ "sexSNP_dual",
      str_detect(Locus, "INDID") ~ "indSNP_dual",
      str_detect(Locus, "LWED") ~ "indSNP_lwed",
      str_detect(Locus, "SIMP") ~ "indSNP_simp",
      str_detect(Locus, "SPECIES") ~ "spSNP_dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
```

     lociSet count
 indSNP_dual   151
 indSNP_lwed    39
 indSNP_simp    35
 sexSNP_dual    11
 sexSNP_lwed     3
 sexSNP_simp     4
  spSNP_dual    18
       Total   261

```{r}
run1_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "LWED") ~ "lwed",
      str_detect(Locus, "SIMP") ~ "simp",
      .default = "dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
```

 lociSet count
    dual   180
    lwed    42
    simp    39
   Total   261

### Sample/loci lists

```{r}
# sample lists
run1_posSamples <- run1_md %>%
  filter(species %in% c("LWED", "SIMP"))

run1_negSamples <- run1_md %>%
  filter(!sampleID %in% run1_posSamples$sampleID)

run1_lwedSamples <- run1_md %>%
  filter(species == "LWED")

run1_simpSamples <- run1_md %>%
  filter(species == "SIMP")

# loci lists
run1_lwedLoci <- run1_primerProbeFile %>%
  filter(!str_detect(Locus, "SIMP"))

run1_simpLoci <- run1_primerProbeFile %>%
  filter(!str_detect(Locus, "LWED"))
```

### Genos

Genos assigned using GTseq_Genotyper_v2.pl

```{r}
# gtseq
run1_genos_gtseq <- 
  # need read_csv vs. read.csv here
  read_csv("00_tamRun1/05_run1GTseq/tamRun1_compiledGenos.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  column_to_rownames("Sample") %>%
  select(c(run1_primerProbeFile$Locus)) %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), ~ str_c(str_sub(., 1, 1), ",", str_sub(., 2, 2)))) %>%
  # recode 0,0 as NA
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))

# gtscore
run1_genos_gtscore0x <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

run1_genos_gtscore10x <- read.table("./00_tamRun1/04_run1GTscore_bp3errorRemoved/fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")
```

### Summaries

**metrics reflect non-negative, target-species only**

```{r}
# gtseq
run1_sampleSum_gtseq <- # need read_csv vs. read.csv here
  read_csv("00_tamRun1/05_run1GTseq/tamRun1_compiledGenos.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  merge(., run1_md[, c("sampleID", "sampleID_md")], by.x = "Sample", by.y = "sampleID") %>%
  relocate(sampleID_md, .after = Sample)

# gtscore
run1_locusSum_gtscore <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/summaryFiles/master_locusSummary.csv")

run1_sampleSum_gtscore <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/summaryFiles/master_sampleSummary.csv")
```

**metrics include other combos - fullset samples/loci, etc.**

```{r}
run1_locusSum_gtscore_lwed_simpSpec <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/LWED_simpSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples") %>%
  relocate(sampleSet, .after = Locus)

run1_locusSum_gtscore_simp_lwedSpec <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/SIMP_lwedSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples") %>%
  relocate(sampleSet, .after = Locus)

run1_sampleSum_gtscore_fullSet <- read.csv("./00_tamRun1/04_run1GTscore_bp3errorRemoved/fullSet_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., run1_md[, c("sampleID", "species")], by.x = "Sample", by.y = "sampleID") %>%
  relocate(species, .after = Sample)
```

## 5.3 Read metrics

### totalReads

```{r}
# gtseq
sum(run1_sampleSum_gtseq$Raw.Reads) # 1885084

# gtscore
sum(run1_sampleSum_gtscore_fullSet$Total.Reads) # 1885084
```

### otReads

```{r}
run1_otReads_gtseq <- run1_sampleSum_gtseq %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  summarise(otReads = sum(On.Target.Reads))

run1_totalReads_gtseq <- run1_sampleSum_gtseq %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  summarise(otReads = sum(Raw.Reads))

(run1_otReads_gtseq / run1_totalReads_gtseq)*100
```


### bySample

**all samples + sp-specific loci sets**

 lociSet otProp otReads primerProp primerReads totalReads
 fullSet   0.01      13       0.12        1049       9048
 lwedSet   0.56  145480       0.25      257509    1030318
 simpSet   0.73  174212       0.28      237728     845718
   Total   0.64  319705       0.26      496286    1885084

```{r}
run1_sampleSum_gtscore %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**pos only (no negs) + sp-specific loci sets**

 lociSet otProp otReads primerProp primerReads totalReads
 lwedSet   0.56  145480       0.25      257509    1030318
 simpSet   0.73  174212       0.28      237728     845718
   Total   0.65  319692       0.26      495237    1876036

```{r}
run1_sampleSum_gtscore %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**pos only (no negs) + all loci**

```{r}
tamRun1_gtscore_sampleSum_lociAll %>%
  filter(!str_detect(species, "Neg")) %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(otProp, otReads, primerProp, primerReads, totalReads)
```

### byLocus

#### Overview

   sampleSet  lociSet otProp otReads primerReads
 lwedSamples lwedSpec   0.95   32991       34634
  posSamples     dual   0.57  222408      391347
 simpSamples simpSpec   0.93   64293       69256
       Total    Total   0.65  319692      495237

```{r}
run1_locusSum_gtscore %>%
  filter(metricUse == "for_panelMetrics") %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  mutate(
    lociSet = case_when(
      sampleSet == "lwedSamples" ~ "lwedSpec",
      sampleSet == "simpSamples" ~ "simpSpec",
      sampleSet == "posSamples" ~ "dual",
      .default = sampleSet
    )
  ) %>%
  relocate(sampleSet, lociSet, otProp, otReads, primerReads)
```

#### dualLoci

lwed vs. simp performance

```{r}
tamRun1_gtscore_locusSum %>%
  filter(sampleSet != "posSamples") %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP")) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### lwedLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun1_gtscore_locusSum %>%
  filter(str_detect(Locus, "LWED")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun1_gtscore_locusSum_simp_lwedSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("lwedSamples", "lwedSamples*", sampleSet)
  )
```

#### simpLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun1_gtscore_locusSum %>%
  filter(str_detect(Locus, "SIMP")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun1_gtscore_locusSum_lwed_simpSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("simpSamples", "simpSamples*", sampleSet)
  )
```

## 5.4 Geno success

### bySample

```{r}
temp1 <- tamRun1_gtscore_sampleSum %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "fullSet")

temp2 <- tamRun1_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "blood")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "blood")

temp3 <- tamRun1_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "hair")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "hair")

rbind(temp1, temp2, temp3) %>%
  relocate(sampleSet, n)
```

### byLocus

#### Overview

```{r}
tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "lwedSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "simpSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )
```

#### dualLoci



#### lwedLoci

**data**

```{r}
tamRun1_genoSuccess_gtscore0x_lwedSpec <- tamRun1_gtscore0x_genos %>%
  select(tamRun1_posSamples$sampleID) %>%
  filter(str_detect(rownames(.), "LWED")) %>%
  mutate(across(everything(), \(x) as.character(x))) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  mutate(
    
    totalGenos_lwed = rowSums(!is.na(select(., tamRun1_lwedSamples$sampleID))),
    totalGenos_simp = rowSums(!is.na(select(., tamRun1_simpSamples$sampleID))),
    
    propSuccess_lwed = round(totalGenos_lwed/nrow(tamRun1_lwedSamples), 2),
    propSuccess_simp = round(totalGenos_simp/nrow(tamRun1_simpSamples), 2)
    
  ) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    
    uniqueGenos_lwed = n_distinct(c_across(tamRun1_lwedSamples$sampleID), na.rm = T),
    uniqueGenos_simp = n_distinct(c_across(tamRun1_simpSamples$sampleID), na.rm = T)
    
  ) %>%
  relocate(contains("tamOpt"), .after = last_col()) %>%
  relocate(contains("uniqueGenos"), .after = locus)
```

**metrics**

```{r}
mean(tamRun1_genoSuccess_gtscore0x_lwedSpec$propSuccess_lwed)
```
```{r}
mean(tamRun1_genoSuccess_gtscore0x_lwedSpec$propSuccess_simp)
```

#### simpLoci

```{r}
tamRun1_genoSuccess_gtscore0x_simpSpec <- tamRun1_gtscore0x_genos %>%
  select(tamRun1_posSamples$sampleID) %>%
  filter(str_detect(rownames(.), "SIMP")) %>%
  mutate(across(everything(), \(x) as.character(x))) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  mutate(
    
    totalGenos_lwed = rowSums(!is.na(select(., tamRun1_lwedSamples$sampleID))),
    totalGenos_simp = rowSums(!is.na(select(., tamRun1_simpSamples$sampleID))),
    
    propSuccess_lwed = round(totalGenos_lwed/nrow(tamRun1_lwedSamples), 2),
    propSuccess_simp = round(totalGenos_simp/nrow(tamRun1_simpSamples), 2)
    
  ) %>%
  rownames_to_column("locus") %>%
  rowwise() %>%
  mutate(
    
    uniqueGenos_lwed = n_distinct(c_across(tamRun1_lwedSamples$sampleID), na.rm = T),
    uniqueGenos_simp = n_distinct(c_across(tamRun1_simpSamples$sampleID), na.rm = T)
    
  ) %>%
  relocate(contains("tamOpt"), .after = last_col()) %>%
  relocate(contains("uniqueGenos"), .after = locus)
```

**metrics**

```{r}
mean(tamRun1_genoSuccess_gtscore0x_simpSpec$propSuccess_lwed)
```

```{r}
mean(tamRun1_genoSuccess_gtscore0x_simpSpec$propSuccess_simp)
```

## 5.5 Geno concordance

```{r}
run1_dups <- run1_md %>%
  get_dupes(animalID)

run1_genoCompare_dups <- run1_genos_gtseq %>%
  select(run1_dups$sampleID) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., run1_md[, c("sampleID", "animalID", "sampleType")], by = "sampleID") %>%
  mutate(
    animalID = str_c(animalID, sampleType, sep = "_")
  ) %>%
  select(-sampleType) %>%
  relocate(animalID) %>%
  arrange(animalID, sampleID) %>%
  group_by(animalID) %>%
  mutate(
    dupID = ifelse(duplicated(animalID), paste0("dup", cumsum(duplicated(animalID))+1), "dup1")
    )%>%
  relocate(dupID) %>%
  select(-sampleID) %>%
  pivot_longer(!c(animalID, dupID),
               names_to = "locus",
               values_to = "genos") %>%
  pivot_wider(id_cols = c(animalID, locus),
              names_from = dupID,
              values_from = genos) %>%
  mutate(
    dup12Match = dup1 == dup2,
    dup13Match = dup1 == dup3,
    dup23Match = dup2 == dup3
  )

temp <- run1_genoCompare_dups %>%
  group_by(animalID) %>%
  summarise(
    genosCommon12 = sum(!is.na(dup12Match)),
    genosMatch12 = sum(dup12Match == TRUE, na.rm = T),
    
    genosCommon13 = sum(!is.na(dup13Match)),
    genosMatch13 = sum(dup13Match == TRUE, na.rm = T),
    
    genosCommon23 = sum(!is.na(dup23Match)),
    genosMatch23 = sum(dup23Match == TRUE, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(
    propMatch12 = genosMatch12/genosCommon12,
    propMatch13 = genosMatch13/genosCommon13,
    propMatch23 = genosMatch23/genosCommon23
  )

temp2 <- temp %>%
  select(animalID, contains("prop")) %>%
  pivot_longer(!animalID,
               names_to = "id",
               values_to = "propMatch") %>%
  filter(propMatch > 0)

mean(temp2$propMatch) # 0.9821244
min(temp2$propMatch) # 0.9557522
max(temp2$propMatch) # 1
median(temp2$propMatch) # 0.986201
```


## 5.6 Panel optimization summary

Removed...

-   41 loci due to bp3 error
-   9 variants not producing genotypes (based on GTseq analysis)



```{r}
tamRun1_lociFail_bp3error <- bp3errorTracking %>%
  filter(pf_bp3error == "fail") %>%
  select(locus)

tamRun1_lociFail_zeroReads <- data.frame(
  locus = c("INDID_23",
            "INDID_51",
            "INDID_58",
            "INDID_96",
            "LWED_287",
            "SPECIES_3",
            "SPECIES_6",
            "SPECIES_7",
            "SPECIES_10")
)

rbind(tamRun1_lociFail_zeroReads, tamRun1_lociFail_bp3error) %>%
  nrow()

View(tamRun1_gtscore_locusSum %>%
       arrange(GenotypeRate_0x))

temp <- tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  select(Locus, sampleSet, AvgReadDepth, GenotypeRate_0x, GenotypeRate_10x, allFreqs_0x, allFreqs_10x) %>%
  dplyr::rename("locus" = "Locus") %>%
  mutate(
    AvgReadDepth = round(AvgReadDepth, 2),
    GenotypeRate_0x = round(GenotypeRate_0x, 2),
    GenotypeRate_10x = round(GenotypeRate_10x, 2)
  ) %>%
  merge(., lociRemoved, by = "locus", all.x = T)
```

# 6 library_2 - run3 metrics

## 6.1 Overview

asdf

## 6.2 Data

### Sample metadata

this is md_run3_v3; see sandbox GTscore_tamRun5_plus_run3.Rmd for details on samples vs. metadata

**Data**

```{r}
run3_md <- read.csv("./03_tamRun3/03_run3GTscore/tamRun3_metadata_abcat_5June2023.csv") #%>%
  #filter(!str_detect(sampleID, "cat")) %>%
  #filter(!str_detect(sampleID, "3b"))
```

**Metrics**

```{r}
run3_sampleOverview <- run3_md %>%
  mutate(
    temp = str_c(sampleType, sex, sep = "_")
  ) %>%
  group_by(species, temp) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = species,
              values_from = count) %>%
  separate(temp, into = c("sampleType", "sex"), sep = "_") %>%
  rowwise() %>%
  mutate(
    TOTAL = sum(LWED,SIMP,pcr1Neg,xtnNeg, na.rm = T)
  ) %>%
  janitor::adorn_totals("row") %>%
  as.data.frame()

run3_sampleOverview
```

```{r}
run3_md %>%
  group_by(sampleType) %>%
  summarise(count = n()) %>%
  as.data.frame()
```

```{r}
str_c(run3_md %>%
  select(animalID) %>%
  filter(!str_detect(animalID, "Neg")) %>%
  distinct() %>%
  nrow(), " unique individuals")
```
```{r}
str_c(run3_md %>%
  filter(species == "LWED") %>%
  select(animalID) %>%
  filter(!str_detect(animalID, "Neg")) %>%
  distinct() %>%
  nrow(), " unique LWED")
```

```{r}
str_c(run3_md %>%
  filter(species == "SIMP") %>%
  select(animalID) %>%
  filter(!str_detect(animalID, "Neg")) %>%
  distinct() %>%
  nrow(), " unique SIMP")
```

### Loci metadata

**Data**

```{r}
run3_primerProbeFile <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt", header = T)
```

**Metrics**

```{r}
run3_lociOverview <- run3_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "SEXID_LWED") ~ "sexSNP_lwed",
      str_detect(Locus, "SEXID_SIMP") ~ "sexSNP_simp",
      str_detect(Locus, "SEXID") ~ "sexSNP_dual",
      str_detect(Locus, "INDID") ~ "indSNP_dual",
      str_detect(Locus, "LWED") ~ "indSNP_lwed",
      str_detect(Locus, "SIMP") ~ "indSNP_simp",
      str_detect(Locus, "SPECIES") ~ "spSNP_dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
  
run3_lociOverview
```

### Sample/loci lists

```{r}
# sample lists
run3_posSamples <- run3_md %>%
  filter(species %in% c("LWED", "SIMP"))

run3_negSamples <- run3_md %>%
  filter(!sampleID %in% run3_posSamples$sampleID)

run3_lwedSamples <- run3_md %>%
  filter(species == "LWED")

run3_simpSamples <- run3_md %>%
  filter(species == "SIMP")

# loci lists
run3_lwedLoci <- run3_primerProbeFile %>%
  filter(!str_detect(Locus, "SIMP"))

run3_simpLoci <- run3_primerProbeFile %>%
  filter(!str_detect(Locus, "LWED"))
```

### Allele reads

**gtscore**

```{r}
run3a_readCounts_gtscore <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

run3b_readCounts_gtscore <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")
```

**gtseq**

```{r}

```

### Genos

**gtscore**

```{r}
# run3a
run3a_gtscore0x_genos <- read.table("./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

run3a_gtscore10x_genos <- read.table("./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

# run3b
run3b_gtscore0x_genos <- read.table("./03_tamRun3/03_run3GTscore/fullSet_b_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

run3b_gtscore10x_genos <- read.table("./03_tamRun3/03_run3GTscore/fullSet_b_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")
```

**gtseq**

```{r}

```

### Summaries

**metrics reflect non-negative, target-species only**

```{r}
run3_gtscore_locusSum <- read.csv("./03_tamRun3/03_run3GTscore/summaryFiles/master_locusSummary_abcat.csv") %>%
  filter(str_detect(sampleSet, "_a"))

run3_gtscore_sampleSum <- read.csv("./03_tamRun3/03_run3GTscore/summaryFiles/master_sampleSummary_abcat.csv") %>%
  filter(!str_detect(sampleID, "3b")) %>%
  filter(!str_detect(sampleID, "cat_"))
```

**metrics include other combos - fullset samples/loci, etc.**

```{r}
run3_gtscore_locusSum_lwed_simpSpec <- read.csv("./03_tamRun3/03_run3GTscore/LWED_simpSpec_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples") %>%
  relocate(sampleSet, .after = Locus)

run3_gtscore_locusSum_simp_lwedSpec <- read.csv("./03_tamRun3/03_run3GTscore/SIMP_lwedSpec_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples") %>%
  relocate(sampleSet, .after = Locus)

run3_gtscore_sampleSum_fullSet <- read.csv("./03_tamRun3/03_run3GTscore/fullSet_a_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., run3_md[, c("sampleID", "species")], by.x = "Sample", by.y = "sampleID") %>%
  relocate(species, .after = Sample)
```

## 6.3 Read metrics

### bySample

**with negs**

```{r}
run3_gtscore_sampleSum %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**without negs**

```{r}
run3_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

### byLocus

#### Overview

```{r}
run3_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### dualLoci

lwed vs. simp vs. both performance

```{r}
run3_gtscore_locusSum %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP")) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### lwedLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
run3_gtscore_locusSum %>%
  filter(str_detect(Locus, "LWED")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., run3_gtscore_locusSum_simp_lwedSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("lwedSamples_a", "lwedSamples*", sampleSet)
  )
```

#### simpLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
run3_gtscore_locusSum %>%
  filter(str_detect(Locus, "SIMP")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., run3_gtscore_locusSum_lwed_simpSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("simpSamples_a", "simpSamples*", sampleSet)
  )
```

## 6.4 Geno success

### bySample

```{r}
# all samples
temp1 <- run3_gtscore_sampleSum %>%
  filter(!str_detect(sampleID_md, "Neg")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "fullSet")

temp2 <- run3_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "blood")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "blood")

temp3 <- run3_gtscore_sampleSum %>%
  filter(str_detect(sampleID_md, "hair")) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    n = nrow(.)
  ) %>%
  as.data.frame() %>%
  mutate(sampleSet = "hair")

rbind(temp1, temp2, temp3) %>%
  relocate(sampleSet, n)
```

### byLocus

```{r}
tamRun1_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "lwedSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )

tamRun1_gtscore_locusSum %>%
  filter(sampleSet == "simpSamples") %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2),
    mean_depth = round(mean(AvgReadDepth, na.rm = T), 2)
  )
```

# X GTseq vs. GTscore

## x.1 Overview

Compare run3a/b w/both pipelines

## x.2 Data

### Allele reads

```{r}
# gtseq


# gtscore
run3a_readCounts_gtscore <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

run3b_readCounts_gtscore <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")
```

### Genos

```{r}
# gtseq
run3ab_genos_gtseq <- read_csv("./03_tamRun3/05_run3GTseq/tamRun3ab_compiledGenos.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  )

run3a_genos_gtseq <- run3ab_genos_gtseq %>%
  filter(!str_detect(Sample, "3b")) %>%
  column_to_rownames("Sample") %>%
  select(run3_primerProbeFile$Locus) %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), ~ str_c(str_sub(., 1, 1), ",", str_sub(., 2, 2)))) %>%
  # recode 0,0 as NA
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))

run3b_genos_gtseq <- run3ab_genos_gtseq %>%
  filter(str_detect(Sample, "3b")) %>%
  mutate(Sample = gsub("tamRun3b", "tamRun3", Sample)) %>%
  column_to_rownames("Sample") %>%
  select(run3_primerProbeFile$Locus) %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), ~ str_c(str_sub(., 1, 1), ",", str_sub(., 2, 2)))) %>%
  # recode 0,0 as NA
  mutate(across(everything(), ~ 
                  case_when(. == "0,0" ~ gsub("0,0", NA, .),
                            .default = .)
                ))

# gtscore
run3a_genos_gtscore0x <- read.table("./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus")

run3b_genos_gtscore0x <- read.table("./03_tamRun3/03_run3GTscore/fullSet_b_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = gsub("tamRun3b", "tamRun3", sampleID)) %>%
  column_to_rownames("sampleID") %>%
  t() %>%
  as.data.frame()
```

### Summaries

```{r}
# gtseq
run3a_sampleSum_gtseq <- run3ab_genos_gtseq %>%
  filter(!str_detect(Sample, "3b"))

run3b_sampleSum_gtseq <- run3ab_genos_gtseq %>%
  filter(str_detect(Sample, "3b")) %>%
  mutate(Sample = gsub("tamRun3b", "tamRun3", Sample))

# gtscore
run3a_sampleSum_gtscore <- read.csv("./03_tamRun3/03_run3GTscore/fullSet_a_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample))

run3b_sampleSum_gtscore <- read.csv("./03_tamRun3/03_run3GTscore/fullSet_b_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("tamRun3b-", "tamRun3.", Sample))
```

## x.3 readCompare

```{r}
# total reads
sum(run3a_sampleSum_gtseq$Raw.Reads) # 23654746
sum(run3b_sampleSum_gtseq$Raw.Reads) # 20799170

sum(run3a_sampleSum_gtscore$Total.Reads) # 23654746
sum(run3b_sampleSum_gtscore$Total.Reads) # 20799170

# on-target reads
run3_sampleList_negs <- run3_md %>%
  filter(str_detect(sampleType, "Neg"))
run3_sampleList_pos <- run3_md %>%
  filter(!str_detect(sampleType, "Neg"))

## gtseq = 7133024
run3a_readCounts_gtseq %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts") %>%
  filter(!sampleID %in% run4_sampleList_negs$sampleID_ill) %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  mutate(readSums = as.numeric(get_readSum(readCounts))) %>%
  summarise(
    otReads = sum(readSums)
  )

## gtscore = 7091829
run4_readCounts_gtscore %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts") %>%
  filter(!sampleID %in% run4_sampleList_negs$sampleID_ill) %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  mutate(readSums = as.numeric(get_readSum(readCounts))) %>%
  summarise(
    otReads = sum(readSums)
  )

# percent difference = |V1 - V2| / (V1 + V2)/2 * 100 = 0.5791975%
( abs(7133024 - 7091829) / ( (7133024 + 7091829)/2) ) * 100
```

## x.4 genoCompare

Comparing a vs. b runs:

gtseq:
  genosCommon genosMatch genosMismatch propMismatch
1       37152      36874           278  0.007482773 -> 0.75%

gtscore0x:
  genosCommon genosMatch genosMismatch propMismatch
1       48603      46882          1721   0.03540934 -> 3.5%

gtscore10x:
  genosCommon genosMatch genosMismatch propMismatch
1       38353      37735           618   0.01611347 > 1.6%

```{r}
# gtseq a vs. b - propMismatch = 0.007482773
run3_genoCompare_gta.gtb <- get_genoCompare(run3a_genos_gtseq, run3b_genos_gtseq)

run3_genoCompare_gta.gtb %>%
  summarise(
    genosCommon = sum(!is.na(match)),
    genosMatch = sum(match == TRUE, na.rm = T),
    genosMismatch = sum(match == FALSE, na.rm = T)
  ) %>%
  mutate(propMismatch = genosMismatch/genosCommon)

# gtscore a vs. b - propMismatch = 0.03540934
run3_genoCompare_gra.grb <- get_genoCompare(run3a_genos_gtscore0x, run3b_genos_gtscore0x)

run3_genoCompare_gra.grb %>%
  summarise(
    genosCommon = sum(!is.na(match)),
    genosMatch = sum(match == TRUE, na.rm = T),
    genosMismatch = sum(match == FALSE, na.rm = T)
  ) %>%
  mutate(propMismatch = genosMismatch/genosCommon)


temp <- run3b_gtscore10x_genos %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = gsub("tamRun3b", "tamRun3", sampleID)) %>%
  column_to_rownames("sampleID") %>%
  t() %>%
  as.data.frame()

test <- get_genoCompare(run3a_gtscore10x_genos, temp)

test %>%
  summarise(
    genosCommon = sum(!is.na(match)),
    genosMatch = sum(match == TRUE, na.rm = T),
    genosMismatch = sum(match == FALSE, na.rm = T)
  ) %>%
  mutate(propMismatch = genosMismatch/genosCommon)
```



### Functions

Calculate via genotyping discordance b/t technical replicates in run3a and run3b

```{r}
# readSum
get_readSum <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# genoError
get_genoError <- function(sampleIDs, genoFile_a, genoFile_b, readCounts_a, readCounts_b) {
  
  # long-format genos
  longGenos_a <- genoFile_a %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_a") %>%
  pivot_longer(!sampleID_a, names_to = "locus", values_to = "genos_a") %>%
  mutate(genos_a = na_if(genos_a, "0"))
  
  longGenos_b <- genoFile_b %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID_b") %>%
  pivot_longer(!sampleID_b, names_to = "locus", values_to = "genos_b") %>%
  mutate(genos_b = na_if(genos_b, "0"))
  
  # long-format allele readSums
  longReads_a <- readCounts_a %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("sampleID_a") %>%
    pivot_longer(!sampleID_a, names_to = "locus", values_to = "readCounts_a") %>%
    mutate(readSums_a = get_readSum(readCounts_a))
  
  longReads_b <- readCounts_b %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("sampleID_b") %>%
    pivot_longer(!sampleID_b, names_to = "locus", values_to = "readCounts_b") %>%
    mutate(readSums_b = get_readSum(readCounts_b))
  
  # compare genos
  comp_df <- sampleIDs %>%
    merge(., longGenos_a, by = "sampleID_a") %>%
    merge(., longGenos_b, by = c("sampleID_b", "locus")) %>%
    mutate(genoMatch = case_when(
      is.na(genos_a) | is.na(genos_b) ~ NA,
      genos_a == genos_b ~ "yes",
      .default = "no"
      )
    ) %>%
    merge(., longReads_a, by = c("sampleID_a", "locus")) %>%
    merge(., longReads_b, by = c("sampleID_b", "locus"))
  
  summary_bySample <- comp_df %>%
    group_by(sample) %>%
    summarise(genosTotal = length(locus),
              genosCommon = sum(!is.na(genoMatch)),
              genosMismatch = sum(genoMatch == "no", na.rm = T)) %>%
    as.data.frame() %>%
    mutate(
      errorRate = (genosMismatch/genosCommon)*100,
      propCommon = (genosCommon/genosTotal)
      )
  
  summary_byLocus <- comp_df %>%
    group_by(locus) %>%
    summarise(genosTotal = length(sample),
              genosCommon = sum(!is.na(genoMatch)),
              genosMismatch = sum(genoMatch == "no", na.rm = T)) %>%
    as.data.frame() %>%
    mutate(
      errorRate = (genosMismatch/genosCommon)*100,
      propCommon = (genosCommon/genosTotal)
      )
    
  
  return(list(comp_df, summary_bySample, summary_byLocus))
   
}

# gtseq alleleRatio
get_gtseqRatio <- function(readCounts) {
  # Function to process a single read count
  process_read_count <- function(rc) {
    # Check if the read count is NA
    if (is.na(rc)) {
      return(NA)
    }
    
    # Split the input string into two numeric values
    counts <- as.numeric(unlist(strsplit(rc, ",")))
    
    # Replace 0 counts with 0.1
    counts[counts == 0] <- 0.1
    
    # Sum the read counts
    readSum <- sum(counts)
    
    # Check the sum and compute the ratio if appropriate
    if (readSum < 10) {
      return(NA)
    } else {
      return(counts[1] / counts[2])
    }
  }
  
  # Apply the processing function to each element in readCounts
  sapply(readCounts, process_read_count)
}

# gtseq genos; based on GTseq_Genotyper_v2.pl
get_gtseqGenos_v2 <- function(ratio, Allele1, Allele2) {
  
  # round ratio to 3 places
  ratio <- round(ratio, 3)
  
  result <- case_when(
    is.na(ratio) ~ NA_character_,
    ratio >= 10 ~ str_c(Allele1, Allele1, sep = ","),
    ratio <= 0.1 ~ str_c(Allele2, Allele2, sep = ","),
    ratio <= 0.2 ~ NA_character_,
    ratio <= 5 ~ str_c(Allele1, Allele2, sep = ","),
    TRUE ~ NA_character_  # Default case
  )
  return(result)
}
```

### Calculate genoError

```{r}
# sampleID file
genoError_sampleIDs <- run3_md %>%
  select(sampleID, sampleID_md) %>%
  dplyr::rename("sample" = "sampleID_md",
                "sampleID_a" = "sampleID") %>%
  mutate(sampleID_b = gsub("run3", "run3b", sampleID_a)) %>%
  relocate(sample)

# gtscore0x
run3_gtscore0x_genoError <- get_genoError(genoError_sampleIDs, run3_gtscore0x_genos, run3b_gtscore0x_genos, run3a_readCounts_gtscore, run3b_readCounts_gtscore)

run3_gtscore0x_genoError_compDF <- run3_gtscore0x_genoError[1] %>%
  as.data.frame()

run3_gtscore0x_genoError_bySample <- run3_gtscore0x_genoError[2] %>%
  as.data.frame() %>%
  filter(propCommon >= 0.5) %>%
  mutate(pipeline = "gtscore0x")

run3_gtscore0x_genoError_byLocus <- run3_gtscore0x_genoError[3] %>%
  as.data.frame() %>%
  filter(propCommon >= 0.5) %>%
  mutate(pipeline = "gtscore0x")

View(run3_gtscore0x_genoError_compDF %>%
       filter(locus == "INDID_153"))

mean(run3_gtscore0x_genoError_bySample$errorRate) # 3.369163
mean(run3_gtscore0x_genoError_byLocus$errorRate) # 2.802502

# gtscore10x
run3_gtscore10x_genoError <- get_genoError(genoError_sampleIDs, run3_gtscore10x_genos, run3b_gtscore10x_genos, run3a_readCounts_gtscore, run3b_readCounts_gtscore)

run3_gtscore10x_genoError_compDF <- run3_gtscore10x_genoError[1] %>%
  as.data.frame()

run3_gtscore10x_genoError_bySample <- run3_gtscore10x_genoError[2] %>%
  as.data.frame() %>%
  filter(propCommon >= 0.5) %>%
  mutate(pipeline = "gtscore10x")

run3_gtscore10x_genoError_byLocus <- run3_gtscore10x_genoError[3] %>%
  as.data.frame() %>%
  filter(propCommon >= 0.5) %>%
  mutate(pipeline = "gtscore10x")

mean(run3_gtscore10x_genoError_bySample$errorRate) # 1.668994
mean(run3_gtscore10x_genoError_byLocus$errorRate) # 1.265634

# gtseq
run3_gtseq_genoError_compDF <- run3_gtscore0x_genoError_compDF %>%
  mutate(
    gtRatio_a = get_gtseqRatio(readCounts_a),
    gtRatio_b = get_gtseqRatio(readCounts_b)
    ) %>%
  merge(., run3_primerProbeFile[, c("Locus", "Allele1", "Allele2")], by.x = "locus", by.y = "Locus") %>%
  mutate(
    genos_a = get_gtseqGenos_v2(gtRatio_a, Allele1, Allele2),
    genos_b = get_gtseqGenos_v2(gtRatio_b, Allele1, Allele2)
  ) %>%
  mutate(genoMatch = case_when(
      is.na(genos_a) | is.na(genos_b) ~ NA,
      genos_a == genos_b ~ "yes",
      .default = "no"
      )
    )

run3_gtseq_genoError_bySample <- run3_gtseq_genoError_compDF %>%
  group_by(sample) %>%
    summarise(genosTotal = length(locus),
              genosCommon = sum(!is.na(genoMatch)),
              genosMismatch = sum(genoMatch == "no", na.rm = T)) %>%
    as.data.frame() %>%
    mutate(
      errorRate = (genosMismatch/genosCommon)*100,
      propCommon = (genosCommon/genosTotal)
      ) %>%
  as.data.frame() %>%
  filter(propCommon >= 0.5) %>%
  mutate(pipeline = "gtseq")

run3_gtseq_genoError_byLocus <- run3_gtseq_genoError_compDF %>%
  group_by(locus) %>%
    summarise(genosTotal = length(sample),
              genosCommon = sum(!is.na(genoMatch)),
              genosMismatch = sum(genoMatch == "no", na.rm = T)) %>%
    as.data.frame() %>%
    mutate(
      errorRate = (genosMismatch/genosCommon)*100,
      propCommon = (genosCommon/genosTotal)
      ) %>%
  as.data.frame() %>%
  filter(propCommon >= 0.5) %>%
  mutate(pipeline = "gtseq")
```

### Figures

```{r}
run3_gtscore0x_genoError_bySample %>%
  select(sample, errorRate, pipeline) %>%
  rbind(., run3_gtscore10x_genoError_bySample[, c("sample", "errorRate", "pipeline")]) %>%
  rbind(., run3_gtseq_genoError_bySample[, c("sample", "errorRate", "pipeline")]) %>%
  ggplot(aes(x = pipeline, y = errorRate, color = pipeline)) +
  geom_boxplot() +
  scale_colour_Publication() +
  theme_Publication() +
  labs(title = "Genotype discordance between L2 technical replicates across pipelines",
       y = "Percent discordant genotypes")
```



# 7 library_3 tamRun4

## 7.1 Overview



## 7.2 Data

### Sample metadata

**Data**

```{r}
run4_md <- read.csv("./04_tamRun4/tamRun4_metadata_illONT.csv")

View(run4_md %>%
       filter(str_detect(sampleID_unique, "fecal")))
```

**Metrics**

```{r}
run4_sampleOverview <- run4_md %>%
  mutate(
    temp = str_c(sampleType, sex, sep = "_")
  ) %>%
  mutate(
    species = case_when(
      str_detect(sampleType, "Neg") ~ "negCtrl",
      .default = species
    ),
    temp = case_when(
      str_detect(sampleType, "Neg") ~ "negCtrl",
      .default = temp
    )
  ) %>%
  group_by(species, temp) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = species,
              values_from = count) %>%
  separate(temp, into = c("sampleType", "sex"), sep = "_") %>%
  rowwise() %>%
  mutate(
    TOTAL = sum(LWED,SIMP,negCtrl, na.rm = T)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals("row")

run4_sampleOverview
```

### Loci metadata

**Data**

```{r}
run4_primerProbeFile <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)
```

**Metrics**

```{r}
run4_lociOverview <- run4_primerProbeFile %>%
  mutate(
    lociSet = case_when(
      str_detect(Locus, "SEXID_LWED") ~ "sexSNP_lwed",
      str_detect(Locus, "SEXID_SIMP") ~ "sexSNP_simp",
      str_detect(Locus, "SEXID") ~ "sexSNP_dual",
      str_detect(Locus, "INDID") ~ "indSNP_dual",
      str_detect(Locus, "LWED") ~ "indSNP_lwed",
      str_detect(Locus, "SIMP") ~ "indSNP_simp",
      str_detect(Locus, "SPECIES") ~ "spSNP_dual"
    )
  ) %>%
  group_by(lociSet) %>%
  summarise(count = n()) %>%
  as.data.frame() %>%
  adorn_totals("row")
  
run4_lociOverview
```

### Sample/loci lists

```{r}
# sample lists
tamRun4_posSamples <- tamRun4_md %>%
  filter(species %in% c("LWED", "SIMP"))

tamRun4_negSamples <- tamRun4_md %>%
  filter(!sampleID_ill %in% tamRun4_posSamples$sampleID_ill)

tamRun4_lwedSamples <- tamRun4_md %>%
  filter(species == "LWED")

tamRun4_simpSamples <- tamRun4_md %>%
  filter(species == "SIMP")

# loci lists
tamRun4_lwedLoci <- tamRun4_primerProbeFile %>%
  filter(!str_detect(Locus, "SIMP"))

tamRun4_simpLoci <- tamRun4_primerProbeFile %>%
  filter(!str_detect(Locus, "LWED"))
```

### Allele reads

```{r}
run4_readCounts_gtseq <- read.csv("./04_tamRun4/00_illumina/04_run4GTseq/gtseq_readCounts.csv") %>%
  select(c(Sample, contains("_"))) %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  column_to_rownames("Sample") %>%
  mutate(across(everything(), \(x) gsub("/", ",", x))) %>%
  t() %>%
  as.data.frame()

run4_readCounts_gtscore <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")
```

### Genos

Includes genos for gtscore0x, gtscore10x, and gtseq (genotyper_v2)

```{r, warning=FALSE}
# gtscore
run4_genos_gtscore0x <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

run4_genos_gtscore10x <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

# gtseq
run4_genos_gtseq <- get_gtseqGenos_v2(readCounts_df = run4_readCounts_gtseq,
                                      lociInfo = run4_primerProbeFile,
                                      mdFile = run4_md,
                                      sampleID_colName = "sampleID_ill",
                                      exclude_nonTargetSp = "yes")
```

### Summaries

**metrics reflect non-negative, target-species only**

```{r}
# gtscore
run4_gtscore_locusSum <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/summaryFiles/ill_master_locusSummary.csv")

run4_gtscore_sampleSum <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/summaryFiles/ill_master_sampleSummary.csv")

# gtseq
run4_gtseq_locusSum <- get_genoSuccess(genoFile = run4_genos_gtseq,
                                           lociFile = run4_primerProbeFile,
                                           mdFile = run4_md,
                                           by = "byLocus",
                                           sampleID_colName = "sampleID_ill",
                                           exclude_nonTargetSp = "yes")

run4_gtseq_sampleSum <- get_genoSuccess(genoFile = run4_genos_gtseq,
                                            lociFile = run4_primerProbeFile,
                                            mdFile = run4_md,
                                            by = "bySample",
                                            sampleID_colName = "sampleID_ill",
                                            exclude_nonTargetSp = "yes") %>%
  merge(., run4_md[, c("sampleID_ill", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill")
```

**metrics include other combos - fullset samples/loci, etc.**

```{r}
# gtscore
run4_gtscore_locusSum_lwed_simpSpec <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_lwed_simpSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples") %>%
  relocate(sampleSet, .after = Locus)

run4_gtscore_locusSum_simp_lwedSpec <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_simp_lwedSpec_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples") %>%
  relocate(sampleSet, .after = Locus)

run4_gtscore_locusSum_fullSet <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_GTscore_locusSummary.txt", sep = "\t")

run4_gtscore_sampleSum_fullSet <- read.csv("./04_tamRun4/00_illumina/03_run4GTscore/ill_fullSet_GTscore_individualSummary.txt", sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., tamRun4_md[, c("sampleID_ill", "species")], by.x = "Sample", by.y = "sampleID_ill") %>%
  relocate(species, .after = Sample)
```

## 7.3 Read metrics

### bySample

#### lociSets

**all samples + sp-specific loci sets**

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

**pos only (no negs) + sp-specific loci sets**

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```
**pos only (no negs) + all loci**

```{r}
tamRun4_ill_gtscore_sampleSum_lociAll %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(otProp, otReads, primerProp, primerReads, totalReads)
```

#### sampleTypes

**performance by sampleType only**

```{r}
tamRun4_ill_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(sampleType) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(sampleType, otProp, otReads, primerProp, primerReads, totalReads)
```

**performance by sampleType + species sets**

```{r, message=FALSE}
tamRun4_ill_gtscore_sampleSum %>%
  filter(lociSet != "fullSet") %>%
  #filter(GenotypeRate_0x >= 0.5) %>%
  group_by(sampleType, lociSet) %>%
  summarise(
    totalReads = sum(Total.Reads),
    primerReads = sum(Primer.Only.Reads) + sum(Primer.Probe.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2),
    primerProp = round(primerReads/totalReads, 2)
  ) %>%
  relocate(sampleType, lociSet, otProp, otReads, primerProp, primerReads, totalReads)
```

### byLocus

#### Overview

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(metricUse == "for_panelMetrics") %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  janitor::adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### dualLoci

lwed vs. simp vs. both performance

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP")) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(sampleSet, otProp, otReads, primerReads)
```

#### lwedLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(str_detect(Locus, "LWED")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun4_ill_gtscore_locusSum_simp_lwedSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("lwedSamples", "lwedSamples*", sampleSet)
  )
```

#### simpLoci

lwed vs. simp samples (* indicates target sampleSet)

```{r}
tamRun4_ill_gtscore_locusSum %>%
  filter(str_detect(Locus, "SIMP")) %>%
  select(Locus, sampleSet, Primer.Reads, Primer.Probe.Reads, Primer.Probe.Proportion) %>%
  rbind(., tamRun4_ill_gtscore_locusSum_lwed_simpSpec) %>%
  group_by(sampleSet) %>%
  summarise(
    primerReads = sum(Primer.Reads),
    otReads = sum(Primer.Probe.Reads)
  ) %>%
  as.data.frame() %>%
  adorn_totals() %>%
  mutate(
    otProp = round(otReads/primerReads, 2)
  ) %>%
  relocate(otProp, otReads, primerReads) %>%
  mutate(
    sampleSet = gsub("simpSamples", "simpSamples*", sampleSet)
  )
```

## 7.4 Geno success

### overall

**gtseq**

```{r}
mean(run4_gtseq_sampleSum$genoSuccess) # 0.6384836
```

**gtscore**

```{r}
run4_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  summarise(mean_genoSuccess_0x = mean(GenotypeRate_0x),
            mean_genoSuccess_10x = mean(GenotypeRate_10x))

# mean_genoSuccess_0x 0.8032102
# mean_genoSuccess_10x 0.703484
```

### by_sampleType

**gtseq**

before ditching failed samples:
  sampleType mean_genoSuccess min_genoSuccess max_genoSuccess med_genoSuccess
  <chr>                 <dbl>           <dbl>           <dbl>           <dbl>
1 blood                 0.816          0                0.977           0.893
2 fecal                 0.727          0.0694           0.983           0.815
3 hair                  0.488          0                0.909           0.601

after:
  sampleType mean_genoSuccess min_genoSuccess max_genoSuccess med_genoSuccess
  <chr>                 <dbl>           <dbl>           <dbl>           <dbl>
1 blood                 0.887           0.571           0.977           0.903
2 fecal                 0.804           0.526           0.983           0.828
3 hair                  0.736           0.583           0.909           0.723

```{r}
run4_gtseq_genoSuccess_bySample <- run4_gtseq_sampleSum %>%
  merge(., run4_md[, c("sampleID_ill", "animalID", "species")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  filter(!str_detect(animalID, "Neg")) %>%
  select(animalID, sampleType, genoSuccess) #%>%
  #pivot_wider(id_cols = animalID,
#              names_from = sampleType,
#              values_from = genoSuccess)

run4_gtseq_genoSuccess_bySample %>%
  group_by(sampleType) %>%
  summarise(
    mean_genoSuccess = mean(genoSuccess),
    min_genoSuccess = min(genoSuccess),
    max_genoSuccess = max(genoSuccess),
    med_genoSuccess = median(genoSuccess)
    )

run4_gtseq_genoSuccess_bySample_over50 <- run4_gtseq_genoSuccess_bySample %>%
  pivot_longer(!animalID, names_to = "sampleType", values_to = "genoSuccess") %>%
  filter(genoSuccess >= 0.5) %>%
  merge(., run4_md[, c("sampleID_ill", "sampleType", "animalID")], by = c("animalID", "sampleType")) %>%
  dplyr::rename("sampleID" = "sampleID_ill")
  
table(run4_gtseq_genoSuccess_bySample_over50$sampleType)

run4_gtseq_genoSuccess_bySample_over50 %>%
  group_by(sampleType) %>%
  summarise(
    mean_genoSuccess = mean(genoSuccess),
    min_genoSuccess = min(genoSuccess),
    max_genoSuccess = max(genoSuccess),
    med_genoSuccess = median(genoSuccess)
    )

run4_gtseq_genoSuccess_bySample_setsOver50 <- run4_gtseq_genoSuccess_bySample %>%
  filter(blood >= 0.5,
         fecal >= 0.5,
         hair >= 0.5)
```

**gtscore**

```{r}
run4_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  group_by(sampleType) %>%
  summarise(
    mean_genoSuccess_0x = round(mean(GenotypeRate_0x), 2),
    mean_genoSuccess_10x = round(mean(GenotypeRate_10x), 2)
  )
```

### figures

make a figure like Burgess et al., 2022 fig. 5a/b

```{r}
# gtseq
run4_gtseq_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  mutate(genoSuccess = genoSuccess*100) %>%
  ggplot(aes(x = sampleType, y = genoSuccess, shape = sampleType, color = sampleType)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(0.2)) +
  scale_shape_manual(values = c(1, 17, 19)) +
  geom_hline(yintercept = 50, linetype = 2) +
  scale_colour_Publication() +
  theme_Publication() +
  labs(title = "Genotyping success across sample types",
       x = "Sample type",
       y = "Percent loci genotyped") +
  theme(legend.position = "none")
```


```{r}
# gtscore0x
run4_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  ggplot(aes(x = sampleType, y = GenotypeRate_0x, shape = sampleType, color = sampleType)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(0.2)) +
  scale_shape_manual(values = c(1, 17, 19)) +
  scale_colour_Publication() +
  theme_Publication()
```

```{r}
# gtscore10x
run4_gtscore_sampleSum %>%
  filter(sampleType %in% c("blood", "fecal", "hair")) %>%
  ggplot(aes(x = sampleType, y = GenotypeRate_10x, shape = sampleType, color = sampleType)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(0.2)) +
  scale_shape_manual(values = c(1, 17, 19)) +
  scale_colour_Publication() +
  theme_Publication()
```

## 7.5 Species/sex assignments

### Species

```{r}
run4_spAssignments_gtseq <- assignSpecies(genoFile = run4_genos_gtseq,
                                         mdFile = run4_md,
                                         sampleID_colName = "sampleID_ill") %>%
  merge(., run4_md[, c("sampleID_ill", "sampleID_ill_md", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_ill_md, .after = sampleID) %>%
  filter(!str_detect(sampleID_ill_md, "Neg")) %>%
  filter(sampleID %in% run4_gtseq_genoSuccess_bySample_over50$sampleID)

mean(run4_spAssignments$totalGenos_sp)
median(run4_spAssignments$totalGenos_sp)
min(run4_spAssignments$totalGenos_sp)
max(run4_spAssignments$totalGenos_sp)

run4_spAssignments %>%
  group_by(sampleType) %>%
  summarise(
    missing = sum(is.na(mdMatch)),
    mdMismatch = sum(mdMatch == FALSE, na.rm = T),
    mix = sum(spAssigned == "MIX", na.rm = T),
    mdMatch = sum(mdMatch == TRUE, na.rm = T)
  ) %>%
  as.data.frame()
```

### Sex

```{r}
run4_sexAssignments_gtseq <- assignSex(genoFile = run4_genos_gtseq,
                                 mdFile = run4_md,
                                 sampleID_colName = "sampleID_ill",
                                 exclude_nonTargetSp = "yes") %>%
  merge(., run4_md[, c("sampleID_ill", "sampleID_ill_md", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_ill_md, .after = sampleID) %>%
  filter(!str_detect(sampleID_ill_md, "Neg")) %>%
  mutate(
    sexMajority = case_when(
      propF > propM ~ "F",
      propM > propF ~ "M",
      .default = NA
    ),
    mdMatch_sexMajority = mdSex == sexMajority
  ) %>%
  relocate(c(sexMajority, mdMatch_sexMajority), .after = sexAssigned) %>%
  # remove samples w/<=50% genoSuccess
  filter(sampleID %in% run4_gtseq_genoSuccess_bySample_over50$sampleID) %>%
  # remove aberrant fecal sample (animalID 210)
  filter(sampleID != "tamRun4.053")

mean(run4_sexAssignments$propGenos_sex)

# lwed avg 5.756757
run4_sexAssignments %>%
  filter(str_detect(sampleID_ill_md, "lwed")) %>%
  summarise(
    avgGenos = mean(totalGenos_sex),
    medGenos = median(totalGenos_sex),
    minGenos = min(totalGenos_sex),
    maxGenos = max(totalGenos_sex)
  )

# simp avg 7.580645
run4_sexAssignments %>%
  filter(str_detect(sampleID_ill_md, "simp")) %>%
  summarise(
    avgGenos = mean(totalGenos_sex),
    medGenos = median(totalGenos_sex),
    minGenos = min(totalGenos_sex),
    maxGenos = max(totalGenos_sex)
  )

run4_sexAssignments %>%
  group_by(sampleType) %>%
  summarise(
    missing = sum(is.na(mdMatch)),
    mdMatch_maj = sum(mdMatch_sexMajority == TRUE, na.rm = T),
    mdMismatch = sum(mdMatch == FALSE, na.rm = T),
    mix = sum(sexAssigned == "MIX", na.rm = T),
    mdMatch = sum(mdMatch == TRUE, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(
    totalGenos = mdMatch + mdMismatch,
    propMatch = mdMatch/totalGenos,
    propMatch_majority = mdMatch_maj/totalGenos
  )
```

## 7.6 Sample type discordance

### bySample

Examining geno concordance for GTseq genos 
1. Set up genoCompare df; set any genos for non-target species to NA
2. Get total/prop common/mismatch genos per animalID; remove indivs whose sample types don't have at least 50% common genos (relative to their species' loci set)

Avg common genos:
  -   blood_vs_fecal = 134.4167 (median = 139)
  -   blood_vs_hair = 119.25 (median = 119.5)
  -   fecal_vs_hair = 122.5833 (median = 123)

```{r}
run4_genoCompare_sampleTypes_sampleList <- run4_md %>%
  filter(animalID %in% run4_gtseq_genoSuccess_bySample_setsOver50$animalID) %>%
  filter(animalID != 210)

run4_genoCompare_sampleTypes_gtseq <- run4_genos_gtseq %>%
  select(run4_genoCompare_sampleTypes_sampleList$sampleID_ill) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID,
               names_to = "locus",
               values_to = "genos") %>%
  merge(., run4_md[, c("sampleID_ill", "animalID", "sampleType", "species")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  mutate(animalID = str_c(animalID, "_", species)) %>%
  pivot_wider(id_cols = c(animalID, locus),
              names_from = sampleType,
              values_from = genos) %>%
  mutate(
    bfMatch = blood == fecal,
    bhMatch = blood == hair,
    fhMatch = fecal == hair
  )

length(unique(run4_genoCompare_sampleTypes_gtseq$animalID))


run4_genoCompare_sampleTypes_gtseq_sampleSum <- run4_genoCompare_sampleTypes_gtseq %>%
#  na.omit() %>%
  group_by(animalID) %>%
  summarise(
#    genosCommon = sum(!is.na(bfMatch)),
    bfCommon = sum(!is.na(bfMatch)),
    bhCommon = sum(!is.na(bhMatch)),
    fhCommon = sum(!is.na(fhMatch)),
    
    bfMatch = sum(bfMatch == TRUE, na.rm = T),
    bhMatch = sum(bhMatch == TRUE, na.rm = T),
    fhMatch = sum(fhMatch == TRUE, na.rm = T)
  ) %>%
  as.data.frame() %>%
  mutate(
#    bf_propMatch = bfMatch/genosCommon,
#    bh_propMatch = bhMatch/genosCommon,
#    fh_propMatch = fhMatch/genosCommon
    
    bf_propMatch = bfMatch/bfCommon,
    bh_propMatch = bhMatch/bhCommon,
    fh_propMatch = fhMatch/fhCommon
  ) %>%
  mutate(
    bf_propCommon = case_when(
#      str_detect(animalID, "LWED") ~ genosCommon/173,
#      str_detect(animalID, "SIMP") ~ genosCommon/175
      
      str_detect(animalID, "LWED") ~ bfCommon/173,
      str_detect(animalID, "SIMP") ~ bfCommon/175
    ),
    bh_propCommon = case_when(
#      str_detect(animalID, "LWED") ~ genosCommon/173,
#      str_detect(animalID, "SIMP") ~ genosCommon/175
      
      str_detect(animalID, "LWED") ~ bhCommon/173,
      str_detect(animalID, "SIMP") ~ bhCommon/175
    ),
    fh_propCommon = case_when(
#      str_detect(animalID, "LWED") ~ genosCommon/173,
#      str_detect(animalID, "SIMP") ~ genosCommon/175
      
      str_detect(animalID, "LWED") ~ fhCommon/173,
      str_detect(animalID, "SIMP") ~ fhCommon/175
    )
  )

# common genos
mean(run4_genoCompare_sampleTypes_gtseq_sampleSum$bfCommon) # 134.4167
mean(run4_genoCompare_sampleTypes_gtseq_sampleSum$bhCommon) # 119.25
mean(run4_genoCompare_sampleTypes_gtseq_sampleSum$fhCommon) # 122.5833

median(run4_genoCompare_sampleTypes_gtseq_sampleSum$bfCommon) # 139
median(run4_genoCompare_sampleTypes_gtseq_sampleSum$bhCommon) # 119.5
median(run4_genoCompare_sampleTypes_gtseq_sampleSum$fhCommon) # 123


# concordance
mean(run4_genoCompare_sampleTypes_gtseq_sampleSum$bf_propMatch) # 0.8871016
mean(run4_genoCompare_sampleTypes_gtseq_sampleSum$bh_propMatch) # 0.860932
mean(run4_genoCompare_sampleTypes_gtseq_sampleSum$fh_propMatch) # 0.908682

median(run4_genoCompare_sampleTypes_gtseq_sampleSum$bf_propMatch) # 0.9104492
median(run4_genoCompare_sampleTypes_gtseq_sampleSum$bh_propMatch) # 0.8583387
median(run4_genoCompare_sampleTypes_gtseq_sampleSum$fh_propMatch) # 0.9388099
```

210_LWED and 74_SIMP fecal samples both seem incorrect
--s053_lwed_210_fecal; rnaL_1166; t8
--s040_simp_74_fecal; rnaL_241; t20

```{r}
View(run4_md %>% filter(sampleID_unique %in% c("210_fecal", "74_fecal")))

run4_genoCompare_sampleTypes_gtseq_sampleSum %>%
  select(animalID, contains("propMatch")) %>%
  dplyr::rename("blood_vs_feces" = "bf_propMatch",
                "blood_vs_hair" = "bh_propMatch",
                "feces_vs_hair" = "fh_propMatch") %>%
  pivot_longer(!animalID,
               names_to = "compType",
               values_to = "propMatch") %>%
  mutate(propMatch = propMatch*100) %>%
  ggplot(aes(x = compType, y = propMatch, color = compType)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position = position_jitterdodge()) +
  scale_colour_Publication() +
  theme_Publication() +
  labs(title = "Genotyping concordance between sample types",
       x = "",
       y = "% concordant genotypes") +
  theme(legend.position = "none") +
  ylim(0, 100)
```

## byLocus

```{r}

```


# 8 GTseq vs. GTscore

## 8.1 Overview

Comparing results from GTseq vs. GTscore pipelines, including:

  -   total reads
  -   on-target reads (dual-loci only)
  -   geno success (gtseq vs. gtscore 0x cutoff)
  -   geno concordance

## 8.1 Data

### Allele reads

```{r}
run4_readCounts_posDual_gtseq <- run4_readCounts_gtseq %>%
  select(run4_sampleList_pos$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_gt") %>%
  mutate(
    readSum_gt = as.numeric(get_readSum(readCounts_gt))
  )

run4_readCounts_posDual_gtscore <- run4_readCounts_gtscore %>%
  select(run4_sampleList_pos$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_gr") %>%
  mutate(
    readSum_gr = as.numeric(get_readSum(readCounts_gr))
  )
```

### Genos

```{r}
run4_genos_posDual_gtseq <- run4_genos_gtseq %>%
  select(run4_sampleList_pos$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_gt")

run4_genos_posDual_gtscore <- run4_genos_gtscore0x %>%
  select(run4_sampleList_pos$sampleID_ill) %>%
  rownames_to_column("locus") %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  #mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_gr")
```

### Summaries

```{r}
# gtseq
#run4_gtseq_sampleSum <- read.csv("./04_tamRun4/00_illumina/04_run4GTseq/summaryFiles/tamRun4_master_sampleSummary.csv") %>%
#  filter(!str_detect(sampleID_md, "Neg"))

## just use for raw reads; was run with genotyper v3 so other metrics are a little off
run4_gtseq_compiledGenos <- read.csv("./04_tamRun4/00_illumina/04_run4GTseq/gtseq_compiledGenos.csv")

run4_gtseq_sampleSum # loaded above


# gtscore
run4_gtscore_sampleSum # loaded in section 7
```

## 8.2 readCompare

```{r}
# total reads
sum(run4_gtseq_compiledGenos$Raw.Reads) # 31663938
sum(run4_gtscore_sampleSum$Total.Reads) # 31663938

# on-target reads
run4_sampleList_negs <- run4_md %>%
  filter(str_detect(sampleType, "Neg"))
run4_sampleList_pos <- run4_md %>%
  filter(!str_detect(sampleType, "Neg"))

## gtseq = 7133024
run4_readCounts_gtseq %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts") %>%
  filter(!sampleID %in% run4_sampleList_negs$sampleID_ill) %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  mutate(readSums = as.numeric(get_readSum(readCounts))) %>%
  summarise(
    otReads = sum(readSums)
  )

## gtscore = 7091829
run4_readCounts_gtscore %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts") %>%
  filter(!sampleID %in% run4_sampleList_negs$sampleID_ill) %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  mutate(readSums = as.numeric(get_readSum(readCounts))) %>%
  summarise(
    otReads = sum(readSums)
  )

# percent difference = |V1 - V2| / (V1 + V2)/2 * 100 = 0.5791975%
( abs(7133024 - 7091829) / ( (7133024 + 7091829)/2) ) * 100
```

## 8.3 genoCompare

```{r}
run4_genoCompare_seqScore <- run4_genos_posDual_gtseq %>%
  merge(., run4_genos_posDual_gtscore, by = c("sampleID", "locus")) %>%
  mutate(
    match = genos_gt == genos_gr
  ) %>%
  merge(., run4_readCounts_posDual_gtseq, by = c("sampleID", "locus")) %>%
  merge(., run4_readCounts_posDual_gtscore, by = c("sampleID", "locus")) %>%
  select(-contains("readCounts")) %>%
  mutate(
    missing = case_when(
      is.na(genos_gt) & is.na(genos_gr) ~ "both",
      is.na(genos_gt) & !is.na(genos_gr) ~ "gt",
      is.na(genos_gr) & !is.na(genos_gt) ~ "gr",
      .default = NA
    ),
    missingReason = case_when(
      missing == "gt" & readSum_gt < 10 ~ "tooLow",
      missing == "gt" & readSum_gt >= 10 ~ "badRatio",
      .default = NA
    )
  )

# both = 2551; gr = 10; gt = 1643
table(run4_genoCompare_seqScore$missing)

# badRatio = 361, tooLow = 1282 (for genos missing in gt, but present in gr)
table(run4_genoCompare_seqScore$missingReason)

# all 8396 common genos matched
table(run4_genoCompare_seqScore$match)

# gtseq gave 8406 genos
sum(!is.na(run4_genoCompare_seqScore$genos_gt))

# gtscore gave 10039 genos
sum(!is.na(run4_genoCompare_gtseqScore$genos_gr))

(abs(8406 - 10039) / ((8406 + 10039)/2))*100 # 17.7067% different

# genoSuccess
sum(!is.na(run4_genoCompare_gtseqScore$genos_gt))/nrow(run4_genoCompare_gtseqScore) # 0.6671429
sum(!is.na(run4_genoCompare_gtseqScore$genos_gr))/nrow(run4_genoCompare_gtseqScore) # 0.796746
```

## 8.4 GTscore fullPrimers

### GTscore quick-pipeline

```{r}
source("./project_scripts/GTscore/GTscore_modified.R")

# count reads for amplicons
system2("perl",
        args="./project_scripts/GTscore/AmpliconReadCounter_modified.pl -p ./05_tamRun5/03_run5GTscore/primerProbeFileV3_dual.txt --files ./04_tamRun4/00_illumina/03_run4GTscore/ill_pos_sampleFiles.txt --useFullPrimer --inDir ./04_tamRun4/00_illumina/02_run4Interleaved/ --outDir ./04_tamRun4/00_illumina/03_run4GTscore/ --prefix ill_posDual_fullPrimer_")

# genotype
## load locus table and 10x allele reads file
posDual_fullPrimer_locusTable <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_posDual_fullPrimer_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

posDual_fullPrimer_alleleReads <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_posDual_fullPrimer_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

## generate singleSNP genotypes using the polyGen algorithm
posDual_fullPrimer_polyGenResults <- polyGen(posDual_fullPrimer_locusTable, posDual_fullPrimer_alleleReads, epsilon = 0.01)

## write results
write.table(posDual_fullPrimer_polyGenResults, "./04_tamRun4/00_illumina/03_run4GTscore/ill_posDual_fullPrimer_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")

# geno success by Sample
posDual_fullPrimer_sampleSum <- summarizeSamples(posDual_fullPrimer_polyGenResults, posDual_fullPrimer_alleleReads)
```

### Data

```{r}
# alleleReads
run4_readCounts_posDual_gtscore_fullPrimer <- read.delim("./04_tamRun4/00_illumina/03_run4GTscore/ill_posDual_fullPrimer_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "readCounts_gr") %>%
  mutate(
    readSum_gr = as.numeric(get_readSum(readCounts_gr))
  )

# genos
run4_genos_posDual_gtscore_fullPrimer <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_posDual_fullPrimer_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_gr")

# sample summary
run4_sampleSum_posDual_gtscore_fullPrimer <- read.table("./04_tamRun4/00_illumina/03_run4GTscore/ill_posDual_fullPrimer_GTscore_individualSummary.txt", header = T, sep = "\t") %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  merge(., posDual_fullPrimer_sampleSum, by.x = "Sample", by.y = "sample")

# gtscore - trimmed primers
temp <- run4_gtscore_sampleSum %>%
  filter(!str_detect(sampleID_unique, "Neg"))

sum(temp$Total.Reads) # 30883594
sum(temp$Primer.Only.Reads) # 2834597
sum(temp$Primer.Probe.Reads) # 9168973
mean(temp$Primer.Probe.Proportion) # 0.2222222

# gtscore - full primers
sum(run4_sampleSum_posDual_gtscore_fullPrimer$Total.Reads) # 30883594
sum(run4_sampleSum_posDual_gtscore_fullPrimer$Primer.Only.Reads) # 2361258
sum(run4_sampleSum_posDual_gtscore_fullPrimer$Primer.Probe.Reads) # 7077368
mean(run4_sampleSum_posDual_gtscore_fullPrimer$Primer.Probe.Proportion) # 0.1665556
```

### genoCompare

```{r}
run4_genoCompare_seqScore_fullPrimer <- run4_genos_posDual_gtseq %>%
  merge(., run4_genos_posDual_gtscore_fullPrimer, by = c("sampleID", "locus")) %>%
  mutate(
    match = genos_gt == genos_gr
  ) %>%
  merge(., run4_readCounts_posDual_gtseq, by = c("sampleID", "locus")) %>%
  merge(., run4_readCounts_posDual_gtscore_fullPrimer, by = c("sampleID", "locus")) %>%
  select(-contains("readCounts")) %>%
  mutate(
    missing = case_when(
      is.na(genos_gt) & is.na(genos_gr) ~ "both",
      is.na(genos_gt) & !is.na(genos_gr) ~ "gt",
      is.na(genos_gr) & !is.na(genos_gt) ~ "gr",
      .default = NA
    ),
    missingReason = case_when(
      missing == "gt" & readSum_gt < 10 ~ "tooLow",
      missing == "gt" & readSum_gt >= 10 ~ "badRatio",
      .default = NA
    )
  )

# both = 2551 > now 2556; gr = 10 > same; gt = 1643 > now 1638
table(run4_genoCompare_seqScore_fullPrimer$missing)

# badRatio = 361 > now 358, tooLow = 1282 > now 1280 (for genos missing in gt, but present in gr)
table(run4_genoCompare_seqScore_fullPrimer$missingReason)

# all 8396 common genos matched >> same
table(run4_genoCompare_seqScore_fullPrimer$match)

# gtseq gave 8406 genos >> same
sum(!is.na(run4_genoCompare_seqScore_fullPrimer$genos_gt))

# gtscore gave 10039 genos >> now 10034
sum(!is.na(run4_genoCompare_seqScore_fullPrimer$genos_gr))

(abs(8406 - 10034) / ((8406 + 10034)/2))*100 # 17.7067% >> now 17.65727% different

# genoSuccess
sum(!is.na(run4_genoCompare_seqScore_fullPrimer$genos_gt))/nrow(run4_genoCompare_seqScore_fullPrimer) # 0.6671429 >> now same
sum(!is.na(run4_genoCompare_seqScore_fullPrimer$genos_gr))/nrow(run4_genoCompare_seqScore_fullPrimer) # 0.796746 >> now 0.7963492
```




```{r}
temp_genos <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_gr") %>%
  filter(!str_detect(locus, "LWED|SIMP"))

ngmerge_genos <- read.table("./sandbox/NGmerge/02_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T, sep = "\t") %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), \(x) na_if(x, "0"))) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genos_ng") %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  mutate(
    sampleID = gsub("_stitched", "", sampleID)
  ) %>%
  merge(., temp_genos, by = c("sampleID", "locus")) %>%
  mutate(
    match = genos_gr == genos_ng
  )

table(ngmerge_genos$match) # FALSE 71 vs. TRUE 34597

(71/(71+34597))*100 # 0.2047998% = mismatch

sum(!is.na(ngmerge_genos$genos_ng)) # 34867 (258 less genos)
sum(!is.na(ngmerge_genos$genos_gr)) # 35125 (258 more genos)
```

# 9 NGmerge

## 9.1 Run NGmerge

```{bash}
# a little name-prep first
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/04_tamRun4/00_illumina/06_NGmerge/00_run4Seqs

# Trim filenames to make them more manageable
for f in *R1_001.fastq.gz; do g="${f%%_S*}_R1.fastq.gz"; mv "${f}" "${g}"; done
for f in *R2_001.fastq.gz; do g="${f%%_S*}_R2.fastq.gz"; mv "${f}" "${g}"; done
```

```{bash}
# run NGmerge
cd ..

for i in ./00_run4Seqs/*_R1.fastq.gz; do name=$(basename $i _R1.fastq.gz); ./NGmerge -1 00_run4Seqs/${name}_R1.fastq.gz -2 00_run4Seqs/${name}_R2.fastq.gz -o 01_NGmergeOutput/00_seqs_stitched/${name}_stitched.fastq.gz -d -f 01_NGmergeOutput/01_seqs_failed/${name}_fail -l 01_NGmergeOutput/02_logs_stitchResults/${name}_log_stitchResults.txt -c 01_NGmergeOutput/03_logs_dovetailedReads/${name}_log_dovetailedReads.txt -j 01_NGmergeOutput/04_logs_stitchAlignments/${name}_log_stitchAlignments.txt; done

# Extract all files
cd 01_NGmergeOutput/00_seqs_stitched
gunzip -r .
```

## 9.2 GTscore quick-pipeline

```{bash}
# create sampleFile
for i in *fastq; do echo $i; done > ./../../02_run4GTscore/sampleFiles_fullSet.txt
```

```{r}
source("./project_scripts/GTscore/GTscore_modified.R")

# count reads for amplicons
system2("perl",
        args="./project_scripts/GTscore/AmpliconReadCounter_modified.pl -p ./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt --files ./04_tamRun4/00_illumina/06_NGmerge/02_run4GTscore/sampleFiles_fullSet.txt --useFullPrimer --inDir ./04_tamRun4/00_illumina/06_NGmerge/01_NGmergeOutput/00_seqs_stitched/ --outDir ./04_tamRun4/00_illumina/06_NGmerge/02_run4GTscore/ --prefix fullSet_")

# genotype
## load locus table and 10x allele reads file
ngmerge_locusTable <- read.delim("./04_tamRun4/00_illumina/06_NGmerge/02_run4GTscore/fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

ngmerge_alleleReads <- read.delim("./04_tamRun4/00_illumina/06_NGmerge/02_run4GTscore/fullSet_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

## generate singleSNP genotypes using the polyGen algorithm
ngmerge_polyGenResults <- polyGen(ngmerge_locusTable, ngmerge_alleleReads, epsilon = 0.01)

## write results
write.table(ngmerge_polyGenResults, "./04_tamRun4/00_illumina/06_NGmerge/02_run4GTscore/fullSet_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")

## then reload
run4_genos_ngmerge_gtscore <- read.table("./04_tamRun4/00_illumina/06_NGmerge/02_run4GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T, sep = "\t") %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = gsub("_stitched", "", sampleID)) %>%
  column_to_rownames("sampleID") %>%
  t() %>%
  as.data.frame() %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), \(x) na_if(x, "0")))

# geno success by Sample
ngmerge_sampleSum <- summarizeSamples(ngmerge_polyGenResults, ngmerge_alleleReads)

NEED TO REDO GENO SUCCESS W/SP-SPECIFIC LOCI SETS

run4_ngmerge_genoSuccess_bySample_over50 <- ngmerge_sampleSum %>%
  filter(Geno)
```

## 9.3 Species/sex assignments

### Species

```{r}
run4_spAssignments_ngmerge_gtscore <- assignSpecies(genoFile = run4_genos_ngmerge_gtscore,
                                         mdFile = run4_md,
                                         sampleID_colName = "sampleID_ill") %>%
  merge(., run4_md[, c("sampleID_ill", "sampleID_ill_md", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_ill_md, .after = sampleID) %>%
  filter(!str_detect(sampleID_ill_md, "Neg")) %>%
  filter(sampleID %in% run4_gtseq_genoSuccess_bySample_over50$sampleID)
```

### Sex

```{r}
run4_sexAssignments_ngmerge_gtscore <- assignSex(genoFile = run4_genos_ngmerge_gtscore,
                                 mdFile = run4_md,
                                 sampleID_colName = "sampleID_ill",
                                 exclude_nonTargetSp = "yes") %>%
  merge(., run4_md[, c("sampleID_ill", "sampleID_ill_md", "sampleType")], by.x = "sampleID", by.y = "sampleID_ill") %>%
  relocate(sampleID_ill_md, .after = sampleID) %>%
  relocate(sampleID_ill_md, .after = sampleID) %>%
  filter(!str_detect(sampleID_ill_md, "Neg")) %>%
  mutate(
    sexMajority = case_when(
      propF > propM ~ "F",
      propM > propF ~ "M",
      .default = NA
    ),
    mdMatch_sexMajority = mdSex == sexMajority
  ) %>%
  relocate(c(sexMajority, mdMatch_sexMajority), .after = sexAssigned)
```