---
title: "tamGenetics_wgs"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

Note that VCF files are 1-based

## RESOURCES:

-   tabix <https://www.htslib.org/doc/tabix.html>

-   vcftools <https://vcftools.github.io/links.html>

-   GATK tool index <https://gatk.broadinstitute.org/hc/en-us/sections/27007892463131-4-6-0-0-Current?page=2#articles>

-   samtools how to extract stuff from VCFs <https://samtools.github.io/bcftools/howtos/query.html>

-   bcftools documentation <https://samtools.github.io/bcftools/bcftools.html#query>

-   eriqande book <https://eriqande.github.io/eca-bioinf-handbook/handle-vcf.html#bcftools>

-   reddit post <https://www.reddit.com/r/bioinformatics/comments/jbw72n/comment/g94sge1/> - youtube vid <https://www.youtube.com/watch?v=7W7hrWNyCaM>

-   biostars <https://www.biostars.org/p/298361/>

-   biostars (less helpful) <https://www.biostars.org/p/9576200/>

-   pysam maybe helpful code bit? <https://github.com/pysam-developers/pysam/issues/469>

# 2 Packages

## command line packages

Install Miniconda if not already installed (install in home directory! makes things easier)

```{bash}
# start in your home directory and do the following:
mkdir conda_install
cd conda_install/
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
chmod u+x Miniconda3-latest-Linux-x86_64.sh 
./Miniconda3-latest-Linux-x86_64.sh 
```

Install mamba

```{bash}
conda install mamba -n base -c conda-forge
```

Install software into bioinformatics environment

```{bash}
mamba create -n bioinf -c bioconda bcftools bwa samtools
```

## R packages

```{r}
library(readxl)
library(tidyverse)
```

# 3 Data

There's some kind of issue with locus "SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_332662-333103_chosen_consensus_modified" -- based on some digging around, I'm *fairly* certain that 333103 is the pre-bp3error stopPos while 333110 is the updated stopPos.

Original colnames for samMasterSNP file: [1] "SET"\
[2] "species"\
[3] "chrom_num"\
[4] "chrom_name"\
[5] "SNP"\
[6] "start (might not be accurate for replaced snps)" [7] "stop (might not be accurate for replaced snps)"\
[8] "bp3_index (UNIQUE)"\
[9] "Old position - for those that will be replaced"\
[10] "Note"\
[11] "BP3_error (no error unless otherwise stated)"\
[12] "GT_seq (\> 280 unless otherwise stated)"\
[13] "To be removed before second amplification round?" [14] "Second revision non informativeness"\
[15] "...15"

```{r}
panel_v3 <- read.table("./project_data/primerProbeFile_panelv3_fullSet.txt", header = T)

# pretty sure this is Sam's file w/the loci info after bp3error interval updates
samMasterSNP <- read_excel("./primers/26Sep2022_masterSNPFile.xlsx", sheet = 4) %>%
  # clean messy colnames
  `colnames<-` (c("set", "species", "chr_num", "chr_name", "snpPos", "beginPos", "endPos", "bp3index_unique", "snpPos_old", "notes1", "bp3_error", "gtseqNotes", "remove_yesNo", "nonInf_yesNo", "notes2")) %>%
  mutate(
    endPos = as.numeric(endPos) # ditch E notation
  ) %>%
  # create snpID for merging later
  mutate(
    snpID = str_c("chr", chr_num, ":", beginPos, "-", endPos)
  ) %>%
  relocate(snpID)

lociInfo <- read.csv("./project_data/seqNames_to_shortNames.csv") %>%
  select(primerName3, seqID3) %>%
  dplyr::rename("locus" = "primerName3",
                "seqID" = "seqID3") %>%
  filter(locus %in% panel_v3$Locus) %>%
  # extract chr number
  mutate(
    temp1 = sub(".*_(CM[0-9]+\\.[0-9]+).*", "\\1", seqID),
    temp2 = str_sub(temp1, 7, 8),
    chr_num = as.numeric(temp2) - 16
  ) %>%
  select(-contains("temp")) %>%
  # extract start/end numbers
  mutate(
    beginPos = sub(".*CM[0-9]+\\.[0-9]+[-:.\\_]([0-9]+)-([0-9]+).*", "\\1", seqID),
    endPos = sub(".*CM[0-9]+\\.[0-9]+[-:.\\_]([0-9]+)-([0-9]+).*", "\\2", seqID)
  ) %>%
  mutate(
    snpID = str_c("chr", chr_num, ":", beginPos, "-", endPos)
  ) %>%
  # update chr23:332662-333103 to 333110
  mutate(
    snpID = case_when(
      snpID == "chr23:332662-333103" ~ "chr23:332662-333110",
      .default = snpID
    ),
    endPos = case_when(
      snpID == "chr23:332662-333110" ~ "333110",
      .default = endPos
    )
  ) %>%
  merge(., samMasterSNP[, c("snpID", "snpPos", "chr_name")], by = "snpID", all.x = T) %>%
  arrange(locus) %>%
  relocate(c(locus, snpID, chr_name, chr_num, snpPos, beginPos, endPos))
```

# 4 Create loci reference files

## 4.1 .txt file for bcftools

```{r}
lociList_bcftools <- lociInfo %>%
  select(chr_name, snpPos) %>%
  arrange(chr_name, snpPos)

write.table(lociList_bcftools, "./DISSERTATION/ch1_panelDev/lociList_bcftools.txt", quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
```

## 4.2 .list file for GATK SelectVariants

Use lociInfo to create .list file for GATK SelectVariants

```{r}
lociList_gatk <- lociInfo %>%
  select(chr, beginPos, endPos) %>%
  mutate(
    temp = str_c(chr, ":", beginPos, "-", endPos)
  )

write.table(lociList_gatk, "./DISSERTATION/ch1_panelDev/lociList_gatk.list", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

# 5 Extract genos from VCF

They literally have an example of what I want on the bcftools documentation page. wowww. https://samtools.github.io/bcftools/bcftools.html#query

-H, --print-header
-f, --format
-R, --regions-file

Notes on output:
/: genotype unphased
|: genotype phased

GQ (Integer):  Conditional genotype quality,  encoded as a phred quality − 10log_10 p(genotype call is wrong, conditioned on the site’s being variant).

DP (Integer):  Read depth at this position for this sample

```{bash}
# 1) activate bioinf environment
conda activate bioinf

# 2) get in correct directory
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/DISSERTATION/ch1_panelDev/

# 3) pring CHROM and POS for each row of VCF file to check format
bcftools query -f '%CHROM\t%POS\n' interval_1.vcf.gz

# 4) extract genos
bcftools query -H -f '%CHROM\t %POS\t %REF\t %ALT [\t%GT\t%PL]\n' -R lociList_bcftools.txt interval_1.vcf.gz > interval_1_genos.txt

# geno + readDepth + genoQuality?
bcftools query -H -f '%CHROM\t %POS\t %REF\t %ALT [\t%GT\t%DP\t%GQ]\n' -R lociList_bcftools.txt interval_1.vcf.gz > interval_1_genos.txt
```

.vcf.gz = vcf files .tbi = indexed file

First need to install miniconda/mamba -- follow <https://eriqande.github.io/eca-bioinf-handbook/working-on-remote-servers.html#miniconda>

How to access VCF files on remote server (from youtube video <https://www.youtube.com/watch?v=7W7hrWNyCaM>):

1.  Copy VCF file link

2.  Assign the link a variable name for easier manipulation:

```{bash}
# assign variable name
link=https://export.uppmax.uu.se/snic2022-6-144/share/tmp_dir_vcfs/interval_1.vcf.gz

# $ indicates that we want the actual value
bcftools query -l $link
```

RESOURCES

reddit thread talking about doing what I think I want to do? <https://www.reddit.com/r/bioinformatics/comments/jbw72n/retrieve_samples_with_specific_snps_from_a_vcf/>
