---
title: "tamAnalyses_generalFiles"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

Meant to create a central holding place for files used across tamGenetics sequencing runs

# 2 Packages

```{r}
library(tidyverse)
```

# 3 Primer-probe file

## 3.1 SNP panel v3

### gtscore

```{r}
pp_fullSet_v3 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)

pp_lwed_v3 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_LWED.txt", header = T)

pp_simp_v3 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_SIMP.txt", header = T)

# export to tamAnalyses_generalFiles
write.table(pp_fullSet_v3, "./tamAnalyses_generalFiles/gtscore/primerProbeFile_panelv3_fullSet.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_lwed_v3, "./tamAnalyses_generalFiles/gtscore/primerProbeFile_panelv3_lwed.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_simp_v3, "./tamAnalyses_generalFiles/gtscore/primerProbeFile_panelv3_simp.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```

### gtseq

```{r, eval = F}
pp_fullSet_v3_gtseq <- pp_fullSet_v3 %>%
  select(c(Locus, Allele1, Allele2, Probe1, Probe2, Primer))

write.table(pp_fullSet_v3_gtseq, "./tamAnalyses_generalFiles/gtseq/gtseq_locusInfo_panelv3.csv", col.names = F, row.names = F, sep = ",", quote = F)
```

# 4 Species & sex assignments

## 4.1 Assign species

### speciesKey

```{r}
speciesKey <- read.csv("./03_tamRun3/04_genoAnalyses/speciesSNP_key_14Dec2022.csv") %>%
  arrange(Locus) %>%
  dplyr::rename("locus" = "Locus")

write.csv(speciesKey, "./tamAnalyses_generalFiles/speciesKey.csv", row.names = F)
```

### Code

```{r}
source("./my_sourceScripts/assignSpecies.R")
```

### Test

```{r}
md_tamRun5_v5 <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv") %>%
  mutate(
    sampleID_unique = str_c(animalID, "_", sampleType)
  ) %>%
  dplyr::rename("foo" = "sampleID")

genos_tamRun5_0x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") 

test_spAssign <- assignSpecies(genos_tamRun5_0x, md_tamRun5_v5, "foo")
```

## 4.2 Assign sex

### sexKey

```{r}
sexKey <- read.csv("./03_tamRun3/04_genoAnalyses/sexSNP_key_8Dec2022.csv") %>%
  na.omit() %>%
  unique() %>%
  filter(Locus %in% primerList_v3$locus) %>%
  dplyr::rename("locus" = "Locus") %>%
  arrange(locus)

write.csv(sexKey, "./tamAnalyses_generalFiles/sexKey.csv", row.names = F)
```

### Code

```{r}
source("./my_sourceScripts/assignSex.R")
```

### Test

```{r}
md_tamRun5_v5 <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv") %>%
  mutate(
    sampleID_unique = str_c(animalID, "_", sampleType)
  )

genos_tamRun5_0x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

test_sexAssign <- assignSex(genos_tamRun5_0x, md_tamRun5_v5, exclude_nonTargetSex = "yes")

test_sexAssign2 <- assignSex(genos_tamRun5_0x, md_tamRun5_v5, exclude_nonTargetSex = "no")
```