---
title: "tamAnalyses_generalFiles"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

Meant to create a central holding place for files used across tamGenetics sequencing runs

# 2 Packages

```{r}
library(tidyverse)
```

# 3 Primers/probes

## 3.1 seqNames_to_shortNames

typo in some seqID2 names? where seqID1 has "*final_rm400s*", seqID2 has "*finalm400s*" and I'm not sure this is correct? dunno where I originally got seqID2 names

species-specific sexid names based on Sam's thesis p. 20

```{r}
seqNames_to_shortNames <- read.csv("./primers/primers_seqName_to_shortName.csv") %>%
  mutate(
    primerName2 = case_when(
      primerName2 == "SEXID_195.3" ~ "SEXID_LWED_195.3",
      primerName2 == "SEXID_208.3" ~ "SEXID_LWED_208.3",
      primerName2 == "SEXID_211.3" ~ "SEXID_LWED_211.3",
      
      primerName2 == "SEXID_197.1" ~ "SEXID_SIMP_197.1",
      primerName2 == "SEXID_198.3" ~ "SEXID_SIMP_198.3",
      primerName2 == "SEXID_203.2" ~ "SEXID_SIMP_203.2",
      primerName2 == "SEXID_218.1" ~ "SEXID_SIMP_218.1",
      
      .default = primerName2
    )
  ) %>%
  mutate(primerName3 = sub('\\..*', '', primerName2)) %>%
  relocate(primerName3, .after = primerName2) %>%
  mutate(
    seqID3 = sub('[_][^_]+$', '', seqID1)
  ) %>%
  dplyr::rename("lociSet" = "type") %>%
  mutate(
    lociSet = case_when(
      str_detect(primerName2, "SEXID_LWED") ~ "SEXID_LWED",
      str_detect(primerName2, "SEXID_SIMP") ~ "SEXID_SIMP",
      .default = lociSet
    )
  )

write.csv(seqNames_to_shortNames, "./project_data/seqNames_to_shortNames.csv", row.names = F)
```

## 3.2 Panel v3

### gtscore

```{r}
pp_fullSet_v3 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_fullSet.txt", header = T)

pp_lwed_v3 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_LWED.txt", header = T)

pp_simp_v3 <- read.table("./05_tamRun5/03_run5GTscore/primerProbeFileV3_SIMP.txt", header = T)

pp_dual_v3 <- pp_fullSet_v3 %>%
  filter(!str_detect(Locus, "LWED|SIMP"))

# export to tamAnalyses_generalFiles
write.table(pp_fullSet_v3, "./tamAnalyses_generalFiles/gtscore/primerProbeFile_panelv3_fullSet.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_lwed_v3, "./project_data/primerProbeFile_panelv3_lwed.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_simp_v3, "./project_data/primerProbeFile_panelv3_simp.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_dual_v3, "./project_data/primerProbeFile_panelv3_dual.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```

### gtseq

```{r, eval = F}
pp_fullSet_v3_gtseq <- pp_fullSet_v3 %>%
  select(c(Locus, Allele1, Allele2, Probe1, Probe2, Primer))

write.table(pp_fullSet_v3_gtseq, "./project_data/gtseq_locusInfo_panelv3.csv", col.names = F, row.names = F, sep = ",", quote = F)
```

## 3.3 Complete lociInfo file

Including locus name, chromosome/SNP position info, primers, probes

### Data

**NOTE** - Sam's Master SNP google sheet file has the chromosome and SNP position info, but I'm not sure yet which tab has the correct information after she made post-bp3error adjustments to these numbers. The "Master_SNP_with_replacement_SNPs" tab *should* have the correct values, but some of the snpPos seem outright wrong (e.g., snpPos = 34 or some other tiny number). I tried looking at the "GTseq" tab to see if perhaps those were the correct/updated values? But still not sure.

I've asked Sam to clarify; waiting for response

```{r}
# final snp panel
panel_v3 <- read.table("./project_data/primerProbeFile_panelv3_fullSet.txt", header = T)
colnames(panel_v3) <- tolower(colnames(panel_v3))

# primers for final snp panel -- just need reverse primers here
primerSet_v3 <- read.csv("./primers/03_lociChoices/primerSet_v3.csv") %>%
  filter(fwdRev == "rev")

# seqNames_to_shortNames
seqNames_to_shortNames <- read.csv("./project_data/seqNames_to_shortNames.csv")

#################
### SAM FILES ###
#################

# "Master_SNP_with_replacement_SNPs" tab
samMasterSNP_masterTab <- read_excel("./primers/26Sep2022_masterSNPFile.xlsx", sheet = 2, skip = 1) %>%
  # clean messy colnames
  `colnames<-` (c("null", "set", "species", "chr_num", "chr_name", "snpPos", "beginPos", "endPos", "snp_dist_to_start", "length", "snp_dist_to_end", "mfe_p1", "mfe_p2", "bp3index_unique", "passed_inSilico", "notes_inSilico", "snp_not_in_amplicon", "amplicon_contains_other_vars", "other_vars_equallyInformative", "other_vars_lessInformative", "snpPos_old", "notes1", "beginPos_old", "endPos_old")) %>%
  # remove second row
  filter(set != "sum") %>%
  # ditch first col
  select(-null) %>%
  # fix E notation
  mutate(
    endPos = as.numeric(endPos)
  ) %>%
  # create chrID for merging later
  mutate(
    chrID = str_c("chr", chr_num, ":", beginPos, "-", endPos)
  ) %>%
  relocate(chrID)

# "GTseq" tab
samMasterSNP_gtseqTab <- read_excel("./primers/26Sep2022_masterSNPFile.xlsx", sheet = 4) %>%
  # clean messy colnames
  `colnames<-` (c("set", "species", "chr_num", "chr_name", "snpPos", "beginPos", "endPos", "bp3index_unique", "snpPos_old", "notes1", "bp3_error", "gtseqNotes", "remove_yesNo", "nonInf_yesNo", "notes2")) %>%
  mutate(
    endPos = as.numeric(endPos) # ditch E notation
  ) %>%
  # create chrID for merging later
  mutate(
    chrID = str_c("chr", chr_num, ":", beginPos, "-", endPos)
  ) %>%
  relocate(chrID)
```

### Create lociInfo file

Creating master file with locus info that includes all chromosome/SNP position stuff + the locus short-names, including:

-   seqID
-   chr_name
-   chr_num
-   snpPos_inChr
-   startPos
-   endPos
-   locus
-   snpPos_inAmp
-   allele1
-   allele2
-   probe1
-   probe2
-   primerF
-   primerR

------------------------------------------------------------------------

There's some kind of issue with locus "SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_332662-333103_chosen_consensus_modified" -- based on some digging around, I'm *fairly* certain that 333103 is the pre-bp3error stopPos while 333110 is the updated stopPos.

```{r}
lociInfo <- seqNames_to_shortNames %>%
  select(primerName3, seqID3) %>%
  filter(primerName3 %in% panel_v3$locus) %>%
  dplyr::rename("locus" = "primerName3",
                "seqID" = "seqID3") %>%
  
  # update seqID
  mutate(
    seqID = case_when(
      seqID == "SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_332662-333103_chosen_consensus_modified" ~ "SEXID_gatk_filtered_new_non_hom_intervals_CM018939.1_332662-333110_chosen_consensus_modified",
      .default = seqID
    )
  ) %>%
 
  # extract chr number
  mutate(
    chr_name = sub(".*_(CM[0-9]+\\.[0-9]+).*", "\\1", seqID),
    chr_num = str_sub(chr_name, 7, 8),
    chr_num = as.numeric(chr_num) - 16
  ) %>%
  
  # extract start/end numbers
  mutate(
    startPos = sub(".*CM[0-9]+\\.[0-9]+[-:.\\_]([0-9]+)-([0-9]+).*", "\\1", seqID),
    endPos = sub(".*CM[0-9]+\\.[0-9]+[-:.\\_]([0-9]+)-([0-9]+).*", "\\2", seqID)
  ) %>%
  mutate(
    chrID = str_c("chr", chr_num, ":", startPos, "-", endPos)
  ) %>%
  
  # add snpID based on chr_name + snpPos (use gtseq tab for now)
  merge(., samMasterSNP_gtseqTab[, c("chrID", "snpPos")], by = "chrID", all.x = T, suffixes = c("_masterTab", "_gtseqTab")) %>%
#  mutate(
#    snpID = str_c(chr_name, "-", snpPos)
#  ) %>%
 
  # add ploidy, alleles, probes, and forward primer
  merge(., panel_v3, by = "locus") %>%
  
  # add reverse primer
  merge(., primerSet_v3[, c("primerName3", "seqPrimer")], by.x = "locus", by.y = "primerName3") %>%
  
  dplyr::rename("snpPos_inChr" = "snpPos",
                "snpPos_inAmp" = "snppos",
                "primerF" = "primer",
                "primerR" = "seqPrimer") %>%
  select(locus, ploidy, snpPos_inAmp, allele1, allele2, probe1, probe2, primerF, primerR, seqID, chr_name, chr_num, snpPos_inChr, startPos, endPos) %>%
  arrange(locus)
```

### Export

```{r}
write.csv(lociInfo, "./project_data/master_lociInfo.csv", row.names = F)
```

# 4 Species & sex assignments

## 4.1 Assign species

### speciesKey

```{r}
speciesKey <- read.csv("./03_tamRun3/04_genoAnalyses/speciesSNP_key_14Dec2022.csv") %>%
  arrange(Locus) %>%
  dplyr::rename("locus" = "Locus")

write.csv(speciesKey, "./project_data/speciesKey.csv", row.names = F)
```

### Code

```{r}
source("./project_scripts/ravo_assignSpecies.R")
```

### Test

```{r}
md_tamRun5_v5 <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv") %>%
  mutate(
    sampleID_unique = str_c(animalID, "_", sampleType)
  ) %>%
  dplyr::rename("foo" = "sampleID")

genos_tamRun5_0x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus") 

test_spAssign <- assignSpecies(genos_tamRun5_0x, md_tamRun5_v5, "foo")
```

## 4.2 Assign sex

### sexKey

```{r}
sexKey <- read.csv("./03_tamRun3/04_genoAnalyses/sexSNP_key_8Dec2022.csv") %>%
  na.omit() %>%
  unique() %>%
  filter(Locus %in% primerList_v3$locus) %>%
  dplyr::rename("locus" = "Locus") %>%
  arrange(locus)

write.csv(sexKey, "./project_data/sexKey.csv", row.names = F)
```

### Code

```{r}
source("./project_scripts/ravo_assignSex.R")
```

### Test

```{r}
md_tamRun5_v5 <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv") %>%
  mutate(
    sampleID_unique = str_c(animalID, "_", sampleType)
  )

genos_tamRun5_0x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_0x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

test_sexAssign <- assignSex(genos_tamRun5_0x, md_tamRun5_v5, exclude_nonTargetSex = "yes")

test_sexAssign2 <- assignSex(genos_tamRun5_0x, md_tamRun5_v5, exclude_nonTargetSex = "no")
```

# 5 Samples

Getting tired of continuously adjusting sample names so I can compare between runs, so going to create a file that has all that info in one spot

## Data

```{r}
run1_md <- read.csv("./00_tamRun1/03_run1GTscore/tamRun1_metadata_v2.csv") %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  mutate(
    sampleRef = tolower(str_c(species, sex, animalID, sep = "_")),
    seqRun = "run1"
    )

run3_md <- read.csv("./03_tamRun3/03_run3GTscore/tamRun3_metadata_abcat_5June2023.csv") %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  mutate(
    sampleRef = tolower(str_c(species, sex, animalID, sep = "_")),
    seqRun = case_when(
      str_detect(sampleID, "cat") ~ "run3_cat",
      str_detect(sampleID, "3b") ~ "run3_b",
      .default = "run3_a"
      )
    )

run4_md_ill <- read.csv("./04_tamRun4/tamRun4_metadata_illONT.csv") %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  mutate(
    sampleID = sampleID_ill,
    sampleRef = tolower(str_c(species, sex, animalID, sep = "_")),
    seqRun = "run4_ill"
    )

run4_md_ont <- read.csv("./04_tamRun4/tamRun4_metadata_illONT.csv") %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  mutate(
    sampleID = sampleID_ont,
    sampleRef = tolower(str_c(species, sex, animalID, sep = "_")),
    seqRun = "run4_ont"
    )

run5_md <- read.csv("./metadataReconciliation/tamRun5_metadata_v5.csv") %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  mutate(
    sampleRef = tolower(str_c(species, sex, animalID, sep = "_")),
    seqRun = "run5"
    )

run_wgs <- read.csv("./06_tamGenetics_wgs/00_data/tamGenetics_wgs_metadata.csv") %>%
  select(-sampleID) %>%
  mutate(
    sampleID = sampleID2,
    sampleRef = tolower(str_c(species, sex, animalID, sep = "_")),
    seqRun = "run_wgs",
    sampleType = "tissue"
  )
```

## sampleRef

```{r}
sampleRef <- run1_md[, c("seqRun", "sampleID", "sampleRef", "sampleType")] %>%
  rbind(., run3_md[, c("seqRun", "sampleID", "sampleRef", "sampleType")]) %>%
  rbind(., run4_md_ill[, c("seqRun", "sampleID", "sampleRef", "sampleType")]) %>%
  rbind(., run4_md_ont[, c("seqRun", "sampleID", "sampleRef", "sampleType")]) %>%
  rbind(., run5_md[, c("seqRun", "sampleID", "sampleRef", "sampleType")]) %>%
  rbind(., run_wgs[, c("seqRun", "sampleID", "sampleRef", "sampleType")])

write.csv(sampleRef, "./project_data/sampleID_reference.csv", row.names = F)
```


