---
title: "GTscorePipeline_tamRun3"
output: html_document
date: "`r Sys.Date()`"
editor_options: 
  markdown: 
    wrap: 72
output: 
  rmdformats::downcute:
    downcute_theme: "chaos"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
```

# 1 Overview

## 1.1 About the pipeline

This pipeline is to analyze data from tamGenetics_run3 and includes all steps needed to analyze MiSeq data output, including how to 1) download MiSeq sequencing files, 2) do quality checks, 3) create the necessary files for analysis in GTscore, and 4) run the GTscore pipeline itself.

The GTscore pipeline was originally created by Garrett McKinney (see the associated github [here](https://github.com/gjmckinney/GTscore)). Here, I combine the scripts included in his original "GTscore_pipeline.R" as well as his original "README.md" - there's a good bit of overlap between these files, but a few scripts exist in one but not the other. I've also included separate analyses for each species, since not all not all loci are informative for both species.

## 1.2 About the data

The sequencing data used in this pipeline is from the full set of hair and blood samples and includes data from two MiSeq runs of this library. There are a total of **438 samples**, including 235 hair samples (138 LWED, 97 SIMP), 194 blood samples (112 LWED, 82 SIMP), and 9 negative controls.

### 1.2.1 Metadata

Metadata below includes separate files for a, b, and cat sequencing results.

To make the metadata files usable in analyses, we need to: 1) make a few formatting adjustments and 2) create a metadata file that includes sampleIDs that allow for analysis of genotypes called for each of the tamRun3 runs separately (tamRun3\_ and tamRun3b\_ sampleIDs) as well as the genotypes called when the sequencing results from the two runs are combined (cat_tamRun3\_ sampleIDs).

```{r}
# Original tamRun3 md
md_a <- read.csv("./03_tamRun3/03_run3GTscore/tamRun3_metadata_5June2023.csv") %>%
  mutate(sampleID = gsub("_","-", sampleID)) %>%
  mutate(sampleFile = paste0(sampleID, ".fastq"))

# Create tamRun3b version of md
md_b <- md_a %>%
  mutate(sampleID = gsub("tamRun3", "tamRun3b", sampleID)) %>%
  mutate(sampleFile = gsub("tamRun3", "tamRun3b", sampleFile))

# Create tamRun3cat version of md
md_cat <- md_a %>%
  mutate(sampleID = gsub("tamRun3", "cat_tamRun3", sampleID)) %>%
  mutate(sampleFile = gsub("tamRun3", "cat_tamRun3", sampleFile))

# Combine all versions of md into one
md <- rbind(md_a, md_b, md_cat) %>%
  mutate_all(str_replace_all, "XTN \\(-\\)", "xtnNeg") %>%
  mutate_all(str_replace_all, "PCR1 \\(-\\)", "pcr1Neg") %>%
  mutate(sampleID = gsub("-", "\\.", sampleID)) %>%
  mutate(
    sampleID_md = str_c("s", str_sub(sampleID, -3, -1), "_", tolower(species), "_", animalID, "_", sampleType),
    sampleID_md = gsub("xtnneg", "xtnNeg", sampleID_md),
    sampleID_md = gsub("pcr1neg", "pcr1Neg", sampleID_md)) %>%
  relocate(sampleID_md, .after = sampleID)

write.csv(md, "./03_tamRun3/03_run3GTscore/tamRun3_metadata_abcat_5June2023.csv", row.names = F)
```

# 2 Packages

```{r load packages, include=FALSE}
library(gsubfn)
library(phylotools)
library(TidyMultiqc)
library(tidyverse)
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager", version = "3.16")
BiocManager::install("Biostrings")
source("./03_tamRun3/03_run3GTscore/GTscore_modified.R") # NOTE- added N=ATGC to script for later analyses (not included in original)
```

GTscore also requires Perl and the packages "Algorithm::Combinatorics" and "Excel-Writer-XLSX". Perform the following in terminal to install on Linux. NOTE -- be sure to install both perl and the modules in root. See how to install perl packages [here](http://www.cpan.org/modules/INSTALL.html).

NOTE - the old way of installing perl through conda (conda install -c conda-forge perl) doesn't play nice anymore; better to install directly from perl:

```{bash, eval = FALSE}
# Install perl
curl -L http://xrl.us/installperlnix | bash

# Install cpanminus (makes installing perl modules easier)
cpan App::cpanminus

# Install necessary perl modules
sudo cpanm Algorithm::Combinatorics
sudo cpanm Excel::Writer::XLSX
```

I also recently had issues with installing perl modules with "cpanm" -- installing with "cpan" instead solved things. Generally speaking, perl continues to be a massive pain in new fun ways every time I open this up to re-run something.

# 3 Prep sequencing results

## 3.1 Download sequences from BaseSpace

To download all sample files, it's easiest to use the BaseSpace Command Line Interface (CLI).

To install, follow the scripts provided here: <https://developer.basespace.illumina.com/docs/content/documentation/cli/cli-overview>

Note that even if it's already installed, you may need to re-authenticate via the command below. Running it will give a link; click to open in a browser and complete authentication.

```{bash, eval = F}
bs auth
```

Once authentication is complete, you can download the desired files as follows:

```{bash, eval = F}
# Check Project IDs (outputs Project IDs for tamRun3_L1Raquel and Unindexed Reads)
bs list projects

# Sample sequences (tamRun3_L1Raquel, tamRun3b_L1)
bs download project -i 384884910 -o /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/03_tamRun3/00_run3Seqs --extension=fastq.g]

# Unindexed Reads
bs download project -i 374582211 -o /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/03_tamRun3/00_run3Seqs/unindexedReads --extension=fastq.gz
```

## 3.2 Extract fast.gz files

```{bash Extract sequencing results, eval = FALSE}
# Navigate to the directory containing the .fastq.gz files from the sequencing run
cd 00_run3Seqs

# Move all .gz files out of their individual folders and into the current directory & remove empty folders
mv */* ./

# Extract all files
gunzip -r .
```

## 3.3 Merge the two runs

```{bash}
# Merge
for i in tamRun3-*; do f=${i##*tamRun3}; cat $i tamRun3b$f > cat_$i ; done

# Check to see that total runs add up
for i in tamRun3-*; do f=${i##*tamRun3}; r1=$[$(cat cat_${i} | wc -l)/4 | bc]; echo ${i%_*} concatenated $r1; r2=$[$(cat ${i} | wc -l)/4 + $(cat tamRun3b${f} | wc -l)/4 | bc]; echo ${i%_*} separated $r2; diff=$[${r1}-${r2}]; echo ${i%_*} diff $diff; done
```

## 3.4 Interleave files

To prep for the GTscore pipeline, we need to combine reads 1 & 2 for each sample into a single interleaved file - leave a set of non-interleaved files in the original folder though to keep for fastq quality checks.

Note that we have interleaved files for tamRun3, tamRun3b, and the merged version of the two.

```{bash Interleave fasta seqs, eval = FALSE}
# Interleave reads 1 & 2 for each sample and place in a new folder
for i in ./00_run3Seqs/*_R1_001.fastq; do name=$(basename $i _R1_001.fastq); reformat.sh in=00_run3Seqs/${name}_R1_001.fastq in2=00_run3Seqs/${name}_R2_001.fastq out=02_run3Interleaved/${name}_interleaved.fastq; done

# Trim filenames to make them more manageable
for f in *.fastq; do g="${f%%_S*}.fastq"; mv "${f}" "${g}"; done
```

# 4 Quality checks

To get sequence quality scores for each sample, we can use 'fastqc' paired with 'MultiQC' -- 'fastp' is better for paired-end reads, but doesn't play nice when it comes to getting quality scores from the MultiQC report. Note that we'll be running quality checks on reads 1 and 2 separately (vs. interleaved).

## 4.1 fastqc + multiqc

*Note*: use "\--interactive" to force plots to tell you which line belongs to which sample, otherwise it will remove this option since there are so many samples in this run

```{bash eval = F}
# Activate fastqc environment
cd ../01_run3QualityChecks
conda activate fastqc-env

# Run fastqc
for i in ../00_run3Seqs/*; do fastqc $i; done

# Move fastqc files to quality checks folder
mv ../00_run3Seqs/*fastqc* .

# Activate multiqc environment
conda activate multiqc-env

# Run MultiQC on all files within the quality checks folder
multiqc --interactive .
```

## 4.2 TidyMultiqc

To convert the MultiQC into something more usable in R, we can use the package 'TidyMultiqc', which converts the 'multiqc_data.json' file into tidy data frames.

Load FastQC MultiQC report to R

```{r}
MultiQCfastQCpath <- file.path("/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/03_tamRun3/01_run3QualityChecks/multiqc_data", "multiqc_data.json")

MultiQCfastQC <- TidyMultiqc::load_multiqc(MultiQCfastQCpath, sections = c("general", "raw"))
```

### 4.2.1 Obtain median sequence quality scores

For the analyses in this section, I followed [this vignette](https://cran.r-project.org/web/packages/TidyMultiqc/vignettes/TidyMultiqc.html) provided by the makers of the TidyMultiqc package.

MultiQC reports do not provide a numerical summary statistic for read quality; they only have mapping quality and pass/fails for the per-base sequence quality. We instead need to pull this data from one of the plots -- I'm using the "Per Sequence Quality Scores" plot here:

```{r}
df <- TidyMultiqc::load_multiqc(
  MultiQCfastQCpath, 
  sections = 'plot',
  plots = "fastqc_per_sequence_quality_scores_plot")
```

This provides a nested data frame as a set of x, y pairs. As it's a histogram plot, we know that the `x` value is the quality score, and `y` is the number of times that score has been counted.

We can use tidyr to unnest the data, HistDat to create a HistDat object for each group, then purr to map each plot data frame into a row of summary statistics:

```{r}
df_unNest <- df %>%
  dplyr::mutate(
    purrr::map_dfr(plot.fastqc_per_sequence_quality_scores_plot, function(plot_df){
      hist = HistDat::HistDat(vals=plot_df$x, counts = plot_df$y)
      list(
        mean_qc = mean(hist),
        median_qc = median(hist),
        max_qc = max(hist)
      )
    }),
    plot.fastqc_per_sequence_quality_scores_plot = NULL
  )
```

This dataframe has separate rows for reads 1 & 2; we can rearrange that to make it easier to work with for our purposes. Since we're most interested in median quality, we'll just pull that value for each sample.

```{r}
seqQC_r1 <- df_unNest %>%
  select(c("metadata.sample_id", "median_qc")) %>%
  filter(grepl("R1", metadata.sample_id)) %>%
  rename(medianQC_r1 = median_qc) %>%
  mutate(sampleNo = substr(metadata.sample_id, start = 9, stop = 11))
seqQC_r2 <- df_unNest %>%
  select(c("metadata.sample_id", "median_qc")) %>%
  filter(grepl("R2", metadata.sample_id)) %>%
  rename(medianQC_r2 = median_qc) %>%
  mutate(sampleNo = substr(metadata.sample_id, start = 9, stop = 11))
seqQC <- merge(seqQC_r1, seqQC_r2, by = "sampleNo") %>%
  select(c("sampleNo", "medianQC_r1", "medianQC_r2")) %>%
  arrange(sampleNo)
```

Let's just take a peek at the distribution of sequence quality scores for now; we'll use it in our other analyses later on:

```{r}
ggplot(seqQC, aes(medianQC_r1)) +
  geom_bar(stat = "count") +
  theme_bw()

ggplot(seqQC, aes(medianQC_r2)) +
  geom_bar(stat = "count") +
  theme_bw()
```

# 5 Create primer-probe & sample files

Prior to running the GTscore pipeline itself, we need to make a few files, including 1) sample files 2) primer-probe files for each species. I'm separating them out like this so that we have more accurate read counts and genotype rates for each sample and locus.

## 5.1 Sample files

We need three sets of sample files total: 1) all samples 2) LWED samples only + all negatives 3) SIMP samples only + all negatives. Note that all sample files contain sampleIDs for separate (tamRun3 and tamRun3b) and combined (cat_tamRun3) sequencing runs.

**Quick-load**

```{r}
sf_fullSet <- read.table("./03_tamRun3/03_run3GTscore/sampleFiles_fullSet.txt")

sf_pos <- read.table("./03_tamRun3/03_run3GTscore/sampleFiles_pos.txt")

sf_lwed <- read.table("./03_tamRun3/03_run3GTscore/sampleFiles_LWED.txt")

sf_simp <- read.table("./03_tamRun3/03_run3GTscore/sampleFiles_SIMP.txt")
```

### 5.1.1 All samples

Create file for all samples in bash

```{bash}
for i in *fastq; do echo $i; done > ./../03_run3GTscore/sampleFiles_fullSet.txt
```

Load into R & create a/b/cat subsets

```{r}
sf_fullSet <- read.table("./03_tamRun3/03_run3GTscore/sampleFiles_fullSet.txt")

sf_fullSet_a <- sf_fullSet %>%
  filter(!str_detect(V1, "cat_")) %>%
  filter(str_detect(V1, "3-"))

sf_fullSet_b <- sf_fullSet %>%
  filter(str_detect(V1, "3b"))

sf_fullSet_cat <- sf_fullSet %>%
  filter(str_detect(V1, "cat"))

write.table(sf_fullSet_a, file = "./03_tamRun3/03_run3GTscore/sampleFiles_fullSet_a.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_fullSet_b, file = "./03_tamRun3/03_run3GTscore/sampleFiles_fullSet_b.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_fullSet_cat, file = "./03_tamRun3/03_run3GTscore/sampleFiles_fullSet_cat.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

Use md file to create sample file species subsets; ditch negatives from species-specific sets

```{r}
# Create subsets
lwed <- filter(md, species == "LWED")
simp <- filter(md, species == "SIMP")
posSamples <- rbind(lwed, simp)

sf_lwed_a <- sf_fullSet_a %>%
  filter(V1 %in% lwed$sampleFile)
sf_lwed_b <- sf_fullSet_b %>%
  filter(V1 %in% lwed$sampleFile)
sf_lwed_cat <- sf_fullSet_cat %>%
  filter(V1 %in% lwed$sampleFile)

sf_simp_a <- sf_fullSet_a %>%
  filter(V1 %in% simp$sampleFile)
sf_simp_b <- sf_fullSet_b %>%
  filter(V1 %in% simp$sampleFile)
sf_simp_cat <- sf_fullSet_cat %>%
  filter(V1 %in% simp$sampleFile)

sf_pos_a <- sf_fullSet_a %>%
  filter(V1 %in% posSamples$sampleFile)
sf_pos_b <- sf_fullSet_b %>%
  filter(V1 %in% posSamples$sampleFile)
sf_pos_cat <- sf_fullSet_cat %>%
  filter(V1 %in% posSamples$sampleFile)

# Export sample files
write.table(sf_lwed_a, file = "./03_tamRun3/03_run3GTscore/sampleFiles_LWED_a.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_lwed_b, file = "./03_tamRun3/03_run3GTscore/sampleFiles_LWED_b.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_lwed_cat, file = "./03_tamRun3/03_run3GTscore/sampleFiles_LWED_cat.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(sf_simp_a, file = "./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_a.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_simp_b, file = "./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_b.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_simp_cat, file = "./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_cat.txt", sep = "\t", row.names = F, col.names = F, quote = F)

write.table(sf_pos_a, file = "./03_tamRun3/03_run3GTscore/sampleFiles_pos_a.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_pos_b, file = "./03_tamRun3/03_run3GTscore/sampleFiles_pos_b.txt", sep = "\t", row.names = F, col.names = F, quote = F)
write.table(sf_pos_cat, file = "./03_tamRun3/03_run3GTscore/sampleFiles_pos_cat.txt", sep = "\t", row.names = F, col.names = F, quote = F)
```

## 4.2 Primer probe files

**Quick-load**

```{r}
pp_fullSet <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt", header = T)

pp_dual <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_dual.txt", header = T)

pp_lwed <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_LWED.txt", header = T)

pp_simp <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_SIMP.txt", header = T)

pp_lwedSpec <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_lwedSpec.txt", header = T)

pp_simpSpec <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_simpSpec.txt", header = T)
```

**File prep**

Create a primerProbeFile w/dual loci only

```{r}
pp_fullSet <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt", header = T)

pp_dual <- pp_fullSet %>%
  filter(!str_detect(Locus, "LWED")) %>%
  filter(!str_detect(Locus, "SIMP"))

pp_lwedSpec <- pp_fullSet %>%
  filter(str_detect(Locus, "LWED"))

pp_simpSpec <- pp_fullSet %>%
  filter(str_detect(Locus, "SIMP"))

write.table(pp_dual, "./03_tamRun3/03_run3GTscore/primerProbeFile_dual.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_lwedSpec, "./03_tamRun3/03_run3GTscore/primerProbeFile_lwedSpec.txt", sep = "\t", row.names = F, col.names = T, quote = F)

write.table(pp_simpSpec, "./03_tamRun3/03_run3GTscore/primerProbeFile_simpSpec.txt", sep = "\t", row.names = F, col.names = T, quote = F)
```

# 6 Count reads for amplicons

## 6.1 tamRun3_a

```{r count reads for amplicons, eval=FALSE}
# all samples, all loci (includes negatives)
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_fullSet_a.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix fullSet_a_")

# pos samples (no negs), dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_dual.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_pos_a.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix posDual_a_")

# lwed samples, lwed + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_LWED.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_LWED_a.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix LWED_a_")

# lwed samples, simp-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_simpSpec.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_LWED_a.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix LWED_simpSpec_a_")

# simp samples, simp + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_SIMP.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_a.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix SIMP_a_")

# simp samples, lwed-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_lwedSpec.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_a.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix SIMP_lwedSpec_a_")
```

## 6.2 tamRun3_b

```{r count reads for amplicons, eval=FALSE}
# all samples, all loci (includes negatives)
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_fullSet_b.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix fullSet_b_")

# pos samples (no negs), dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_dual.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_pos_b.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix posDual_b_")

# lwed samples, lwed + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_LWED.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_LWED_b.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix LWED_b_")

# lwed samples, simp-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_simpSpec.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_LWED_b.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix LWED_simpSpec_b_")

# simp samples, simp + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_SIMP.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_b.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix SIMP_b_")

# simp samples, lwed-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_lwedSpec.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_b.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix SIMP_lwedSpec_b_")
```

## 6.3 tamRun3_cat

```{r count reads for amplicons, eval=FALSE}
# all samples, all loci (includes negatives)
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_fullSet_cat.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix fullSet_cat_")

# pos samples (no negs), dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_dual.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_pos_cat.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix posDual_cat_")

# lwed samples, lwed + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_LWED.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_LWED_cat.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix LWED_cat_")

# lwed samples, simp-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_simpSpec.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_LWED_cat.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix LWED_simpSpec_cat_")

# simp samples, simp + dual loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_SIMP.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_cat.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix SIMP_cat_")

# simp samples, lwed-specific loci
system2("perl",
        args="./project_scripts/GTscore_AmpliconReadCounter_modified.pl -p ./03_tamRun3/03_run3GTscore/primerProbeFile_lwedSpec.txt --files ./03_tamRun3/03_run3GTscore/sampleFiles_SIMP_cat.txt --inDir ./03_tamRun3/02_run3Interleaved/ --outDir ./03_tamRun3/03_run3GTscore/ --prefix SIMP_lwedSpec_cat_")
```

# 7 Genotyping

## 7.1 0x cutoff

### Full set genos

```{r}
#load locus table and 0x allele reads file
fullSet_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

fullSet_a_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
fullSet_b_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
fullSet_cat_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm, adjust "0" formatting
fullSet_a_polyGenResults_singleSNP_0x <- polyGen(fullSet_singleSNP_locusTable, fullSet_a_singleSNP_alleleReads_0x)
fullSet_b_polyGenResults_singleSNP_0x <- polyGen(fullSet_singleSNP_locusTable, fullSet_b_singleSNP_alleleReads_0x)
fullSet_cat_polyGenResults_singleSNP_0x <- polyGen(fullSet_singleSNP_locusTable, fullSet_cat_singleSNP_alleleReads_0x)

# write results
write.table(fullSet_a_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(fullSet_b_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/fullSet_b_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(fullSet_cat_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/fullSet_cat_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

### posDual

```{r Genotyping}
#load locus table and 10x allele reads file
posDual_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/posDual_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

posDual_a_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/posDual_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
posDual_b_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/posDual_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
posDual_cat_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/posDual_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
posDual_a_polyGenResults_singleSNP_0x <- polyGen(posDual_singleSNP_locusTable, posDual_a_singleSNP_alleleReads_0x)
posDual_b_polyGenResults_singleSNP_0x <- polyGen(posDual_singleSNP_locusTable, posDual_b_singleSNP_alleleReads_0x)
posDual_cat_polyGenResults_singleSNP_0x <- polyGen(posDual_singleSNP_locusTable, posDual_cat_singleSNP_alleleReads_0x)

#write results
write.table(posDual_a_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/posDual_a_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(posDual_b_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/posDual_b_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(posDual_cat_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/posDual_cat_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

### LWED

```{r Genotyping}
#load locus table and 10x allele reads file
LWED_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

LWED_a_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
LWED_b_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/LWED_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
LWED_cat_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/LWED_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
LWED_a_polyGenResults_singleSNP_0x <- polyGen(LWED_singleSNP_locusTable, LWED_a_singleSNP_alleleReads_0x)
LWED_b_polyGenResults_singleSNP_0x <- polyGen(LWED_singleSNP_locusTable, LWED_b_singleSNP_alleleReads_0x)
LWED_cat_polyGenResults_singleSNP_0x <- polyGen(LWED_singleSNP_locusTable, LWED_cat_singleSNP_alleleReads_0x)

#write results
write.table(LWED_a_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/LWED_a_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(LWED_b_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/LWED_b_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(LWED_cat_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/LWED_cat_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

### SIMP

```{r Genotyping}
#load locus table and 10x allele reads file
SIMP_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

SIMP_a_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
SIMP_b_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
SIMP_cat_singleSNP_alleleReads_0x <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
SIMP_a_polyGenResults_singleSNP_0x <- polyGen(SIMP_singleSNP_locusTable, SIMP_a_singleSNP_alleleReads_0x)
SIMP_b_polyGenResults_singleSNP_0x <- polyGen(SIMP_singleSNP_locusTable, SIMP_b_singleSNP_alleleReads_0x)
SIMP_cat_polyGenResults_singleSNP_0x <- polyGen(SIMP_singleSNP_locusTable, SIMP_cat_singleSNP_alleleReads_0x)

#write results
write.table(SIMP_a_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/SIMP_a_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(SIMP_b_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/SIMP_b_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
write.table(SIMP_cat_polyGenResults_singleSNP_0x, "./03_tamRun3/03_run3GTscore/SIMP_cat_polyGenResults_singleSNP_0x.txt", quote = FALSE, sep = "\t")
```

## 7.2 10x cutoff

### Identify loci \<10x coverage

Step 1: Recode loci with \<10x coverage in the full set of allele read counts

```{r}
# Read in allele counts for the full dataset
readCounts_a <- read.table("./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs.txt")
readCounts_b <- read.table("./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs.txt")
readCounts_cat <- read.table("./03_tamRun3/03_run3GTscore/fullSet_cat_AlleleReads_singleSNPs.txt")
#readCounts <- cbind(readCounts_a, readCounts_b, readCounts_cat)

# Set up a function to sum the read counts per allele for each locus, using package gsubfn
repl <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

# Then apply the function to readCounts to sum each set of allele reads for each locus
readCounts_a_sum <- replace(readCounts_a, TRUE, lapply(readCounts_a, repl)) %>%
  mutate(across(everything(),as.numeric))
readCounts_b_sum <- replace(readCounts_b, TRUE, lapply(readCounts_b, repl)) %>%
  mutate(across(everything(),as.numeric))
readCounts_cat_sum <- replace(readCounts_cat, TRUE, lapply(readCounts_cat, repl)) %>%
  mutate(across(everything(),as.numeric))

# Recode <10x loci with "0"
readCounts_a[readCounts_a_sum < 10] <- "0,0"
readCounts_b[readCounts_b_sum < 10] <- "0,0"
readCounts_cat[readCounts_cat_sum < 10] <- "0,0"
```

Step 2: Create posDual, lwed & simp subsets & create new AlleleReads_singleSNPs files

```{r}
# lwed and simp loci sets
lwed_loci <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)

simp_loci <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_LocusTable_singleSNPs.txt",header=TRUE,stringsAsFactors=FALSE)

# subsets
readCounts_posDual_a <- readCounts_a %>%
  select(colnames(posDual_a_singleSNP_alleleReads_0x)) %>%
  filter(rownames(.) %in% rownames(posDual_a_singleSNP_alleleReads_0x))
readCounts_posDual_b <- readCounts_b %>%
  select(colnames(posDual_b_singleSNP_alleleReads_0x)) %>%
  filter(rownames(.) %in% rownames(posDual_b_singleSNP_alleleReads_0x))
readCounts_posDual_cat <- readCounts_cat %>%
  select(colnames(posDual_cat_singleSNP_alleleReads_0x)) %>%
  filter(rownames(.) %in% rownames(posDual_cat_singleSNP_alleleReads_0x))

readCounts_lwed_a <- readCounts_a %>%
  select(any_of(lwed$sampleID)) %>%
  filter(rownames(.) %in% lwed_loci$Locus_ID)
readCounts_lwed_b <- readCounts_b %>%
  select(any_of(lwed$sampleID)) %>%
  filter(rownames(.) %in% lwed_loci$Locus_ID)
readCounts_lwed_cat <- readCounts_cat %>%
  select(any_of(lwed$sampleID)) %>%
  filter(rownames(.) %in% lwed_loci$Locus_ID)

readCounts_simp_a <- readCounts_a %>%
  select(any_of(simp$sampleID)) %>%
  filter(rownames(.) %in% simp_loci$Locus_ID)
readCounts_simp_b <- readCounts_b %>%
  select(any_of(simp$sampleID)) %>%
  filter(rownames(.) %in% simp_loci$Locus_ID)
readCounts_simp_cat <- readCounts_cat %>%
  select(any_of(simp$sampleID)) %>%
  filter(rownames(.) %in% simp_loci$Locus_ID)
  
# Export new AlleleReads_singleSNPs files
write.table(readCounts_a,"./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_b,"./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_cat,"./03_tamRun3/03_run3GTscore/fullSet_cat_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")

write.table(readCounts_posDual_a,"./03_tamRun3/03_run3GTscore/posDual_a_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_posDual_b,"./03_tamRun3/03_run3GTscore/posDual_b_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_posDual_cat,"./03_tamRun3/03_run3GTscore/posDual_cat_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")

write.table(readCounts_lwed_a,"./03_tamRun3/03_run3GTscore/LWED_a_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_lwed_b,"./03_tamRun3/03_run3GTscore/LWED_b_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_lwed_cat,"./03_tamRun3/03_run3GTscore/LWED_cat_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")

write.table(readCounts_simp_a,"./03_tamRun3/03_run3GTscore/SIMP_a_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_simp_b,"./03_tamRun3/03_run3GTscore/SIMP_b_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
write.table(readCounts_simp_cat,"./03_tamRun3/03_run3GTscore/SIMP_cat_AlleleReads_singleSNPs_10x.txt",quote=FALSE,sep="\t")
```

### Full set genos

```{r}
#load locus table and 0x allele reads file
fullSet_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

fullSet_a_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
fullSet_b_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
fullSet_cat_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_cat_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm, adjust "0" formatting
fullSet_a_polyGenResults_singleSNP_10x <- polyGen(fullSet_singleSNP_locusTable, fullSet_a_singleSNP_alleleReads_10x)
fullSet_b_polyGenResults_singleSNP_10x <- polyGen(fullSet_singleSNP_locusTable, fullSet_b_singleSNP_alleleReads_10x)
fullSet_cat_polyGenResults_singleSNP_10x <- polyGen(fullSet_singleSNP_locusTable, fullSet_cat_singleSNP_alleleReads_10x)

# write results
write.table(fullSet_a_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/fullSet_a_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(fullSet_b_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/fullSet_b_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(fullSet_cat_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/fullSet_cat_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

### posDual

```{r Genotyping}
#load locus table and 10x allele reads file
posDual_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/posDual_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

posDual_a_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/posDual_a_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
posDual_b_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/posDual_b_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
posDual_cat_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/posDual_cat_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
posDual_a_polyGenResults_singleSNP_10x <- polyGen(posDual_singleSNP_locusTable, posDual_a_singleSNP_alleleReads_10x)
posDual_b_polyGenResults_singleSNP_10x <- polyGen(posDual_singleSNP_locusTable, posDual_b_singleSNP_alleleReads_10x)
posDual_cat_polyGenResults_singleSNP_10x <- polyGen(posDual_singleSNP_locusTable, posDual_cat_singleSNP_alleleReads_10x)

#write results
write.table(posDual_a_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/posDual_a_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(posDual_b_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/posDual_b_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(posDual_cat_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/posDual_cat_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

### LWED

```{r Genotyping}
#load locus table and 10x allele reads file
LWED_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

LWED_a_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
LWED_b_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/LWED_b_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
LWED_cat_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/LWED_cat_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
LWED_a_polyGenResults_singleSNP_10x <- polyGen(LWED_singleSNP_locusTable, LWED_a_singleSNP_alleleReads_10x)
LWED_b_polyGenResults_singleSNP_10x <- polyGen(LWED_singleSNP_locusTable, LWED_b_singleSNP_alleleReads_10x)
LWED_cat_polyGenResults_singleSNP_10x <- polyGen(LWED_singleSNP_locusTable, LWED_cat_singleSNP_alleleReads_10x)

#write results
write.table(LWED_a_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/LWED_a_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(LWED_b_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/LWED_b_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(LWED_cat_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/LWED_cat_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

### SIMP

```{r Genotyping}
#load locus table and 10x allele reads file
SIMP_singleSNP_locusTable <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)

SIMP_a_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
SIMP_b_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_b_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
SIMP_cat_singleSNP_alleleReads_10x <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_cat_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm
SIMP_a_polyGenResults_singleSNP_10x <- polyGen(SIMP_singleSNP_locusTable, SIMP_a_singleSNP_alleleReads_10x)
SIMP_b_polyGenResults_singleSNP_10x <- polyGen(SIMP_singleSNP_locusTable, SIMP_b_singleSNP_alleleReads_10x)
SIMP_cat_polyGenResults_singleSNP_10x <- polyGen(SIMP_singleSNP_locusTable, SIMP_cat_singleSNP_alleleReads_10x)

#write results
write.table(SIMP_a_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/SIMP_a_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(SIMP_b_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/SIMP_b_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
write.table(SIMP_cat_polyGenResults_singleSNP_10x, "./03_tamRun3/03_run3GTscore/SIMP_cat_polyGenResults_singleSNP_10x.txt", quote = FALSE, sep = "\t")
```

# 8 Data summaries

## 8.1 Locus summaries

### Summarize single SNP results for loci

**gtscore0x**

```{r}
fullSet_a_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
fullSet_b_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
fullSet_cat_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

posDual_a_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/posDual_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
posDual_b_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/posDual_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
posDual_cat_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/posDual_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

LWED_a_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
LWED_b_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/LWED_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
LWED_cat_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/LWED_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

SIMP_a_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
SIMP_b_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_b_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
SIMP_cat_singleSNP_alleleReads <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_cat_AlleleReads_singleSNPs.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#summarize single SNP results
fullSet_a_singleSNP_summary_0x <- summarizeGTscore(fullSet_a_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_a_polyGenResults_singleSNP_0x)
fullSet_b_singleSNP_summary_0x <- summarizeGTscore(fullSet_b_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_b_polyGenResults_singleSNP_0x)
fullSet_cat_singleSNP_summary_0x <- summarizeGTscore(fullSet_cat_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_cat_polyGenResults_singleSNP_0x)

posDual_a_singleSNP_summary_0x <- summarizeGTscore(posDual_a_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_a_polyGenResults_singleSNP_0x)
posDual_b_singleSNP_summary_0x <- summarizeGTscore(posDual_b_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_b_polyGenResults_singleSNP_0x)
posDual_cat_singleSNP_summary_0x <- summarizeGTscore(posDual_cat_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_cat_polyGenResults_singleSNP_0x)

LWED_a_singleSNP_summary_0x <- summarizeGTscore(LWED_a_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_a_polyGenResults_singleSNP_0x)
LWED_b_singleSNP_summary_0x <- summarizeGTscore(LWED_b_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_b_polyGenResults_singleSNP_0x)
LWED_cat_singleSNP_summary_0x <- summarizeGTscore(LWED_cat_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_cat_polyGenResults_singleSNP_0x)

SIMP_a_singleSNP_summary_0x <- summarizeGTscore(SIMP_a_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_a_polyGenResults_singleSNP_0x)
SIMP_b_singleSNP_summary_0x <- summarizeGTscore(SIMP_b_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_b_polyGenResults_singleSNP_0x)
SIMP_cat_singleSNP_summary_0x <- summarizeGTscore(SIMP_cat_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_cat_polyGenResults_singleSNP_0x)

#write results
write.table(fullSet_a_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/fullSet_a_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(fullSet_b_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/fullSet_b_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(fullSet_cat_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/fullSet_cat_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(posDual_a_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/posDual_a_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(posDual_b_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/posDual_b_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(posDual_cat_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/posDual_cat_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(LWED_a_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/LWED_a_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(LWED_b_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/LWED_b_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(LWED_cat_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/LWED_cat_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(SIMP_a_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/SIMP_a_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(SIMP_b_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/SIMP_b_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(SIMP_cat_singleSNP_summary_0x, "./03_tamRun3/03_run3GTscore/SIMP_cat_singleSNP_summary_0x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

**gtscore10x**

```{r}
#summarize single SNP results
fullSet_a_singleSNP_summary_10x <- summarizeGTscore(fullSet_a_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_a_polyGenResults_singleSNP_10x)
fullSet_b_singleSNP_summary_10x <- summarizeGTscore(fullSet_b_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_b_polyGenResults_singleSNP_10x)
fullSet_cat_singleSNP_summary_10x <- summarizeGTscore(fullSet_cat_singleSNP_alleleReads, fullSet_singleSNP_locusTable, fullSet_cat_polyGenResults_singleSNP_10x)

posDual_a_singleSNP_summary_10x <- summarizeGTscore(posDual_a_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_a_polyGenResults_singleSNP_10x)
posDual_b_singleSNP_summary_10x <- summarizeGTscore(posDual_b_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_b_polyGenResults_singleSNP_10x)
posDual_cat_singleSNP_summary_10x <- summarizeGTscore(posDual_cat_singleSNP_alleleReads, posDual_singleSNP_locusTable, posDual_cat_polyGenResults_singleSNP_10x)

LWED_a_singleSNP_summary_10x <- summarizeGTscore(LWED_a_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_a_polyGenResults_singleSNP_10x)
LWED_b_singleSNP_summary_10x <- summarizeGTscore(LWED_b_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_b_polyGenResults_singleSNP_10x)
LWED_cat_singleSNP_summary_10x <- summarizeGTscore(LWED_cat_singleSNP_alleleReads, LWED_singleSNP_locusTable, LWED_cat_polyGenResults_singleSNP_10x)

SIMP_a_singleSNP_summary_10x <- summarizeGTscore(SIMP_a_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_a_polyGenResults_singleSNP_10x)
SIMP_b_singleSNP_summary_10x <- summarizeGTscore(SIMP_b_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_b_polyGenResults_singleSNP_10x)
SIMP_cat_singleSNP_summary_10x <- summarizeGTscore(SIMP_cat_singleSNP_alleleReads, SIMP_singleSNP_locusTable, SIMP_cat_polyGenResults_singleSNP_10x)

#write results
write.table(fullSet_a_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/fullSet_a_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(fullSet_b_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/fullSet_b_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(fullSet_cat_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/fullSet_cat_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(posDual_a_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/posDual_a_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(posDual_b_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/posDual_b_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(posDual_cat_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/posDual_cat_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(LWED_a_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/LWED_a_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(LWED_b_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/LWED_b_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(LWED_cat_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/LWED_cat_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)

write.table(SIMP_a_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/SIMP_a_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(SIMP_b_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/SIMP_b_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
write.table(SIMP_cat_singleSNP_summary_10x, "./03_tamRun3/03_run3GTscore/SIMP_cat_singleSNP_summary_10x.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

### Master locus summary

Generate locus summary files with single SNP summaries + GTscore locus summaries created earlier

```{r}
# per-locus read summaries
ls_posDual_a <- read.table("./03_tamRun3/03_run3GTscore/posDual_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "posSamples_a")
ls_posDual_b <- read.table("./03_tamRun3/03_run3GTscore/posDual_b_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "posSamples_b")
ls_posDual_cat <- read.table("./03_tamRun3/03_run3GTscore/posDual_cat_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "posSamples_cat")

ls_lwed_a <- read.table("./03_tamRun3/03_run3GTscore/LWED_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples_a")
ls_lwed_b <- read.table("./03_tamRun3/03_run3GTscore/LWED_b_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples_b")
ls_lwed_cat <- read.table("./03_tamRun3/03_run3GTscore/LWED_cat_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "lwedSamples_cat")

ls_simp_a <- read.table("./03_tamRun3/03_run3GTscore/SIMP_a_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples_a")
ls_simp_b <- read.table("./03_tamRun3/03_run3GTscore/SIMP_b_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples_b")
ls_simp_cat <- read.table("./03_tamRun3/03_run3GTscore/SIMP_cat_GTscore_locusSummary.txt", header = T, sep = "\t") %>%
  mutate(sampleSet = "simpSamples_cat")

# gtscore0x per-locus depth/geno/etc summaries
ss_posDual_a_0x <- posDual_a_singleSNP_summary_0x %>%
  mutate(sampleSet = "posSamples_a")
ss_posDual_b_0x <- posDual_b_singleSNP_summary_0x %>%
  mutate(sampleSet = "posSamples_b")
ss_posDual_cat_0x <- posDual_cat_singleSNP_summary_0x %>%
  mutate(sampleSet = "posSamples_cat")

ss_lwed_a_0x <- LWED_a_singleSNP_summary_0x %>%
  mutate(sampleSet = "lwedSamples_a")
ss_lwed_b_0x <- LWED_b_singleSNP_summary_0x %>%
  mutate(sampleSet = "lwedSamples_b")
ss_lwed_cat_0x <- LWED_cat_singleSNP_summary_0x %>%
  mutate(sampleSet = "lwedSamples_cat")

ss_simp_a_0x <- SIMP_a_singleSNP_summary_0x %>%
  mutate(sampleSet = "simpSamples_a")
ss_simp_b_0x <- SIMP_b_singleSNP_summary_0x %>%
  mutate(sampleSet = "simpSamples_b")
ss_simp_cat_0x <- SIMP_cat_singleSNP_summary_0x %>%
  mutate(sampleSet = "simpSamples_cat")

# gtscore10x per-locus depth/geno/etc summaries
ss_posDual_a_10x <- posDual_a_singleSNP_summary_10x %>%
  mutate(sampleSet = "posSamples_a")
ss_posDual_b_10x <- posDual_b_singleSNP_summary_10x %>%
  mutate(sampleSet = "posSamples_b")
ss_posDual_cat_10x <- posDual_cat_singleSNP_summary_10x %>%
  mutate(sampleSet = "posSamples_cat")

ss_lwed_a_10x <- LWED_a_singleSNP_summary_10x %>%
  mutate(sampleSet = "lwedSamples_a")
ss_lwed_b_10x <- LWED_b_singleSNP_summary_10x %>%
  mutate(sampleSet = "lwedSamples_b")
ss_lwed_cat_10x <- LWED_cat_singleSNP_summary_10x %>%
  mutate(sampleSet = "lwedSamples_cat")

ss_simp_a_10x <- SIMP_a_singleSNP_summary_10x %>%
  mutate(sampleSet = "simpSamples_a")
ss_simp_b_10x <- SIMP_b_singleSNP_summary_10x %>%
  mutate(sampleSet = "simpSamples_b")
ss_simp_cat_10x <- SIMP_cat_singleSNP_summary_10x %>%
  mutate(sampleSet = "simpSamples_cat")

# master summary
ls_combo <- rbind(ls_posDual_a, ls_lwed_a, ls_simp_a,
                  ls_posDual_b, ls_lwed_b, ls_simp_b,
                  ls_posDual_cat, ls_lwed_cat, ls_simp_cat)

ss_combo_0x <- rbind(ss_posDual_a_0x, ss_lwed_a_0x, ss_simp_a_0x,
                     ss_posDual_b_0x, ss_lwed_b_0x, ss_simp_b_0x,
                     ss_posDual_cat_0x, ss_lwed_cat_0x, ss_simp_cat_0x) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

ss_combo_10x <- rbind(ss_posDual_a_10x, ss_lwed_a_10x, ss_simp_a_10x,
                     ss_posDual_b_10x, ss_lwed_b_10x, ss_simp_b_10x,
                     ss_posDual_cat_10x, ss_lwed_cat_10x, ss_simp_cat_10x) %>%
  mutate(Locus_ID = sub('[_][^_]+$', '', Locus_ID))

master_locusSummary <- ls_combo %>%
  merge(., ss_combo_0x, by.x = c("Locus", "sampleSet"), by.y = c("Locus_ID", "sampleSet")) %>%
  merge(., ss_combo_10x[, c("Locus_ID", "sampleSet", "GenotypeRate", "minAF", "majAF", "allFreqs", "conScore")], by.x = c("Locus", "sampleSet"), by.y = c("Locus_ID", "sampleSet"), suffixes = c("_0x", "_10x")) %>%
  mutate(
    metricUse = case_when(
      str_detect(Locus, "LWED") & str_detect(sampleSet, "lwedSamples") ~ "for_panelMetrics",
      str_detect(Locus, "SIMP") & str_detect(sampleSet, "simpSamples") ~ "for_panelMetrics",
      str_detect(sampleSet, "posSamples") ~ "for_panelMetrics",
      .default = "for_spMetrics"
    )
  )

write.csv(master_locusSummary, "./03_tamRun3/03_run3GTscore/summaryFiles/master_locusSummary_abcat.csv", row.names = F)
```

### Plots [not updated]

plot genotype rate

```{r locus genotype rate, warning=FALSE}
ggplot()+geom_histogram(data=ls_ss,aes(x=GenotypeRate),binwidth=0.03)+xlim(-0.01,1.01)+
  labs(title="Locus Genotype Rate", x="Genotype Rate", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot average read depth for single SNP data

```{r locus read depth}
ggplot()+geom_histogram(data=ls_ss,aes(x=AvgReadDepth),binwidth=1)+
  labs(title="Average Read Depth per SNP", x="Average Read Depth", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot genotype rate relative to average depth

```{r locus genotype rate vs read depth}
ggplot()+geom_point(data=ls_ss,aes(x=AvgReadDepth,y=GenotypeRate))+ylim(0,1)+
  labs(title="Genotype Rate vs Average Depth per SNP", x="Average Depth", y="Genotype Rate")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot distribution of minor allele frequency

```{r histogram of MAF, warning=FALSE}
ggplot()+geom_histogram(data=ls_ss,aes(x=minAF),binwidth=0.01)+
  labs(title="Minor Allele Frequency Single SNP", x="Minor Allele Frequency", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot distribution of major allele frequency

```{r histogram of MajAF, warning=FALSE}
ggplot()+geom_histogram(data=ls_ss,aes(x=majAF),binwidth=0.01)+
  labs(title="Major Allele Frequency Single SNP", x="Major Allele Frequency", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

## 8.2 Sample summaries

### Summarize single SNP summaries for samples

**gtscore 0x**

```{r}
fullSet_a_singleSNP_sampleSummary_0x <- summarizeSamples(fullSet_a_polyGenResults_singleSNP_0x, fullSet_a_singleSNP_alleleReads_0x)
fullSet_b_singleSNP_sampleSummary_0x <- summarizeSamples(fullSet_b_polyGenResults_singleSNP_0x, fullSet_b_singleSNP_alleleReads_0x)
fullSet_cat_singleSNP_sampleSummary_0x <- summarizeSamples(fullSet_cat_polyGenResults_singleSNP_0x, fullSet_cat_singleSNP_alleleReads_0x)

posDual_a_singleSNP_sampleSummary_0x <- summarizeSamples(posDual_a_polyGenResults_singleSNP_0x, posDual_a_singleSNP_alleleReads_0x)
posDual_b_singleSNP_sampleSummary_0x <- summarizeSamples(posDual_b_polyGenResults_singleSNP_0x, posDual_b_singleSNP_alleleReads_0x)
posDual_cat_singleSNP_sampleSummary_0x <- summarizeSamples(posDual_cat_polyGenResults_singleSNP_0x, posDual_cat_singleSNP_alleleReads_0x)

LWED_a_singleSNP_sampleSummary_0x <- summarizeSamples(LWED_a_polyGenResults_singleSNP_0x, LWED_a_singleSNP_alleleReads_0x)
LWED_b_singleSNP_sampleSummary_0x <- summarizeSamples(LWED_b_polyGenResults_singleSNP_0x, LWED_b_singleSNP_alleleReads_0x)
LWED_cat_singleSNP_sampleSummary_0x <- summarizeSamples(LWED_cat_polyGenResults_singleSNP_0x, LWED_cat_singleSNP_alleleReads_0x)

SIMP_a_singleSNP_sampleSummary_0x <- summarizeSamples(SIMP_a_polyGenResults_singleSNP_0x, SIMP_a_singleSNP_alleleReads_0x)
SIMP_b_singleSNP_sampleSummary_0x <- summarizeSamples(SIMP_b_polyGenResults_singleSNP_0x, SIMP_b_singleSNP_alleleReads_0x)
SIMP_cat_singleSNP_sampleSummary_0x <- summarizeSamples(SIMP_cat_polyGenResults_singleSNP_0x, SIMP_cat_singleSNP_alleleReads_0x)

write.csv(fullSet_a_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/fullSet_a_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(fullSet_b_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/fullSet_b_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(fullSet_cat_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/fullSet_cat_singleSNP_sampleSummary_0x.csv", row.names = F)

write.csv(posDual_a_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/posDual_a_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(posDual_b_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/posDual_b_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(posDual_cat_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/posDual_cat_singleSNP_sampleSummary_0x.csv", row.names = F)

write.csv(LWED_a_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/LWED_a_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(LWED_b_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/LWED_b_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(LWED_cat_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/LWED_cat_singleSNP_sampleSummary_0x.csv", row.names = F)

write.csv(SIMP_a_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/SIMP_a_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(SIMP_b_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/SIMP_b_singleSNP_sampleSummary_0x.csv", row.names = F)
write.csv(SIMP_cat_singleSNP_sampleSummary_0x, "./03_tamRun3/03_run3GTscore/SIMP_cat_singleSNP_sampleSummary_0x.csv", row.names = F)
```

**gtscore 10x**

```{r}
fullSet_a_singleSNP_sampleSummary_10x <- summarizeSamples(fullSet_a_polyGenResults_singleSNP_10x, fullSet_a_singleSNP_alleleReads_10x)
fullSet_b_singleSNP_sampleSummary_10x <- summarizeSamples(fullSet_b_polyGenResults_singleSNP_10x, fullSet_b_singleSNP_alleleReads_10x)
fullSet_cat_singleSNP_sampleSummary_10x <- summarizeSamples(fullSet_cat_polyGenResults_singleSNP_10x, fullSet_cat_singleSNP_alleleReads_10x)

posDual_a_singleSNP_sampleSummary_10x <- summarizeSamples(posDual_a_polyGenResults_singleSNP_10x, posDual_a_singleSNP_alleleReads_10x)
posDual_b_singleSNP_sampleSummary_10x <- summarizeSamples(posDual_b_polyGenResults_singleSNP_10x, posDual_b_singleSNP_alleleReads_10x)
posDual_cat_singleSNP_sampleSummary_10x <- summarizeSamples(posDual_cat_polyGenResults_singleSNP_10x, posDual_cat_singleSNP_alleleReads_10x)

LWED_a_singleSNP_sampleSummary_10x <- summarizeSamples(LWED_a_polyGenResults_singleSNP_10x, LWED_a_singleSNP_alleleReads_10x)
LWED_b_singleSNP_sampleSummary_10x <- summarizeSamples(LWED_b_polyGenResults_singleSNP_10x, LWED_b_singleSNP_alleleReads_10x)
LWED_cat_singleSNP_sampleSummary_10x <- summarizeSamples(LWED_cat_polyGenResults_singleSNP_10x, LWED_cat_singleSNP_alleleReads_10x)

SIMP_a_singleSNP_sampleSummary_10x <- summarizeSamples(SIMP_a_polyGenResults_singleSNP_10x, SIMP_a_singleSNP_alleleReads_10x)
SIMP_b_singleSNP_sampleSummary_10x <- summarizeSamples(SIMP_b_polyGenResults_singleSNP_10x, SIMP_b_singleSNP_alleleReads_10x)
SIMP_cat_singleSNP_sampleSummary_10x <- summarizeSamples(SIMP_cat_polyGenResults_singleSNP_10x, SIMP_cat_singleSNP_alleleReads_10x)

write.csv(fullSet_a_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/fullSet_a_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(fullSet_b_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/fullSet_b_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(fullSet_cat_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/fullSet_cat_singleSNP_sampleSummary_10x.csv", row.names = F)

write.csv(posDual_a_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/posDual_a_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(posDual_b_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/posDual_b_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(posDual_cat_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/posDual_cat_singleSNP_sampleSummary_10x.csv", row.names = F)

write.csv(LWED_a_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/LWED_a_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(LWED_b_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/LWED_b_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(LWED_cat_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/LWED_cat_singleSNP_sampleSummary_10x.csv", row.names = F)

write.csv(SIMP_a_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/SIMP_a_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(SIMP_b_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/SIMP_b_singleSNP_sampleSummary_10x.csv", row.names = F)
write.csv(SIMP_cat_singleSNP_sampleSummary_10x, "./03_tamRun3/03_run3GTscore/SIMP_cat_singleSNP_sampleSummary_10x.csv", row.names = F)
```

### Master sample summary

Sample metrics for genotypes and such are based on species sample/loci sets -

-   LWED sample metrics are based on dual loci + lwed-specific loci
-   SIMP sample metrics are based on dual loci + simp-specific loci
-   negCtrl sample metrics are based on all loci

```{r}
# import individual summaries from AmpliconReadCounter.pl output
is_neg_a <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_a_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  filter(!Sample %in% lwed$sampleID) %>%
  filter(!Sample %in% simp$sampleID)
is_neg_b <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_b_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  filter(!Sample %in% lwed$sampleID) %>%
  filter(!Sample %in% simp$sampleID)
is_neg_cat <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_cat_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample)) %>%
  filter(!Sample %in% lwed$sampleID) %>%
  filter(!Sample %in% simp$sampleID)

is_lwed_a <- read.delim("./03_tamRun3/03_run3GTscore/LWED_a_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))
is_lwed_b <- read.delim("./03_tamRun3/03_run3GTscore/LWED_b_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))
is_lwed_cat <- read.delim("./03_tamRun3/03_run3GTscore/LWED_cat_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))

is_simp_a <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_a_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))
is_simp_b <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_b_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))
is_simp_cat <- read.delim("./03_tamRun3/03_run3GTscore/SIMP_cat_GTscore_individualSummary.txt", header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Sample = gsub("-", "\\.", Sample))

# gtscore0x sample summary (from above)
sampleSum_fullSet_a_0x <- fullSet_a_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "fullSet")
sampleSum_fullSet_b_0x <- fullSet_b_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "fullSet")
sampleSum_fullSet_cat_0x <- fullSet_cat_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "fullSet")

sampleSum_neg_a_0x <- sampleSum_fullSet_a_0x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)
sampleSum_neg_b_0x <- sampleSum_fullSet_b_0x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)
sampleSum_neg_cat_0x <- sampleSum_fullSet_cat_0x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)

sampleSum_lwed_a_0x <- LWED_a_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "lwedSet")
sampleSum_lwed_b_0x <- LWED_b_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "lwedSet")
sampleSum_lwed_cat_0x <- LWED_cat_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "lwedSet")

sampleSum_simp_a_0x <- SIMP_a_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "simpSet")
sampleSum_simp_b_0x <- SIMP_b_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "simpSet")
sampleSum_simp_cat_0x <- SIMP_cat_singleSNP_sampleSummary_0x %>%
  mutate(lociSet = "simpSet")

# gtscore10x sample summary (from above)
sampleSum_fullSet_a_10x <- fullSet_a_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "fullSet")
sampleSum_fullSet_b_10x <- fullSet_b_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "fullSet")
sampleSum_fullSet_cat_10x <- fullSet_cat_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "fullSet")

sampleSum_neg_a_10x <- sampleSum_fullSet_a_10x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)
sampleSum_neg_b_10x <- sampleSum_fullSet_b_10x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)
sampleSum_neg_cat_10x <- sampleSum_fullSet_cat_10x %>%
  filter(!sample %in% lwed$sampleID) %>%
  filter(!sample %in% simp$sampleID)

sampleSum_lwed_a_10x <- LWED_a_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "lwedSet")
sampleSum_lwed_b_10x <- LWED_b_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "lwedSet")
sampleSum_lwed_cat_10x <- LWED_cat_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "lwedSet")

sampleSum_simp_a_10x <- SIMP_a_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "simpSet")
sampleSum_simp_b_10x <- SIMP_b_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "simpSet")
sampleSum_simp_cat_10x <- SIMP_cat_singleSNP_sampleSummary_10x %>%
  mutate(lociSet = "simpSet")

# combos
indSum_combo <- rbind(is_neg_a, is_lwed_a, is_simp_a,
                      is_neg_b, is_lwed_b, is_simp_b,
                      is_neg_cat, is_lwed_cat, is_simp_cat) %>%
  arrange(Sample)

sampleSum_combo_0x <- rbind(sampleSum_neg_a_0x, sampleSum_lwed_a_0x, sampleSum_simp_a_0x,
                            sampleSum_neg_b_0x, sampleSum_lwed_b_0x, sampleSum_simp_b_0x,
                            sampleSum_neg_cat_0x, sampleSum_lwed_cat_0x, sampleSum_simp_cat_0x) %>%
  rbind() %>%
  arrange(sample)

sampleSum_combo_10x <- rbind(sampleSum_neg_a_10x, sampleSum_lwed_a_10x, sampleSum_simp_a_10x,
                            sampleSum_neg_b_10x, sampleSum_lwed_b_10x, sampleSum_simp_b_10x,
                            sampleSum_neg_cat_10x, sampleSum_lwed_cat_10x, sampleSum_simp_cat_10x) %>%
  rbind() %>%
  arrange(sample) %>%
  select(-lociSet)

# master summary
master_sampleSummary <- indSum_combo %>%
  merge(., sampleSum_combo_0x, by.x = "Sample", by.y = "sample") %>%
  merge(., sampleSum_combo_10x, by.x = "Sample", by.y = "sample", suffixes = c("_0x", "_10x")) %>%
  relocate(lociSet, .after = Sample) %>%
  dplyr::rename("sampleID" = "Sample") %>%
  # add metadata
  merge(., md[, c("sampleID", "sampleID_md")], by = "sampleID") %>%
  relocate(sampleID_md, .after = sampleID)

write.csv(master_sampleSummary, "./03_tamRun3/03_run3GTscore/summaryFiles/master_sampleSummary_abcat.csv", row.names = F)
```

### Plots [not updated]

plot histogram of genotype rate

```{r histogram of genotype rate, warning=FALSE}
ggplot()+geom_histogram(data=master_sampleSummary,aes(x=GenotypeRate),binwidth=0.01)+xlim(-0.01,1.01)+
  labs(title="Sample Genotype Rate", x="Genotype Rate", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot histogram of heterozygosity

```{r histogram of heterozygosity, warning=FALSE}
ggplot()+geom_histogram(data=master_sampleSummary,aes(x=Heterozygosity),binwidth=0.03)+xlim(-0.01,1.01)+
  labs(title="Sample Heterozygosity", x="Heterozygosity", y="Count")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

plot genotype rate vs primer probe reads

```{r genotype rate vs primer probe reads}
#dashed line added at 90% genotype rate, this is not a strict threshold, just a goal to aim for
ggplot(data=master_sampleSummary,aes(x=Primer.Probe.Reads,y=GenotypeRate,color=sampleType))+
  geom_point(stat = "identity")+
  geom_smooth(method = "loess")+
  labs(title="Genotype Rate vs Total Reads per Sample", x="Primer Probe Reads", y="Genotype Rate")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))+
  geom_hline(yintercept=0.5,lty="dashed")
```

Samples with unusually high heterozygosity may be contaminated or have elevated ploidy. [[Given chimerism, however, this isn't super helpful for blood samples]]

plot heterozygosity vs primer probe reads

```{r heterozygosity vs primer probe reads}
ggplot()+geom_point(data=master_sampleSummary,aes(x=Primer.Probe.Reads,y=Heterozygosity))+
  labs(title="Heterozygosity vs Total Reads per Sample", x="Primer Probe Reads", y="Heterozygosity")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))+
  geom_hline(yintercept=0.3, lty="dashed")
```

# 8 Quality control

## 8.1 Identify duplicate samples

The IDduplicateSamples function does all pairwise comparisons of samples and outputs two metrics:

-   proportionCommon - The proportion of loci that had genotypes for both samples in a pair
-   proportionMatch - The proportion of loci that have identical genotypes in the sample pair

```{r identify duplicate samples,eval=FALSE}
#convert missing genotypes "0" to NA
polyGenResults_singleSNP_NA<-fullSet_cat_polyGenResults_singleSNP
polyGenResults_singleSNP_NA[polyGenResults_singleSNP_NA=="0"]<-NA

#compare all samples, for each comparison get proportion of loci that have genotypes
#in both samples (proportionCommon) and proportion of shared loci that have identical
#genotypes (proportionMatch).
#Samples with a high proportionCommon and high proportionMatch are likely duplicates
polyGenResults_dupTest<-IDduplicateSamples(polyGenResults_singleSNP_NA)

write.table(polyGenResults_dupTest,"run3_cat_polyGenResults_dupTest.txt",quote=FALSE,sep="\t",row.names=FALSE)

```

```{r duplicate sample results}
polyGenResults_dupTest<-read.delim("run2_polyGenResults_dupTest.txt",stringsAsFactors=FALSE)
head(polyGenResults_dupTest)
```

polyGenResults_dupTest

Plot results of IDduplicateSamples

McKinney says that he generally sets thresholds of proportionMatch \> 0.8 and proportionCommon \> 0.75 to identify duplicate sample pairs, but the appropriate thresholds will depend on the marker set being used and the relatedness among samples. In this plot, he notes that it is possible that individuals at the lower range of identical genotypes (\~0.85) are relatives rather than duplicated samples.

```{r plot IDduplicateSamples results}
#potential duplicates are proportionMatch>0.8 and proportionCommon>0.75
#feel free to adjust these thresholds as needed
matchThresh=0.8
commonThresh=0.75 #switched to 0, looks like some loci were successfully genotyped for one sample and some for the other

#plot results
ggplot()+geom_point(data=polyGenResults_dupTest,aes(x=proportionCommon,y=proportionMatch))+
  geom_segment(aes(x=commonThresh,xend=1,y=matchThresh,yend=matchThresh),lty="dashed")+
  geom_segment(aes(x=commonThresh,xend=commonThresh,y=matchThresh,yend=1),lty="dashed")
```

Identify duplicate sample pairs with thresholds set previously

```{r}
#filter to potentially duplicated samples using thresholds above
polyGenResults_dupTest %>% filter(proportionMatch>=matchThresh,proportionCommon>=commonThresh)
```

## 8.2 Identify contaminated samples

Contaminated samples can be identified through elevated heterozygosity. NOTE- this isn't going to be helpful for us given the blood chimerism.

```{r identify contaminated samples (heterozygosity)}
#plot heterozygosity vs genotype rate per sample
#samples with unusually high heterozygosity relative to others are candidates for contamination
#the sample above the dashed line in this example is likely contaminated
ggplot()+geom_point(data=master_sampleSummary,aes(x=GenotypeRate,y=Heterozygosity))+
  labs(title="Heterozygosity vs Genotype Rate per Sample", x="Genotype Rate", y="Heterozygosity")+
  theme_bw()+theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))+
  geom_hline(yintercept=0.30, lty="dashed")
```

Set heterozygosity treshold to identify likely contaminated sample

```{r}
#identify likely contaminated sample
contaminatedSample<-master_sampleSummary %>% filter(Heterozygosity>0.30) %>% pull(Sample)
contaminatedSample
```

Contaminated samples can also be identified by an elevated contamination score. In this example, samples with a contamination score \> 0.3 are known to be contaminated.

```{r }
#plot histogram of contamination score
ggplot()+geom_histogram(data=master_sampleSummary,aes(x=conScore),binwidth=0.02)+geom_vline(xintercept=0.3,lty="dashed")

```

Set contamination score threshold to identify contaminated samples

```{r}
#identify likely contaminated samples
contaminatedSamples2<-master_sampleSummary %>% filter(conScore>=0.3) %>% pull(Sample)
contaminatedSamples2
```

Plot genotype scatterplots for each sample

```{r}
#Scatter Plots can show evidence of contamination or elevated ploidy
plotGenotypes_sample(fullSet_singleSNP_locusTable, fullSet_singleSNP_alleleReads, fullSet_polyGenResults_singleSNP, type='scatter', savePlot="Y", saveDir="./scatterPlots_sample")
#look at plots for samples, particularly for putatively contaminated samples identified by high heterozygosity
```

## 8.3 Identify poor quality loci to remove

Flag SNPs with \< 50% genotype rate as candidates for removal. Will need to do this with LWED and SIMP samples separately, then recombine.

```{r}
poorQualitySNPs <- ls_ss %>% filter(GenotypeRate<0.5) %>% mutate(Locus=as.character(Locus)) %>% pull(Locus)

View(poorQualitySNPs)

pq_lwed <- data.frame(name=poorQualitySNPs_LWED) %>%
  mutate(species="LWED")
pq_simp<- data.frame(name=poorQualitySNPs_SIMP) %>%
  mutate(species="SIMP")

pq_both <- rbind(pq_lwed, pq_simp)
View(pq_both)

pq_both2 <- pq_both %>%
  group_by(name) %>%
  mutate(count=n()) %>%
  distinct(name, .keep_all = T) %>%
  within(species[count == '2'] <- "both")
View(pq_both2)

# Export file
write.csv(pq_both2, "./summaryFiles/poorQualitySNPs_fullSet.csv", row.names = F)
```

## 8.4 Plots

Ratio & Scatter Plots can show evidence of systemic issues in read counting (off-target reads, biases) or elevated ploidy

Plot genotype ratio & scatterplots for each locus

```{r}
#Allele Ratio Plots  
plotGenotypes(singleSNP_locusTable, singleSNP_alleleReads, polyGenResults_singleSNP, type='ratio', savePlot="Y", saveDir="ratioPlots_loci")
#Scatter Plots 
plotGenotypes(singleSNP_locusTable, singleSNP_alleleReads, polyGenResults_singleSNP, type='scatter', savePlot="Y", saveDir="scatterPlots_loci")
```

Plot histogram of contamination score for loci. An arbitrary score of 0.3 was set to identify problematic loci, but a lower or higher threshold may be more suitable. Visual examination of loci with elevated scores is recommended. In cases where there are very few heterozygous individuals, a high contamination score may be not truly reflect locus performance..

```{r, warning=FALSE}
#loci with high contamination scores should have scatterplots examined to ensure genotypes look accurate
fullSet_singleSNP_summary %>% ggplot(data=.) + geom_histogram(aes(x=conScore),binwidth=0.01)+geom_vline(xintercept=0.3)
```

Plot contamination score vs minor allele frequency. Loci with both a high contamination score and high minor allele frequency are likely to be problematic.

```{r, warning=FALSE}
fullSet_singleSNP_summary %>% ggplot(data=.) + geom_point(aes(x=conScore,y=minAF))
```

Get list of loci with high contamination score

```{r}
singleSNP_summary %>% filter(conScore>0.3)
```

Look at scatterplots for loci with high contamination score. See orignal GTscore README.Rmd for examples of a paralog, locus w/likely off-target sequence, locus w/high contamination score but may be fine.

# 9 Export genotype files

When exporting files, you can either export the full results and remove any poor quality samples or loci afterwards, or you can use whitelists or blacklists to filter during export. Both the exportGenepop and exportRubias functions have the following options: sampleWhitelist, locusWhitelist, sampleBlacklist, locusBlacklist. The whitelists are a list of samples or loci to retain, the blacklists are lists of samples or loci to discard.

## 9.1 Genepop format



export genepop format - poor quality loci removed

```{r}
#example with poor quality loci and contaminated sample removed
exportGenepop(polyGenResults_singleSNP,singleSNP_locusTable,
              locusBlacklist=poorQualitySNPs,
              filename="polyGenResults_singleSNP_genepop_filtered.txt")
```

## 9.2 Rubias format

export Rubias format with poor quality loci and contaminated sample removed

```{r}
exportRubias(polyGenResults_singleSNP,singleSNP_locusTable,sampleMetaData,
             locusBlacklist=poorQualitySNPs,
             filename="polyGenResults_singleSNP_rubias_filtered.txt")
exportRubias(polyGenResults_singleSNP_LWED,singleSNP_locusTable_LWED,sampleMetaData,
             locusBlacklist=poorQualitySNPs_LWED,
             filename="polyGenResults_singleSNP_rubias_filtered_LWED.txt")
exportRubias(polyGenResults_singleSNP_SIMP,singleSNP_locusTable_SIMP,sampleMetaData,
             locusBlacklist=poorQualitySNPs_SIMP,
             filename="polyGenResults_singleSNP_rubias_filtered_SIMP.txt")
```
