---
title: "tamRun5_genoAnalyses"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: 
  rmdformats::downcute:
    downcute_theme: "chaos"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
```

```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto
}

pre[class] {
  max-height: 100px;
}
```


# 1 Overview

This pipeline is for the analysis of genotypes acquired via GTscorePipeline_tamRun5.Rmd, including comparisons to metadata, investigation of species/sex mismatches, and comparisons to genotypes from tamRun1 and tamRun2.

# 2 Packages

```{r}
library(forcats) # reorder factor by another variable
library(huxtable) # for pretty number formatting
library(janitor) # duptest
library(kableExtra) # pretty tables
library(plotrix) # plotrix::std.error()
library(purrr) # for pmap
library(scales) # scales::percentage()
library(tidyverse) # general data manipulation
```

Fix scrolling issue with large dataframes:

```{r}
rstudioapi::writeRStudioPreference("data_viewer_max_columns", 150L)
```

# 3 Data

## 3.1 Metadata

### General metadata

Below is the metadata for samples included in tamRun5 - note that this metadata file has been updated in metadataReconciliation.Rmd to address sampleName issues; the alternate animalIDs + their metadata are included in the md file below. I've also included columns with sample names in previous runs if the same samples were run in tamRun1, tamRun2, or tamRun3 so that I can compare genotypes when possible.

```{r}
# tamRun5 md - updated with metadataReconciliation.Rmd
md_tamRun5 <- read.csv("./metadataReconciliation/tamRun5_metadata_v2.csv") %>%
  # add unique sampleID w/animalID + sampleType
  mutate(sampleID_unique = str_c(animalID, sampleType, sep = "_")) %>%
  relocate(sampleID_unique)

View(md_tamRun5 %>%
       filter(sampleID %in% c("tamRun5.280", "tamRun5.288")))
```

Below is the original script for tamRun5 md; I have a more recently updated copy imported above, but keeping the stuff below for records

```{r, include=FALSE}
# tamRun1 metadata
md_tamRun1 <- read.csv("./00_tamRun1_optimization/metadata_run1.csv") %>%
  mutate(sampleID = gsub("_","\\.", sampleID)) %>%
  mutate(species = sub("SFUS", "LWED", species)) %>%
  unite("sample", c(animalID,sampleType), sep = "_", remove = F)

# tamRun1 sample list + animalID & sampletype
samples_tamRun1 <- md_tamRun1 %>%
  select(c(sampleID, animalID, sampleType)) %>%
  distinct() %>%
  na.omit() %>%
  group_by(animalID, sampleType) %>%
  summarise(tamRun1 = toString(sampleID)) %>%
  ungroup()

# tamRun2 metadata
md_tamRun2 <- read.csv("./01_run2_fecalHairBlood/03_run2GTscore/metadata_tamRun2.csv") %>%
  filter(included_inRun == "yes") %>%
  mutate(sampleID = gsub("_","\\.", sampleID)) %>%
  mutate(species = sub("SFUS", "LWED", species)) %>%
  unite("sample", c(animalID,sampleType), sep = "_", remove = F)

# tamRun2 sample list + animalID & sampletype
samples_tamRun2 <- md_tamRun2 %>%
  select(c(sampleID, animalID, sampleType)) %>%
  distinct() %>%
  na.omit() %>%
  group_by(animalID, sampleType) %>%
  summarise(tamRun2 = toString(sampleID)) %>%
  ungroup()

# tamRun5 md
md_tamRun5 <- read.csv("./05_tamRun5/03_run5GTscore/tamRun5_metadata.csv") %>%
  mutate(
    animalID = 
      case_when(
        sampleID == "tamRun5-411" ~ "pcr1Neg_p5.C4",
        .default = animalID
      ),
    sampleType =
      case_when(
        sampleID == "tamRun5-411" ~ "pcr1Neg",
        .default = sampleType
      )
  ) %>%
  # add sampleIDs from samples run in tamRun1
  merge(., samples_tamRun1, by = c("animalID", "sampleType"), all.x = T) %>%
  # add sampleIDs from samples run in tamRun2
  merge(., samples_tamRun2, by = c("animalID", "sampleType"), all.x = T) %>%
  mutate(sampleID = gsub("-", "\\.", sampleID))

# Export updated tamRun5 metadata
write.csv(md_tamRun5, "./05_tamRun5/03_run5GTscore/tamRun5_metadata.csv", row.names = F)
```

### Primer list

```{r}
primerList_v3 <- read.csv("./primers/03_lociChoices/tamGenetics_primerList_v3.csv")
```

### Sample lists

```{r}
lwedSamples <- md_tamRun5 %>%
  filter(species == "LWED") %>%
  select(sampleID)

simpSamples <- md_tamRun5 %>%
  filter(species == "SIMP") %>%
  select(sampleID)

bloodSamples <- md_tamRun5 %>%
  filter(sampleType == "blood") %>%
  select(sampleID)

hairSamples <- md_tamRun5 %>%
  filter(sampleType == "hair") %>%
  select(sampleID)
```

## 3.2 Genotypes

Genotypes for all runs are below.

**NOTE:**

-   Except for tamRun1, genotypes were only called for loci with at least 10x coverage
-   For tamRun1 and tamRun2, some locus names need to be recoded to reflect species-specificity
-   tamRun3 genotypes include those called separately for each library run as well as those called when the sequencing results from both runs were combined

```{r}
genos_tamRun1 <- read.table("./00_tamRun1_optimization/polyGenResults_singleSNP.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  mutate(locus = recode(locus,
                        'SEXID_195' = 'SEXID_LWED_195',
                        'SEXID_197' = 'SEXID_SIMP_197',
                        'SEXID_198' = 'SEXID_SIMP_198',
                        'SEXID_203' = 'SEXID_SIMP_203',
                        'SEXID_208' = 'SEXID_LWED_208',
                        'SEXID_211' = 'SEXID_LWED_211',
                        'SEXID_218' = 'SEXID_SIMP_218')) %>%
  column_to_rownames("locus")

genos_tamRun2 <- read.table("./01_run2_fecalHairBlood/03_run2GTscore/fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('\\..*', '', locus)) %>%
  mutate(locus = recode(locus,
                        'SEXID_195' = 'SEXID_LWED_195',
                        'SEXID_197' = 'SEXID_SIMP_197',
                        'SEXID_198' = 'SEXID_SIMP_198',
                        'SEXID_203' = 'SEXID_SIMP_203',
                        'SEXID_208' = 'SEXID_LWED_208',
                        'SEXID_211' = 'SEXID_LWED_211',
                        'SEXID_218' = 'SEXID_SIMP_218')) %>%
  column_to_rownames("locus")

genos_tamRun5_10x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_10x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

genos_tamRun5_30x <- read.table("./05_tamRun5/03_run5GTscore/fullSet_polyGenResults_singleSNP_30x.txt", header = T) %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub('[_][^_]+$', '', locus)) %>%
  column_to_rownames("locus")

genos_tamRun5_gtseq <- read_csv("./05_tamRun5/05_run5GTseq/genos/compiledGenos_gtseq_tamRun5.csv") %>%
  as.data.frame() %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  # read_csv adds a comma to last column - remove this
  mutate(
    SPECIESID_9 = gsub(",", "", SPECIESID_9)
  ) %>%
  select(-Raw.Reads, -On.Target.Reads, -X.On.Target, -X.GT, -IFI) %>%
  column_to_rownames("Sample") %>%
  t() %>%
  as.data.frame() %>%
  mutate_all(funs(str_replace(., "00", "0")))
```

## 3.3 Summaries

The locus and sample summaries below are both products of GTscorePipeline_tamRun5.Rmd and provide metrics such as genotype rate and on-/off-target coverage depth. The sample master summary includes all metadata as well.

### 10x summaries

```{r}
# Locus summary
locusSum_10x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/master_locusSummary_10x.csv") %>%
  mutate(set =
           case_when(
             str_detect(Locus, "INDID") ~ "INDID",
             str_detect(Locus, "SEXID_LWED") ~ "SEXID_LWED",
             str_detect(Locus, "SEXID_SIMP") ~ "SEXID_SIMP",
             str_detect(Locus, "SEXID") ~ "SEXID",
             str_detect(Locus, "LWED") ~ "LWED",
             str_detect(Locus, "SIMP") ~ "SIMP",
             str_detect(Locus, "SPECIES") ~ "SPECIES"
           ))

# Sample summaries
sampleSum_10x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/master_sampleSummary_10x.csv") %>%
  mutate(sampleID = gsub("-", "\\.", sampleID))

sampleSum_noNegs_10x <- sampleSum_10x %>%
  filter(!str_detect(sampleType, "Neg"))

sampleSum_blood_10x <- sampleSum_10x %>%
  filter(sampleType == "blood")

sampleSum_hair_10x <- sampleSum_10x %>%
  filter(sampleType == "hair")
```

### 30x summaries

```{r}
# Locus summary
locusSum_30x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/master_locusSummary_30x.csv") %>%
  mutate(set =
           case_when(
             str_detect(Locus, "INDID") ~ "INDID",
             str_detect(Locus, "SEXID_LWED") ~ "SEXID_LWED",
             str_detect(Locus, "SEXID_SIMP") ~ "SEXID_SIMP",
             str_detect(Locus, "SEXID") ~ "SEXID",
             str_detect(Locus, "LWED") ~ "LWED",
             str_detect(Locus, "SIMP") ~ "SIMP",
             str_detect(Locus, "SPECIES") ~ "SPECIES"
           ))

# Sample summaries
sampleSum_30x <- read.csv("./05_tamRun5/03_run5GTscore/summaryFiles/master_sampleSummary_30x.csv") %>%
  mutate(sampleID = gsub("-", "\\.", sampleID))

sampleSum_noNegs_30x <- sampleSum_30x %>%
  filter(!str_detect(sampleType, "Neg"))

sampleSum_blood_30x <- sampleSum_30x %>%
  filter(sampleType == "blood")

sampleSum_hair_30x <- sampleSum_30x %>%
  filter(sampleType == "hair")
```

### gtseq summaries

```{r}
# Locus summaries

# Sample summaries
sampleSum_gtseq <- read_csv("./05_tamRun5/05_run5GTseq/genos/compiledGenos_gtseq_tamRun5.csv") %>%
  as.data.frame() %>%
  dplyr::rename_all(funs(make.names(.))) %>%
  mutate(
    sampleID = sub("-", "\\.", Sample),
    GenotypeRate = round(X.GT/100, 2)
    ) %>%
  select(c("sampleID", "Raw.Reads", "On.Target.Reads", "X.On.Target", "GenotypeRate")) %>%
  merge(., md_tamRun5, by = "sampleID")

sampleSum_noNegs_gtseq <- sampleSum_gtseq %>%
  filter(!str_detect(sampleType, "Neg"))
  
sampleSum_blood_gtseq <- sampleSum_gtseq %>%
  filter(sampleType == "blood")

sampleSum_hair_gtseq <- sampleSum_gtseq %>%
  filter(sampleType == "hair")
```

## 3.4 Sample counts

```{r, echo=FALSE}
# LWED metrics
tamRun5_lwedCount <- sum(md_tamRun5$species == "LWED",
                         na.rm = T)
tamRun5_lwedHairCount <- sum(md_tamRun5$sampleType == "hair"  &
                               md_tamRun5$species == "LWED",
                             na.rm = T)
tamRun5_lwedHairFCount <- sum(md_tamRun5$sampleType == "hair"  &
      md_tamRun5$species == "LWED" &
      md_tamRun5$sex == "F",
                             na.rm = T)
tamRun5_lwedHairMCount <- sum(md_tamRun5$sampleType == "hair"  &
      md_tamRun5$species == "LWED" &
      md_tamRun5$sex == "M",
                             na.rm = T)
tamRun5_lwedBloodCount <- sum(md_tamRun5$sampleType == "blood"  &
                               md_tamRun5$species == "LWED",
                             na.rm = T)
tamRun5_lwedBloodFCount <- sum(md_tamRun5$sampleType == "blood"  &
      md_tamRun5$species == "LWED" &
      md_tamRun5$sex == "F",
                             na.rm = T)
tamRun5_lwedBloodMCount <- sum(md_tamRun5$sampleType == "blood"  &
      md_tamRun5$species == "LWED" &
      md_tamRun5$sex == "M",
                             na.rm = T)

# SIMP metrics
tamRun5_simpCount <- sum(md_tamRun5$species == "SIMP",
                             na.rm = T)
tamRun5_simpHairCount <- sum(md_tamRun5$sampleType == "hair"  &
                               md_tamRun5$species == "SIMP",
                             na.rm = T)
tamRun5_simpHairFCount <- sum(md_tamRun5$sampleType == "hair"  &
      md_tamRun5$species == "SIMP" &
      md_tamRun5$sex == "F",
                             na.rm = T)
tamRun5_simpHairMCount <- sum(md_tamRun5$sampleType == "hair"  &
      md_tamRun5$species == "SIMP" &
      md_tamRun5$sex == "M",
                             na.rm = T)
tamRun5_simpBloodCount <- sum(md_tamRun5$sampleType == "blood"  &
                               md_tamRun5$species == "SIMP",
                             na.rm = T)
tamRun5_simpBloodFCount <- sum(md_tamRun5$sampleType == "blood"  &
      md_tamRun5$species == "SIMP" &
      md_tamRun5$sex == "F",
                             na.rm = T)
tamRun5_simpBloodMCount <- sum(md_tamRun5$sampleType == "blood"  &
      md_tamRun5$species == "SIMP" &
      md_tamRun5$sex == "M",
                             na.rm = T)

# Negatives
tamRun5_negCount <- sum(str_detect(md_tamRun5$sampleType, "Neg"),
                             na.rm = T)

# Table
sampleOverview_tamRun5 <- matrix(c(tamRun5_lwedHairCount,
                                   tamRun5_lwedHairFCount,
                                   tamRun5_lwedHairMCount,
                                   tamRun5_lwedBloodCount,
                                   tamRun5_lwedBloodFCount,
                                   tamRun5_lwedBloodMCount,
                                   tamRun5_lwedCount,
                                   tamRun5_simpHairCount,
                                   tamRun5_simpHairFCount,
                                   tamRun5_simpHairMCount,
                                   tamRun5_simpBloodCount,
                                   tamRun5_simpBloodFCount,
                                   tamRun5_simpBloodMCount,
                                   tamRun5_simpCount,
                                   NA, NA, NA, NA, NA, NA,
                                   tamRun5_negCount),
                                    nrow = 7,
                                    ncol = 3) %>%
  `colnames<-`(., c("LWED", "SIMP", "Negatives")) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric) %>%
  rowwise() %>%
  mutate(TOTAL = sum(LWED, SIMP, Negatives, na.rm = T)) %>%
  as.data.frame() %>%
  `rownames<-`(., c("Total hair samples", "Female hair samples", "Male hair samples", "Total blood samples", "Female blood samples", "Male blood samples", "TOTAL"))

sampleOverview_tamRun5 %>%
  kable(caption = "Counts of samples included in tamRun5 by species, sample type, and sex") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

# 4 Quality control

## 4.1 Duplicates

### Run duptest

The IDduplicateSamples function does all pairwise comparisons of samples and outputs two metrics:

-   proportionCommon - The proportion of loci that had genotypes for both samples in a pair
-   proportionMatch - The proportion of loci that have identical genotypes in the sample pair

```{r identify duplicate samples,eval=FALSE}
source("./GTscore_sourceScripts/GTscore_modified.R") # NOTE- added N=ATGC to script for later analyses (not included in original)

#load locus table and 10x allele reads file
fullSet_singleSNP_locusTable <- read.delim("./05_tamRun5/03_run5GTscore/fullSet_LocusTable_singleSNPs.txt", header = TRUE, stringsAsFactors = FALSE)
fullSet_singleSNP_alleleReads_10x <- read.delim("./05_tamRun5/03_run5GTscore/fullSet_AlleleReads_singleSNPs_10x.txt", header = TRUE, row.names = 1, stringsAsFactors = FALSE)

#generate singleSNP genotypes using the polyGen algorithm, adjust "0" formatting
fullSet_polyGenResults_singleSNP_10x <- polyGen(fullSet_singleSNP_locusTable, fullSet_singleSNP_alleleReads_10x)

#convert missing genotypes "0" to NA
polyGenResults_singleSNP_NA <- fullSet_polyGenResults_singleSNP_10x
polyGenResults_singleSNP_NA[polyGenResults_singleSNP_NA=="0"] <- NA

#compare all samples, for each comparison get proportion of loci that have genotypes
#in both samples (proportionCommon) and proportion of shared loci that have identical
#genotypes (proportionMatch).
#Samples with a high proportionCommon and high proportionMatch are likely duplicates
polyGenResults_dupTest <- IDduplicateSamples(polyGenResults_singleSNP_NA)
```

Export results

```{r}
write.table(polyGenResults_dupTest, "./05_tamRun5/04_run5GenoAnalyses/tamRun5_polyGenResults_dupTest.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

Quick read-in

```{r duplicate sample results}
polyGenResults_dupTest <- read.delim("./05_tamRun5/04_run5GenoAnalyses/tamRun5_polyGenResults_dupTest.txt", stringsAsFactors = FALSE)
head(polyGenResults_dupTest)
```

## Likely duplicates

```{r}
sample_shortNames <- md_tamRun5 %>%
  mutate(
    sample_shortName = str_c(animalID, species, sex, sampleType, sep = "_")
  ) %>%
  select(c(sampleID, sample_shortName))

duptest_sampleShortNames <- polyGenResults_dupTest %>%
  merge(., sample_shortNames, by.x = "Sample1", by.y = "sampleID") %>%
  merge(., sample_shortNames, by.x = "Sample2", by.y = "sampleID") %>%
  relocate(Sample1) %>%
  dplyr::rename(
    "sample_md.1" = "sample_shortName.x",
    "sample_md.2" = "sample_shortName.y"
  ) %>%
  arrange(desc(proportionMatch)) %>%
  na.omit()

View(duptest_sampleShortNames %>%
       filter(Sample1 == "tamRun5.067") %>%
       filter(commonGenotypes > 10) %>%
       filter(proportionMatch > 0.9))

duptest_likelyDups <- duptest_sampleShortNames %>%
  filter(commonGenotypes > 10) %>%
  filter(proportionMatch > 0.9)

duptest_likelyDups_troubleshoot <- duptest_likelyDups %>%
  mutate(
    pairType = case_when(
      sample_md.1 == sample_md.2 ~ "sampleDup",
      sub("\\_.*", "", sample_md.1) == sub("\\_.*", "", sample_md.2) ~ "bloodHair",
      .default = "troubleshoot"
    )
  ) %>%
  filter(pairType == "troubleshoot") %>%
  merge(., md_tamRun5[, c("sampleID", "captureDate", "animalName", "sampleName")], by.x = "Sample1", by.y = "sampleID") %>%
  dplyr::rename(
    "captureDate1" = "captureDate",
    "animalName1" = "animalName",
    "sampleName1" = "sampleName"
  ) %>%
  merge(., md_tamRun5[, c("sampleID", "captureDate", "animalName", "sampleName")], by.x = "Sample2", by.y = "sampleID") %>%
  dplyr::rename(
    "captureDate2" = "captureDate",
    "animalName2" = "animalName",
    "sampleName2" = "sampleName"
  ) %>%
  mutate(
    pairID = row_number()
  ) %>%
  select(c(pairID, pairType, Sample1, Sample2, matchedGenotypes, commonGenotypes, proportionMatch, proportionCommon, starts_with("sample_md"), starts_with("captureDate"), starts_with("animalName"), starts_with("sampleName")))

install.packages("xlsx")
library(xlsx)
write.xlsx(duptest_likelyDups_troubleshoot, "./05_tamRun5/04_run5GenoAnalyses/tamRun5_likelyDups_troubleshooting.xlsx", col.names = T, row.names = F)
```

## Compare hair vs. blood

animalID 196

```{r}
sample_shortNames <- md_tamRun5 %>%
  mutate(
    sample_shortName = str_c(animalID, species, sex, sampleType, sep = "_")
  ) %>%
  select(c(sampleID, sample_shortName))

duptest <- polyGenResults_dupTest %>%
  merge(., sample_shortNames, by.x = "Sample1", by.y = "sampleID") %>%
  merge(., sample_shortNames, by.x = "Sample2", by.y = "sampleID") %>%
  relocate(Sample1) %>%
  dplyr::rename(
    "sample_md.1" = "sample_shortName.x",
    "sample_md.2" = "sample_shortName.y"
  ) %>%
  arrange(desc(proportionMatch)) %>%
  na.omit()

dupTest_dupAnimalIDs <- duptest %>%
  mutate(
    sampleName_match = case_when(
      sample_md.1 == sample_md.2 ~ "yes",
      .default = "no"
    )
  ) %>%
  filter(sampleName_match == "yes")

dupTest_bloodHairPairs <- duptest %>%
  separate(sample_md.1, into = c("animalID.1", "sampleType.1"), sep = "_") %>%
  separate(sample_md.2, into = c("animalID.2", "sampleType.2"), sep = "_") %>%
  mutate(
    animalID_match = case_when(
      animalID.1 == animalID.2 ~ "yes",
      .default = "no"
    ),
    sampleType_match = case_when(
      sampleType.1 == sampleType.2 ~ "yes",
      .default = "no"
    )
  ) %>%
  filter(animalID_match == "yes") %>%
  filter(sampleType_match == "no")
```

# 4 Sample genotype success

## 10x coverage

### Overview

```{r, echo=FALSE}
genoRate_summary_tamRun5_10x <- matrix(
  # blood
  c(length(sampleSum_blood_10x$sampleFile),
    sum(sampleSum_blood_10x$GenotypeRate == 0),
    scales::percent(median(sampleSum_blood_10x$GenotypeRate)),
    scales::percent(min(sampleSum_blood_10x$GenotypeRate)),
    scales::percent(max(sampleSum_blood_10x$GenotypeRate)),
    # hair
    length(sampleSum_hair_10x$sampleFile),
    sum(sampleSum_hair_10x$GenotypeRate == 0),
    scales::percent(median(sampleSum_hair_10x$GenotypeRate)),
    scales::percent(min(sampleSum_hair_10x$GenotypeRate)),
    scales::percent(max(sampleSum_hair_10x$GenotypeRate))),
  nrow = 5,
  ncol = 2) %>%
  `colnames<-`(., c("blood", "hair")) %>%
  `rownames<-`(., c("Total samples", "Samples w/0% geno success", "Median geno success", "Min geno success", "Max geno success")) %>%
  as.data.frame()
  
genoRate_summary_tamRun5_10x %>%
  kable(caption = "10x: Overview of genotype success by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### Distribution of geno success

```{r, echo=FALSE}
# Blood
ggplot() + 
  geom_histogram(data = sampleSum_blood_10x, aes(x = GenotypeRate), binwidth = 0.01) +
  #xlim(-0.01,1.01) +
  geom_vline(xintercept = median(sampleSum_blood_10x$GenotypeRate), linetype = "dashed") +
  annotate(x = 0.54, y = 27, label = "Median = 0.65", geom = "label") +
  labs(title="GTscore 10x genotype success: Blood samples (n = 194)", x="Proportion of loci genotyped", y="Number of blood samples") +
  theme_bw() + 
  theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))

# Hair
ggplot() + 
  geom_histogram(data = sampleSum_hair_10x, aes(x = GenotypeRate), binwidth = 0.01) + 
  geom_vline(xintercept = median(sampleSum_hair_10x$GenotypeRate), linetype = "dashed") +
  annotate(x = 0.44, y = 27, label = "Median = 0.37", geom = "label") +
  labs(title="GTscore 10x proportion of loci genotyped: Hair samples (n = 237)", x="Proportion of loci genotyped", y="Number of hair samples") +
  theme_bw() + 
  theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

### By sample collection year

#### Blood samples

Based on Kruskal-Wallis test results combined with pairwise.wilcox.test, there appears to be a statistical difference in genotype rates in blood samples between capture years 2014-2018, 2016-2018, and 2017-2018.

Looking at boxplots comparing genotype rates per sample across capture years, 2018 in general seems lower than the rest - otherwise, genotype rates seem fairly consistent across capture years.

```{r}
# ANOVA model
aovModel_blood_10x <- aov(GenotypeRate ~ as.factor(captureYear), data = sampleSum_blood_10x)

# check normality assumption - FAIL
hist(sampleSum_blood_10x$GenotypeRate) # sort of bell-shaped?
qqnorm(aovModel_blood_10x$residuals)
qqline(aovModel_blood_10x$residuals) # tail suggests not normal
shapiro.test(sampleSum_blood_10x$GenotypeRate) # p-value suggests NOT from a normal dist

# Kruskal-Wallis (non-parametric test)
kruskal.test(GenotypeRate ~ as.factor(captureYear), data = sampleSum_blood_10x) # p-value suggests diff b/t groups

# see which groups here:
pairwise.wilcox.test(sampleSum_blood_10x$GenotypeRate, as.factor(sampleSum_blood_10x$captureYear), p.adjust.method = "BH")
```

```{r, echo=FALSE}
ggplot(sampleSum_blood_10x, aes(x = captureYear, y = GenotypeRate, group = captureYear)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color=factor(captureYear))) +
  geom_text(data=sampleSum_blood_10x %>% 
              group_by(captureYear) %>% 
              tally(),
#              summarise(q3 = quantile(GenotypeRate, 0.75),
#                        q1 = quantile(GenotypeRate, 0.25),
#                        iqr = q3 - q1,
#                        top = min(q3 + 1.5*iqr, max(GenotypeRate)), 
#                        n=n()), 
            aes(x=captureYear, y=0, label= paste0("n = ", n)), nudge_y=1) +
  labs(title = "GTscore 10x blood sample genotype success by sample collection year",
       x = "Sample collection year",
       y = "Proportion of loci genotyped") +
  theme_bw()
```

#### Hair samples

Hair sample genotype rates, however, do differ significantly by the year the sample was collected, with higher genotype success for samples collected more recently. **Note** however, that this trend may be at least partly influenced by the extraction protocols used as well, especially since many of the older samples were extracted in 2015/2016.

tukey: (x = also in kruskal wallis)
2015-2009 x
2015-2010 x
2016-2010 -
2017-2010 x
2014-2011 x
2015-2011 x
2016-2011 x
2017-2011 x
2015-2013 x
2016-2013 x
2017-2013 x
2019-2015 x

```{r}
# ANOVA model
aovModel_hair_10x <- aov(GenotypeRate ~ as.factor(captureYear), data = sampleSum_hair_10x)

test <- TukeyHSD(aovModel_hair_10x)
test2 <- as.data.frame(test$`as.factor(captureYear)`) %>%
  rename_all(~str_replace_all(., "\\s+", "")) %>%
  filter(padj < 0.05)

# check normality assumption - FAIL
hist(sampleSum_hair_10x$GenotypeRate) # def not bell shaped
qqnorm(aovModel_hair_10x$residuals)
qqline(aovModel_hair_10x$residuals) # doesn't look terrible?
shapiro.test(sampleSum_hair_10x$GenotypeRate) # p-value suggests NOT from a normal dist

# Kruskal-Wallis (non-parametric test)
kruskal.test(GenotypeRate ~ as.factor(captureYear), data = sampleSum_hair_10x) # p-value suggests diff b/t groups

# see which groups here:
pairwise.wilcox.test(sampleSum_hair_10x$GenotypeRate, as.factor(sampleSum_hair_10x$captureYear), p.adjust.method = "BH")
```

```{r, echo=FALSE}
ggplot(sampleSum_hair_10x, aes(x = captureYear, y = GenotypeRate, group = captureYear)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color=factor(captureYear))) +
  geom_text(data=sampleSum_hair_10x %>% 
              group_by(captureYear) %>% 
              tally(),
#              summarise(q3 = quantile(GenotypeRate, 0.75),
#                        q1 = quantile(GenotypeRate, 0.25),
#                        iqr = q3 - q1,
#                        top = min(q3 + 1.5*iqr, max(GenotypeRate)), 
#                        n=n()), 
            aes(x=captureYear, y=0, label= paste0("n=", n)), nudge_y=1) +
  labs(title = "GTscore 10x hair sample genotype success by sample collection year",
       x = "Sample collection year",
       y = "Proportion of loci genotyped") +
  theme_bw()
```

```{r, echo=FALSE}
ggplot(sampleSum_hair_10x, aes(x = xtnYear, y = GenotypeRate, group = xtnYear)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color=factor(captureYear))) +
  labs(title = "GTscore 10x hair sample genotype success by extraction year", 
       x = "Extraction year",
       y = "Propotion of loci genotyped") +
  theme_bw()
```

## 30x coverage

### Overview

```{r, echo=FALSE}
genoRate_summary_tamRun5_30x <- matrix(
  # blood
  c(length(sampleSum_blood_30x$sampleFile),
    sum(sampleSum_blood_30x$GenotypeRate == 0),
    scales::percent(median(sampleSum_blood_30x$GenotypeRate)),
    scales::percent(min(sampleSum_blood_30x$GenotypeRate)),
    scales::percent(max(sampleSum_blood_30x$GenotypeRate)),
    # hair
    length(sampleSum_hair_30x$sampleFile),
    sum(sampleSum_hair_30x$GenotypeRate == 0),
    scales::percent(median(sampleSum_hair_30x$GenotypeRate)),
    scales::percent(min(sampleSum_hair_30x$GenotypeRate)),
    scales::percent(max(sampleSum_hair_30x$GenotypeRate))),
  nrow = 5,
  ncol = 2) %>%
  `colnames<-`(., c("blood", "hair")) %>%
  `rownames<-`(., c("Total samples", "Samples w/0% geno success", "Median geno success", "Min geno success", "Max geno success")) %>%
  as.data.frame()
  
genoRate_summary_tamRun5_30x %>%
  kable(caption = "30x: Overview of genotype success by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### Distribution of geno success

```{r, echo=FALSE}
# Blood
ggplot() + 
  geom_histogram(data = sampleSum_blood_30x, aes(x = GenotypeRate), binwidth = 0.01) +
  #xlim(-0.01,1.01) +
  geom_vline(xintercept = median(sampleSum_blood_30x$GenotypeRate), linetype = "dashed") +
  annotate(x = 0.54, y = 27, label = "Median = 0.50", geom = "label") +
  labs(title="GTscore 30x genotype success: Blood samples (n = 194)", x="Proportion of loci genotyped", y="Number of blood samples") +
  theme_bw() + 
  theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))

# Hair
ggplot() + 
  geom_histogram(data = sampleSum_hair_30x, aes(x = GenotypeRate), binwidth = 0.01) + 
  geom_vline(xintercept = median(sampleSum_hair_30x$GenotypeRate), linetype = "dashed") +
  annotate(x = 0.44, y = 27, label = "Median = 0.14", geom = "label") +
  labs(title="GTscore 30x proportion of loci genotyped: Hair samples (n = 237)", x="Proportion of loci genotyped", y="Number of hair samples") +
  theme_bw() + 
  theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

### By sample collection year

**Note** that plots and statistical tests are based on "cat" tamRun3 data only.

#### Blood samples

Blood genotype rate does not appear to differ based on the year the sample was collected.

```{r}
summary.aov(aov(GenotypeRate ~ captureYear, data = sampleSum_blood_30x))
```

```{r, echo=FALSE}
ggplot(sampleSum_blood_30x, aes(x = captureYear, y = GenotypeRate, group = captureYear)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color=factor(captureYear))) +
  geom_text(data=sampleSum_blood_30x %>% 
              group_by(captureYear) %>% 
              tally(),
#              summarise(q3 = quantile(GenotypeRate, 0.75),
#                        q1 = quantile(GenotypeRate, 0.25),
#                        iqr = q3 - q1,
#                        top = min(q3 + 1.5*iqr, max(GenotypeRate)), 
#                        n=n()), 
            aes(x=captureYear, y=0, label= paste0("n = ", n)), nudge_y=1) +
  labs(title = "GTscore 30x blood sample genotype success by sample collection year",
       x = "Sample collection year",
       y = "Proportion of loci genotyped") +
  theme_bw()
```

#### Hair samples

Hair sample genotype rates, however, do differ significantly by the year the sample was collected, with higher genotype success for samples collected more recently. **Note** however, that this trend may be at least partly influenced by the extraction protocols used as well, especially since many of the older samples were extracted in 2015/2016.

```{r}
summary.aov(aov(GenotypeRate ~ captureYear, data = sampleSum_hair_30x))
```

```{r, echo=FALSE}
ggplot(sampleSum_hair_30x, aes(x = captureYear, y = GenotypeRate, group = captureYear)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color=factor(captureYear))) +
  geom_text(data=sampleSum_hair_30x %>% 
              group_by(captureYear) %>% 
              tally(),
#              summarise(q3 = quantile(GenotypeRate, 0.75),
#                        q1 = quantile(GenotypeRate, 0.25),
#                        iqr = q3 - q1,
#                        top = min(q3 + 1.5*iqr, max(GenotypeRate)), 
#                        n=n()), 
            aes(x=captureYear, y=0, label= paste0("n=", n)), nudge_y=1) +
  labs(title = "GTscore 30x hair sample genotype success by sample collection year",
       x = "Sample collection year",
       y = "Proportion of loci genotyped") +
  theme_bw()
```

```{r, echo=FALSE}
ggplot(sampleSum_hair_30x, aes(x = xtnYear, y = GenotypeRate, group = xtnYear)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(aes(color=factor(captureYear))) +
  labs(title = "GTscore 30x hair sample genotype success by extraction year", 
       x = "Extraction year",
       y = "Propotion of loci genotyped") +
  theme_bw()
```

# 5 Geno consistency across runs

Before looking at the specifics of tamRun3, I first want to get a sense of genotype consistency between sequencing runs. Assessments include the following:
-   tamRun3 a vs. b vs. cat datasets
-   tamRun1 vs. tamRun2
-   tamRun3 vs. tamRun1 & 2

## 5.1 tamRun3 a, b, cat

First, I'm comparing genotypes assigned to samples from tamRun3a, b, and cat datasets. Samples in these comparisons are **technical replicates**, where all wet-lab processes are the same for each sample.
-   tamRun3 a vs. b comparisons show whether genotypes differ as a result of differences between sequencing runs
-   tamRun3a/b vs. cat comparisons show whether genotypes differ as a result of increased reads

### 5.1.1 Comparison dataframe

**1) Create long dataframes**

```{r}
# tamRun3 a - long format
genos_tamRun3al <- genos_tamRun3 %>%
  select(., !contains(c("tamRun3b", "cat_"))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "geno_3a") # %>%
#  filter(geno_3a != "0")

# tamRun3 b - long format
genos_tamRun3bl <- genos_tamRun3 %>%
  select(., contains("tamRun3b")) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = str_replace(sampleID, "tamRun3b", "tamRun3")) %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "geno_3b") #%>%
#  filter(geno_3b != "0")

# tamRun3 cat - long format
genos_tamRun3catl <- genos_tamRun3 %>%
  select(., contains("cat_")) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = str_replace(sampleID, "cat_tamRun3", "tamRun3")) %>%
  pivot_longer(!sampleID, names_to = "locus", values_to = "geno_3cat") #%>%
#  filter(geno_3cat != "0")
```

**2) Comparison dataframe**

```{r}
tamRun3_genos_abcat <- merge(genos_tamRun3al, genos_tamRun3bl, by = c("locus", "sampleID")) %>%
  merge(., genos_tamRun3catl, by = c("locus", "sampleID")) %>%
  mutate(totalCalled = rowSums(select(., -sampleID, -locus) != 0, na.rm = T)) %>%
  mutate(zerosCalled = rowSums(select(., -sampleID, -locus, -totalCalled) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos = ifelse(zerosCalled > 0,
                              (n_distinct(c_across(geno_3a:geno_3cat), na.rm = T)) - 1,
                              (n_distinct(c_across(geno_3a:geno_3cat), na.rm = T))
                              ))
```

### 5.1.2 General comparison

```{r, echo=FALSE}
# tamRun3 a vs. b
genoMatch_tamRun3ab <- tamRun3_genos_abcat %>%
  select(c(locus, sampleID, geno_3a, geno_3b)) %>%
  filter(geno_3a != 0, geno_3b != 0) %>%
  mutate(match_3ab = geno_3a == geno_3b) %>%
  merge(., md_tamRun3[,c("sampleID", "sampleType")], by = "sampleID", all.x = T, all.y = F) %>%
#  select(-sampleType) %>%
  arrange(sampleID, locus)

# tamRun3a vs. cat
genoMatch_tamRun3acat <- tamRun3_genos_abcat %>%
  select(c(locus, sampleID, geno_3a, geno_3cat)) %>%
  filter(geno_3a != 0, geno_3cat != 0) %>%
  mutate(match_3acat = geno_3a == geno_3cat) %>%
  merge(., md_tamRun3[,c("sampleID", "sampleType")], by = "sampleID", all.x = T, all.y = F) %>%
  select(-sampleType) %>%
  arrange(sampleID, locus)

# tamRun3b vs. cat
genoMatch_tamRun3bcat <- tamRun3_genos_abcat %>%
  select(c(locus, sampleID, geno_3b, geno_3cat)) %>%
  filter(geno_3b != 0, geno_3cat != 0) %>%
  mutate(match_3bcat = geno_3b == geno_3cat) %>%
  merge(., md_tamRun3[,c("sampleID", "sampleType")], by = "sampleID", all.x = T, all.y = F) %>%
#  select(-sampleType) %>%
  arrange(sampleID, locus)
```

#### Summary table

```{r, echo=FALSE}
genoMatch_tamRun3abcat_summary <- matrix(c(
  # Total genotypes in each dataset
  length(genos_tamRun3al$sampleID[genos_tamRun3al$geno_3a != 0]),
  length(genos_tamRun3al$sampleID[genos_tamRun3al$geno_3a != 0]),
  NA,
  length(genos_tamRun3bl$sampleID[genos_tamRun3bl$geno_3b != 0]),
  NA,
  length(genos_tamRun3bl$sampleID[genos_tamRun3bl$geno_3b != 0]),
  NA,
  length(genos_tamRun3catl$sampleID[genos_tamRun3catl$geno_3cat != 0]),
  length(genos_tamRun3catl$sampleID[genos_tamRun3catl$geno_3cat != 0]),
  # Non-common genos
  length(genos_tamRun3al$sampleID[genos_tamRun3al$geno_3a != 0]) - length(genos_tamRun3bl$sampleID[genos_tamRun3bl$geno_3b != 0]),
  length(genos_tamRun3catl$sampleID[genos_tamRun3catl$geno_3cat != 0]) - length(genos_tamRun3al$sampleID[genos_tamRun3al$geno_3a != 0]),
  length(genos_tamRun3catl$sampleID[genos_tamRun3catl$geno_3cat != 0]) - length(genos_tamRun3bl$sampleID[genos_tamRun3bl$geno_3b != 0]),
  # Common genos
  length(genoMatch_tamRun3ab$sampleID),
  length(genoMatch_tamRun3acat$sampleID),
  length(genoMatch_tamRun3bcat$sampleID),
  length(genoMatch_tamRun3ab$sampleID[genoMatch_tamRun3ab$match_3ab == FALSE]),
  length(genoMatch_tamRun3acat$sampleID[genoMatch_tamRun3acat$match_3acat == FALSE]),
  length(genoMatch_tamRun3bcat$sampleID[genoMatch_tamRun3bcat$match_3bcat == FALSE])),
  nrow = 3,
  ncol = 6) %>%
  `rownames<-`(., c("tamRun3a vs. b", "tamRun3a vs. cat", "tamRun3b vs. cat")) %>%
  `colnames<-`(., c("totalGenos_3a", "totalGenos_3b", "totalGenos_3cat", "nonCommonGenos", "commonGenos", "diffGenos")) %>%
  as.data.frame() %>%
  mutate(matchGenos = commonGenos - diffGenos) %>%
  relocate(matchGenos, .after = commonGenos) %>%
  mutate(propMatch = matchGenos/commonGenos) %>%
  relocate(propMatch, .after = matchGenos) %>%
  mutate(propDiff = diffGenos/commonGenos)

genoMatch_tamRun3abcat_summary %>%
  kable(caption = "Overview of geno differences between tamRun3 a, b, and cat datasets") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

**Note** when comparing genotype consistency between tamRun3a and tamRun3b vs. consistency between tamRun3a/b and tamRun3_cat, keep in mind the following:

-   90 genotypes that occur in both tamRun3a and tamRun3b are NOT present when the two runs are combined
-   2780 genotypes that occur in both tamRun3a and tamRun3cat are NOT present in tamRun3b
-   1415 genotypes that occur in both tamRun3b and tamRun3cat are NOT present in tamRun3a

```{r}
tamRun3_genos_abcat %>%
  filter(geno_3a !=0, geno_3b !=0, geno_3cat == 0) %>%
  nrow()

tamRun3_genos_abcat %>%
  filter(geno_3a !=0, geno_3b == 0, geno_3cat != 0) %>%
  nrow()

tamRun3_genos_abcat %>%
  filter(geno_3a == 0, geno_3b != 0, geno_3cat != 0) %>%
  nrow()
```

### 5.1.3 Sample consistency

```{r, echo=FALSE}
# tamRun3a vs. b
genoMatch_bySample_tamRun3ab <- genoMatch_tamRun3ab %>%
  group_by(sampleID) %>%
  summarise(lociGenotyped_3ab = length(locus),
            lociDiff_3ab = sum(match_3ab == "FALSE")) %>%
  as.data.frame() %>%
  mutate(propDiff_3ab = lociDiff_3ab/lociGenotyped_3ab) %>%
  merge(., md_tamRun3[,c("sampleID", "sampleType")], by = "sampleID", all.x = T, all.y = F)

# tamRun3a vs. cat
genoMatch_bySample_tamRun3acat <- genoMatch_tamRun3acat %>%
  group_by(sampleID) %>%
  summarise(lociGenotyped_3acat = length(locus),
            lociDiff_3acat = sum(match_3acat == "FALSE")) %>%
  as.data.frame() %>%
  mutate(propDiff_3acat = lociDiff_3acat/lociGenotyped_3acat) %>%
  merge(., md_tamRun3[,c("sampleID", "sampleType")], by = "sampleID", all.x = T, all.y = F)

# tamRun3b vs. cat
genoMatch_bySample_tamRun3bcat <- genoMatch_tamRun3bcat %>%
  group_by(sampleID) %>%
  summarise(lociGenotyped_3bcat = length(locus),
            lociDiff_3bcat = sum(match_3bcat == "FALSE")) %>%
  as.data.frame() %>%
  mutate(propDiff_3bcat = lociDiff_3bcat/lociGenotyped_3bcat) %>%
  merge(., md_tamRun3[,c("sampleID", "sampleType")], by = "sampleID", all.x = T, all.y = F)
```

#### Summary table

```{r, echo=FALSE}
genoMatch_bySample_summary_tamRun3abcat <- matrix(c(
  # a vs. b
  length(genoMatch_bySample_tamRun3ab$sampleID),
  genoMatch_bySample_tamRun3ab %>%
    filter(lociDiff_3ab == 0) %>%
    nrow(),
  genoMatch_bySample_tamRun3ab %>%
    filter(lociDiff_3ab > 0) %>%
    nrow(),
  max(genoMatch_bySample_tamRun3ab$lociDiff_3ab),
  "tamRun3.428 (9%)",
  max(genoMatch_bySample_tamRun3ab$propDiff_3ab),
  "tamRun3.098 (1/8)",
  # a vs. cat
  length(genoMatch_bySample_tamRun3acat$sampleID),
  genoMatch_bySample_tamRun3acat %>%
    filter(lociDiff_3acat == 0) %>%
    nrow(),
  genoMatch_bySample_tamRun3acat %>%
    filter(lociDiff_3acat > 0) %>%
    nrow(),
  max(genoMatch_bySample_tamRun3acat$lociDiff_3acat),
  "tamRun3.289, 353, & 428 (all 4%)",
  max(genoMatch_bySample_tamRun3acat$propDiff_3acat),
  "tamRun3.100 (3/53)",
  # b vs. cat
  length(genoMatch_bySample_tamRun3bcat$sampleID),
  genoMatch_bySample_tamRun3bcat %>%
    filter(lociDiff_3bcat == 0) %>%
    nrow(),
  genoMatch_bySample_tamRun3bcat %>%
    filter(lociDiff_3bcat > 0) %>%
    nrow(),
  max(genoMatch_bySample_tamRun3bcat$lociDiff_3bcat),
  "tamRun3.401 (4%)",
  max(genoMatch_bySample_tamRun3bcat$propDiff_3bcat),
  "tamRun3.098 (1/9)"),
  nrow = 7,
  ncol = 3, 
  byrow = F) %>%
  `colnames<-`(., c("a vs. b", "a vs. cat", "b vs. cat")) %>%
  `rownames<-`(., c("Total samples w/genos in both datasets",
                    "Samples w/100% matched genos across loci",
                    "Samples w/mismatched genos across loci",
                    "Max # mismatched genos per sample",
                    "Sample w/max # mismatches",
                    "Max prop. mismatched genos per sample", 
                    "Sample w/max prop. mismatches"))

genoMatch_bySample_summary_tamRun3abcat %>%
  kable(caption = "Overview of sample genotype consistency in tamRun3abcat") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### 5.1.4 Locus consistency

#### All samples combined

```{r, echo=FALSE}
# tamRun3a vs. b
genoMatch_byLocus_tamRun3ab <- genoMatch_tamRun3ab %>%
  group_by(locus) %>%
  summarise(totalGenos = length(sampleID),
            sumMatch = length(sampleID[match_3ab == "TRUE"]),
            sumDiff = length(sampleID[match_3ab == "FALSE"])) %>%
  as.data.frame() %>%
  mutate(propDiff = sumDiff/totalGenos) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  arrange(desc(propDiff))

# tamRun3a vs. cat
genoMatch_byLocus_tamRun3acat <- genoMatch_tamRun3acat %>%
  group_by(locus) %>%
  summarise(totalGenos = length(sampleID),
            sumMatch = length(sampleID[match_3acat == "TRUE"]),
            sumDiff = length(sampleID[match_3acat == "FALSE"])) %>%
  as.data.frame() %>%
  mutate(propDiff = sumDiff/totalGenos) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  arrange(desc(propDiff))

# tamRun3b vs. cat
genoMatch_byLocus_tamRun3bcat <- genoMatch_tamRun3bcat %>%
  group_by(locus) %>%
  summarise(totalGenos = length(sampleID),
            sumMatch = length(sampleID[match_3bcat == "TRUE"]),
            sumDiff = length(sampleID[match_3bcat == "FALSE"])) %>%
  as.data.frame() %>%
  mutate(propDiff = sumDiff/totalGenos) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  arrange(desc(propDiff))
```

```{r, eval=FALSE}
write.csv(genoMatch_byLocus_tamRun3ab, "./03_tamRun3/04_genoAnalyses/genoMatch_byLocus_tamRun3ab.csv", row.names = F)
write.csv(genoMatch_byLocus_tamRun3acat, "./03_tamRun3/04_genoAnalyses/genoMatch_byLocus_tamRun3acat.csv", row.names = F)
write.csv(genoMatch_byLocus_tamRun3bcat, "./03_tamRun3/04_genoAnalyses/genoMatch_byLocus_tamRun3bcat.csv", row.names = F)
```

##### Summary table

```{r, echo=FALSE}
genoMatch_byLocus_summary_tamRun3abcat <- matrix(c(
  # tamRun3a vs. b
  length(genoMatch_byLocus_tamRun3ab$locus),
  genoMatch_byLocus_tamRun3ab %>%
    filter(sumDiff == 0) %>%
    nrow(),
  genoMatch_byLocus_tamRun3ab %>%
    filter(sumDiff > 0) %>%
    nrow(),
  max(genoMatch_byLocus_tamRun3ab$sumDiff),
  "INDID_49 (5%)",
  max(genoMatch_byLocus_tamRun3ab$propDiff),
  "SEXID_215 (6/25)",
  # tamRun3a vs. cat
  length(genoMatch_byLocus_tamRun3acat$locus),
  genoMatch_byLocus_tamRun3acat %>%
    filter(sumDiff == 0) %>%
    nrow(),
  genoMatch_byLocus_tamRun3acat %>%
    filter(sumDiff > 0) %>%
    nrow(),
  max(genoMatch_byLocus_tamRun3acat$sumDiff),
  "INDID_49 (3%), SIMP_306 (2%)",
  max(genoMatch_byLocus_tamRun3acat$propDiff),
  "INDID_146 (2/13)",
  # tamRun3b vs. cat
  length(genoMatch_byLocus_tamRun3bcat$locus),
  genoMatch_byLocus_tamRun3bcat %>%
    filter(sumDiff == 0) %>%
    nrow(),
  genoMatch_byLocus_tamRun3bcat %>%
    filter(sumDiff > 0) %>%
    nrow(),
  max(genoMatch_byLocus_tamRun3bcat$sumDiff),
  "INDID_396 (3%)",
  max(genoMatch_byLocus_tamRun3bcat$propDiff),
  "LWED_270 (1/8)"),
  nrow = 7,
  ncol = 3) %>%
  `rownames<-`(., c("Total loci w/genos in both runs",
                    "Loci w/100% matched genos across samples",
                    "Loci w/mismatched genos across samples",
                    "Max # mismatched genos per locus",
                    "Locus w/max # mismatches",
                    "Max prop. mismatched genos per locus",
                    "Locus w/max prop. mismatches")) %>%
  `colnames<-`(., c("a vs. b","a vs. cat", "b vs. cat"))

genoMatch_byLocus_summary_tamRun3abcat %>%
  kable(caption = "Overview of locus genotype consistency in tamRun3a vs. b") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

##### Plots

**tamRun3a vs. b**

```{r, echo=FALSE}
# Reformat for plotting
genoMatch_byLocus_tamRun3ab_totalGenos <- genoMatch_byLocus_tamRun3ab %>%
  select(c(locus, totalGenos, set)) %>%
  dplyr::rename("count" = "totalGenos") %>%
  mutate(type = "totalGenos")

genoMatch_byLocus_tamRun3ab_sumDiff <- genoMatch_byLocus_tamRun3ab %>%
  select(c(locus, sumDiff, set)) %>%
  dplyr::rename("count" = "sumDiff") %>%
  mutate(type = "sumDiff")

genoMatch_byLocus_tamRun3ab_totalvdiff <- rbind(genoMatch_byLocus_tamRun3ab_totalGenos, genoMatch_byLocus_tamRun3ab_sumDiff) %>%
  merge(., genoMatch_byLocus_tamRun3ab[, c("locus", "propDiff")], by = "locus") %>%
  arrange(desc(type), desc(propDiff)) %>%
  mutate(row_id = row_number())

# Scatterplot for all loci
ggplot(genoMatch_byLocus_tamRun3ab, aes(x = totalGenos, y = propDiff, color = set)) +
  geom_point() +
  theme_bw()

# Barplots
## INDID - general
genoMatch_byLocus_tamRun3ab_totalvdiff %>%
  filter(set %in% "INDID") %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  labs(x = "Number of samples genotyped",
       y = "Locus ID",
       title = "INDID genotype consistency across samples in tamRun3a vs. b", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

## INDID - species-specific
genoMatch_byLocus_tamRun3ab_totalvdiff %>%
  filter(set %in% c("LWED", "SIMP")) %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Number of samples genotyped",
       y = "Locus ID",
       title = "Species-specific INDID genotype consistency in tamRun3a vs. b", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

## SEXID - general & species-specific
genoMatch_byLocus_tamRun3ab_totalvdiff %>%
  filter(set %in% c("SEXID", "SEXID_LWED", "SEXID_SIMP")) %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Number of samples genotyped",
       y = "Locus ID",
       title = "SEXID genotype consistency in tamRun3a vs. b", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

## SPECIES
genoMatch_byLocus_tamRun3ab_totalvdiff %>%
  filter(set == "SPECIES") %>%
  ggplot(aes(x = count, y = locus, fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Number of samples genotyped",
       y = "Locus ID",
       title = "SPECIES genotype consistency in tamRun3a vs. b", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()
```

**tamRun3a vs. cat**

```{r, echo=FALSE}
# Reformat for plotting
genoMatch_byLocus_tamRun3acat_totalGenos <- genoMatch_byLocus_tamRun3acat %>%
  select(c(locus, totalGenos, set)) %>%
  dplyr::rename("count" = "totalGenos") %>%
  mutate(type = "totalGenos")

genoMatch_byLocus_tamRun3acat_sumDiff <- genoMatch_byLocus_tamRun3acat %>%
  select(c(locus, sumDiff, set)) %>%
  dplyr::rename("count" = "sumDiff") %>%
  mutate(type = "sumDiff")

genoMatch_byLocus_tamRun3acat_totalvdiff <- rbind(genoMatch_byLocus_tamRun3acat_totalGenos, genoMatch_byLocus_tamRun3acat_sumDiff) %>%
  merge(., genoMatch_byLocus_tamRun3acat[, c("locus", "propDiff")], by = "locus") %>%
  arrange(desc(type), desc(propDiff)) %>%
  mutate(row_id = row_number())

# Scatterplot for all loci
ggplot(genoMatch_byLocus_tamRun3acat, aes(x = totalGenos, y = propDiff, color = set)) +
  geom_point() +
  theme_bw()

# Barplots
## INDID - general
genoMatch_byLocus_tamRun3acat_totalvdiff %>%
  filter(set %in% "INDID") %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different INDID genotypes between tamRun3a vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

# INDID - species-specific
genoMatch_byLocus_tamRun3acat_totalvdiff %>%
  filter(set %in% c("LWED", "SIMP")) %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different species-specific INDID genotypes between tamRun3a vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

# SEXID - general & species-specific
genoMatch_byLocus_tamRun3acat_totalvdiff %>%
  filter(set %in% c("SEXID", "SEXID_LWED", "SEXID_SIMP")) %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different SEXID genotypes between tamRun3a vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

# SPECIES
genoMatch_byLocus_tamRun3acat_totalvdiff %>%
  filter(set == "SPECIES") %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different SPECIES genotypes between tamRun3a vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()
```

**tamRun3b vs. cat**

```{r}
# Reformat for plotting
genoMatch_byLocus_tamRun3bcat_totalGenos <- genoMatch_byLocus_tamRun3bcat %>%
  select(c(locus, totalGenos, set)) %>%
  dplyr::rename("count" = "totalGenos") %>%
  mutate(type = "totalGenos")

genoMatch_byLocus_tamRun3bcat_sumDiff <- genoMatch_byLocus_tamRun3bcat %>%
  select(c(locus, sumDiff, set)) %>%
  dplyr::rename("count" = "sumDiff") %>%
  mutate(type = "sumDiff")

genoMatch_byLocus_tamRun3bcat_totalvdiff <- rbind(genoMatch_byLocus_tamRun3bcat_totalGenos, genoMatch_byLocus_tamRun3bcat_sumDiff) %>%
  merge(., genoMatch_byLocus_tamRun3bcat[, c("locus", "propDiff")], by = "locus") %>%
  arrange(desc(type), desc(propDiff)) %>%
  mutate(row_id = row_number())

# Scatterplot for all loci
ggplot(genoMatch_byLocus_tamRun3bcat, aes(x = totalGenos, y = propDiff, color = set)) +
  geom_point() +
  theme_bw()

# Barplots
## INDID - general
genoMatch_byLocus_tamRun3bcat_totalvdiff %>%
  filter(set %in% "INDID") %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different INDID genotypes between tamRun3b vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

# INDID - species-specific
genoMatch_byLocus_tamRun3bcat_totalvdiff %>%
  filter(set %in% c("LWED", "SIMP")) %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different species-specific INDID genotypes between tamRun3a vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

# SEXID - general & species-specific
genoMatch_byLocus_tamRun3bcat_totalvdiff %>%
  filter(set %in% c("SEXID", "SEXID_LWED", "SEXID_SIMP")) %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different SEXID genotypes between tamRun3b vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()

# SPECIES
genoMatch_byLocus_tamRun3bcat_totalvdiff %>%
  filter(set == "SPECIES") %>%
  ggplot(aes(x = count, y = forcats::fct_reorder(locus, row_id), fill = forcats::fct_reorder(type, row_id))) +
  geom_bar(position = "stack", stat = "identity") +
  facet_grid(rows = vars(set), scales = "free") +
  labs(x = "Proportion of samples genotyped",
       y = "Locus ID",
       title = "Proportion of samples with different SPECIES genotypes between tamRun3a vs. cat", 
       fill = "Key") +
  scale_fill_discrete(labels = c("genosMatch", "genosMismatch")) +
  theme_bw()
```

#### By sample type

```{r, echo=FALSE}
# sample lists
sampleList_blood <- sampleSum_blood_a %>%
  select(sampleID)
sampleList_hair <- sampleSum_hair_a %>%
  select(sampleID)

# tamRun3a vs. b - blood
genoMatch_byLocus_tamRun3ab_blood <- genoMatch_tamRun3ab %>%
  filter(sampleID %in% sampleList_blood$sampleID) %>%
  group_by(locus) %>%
  summarise(totalGenos = length(sampleID),
            sumMatch = length(sampleID[match_3ab == "TRUE"]),
            sumDiff = length(sampleID[match_3ab == "FALSE"])) %>%
  as.data.frame() %>%
  mutate(propDiff = sumDiff/totalGenos) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  arrange(desc(propDiff))

# tamRun3a vs. b - hair
genoMatch_byLocus_tamRun3ab_hair <- genoMatch_tamRun3ab %>%
  filter(sampleID %in% sampleList_hair$sampleID) %>%
  group_by(locus) %>%
  summarise(totalGenos = length(sampleID),
            sumMatch = length(sampleID[match_3ab == "TRUE"]),
            sumDiff = length(sampleID[match_3ab == "FALSE"])) %>%
  as.data.frame() %>%
  mutate(propDiff = sumDiff/totalGenos) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  arrange(desc(propDiff))
```

```{r, eval=FALSE}
write.csv(genoMatch_byLocus_tamRun3ab_blood, "./03_tamRun3/04_genoAnalyses/genoMatch_byLocus_tamRun3ab_blood.csv", row.names = F)
write.csv(genoMatch_byLocus_tamRun3ab_hair, "./03_tamRun3/04_genoAnalyses/genoMatch_byLocus_tamRun3ab_hair.csv", row.names = F)
```

## 5.2 tamRun1 vs. 2

Next I'll be comparing genotypes assigned to samples from tamRun1 & 2. Samples in these comparisons include different kinds of **technical replicates**, including replicates both within and between sequencing runs.

### 5.2.1 Comparison dataframe

**1) Create general names for each run based on animalID and sampleType for merging**

```{r}
# tamRun1
nameType_tamRun1 <- md_tamRun1 %>%
  mutate(sampleID = gsub("_", "\\.", sampleID)) %>%
  select(c(sampleID, animalID, sampleType)) %>%
  distinct() %>%
  na.omit() %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  unite("sample", sampleID:sample, sep = "--", remove = F)

genos_tamRun1t <- genos_tamRun1 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "run1") %>%
  mutate(run1 = sub('_[^.]+', '', run1)) %>%
  merge(., nameType_tamRun1, by.x = "run1", by.y = "sampleID") %>%
  relocate(sample, .after = run1) %>%
  rename_all(~sub("\\..*", "", .x)) %>%
  select(!c(run1, animalID, sampleType))

# tamRun2
nameType_tamRun2 <- md_tamRun2 %>%
  mutate(sampleID = gsub("_", "\\.", sampleID)) %>%
  filter(included_inRun == "yes") %>%
  select(c(sampleID, animalID, sampleType)) %>%
  distinct() %>%
  na.omit() %>%
  mutate(sampleType = gsub("hair \\(chelex\\)", "hair", sampleType)) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  unite("sample", sampleID:sample, sep = "--", remove = F)

genos_tamRun2t <- genos_tamRun2 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "run2") %>%
  mutate(run2 = sub('_[^.]+', '', run2)) %>%
  merge(., nameType_tamRun2, by.x = "run2", by.y = "sampleID") %>%
  relocate(sample, .after = run2) %>%
  rename_all(~sub("\\..*", "", .x)) %>%
  select(!c(run2, animalID, sampleType))

# tamRun3
nameType_tamRun3_cat <- md_tamRun3_cat %>%
  filter(!sampleType %in% c("XTN (-)", "PCR1 (-)")) %>%
  select(c(sampleID, animalID, sampleType)) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  unite("sample", sampleID:sample, sep = "--", remove = F)

genos_tamRun3_catt <- genos_tamRun3 %>%
  select(starts_with("cat")) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "run3") %>%
#  mutate(run3 = sub('_[^.]+', '', run3)) %>%
  merge(., nameType_tamRun3_cat, by.x = "run3", by.y = "sampleID") %>%
  relocate(sample, .after = run3) %>%
  #rename_all(~sub("[_][^_]+$", "", .x)) %>%
  select(!c(run3, animalID, sampleType))
```

**2) Create long dataframes**

```{r}
genos_tamRun1l <- genos_tamRun1t %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("run1Geno" = "genotype") %>%
  separate(sample, into = c("run1", "sample"), sep = "--") %>%
  select(!run1) %>%
  group_by(sample, locus) %>%
  mutate(run1GenoID = paste0("run1Geno", row_number())) %>%
  spread(run1GenoID, run1Geno)

genos_tamRun2l <- genos_tamRun2t %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("run2Geno" = "genotype") %>%
  separate(sample, into = c("run2", "sample"), sep = "--") %>%
  select(!run2) %>%
  group_by(sample, locus) %>%
  mutate(run2GenoID = paste0("run2Geno", row_number())) %>%
  spread(run2GenoID, run2Geno)

genos_tamRun3_catl <- genos_tamRun3_catt %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("run3Geno" = "genotype") %>%
  separate(sample, into = c("run3", "sample"), sep = "--") %>%
  select(!run3) %>%
  group_by(sample, locus) %>%
  mutate(run3GenoID = paste0("run3Geno", row_number())) %>%
  spread(run3GenoID, run3Geno)
```

**3) Comparison dataframe**

```{r}
genos_tamRun125 <- merge(genos_tamRun1l, genos_tamRun2l, by = c("sample", "locus"), all = T) %>%
  merge(., genos_tamRun3_catl, by = c("sample", "locus"), all = T) %>%
  filter(!str_detect(sample, "fecal")) %>%
  mutate(genosCalled_125 = 
           rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  mutate(genosCalled_12 = 
           rowSums(select(., -sample, -locus, -run3Geno1, -genosCalled_125) != 0, na.rm = T)) %>%
  mutate(zerosCalled_125 = 
           rowSums(select(., -sample, -locus, -genosCalled_125, -genosCalled_12) == 0, na.rm = T)) %>%
  mutate(zerosCalled_12 = 
           rowSums(select(., -sample, -locus, -run3Geno1, -genosCalled_125, -genosCalled_12, -zerosCalled_125) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos_125 = ifelse(zerosCalled_125 > 0,
                              (n_distinct(c_across(run1Geno1:run3Geno1), na.rm = T)) - 1,
                              (n_distinct(c_across(run1Geno1:run3Geno1), na.rm = T))
                              )) %>%
  rowwise() %>%
  mutate(uniqueGenos_12 = ifelse(zerosCalled_12 > 0,
                              (n_distinct(c_across(run1Geno1:run2Geno8), na.rm = T)) - 1,
                              (n_distinct(c_across(run1Geno1:run2Geno8), na.rm = T))
                              )) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T)

genos_tamRun12 <- genos_tamRun125 %>%
  filter(genosCalled_12 > 1) %>%
  select(-run3Geno1, -genosCalled_125, -zerosCalled_125, -uniqueGenos_125) %>%
  mutate(genosCalled_1 = 
           rowSums(select(., starts_with("run1Geno")) != 0, na.rm = T)) %>%
  mutate(genosCalled_2 = 
           rowSums(select(., starts_with("run2Geno")) != 0, na.rm = T))
```

### 5.3.2 Within-run comparison

```{r}
genoMatch_tamRun1dups <- genos_tamRun12 %>%
  select(!starts_with("run2Geno")) %>%
  select(-genosCalled_12, -zerosCalled_12, -uniqueGenos_12, -genosCalled_2) %>%
  filter(genosCalled_1 > 1) %>%
  mutate(zerosCalled_1 = 
           rowSums(select(., starts_with("run1Geno")) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos_1 = ifelse(zerosCalled_1 > 0,
                              (n_distinct(c_across(run1Geno1:run1Geno3), na.rm = T)) - 1,
                              (n_distinct(c_across(run1Geno1:run1Geno3), na.rm = T))
                              ))

genoMatch_tamRun2dups <- genos_tamRun12 %>%
  select(!starts_with("run1Geno")) %>%
  select(-genosCalled_12, -zerosCalled_12, -uniqueGenos_12, -genosCalled_1) %>%
  filter(genosCalled_2 > 1) %>%
  mutate(zerosCalled_2 = 
           rowSums(select(., starts_with("run2Geno")) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos_2 = ifelse(zerosCalled_2 > 0,
                              (n_distinct(c_across(run2Geno1:run2Geno8), na.rm = T)) - 1,
                              (n_distinct(c_across(run2Geno1:run2Geno8), na.rm = T))
                              ))
```

#### Sample consistency

```{r}
genoMatch_bySample_tamRun1dups <- genoMatch_tamRun1dups %>%
  group_by(sample) %>%
  summarise(lociGenotyped_1 = length(locus),
            lociDiff_1 = length(locus[uniqueGenos_1 > 1])) %>%
  as.data.frame() %>%
  mutate(propDiff_1 = lociDiff_1/lociGenotyped_1)

genoMatch_bySample_tamRun2dups <- genoMatch_tamRun2dups %>%
  group_by(sample) %>%
  summarise(lociGenotyped_2 = length(locus),
            lociDiff_2 = length(locus[uniqueGenos_2 > 1])) %>%
  as.data.frame() %>%
  mutate(propDiff_2 = lociDiff_2/lociGenotyped_2)
```


#### Locus consistency


Between tamRun1 & 2, 23 samples had loci with more than one genotype between runs (including if the sample was duplicated within a run). Of these, 15 had at least one locus with different genotypes, with at most 27 loci with differing genotypes (19.6% of loci genotyped for that sample).

The number of loci with different calls between datasets does not appear to be related to the number of on-target reads (from "cat" data) per sample.

```{r}
# Summary stats per sample tamRun1 vs. tamRun2
uniqueGenos_perSample_summary_tamRun12 <- genos_tamRun12 %>%
  group_by(sample) %>%
  summarise(lociGenotyped = length(locus),
            lociDiff = length(locus[uniqueGenos_12 > 1]),
            min_uniqueGenos_12 = min(uniqueGenos_12),
            max_uniqueGenos_12 = max(uniqueGenos_12),
            mean_uniqueGenos_12 = mean(uniqueGenos_12),
            var_uniqueGenos_12 = var(uniqueGenos_12)) %>%
  as.data.frame() %>%
  mutate(propDiff = (mean_uniqueGenos_12 - 1)) %>%
  relocate(propDiff, .after = mean_uniqueGenos_12) %>%
  mutate(sampleType = gsub("^.*_","", sample))

uniqueGenos_perSample_byLocusSet_tamRun12 <- genos_tamRun12 %>%
  filter(uniqueGenos_12 > 1) %>%
  group_by(set, sample) %>%
  summarise(countUnique = length(sample)) %>%
  as.data.frame() %>%
  group_by(set) %>%
  summarise(countSamples = length(sample)) %>%
  na.omit() %>%
  column_to_rownames("set")
```

**Variation in number of unique genotypes called per locus across samples**

```{r}
summary(genos_tamRun12$uniqueGenos_12[genos_tamRun125$genosCalled_12 > 1])
```

**Overview of genotype differences PER SAMPLE between tamRun1 & 2**

```{r, echo=FALSE}
uniqueGenos_perSample_summaryTable_tamRun12 <- matrix(c(length(uniqueGenos_perSample_summary_tamRun12$sample),
         uniqueGenos_perSample_summary_tamRun12 %>%
           filter(lociDiff > 1) %>%
         nrow(),
         mean(uniqueGenos_perSample_summary_tamRun12$lociDiff),
         mean(uniqueGenos_perSample_summary_tamRun12$propDiff),
         max(uniqueGenos_perSample_summary_tamRun12$lociDiff),
         max(uniqueGenos_perSample_summary_tamRun12$propDiff)),
         nrow = 6,
         ncol = 1) %>%
  `rownames<-`(., c("Total samples", "Samples w/multiple unique genos", "Avg # of loci w/unique genos per sample", "Avg prop. loci w/unique genos per sample", "Max # loci w/unique genos per sample", "Max prop. loci w/unique genos per sample")) %>%
  as.data.frame() %>%
  mutate(V1 = signif(V1, digits = 3)) %>%
  mutate(V1 = as.character(V1))

names(uniqueGenos_perSample_summaryTable_tamRun12) <- NULL

uniqueGenos_perSample_summaryTable_tamRun12 %>%
  kable(caption = "Overview of genos called across SAMPLES between tamRun1 & 2") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

#### Per locus

Across tamRun1 & 2, 135 out of 221 loci had samples genotyped in more than one dataset. Of these, ALL loci had at least one sample with different genotypes in either a, b, or cat datasets, with at most 11 samples with differing genotypes.

The number of samples with different calls between datasets does not appear to be related to the number of on-target reads (from "cat" data) per sample, though most loci with differing genotypes appear to be from the INDID set.

```{r, echo=FALSE}
# Summary stats per locus
uniqueGenos_perLocus_summary_tamRun12 <- genos_tamRun12 %>%
  group_by(locus) %>%
  summarise(samplesGenotyped = length(sample),
            countSamples_lociDiff_12 = length(sample[uniqueGenos_12 > 1]),
            min_uniqueGenos_12 = min(uniqueGenos_12),
            max_uniqueGenos_12 = max(uniqueGenos_12),
            mean_uniqueGenos_12 = mean(uniqueGenos_12),
            var_uniqueGenos_12 = var(uniqueGenos_12)) %>%
  as.data.frame() %>%
  mutate(propDiff_12 = (mean_uniqueGenos_12 - 1)) %>%
  relocate(propDiff_12, .after = mean_uniqueGenos_12) %>%
  filter(locus %in% locusSum$Locus) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  group_by(set) %>%
  arrange(desc(countSamples_lociDiff_12), .by_group = T)

# Table
uniqueGenos_perLocus_summary_tamRun12_table <- matrix(c(length(uniqueGenos_perLocus_summary_tamRun12$locus),
         uniqueGenos_perLocus_summary_tamRun12 %>%
           filter(countSamples_lociDiff_12 > 1) %>%
         nrow(),
         max(uniqueGenos_perLocus_summary_tamRun12$countSamples_lociDiff_12)),
         nrow = 3,
         ncol = 1) %>%
  `rownames<-`(., c("Total loci", "Loci w/multiple unique genos", "Max # loci w/unique genos")) %>%
  as.data.frame() %>%
  dplyr::rename("countSamples" = "V1") %>%
  rbind(., uniqueGenos_perLocus_bySet_tamRun12)

uniqueGenos_perLocus_summary_tamRun12_table %>%
  kable(caption = "Overview of unique genos called PER LOCUS between tamRun1 & 2") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# Plots
## vs. primer-probe reads 
ggplot(uniqueGenos_perLocus_summary_tamRun3, aes(x = Primer.Probe.Reads, y = countSamples_lociDiff, color = set)) +
  geom_point() +
  theme_bw()

## change locus to factor to keep df order
uniqueGenos_perLocus_summary_tamRun3$locus <- factor(uniqueGenos_perLocus_summary_tamRun3$locus, levels = uniqueGenos_perLocus_summary_tamRun3$locus)

## INDID - general & species-specific
uniqueGenos_perLocus_summary_tamRun3 %>%
  filter(set %in% c("INDID", "LWED", "SIMP")) %>%
ggplot(aes(x = locus, y = countSamples_lociDiff, fill = set)) +
  geom_bar(stat = "identity", color = "black") +
  coord_flip() +
  theme_bw() +
  facet_grid(rows = vars(set), scales = "free") +
  labs(y = "Number of samples",
       title = "Number of samples with different INDID genotypes between tamRun3 a, b, and cat")

## SEXID - general & species-specific
uniqueGenos_perLocus_summary_tamRun3 %>%
  filter(set %in% c("SEXID", "SEXID_LWED", "SEXID_SIMP")) %>%
ggplot(aes(x = locus, y = countSamples_lociDiff, fill = set)) +
  geom_bar(stat = "identity", color = "black") +
  coord_flip() +
  theme_bw() +
  facet_grid(rows = vars(set), scales = "free") +
  labs(y = "Number of samples",
       title = "Number of samples with different SEXID genotypes between tamRun3 a, b, and cat")

## SPECIES
uniqueGenos_perLocus_summary_tamRun3 %>%
  filter(set == "SPECIES") %>%
ggplot(aes(x = locus, y = countSamples_lociDiff, fill = set)) +
  geom_bar(stat = "identity", color = "black") +
  coord_flip() +
  theme_bw() +
  facet_grid(rows = vars(set), scales = "free") +
  labs(y = "Number of samples",
       title = "Number of samples with different SPECIESID genotypes between tamRun3 a, b, and cat")
```

## 5.2 tamRun5 vs. 1 & 2

For samples genotyped in tamRun5 as well as tamRun1/2 (including duplicates within runs 1 & 2), how well do their genotype calls match up?

### 5.2.1 Comparison dataframe

**1) Create general names for each run based on animalID and sampleType for merging**

```{r}
# tamRun1
nameType_tamRun1 <- md_tamRun1 %>%
  mutate(sampleID = gsub("_", "\\.", sampleID)) %>%
  select(c(sampleID, animalID, sampleType)) %>%
  distinct() %>%
  na.omit() %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  unite("sample", sampleID:sample, sep = "--", remove = F)

genos_tamRun1t <- genos_tamRun1 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "run1") %>%
  mutate(run1 = sub('_[^.]+', '', run1)) %>%
  merge(., nameType_tamRun1, by.x = "run1", by.y = "sampleID") %>%
  relocate(sample, .after = run1) %>%
  rename_all(~sub("\\..*", "", .x)) %>%
  select(!c(run1, animalID, sampleType))

# tamRun2
nameType_tamRun2 <- md_tamRun2 %>%
  mutate(sampleID = gsub("_", "\\.", sampleID)) %>%
  filter(included_inRun == "yes") %>%
  select(c(sampleID, animalID, sampleType)) %>%
  distinct() %>%
  na.omit() %>%
  mutate(sampleType = gsub("hair \\(chelex\\)", "hair", sampleType)) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  unite("sample", sampleID:sample, sep = "--", remove = F)

genos_tamRun2t <- genos_tamRun2 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "run2") %>%
  mutate(run2 = sub('_[^.]+', '', run2)) %>%
  merge(., nameType_tamRun2, by.x = "run2", by.y = "sampleID") %>%
  relocate(sample, .after = run2) %>%
  rename_all(~sub("\\..*", "", .x)) %>%
  select(!c(run2, animalID, sampleType))

# tamRun3
nameType_tamRun5 <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  filter(!sampleType %in% c("xtnNeg", "pcr1Neg")) %>%
  select(c(sampleID, animalID, sampleType)) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  unite("sample", sampleID:sample, sep = "--", remove = F)

genos_tamRun5t <- genos_tamRun5 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "run5") %>%
#  mutate(run3 = sub('_[^.]+', '', run3)) %>%
  merge(., nameType_tamRun5, by.x = "run5", by.y = "sampleID") %>%
  relocate(sample, .after = run5) %>%
  #rename_all(~sub("[_][^_]+$", "", .x)) %>%
  select(!c(run5, animalID, sampleType))
```

**2) Create long dataframes**

```{r}
genos_tamRun1l <- genos_tamRun1t %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("run1Geno" = "genotype") %>%
  separate(sample, into = c("run1", "sample"), sep = "--") %>%
  select(!run1) %>%
  group_by(sample, locus) %>%
  mutate(run1GenoID = paste0("run1Geno", row_number())) %>%
  spread(run1GenoID, run1Geno)

genos_tamRun2l <- genos_tamRun2t %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("run2Geno" = "genotype") %>%
  separate(sample, into = c("run2", "sample"), sep = "--") %>%
  select(!run2) %>%
  group_by(sample, locus) %>%
  mutate(run2GenoID = paste0("run2Geno", row_number())) %>%
  spread(run2GenoID, run2Geno)

genos_tamRun5l <- genos_tamRun5t %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("run5Geno" = "genotype") %>%
  separate(sample, into = c("run5", "sample"), sep = "--") %>%
  select(!run5) %>%
  group_by(sample, locus) %>%
  mutate(run5GenoID = paste0("run5Geno", row_number())) %>%
  spread(run5GenoID, run5Geno)
```

**3) Comparison dataframe**

```{r}
genos_tamRun125 <- merge(genos_tamRun1l, genos_tamRun2l, by = c("sample", "locus"), all = T) %>%
  merge(., genos_tamRun5l, by = c("sample", "locus"), all = T) %>%
  filter(!str_detect(sample, "fecal")) %>%
  mutate(genosCalled_125 = 
           rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  mutate(genosCalled_12 = 
           rowSums(select(., -sample, -locus, -run5Geno1, -genosCalled_125) != 0, na.rm = T)) %>%
  mutate(zerosCalled_125 = 
           rowSums(select(., -sample, -locus, -genosCalled_125, -genosCalled_12) == 0, na.rm = T)) %>%
  mutate(zerosCalled_12 = 
           rowSums(select(., -sample, -locus, -run5Geno1, -genosCalled_125, -genosCalled_12, -zerosCalled_125) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos_125 = ifelse(zerosCalled_125 > 0,
                              (n_distinct(c_across(run1Geno1:run5Geno1), na.rm = T)) - 1,
                              (n_distinct(c_across(run1Geno1:run5Geno1), na.rm = T))
                              )) %>%
  rowwise() %>%
  mutate(uniqueGenos_12 = ifelse(zerosCalled_12 > 0,
                              (n_distinct(c_across(run1Geno1:run2Geno8), na.rm = T)) - 1,
                              (n_distinct(c_across(run1Geno1:run2Geno8), na.rm = T))
                              )) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T)
```

### 5.2.2 Sample consistency

Between tamRun1 & 2, 23 samples had loci with more than one genotype between runs (including if the sample was duplicated within a run). Of these, 15 had at least one locus with different genotypes, with at most 27 loci with differing genotypes (19.6% of loci genotyped for that sample).

The number of loci with different calls between datasets does not appear to be related to the number of on-target reads (from "cat" data) per sample.

```{r}
# Summary stats per sample tamRun1 vs. tamRun2
genoMatch_bySample_summary_tamRun125 <- genos_tamRun125 %>%
  group_by(sample) %>%
  summarise(lociGenotyped = length(locus),
            lociDiff = length(locus[uniqueGenos_125 > 1]),
            min_uniqueGenos_125 = min(uniqueGenos_125),
            max_uniqueGenos_125 = max(uniqueGenos_125),
            mean_uniqueGenos_125 = mean(uniqueGenos_125),
            var_uniqueGenos_125 = var(uniqueGenos_125)) %>%
  as.data.frame() %>%
  mutate(propDiff = (mean_uniqueGenos_125 - 1)) %>%
  relocate(propDiff, .after = mean_uniqueGenos_125) %>%
  mutate(sampleType = gsub("^.*_","", sample))

uniqueGenos_perSample_byLocusSet_tamRun125 <- genos_tamRun125 %>%
  filter(uniqueGenos_125 > 1) %>%
  group_by(set, sample) %>%
  summarise(countUnique = length(sample)) %>%
  as.data.frame() %>%
  group_by(set) %>%
  summarise(countSamples = length(sample)) %>%
  na.omit() %>%
  column_to_rownames("set")
```

**Variation in number of unique genotypes called per locus across samples**

```{r}
summary(genos_tamRun125$uniqueGenos_125[genos_tamRun125$genosCalled_125 > 1])
```

**Overview of genotype differences PER SAMPLE between tamRun1 & 2**

```{r, echo=FALSE}
uniqueGenos_perSample_summaryTable_tamRun125 <- matrix(c(length(uniqueGenos_perSample_byLocusSet_tamRun125$sample),
         uniqueGenos_perSample_byLocusSet_tamRun125 %>%
           filter(lociDiff > 1) %>%
         nrow(),
         mean(uniqueGenos_perSample_byLocusSet_tamRun125$lociDiff),
         mean(uniqueGenos_perSample_byLocusSet_tamRun125$propDiff),
         max(uniqueGenos_perSample_byLocusSet_tamRun125$lociDiff),
         max(uniqueGenos_perSample_byLocusSet_tamRun125$propDiff)),
         nrow = 6,
         ncol = 1) %>%
  `rownames<-`(., c("Total samples", "Samples w/multiple unique genos", "Avg # of loci w/unique genos per sample", "Avg prop. loci w/unique genos per sample", "Max # loci w/unique genos per sample", "Max prop. loci w/unique genos per sample")) %>%
  as.data.frame() %>%
  mutate(V1 = signif(V1, digits = 3)) %>%
  mutate(V1 = as.character(V1))

names(uniqueGenos_perSample_summaryTable_tamRun125) <- NULL

uniqueGenos_perSample_summaryTable_tamRun125 %>%
  kable(caption = "Overview of genos called across SAMPLES between tamRun3 vs. 1 & 2") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

#### Per locus

Across tamRun1 & 2, 135 out of 221 loci had samples genotyped in more than one dataset. Of these, ALL loci had at least one sample with different genotypes in either a, b, or cat datasets, with at most 11 samples with differing genotypes.

The number of samples with different calls between datasets does not appear to be related to the number of on-target reads (from "cat" data) per sample, though most loci with differing genotypes appear to be from the INDID set.

```{r, echo=FALSE}
# Summary stats per locus
uniqueGenos_perLocus_summary_tamRun125 <- genos_tamRun125 %>%
  group_by(locus) %>%
  summarise(samplesGenotyped = length(sample),
            countSamples_lociDiff_125 = length(sample[uniqueGenos_125 > 1]),
            min_uniqueGenos_125 = min(uniqueGenos_125),
            max_uniqueGenos_125 = max(uniqueGenos_125),
            mean_uniqueGenos_125 = mean(uniqueGenos_125),
            var_uniqueGenos_125 = var(uniqueGenos_125)) %>%
  as.data.frame() %>%
  mutate(propDiff_125 = (mean_uniqueGenos_125 - 1)) %>%
  relocate(propDiff_125, .after = mean_uniqueGenos_125) %>%
  filter(locus %in% locusSum$Locus) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  group_by(set) %>%
  arrange(desc(countSamples_lociDiff_125), .by_group = T)

# Table
uniqueGenos_perLocus_summary_tamRun125_table <- matrix(c(length(uniqueGenos_perLocus_summary_tamRun125$locus),
         uniqueGenos_perLocus_summary_tamRun125 %>%
           filter(countSamples_lociDiff_12 > 1) %>%
         nrow(),
         max(uniqueGenos_perLocus_summary_tamRun125$countSamples_lociDiff_125)),
         nrow = 3,
         ncol = 1) %>%
  `rownames<-`(., c("Total loci", "Loci w/multiple unique genos", "Max # loci w/unique genos")) %>%
  as.data.frame() %>%
  dplyr::rename("countSamples" = "V1") %>%
  rbind(., uniqueGenos_perLocus_bySet_tamRun125)

uniqueGenos_perLocus_summary_tamRun125_table %>%
  kable(caption = "Overview of unique genos called PER LOCUS between tamRun1 & 2") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# Plots
## vs. primer-probe reads 
ggplot(uniqueGenos_perLocus_summary_tamRun3, aes(x = Primer.Probe.Reads, y = countSamples_lociDiff, color = set)) +
  geom_point() +
  theme_bw()

## change locus to factor to keep df order
uniqueGenos_perLocus_summary_tamRun3$locus <- factor(uniqueGenos_perLocus_summary_tamRun3$locus, levels = uniqueGenos_perLocus_summary_tamRun3$locus)

## INDID - general & species-specific
uniqueGenos_perLocus_summary_tamRun3 %>%
  filter(set %in% c("INDID", "LWED", "SIMP")) %>%
ggplot(aes(x = locus, y = countSamples_lociDiff, fill = set)) +
  geom_bar(stat = "identity", color = "black") +
  coord_flip() +
  theme_bw() +
  facet_grid(rows = vars(set), scales = "free") +
  labs(y = "Number of samples",
       title = "Number of samples with different INDID genotypes between tamRun3 a, b, and cat")

## SEXID - general & species-specific
uniqueGenos_perLocus_summary_tamRun3 %>%
  filter(set %in% c("SEXID", "SEXID_LWED", "SEXID_SIMP")) %>%
ggplot(aes(x = locus, y = countSamples_lociDiff, fill = set)) +
  geom_bar(stat = "identity", color = "black") +
  coord_flip() +
  theme_bw() +
  facet_grid(rows = vars(set), scales = "free") +
  labs(y = "Number of samples",
       title = "Number of samples with different SEXID genotypes between tamRun3 a, b, and cat")

## SPECIES
uniqueGenos_perLocus_summary_tamRun3 %>%
  filter(set == "SPECIES") %>%
ggplot(aes(x = locus, y = countSamples_lociDiff, fill = set)) +
  geom_bar(stat = "identity", color = "black") +
  coord_flip() +
  theme_bw() +
  facet_grid(rows = vars(set), scales = "free") +
  labs(y = "Number of samples",
       title = "Number of samples with different SPECIESID genotypes between tamRun3 a, b, and cat")
```

# 6 Species assignments

## 6.1 Code

### Species key

```{r}
speciesKey <- read.csv("./03_tamRun3/04_genoAnalyses/speciesSNP_key_14Dec2022.csv")

speciesKey_gtseq <- speciesKey %>%
  mutate(
    genotype = gsub(",", "", genotype)
  )
```

### Species assignments

#### 10x coverage

```{r}
genos_species_10x <- genos_tamRun5_10x %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SPECIES")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  
  # assign species to each genotype
  left_join(speciesKey, by = c("Locus", "genotype")) %>%
  mutate(species = coalesce(species, genotype)) %>%
  select(c(Locus, sampleID, species)) %>%
  pivot_wider(names_from = sampleID,
              values_from = species) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  
  # count total loci w/species assignment
  mutate(totalGenos_sp = (rowSums(. == "LWED") + rowSums(. == "SIMP"))) %>%
  # get prop of loci w/species assignments
  mutate(propGenos_sp = (totalGenos_sp/12)) %>%
  
  # get prop LWED assignments
  mutate(propLWED = (rowSums(. == "LWED")/totalGenos_sp)) %>%
  
  # get prop SIMP assignments
  mutate(propSIMP = (rowSums(. == "SIMP")/totalGenos_sp)) %>%
  
  # get prop mismatched assignments
  mutate(propMismatch_species = (1 - abs(propLWED - propSIMP))) %>%
  
  # assign species
  mutate(speciesAssigned = ifelse(propLWED == 1, "LWED", 
                              ifelse(propSIMP == 1, "SIMP", 
                                     ifelse(propLWED < 1 & propSIMP > 0, "MIX", NA)))) %>%
  
  rownames_to_column("sampleID") %>%
  relocate(c(speciesAssigned, totalGenos_sp, propGenos_sp, propMismatch_species, propLWED, propSIMP), .after = sampleID)
```

#### 30x coverage

```{r}
genos_species_30x <- genos_tamRun5_30x %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SPECIES")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  
  # assign species to each genotype
  left_join(speciesKey, by = c("Locus", "genotype")) %>%
  mutate(species = coalesce(species, genotype)) %>%
  select(c(Locus, sampleID, species)) %>%
  pivot_wider(names_from = sampleID,
              values_from = species) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  
  # count total loci w/species assignment
  mutate(totalGenos_sp = (rowSums(. == "LWED") + rowSums(. == "SIMP"))) %>%
  # get prop of loci w/species assignments
  mutate(propGenos_sp = (totalGenos_sp/12)) %>%
  
  # get prop LWED assignments
  mutate(propLWED = (rowSums(. == "LWED")/totalGenos_sp)) %>%
  
  # get prop SIMP assignments
  mutate(propSIMP = (rowSums(. == "SIMP")/totalGenos_sp)) %>%
  
  # get prop mismatched assignments
  mutate(propMismatch_species = (1 - abs(propLWED - propSIMP))) %>%
  
  # assign species
  mutate(speciesAssigned = ifelse(propLWED == 1, "LWED", 
                              ifelse(propSIMP == 1, "SIMP", 
                                     ifelse(propLWED < 1 & propSIMP > 0, "MIX", NA)))) %>%
  
  rownames_to_column("sampleID") %>%
  relocate(c(speciesAssigned, totalGenos_sp, propGenos_sp, propMismatch_species, propLWED, propSIMP), .after = sampleID)
```

#### gtseq

```{r}
genos_species_gtseq <- genos_tamRun5_gtseq %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SPECIES")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  
  # assign species to each genotype
  left_join(speciesKey_gtseq, by = c("Locus", "genotype")) %>%
  mutate(species = coalesce(species, genotype)) %>%
  select(c(Locus, sampleID, species)) %>%
  pivot_wider(names_from = sampleID,
              values_from = species) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  
  # count total loci w/species assignment
  mutate(totalGenos_sp = (rowSums(. == "LWED") + rowSums(. == "SIMP"))) %>%
  # get prop of loci w/species assignments
  mutate(propGenos_sp = (totalGenos_sp/12)) %>%
  
  # get prop LWED assignments
  mutate(propLWED = (rowSums(. == "LWED")/totalGenos_sp)) %>%
  
  # get prop SIMP assignments
  mutate(propSIMP = (rowSums(. == "SIMP")/totalGenos_sp)) %>%
  
  # get prop mismatched assignments
  mutate(propMismatch_species = (1 - abs(propLWED - propSIMP))) %>%
  
  # assign species
  mutate(speciesAssigned = ifelse(propLWED == 1, "LWED", 
                              ifelse(propSIMP == 1, "SIMP", 
                                     ifelse(propLWED < 1 & propSIMP > 0, "MIX", NA)))) %>%
  
  rownames_to_column("sampleID") %>%
  relocate(c(speciesAssigned, totalGenos_sp, propGenos_sp, propMismatch_species, propLWED, propSIMP), .after = sampleID)
```

## 6.2 Results

### 10x coverage

80% of all blood/hair samples (n = 344 out of 431) were assigned a species, including 93% of blood samples (n = 181 out of 194) and 69% of hair samples (n = 163 out of 235). No samples had a mix of species assignments, so any species assigned had either 100% LWED or 100% SIMP assignments.

None of the 9 negatives were genotyped at any SPECIES loci.

```{r, echo=FALSE}
# Connect metadata
genos_species_md_10x <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  merge(., genos_species_10x, by = "sampleID") %>%
  mutate(mdMatch_sp5 = ifelse(species == speciesAssigned,"TRUE", "FALSE")) %>%
  relocate(animalID, .after = sampleID) %>%
  relocate(mdMatch_sp5, .after = animalID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F)

genos_species_md_noNegs_10x <- genos_species_md_10x %>%
  filter(!str_detect(sampleType, "Neg"))

# How many assignments were made?
speciesCount_all_10x <- genos_species_md_10x %>%
  filter(sampleType %in% c("hair", "blood")) %>%
  summarise(totalAll_10x = sum(!is.na(speciesAssigned)))%>%
  mutate(propAll_10x = totalAll_10x/(440 - 9)) %>%
  mutate(propAll_10x = signif(propAll_10x, digits = 2)) %>%
  mutate(propAll_10x = scales::percent(propAll_10x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_blood_10x <- genos_species_md_10x %>%
  filter(sampleType == "blood") %>%
#  group_by(seqRun) %>%
  summarise(totalBlood_10x = sum(!is.na(speciesAssigned))) %>%
  mutate(propBlood_10x = totalBlood_10x/194) %>%
  mutate(propBlood_10x = signif(propBlood_10x, digits = 2)) %>%
  mutate(propBlood_10x = scales::percent(propBlood_10x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_hair_10x <- genos_species_md_10x %>%
  filter(sampleType == "hair") %>%
#  group_by(seqRun) %>%
  summarise(totalHair_10x = sum(!is.na(speciesAssigned))) %>%
  mutate(propHair_10x = totalHair_10x/237) %>%
  mutate(propHair_10x = signif(propHair_10x, digits = 2)) %>%
  mutate(propHair_10x = scales::percent(propHair_10x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_df_10x <- rbind(speciesCount_blood_10x, speciesCount_hair_10x, speciesCount_all_10x) %>%
  t() %>%
  as.data.frame() #%>%
#  rownames_to_column("seqRun")

speciesCount_df_10x %>%
#  dplyr::rename("tamRun5 data" = "seqRun") %>%
  kable(caption = "10x: Successful species assignments by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# any with mixed assignments?
genos_species_md_10x %>%
  filter(speciesAssigned == "MIX") %>%
  nrow()

# any negatives assigned a sex?
genos_species_negs_10x <- genos_species_md_10x %>%
  filter(str_detect(sampleType, "Neg"))
```

### 30x coverage

76% of all blood/hair samples (n = 329 out of 431) were assigned a species, including 92% of blood samples (n = 179 out of 194) and 63% of hair samples (n = 150 out of 235). No samples had a mix of species assignments, so any species assigned had either 100% LWED or 100% SIMP assignments.

None of the 9 negatives were genotyped at any SPECIES loci.

```{r, echo=FALSE}
# Connect metadata
genos_species_md_30x <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  merge(., genos_species_30x, by = "sampleID") %>%
  mutate(mdMatch_sp5 = ifelse(species == speciesAssigned,"TRUE", "FALSE")) %>%
  relocate(animalID, .after = sampleID) %>%
  relocate(mdMatch_sp5, .after = animalID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F)

genos_species_md_noNegs_30x <- genos_species_md_30x %>%
  filter(!str_detect(sampleType, "Neg"))

# How many assignments were made?
speciesCount_all_30x <- genos_species_md_30x %>%
  filter(sampleType %in% c("hair", "blood")) %>%
  summarise(totalAll_30x = sum(!is.na(speciesAssigned)))%>%
  mutate(propAll_30x = totalAll_30x/(440 - 9)) %>%
  mutate(propAll_30x = signif(propAll_30x, digits = 2)) %>%
  mutate(propAll_30x = scales::percent(propAll_30x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_blood_30x <- genos_species_md_30x %>%
  filter(sampleType == "blood") %>%
#  group_by(seqRun) %>%
  summarise(totalBlood_30x = sum(!is.na(speciesAssigned))) %>%
  mutate(propBlood_30x = totalBlood_30x/194) %>%
  mutate(propBlood_30x = signif(propBlood_30x, digits = 2)) %>%
  mutate(propBlood_30x = scales::percent(propBlood_30x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_hair_30x <- genos_species_md_30x %>%
  filter(sampleType == "hair") %>%
#  group_by(seqRun) %>%
  summarise(totalHair_30x = sum(!is.na(speciesAssigned))) %>%
  mutate(propHair_30x = totalHair_30x/237) %>%
  mutate(propHair_30x = signif(propHair_30x, digits = 2)) %>%
  mutate(propHair_30x = scales::percent(propHair_30x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_df_30x <- rbind(speciesCount_blood_30x, speciesCount_hair_30x, speciesCount_all_30x) %>%
  t() %>%
  as.data.frame() #%>%
#  rownames_to_column("seqRun")

speciesCount_df_30x %>%
#  dplyr::rename("tamRun5 data" = "seqRun") %>%
  kable(caption = "30x: Successful species assignments by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# any with mixed assignments?
genos_species_md_30x %>%
  filter(speciesAssigned == "MIX") %>%
  nrow()

# any negatives assigned a sex?
genos_species_negs_30x <- genos_species_md_30x %>%
  filter(str_detect(sampleType, "Neg"))
```

### gtseq

```{r}
# Connect metadata
genos_species_md_gtseq <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  merge(., genos_species_gtseq, by = "sampleID") %>%
  mutate(mdMatch_sp5 = ifelse(species == speciesAssigned,"TRUE", "FALSE")) %>%
  relocate(animalID, .after = sampleID) %>%
  relocate(mdMatch_sp5, .after = animalID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F)

genos_species_md_noNegs_gtseq <- genos_species_md_gtseq %>%
  filter(!str_detect(sampleType, "Neg"))

# How many assignments were made?
speciesCount_all_gtseq <- genos_species_md_gtseq %>%
  filter(sampleType %in% c("hair", "blood")) %>%
  summarise(totalAll_gtseq = sum(!is.na(speciesAssigned)))%>%
  mutate(propAll_gtseq = totalAll_gtseq/(440 - 9)) %>%
  mutate(propAll_gtseq = signif(propAll_gtseq, digits = 2)) %>%
  mutate(propAll_gtseq = scales::percent(propAll_gtseq)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_blood_gtseq <- genos_species_md_gtseq %>%
  filter(sampleType == "blood") %>%
#  group_by(seqRun) %>%
  summarise(totalBlood_gtseq = sum(!is.na(speciesAssigned))) %>%
  mutate(propBlood_gtseq = totalBlood_gtseq/194) %>%
  mutate(propBlood_gtseq = signif(propBlood_gtseq, digits = 2)) %>%
  mutate(propBlood_gtseq = scales::percent(propBlood_gtseq)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_hair_gtseq <- genos_species_md_gtseq %>%
  filter(sampleType == "hair") %>%
#  group_by(seqRun) %>%
  summarise(totalHair_gtseq = sum(!is.na(speciesAssigned))) %>%
  mutate(propHair_gtseq = totalHair_gtseq/237) %>%
  mutate(propHair_gtseq = signif(propHair_gtseq, digits = 2)) %>%
  mutate(propHair_gtseq = scales::percent(propHair_gtseq)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

speciesCount_df_gtseq <- rbind(speciesCount_blood_gtseq, speciesCount_hair_gtseq, speciesCount_all_gtseq) %>%
  t() %>%
  as.data.frame() #%>%
#  rownames_to_column("seqRun")

speciesCount_df_gtseq %>%
#  dplyr::rename("tamRun5 data" = "seqRun") %>%
  kable(caption = "gtseq: Successful species assignments by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# any with mixed assignments?
genos_species_md_gtseq %>%
  filter(speciesAssigned == "MIX") %>%
  nrow()

# any negatives assigned a sex?
genos_species_negs_gtseq <- genos_species_md_gtseq %>%
  filter(str_detect(sampleType, "Neg"))
```

# 7 Sex assignments

## 7.1 Code

### Sex key

```{r}
sexKey <- read.csv("./03_tamRun3/04_genoAnalyses/sexSNP_key_8Dec2022.csv") %>%
  na.omit() %>%
  unique() %>%
  filter(Locus %in% primerList_v3$locus) %>%
  dplyr::rename("locus" = "Locus")

loci_sex_pp2 <- c('SEXID_SIMP_197', 'SEXID_SIMP_198', 'SEXID_200', 'SEXID_SIMP_203', 'SEXID_SIMP_218', 'SEXID_222')
loci_sex_pp3 <- c('SEXID_LWED_195', 'SEXID_205', 'SEXID_LWED_208', 'SEXID_209', 'SEXID_215')

sexKey_gtseq <- sexKey %>%
  mutate(
    genotype = gsub(",", "", genotype)
  )
```

### Sex assignments

#### 10x coverage

REDO THIS SO GET ASSIGNMENTS FOR LWED AND SIMP SEXID SETS

```{r}
genos_sex_spSpecific_10x <- genos_tamRun5_10x %>%
  rownames_to_column("locus") %>%
  filter(str_detect(locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey, by = c("locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun5[, c("sampleID", "species")], by = "sampleID") %>%
  mutate(species = 
           case_when(
             sampleID == "tamRun5.411" ~ "pcr1Neg",
             is.na(species) ~ "xtnNeg",
             .default = species
           )) %>%
  mutate(
    SEXID_LWED_195 = case_when(
      species == "SIMP" ~ NA,
      .default = SEXID_LWED_195
    ),
    SEXID_LWED_208 = case_when(
      species == "SIMP" ~ NA,
      .default = SEXID_LWED_208
    ),
    SEXID_SIMP_197 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_197
    ),
    SEXID_SIMP_198 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_198
    ),
    SEXID_SIMP_203 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_203
    ),
    SEXID_SIMP_218 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_218
    )
  ) %>%
  column_to_rownames("sampleID") %>%
  # assign sex
  mutate(totalGenos_sex = (rowSums(. == "F", na.rm = T) + rowSums(. == "M", na.rm = T))) %>%
  mutate(propGenos_sex = 
           case_when(
             species == "LWED" ~ (totalGenos_sex/7),
             species == "SIMP" ~ (totalGenos_sex/9),
             .default = totalGenos_sex/11
             )
         ) %>%
  mutate(propF = (rowSums(. == "F", na.rm = T)/totalGenos_sex)) %>%
  mutate(propM = (rowSums(. == "M", na.rm = T)/totalGenos_sex)) %>%
  mutate(propMismatch_sex = (1 - abs(propF - propM))) %>%
  mutate(sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  select(c("sampleID",
           "sexAssigned",
           "totalGenos_sex",
           "propGenos_sex",
           "propMismatch_sex",
           "propF",
           "propM",
           "SEXID_200",
           "SEXID_205",
           "SEXID_209",
           "SEXID_215",
           "SEXID_222",
           # LWED-specific SEXID
           "SEXID_LWED_195",
           "SEXID_LWED_208",
           # SIMP-specific SEXID
           "SEXID_SIMP_197",
           "SEXID_SIMP_198",
           "SEXID_SIMP_203",
           "SEXID_SIMP_218"))
```

#### 30x coverage

```{r}
genos_sex_spSpecific_30x <- genos_tamRun5_30x %>%
  rownames_to_column("locus") %>%
  filter(str_detect(locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey, by = c("locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun5[, c("sampleID", "species")], by = "sampleID") %>%
  mutate(species = 
           case_when(
             sampleID == "tamRun5.411" ~ "pcr1Neg",
             is.na(species) ~ "xtnNeg",
             .default = species
           )) %>%
  mutate(
    SEXID_LWED_195 = case_when(
      species == "SIMP" ~ NA,
      .default = SEXID_LWED_195
    ),
    SEXID_LWED_208 = case_when(
      species == "SIMP" ~ NA,
      .default = SEXID_LWED_208
    ),
    SEXID_SIMP_197 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_197
    ),
    SEXID_SIMP_198 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_198
    ),
    SEXID_SIMP_203 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_203
    ),
    SEXID_SIMP_218 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_218
    )
  ) %>%
  column_to_rownames("sampleID") %>%
  # assign sex
  mutate(totalGenos_sex = (rowSums(. == "F", na.rm = T) + rowSums(. == "M", na.rm = T))) %>%
  mutate(propGenos_sex = 
           case_when(
             species == "LWED" ~ (totalGenos_sex/7),
             species == "SIMP" ~ (totalGenos_sex/9),
             .default = totalGenos_sex/11
             )
         ) %>%
  mutate(propF = (rowSums(. == "F", na.rm = T)/totalGenos_sex)) %>%
  mutate(propM = (rowSums(. == "M", na.rm = T)/totalGenos_sex)) %>%
  mutate(propMismatch_sex = (1 - abs(propF - propM))) %>%
  mutate(sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  select(c("sampleID",
           "sexAssigned",
           "totalGenos_sex",
           "propGenos_sex",
           "propMismatch_sex",
           "propF",
           "propM",
           "SEXID_200",
           "SEXID_205",
           "SEXID_209",
           "SEXID_215",
           "SEXID_222",
           # LWED-specific SEXID
           "SEXID_LWED_195",
           "SEXID_LWED_208",
           # SIMP-specific SEXID
           "SEXID_SIMP_197",
           "SEXID_SIMP_198",
           "SEXID_SIMP_203",
           "SEXID_SIMP_218"))
```

#### gtseq

```{r}
genos_sex_spSpecific_gtseq <- genos_tamRun5_gtseq %>%
  rownames_to_column("locus") %>%
  filter(str_detect(locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey_gtseq, by = c("locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun5[, c("sampleID", "species")], by = "sampleID") %>%
  mutate(species = 
           case_when(
             sampleID == "tamRun5.411" ~ "pcr1Neg",
             is.na(species) ~ "xtnNeg",
             .default = species
           )) %>%
  mutate(
    SEXID_LWED_195 = case_when(
      species == "SIMP" ~ NA,
      .default = SEXID_LWED_195
    ),
    SEXID_LWED_208 = case_when(
      species == "SIMP" ~ NA,
      .default = SEXID_LWED_208
    ),
    SEXID_SIMP_197 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_197
    ),
    SEXID_SIMP_198 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_198
    ),
    SEXID_SIMP_203 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_203
    ),
    SEXID_SIMP_218 = case_when(
      species == "LWED" ~ NA,
      .default = SEXID_SIMP_218
    )
  ) %>%
  column_to_rownames("sampleID") %>%
  # assign sex
  mutate(totalGenos_sex = (rowSums(. == "F", na.rm = T) + rowSums(. == "M", na.rm = T))) %>%
  mutate(propGenos_sex = 
           case_when(
             species == "LWED" ~ (totalGenos_sex/7),
             species == "SIMP" ~ (totalGenos_sex/9),
             .default = totalGenos_sex/11
             )
         ) %>%
  mutate(propF = (rowSums(. == "F", na.rm = T)/totalGenos_sex)) %>%
  mutate(propM = (rowSums(. == "M", na.rm = T)/totalGenos_sex)) %>%
  mutate(propMismatch_sex = (1 - abs(propF - propM))) %>%
  mutate(sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  select(c("sampleID",
           "sexAssigned",
           "totalGenos_sex",
           "propGenos_sex",
           "propMismatch_sex",
           "propF",
           "propM",
           "SEXID_200",
           "SEXID_205",
           "SEXID_209",
           "SEXID_215",
           "SEXID_222",
           # LWED-specific SEXID
           "SEXID_LWED_195",
           "SEXID_LWED_208",
           # SIMP-specific SEXID
           "SEXID_SIMP_197",
           "SEXID_SIMP_198",
           "SEXID_SIMP_203",
           "SEXID_SIMP_218"))
```

## 7.2 Results

### 10x coverage

77% of blood/hair samples (n = 330 out of 431) were assigned a sex, including 92% of blood samples (n = 179 out of 194) and 64% of hair samples (n = 151 out of 235).

2 out of 9 negatives were genotyped at a SEXID locus. xtnNeg_p2.F9_xtnNeg was genotyped at SEXID_222, and received a sex assignment of M; xtnNeg_p4.F9_xtnNeg was genotyped at SEXID_SIMP_218, but received no sex assignment.

```{r, echo=FALSE}
# Connect metadata
genos_sex_md_10x <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  merge(., genos_sex_spSpecific_10x, by = "sampleID") %>%
  mutate(mdMatch_sex = 
           case_when(
             sexAssigned == "MIX" ~ "MIX",
             sex == sexAssigned ~ "TRUE",
             sex != sexAssigned ~ "FALSE",
             is.na(sex) & !is.na(sexAssigned) ~ "FALSE")) %>%
  relocate(mdMatch_sex, .after = sampleID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  relocate(c(sex, sexAssigned, totalGenos_sex, propGenos_sex, propMismatch_sex, propF, propM), .after = mdMatch_sex) %>%
  relocate(sample, .after = sampleID)

genos_sex_md_noNegs_10x <- genos_sex_md_10x %>%
  filter(!str_detect(sampleType, "Neg"))

# How many assignments were made?
sexCount_all_10x <- genos_sex_md_10x %>%
  filter(sampleType %in% c("hair", "blood")) %>%
#  group_by(seqRun) %>%
  summarise(totalAll_10x = sum(!is.na(sexAssigned)))%>%
  mutate(propAll_10x = totalAll_10x/(440 - 9)) %>%
  mutate(propAll_10x = signif(propAll_10x, digits = 2)) %>%
  mutate(propAll_10x = scales::percent(propAll_10x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_blood_10x <- genos_sex_md_10x %>%
  filter(sampleType == "blood") %>%
#  group_by(seqRun) %>%
  summarise(totalBlood_10x = sum(!is.na(sexAssigned))) %>%
  mutate(propBlood_10x = totalBlood_10x/194) %>%
  mutate(propBlood_10x = signif(propBlood_10x, digits = 2)) %>%
  mutate(propBlood_10x = scales::percent(propBlood_10x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_hair_10x <- genos_sex_md_10x %>%
  filter(sampleType == "hair") %>%
#  group_by(seqRun) %>%
  summarise(totalHair_10x = sum(!is.na(sexAssigned))) %>%
  mutate(propHair_10x = totalHair_10x/235) %>%
  mutate(propHair_10x = signif(propHair_10x, digits = 2)) %>%
  mutate(propHair_10x = scales::percent(propHair_10x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_df_10x <- rbind(sexCount_blood_10x, sexCount_hair_10x, sexCount_all_10x) %>%
  t() %>%
  as.data.frame()

sexCount_df_10x %>%
  kable(caption = "10x: Successful sex assignments by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# any negatives assigned a sex?
genos_sex_negs_10x <- genos_sex_md_10x %>%
  filter(str_detect(sampleType, "Neg"))
```

### 30x coverage

73% of blood/hair samples (n = 315 out of 431) were assigned a sex, including 90% of blood samples (n = 174 out of 194) and 60% of hair samples (n = 141 out of 235).

No negatives were genotyped at a SEXID locus.

```{r, echo=FALSE}
# Connect metadata
genos_sex_md_30x <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  merge(., genos_sex_spSpecific_30x, by = "sampleID") %>%
  mutate(mdMatch_sex = 
           case_when(
             sexAssigned == "MIX" ~ "MIX",
             sex == sexAssigned ~ "TRUE",
             sex != sexAssigned ~ "FALSE",
             is.na(sex) & !is.na(sexAssigned) ~ "FALSE")) %>%
  relocate(mdMatch_sex, .after = sampleID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  relocate(c(sex, sexAssigned, totalGenos_sex, propGenos_sex, propMismatch_sex, propF, propM), .after = mdMatch_sex) %>%
  relocate(sample, .after = sampleID)

genos_sex_md_noNegs_30x <- genos_sex_md_30x %>%
  filter(!str_detect(sampleType, "Neg"))

# How many assignments were made?
sexCount_all_30x <- genos_sex_md_30x %>%
  filter(sampleType %in% c("hair", "blood")) %>%
#  group_by(seqRun) %>%
  summarise(totalAll_30x = sum(!is.na(sexAssigned)))%>%
  mutate(propAll_30x = totalAll_30x/(440 - 9)) %>%
  mutate(propAll_30x = signif(propAll_30x, digits = 2)) %>%
  mutate(propAll_30x = scales::percent(propAll_30x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_blood_30x <- genos_sex_md_30x %>%
  filter(sampleType == "blood") %>%
#  group_by(seqRun) %>%
  summarise(totalBlood_30x = sum(!is.na(sexAssigned))) %>%
  mutate(propBlood_30x = totalBlood_30x/194) %>%
  mutate(propBlood_30x = signif(propBlood_30x, digits = 2)) %>%
  mutate(propBlood_30x = scales::percent(propBlood_30x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_hair_30x <- genos_sex_md_30x %>%
  filter(sampleType == "hair") %>%
#  group_by(seqRun) %>%
  summarise(totalHair_30x = sum(!is.na(sexAssigned))) %>%
  mutate(propHair_30x = totalHair_30x/235) %>%
  mutate(propHair_30x = signif(propHair_30x, digits = 2)) %>%
  mutate(propHair_30x = scales::percent(propHair_30x)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_df_30x <- rbind(sexCount_blood_30x, sexCount_hair_30x, sexCount_all_30x) %>%
  t() %>%
  as.data.frame()

sexCount_df_30x %>%
  kable(caption = "30x: Successful sex assignments by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# any negatives assigned a sex?
genos_sex_negs_30x <- genos_sex_md_30x %>%
  filter(str_detect(sampleType, "Neg"))
```

### gtseq

```{r}
# Connect metadata
genos_sex_md_gtseq <- md_tamRun5 %>%
  mutate(sampleID = gsub("-","\\.", sampleID)) %>%
  merge(., genos_sex_spSpecific_gtseq, by = "sampleID") %>%
  mutate(mdMatch_sex = 
           case_when(
             sexAssigned == "MIX" ~ "MIX",
             sex == sexAssigned ~ "TRUE",
             sex != sexAssigned ~ "FALSE",
             is.na(sex) & !is.na(sexAssigned) ~ "FALSE")) %>%
  relocate(mdMatch_sex, .after = sampleID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F) %>%
  relocate(c(sex, sexAssigned, totalGenos_sex, propGenos_sex, propMismatch_sex, propF, propM), .after = mdMatch_sex) %>%
  relocate(sample, .after = sampleID)

genos_sex_md_noNegs_gtseq <- genos_sex_md_gtseq %>%
  filter(!str_detect(sampleType, "Neg"))

# How many assignments were made?
sexCount_all_gtseq <- genos_sex_md_gtseq %>%
  filter(sampleType %in% c("hair", "blood")) %>%
#  group_by(seqRun) %>%
  summarise(totalAll_gtseq = sum(!is.na(sexAssigned)))%>%
  mutate(propAll_gtseq = totalAll_gtseq/(440 - 9)) %>%
  mutate(propAll_gtseq = signif(propAll_gtseq, digits = 2)) %>%
  mutate(propAll_gtseq = scales::percent(propAll_gtseq)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_blood_gtseq <- genos_sex_md_gtseq %>%
  filter(sampleType == "blood") %>%
#  group_by(seqRun) %>%
  summarise(totalBlood_gtseq = sum(!is.na(sexAssigned))) %>%
  mutate(propBlood_gtseq = totalBlood_gtseq/194) %>%
  mutate(propBlood_gtseq = signif(propBlood_gtseq, digits = 2)) %>%
  mutate(propBlood_gtseq = scales::percent(propBlood_gtseq)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_hair_gtseq <- genos_sex_md_gtseq %>%
  filter(sampleType == "hair") %>%
#  group_by(seqRun) %>%
  summarise(totalHair_gtseq = sum(!is.na(sexAssigned))) %>%
  mutate(propHair_gtseq = totalHair_gtseq/235) %>%
  mutate(propHair_gtseq = signif(propHair_gtseq, digits = 2)) %>%
  mutate(propHair_gtseq = scales::percent(propHair_gtseq)) %>%
#  column_to_rownames("seqRun") %>%
  t() %>%
  as.data.frame()

sexCount_df_gtseq <- rbind(sexCount_blood_gtseq, sexCount_hair_gtseq, sexCount_all_gtseq) %>%
  t() %>%
  as.data.frame()

sexCount_df_gtseq %>%
  kable(caption = "gtseq: Successful sex assignments by sample type") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))

# any negatives assigned a sex?
genos_sex_negs_gtseq <- genos_sex_md_gtseq %>%
  filter(str_detect(sampleType, "Neg"))
```

# 8 Investigating mismatches

As a data quality check, especially with regard to potential sample contamination, we can look at whether the species and sex assignments based on genotypes match those listed for each sample in the metadata.

## 8.1 Species mismatches

For all samples, including chimeric blood samples, we should expect 100% congruence with metadata for species.

**NOTE** that all SPECIESID loci are in primer pool 2

### 10x coverage

#### Overview

Of the 344 samples that received species assignments, 9 samples (2.2% of those called) did NOT match the species listed in the metadata. For all mismatches, 100% of SPECIESID loci called were either LWED or SIMP, with no mixed species calls.

**Species mismatches in tamRun5**

```{r, echo=FALSE}
# Extract samples with mismatches
genos_species_mismatch_10x <- genos_species_md_10x %>%
  filter(!mdMatch_sp5 == "TRUE")

write.csv(genos_species_mismatch_10x, "./05_tamRun5/03_run5GTscore/summaryFiles/genos_species_mismatch_gtscore10x.csv", row.names = F)

# By sample type
speciesMismatchAll_10x <- genos_species_mismatch_10x %>%
#  group_by(seqRun) %>%
  summarise(total = length(sampleID))

speciesMismatchBlood_10x <- genos_species_mismatch_10x %>%
#  group_by(seqRun) %>%
  summarise(blood = sum(sampleType == "blood"))

speciesMismatchHair_10x <- genos_species_mismatch_10x %>%
#  group_by(seqRun) %>%
  summarise(hair = sum(sampleType == "hair"))

speciesMismatchNeg_10x <- genos_species_mismatch_10x %>%
#  group_by(seqRun) %>%
  summarise("(-)" = sum(!sampleType %in% c("blood", "hair")))

# How many mismatches for each data set?
speciesMismatchCount_df_10x <- cbind(speciesMismatchAll_10x, speciesMismatchBlood_10x, speciesMismatchHair_10x, speciesMismatchNeg_10x)

speciesMismatchCount_df_10x %>%
  kable(caption = "10x: Species mismatch overview") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

**Species mismatches in tamRun5 by sample type according to species listed in metadata**

```{r, echo=FALSE}
speciesMismatch_summary_10x <- matrix(c(
  # LWED
  genos_species_md_10x %>%
    filter(species == "LWED") %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "LWED") %>%
    filter(!is.na(speciesAssigned)) %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "LWED") %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "LWED") %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    filter(sampleType == "blood") %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "LWED") %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    filter(sampleType == "hair") %>%
    nrow(),
  NA,
  # SIMP
  genos_species_md_10x %>%
    filter(species == "SIMP") %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "SIMP") %>%
    filter(!is.na(speciesAssigned)) %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "SIMP") %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "SIMP") %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    filter(sampleType == "blood") %>%
    nrow(),
  genos_species_md_10x %>%
    filter(species == "SIMP") %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    filter(sampleType == "hair") %>%
    nrow(),
  NA,
  # NEGATIVES
  genos_species_md_10x %>%
    filter(sampleType %in% c("xtnNeg", "pcr1Neg")) %>%
    nrow(),
  genos_species_md_10x %>%
    filter(sampleType %in% c("xtnNeg", "pcr1Neg")) %>%
    filter(!is.na(speciesAssigned)) %>%
    nrow(),
  genos_species_md_10x %>%
    filter(sampleType %in% c("xtnNeg", "pcr1Neg")) %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    nrow(),
  NA, NA,
  genos_species_md_10x %>%
    filter(sampleType %in% c("xtnNeg", "pcr1Neg")) %>%
    filter(mdMatch_sp5 == "FALSE") %>%
    nrow()),
       nrow = 6,
       ncol = 3) %>%
       `rownames<-`(., c("totalSamples", "totalAssigned", "totalMismatch", "mismatchBlood", "mismatchHair", "mismatchNeg")) %>%
       `colnames<-`(., c("LWED", "SIMP", "(-)")) %>%
  as.data.frame() %>%
  mutate(TOTAL = rowSums(.[1:3], na.rm = T)) %>%
  t() %>%
  as.data.frame() %>%
  mutate(propMismatch = scales::percent(totalMismatch/totalAssigned)) %>%
  relocate(propMismatch, .after = totalMismatch) %>%
  mutate(propMismatch = gsub("100.00%", scales::percent(1/9), propMismatch))
  
speciesMismatch_summary_10x %>%
  kable(caption = "10x: Mismatched species assignments") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

#### Plate maps

Below are the tamRun5 plate maps with the species of each sample as per the metadata. Because all species loci are in primer pool 2, hair sample plates are are limited to those run with primer pool 2, which include plates 1, 3, and the second half of plate 5 (first half was run with primer pool 3, second half with primer pool 2).

Green samples matched metadata, red samples did NOT match metadata, and grey samples received no species assignment.

```{r}
condFormat_species <- genos_species_md_10x %>%
  mutate(
    species = case_when(
      mdMatch_sp5 == "FALSE" ~ cell_spec(species, background = "red"),
      mdMatch_sp5 == "TRUE" ~ cell_spec(species, background = "green"),
      is.na(mdMatch_sp5) ~ cell_spec(species, background = "grey")
    ))

plate1_species <- matrix(condFormat_species[1:96,]$species,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N701', 'N702', 'N703', 'N704', 'N705', 'N706', 'N707', 'N710', 'N711', 'N712', 'N714', 'N715'),.) %>%
  rbind(c(1, 9, 17, 25, 33, 41, 49, 57, 65, 73, 81, 89),.) %>%
  cbind(c("", "", 'S502', 'S503', 'S505', 'S506', 'S507', 'S508', 'S510', 'S511'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 1 (species/blood)", escape = FALSE) %>%
  kable_paper(full_width = F)

plate2_species <- matrix(condFormat_species[97:192,]$species,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N716',	'N718',	'N719',	'N720',	'N721',	'N722',	'N723',	'N724',	'N726',	'N727',	'N728', 'N729'),.) %>%
  rbind(c(97, 105, 113, 121, 129, 137, 145, 153, 161, 169, 177, 185),.) %>%
  cbind(c("", "", 'S502', 'S503', 'S505', 'S506', 'S507', 'S508', 'S510', 'S511'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 2 (species/blood)", escape = FALSE) %>%
  kable_paper(full_width = F)

plate3_species <- matrix(condFormat_species[193:288,]$species,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N701', 'N702', 'N703', 'N704', 'N705', 'N706'),.) %>%
  rbind(c(193,	201,	209,	217,	225,	233, 241, 249, 257, 265, 273, 281),.) %>%
  cbind(c("", "", 'S513', 'S515', 'S516', 'S517', 'S518', 'S520', 'S521', 'S522'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 3 (species/hair)", escape = FALSE) %>%
  kable_paper(full_width = F)

plate4_species <- matrix(condFormat_species[289:384,]$species,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N716',	'N718',	'N719',	'N720',	'N721',	'N722',	'N723',	'N724',	'N726',	'N727',	'N728', 'N729'),.) %>%
  rbind(c(289,	297,	305,	313,	321,	329, 337, 345, 353, 361, 369, 377),.) %>%
  cbind(c("", "", 'S513', 'S515', 'S516', 'S517', 'S518', 'S520', 'S521', 'S522'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 4 (species/hair)", escape = FALSE) %>%
  kable_paper(full_width = F)

plate5_species <- matrix(condFormat_species[385:440,]$species,
            nrow = 8,
            ncol = 7,
         byrow = F) %>%
  rbind(c("cust", "cust", "cust", "cust", "cust", "cust", "cust"),.) %>%
  rbind(c(385,	393,	401,	409,	417,	425, 433),.) %>%
  cbind(c("", "", "cust", "cust", "cust", "cust", "cust", "cust", "cust", "cust"),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 5 (species/hair+blood)", escape = FALSE) %>%
  kable_paper(full_width = F)
```

```{r, echo=FALSE}
plate1_species
plate2_species
plate3_species
plate4_species
plate5_species
```

#### Across sample types

None of the mismatches occurred in both blood and hair samples from the same animalID

```{r}
genos_species_mismatch %>%
  get_dupes(animalID)
```

#### Previous runs

We can check to see if any of the mismatched samples were included in tamRun1 and/or tamRun2 to see if the mismatches existed there as well. However, we see that **none of the 9 samples with mismatched species were included in previous runs.**

```{r, echo=FALSE}
# tamRun1 species assignments
genos_species_run1 <- genos_tamRun1 %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SPECIES")) %>%
  mutate(Locus = sub('[.][^.]+$', '', Locus)) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(speciesKey, by = c("Locus", "genotype")) %>%
  mutate(species = coalesce(species, genotype)) %>%
  select(c(Locus, sampleID, species)) %>%
  pivot_wider(names_from = sampleID,
              values_from = species) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  mutate(totalGenos_sp = (rowSums(. == "LWED") + rowSums(. == "SIMP"))) %>%
  mutate(propGenos_sp = (totalGenos_sp/12)) %>%
  mutate(propLWED = (rowSums(. == "LWED")/totalGenos_sp)) %>%
  mutate(propSIMP = (rowSums(. == "SIMP")/totalGenos_sp)) %>%
  mutate(propMismatch_species = (1 - abs(propLWED - propSIMP))) %>%
  mutate(speciesAssigned = ifelse(propLWED == 1, "LWED", ifelse(propSIMP == 1, "SIMP", NA))) %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = sub('_[^.]+', '', sampleID)) %>%
#  mutate(sampleID = sub("\\.", "_", sampleID)) %>%
  relocate(c(speciesAssigned, totalGenos_sp, propMismatch_species, propLWED, propSIMP, propGenos_sp), .after = sampleID)

# tamRun1 species match check
genos_species_md_tamRun1 <- md_tamRun1 %>%
  select(c(sampleID, sample, animalID, species, sampleType)) %>%
  merge(., genos_species_run1, by = "sampleID") %>%
  mutate(mdMatch_sp1 = ifelse(species == speciesAssigned,"TRUE", "FALSE")) %>%
  select(c(sample, mdMatch_sp1))

# tamRun2 species assignments
genos_species_run2 <- genos_tamRun2 %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SPECIES")) %>%
  mutate(Locus = sub('[.][^.]+$', '', Locus)) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(speciesKey, by = c("Locus", "genotype")) %>%
  mutate(species = coalesce(species, genotype)) %>%
  select(c(Locus, sampleID, species)) %>%
  pivot_wider(names_from = sampleID,
              values_from = species) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  mutate(totalGenos_sp = (rowSums(. == "LWED") + rowSums(. == "SIMP"))) %>%
  mutate(propGenos_sp = (totalGenos_sp/12)) %>%
  mutate(propLWED = (rowSums(. == "LWED")/totalGenos_sp)) %>%
  mutate(propSIMP = (rowSums(. == "SIMP")/totalGenos_sp)) %>%
  mutate(propMismatch_species = (1 - abs(propLWED - propSIMP))) %>%
  mutate(speciesAssigned = ifelse(propLWED == 1, "LWED", ifelse(propSIMP == 1, "SIMP", NA))) %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = sub('_[^.]+', '', sampleID)) %>%
#  mutate(sampleID = sub("\\.", "_", sampleID)) %>%
  relocate(c(speciesAssigned, totalGenos_sp, propMismatch_species, propLWED, propSIMP, propGenos_sp), .after = sampleID)

# tamRun2 species match check
genos_species_md_tamRun2 <- md_tamRun2 %>%
  select(c(sampleID, sample, animalID, species, sampleType)) %>%
  merge(., genos_species_run2, by = "sampleID") %>%
  mutate(mdMatch_sp2 = ifelse(species == speciesAssigned,"TRUE", "FALSE")) %>%
  select(c(sample, mdMatch_sp2))

speciesMismatch_tamRun125 <- genos_species_md %>%
  filter(mdMatch_sp5 == "FALSE") %>%
  filter(if_any(c(tamRun1, tamRun2), ~ !is.na(.))) %>%
  select(c(sampleID, sample, mdMatch_sp5)) %>%
  merge(., genos_species_md_tamRun1, by = "sample", all.x = T) %>%
  merge(., genos_species_md_tamRun2, by = "sample", all.x = T) %>%
  dplyr::rename("spMatch_tamRun1" = "mdMatch_sp1",
                "spMatch_tamRun2" = "mdMatch_sp2",
                "spMatch_tamRun5" = "mdMatch_sp5")

speciesMismatch_tamRun125 %>%
  kable(caption = "tamRun5 species mismatches vs. tamRun1/2 assignments") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### 30x coverage

#### Overview

Of the 329 samples that received species assignments, 8 samples (2.4% of those called) did NOT match the species listed in the metadata. For all mismatches, 100% of SPECIESID loci called were either LWED or SIMP, with no mixed species calls.

**Species mismatches in tamRun5**

```{r, echo=FALSE}
# Extract samples with mismatches
genos_species_mismatch_30x <- genos_species_md_30x %>%
  filter(!mdMatch_sp5 == "TRUE")

write.csv(genos_species_mismatch_30x, "./05_tamRun5/03_run5GTscore/summaryFiles/genos_species_mismatch_gtscore30x.csv", row.names = F)

# By sample type
speciesMismatchAll_30x <- genos_species_mismatch_30x %>%
#  group_by(seqRun) %>%
  summarise(total = length(sampleID))

speciesMismatchBlood_30x <- genos_species_mismatch_30x %>%
#  group_by(seqRun) %>%
  summarise(blood = sum(sampleType == "blood"))

speciesMismatchHair_30x <- genos_species_mismatch_30x %>%
#  group_by(seqRun) %>%
  summarise(hair = sum(sampleType == "hair"))

speciesMismatchNeg_30x <- genos_species_mismatch_30x %>%
#  group_by(seqRun) %>%
  summarise("(-)" = sum(!sampleType %in% c("blood", "hair")))

# How many mismatches for each data set?
speciesMismatchCount_df_30x <- cbind(speciesMismatchAll_30x, speciesMismatchBlood_30x, speciesMismatchHair_30x, speciesMismatchNeg_30x)

speciesMismatchCount_df_30x %>%
  kable(caption = "30x: Species mismatch overview") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### gtseq

#### Overview

```{r, echo=FALSE}
# Extract samples with mismatches
genos_species_mismatch_gtseq <- genos_species_md_gtseq %>%
  filter(!mdMatch_sp5 == "TRUE")

write.csv(genos_species_mismatch_gtseq, "./05_tamRun5/05_run5GTseq/summaryFiles/genos_species_mismatch_gtseq.csv", row.names = F)

# By sample type
speciesMismatchAll_gtseq <- genos_species_mismatch_gtseq %>%
#  group_by(seqRun) %>%
  summarise(total = length(sampleID))

speciesMismatchBlood_gtseq <- genos_species_mismatch_gtseq %>%
#  group_by(seqRun) %>%
  summarise(blood = sum(sampleType == "blood"))

speciesMismatchHair_gtseq <- genos_species_mismatch_gtseq %>%
#  group_by(seqRun) %>%
  summarise(hair = sum(sampleType == "hair"))

speciesMismatchNeg_gtseq <- genos_species_mismatch_gtseq %>%
#  group_by(seqRun) %>%
  summarise("(-)" = sum(!sampleType %in% c("blood", "hair")))

# How many mismatches for each data set?
speciesMismatchCount_df_gtseq <- cbind(speciesMismatchAll_gtseq, speciesMismatchBlood_gtseq, speciesMismatchHair_gtseq, speciesMismatchNeg_gtseq)

speciesMismatchCount_df_gtseq %>%
  kable(caption = "gtseq: Species mismatch overview") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### 10x vs. 30x vs. gtseq

```{r}
speciesMismatch_10x30x <- rbind(speciesMismatchCount_df_10x, speciesMismatchCount_df_30x) %>%
  mutate(
    coverage = c("mismatchAll_10x", "mismatchAll_30x")
  ) %>%
  column_to_rownames("coverage") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleType") %>%
  mutate(
    mismatchDiff = mismatchAll_10x - mismatchAll_30x
  )

speciesMismatch_10x30x %>%
  kable(caption = "Species mismatch overview: 10x - 30x") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

## 8.2 Sex mismatches

While we expect some sex mismatches for blood samples due to chimerism, all sex assignments for hair samples should match metadata.

**NOTE** that we have 11 SEXID total, with 4 general, 4 SIMP-specific, and 2 LWED-specific SEXID loci. These are split between primer pool 2 and 3 as follows:

-   7 SEXID loci are in primer pool 2:
    -   197 SIMP
    -   198 SIMP
    -   203 SIMP
    -   218 SIMP
    -   219
    -   222
-   5 SEXID loci are in primer pool 3:
    -   195 LWED
    -   205
    -   208 LWED
    -   209
    -   215
    
### 10x coverage

#### Overview

Out of the 440 samples, 330 were assigned a sex with 100% of loci in agreement. Of these 330, 61 samples did not match sex metadata, including 44 blood samples (26 complete mismatch, 18 mixed calls), 16 hair samples (3 complete mismatch, 13 mixed calls), and 1 negative (complete mismatch).

**Sex mismatches in tamRun5**

```{r}
# Extract samples with mismatches
genos_sex_mismatchAll_10x <- genos_sex_md_10x %>%
  filter(!mdMatch_sex == "TRUE")
genos_sex_mismatchBlood_10x <- genos_sex_mismatchAll_10x %>%
  filter(sampleType == "blood")
genos_sex_mismatchHair_10x <- genos_sex_mismatchAll_10x %>%
  filter(sampleType == "hair")
genos_sex_mismatchNeg_10x <- genos_sex_mismatchAll_10x %>%
  filter(str_detect(sampleType, "Neg"))

write.csv(genos_sex_mismatchAll_10x, "./05_tamRun5/03_run5GTscore/summaryFiles/genos_sex_mismatchAll_gtscore10x.csv", row.names = F)

# By sample type
sexMismatchAll_10x <- data.frame(
    mismatchAll = length(genos_sex_mismatchAll_10x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchAll_10x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchAll_10x$mdMatch_sex)),
    sampleType = "all"
  )

sexMismatchBlood_10x <- data.frame(
    mismatchAll = length(genos_sex_mismatchBlood_10x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchBlood_10x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchBlood_10x$mdMatch_sex)),
    sampleType = "blood"
  )

sexMismatchHair_10x <- data.frame(
    mismatchAll = length(genos_sex_mismatchHair_10x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchHair_10x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchHair_10x$mdMatch_sex)),
    sampleType = "hair"
  )

sexMismatchNeg_10x <- data.frame(
    mismatchAll = length(genos_sex_mismatchNeg_10x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchNeg_10x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchNeg_10x$mdMatch_sex)),
    sampleType = "neg"
  )

# How many mismatches for each data set?
sexMismatchCount_df_10x <- rbind(sexMismatchAll_10x, sexMismatchBlood_10x, sexMismatchHair_10x, sexMismatchNeg_10x) %>%
  relocate(sampleType)

sexMismatchCount_df_10x %>%
  kable(caption = "10x: Sex mismatch overview") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

**Sex mismatches in tamRun5 by sample type according to species listed in metadata**

```{r, echo=FALSE}
sexMismatch_summary_10x <- matrix(c(
  # FEMALES
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    filter(sampleType == "blood") %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    filter(sampleType == "hair") %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    filter(sampleType == "blood") %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "F") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    filter(sampleType == "hair") %>%
    nrow(),
  NA,
  # MALES
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    filter(sampleType == "blood") %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    filter(sampleType == "hair") %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    filter(sampleType == "blood") %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(sex == "M") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    filter(sampleType == "hair") %>%
    nrow(),
  NA,
  # NEGATIVES
  genos_sex_md_10x %>%
    filter(str_detect(sampleType, "Neg")) %>%
    nrow(),
  genos_sex_md_10x %>%
    filter(str_detect(sampleType, "Neg")) %>%
    filter(!is.na(sexAssigned)) %>%
    nrow(),
  NA, NA,
  genos_sex_md_10x %>%
    filter(str_detect(sampleType, "Neg")) %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  NA, NA,
  genos_sex_md_10x %>%
    filter(str_detect(sampleType, "Neg")) %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow()),
       nrow = 8,
       ncol = 3) %>%
       `rownames<-`(., c("totalSamples", "totalAssigned", "bloodAssigned", "hairAssigned", "totalMismatch", "bloodMismatch", "hairMismatch", "negMismatch")) %>%
       `colnames<-`(., c("F", "M", "(-)")) %>%
  as.data.frame() %>%
  mutate(TOTAL = 
           rowSums(.[1:3], na.rm = T)) %>%
  t() %>%
  as.data.frame() %>%
  mutate(propMismatch = 
           scales::percent(totalMismatch/totalAssigned)) %>%
  mutate(propMismatch_blood = 
           scales::percent(bloodMismatch/bloodAssigned)) %>%
  mutate(propMismatch_hair = 
           scales::percent(hairMismatch/hairAssigned)) %>%
  mutate(propMismatch = gsub("100.0%", scales::percent(1/9), propMismatch)) %>%
  select(c(totalSamples, totalAssigned, totalMismatch, propMismatch,
           bloodAssigned, bloodMismatch, propMismatch_blood,
           hairAssigned, hairMismatch, propMismatch_hair,
           negMismatch)) %>%
  rownames_to_column("Metadata assignment")
  
sexMismatch_summary_10x %>%
  kable(caption = "Mismatched sex assignments") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

**Mismatch type (F, M, or MIX)**

```{r}
sexMismatch_types <- matrix(c(
  # FEMALES (general)
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(),
  # FEMALES (blood)
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(sampleType == "blood") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(sampleType == "blood") %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(sampleType == "blood") %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(),
  # FEMALES (hair)
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(sampleType == "hair") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(sampleType == "hair") %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "F") %>%
    filter(sampleType == "hair") %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(),
  # MALES (general)
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(),
  # MALES (blood)
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(sampleType == "blood") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(sampleType == "blood") %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(sampleType == "blood") %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(),
  # MALES (hair)
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(sampleType == "hair") %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(sampleType == "hair") %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(sex == "M") %>%
    filter(sampleType == "hair") %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(), 
  # NEGATIVES
  genos_sex_md %>%
    filter(str_detect(sampleType, "Neg")) %>%
    filter(!mdMatch_sex == "TRUE") %>%
    nrow(),
  genos_sex_md %>%
    filter(str_detect(sampleType, "Neg")) %>%
    filter(mdMatch_sex == "FALSE") %>%
    nrow(),
  genos_sex_md %>%
    filter(str_detect(sampleType, "Neg")) %>%
    filter(mdMatch_sex == "MIX") %>%
    nrow(), 
  NA, NA, NA, NA, NA, NA
  ),
  nrow = 9, 
  ncol = 3
) %>%
       `rownames<-`(., c("totalMismatch", "mismatchFALSE", "mismatchMIX", "bloodMismatch", "bloodMismatch_FALSE", "bloodMismatch_MIX", "hairMismatch", "hairMismatch_FALSE", "hairMismatch_MIX")) %>%
       `colnames<-`(., c("F", "M", "(-)")) %>%
  as.data.frame() %>%
  mutate(TOTAL = 
           rowSums(.[1:3], na.rm = T)) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Metadata assignment")
  
sexMismatch_types %>%
  kable(caption = "Mismatch types") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

#### Compare species sets

If we look at sex assigned for genotypes in the species-specific sex loci set for the OTHER species, 4 of the originally mismatched samples end up matching metadata -- HOWEVER, this isn't b/c they're being genotyped correctly with the other species sex loci; rather, it's b/c the mismatches were in their own species-specific sex loci.

Get sex assignments for LWED and SIMP-specific sex loci sets:

```{r}
genos_sex_spNeutral <- genos_tamRun5 %>%
  rownames_to_column("locus") %>%
  filter(str_detect(locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey, by = c("locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  # assign sex
  mutate(totalGenos_sex = (rowSums(. == "F", na.rm = T) + rowSums(. == "M", na.rm = T))) %>%
  mutate(propGenos_sex = totalGenos_sex/11) %>%
  mutate(propF = (rowSums(. == "F", na.rm = T)/totalGenos_sex)) %>%
  mutate(propM = (rowSums(. == "M", na.rm = T)/totalGenos_sex)) %>%
  mutate(propMismatch_sex = (1 - abs(propF - propM))) %>%
  mutate(sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  select(c("sampleID",
           "SEXID_200",
           "SEXID_205",
           "SEXID_209",
           "SEXID_215",
           "SEXID_222",
           # LWED-specific SEXID
           "SEXID_LWED_195",
           "SEXID_LWED_208",
           # SIMP-specific SEXID
           "SEXID_SIMP_197",
           "SEXID_SIMP_198",
           "SEXID_SIMP_203",
           "SEXID_SIMP_218"))

genos_sex_spLWED <- genos_tamRun5 %>%
  rownames_to_column("locus") %>%
  filter(str_detect(locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey, by = c("locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun5[, c("sampleID", "species")], by = "sampleID") %>%
  column_to_rownames("sampleID") %>%
  # remove SIMP-specific sex loci
  mutate(SEXID_SIMP_197 = NA,
         SEXID_SIMP_198 = NA,
         SEXID_SIMP_203 = NA,
         SEXID_SIMP_218 = NA) %>%
  # assign sex
  mutate(
    totalGenos_sex = (rowSums(. == "F", na.rm = T) + rowSums(. == "M", na.rm = T)),
    propGenos_sex = totalGenos_sex/7,
    propF = (rowSums(. == "F", na.rm = T)/totalGenos_sex),
    propM = (rowSums(. == "M", na.rm = T)/totalGenos_sex),
    propMismatch_sex = (1 - abs(propF - propM)),
    sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  select(c("sampleID",
           "sexAssigned",
           "totalGenos_sex",
           "propGenos_sex",
           "propMismatch_sex",
           "propF",
           "propM")) %>%
  dplyr::rename(
    "sexAssigned_spLWED" = "sexAssigned",
    "totalGenos_spLWED" = "totalGenos_sex",
    "propGenos_sex_spLWED" = "propGenos_sex",
    "propMismatch_sex_spLWED" = "propMismatch_sex",
    "propF_spLWED" = "propF",
    "propM_spLWED" = "propM"
  )

genos_sex_spSIMP <- genos_tamRun5 %>%
  rownames_to_column("locus") %>%
  filter(str_detect(locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey, by = c("locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun5[, c("sampleID", "species")], by = "sampleID") %>%
  column_to_rownames("sampleID") %>%
  # remove SIMP-specific sex loci
  mutate(SEXID_LWED_195 = NA,
         SEXID_LWED_208 = NA) %>%
  # assign sex
  mutate(
    totalGenos_sex = (rowSums(. == "F", na.rm = T) + rowSums(. == "M", na.rm = T)),
    propGenos_sex = totalGenos_sex/9,
    propF = (rowSums(. == "F", na.rm = T)/totalGenos_sex),
    propM = (rowSums(. == "M", na.rm = T)/totalGenos_sex),
    propMismatch_sex = (1 - abs(propF - propM)),
    sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  select(c("sampleID",
           "sexAssigned",
           "totalGenos_sex",
           "propGenos_sex",
           "propMismatch_sex",
           "propF",
           "propM")) %>%
  dplyr::rename(
    "sexAssigned_spSIMP" = "sexAssigned",
    "totalGenos_sex_spSIMP" = "totalGenos_sex",
    "propGenos_sex_spSIMP" = "propGenos_sex",
    "propMismatch_sex_spSIMP" = "propMismatch_sex",
    "propF_spSIMP" = "propF",
    "propM_spSIMP" = "propM"
  )
```

Do species that turn up as mismatched match up with different species sets?

```{r}
genos_sex_mismatch_bySpeciesSet <- genos_sex_mismatchAll %>%
  select(c(
    "sampleID",
    "sample",
    "sampleType",
    "species",
    "mdMatch_sex",
    "sex",
    "sexAssigned",
    "totalGenos_sex",
    "propGenos_sex",
    "propMismatch_sex",
    "propF",
    "propM"
  )) %>%
  merge(., genos_sex_spNeutral, by = "sampleID") %>%
  merge(., genos_sex_spLWED, by = "sampleID") %>%
  merge(., genos_sex_spSIMP, by = "sampleID") %>%
  relocate(c(sexAssigned_spLWED, sexAssigned_spSIMP), .after = sexAssigned) %>%
  mutate(mdMatch_sex_spLWED = 
           case_when(
             sexAssigned_spLWED == "MIX" ~ "MIX",
             sex == sexAssigned_spLWED ~ "TRUE",
             sex != sexAssigned_spLWED ~ "FALSE",
             is.na(sex) & !is.na(sexAssigned_spLWED) ~ "FALSE"),
         mdMatch_sex_spSIMP = 
           case_when(
             sexAssigned_spSIMP == "MIX" ~ "MIX",
             sex == sexAssigned_spSIMP ~ "TRUE",
             sex != sexAssigned_spSIMP ~ "FALSE",
             is.na(sex) & !is.na(sexAssigned_spSIMP) ~ "FALSE")) %>%
  relocate(c(mdMatch_sex_spLWED, mdMatch_sex_spSIMP), .after = mdMatch_sex)

genos_sex_mismatchBlood_bySpeciesSet <- genos_sex_mismatch_bySpeciesSet %>%
  filter(sampleType == "blood")

genos_sex_mismatchHair_bySpeciesSet <- genos_sex_mismatch_bySpeciesSet %>%
  filter(sampleType == "hair")
```

#### Loci-specific (mis)matches

Are there particular loci that are resulting in mismatches?

meh, doesn't seem like it - mismatch prop ranges from 30-66%; but also, if there was an issue it should be popping up with more samples no?

```{r}
genos_sex_mismatchHair_byLocus <- genos_sex_mismatchHair %>%
  select(c("sampleID",
           "sex",
           "SEXID_200",
           "SEXID_205",
           "SEXID_209",
           "SEXID_215",
           "SEXID_222",
           # LWED-specific SEXID
           "SEXID_LWED_195",
           "SEXID_LWED_208",
           # SIMP-specific SEXID
           "SEXID_SIMP_197",
           "SEXID_SIMP_198",
           "SEXID_SIMP_203",
           "SEXID_SIMP_218")) %>%
  mutate_all(., as.character()) %>%
  mutate(
    SEXID_200 = case_when(is.na(SEXID_200) ~ NA,
                          SEXID_200 == "0" ~ NA,
                          SEXID_200 == sex ~ "0",
                          .default = "1"),
    SEXID_205 = case_when(is.na(SEXID_205) ~ NA,
                          SEXID_205 == "0" ~ NA,
                          SEXID_205 == sex ~ "0",
                          .default = "1"),
    SEXID_209 = case_when(is.na(SEXID_209) ~ NA,
                          SEXID_209 == "0" ~ NA,
                          SEXID_209 == sex ~ "0",
                          .default = "1"),
    SEXID_215 = case_when(is.na(SEXID_215) ~ NA,
                          SEXID_215 == "0" ~ NA,
                          SEXID_215 == sex ~ "0",
                          .default = "1"),
    SEXID_222 = case_when(is.na(SEXID_222) ~ NA,
                          SEXID_222 == "0" ~ NA,
                          SEXID_222 == sex ~ "0",
                          .default = "1"),
    SEXID_LWED_195 = case_when(is.na(SEXID_LWED_195) ~ NA,
                          SEXID_LWED_195 == "0" ~ NA,
                          SEXID_LWED_195 == sex ~ "0",
                          .default = "1"),
    SEXID_LWED_208 = case_when(is.na(SEXID_LWED_208) ~ NA,
                          SEXID_LWED_208 == "0" ~ NA,
                          SEXID_LWED_208 == sex ~ "0",
                          .default = "1"),
    SEXID_SIMP_197 = case_when(is.na(SEXID_SIMP_197) ~ NA,
                          SEXID_SIMP_197 == "0" ~ NA,
                          SEXID_SIMP_197 == sex ~ "0",
                          .default = "1"),
    SEXID_SIMP_198 = case_when(is.na(SEXID_SIMP_198) ~ NA,
                          SEXID_SIMP_198 == "0" ~ NA,
                          SEXID_SIMP_198 == sex ~ "0",
                          .default = "1"),
    SEXID_SIMP_203 = case_when(is.na(SEXID_SIMP_203) ~ NA,
                          SEXID_SIMP_203 == "0" ~ NA,
                          SEXID_SIMP_203 == sex ~ "0",
                          .default = "1"),
    SEXID_SIMP_218 = case_when(is.na(SEXID_SIMP_218) ~ NA,
                          SEXID_SIMP_218 == "0" ~ NA,
                          SEXID_SIMP_218 == sex ~ "0",
                          .default = "1")
  ) %>%
  select(-sampleID, -sex) %>%
  t() %>%
  as.data.frame() %>%
  mutate(
    totalMismatch = rowSums(. == 1, na.rm = T),
    totalMatch = rowSums(. == 0, na.rm = T),
    propMismatch = totalMismatch/(totalMatch + totalMismatch))
```

#### Plate maps

```{r}
condFormat_sex_pp1 <- genos_sex_md %>%
  mutate(
    sex = case_when(
      mdMatch_sex == "TRUE" ~ cell_spec(sex, background = "green"),
      mdMatch_sex == "FALSE" ~ cell_spec(sex, background = "red"),
      mdMatch_sex == "MIX" ~ cell_spec(sex, background = "yellow"),
      is.na(mdMatch_sex) ~ cell_spec(sex, background = "grey")
    ))

# condFormat_sex_pp2 <- genos_sex_md %>%
#  mutate(
#    sex = case_when(
#      mdMatch_sex_pp2 == "TRUE" ~ cell_spec(sex, background = "green"),
#      mdMatch_sex_pp2 == "FALSE" ~ cell_spec(sex, background = "red"),
#      mdMatch_sex_pp2 == "MIX" ~ cell_spec(sex, background = "yellow"),
#      is.na(mdMatch_sex_pp2) ~ cell_spec(sex, background = "grey")
#    ))

#condFormat_sex_pp3 <- genos_sex_md %>%
#  mutate(
#    sex = case_when(
#      mdMatch_sex_pp3 == "TRUE" ~ cell_spec(sex, background = "green"),
#      mdMatch_sex_pp3 == "FALSE" ~ cell_spec(sex, background = "red"),
#      mdMatch_sex_pp3 == "MIX" ~ cell_spec(sex, background = "yellow"),
#      is.na(mdMatch_sex_pp3) ~ cell_spec(sex, background = "grey")
#    ))

plate1_sex <- matrix(condFormat_sex_pp1[1:96,]$sex,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N701', 'N702', 'N703', 'N704', 'N705', 'N706', 'N707', 'N710', 'N711', 'N712', 'N714', 'N715'),.) %>%
  rbind(c(1, 9, 17, 25, 33, 41, 49, 57, 65, 73, 81, 89),.) %>%
  cbind(c("", "", 'S502', 'S503', 'S505', 'S506', 'S507', 'S508', 'S510', 'S511'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 1 (sex/blood)", escape = FALSE) %>%
  kable_paper(full_width = F)

plate2_sex <- matrix(condFormat_sex_pp1[97:192,]$sex,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N716',	'N718',	'N719',	'N720',	'N721',	'N722',	'N723',	'N724',	'N726',	'N727',	'N728', 'N729'),.) %>%
  rbind(c(97, 105, 113, 121, 129, 137, 145, 153, 161, 169, 177, 185),.) %>%
  cbind(c("", "", 'S502', 'S503', 'S505', 'S506', 'S507', 'S508', 'S510', 'S511'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 2 (sex/blood)", escape = FALSE) %>%
  kable_paper(full_width = F)

plate3a_sex <- matrix(condFormat_sex_pp1[193:288,]$sex,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N701', 'N702', 'N703', 'N704', 'N705', 'N706'),.) %>%
  rbind(c(193,	201,	209,	217,	225,	233, 241, 249, 257, 265, 273, 281),.) %>%
  cbind(c("", "", 'S513', 'S515', 'S516', 'S517', 'S518', 'S520', 'S521', 'S522'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 3a (sex/hair)", escape = FALSE) %>%
  kable_paper(full_width = F)

#plate3b_sex <- matrix(condFormat_sex_pp3[193:288,]$sex,
#            nrow = 8,
#            ncol = 12,
#         byrow = F) %>%
#  rbind(c('N701', 'N702', 'N703', 'N704', 'N705', 'N706'),.) %>%
#  rbind(c(193,	201,	209,	217,	225,	233, 241, 249, 257, 265, 273, 281),.) %>%
#  cbind(c("", "", 'S513', 'S515', 'S516', 'S517', 'S518', 'S520', 'S521', 'S522'),.) %>%
#  cbind(c("", "", 1:8),.) %>%
#  kable(caption = "Plate 3b (sex/hair)", escape = FALSE) %>%
#  kable_paper(full_width = F)

plate4a_sex <- matrix(condFormat_sex_pp1[289:384,]$sex,
            nrow = 8,
            ncol = 12,
         byrow = F) %>%
  rbind(c('N716',	'N718',	'N719',	'N720',	'N721',	'N722',	'N723',	'N724',	'N726',	'N727',	'N728', 'N729'),.) %>%
  rbind(c(289,	297,	305,	313,	321,	329, 337, 345, 353, 361, 369, 377),.) %>%
  cbind(c("", "", 'S513', 'S515', 'S516', 'S517', 'S518', 'S520', 'S521', 'S522'),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 4a (sex/hair)", escape = FALSE) %>%
  kable_paper(full_width = F)

#plate4b_sex <- matrix(condFormat_sex_pp3[289:384,]$sex,
#            nrow = 8,
#            ncol = 12,
#         byrow = F) %>%
#  rbind(c('N716',	'N718',	'N719',	'N720',	'N721',	'N722',	'N723',	'N724',	'N726',	'N727',	'N728', 'N729'),.) %>%
#  rbind(c(289,	297,	305,	313,	321,	329, 337, 345, 353, 361, 369, 377),.) %>%
#  cbind(c("", "", 'S513', 'S515', 'S516', 'S517', 'S518', 'S520', 'S521', 'S522'),.) %>%
#  cbind(c("", "", 1:8),.) %>%
#  kable(caption = "Plate 4b (sex/hair)", escape = FALSE) %>%
#  kable_paper(full_width = F)

plate5a_sex <- matrix(condFormat_sex_pp1[385:440,]$sex,
            nrow = 8,
            ncol = 7,
         byrow = F) %>%
  rbind(c("cust", "cust", "cust", "cust", "cust", "cust", "cust"),.) %>%
  rbind(c(385,	393,	401,	409,	417,	425, 433),.) %>%
  cbind(c("", "", "cust", "cust", "cust", "cust", "cust", "cust", "cust", "cust"),.) %>%
  cbind(c("", "", 1:8),.) %>%
  kable(caption = "Plate 5a (sex/hair+blood)", escape = FALSE) %>%
  kable_paper(full_width = F)

#plate5b_sex <- matrix(condFormat_sex_pp3[385:440,]$sex,
#            nrow = 8,
#            ncol = 7,
#         byrow = F) %>%
#  rbind(c("cust", "cust", "cust", "cust", "cust", "cust", "cust"),.) %>%
#  rbind(c(385,	393,	401,	409,	417,	425, 433),.) %>%
#  cbind(c("", "", "cust", "cust", "cust", "cust", "cust", "cust", "cust", "cust"),.) %>%
#  cbind(c("", "", 1:8),.) %>%
#  kable(caption = "Plate 5b (sex/hair)", escape = FALSE) %>%
#  kable_paper(full_width = F)
```

```{r}
plate1_sex
plate2_sex
plate3a_sex
#plate3b_sex
plate4a_sex
#plate4b_sex
plate5a_sex
#plate5b_sex
```

#### Previous runs

Looking at hair samples only, we can check to see if any of the mismatched samples were included in tamRun1 and/or tamRun2 to see if the mismatches existed there as well.

In total, 4 of the tamRun5 samples with species mismatches were run in either tamRun1 or tamRun2, and 2 of these (hair_167 & hair_55) received species assignments in the previous runs BUT these different species assignments were only for 1 locus apiece (1/6 for hair_167 and 1/1 for hair_55). **None of these four were mismatched in previous runs.**

```{r, echo=FALSE}
genos_tamRun125_mismatchSex_hair <- genos_tamRun125 %>%
  filter(sample %in% genos_sex_mismatch$sample) %>%
  filter(str_detect(sample, "hair")) %>%
  filter(str_detect(locus, "SEX")) %>%
  filter(run5Geno1 != 0) %>%
  filter(genosCalled_125 > 1) %>%
  # recode zeros to NA
  replace(., .==0, NA) %>%
  # ditch columns that are completely NA
  .[, colSums(is.na(.)) < nrow(.)] %>%
#  merge(., sexKey, by.x = c("locus", "run1Geno1"), by.y = c("locus", "genotype"), all.x = T) %>%
#  dplyr::rename("sexRun1a" = "sex") %>%
#  merge(., sexKey, by.x = c("locus", "run1Geno2"), by.y = c("Locus", "genotype"), all.x = T) %>%
#  dplyr::rename("sexRun1b" = "sex") %>%
  merge(., sexKey, by.x = c("locus", "run2Geno1"), by.y = c("locus", "genotype"), all.x = T) %>%
  dplyr::rename("sexRun2a" = "sex") %>%
  merge(., sexKey, by.x = c("locus", "run2Geno2"), by.y = c("locus", "genotype"), all.x = T) %>%
  dplyr::rename("sexRun2b" = "sex") %>%
  merge(., sexKey, by.x = c("locus", "run2Geno3"), by.y = c("locus", "genotype"), all.x = T) %>%
  dplyr::rename("sexRun2c" = "sex") %>%
  merge(., sexKey, by.x = c("locus", "run2Geno4"), by.y = c("locus", "genotype"), all.x = T) %>%
  dplyr::rename("sexRun2d" = "sex") %>%
#  merge(., sexKey, by.x = c("locus", "run2Geno5"), by.y = c("locus", "genotype"), all.x = T) %>%
#  dplyr::rename("sexRun2e" = "sex") %>%
#  merge(., sexKey, by.x = c("locus", "run2Geno6"), by.y = c("locus", "genotype"), all.x = T) %>%
#  dplyr::rename("sexRun2f" = "sex") %>%
  merge(., sexKey, by.x = c("locus", "run5Geno1"), by.y = c("locus", "genotype"), all.x = T) %>%
  dplyr::rename("sexRun5" = "sex") %>%
  select(!contains(c("geno","zeros", "set"))) %>%
  mutate(totalAssigned = rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  filter(totalAssigned > 1) %>%
  rowwise() %>%
  mutate(uniqueSex = n_distinct(c_across(sexRun2a:sexRun5), na.rm = T)) %>%
  select(sample, locus, sexRun2a, sexRun2b, sexRun2c, sexRun2d, sexRun5, totalAssigned, uniqueSex)

# Number of samples with sex genotypes in tamRun3 and tamRun1/2
length(unique(genos_tamRun125_mismatchSex_hair$sample))
```

### 30x coverage

#### Overview

Out of the 440 samples, 330 were assigned a sex with 100% of loci in agreement. Of these 330, 61 samples did not match sex metadata, including 44 blood samples (26 complete mismatch, 18 mixed calls), 16 hair samples (3 complete mismatch, 13 mixed calls), and 1 negative (complete mismatch).

**Sex mismatches in tamRun5**

```{r}
# Extract samples with mismatches
genos_sex_mismatchAll_30x <- genos_sex_md_30x %>%
  filter(!mdMatch_sex == "TRUE")
genos_sex_mismatchBlood_30x <- genos_sex_mismatchAll_30x %>%
  filter(sampleType == "blood")
genos_sex_mismatchHair_30x <- genos_sex_mismatchAll_30x %>%
  filter(sampleType == "hair")
genos_sex_mismatchNeg_30x <- genos_sex_mismatchAll_30x %>%
  filter(str_detect(sampleType, "Neg"))

write.csv(genos_sex_mismatchAll_30x, "./05_tamRun5/03_run5GTscore/summaryFiles/genos_sex_mismatchAll_gtscore30x.csv", row.names = F)

# By sample type
sexMismatchAll_30x <- data.frame(
    mismatchAll = length(genos_sex_mismatchAll_30x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchAll_30x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchAll_30x$mdMatch_sex)),
    sampleType = "all"
  )

sexMismatchBlood_30x <- data.frame(
    mismatchAll = length(genos_sex_mismatchBlood_30x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchBlood_30x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchBlood_30x$mdMatch_sex)),
    sampleType = "blood"
  )

sexMismatchHair_30x <- data.frame(
    mismatchAll = length(genos_sex_mismatchHair_30x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchHair_30x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchHair_30x$mdMatch_sex)),
    sampleType = "hair"
  )

sexMismatchNeg_30x <- data.frame(
    mismatchAll = length(genos_sex_mismatchNeg_30x$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchNeg_30x$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchNeg_30x$mdMatch_sex)),
    sampleType = "neg"
  )

# How many mismatches for each data set?
sexMismatchCount_df_30x <- rbind(sexMismatchAll_30x, sexMismatchBlood_30x, sexMismatchHair_30x, sexMismatchNeg_30x) %>%
  relocate(sampleType)

sexMismatchCount_df_30x %>%
  kable(caption = "30x: Sex mismatch overview") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### gtseq

#### Overview

```{r}
# Extract samples with mismatches
genos_sex_mismatchAll_gtseq <- genos_sex_md_gtseq %>%
  filter(!mdMatch_sex == "TRUE")
genos_sex_mismatchBlood_gtseq <- genos_sex_mismatchAll_gtseq %>%
  filter(sampleType == "blood")
genos_sex_mismatchHair_gtseq <- genos_sex_mismatchAll_gtseq %>%
  filter(sampleType == "hair")
genos_sex_mismatchNeg_gtseq <- genos_sex_mismatchAll_gtseq %>%
  filter(str_detect(sampleType, "Neg"))

write.csv(genos_sex_mismatchAll_gtseq, "./05_tamRun5/05_run5GTseq/summaryFiles/genos_sex_mismatchAll_gtseq.csv", row.names = F)

# By sample type
sexMismatchAll_gtseq <- data.frame(
    mismatchAll = length(genos_sex_mismatchAll_gtseq$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchAll_gtseq$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchAll_gtseq$mdMatch_sex)),
    sampleType = "all"
  )

sexMismatchBlood_gtseq <- data.frame(
    mismatchAll = length(genos_sex_mismatchBlood_gtseq$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchBlood_gtseq$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchBlood_gtseq$mdMatch_sex)),
    sampleType = "blood"
  )

sexMismatchHair_gtseq <- data.frame(
    mismatchAll = length(genos_sex_mismatchHair_gtseq$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchHair_gtseq$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchHair_gtseq$mdMatch_sex)),
    sampleType = "hair"
  )

sexMismatchNeg_gtseq <- data.frame(
    mismatchAll = length(genos_sex_mismatchNeg_gtseq$sampleID),
    mismatchComplete = sum(grepl("FALSE", genos_sex_mismatchNeg_gtseq$mdMatch_sex)),
    mismatchMix = sum(grepl("MIX", genos_sex_mismatchNeg_gtseq$mdMatch_sex)),
    sampleType = "neg"
  )

# How many mismatches for each data set?
sexMismatchCount_df_gtseq <- rbind(sexMismatchAll_gtseq, sexMismatchBlood_gtseq, sexMismatchHair_gtseq, sexMismatchNeg_gtseq) %>%
  relocate(sampleType)

sexMismatchCount_df_gtseq %>%
  kable(caption = "gtseq: Sex mismatch overview") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

### 10x vs. 30x vs. gtseq

```{r}
sexMismatch_10x30x <- merge(sexMismatchCount_df_10x, sexMismatchCount_df_30x, by = "sampleType") %>%
  dplyr::rename(
    "mismatchAll_10x" = "mismatchAll.x",
    "mismatchComplete_10x" = "mismatchComplete.x",
    "mismatchMix_10x" = "mismatchMix.x",
    "mismatchAll_30x" = "mismatchAll.y",
    "mismatchComplete_30x" = "mismatchComplete.y",
    "mismatchMix_30x" = "mismatchMix.y"
  ) %>%
  mutate(
    mismatchAll_diff = mismatchAll_10x - mismatchAll_30x,
    mismatchComplete_diff = mismatchComplete_10x - mismatchComplete_30x,
    mismatchMix_diff = mismatchMix_10x - mismatchMix_30x
  ) %>%
  select(order(colnames(.))) %>%
  relocate(sampleType)

sexMismatch_10x30x %>%
  kable(caption = "Sex mismatch overview: 10x - 30x") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

## 8.3 Species/sex mismatch comparison

```{r}
genos_spSex <- genos_species_md %>%
  select(c(sampleID, sampleType, species, speciesAssigned, mdMatch_sp5)) %>%
  merge(., genos_sex_md[, c("sampleID", "mdMatch_sex", "sex", "sexAssigned")], by = "sampleID") %>%
  select(c(sampleID, sampleType, mdMatch_sp5, mdMatch_sex, species, speciesAssigned, sex, sexAssigned))

genos_spSex_hair <- genos_spSex %>%
  filter(sampleType == "hair")
```

### Species-specific sex loci - Mismatches

For the 20 samples with a species mismatch, two samples (both SIMP; cat_tamRun3.369 and cat_tamRun3.371) were genotyped at SEXID_LWED_208 (but not SEXID_LWED_195).

```{r}
genos_spSex_spMismatch <- genos_spSex %>%
  filter(str_detect(sampleID, "cat_")) %>%
  filter(mdMatch_sp5 == "FALSE")

matrix(c(sum(genos_spSex_spMismatch$speciesAssigned == "LWED"),
         genos_spSex_spMismatch %>%
           filter(speciesAssigned == "LWED") %>%
           filter(lwedSex > 0) %>%
           nrow(),
         genos_spSex_spMismatch %>%
           filter(speciesAssigned == "LWED") %>%
           filter(simpSex > 0) %>%
           nrow(),
       sum(genos_spSex_spMismatch$speciesAssigned == "SIMP"),
       genos_spSex_spMismatch %>%
           filter(speciesAssigned == "SIMP") %>%
           filter(lwedSex > 0) %>%
           nrow(),
       genos_spSex_spMismatch %>%
           filter(speciesAssigned == "SIMP") %>%
           filter(simpSex > 0) %>%
           nrow()),
       ncol = 2,
       nrow = 3) %>%
       `rownames<-`(., c("totalMismatch", "lwedSex", "simpSex")) %>%
       `colnames<-`(., c("lwedAssigned", "simpAssigned"))

simpAssigned_mismatch_lwedSex_sampleIDs <- genos_spSex_spMismatch %>%
  filter(speciesAssigned == "SIMP") %>%
  filter(lwedSex > 0)

genos_sex_md %>%
  filter(sampleID %in% simpAssigned_mismatch_lwedSex_sampleIDs$sampleID) %>%
  select(., c(sampleID, sample, sampleType, species, sex, contains("LWED")))
```

### Species-specific sex loci: Match samples

If we look at samples whose species DOES match the metadata though, there are quite a few samples that have genotypes for species-specific sex loci from the other species. Of the 67 samples genotyped at species-specific sex loci NOT for their species, 66 were SIMP samples and 1 was LWED.

```{r}
genos_spSex_spMatch <- genos_spSex %>%
  filter(str_detect(sampleID, "cat_")) %>%
  filter(mdMatch_sp5 == "TRUE")

matrix(c(sum(genos_spSex_spMatch$speciesAssigned == "LWED"),
         genos_spSex_spMatch %>%
           filter(speciesAssigned == "LWED") %>%
           filter(lwedSex > 0) %>%
           nrow(),
         genos_spSex_spMatch %>%
           filter(speciesAssigned == "LWED") %>%
           filter(simpSex > 0) %>%
           nrow(),
       sum(genos_spSex_spMatch$speciesAssigned == "SIMP"),
       genos_spSex_spMatch %>%
           filter(speciesAssigned == "SIMP") %>%
           filter(lwedSex > 0) %>%
           nrow(),
       genos_spSex_spMismatch %>%
           filter(speciesAssigned == "SIMP") %>%
           filter(simpSex > 0) %>%
           nrow()),
       ncol = 2,
       nrow = 3) %>%
       `rownames<-`(., c("totalMatch", "lwedSex", "simpSex")) %>%
       `colnames<-`(., c("lwedAssigned", "simpAssigned"))
```

SEXID_SIMP_218 was genotyped once for LWED-assigned samples, while SEXID_LWED_195 was genotyped 3 times (F: n = 0, M: n = 3) and SEXID_LWED_208 66 times (F: n = 65, M: n = 1) for SIMP-assigned samples.

```{r}
lwedAssigned_match_simpSex <- genos_spSex_spMatch %>%
  filter(speciesAssigned == "LWED") %>%
  filter(simpSex > 0) %>%
  select(sampleID) %>%
  merge(., genos_sex_md %>%
          select(sampleID, sample, sampleType, species, sex, contains("SIMP")),
        by = "sampleID")

simpAssigned_match_lwedSex <- genos_spSex_spMatch %>%
  filter(speciesAssigned == "SIMP") %>%
  filter(lwedSex > 0) %>%
  select(sampleID) %>%
  merge(., genos_sex_md %>%
          select(sampleID, sample, sampleType, species, sex, contains("LWED")),
        by = "sampleID")

matrix(c(nrow(lwedAssigned_match_simpSex),
         NA, NA,
         lwedAssigned_match_simpSex %>%
           filter(SEXID_SIMP_197 %in% c("F", "M")) %>%
           nrow(),
         lwedAssigned_match_simpSex %>%
           filter(SEXID_SIMP_198 %in% c("F", "M")) %>%
           nrow(),
         lwedAssigned_match_simpSex %>%
           filter(SEXID_SIMP_203 %in% c("F", "M")) %>%
           nrow(),
         lwedAssigned_match_simpSex %>%
           filter(SEXID_SIMP_218 %in% c("F", "M")) %>%
           nrow(),
         nrow(simpAssigned_match_lwedSex),
         simpAssigned_match_lwedSex %>%
           filter(SEXID_LWED_195 %in% c("F", "M")) %>%
           nrow(),
         simpAssigned_match_lwedSex %>%
           filter(SEXID_LWED_208 %in% c("F", "M")) %>%
           nrow(),
         NA, NA, NA, NA),
       nrow = 7,
       ncol = 2,
       byrow = F) %>%
       `rownames<-`(., c("totalSamples", "SEXID_LWED_195", "SEXID_LWED_208", "SEXID_SIMP_197", "SEXID_SIMP_198", "SEXID_SIMP_203", "SEXID_SIMP_218")) %>%
       `colnames<-`(., c("lwedAssigned", "simpAssigned"))
```

# X 10x vs. 30x vs. gtseq

```{r}
summary_10x30xgtseq <- data.frame(
  # avg sample geno rate
  avg_sampleGenoRate = c(mean(sampleSum_10x$GenotypeRate), mean(sampleSum_30x$GenotypeRate), mean(sampleSum_gtseq$GenotypeRate)),
  # avg BLOOD sample geno rate
  avg_sampleGenoRate_blood = c(mean(sampleSum_blood_10x$GenotypeRate), mean(sampleSum_blood_30x$GenotypeRate), mean(sampleSum_blood_gtseq$GenotypeRate)),
  # avg HAIR sample geno rate
  avg_sampleGenoRate_hair = c(mean(sampleSum_hair_10x$GenotypeRate), mean(sampleSum_hair_30x$GenotypeRate), mean(sampleSum_hair_gtseq$GenotypeRate)),
  # number of species assignments
  spAssignments = c(speciesCount_df_10x$totalAll_10x, speciesCount_df_30x$totalAll_30x, speciesCount_df_gtseq$totalAll_gtseq),
  # proportion of species mismatches
  spMismatches = c(speciesMismatchCount_df_10x$total, speciesMismatchCount_df_30x$total, speciesMismatchCount_df_gtseq$total),
  # number of sex assignments
  sexAssignments = c(sexCount_df_10x$totalAll_10x, sexCount_df_30x$totalAll_30x, sexCount_df_gtseq$totalAll_gtseq),
  # number of sex mismatches
  sexMismatches = c(sexMismatchAll_10x$mismatchAll, sexMismatchAll_30x$mismatchAll, sexMismatchAll_gtseq$mismatchAll)
) %>%
  mutate(across(everything(), ~ as.numeric(.))) %>%
  mutate(
    prop_spMismatches = spMismatches/spAssignments,
    prop_sexMismatches = sexMismatches/sexAssignments
  ) %>%
  select(-spMismatches, -sexMismatches) %>%
  relocate(prop_spMismatches, .after = spAssignments) %>%
  relocate(prop_sexMismatches, .after = sexAssignments) %>%
  mutate(across(everything(), round, 2)) %>%
  mutate(across(everything(), ~ prettyNum(.))) %>%
  mutate(
    avg_sampleGenoRate = cell_spec(avg_sampleGenoRate, background = ifelse(avg_sampleGenoRate == max(avg_sampleGenoRate), "green", "white")),
    avg_sampleGenoRate_blood = cell_spec(avg_sampleGenoRate_blood, background = ifelse(avg_sampleGenoRate_blood == max(avg_sampleGenoRate_blood), "green", "white")),
    avg_sampleGenoRate_hair = cell_spec(avg_sampleGenoRate_hair, background = ifelse(avg_sampleGenoRate_hair == max(avg_sampleGenoRate_hair), "green", "white")),
    spAssignments = cell_spec(spAssignments, background = ifelse(spAssignments == max(spAssignments), "green", "white")),
    prop_spMismatches = cell_spec(prop_spMismatches, background = ifelse(prop_spMismatches == min(prop_spMismatches), "green", "white")),
    sexAssignments = cell_spec(sexAssignments, background = ifelse(sexAssignments == max(sexAssignments), "green", "white")),
    prop_sexMismatches = cell_spec(prop_sexMismatches, background = ifelse(prop_sexMismatches == min(prop_sexMismatches), "green", "white"))
  ) %>%
  t() %>%
  as.data.frame() %>%
  dplyr::rename(
    "gtscore_10x" = "V1",
    "gtscore_30x" = "V2",
    "gtseq_10x" = "V3"
  )

summary_10x30xgtseq %>%
  kable(escape = FALSE) %>%
  kable_classic()
```

# Sample performance summary

Note that sample performance summaries EXCLUDE negatives.

## 10x coverage

```{r}
sum_samplePerfomance_gtscore10x <- sampleSum_noNegs_10x %>%
  select(c("animalID", "sampleType", "GenotypeRate")) %>%
  merge(., genos_species_md_noNegs_10x[, c("animalID", "sampleType", "mdMatch_sp5")], by = c("animalID", "sampleType")) %>%
  merge(., genos_sex_md_noNegs_10x[, c("animalID", "sampleType", "mdMatch_sex")], by = c("animalID", "sampleType")) %>%
  dplyr::rename(
    "genoSuccess_gtscore10x" = "GenotypeRate",
    "mdMatch_species_gtscore10x" = "mdMatch_sp5",
    "mdMatch_sex_gtscore10x" = "mdMatch_sex") %>%
  mutate(
    genoSuccess_gtscore10x = round(genoSuccess_gtscore10x, 2),
    redo_gtscore10x = 
      case_when(
        genoSuccess_gtscore10x < 0.5 ~ "yes",
        mdMatch_species_gtscore10x == "FALSE" ~ "yes",
        mdMatch_sex_gtscore10x == "FALSE" ~ "yes",
        .default = "no"
      )
  )

samples_toRedo_10x <- sum_samplePerfomance_gtscore10x %>%
  filter(redo_gtscore10x == "yes")
```

## 30x coverage

```{r}
sum_samplePerfomance_gtscore30x <- sampleSum_noNegs_30x %>%
  select(c("animalID", "sampleType", "GenotypeRate")) %>%
  merge(., genos_species_md_noNegs_30x[, c("animalID", "sampleType", "mdMatch_sp5")], by = c("animalID", "sampleType")) %>%
  merge(., genos_sex_md_noNegs_30x[, c("animalID", "sampleType", "mdMatch_sex")], by = c("animalID", "sampleType")) %>%
  dplyr::rename(
    "genoSuccess_gtscore30x" = "GenotypeRate",
    "mdMatch_species_gtscore30x" = "mdMatch_sp5",
    "mdMatch_sex_gtscore30x" = "mdMatch_sex") %>%
  mutate(
    genoSuccess_gtscore30x = round(genoSuccess_gtscore30x, 2),
    redo_gtscore30x = 
      case_when(
        genoSuccess_gtscore30x < 0.5 ~ "yes",
        mdMatch_species_gtscore30x == "FALSE" ~ "yes",
        mdMatch_sex_gtscore30x == "FALSE" ~ "yes",
        .default = "no"
      )
  )

samples_toRedo_30x <- sum_samplePerfomance_gtscore30x %>%
  filter(redo_gtscore30x == "yes")
```

## gtseq

```{r}
sum_samplePerfomance_gtseq10x <- sampleSum_noNegs_gtseq %>%
  select(c("animalID", "sampleType", "GenotypeRate")) %>%
  merge(., genos_species_md_noNegs_gtseq[, c("animalID", "sampleType", "mdMatch_sp5")], by = c("animalID", "sampleType")) %>%
  merge(., genos_sex_md_noNegs_gtseq[, c("animalID", "sampleType", "mdMatch_sex")], by = c("animalID", "sampleType")) %>%
  dplyr::rename(
    "genoSuccess_gtseq10x" = "GenotypeRate",
    "mdMatch_species_gtseq10x" = "mdMatch_sp5",
    "mdMatch_sex_gtseq10x" = "mdMatch_sex") %>%
  mutate(
    genoSuccess_gtseq10x = round(genoSuccess_gtseq10x, 2),
    redo_gtseq10x = 
      case_when(
        genoSuccess_gtseq10x < 0.5 ~ "yes",
        mdMatch_species_gtseq10x == "FALSE" ~ "yes",
        mdMatch_sex_gtseq10x == "FALSE" ~ "yes",
        .default = "no"
      )
  )

samples_toRedo_gtseq10x <- sum_samplePerfomance_gtseq10x %>%
  filter(redo_gtseq10x == "yes")
```

## 10x vs. 30x vs. gtseq

```{r}
sampleRedo_10x30xgtseq <- sum_samplePerfomance_gtscore10x %>%
  merge(., sum_samplePerfomance_gtscore30x, by = c("animalID", "sampleType")) %>%
  merge(., sum_samplePerfomance_gtseq10x, by = c("animalID", "sampleType")) %>%
  select(order(colnames(.))) %>%
  relocate(sampleType, .after = animalID) %>%
  mutate(
    redo_all = 
      case_when(
        redo_gtscore10x == "yes" & redo_gtscore30x == "yes" & redo_gtseq10x == "yes" ~ "yes",
        .default = "no"
        )
  )
  
sum_sampleRedo_10x30xgtseq <- rbind(
  table(sum_samplePerfomance_gtscore10x$redo_gtscore10x),
  table(sum_samplePerfomance_gtscore30x$redo_gtscore30x),
  table(sum_samplePerfomance_gtseq10x$redo_gtseq10x)
) %>%
  as.data.frame() %>%
  dplyr::rename(
    "redo_yes" = "yes",
    "redo_no" = "no"
  ) %>%
  mutate(
    pipeline = c("gtscore_10x", "gtscore_30x", "gtseq_10x"),
    avg_genoSuccess = c(
      mean(sampleSum_blood_10x$GenotypeRate),
      mean(sampleSum_blood_30x$GenotypeRate),
      mean(sampleSum_blood_gtseq$GenotypeRate)),
    prop_mismatchSpecies = c(
      as.numeric(speciesMismatchCount_df_10x$total)/as.numeric(speciesCount_df_10x$totalAll_10x),
      as.numeric(speciesMismatchCount_df_30x$total)/as.numeric(speciesCount_df_30x$totalAll_30x),
      as.numeric(speciesMismatchCount_df_gtseq$total)/as.numeric(speciesCount_df_gtseq$totalAll_gtseq)
    ),
    prop_mismatchSex = c(
      as.numeric(sexMismatchAll_10x$mismatchAll)/as.numeric(sexCount_df_10x$totalAll_10x),
      as.numeric(sexMismatchAll_30x$mismatchAll)/as.numeric(sexCount_df_30x$totalAll_30x),
      as.numeric(sexMismatchAll_gtseq$mismatchAll)/as.numeric(sexCount_df_gtseq$totalAll_gtseq)
      )
  ) %>%
  mutate(
    avg_genoSuccess = round(avg_genoSuccess, 2),
    prop_mismatchSpecies = round(prop_mismatchSpecies, 2),
    prop_mismatchSex = round(prop_mismatchSex, 2)
  ) %>%
  select(c("pipeline", "redo_yes", "redo_no", "avg_genoSuccess", "prop_mismatchSpecies", "prop_mismatchSex"))

table(sampleRedo_10x30xgtseq$redo_all)
```



```{r}
knitr::knit_exit()
```

# 7 Duptest

Checking duptest results to see if tamRun3 mismatch samples match any others in the dataset.

```{r}
dup- read.table("./05_tamRun5/03_run5GTscore/run3_polyGenResults_dupTest.txt", header = T) 

duptest_md <- duptest %>%
  merge(., md_tamRun3[, c("sampleID","animalID", "sampleType", "species")], by.x = "Sample1", by.y = "sampleID") %>%
  dplyr::rename("animalID1" = "animalID",
                "sampleType1" = "sampleType",
                "species1" = "species") %>%
  relocate(c("animalID1", "sampleType1"), .after = "Sample2")  %>%
  merge(., md_tamRun3[, c("sampleID","animalID", "sampleType", "species")], by.x = "Sample2", by.y = "sampleID") %>%
  dplyr::rename("animalID2" = "animalID",
                "sampleType2" = "sampleType",
                "species2" = "species") %>%
  merge(., genos_species_md[, c("sampleID", "mdMatch_sp5")], by.x = "Sample1", by.y = "sampleID", all.x = T) %>%
  dplyr::rename("mdMatch_sp1" = "mdMatch_sp5") %>%
  merge(., genos_species_md[, c("sampleID", "mdMatch_sp5")], by.x = "Sample2", by.y = "sampleID", all.x = T) %>%
  dplyr::rename("mdMatch_sp2" = "mdMatch_sp5") %>%
  select(c(Sample1, Sample2, animalID1, animalID2, sampleType1, sampleType2, species1, species2, mdMatch_sp1, mdMatch_sp2, matchedGenotypes, commonGenotypes, proportionMatch, proportionCommon))

duptest_animIDMatch <- duptest_md %>%
  filter(animalID1 == animalID2)

duptest_mismatch_sp1 <- duptest_md %>%
  filter(mdMatch_sp1 == "FALSE")

duptest_highMatch <- duptest_md %>%
  filter(proportionCommon >= 0.5) %>%
  filter(proportionMatch >= 0.5) %>%
  mutate(dupMatch_sp = case_when(species1 == species2 ~ "TRUE",
                                  .default = "FALSE"))

table(duptest_highMatch$mdMatch_sp1)

tr3_069 <- duptest_md %>%
  filter(Sample1 == "tamRun3.069")
```

### 5.4.2 Species-specific sex loci

We can check to see if samples in previous runs have been genotyped at species-specific loci NOT for their species. While no LWED samples were genotyped at SIMP-sex loci, all 14 SIMP-assigned samples were genotyped at SEXID_LWED_208 (but not 195).

Perhaps important to keep in mind, all SIMP-assigned samples were genotyped as F at SEXID_LWED_208.

```{r}
genos_species_tamRun1 <- genos_tamRun1 %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SPECIES")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(speciesKey, by = c("Locus", "genotype")) %>%
  mutate(species = coalesce(species, genotype)) %>%
  select(c(Locus, sampleID, species)) %>%
  pivot_wider(names_from = sampleID,
              values_from = species) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  mutate(totalGenos_sp = (rowSums(. == "LWED") + rowSums(. == "SIMP"))) %>%
  mutate(propGenos_sp = (totalGenos_sp/12)) %>%
  mutate(propLWED = (rowSums(. == "LWED")/totalGenos_sp)) %>%
  mutate(propSIMP = (rowSums(. == "SIMP")/totalGenos_sp)) %>%
  mutate(propMismatch_species = (1 - abs(propLWED - propSIMP))) %>%
  mutate(speciesAssigned = ifelse(propLWED == 1, "LWED", 
                              ifelse(propSIMP == 1, "SIMP", 
                                     ifelse(propLWED < 1 & propSIMP > 0, "MIX", NA)))) %>%
  rownames_to_column("sampleID") %>%
  mutate(sampleID = sub('_S.*', '', sampleID))

genos_species_md_tamRun1 <- md_tamRun1 %>%
  select(c(sampleID, animalID, species, sampleType)) %>%
  merge(., genos_species_tamRun1, by = "sampleID") %>%
  mutate(mdMatch_sp = ifelse(species == speciesAssigned, "TRUE", "FALSE")) %>%
  relocate(animalID, .after = sampleID) %>%
  relocate(mdMatch_sp, .after = animalID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F)

genos_sex_tamRun1 <- genos_tamRun1 %>%
  rownames_to_column("Locus") %>%
  filter(str_detect(Locus, "SEXID")) %>%
  mutate(across(everything(), as.character)) %>%
  pivot_longer(!Locus,
               names_to = "sampleID",
               values_to = "genotype") %>%
  left_join(sexKey, by = c("Locus", "genotype")) %>%
  mutate(sex = coalesce(sex, genotype)) %>%
  select(c(Locus, sampleID, sex)) %>%
  pivot_wider(names_from = sampleID,
              values_from = sex) %>%
  column_to_rownames("Locus") %>%
  t() %>%
  as.data.frame() %>%
  
  # BOTH PRIMER POOLS
  mutate(totalGenos_sex = (rowSums(. == "F") + rowSums(. == "M"))) %>%
  mutate(propGenos_sex = (totalGenos_sex/12)) %>%
  mutate(propF = (rowSums(. == "F")/totalGenos_sex)) %>%
  mutate(propM = (rowSums(. == "M")/totalGenos_sex)) %>%
  mutate(propMismatch_sex = (1 - abs(propF - propM))) %>%
  mutate(sexAssigned = ifelse(propF == 1, "F", 
                              ifelse(propM == 1, "M", 
                                     ifelse(propF < 1 & propF > 0, "MIX", NA)))) %>%
  
  # PRIMER POOL 2
  mutate(totalGenos_sex_pp2 = 
           (rowSums(across(matches(loci_sex_pp2)) == "F")) + 
           (rowSums(across(matches(loci_sex_pp2)) == "M"))) %>%
  mutate(propGenos_sex_pp2 = 
           (totalGenos_sex_pp2/length(loci_sex_pp2))) %>%
  mutate(propF_pp2 = 
           (rowSums(across(matches(loci_sex_pp2)) == "F"))/totalGenos_sex_pp2) %>%
  mutate(propM_pp2 = 
           (rowSums(across(matches(loci_sex_pp2)) == "M"))/totalGenos_sex_pp2) %>%
  mutate(propMismatch_sex_pp2 = 
           (1 - abs(propF_pp2 - propM_pp2))) %>%
  mutate(sexAssigned_pp2 = 
           ifelse(propF_pp2 == 1, "F", 
                              ifelse(propM_pp2 == 1, "M", 
                                     ifelse(propF_pp2 < 1 & propF_pp2 > 0, "MIX", NA)))) %>%
  
  # PRIMER POOL 3
  mutate(totalGenos_sex_pp3 = 
           (rowSums(across(matches(loci_sex_pp3)) == "F")) + 
           (rowSums(across(matches(loci_sex_pp3)) == "M"))) %>%
  mutate(propGenos_sex_pp3 = 
           (totalGenos_sex_pp3/length(loci_sex_pp3))) %>%
  mutate(propF_pp3 = 
           (rowSums(across(matches(loci_sex_pp3)) == "F"))/totalGenos_sex_pp3) %>%
  mutate(propM_pp3 = 
           (rowSums(across(matches(loci_sex_pp3)) == "M"))/totalGenos_sex_pp3) %>%
  mutate(propMismatch_sex_pp3 = 
           (1 - abs(propF_pp3 - propM_pp3))) %>%
  mutate(sexAssigned_pp3 = 
           ifelse(propF_pp3 == 1, "F", 
                              ifelse(propM_pp3 == 1, "M", 
                                     ifelse(propF_pp3 < 1 & propF_pp3 > 0, "MIX", NA)))) %>%
  # LWED SEX LOCI
  mutate(lwedSex = 
           (rowSums(across(contains("LWED")) == "F")) + 
           (rowSums(across(contains("LWED")) == "M"))) %>%
  # SIMP SEX LOCI
  mutate(simpSex = 
           (rowSums(across(contains("SIMP")) == "F")) + 
           (rowSums(across(contains("SIMP")) == "M"))) %>%
  
  rownames_to_column("sampleID") %>%
  mutate(sampleID = sub('_S.*', '', sampleID))

genos_sex_tamRun1_md <- md_tamRun1 %>%
  select(c(sampleID, animalID, species, sex, sampleType)) %>%
  merge(., genos_sex_tamRun1, by = "sampleID") %>%
  mutate(mdMatch_sex = case_when(sexAssigned == "MIX" ~ "MIX",
                                 sex == sexAssigned ~ "TRUE",
                                 sex != sexAssigned ~ "FALSE")) %>%
  mutate(mdMatch_sex_pp2 = case_when(sexAssigned_pp2 == "MIX" ~ "MIX",
                                 sex == sexAssigned_pp2 ~ "TRUE",
                                 sex != sexAssigned_pp2 ~ "FALSE")) %>%
  mutate(mdMatch_sex_pp3 = case_when(sexAssigned_pp3 == "MIX" ~ "MIX",
                                 sex == sexAssigned_pp3 ~ "TRUE",
                                 sex != sexAssigned_pp3 ~ "FALSE")) %>%
  relocate(c(mdMatch_sex, mdMatch_sex_pp2, mdMatch_sex_pp3), .after = sampleID) %>%
  relocate(sampleType, .after = animalID) %>%
  unite("sample", animalID:sampleType, sep = "_", remove = F)

genos_spSex_tamRun1 <- genos_species_md_tamRun1 %>%
  select(c(sampleID, sampleType, species, speciesAssigned, mdMatch_sp)) %>%
  merge(genos_sex_tamRun1_md %>% select(sampleID, sex, sexAssigned_pp2, sexAssigned_pp3, mdMatch_sex_pp2, mdMatch_sex_pp3, lwedSex, simpSex), by = "sampleID") %>%
  select(c(sampleID, sampleType, species, sex, mdMatch_sp, mdMatch_sex_pp2, mdMatch_sex_pp3, speciesAssigned, sexAssigned_pp2, simpSex, sexAssigned_pp3, lwedSex))

matrix(c(sum(genos_spSex_tamRun1$speciesAssigned %in% "LWED"),
         genos_spSex_tamRun1 %>%
           filter(speciesAssigned == "LWED") %>%
           filter(lwedSex > 0) %>%
           nrow(),
         genos_spSex_tamRun1 %>%
           filter(speciesAssigned == "LWED") %>%
           filter(simpSex > 0) %>%
           nrow(),
       sum(genos_spSex_tamRun1$speciesAssigned %in% "SIMP"),
       genos_spSex_tamRun1 %>%
           filter(speciesAssigned == "SIMP") %>%
           filter(lwedSex > 0) %>%
           nrow(),
       genos_spSex_spMismatch %>%
           filter(speciesAssigned == "SIMP") %>%
           filter(simpSex > 0) %>%
           nrow()),
       ncol = 2,
       nrow = 3) %>%
       `rownames<-`(., c("totalMatch", "lwedSex", "simpSex")) %>%
       `colnames<-`(., c("lwedAssigned", "simpAssigned"))
```

## 5.1 Split primer pools

In primer pool 2, we have:

-   all SPECIESID loci
-   6 LWED-specific loci
-   no SIMP-specific loci

In primer pool 3, we have:

-   no SPECIESID loci
-   26 LWED-specific loci
-   all SIMP-specific loci

For hair samples with mismatches, let's take a closer look at the genotypes from loci in primer pool 2 vs. primer pool 3

```{r}
pp2 <- read.table("./01_run2_fecalHairBlood/03_run2GTscore/p2_combined_primerProbeFile.txt", header = T) %>%
  mutate(Locus = sub('[.][^.]+$', '', Locus)) %>%
  mutate(primerPool = "pool2") %>%
  select(c(Locus, primerPool))
pp3 <- read.table("./01_run2_fecalHairBlood/03_run2GTscore/p3_combined_primerProbeFile.txt", header = T) %>%
  mutate(Locus = sub('[.][^.]+$', '', Locus)) %>%
  mutate(primerPool = "pool3") %>%
  select(c(Locus, primerPool))
splitPools <- rbind(pp2, pp3) %>%
  mutate(Locus = str_replace(Locus, 'SEXID_195', 'SEXID_LWED_195')) %>%
  mutate(Locus = str_replace(Locus, 'SEXID_197', 'SEXID_SIMP_197')) %>%
  mutate(Locus = str_replace(Locus, 'SEXID_198', 'SEXID_SIMP_198')) %>%
  mutate(Locus = str_replace(Locus, 'SEXID_203', 'SEXID_SIMP_203')) %>%
  mutate(Locus = str_replace(Locus, 'SEXID_208', 'SEXID_LWED_208')) %>%
#  mutate(Locus = str_replace(Locus, 'SEXID_211', 'SEXID_LWED_211')) %>%
  mutate(Locus = str_replace(Locus, 'SEXID_218', 'SEXID_SIMP_218'))

genos_mismatches_hair <- genos_fullSet %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  filter(sampleID %in% speciesSex_mismatches_hair$sampleID) %>%
  column_to_rownames("sampleID") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Locus") %>%
  mutate(Locus = sub('[_][^_]+$', '', Locus)) %>%
  merge(., splitPools, by = "Locus") %>%
  relocate(primerPool, .after = "Locus")
```

# X Hair-blood pairs

Which hair samples have corresponding blood samples?

```{r}
genos_md <- genos_fullSet %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  merge(., md_tamRun3[,c("sampleID", "animalID", "sampleType")], by = "sampleID")

genos_hair <- genos_md %>%
  filter(sampleType == "hair")

genos_blood <- genos_md %>%
  filter(sampleType == "blood")
```

We have 154 individuals with both hair and blood samples available; 103 of those samples have at least one common locus genotyped

What proportion of loci are genotyped in both samples? Here we can use the results from GTscore's duptest.

[[NOTE - may want to redo this after coding species-specific loci with zeros]]

```{r}
sampleSummary <- read.csv("./../03_run3GTscore/summaryFiles/master_sampleSummary.csv") %>%
  mutate(sampleID = gsub("-","\\.", sampleID))
dup- read.table("./../03_run3GTscore/run3_polyGenResults_dupTest.txt", header = T)

hairBlood_pairs <- genos_hair$animalID[genos_hair$animalID %in% genos_blood$animalID] %>%
  as.data.frame() %>%
  rename("animalID" = ".") %>%
  distinct() %>%
  # add sampleIDs
  merge(., genos_blood[,c("sampleID", "animalID")], by = "animalID", .keep_all = T) %>%
  rename("sampleID_blood" = "sampleID") %>%
  merge(., genos_hair[,c("sampleID", "animalID")], by = "animalID", .keep_all = T) %>%
  rename("sampleID_hair" = "sampleID") %>%
  # add genotype success
  merge(., sampleSummary[,c("sampleID", "GenotypeRate")], by.x = "sampleID_blood", by.y = "sampleID", .keep_all = T) %>%
  rename("genoRate_blood" = "GenotypeRate") %>%
  merge(., sampleSummary[,c("sampleID", "GenotypeRate")], by.x = "sampleID_hair", by.y = "sampleID", .keep_all = T) %>%
  rename("genoRate_hair" = "GenotypeRate") %>%
  # Add proportion common genotypes
  dplyr::select(c(animalID, sampleID_hair, sampleID_blood, genoRate_hair, genoRate_blood)) %>%
  merge(., dupTest[,c("Sample1", "Sample2", "commonGenotypes", "proportionCommon")], by.x = c("sampleID_hair", "sampleID_blood"), by.y = c("Sample1", "Sample2"), .keep_all = T)

length(hairBlood_pairs$animalID)
sum(hairBlood_pairs$proportionCommon > 0)
sum(hairBlood_pairs$proportionCommon == 0)
```

Visualize hair-blood pair common genotypes

median: 51 common genotypes (23% of all loci) range: 0 to 0.78

```{r}
median(hairBlood_pairs$commonGenotypes)
median(hairBlood_pairs$proportionCommon)
range(hairBlood_pairs$proportionCommon)

ggplot() + 
  geom_histogram(data = hairBlood_pairs, aes(x = proportionCommon), binwidth = 0.01) +
  xlim(-0.01,1.01) +
  geom_vline(xintercept = median(hairBlood_pairs$proportionCommon), linetype = "dashed") +
  annotate(x = 0.23, y = 50, label = "Median = 0.23", geom = "label") +
  labs(title="Proportion of common genotypes shared by blood-hair sample pairs", x="Proportion of common genotypes", y="Number of pairs") +
  theme_bw() + 
  theme(plot.title=element_text(hjust=0.5),plot.subtitle=element_text(hjust=0.5))
```

# TROUBLESHOOTING

There are several samples from tamRun3 that do not match the species/sex in the metadata. This section is meant to figure out what's going on and what to do moving forward.

## General metrics tamRun1,2,3

```{r}
# tamRun1
metrics_tamRun1 <- c(length(md_tamRun1$sampleID),
                     length(which(is.na(md_tamRun1$sampleType))),
                     length(md_tamRun1$animalID[md_tamRun1$sampleType %in% "hair"]),
                     length(unique(md_tamRun1$animalID[md_tamRun1$sampleType %in% "hair"])),
                     length(md_tamRun1$animalID[md_tamRun1$sampleType %in% "blood"]),
                     length(unique(md_tamRun1$animalID[md_tamRun1$sampleType %in% "blood"])))

metrics_tamRun2 <- c(length(md_tamRun2$sampleID),
                     length(which(is.na(md_tamRun2$sampleType))),
                     length(md_tamRun2$animalID[md_tamRun2$sampleType %in% "hair"]),
                     length(unique(md_tamRun2$animalID[md_tamRun2$sampleType %in% "hair"])),
                     length(md_tamRun2$animalID[md_tamRun2$sampleType %in% "blood"]),
                     length(unique(md_tamRun2$animalID[md_tamRun2$sampleType %in% "blood"])))

metrics_tamRun3 <- c(length(md_tamRun3$sampleID),
                     length(unique(md_tamRun3$index_combo[md_tamRun3$sampleType %in% c("PCR1 (-)", "XTN (-)")])),
                     length(md_tamRun3$animalID[md_tamRun3$sampleType %in% "hair"]),
                     length(unique(md_tamRun3$animalID[md_tamRun3$sampleType %in% "hair"])),
                     length(md_tamRun3$animalID[md_tamRun3$sampleType %in% "blood"]),
                     length(unique(md_tamRun3$animalID[md_tamRun3$sampleType %in% "blood"])))

matrix(rbind(metrics_tamRun1, metrics_tamRun2, metrics_tamRun3),
       nrow = 3,
       ncol = 6) %>%
  `rownames<-`(., c("tamRun1", "tamRun2", "tamRun3")) %>%
  `colnames<-`(., c("totalSamples", "totalNeg", "totalHair", "uniqueHair", "totalBlood", "uniqueBlood")) %>%
  t() %>%
  kbl() %>%
  kable_classic(full_width = F)
```

note on rownames-- When you write this: rownames(df) \<- c(1:5)

What R is actually doing is running this function: `rownames<-`(df, c(1:5))

<https://stackoverflow.com/questions/56014722/how-can-i-assign-rownames-with-while-using-a-pipe>

## Genotype comparison

Here I'm comparing the genotypes called in tamRun3 to those called for the same samples in runs 1 and 2. Note that I'm not filtering out any loci that are species-specific, but I am removing negative controls.

### tamRun3 vs. tamRun1/2

## Sample by sample

10_hair (tamRun3_177) \* should be: LWED-M \* tamRun3 says: SIMP-F (6 F-pp2, 1 F-pp3)

```{r}
duptest_10_hair <- duptest_md %>%
  filter(animalID1 == "10") %>%
  filter(sampleType1 == "hair")

genos_10_hair <- genos_tamRun3 %>%
  select(c(tamRun3.177, tamRun3.185, tamRun3.417))
```

Likeliest match according to duptest is 15_blood (tamRun3_417) -- so what does 15_hair (tamRun3.185) look like? \* should be: SIMP-F \* tamRun3 says: LWED-[1F-pp2, 1M-pp3, 1F-pp3] \* 47 loci have calls in another run (run2Geno1); most of these are different from tamRun3

```{r}
duptest_15_hair <- duptest_md %>%
  filter(animalID1 == "15") %>%
  filter(sampleType1 == "hair")

multipleCalls_15_hair <- multipleCalls_hair %>%
  filter(sample == "15_hair") %>%
  .[, colSums(is.na(.)) < nrow(.)]

# Does tamRun3_10_hair match tamRun2_15_hair?
genos_10vs15_hair <- genos_10_hair %>%
  merge(., genos_tamRun2 %>% select(tamRun2.100_S84_L001_interleaved), by = 'row.names', all.x = T) %>%
  column_to_rownames("Row.names") %>%
  select(c(tamRun3.177, tamRun2.100_S84_L001_interleaved)) %>%
  filter(!tamRun3.177 == 0) %>%
  filter(!tamRun2.100_S84_L001_interleaved == 0) %>%
  mutate(match = ifelse(tamRun3.177 == tamRun2.100_S84_L001_interleaved, "TRUE", "FALSE"))
```



# SCRAPS

propMatch_run12 shows...

-   the average number of UNIQUE genotypes per locus...
-   for sample/locus combinations with genotypes called at least once across tamRun1 and tamRun2
-   **NOTE** that samples may not be in both tamRun1 and 2; they may have just been duplicated within one of these runs

propMatch_run125 shows...

-   the average number of UNIQUE genotypes per locus...
-   for sample/locus combinations with genotypes called at least once in tamRun3_cat AND at least once in either tamRun1 or tamRun2 (N = 66)

We can use this comparison dataframe as follows (for example):
-   "Sample 1_blood had 150 genotypes called in tamrun3_cat that were also called in tamRun1/2. Of these 150 genotypes, 5.33% did not match between tamRun3 and tamRun1/2."

```{r, include=FALSE}
samples_spMatch_tamRun3cat <- genos_species_md_cat %>%
  filter(mdMatch_sp5 == "TRUE")
samples_spMismatch_tamRun3cat <- genos_species_md_cat %>%
  filter(mdMatch_sp5 == "FALSE")
samples_sexMatch_tamRun3cat <- genos_sex_md_cat %>%
  filter(mdMatch_sex == "TRUE")
samples_sexMismatch_tamRun3cat <- genos_sex_md_cat %>%
  filter(mdMatch_sex == "FALSE")

propMatch_run12 <- genos_tamRun125 %>%
  filter(genosCalled_12 > 1) %>%
  group_by(sample) %>%
  summarise(meanUnique_12 = mean(uniqueGenos_12),
            sumGenos_12 = sum(genosCalled_12)) %>%
  as.data.frame()

propMatch_run125 <- genos_tamRun125 %>%
  filter(genosCalled_125 > 1) %>%
  group_by(sample) %>%
  summarise(meanUnique_125 = mean(uniqueGenos_125),
            sumGenos_125 = sum(genosCalled_125),
            sumGenos_12 = sum(genosCalled_12)) %>%
  as.data.frame() %>%
  mutate(sumGenos_3 = sumGenos_125 - sumGenos_12) %>%
  filter(sumGenos_3 > 0) %>%
  merge(., propMatch_run12[,c("sample", "meanUnique_12")], by = "sample", all.x = T) %>%
  mutate(spMatch_run3 = 
           case_when(sample %in% samples_spMatch_tamRun3cat$sample ~ "TRUE", sample %in% samples_spMismatch_tamRun3cat$sample ~ "FALSE")) %>%
  mutate(sexMatch_run3 = 
           case_when(sample %in% samples_sexMatch_tamRun3cat$sample ~ "TRUE", sample %in% samples_sexMismatch_tamRun3cat$sample ~ "FALSE"))
```

Across tamRun3 a, b, and cat results, 369 samples had loci genotyped in more than one dataset. Of these, 123 had at least one locus with different genotypes in either a, b, or cat datasets, with at most 15 loci with differing genotypes (7.5% of loci genotyped for that sample).

The number of loci with different calls between datasets does not appear to be related to the number of on-target reads (from "cat" data) per sample.

```{r, echo=FALSE}
# Remove sample/loci with no calls in any dataset
tamRun3_genos_abcat_noZeros <- tamRun3_genos_abcat %>%
  filter(uniqueGenos != 0) %>%
  merge(., locusSum[,c("Locus", "set")], by.x = "locus", by.y = "Locus", all.x = T) %>%
  merge(., sampleSum_cat[,c("sampleID2", "sampleType")], by.x = "sampleID", by.y = "sampleID2", all.x = T)

# Summary stats per sample
genoDiffs_perSample_summary_tamRun3 <- tamRun3_genos_abcat_noZeros %>%
  group_by(sampleID) %>%
  summarise(lociGenotyped = length(locus),
            # no. loci w/one unique geno called
            oneUnique = length(locus[uniqueGenos == 1]),
            # no. loci w/two unique genos called
            twoUnique = length(locus[uniqueGenos == 2]),
            # no. loci w/three unique genos called
            threeUnique = length(locus[uniqueGenos == 3]),
            # min unique genos called per locus per sample
            min_uniqueGenos = min(uniqueGenos),
            # max unique genos called per locus per sample
            max_uniqueGenos = max(uniqueGenos),
            # mean unique genos called per locus per sample
            mean_uniqueGenos = mean(uniqueGenos),
            # variance in unique genos called per locus per sample
            var_uniqueGenos = var(uniqueGenos)) %>%
  as.data.frame() %>%
  mutate(propDiff = (mean_uniqueGenos - 1)) %>%
  relocate(propDiff, .after = mean_uniqueGenos) %>%
  mutate(count_lociDiff = twoUnique + threeUnique) %>%
  relocate(count_lociDiff, .after = threeUnique) %>%
  merge(., sampleSum_cat[,c("sampleID2", "sampleType", "Primer.Probe.Reads")], by.x = "sampleID", by.y = "sampleID2", all.x = T)

# How many loci per set are giving different genos per sample?
countDiffLoci_byLocusSet_perSample_tamRun3 <- tamRun3_genos_abcat_noZeros %>%
  filter(uniqueGenos > 1) %>%
  group_by(set, sampleID) %>%
  summarise(countLoci_diffGenos = length(unique(locus))) %>%
  as.data.frame() %>%
  pivot_wider(names_from = set,
              values_from = countLoci_diffGenos)
```

**Variation in number of unique genotypes called per locus across samples**

```{r}
summary(tamRun3_genos_abcat_noZeros$uniqueGenos)
```

**Overview of genotype differences PER SAMPLE between tamRun3 datasets**

```{r, echo=FALSE}
uniqueGenos_perSample_summaryTable_tamRun3 <- matrix(c(length(genoDiffs_perSample_summary_tamRun3$sampleID),
         genoDiffs_perSample_summary_tamRun3 %>%
           filter(count_lociDiff > 1) %>%
         nrow(),
         mean(genoDiffs_perSample_summary_tamRun3$count_lociDiff),
         mean(genoDiffs_perSample_summary_tamRun3$mean_uniqueGenos - 1),
         max(genoDiffs_perSample_summary_tamRun3$count_lociDiff),
         max(genoDiffs_perSample_summary_tamRun3$mean_uniqueGenos - 1)),
         nrow = 6,
         ncol = 1) %>%
  `rownames<-`(., c("Total samples",
                    "Samples w/multiple unique genos",
                    "Avg # of loci w/unique genos per sample",
                    "Avg prop. loci w/unique genos per sample",
                    "Max # loci w/unique genos per sample",
                    "Max prop. loci w/unique genos per sample")) %>%
  as.data.frame() %>%
  mutate(V1 = signif(V1, digits = 3)) %>%
  mutate(V1 = as.character(V1))
names(uniqueGenos_perSample_summaryTable_tamRun3) <- NULL

uniqueGenos_perSample_summaryTable_tamRun3 %>%
  kable(caption = "Overview of genos called across SAMPLES between tamRun3 datasets") %>%
  kable_material(full_width = F) %>%
  scroll_box(fixed_thead = list(enabled = T, background = "#363636"))
```

**Plots**

```{r, echo=FALSE}
# Boxplot
ggplot(genoDiffs_perSample_summary_tamRun3,
       aes(x = (mean_uniqueGenos - 1),
           y = factor(0))) +
  geom_boxplot() +
  theme_bw() +
  labs(x = "Proportion of genotyped loci",
       y = "",
       title = "Proportion of loci per sample with different genotypes between tamRun3 a, b, and cat") +
  theme(axis.text.y=element_blank(),axis.ticks.y=element_blank())

# Scatterplot by sampleType
ggplot(genoDiffs_perSample_summary_tamRun3, aes(x = count_lociDiff, y = Primer.Probe.Reads, color = sampleType)) +
  geom_point() +
  theme_bw() +
  labs(y = "On-target primer-probe reads",
       x = "Number of loci w/differing genotypes",
       title = "Number of loci with different genotypes between tamRun3 a, b, and cat vs. on target primer-probe reads")
```
