---
title: "GTseq vs. GTscore data analysis pipelines"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: 
  rmdformats::downcute:
    downcute_theme: "chaos"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto
}

pre[class] {
  max-height: 100px;
}
```

# 1 Overview

## 1.1 General info

The current document provides a comparison of GTscore vs. GTseq data analysis pipelines, with the goal of determining whether one or the other (or a combination of both) is best to use for the analysis of GT-seq data.

## 1.2 Background on GTseq vs. GTscore pipelines

There are two main pipelines for the analysis of data acquired via genotyping-in-thousands by sequencing (GT-seq). The original pipeline is the "GTseq pipeline" ([GitHub link](https://github.com/GTseq/GTseq-Pipeline)) created by Nate Campbell, who developed the original protocol for GT-seq. The second pipeline is the "GTscore pipeline" ([Github link](https://github.com/gjmckinney/GTscore)) created by Garrett McKinney, Carita Pascal, and Lisa Seeb, which follows the same general format as GTseq, but includes the following additions:

-   Determines genotypes for both multi-SNP haplotypes and single-SNP genotypes
-   Retains phase for multi-SNP haplotypes
-   Handles genotypes for loci with varying ploidy levels
-   Provides detailed summaries and plots and optional diagnostics

Beyond these additions, there are a few other differences to keep in mind:

-   Overall, documentation for the GTscore pipeline is much more thorough; as a bioinformatics novice, I found it much easier to follow
-   Both pipelines rely on perl scripts, but the GTscore pipeline is based in R
-   Both pipelines provide scripts to help with locus/sample diagnostics and primer optimization; these scripts are built into the GTscore pipeline, but analysis via the GTseq pipeline requires additional steps
-   GTscore does not include a coverage cutoff when genotyping samples, but GTseq excludes loci with less than 10x coverage

## 1.3 Summary of findings

The key findings from my comparison of GTseq and GTscore data analysis pipelines are below:

-   Sample performance
    -   Both pipelines give the same number of total reads per sample

    -   On-target reads per sample were on average 0.34% HIGHER for GTseq vs. GTscore

    -   Genotype success per sample was on average 3.2% LOWER for GTseq vs. GTscore

    -   Of the 96,798 potential sample/loci combinations for which a genotype could be called, 38,007 genotypes (39.3%) were called by both pipelines

        -   GTseq called 87 genotypes where GTscore did not, providing genotypes for 39.4% of all sample/loci combinations

        -   GTscore called 3,262 genotypes where GTseq did not, providing genotypes for 42.6% of all sample/loci combinations

        -   *Note - keep in mind that the genotype percentages do NOT take species-specific loci into account, so actual values are likely a bit higher*

    -   Of the 38,007 genotypes called by both pipelines, there were no differences in genotype calls between pipelines

    -   Removing the 10x coverage cutoff that I added to the GTscore pipeline yields a 21.8% increase in genotypes called by GTscore

        -   30 of these additional genotypes were originally missing from GTscore genotype results, but present from GTseq genotype results

        -   There were still no differences in genotype calls between pipelines after removing the 10x cutoff
-   Locus performance:
    -   The proportion of primer-probe reads to primer reads was 0.47% (0.0047) HIGHER with GTseq vs. GTscore

    -   GTscore reports two loci as being fixed (LWED_248, SIMP_318), but GTseq shows these same loci as *almost* fixed (and reports no fixed loci overall). Given that there aren't any mismatched genotypes (see section 4.5), we can assume that the allele reads for those "almost-fixed" loci are from the small handful of cases where GTseq gave a genotype, but GTscore did not.

    -   GTseq and GTscore show the same loci as overamplifying relative to the others in the primer pool

        -   Includes 28 loci total, with 6 generic loci overamplifying in both species, 7 generic loci overamplifying in LWED samples only, 5 LWED-specific loci, and 10 SIMP-specific loci

## 1.4 Recommendations

Based on my comparisons of GTseq and GTscore data analysis pipelines, **I recommend using the GTscore pipeline for the analysis of GT-seq data moving forward**. Given that the sample/locus metrics are comparable between the two pipelines, I'm primarily basing this recommendation on the higher number of genotypes provided by GTscore. While the lower number of genotypes from GTseq suggests that its genotyping algorithm is more conservative, I don't think a more conservative approach is necessary given that there were no differences in the genotypes called by both pipelines.

Beyond the higher percentage of genotypes called, I also think that the detailed documentation and incorporation of troubleshooting scripts within the GTscore pipeline make it a better choice, particularly when first developing a GT-seq protocol.

# 2 Packages

```{r}
library(data.table, quietly = T)
library(ggpubr, quietly = T)
library(ggthemes, quietly = T)
library(gsubfn, quietly = T)
library(kableExtra, quietly = T)
library(magrittr, quietly = T)
suppressPackageStartupMessages(library(tidyverse))
```

# 3 Data

To compare the GTseq and GTscore pipelines, I'm using sequencing data from tamRun3, the third MiSeq sequencing run for the saddleback (LWED) and emperor (SIMP) tamarin population at the Estación Biológica Río Los Amigos in Perú.

This run has a total of **438 samples**, including 235 hair samples (138 LWED, 97 SIMP), 194 blood samples (112 LWED, 82 SIMP), and 9 negative controls.

## 3.1 Metadata & sample lists

**Metadata**

```{r}
md_tamRun3 <- read.csv("./03_tamRun3/03_run3GTscore/tamRun3_metadata_12Jan2023.csv") %>%
  mutate(sampleID = gsub("_","\\.", sampleID))
```

**Sample lists**

```{r}
samples_noNeg <- md_tamRun3 %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  select(sampleID)

samplesLWED <- md_tamRun3 %>%
  filter(species == "LWED") %>%
  select(sampleID)

samplesSIMP <- md_tamRun3 %>%
  filter(species == "SIMP") %>%
  select(sampleID)
```

## 3.2 GTseq data

The genotypes and locus/sample metrics for GTseq are from the GT-seq pipeline. For easy reference, I've compiled the steps to this pipeline - including scripts, their descriptions, and detailed notes on how to run them - in a separate file (gtseqPipeline_tamRun3.Rmd).

### Genotypes

```{r}
genos_gtseq <- read.csv("./gtscore_vs_gtseq/compiledGenos_gtseq_tamRun3.csv", header = T) %>%
  mutate(Sample = sub("\\_.*", "", Sample))
```

### Locus summaries

From what I could tell, the GTseq pipeline doesn't have a script that provides locus-specific summary metrics (e.g., total reads per locus across samples). Instead, we need to create this based on the results from the individual sample seqTest and genos files. We can use the seqTest files to give us total read counts per locus and the genos files to give read counts for each allele per locus.

#### Locus read counts

GTseq_seqTest.pl outputs files for each sample that contain three columns, which represent counts of the following:

-   primer reads
-   probe reads
-   primer-probe reads (i.e., reads when both primer and probe occur in the sequence)

We can sum these sample-specific read counts to get loci-specific read counts by combining all of the different seqTest sample files. I've also included species-specific subsets, where read counts per locus include only those from LWED or SIMP samples.

```{r}
# use full.names = T to list full sample path (avoid changing wd)
seqTestFiles_gtseq <- list.files(path = "/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq/seqTest", pattern = "*.csv", full.names = T)

# Read in individual sample seqTest files and combine
seqTest_compiled <- lapply(seqTestFiles_gtseq, fread, select = c("V1", "V2", "V3", "V4")) %>%
  set_names(seqTestFiles_gtseq) %>%
  rbindlist(idcol = "sample") %>%
  as.data.frame() %>%
  dplyr::mutate(sample = sub(".*seqTest/", "", sample)) %>%
  dplyr::mutate(sample = sub("\\_.*", "", sample)) %>%
  dplyr::mutate(sample = gsub("-","\\.", sample)) %>%
  dplyr::rename("locus" = "V1",
                "primerReads" = "V2",
                "probeReads" = "V3", 
                "primerProbeReads" = "V4") %>%
  dplyr::mutate(sample = sub("\\_.*", "", sample)) %>%
  dplyr::mutate(sample = sub("-", "\\.", sample))

# Remove negative controls
lociReadCounts_lwedSIMP <- seqTest_compiled %>%
  filter(sample %in% samples_noNeg$sample) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  dplyr::mutate(primerProbeProportion = primerProbeReads/primerReads)

# Loci counts for LWED samples only
lociReadCounts_lwed <- seqTest_compiled %>%
  filter(sample %in% samplesLWED$sampleID) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  dplyr::mutate(primerProbeProportion = primerProbeReads/primerReads)

# Loci counts for SIMP samples only
lociReadCounts_simp <- seqTest_compiled %>%
  filter(sample %in% samplesSIMP$sampleID) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  dplyr::mutate(primerProbeProportion = primerProbeReads/primerReads)
```

Using the species-subsets created above, we can then create a final locus read-count summary that's species-specific, such that:

-   values for LWED-specific loci reflect performance with LWED individuals only
-   values for SIMP-specific loci reflect performance with SIMP individuals only
-   all other loci reflect performance with LWED and SIMP individuals combined

```{r}
lociRC <- lociReadCounts_lwedSIMP %>%
  filter(!str_detect(locus, "LWED|SIMP"))
lociRC_lwed <- lociReadCounts_lwed %>%
  filter(str_detect(locus, "LWED"))
lociRC_simp <- lociReadCounts_simp %>%
  filter(str_detect(locus, "SIMP"))

lociReadCounts_combine <- rbind(lociRC, lociRC_lwed, lociRC_simp)
```

#### Allele frequencies

We also need allele frequency metrics per locus. To do this, we need to get the allele count data from the .genos files and compile.

*Note that reading in all of the .genos files will give a warning; this is because these files are output with two extra (empty) columns*

```{r, warning = F}
# use full.names = T to list full sample path (avoid changing wd)
genoFiles_gtseq <- list.files(path = "/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq/genos", pattern = "*interleaved.genos", full.names = T)

# use skip = 1 to skip reading the header
alleleReads_gtseq <- lapply(genoFiles_gtseq, fread, skip = 1, sep = ",", header = F, select = c("V1", "V2", "V3")) %>%
  set_names(genoFiles_gtseq) %>%
  rbindlist(idcol = "sample") %>%
  as.data.frame() %>%
  dplyr::mutate(sample = sub(".*genos/", "", sample)) %>%
  dplyr::mutate(sample = sub("\\_.*", "", sample)) %>%
  dplyr::mutate(sample = gsub("-","\\.", sample)) %>%
  dplyr::rename("locus" = "V1") %>%
  dplyr::mutate(allele1 = substr(V2, 1, 1)) %>%
  dplyr::mutate(allele2 = substr(V3, 1, 1)) %>%
  dplyr::mutate(allele1Reads = sub(".*=", "", V2)) %>%
  dplyr::mutate(allele2Reads = sub(".*=", "", V3)) %>%
  dplyr::mutate_at(c("allele1Reads", "allele2Reads"), as.numeric) %>%
  select(c(sample, locus, allele1, allele2, allele1Reads, allele2Reads))

alleleReads_lwedSIMP <- alleleReads_gtseq %>%
  filter(sample %in% samples_noNeg$sampleID) %>%
  dplyr::group_by(locus, allele1, allele2) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame()

alleleReads_lwed <- alleleReads_gtseq %>%
  filter(sample %in% samplesLWED$sampleID) %>%
  dplyr::group_by(locus, allele1, allele2) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame()

alleleReads_simp <- alleleReads_gtseq %>%
  filter(sample %in% samplesSIMP$sampleID) %>%
  dplyr::group_by(locus, allele1, allele2) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame()
```

Then we can combine the allele read-counts per locus, again making species-specific subsets with species-specific loci that only reflect read counts for samples of that species:

```{r}
alleleRC <- alleleReads_lwedSIMP %>%
  filter(!str_detect(locus, "LWED|SIMP"))
alleleRC_lwed <- alleleReads_lwed %>%
  filter(str_detect(locus, "LWED"))
alleleRC_simp <- alleleReads_simp %>%
  filter(str_detect(locus, "SIMP"))

alleleReads_combine <- rbind(alleleRC, alleleRC_lwed, alleleRC_simp)
```

#### Full locus summary

And finally we can combine total read counts and allele read counts into a single locus summary file for GTseq:

```{r}
# Combined
locusSum_gtseq <- merge(lociReadCounts_combine, alleleReads_combine, by = "locus") %>%
  dplyr::mutate(alleleReadsTotal = allele1Reads + allele2Reads) %>%
  dplyr::mutate(maf_gtseq = pmax(allele1Reads, allele2Reads)/alleleReadsTotal)
```

View GTseq locus summary:

```{r, echo = F}
locusSum_gtseq %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 3.3 GTscore data

The genotypes and locus/sample metrics for GTscore are from the GT-score pipeline, which includes the original GT-score scripts as well as my own additions. See GTscorePipeline_tamRun3.Rmd for the full detailed pipeline.

### Genotypes

*Note that I adjusted the original GTscore pipeline so that genotypes are only called for loci with at least 10x coverage.*

```{r}
# Genotypes
genos_gtscore <- read.table("./03_tamRun3/03_run3GTscore/fullSet_polyGenResults_singleSNP_10x.txt", header = T)
```

### Locus summaries

The GTscore pipeline includes scripts that provide locus-specific summary metrics; e.g., total primer reads, probe reads, genotype rate, etc. per locus across samples.

I compiled the locus-specific metrics from the GTscore pipeline into a "master summary", in which...

-   values for LWED-specific loci reflect performance with LWED individuals only
-   values for SIMP-specific loci reflect performance with SIMP individuals only
-   all other loci reflect performance with LWED and SIMP individuals combined

```{r}
locusSum_gtscore <- read.csv("/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/03_tamRun3/03_run3GTscore/summaryFiles/master_locusSummary.csv")
```

View GTscore locus summary:

```{r, echo = F}
locusSum_gtscore %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Sample summaries

GTscore individual (sample) summary

```{r}
indivSum_gtscore <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)
```

View GTscore individual (sample) summary:

```{r, echo = F}
indivSum_gtscore %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

GTscore single-SNP summary

```{r}
ssSum_gtscore <- read.csv("./03_tamRun3/03_run3GTscore/fullSet_singleSNP_sampleSummary.csv")
```

View GTscore single-SNP summary:

```{r, echo = F}
ssSum_gtscore %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

# 4 Sample performance comparison

For the first set of comparisons between the two pipelines, I'm looking at results relative to samples (vs. relative to loci; see section 5). Specifically, I'm looking at what each pipeline shows relative to:

1.  Total reads per sample
2.  On-target reads per sample
3.  Genotype success per sample
4.  Genotypes assigned per sample

## 4.1 Total reads

Both GTseq and GTscore pipelines give the same number of total reads per sample.

```{r}
totalReads <- indivSum_gtscore %>%
  select(c(Sample, Total.Reads)) %>%
  dplyr::rename("totalReads_gtscore" = "Total.Reads") %>%
  merge(., genos_gtseq[, c("Sample", "Raw.Reads")], by = "Sample") %>%
  dplyr::rename("totalReads_gtseq" = "Raw.Reads") %>%
  mutate(totalReads_diff = totalReads_gtseq - totalReads_gtscore) %>%
  mutate(totalReads_avg = (totalReads_gtseq + totalReads_gtscore)/2)
```

```{r}
summary(totalReads$totalReads_diff)
```

## 4.2 On-target reads

On-target read counts are on average 0.34% HIGHER for GTseq vs. GTscore (p-value \< 2.2e-16).

```{r}
otReads <- indivSum_gtscore %>%
  select(c(Sample, Primer.Probe.Reads)) %>%
  dplyr::rename("otReads_gtscore" = "Primer.Probe.Reads") %>%
  select(c(Sample, otReads_gtscore)) %>%
  merge(., genos_gtseq[, c("Sample", "On.Target.Reads")], by = "Sample") %>%
  dplyr::rename("otReads_gtseq" = "On.Target.Reads") %>%
  mutate(otReads_avg = (otReads_gtseq + otReads_gtscore)/2) %>%
  mutate(otReads_diff = otReads_gtseq - otReads_gtscore) %>%
  mutate(otReads_propDiff = (otReads_diff)/otReads_avg) %>%
  relocate(otReads_gtseq, .after = Sample)
```

View GTseq vs. GTscore on-target reads (highest to lowest % difference):

```{r, echo = F}
otReads %>%
  arrange(desc(otReads_propDiff)) %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

```{r}
# Summarize percent difference b/t GTseq and GTscore on-target reads
summary(otReads$otReads_propDiff)

# t-test for GTseq vs GTscore on-target reads
t.test(otReads$otReads_gtseq, otReads$otReads_gtscore, paired = T)
```

```{r, include = F}
# Quick normality test - neither GTseq or GTscore on-target reads have a normal distribution
shapiro.test(otReads$otReads_gtseq)
shapiro.test(otReads$otReads_gtscore)
```

### Visualize on-target reads

```{r, include = F, fig.show = F}
# Plotting the log-transformed on-target reads show that GTseq on-target reads increase 1% for every 1% increase in GTscore on-target reads
ggplot(otReads, aes(x = log(otReads_gtseq), y = log(otReads_gtscore))) +
  geom_point() +
  geom_smooth(method = "lm", se=FALSE, color="mediumvioletred", formula = y ~ x, size = 0.75) +
  ggpubr::stat_regline_equation(label.y = 11, aes(label = ..eq.label..)) +
  ggpubr::stat_regline_equation(label.y = 10, aes(label = ..rr.label..)) +
  xlab("log(GTseq) on-target read counts per sample") +
  ylab("log(GTscore) on-target read counts per sample") +
  theme_pander()
```

We can visualize the differences in on-target read counts using a Bland-Altman plot. Here I'm plotting the mean (pink line) +/- standard deviation (blue lines) of the percent difference in reads between GTseq and GTscore (grey circles) relative to the total on-target reads for that sample (an average of the values provided by each pipeline).

*Note that the "Percent difference" here is equivalent to log(GTseq) - log(GTscore)*

```{r, echo = F, warning = F}
ggplot(otReads, aes(x = otReads_avg, y = otReads_propDiff)) +
  geom_point(alpha = 0.5) +
  geom_hline(yintercept = mean(otReads$otReads_propDiff, na.rm = T), colour = "mediumvioletred", size = 0.5) +
  geom_hline(yintercept = mean(otReads$otReads_propDiff, na.rm = T) - (1.96 * sd(otReads$otReads_propDiff, na.rm = T)), colour = "cornflowerblue", size = 0.5) +
  geom_hline(yintercept = mean(otReads$otReads_propDiff, na.rm = T) + (1.96 * sd(otReads$otReads_propDiff, na.rm = T)), colour = "cornflowerblue", size = 0.5) +
  annotate(geom = "label", label = "Mean diff = 0.034", x = 90000, y = mean(otReads$otReads_propDiff, na.rm = T), hjust = 0, vjust = 0) +
  ylab("Percent difference \n (GTseq - GTscore)/((GTseq + GTscore)/2)") +
  xlab("Mean on-target read counts per sample \n ((GTseq + GTscore)/2)") +
  theme_pander()
```

## 4.4 Genotype success

The percent of loci successfully genotyped per sample was 3.2% LOWER for GTseq vs. GTscore (p-value \< 2.2e-16).

```{r}
genoRate <- ssSum_gtscore %>%
  select(c(sample, GenotypeRate)) %>%
  dplyr::rename("Sample" = "sample") %>%
  dplyr::rename("genoRate_gtscore" = "GenotypeRate") %>%
  select(c(Sample, genoRate_gtscore)) %>%
  merge(., genos_gtseq[, c("Sample", "X.GT")], by = "Sample") %>%
  dplyr::rename("genoRate_gtseq" = "X.GT") %>%
  mutate(genoRate_gtseq = genoRate_gtseq/100) %>%
  mutate_if(is.numeric, round, 2) %>%
  mutate(genoRate_avg = (genoRate_gtseq + genoRate_gtscore)/2) %>%
  mutate(genoRate_diff = genoRate_gtseq - genoRate_gtscore) %>%
  mutate(genoRate_propDiff = genoRate_diff/genoRate_avg) %>%
  relocate(genoRate_gtseq, .after = Sample)
```

View GTseq vs. GTscore genotype success:

```{r, echo = F}
genoRate %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

```{r}
summary(genoRate$genoRate_diff, na.rm = T)
t.test(genoRate$genoRate_gtseq, genoRate$genoRate_gtscore, paired = T)
```

```{r, include = F}
# Quick normality test - neither GTseq or GTscore genotype rates have a normal distribution
shapiro.test(genoRate$genoRate_gtseq)
shapiro.test(genoRate$genoRate_gtscore)
```

### Visualize genotype success

```{r, include = F, fig.show = F}
# Plotting the log-transformed genotype rates show that GTseq on-target reads increase 1% for every 1% increase in GTscore on-target reads
ggplot(genoRate, aes(x = log(genoRate_gtseq), y = log(genoRate_gtscore))) +
  geom_point() +
  geom_smooth(method = "lm", se=FALSE, color="mediumvioletred", formula = y ~ x, size = 0.75) +
  ggpubr::stat_regline_equation(label.y = -0.5, aes(label = ..eq.label..)) +
  ggpubr::stat_regline_equation(label.y = -1, aes(label = ..rr.label..)) +
  xlab("log(GTseq) % loci genotyped per sample") +
  ylab("log(GTscore) % loci genotyped per sample") +
  theme_pander()
```

We can use a Bland-Altman plot again to visualize the difference in genotype success between pipelines. Here I'm plotting the mean (pink line) +/- standard deviation (blue lines) of the differences in % loci genotyped per sample between GTseq and GTscore (grey circles) relative to the total % loci genotyped for that sample (an average of the values provided by each pipeline).

```{r, echo = F, warning= F}
ggplot(genoRate, aes(x = genoRate_avg, y = genoRate_diff)) +
  geom_point(alpha = 0.5) +
  geom_hline(yintercept = mean(genoRate$genoRate_diff, na.rm = T), colour = "springgreen4", size = 0.5) +
  geom_hline(yintercept = mean(genoRate$genoRate_diff, na.rm = T) - (1.96 * sd(genoRate$genoRate_diff, na.rm = T)), colour = "cornflowerblue", size = 0.5) +
  geom_hline(yintercept = mean(genoRate$genoRate_diff, na.rm = T) + (1.96 * sd(genoRate$genoRate_diff, na.rm = T)), colour = "cornflowerblue", size = 0.5) +
  annotate(geom = "label", label = "Mean diff = -0.032", x = 0, y = mean(genoRate$genoRate_diff, na.rm = T), hjust = 0, vjust = 0, fontface = "bold") +
  ylab("Difference in % loci genotyped \n (GTseq - GTscore)") +
  xlab("Mean % loci genotyped per sample \n ((GTseq + GTscore)/2)") +
  theme_pander()
```

## 4.5 Genotypes assigned

To compare genotypes called by each pipeline, we first need to set up a genotype comparison dataframe:

```{r}
genos_gtscore_l <- genos_gtscore %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub("_[^_]+$", "", locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("gtscore" = "genotype") %>%
  mutate(gtscore = sub(",", "", gtscore))
  
genos_gtseq_l <- genos_gtseq %>%
  select(!c(Raw.Reads, On.Target.Reads, X.On.Target, X.GT, IFI)) %>%
  pivot_longer(!Sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("gtseq" = "genotype") %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  rename("sample" = "Sample")

genos_seqScore <- merge(genos_gtscore_l, genos_gtseq_l, by = c("sample", "locus"), all = T) %>%
  mutate(genosCalled = rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  mutate(zerosCalled = rowSums(select(., -sample, -locus, -genosCalled) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos = ifelse(zerosCalled > 0,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T)) - 1,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T))
                              )) %>%
  relocate(gtseq, .after = locus)
```

View head() GTseq vs. GTscore genotypes:

```{r, echo = F}
head(genos_seqScore) %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

Our comparison dataframe shows that of the **96,798** potential sample/loci combinations for which a genotype could be called, **38,007** genotypes were called by both pipelines.

```{r, collapse = T}
length(genos_seqScore$sample)
length(which(genos_seqScore$genosCalled == 2))
```

Of these **38,007** genotypes, there were no sample/loci combinations where GTseq and GTscore made different genotype calls (maximum number of unique genotypes called between both pipelines = 1).

```{r, collapse = T}
max(genos_seqScore$uniqueGenos)
```

There were **3,262** sample/loci combinations where only one pipeline called a genotype.

Of these, **GTseq called 87 genotypes where GTscore did not**, and **GTscore called 3,175 genotypes when GTseq did not**.

```{r, collapse = T}
genos_seqScore_zeros <- genos_seqScore %>%
  filter(genosCalled == 1)
length(genos_seqScore_zeros$sample)
length(which(genos_seqScore_zeros$gtscore == 0))
length(which(genos_seqScore_zeros$gtseq == 0))
```

Perhaps of note, of the instances where GTseq called genotypes when GTscore didn't, the vast majority were homozygous.

```{r, collapse = T}
genos_zeros_score <- genos_seqScore_zeros %>%
  filter(gtscore == 0)
table(genos_zeros_score$gtseq)
```

## 4.6 Drop 10x cutoff

Since the 10x coverage cutoff was something that I added to the GTscore pipeline, I also wanted to see if there were any differences if I removed this cutoff.

```{r}
genos_gtscore_no10x <- read.table("./03_tamRun3/03_run3GTscore/fullSet_polyGenResults_singleSNP_no10x.txt", header = T)

genos_gtscore_no10x_l <- genos_gtscore_no10x %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub("_[^_]+$", "", locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("gtscore" = "genotype") %>%
  mutate(gtscore = sub(",", "", gtscore))

genos_seqScore_no10x <- merge(genos_gtscore_no10x_l, genos_gtseq_l, by = c("sample", "locus"), all = T) %>%
  mutate(genosCalled = rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  mutate(zerosCalled = rowSums(select(., -sample, -locus, -genosCalled) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos = ifelse(zerosCalled > 0,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T)) - 1,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T))
                              ))
```

Removing the 10x cutoff gives **11,498 additional genotypes** (a 21.8% increase) from GTscore (52,680 genotypes without 10x cutoff vs. 41,182 genotypes with cutoff).

```{r}
# Number of GTscore genotypes WITHOUT 10x cutoff
length(which(!genos_gtscore_no10x_l$gtscore == 0))

# Number of GTscore genotypes WITH 10x cutoff
length(which(!genos_gtscore_l$gtscore == 0))

# Difference b/t the two
length(which(!genos_gtscore_no10x_l$gtscore == 0)) - length(which(!genos_gtscore_l$gtscore == 0))

# Percent increase
((length(which(!genos_gtscore_no10x_l$gtscore == 0)) - length(which(!genos_gtscore_l$gtscore == 0)))/length(which(!genos_gtscore_no10x_l$gtscore == 0)))*100
```

After removing the 10x cutoff for the GTscore genotypes, there are still no sample/loci combinations where GTseq and GTscore made different genotype calls (maximum number of unique genotypes called between both pipelines = 1).

```{r, collapse = T}
max(genos_seqScore_no10x$uniqueGenos)
```

Overall, there were **14,700 (vs. 3,262)** sample/loci combinations where only one pipeline called a genotype.

Of these, **GTseq called 57 (vs. 87) genotypes where GTscore did not**, and **GTscore called 14,643 (vs. 3,175) genotypes when GTseq did not**.

```{r, collapse = T}
genos_seqScore_zeros_no10x <- genos_seqScore_no10x %>%
  filter(genosCalled == 1)
length(genos_seqScore_zeros_no10x$sample)
length(which(genos_seqScore_zeros_no10x$gtscore == 0))
length(which(genos_seqScore_zeros_no10x$gtseq == 0))
```

Similar to the 10x gtseq genotype comparison, of the instances where GTseq called genotypes when GTscore didn't, all calls were homozygous.

```{r, collapse = T}
genos_no10x_zeros_score <- genos_seqScore_zeros_no10x %>%
  filter(gtscore == 0)
table(genos_no10x_zeros_score$gtseq)
```

# 5 Locus performance comparison

For the second set of comparisons between the two pipelines, I'm looking at results relative to loci (vs. relative to samples; see section 4). Specifically, I'm looking at what each pipeline shows in relation to:

1.  The proportion of primer-probe reads relative to primer reads
2.  Whether any loci are fixed
3.  Whether any loci are overamplifying relative to the others

## 5.1 Primer-probe proportion

The proportion of primer-probe reads to primer reads is helpful in assessing primer performance, where large discrepancies in primer, probe, and primer-probe reads can indicate artifact sequences, unaccounted for in-silico probe variations, or an in-silico probe sequence that's not specific to the target amplicon. As such, it's important to determine whether the GTseq and GTscore pipelines are providing similar values for primer, probe, and primer-probe reads.

GTseq documentation provides the following as examples of read counts for poorly performing loci (locus name, primer reads, probe reads, primer-probe reads):

-   Example 1: Artifact sequences.
    -   LocusName1, 325096 ,2609, 2367\
-   Example 2: Unaccounted for in-silico probe variation or possibly entirely off-target amplification.
    -   LocusName2, 32252, 0, 0
-   Example 3: in-silico probe not specific enough.
    -   LocusName3, 34435, 3400826, 33213

The proportion of primer-probe reads to primer reads was 0.47% (0.0047) HIGHER with GTseq vs. GTscore (p-value = 2.182e-08). Given this small difference, it looks like both are providing similar values for primer-probe reads:primer reads.

```{r}
ppProp <- locusSum_gtscore %>%
  select(c(Locus, Primer.Probe.Proportion)) %>%
  dplyr::rename("locus" = "Locus") %>%
  dplyr::rename("ppProp_gtscore" = "Primer.Probe.Proportion") %>%
  merge(., locusSum_gtseq[, c("locus", "primerProbeProportion")], by = "locus") %>%
  dplyr::rename("ppProp_gtseq" = "primerProbeProportion") %>%
  mutate(ppProp_diff = ppProp_gtseq - ppProp_gtscore) %>%
  mutate_if(is.numeric, round, 2) %>%
  relocate(ppProp_gtseq, .after = locus)
```

View GTseq vs. GTscore primer-probe:primer proportions:

```{r, echo = F}
ppProp %>%
  kbl(row.names = F) %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

```{r}
summary(ppProp$ppProp_diff)
t.test(ppProp$ppProp_gtseq, ppProp$ppProp_gtscore, paired = T)
```

Of note, there are 35 loci for which the proportion of primer-probe to primer reads is under 0.5. These loci should be assessed on a case-by-case basis to determine why the primers are amplifying sequences that don't contain our SNP of interest, and may be candidates for removal from the final primer pool.

View loci with primer-probe:primer reads \< 0.5:

```{r, echo = F}
ppProp %>%
  filter(ppProp_gtscore < 0.5) %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 5.2 Fixed loci

Any loci with fixed alleles are ones that we would ultimately want to take out of the primer pool. However, if they're not 100% fixed it would be better to keep them since these small variations would be informative.

GTscore gives two loci as fixed. GTseq, however, shows these same loci as *almost* fixed, but not 100%. Given that there aren't any mismatched genotypes (see section 4.5), we can assume that the allele reads for those "almost-fixed" loci are from the small handful of cases where GTseq gave a genotype, but GTscore did not.

```{r}
# Dataframe comparing major allele frequencies
maf <- locusSum_gtscore %>%
  dplyr::rename("locus" = "Locus") %>%
  dplyr::rename("maf_gtscore" = "majAF") %>%
  select(c(locus, maf_gtscore)) %>%
  merge(., locusSum_gtseq[, c("locus", "maf_gtseq")], by = "locus") %>%
  dplyr::mutate(mafDiff = maf_gtseq - maf_gtscore) %>%
  relocate(maf_gtseq, .after = locus) %>%
  arrange(desc(maf_gtscore))
summary(maf$mafDiff)
```

View GTseq vs. GTscore major allele frequencies (highest to lowest):

```{r, echo = F}
maf %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

## 5.3 Overamplifiers

Loci that overamplify relative to others in the pool are also important to identify, either as candidates for removal or to adjust the volume in which they're added to the pool so that they're not taking reads away from other loci.

To identify loci that are overamplifying in each pipeline, I'm looking at primer-probe read counts for each locus in each pipeline with species-specific subsets.

### Primer-probe reads

First we'll set up species-specific dataframes for primer-probe reads from GTseq and GTscore and GTscore pipelines

**Full set of primer-probe read counts from each pipeline**

Note that I'm recoding all species-specific primer-probe reads as NA for species-specific loci that do not match the sample species.

```{r}
# GTseq
ppReads_gtseq <- seqTest_compiled %>%
  select(!c(primerReads, probeReads)) %>%
  mutate(primerProbeReads =
           ifelse(sample %in% samplesLWED$sampleID & grepl("SIMP", locus), NA, 
                  ifelse(sample %in% samplesSIMP$sampleID & grepl("LWED", locus), NA, primerProbeReads))) %>%
  dplyr::rename("ppReads_gtseq" = "primerProbeReads")
  
# GTscore
repl <- function(x) gsubfn("(\\d+),(\\d+)", ~ as.numeric(x) + as.numeric(y), paste(x))

ppReads_gtscore <- read.table("./03_tamRun3/03_run3GTscore/fullSet_AlleleReads_singleSNPs.txt") %>%
  replace(., TRUE, lapply(., repl)) %>%
  rownames_to_column("Locus") %>%
  mutate(Locus = sub("_[^_]+$", "\\1", Locus)) %>%
  column_to_rownames("Locus") %>%
  rownames_to_column("locus") %>%
  pivot_longer(!locus,
               names_to = "sample",
               values_to = "ppReads_gtscore") %>%
  mutate(ppReads_gtscore =
           ifelse(sample %in% samplesLWED$sampleID & grepl("SIMP", locus), NA, 
                  ifelse(sample %in% samplesSIMP$sampleID & grepl("LWED", locus), NA, ppReads_gtscore)))
```

**Species subsets**

```{r}
# Merge dataframes and add species
ppReads_seqScore <- merge(ppReads_gtseq, ppReads_gtscore, by = c("sample", "locus")) %>%
  mutate(species = ifelse(sample %in% samplesLWED$sampleID, "LWED",
                          ifelse(sample %in% samplesSIMP$sampleID, "SIMP", NA)))

# LWED primer-probe reads
ppReads_seqScore_lwed <- ppReads_seqScore %>%
  filter(species == "LWED") %>%
  select(!c(sample, species)) %>%
  mutate(ppReads_gtseq = as.numeric(ppReads_gtseq)) %>%
  mutate(ppReads_gtscore = as.numeric(ppReads_gtscore)) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  na.omit()

# SIMP primer-probe reads
ppReads_seqScore_simp <- ppReads_seqScore %>%
  filter(species == "SIMP") %>%
  select(!c(sample, species)) %>%
  mutate(ppReads_gtseq = as.numeric(ppReads_gtseq)) %>%
  mutate(ppReads_gtscore = as.numeric(ppReads_gtscore)) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  na.omit()
```

**Summary dataframe with species-specific read counts for each locus**

```{r}
ppReads_seqScore_summary <- rbind(ppReads_seqScore_lwed %>%
                                    mutate(species = "LWED"), ppReads_seqScore_simp %>%
                                    mutate(species = "SIMP"))
```

View GTseq vs. GTscore primer-probe reads (highest to lowest):

```{r, echo = F}
ppReads_seqScore_summary %>%
  arrange(desc(ppReads_gtseq)) %>%
  kbl(row.names = F) %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

### Identify overamplifiers

Next we need define a function to find outliers (i.e., overamplifying loci). The function is set so that it will return the name of the locus when applied to our dataset.

```{r}
findoutlier <- function(x) {
  return(x < quantile(x, .25) - 1.5*IQR(x) | x > quantile(x, .75) + 1.5*IQR(x))
}
```

After applying the function, we can see that the same overamplifiers occur in both pipelines for each species.

```{r}
# LWED
ppReads_outliers_lwed <- ppReads_seqScore_lwed %>%
  mutate(gtseqOutlier = ifelse(findoutlier(ppReads_gtseq), "yes", NA)) %>%
  mutate(gtscoreOutlier = ifelse(findoutlier(ppReads_gtscore), "yes", NA)) %>%
  filter_at(vars(gtseqOutlier, gtscoreOutlier), all_vars(!is.na(.))) %>%
  mutate(species = "LWED")

# SIMP
ppReads_outliers_simp <- ppReads_seqScore_simp %>%
  mutate(gtseqOutlier = ifelse(findoutlier(ppReads_gtseq), "yes", NA)) %>%
  mutate(gtscoreOutlier = ifelse(findoutlier(ppReads_gtscore), "yes", NA)) %>%
  filter_at(vars(gtseqOutlier, gtscoreOutlier), all_vars(!is.na(.))) %>%
  mutate(species = "SIMP")
```

### Overamplifier summary

We can view the primer-probe read counts for each overamplifier via the summary dataframe below, which contains:

-   Generic loci overamplifiers (i.e., those that are informative for both species), with species-specific read counts from each pipeline
-   Species-specific loci overamplifiers, with species-specific read counts from each pipeline

```{r}
ppReads_outliers_summary <- ppReads_outliers_lwed %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  rbind(.,
        ppReads_outliers_simp %>%
          filter(!str_detect(locus, "LWED|SIMP"))) %>%
  rbind(.,
        ppReads_outliers_lwed %>%
          filter(str_detect(locus, "LWED"))) %>%
  rbind(.,
        ppReads_outliers_simp %>%
          filter(str_detect(locus, "SIMP"))) %>%
  na.omit()
```

View GTseq vs. GTscore overamplifiers (highest to lowest primer-probe read counts):

```{r, echo = F}
ppReads_outliers_summary %>%
  arrange(desc(ppReads_gtseq)) %>%
  kbl() %>%
  kable_material(full_width = F) %>%
  row_spec(0, color = "black", background = "darkgrey") %>%
  scroll_box(width = "100%", height = "200px", fixed_thead = list(enabled = T))
```

Across both species, there are **28 overamplifiers total**. This includes **13 generic loci** (6 of which appear as overamplifiers for both species, 7 of which are overamplifiers only among LWED samples), **5 LWED-specific loci**, and **10 SIMP-specific loci**.

```{r, collapse = T}
# Total overamplifiers across species
length(unique(ppReads_outliers_summary$locus))

# Total generic overamplifiers
ppReads_outliers_summary %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  distinct(locus) %>%
  tally()

# Shared generic overamplifiers
ppReads_outliers_shared <- ppReads_outliers_lwed %>%
  filter(locus %in% ppReads_outliers_simp$locus) %>%
  rbind(.,
        ppReads_outliers_simp %>%
          filter(locus %in% ppReads_outliers_lwed$locus))

ppReads_outliers_shared %>%
  distinct(locus) %>%
  tally()

# LWED overamplifiers
## LWED generic loci
ppReads_outliers_lwed %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  filter(!locus %in% ppReads_outliers_simp$locus) %>%
  tally()

## LWED-specific loci
ppReads_outliers_summary %>%
  filter(str_detect(ppReads_outliers_summary$locus, "LWED")) %>%
  tally()

# SIMP overamplifiers
## SIMP generic loci
ppReads_outliers_simp %>%
  filter(!str_detect(locus, "LWED|SIMP")) %>%
  filter(!locus %in% ppReads_outliers_lwed$locus) %>%
  tally()

## SIMP-specific loci
ppReads_outliers_summary %>%
  filter(str_detect(ppReads_outliers_summary$locus, "SIMP")) %>%
  tally()
```

### Visualize overamplifiers

Below are boxplots showing primer-probe reads per locus within LWED and SIMP sample subsets, where generic overamplifiers present in both species are highlighted. Primer-probe reads counts are based on GTscore values.

```{r, echo = F}
ggplot(ppReads_seqScore_summary, aes(x=ppReads_gtscore, y=species)) +
  geom_boxplot() +
  geom_point(data = ppReads_outliers_shared,
             aes(color = locus)) +
  labs(title = "Total primer-probe reads per locus (GTscore)", x = "Total primer-probe reads per locus", y = "", color = "Shared overamplifiers") +
  theme_pander()
```
