---
title: "tamGenetics_gtseq_vs_gtscore"
author: "Rachel Voyt"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Overview

This document is to assess the performance of GTscore vs. GTseq data analysis pipelines. I'm using the data from tamRun3 to compare genotype success, the actual genotypes called, and loci performance. 

# 2 Packages

```{r}
library(tidyverse)
```

# 3 Metadata

```{r}
md_tamRun3 <- read.csv("./03_tamRun3/03_run3GTscore/tamRun3_metadata_12Jan2023.csv") %>%
  mutate(sampleID = gsub("_","\\.", sampleID))

samples_noNeg <- md_tamRun3 %>%
  filter(species %in% c("LWED", "SIMP")) %>%
  select(sampleID)

samplesLWED <- md_tamRun3 %>%
  filter(species == "LWED") %>%
  select(sampleID)

samplesSIMP <- md_tamRun3 %>%
  filter(species == "SIMP") %>%
  select(sampleID)
```

# 4 GT-seq pipeline

I acquired the scripts for the GT-seq data analysis pipeline from Nate Campbell's github (here), following the steps outlined in GTseq_Pipeline.txt. 

Note that while the perl scripts are taken directly from the original pipeline, I've added information below on how to run each as a loop to allow analysis of all sample files at once, as this was not included in the original GTseq pipeline. 

## 4.1 Split raw fastq sequence file into individual fastq files using dual index sequences

No need to complete this step; fastq files were split automatically by MiSeq

## 4.2 Primer-probe counts

This script counts the occurrence of each forward primer sequences, in-silico probe sequences, and when both occur within the same sequence read. This can help identify primers that create large numbers of artifact sequences.

Note - this is included as an optional step in the GT-seq pipeline

### GTseq_HashSeqs.pl

First use the GTseq_HashSeqs.pl script on plate .fastq files.
This script collects and counts unique reads within the fastq file and reduces the compute time for the GTseq_SeqTest.pl script.
\$ perl GTseq_HashSeqs.pl i20_plate123.fastq \> i20_plate123.hash

```{bash}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq

for i in ./../03_tamRun3/02_run3Interleaved/tamRun3*; do perl GTseq_HashSeqs.pl $i > ./hashSeqs/$(basename "${i/.fastq/.hash}"); done
```

### GTseq_SeqTest.pl

Next use the GTseq_SeqTest.pl script by supplying a tab delimited text file containing loci name, forward primer sequence, allele 1 probe, and allele 2 probe.

Note: The script checks for both the supplied sequence and the reverse complement of the in-silico probes.

#### Primer-probe file

Create primer-probe file formatted for the GTseq_SeqTest.pl script

```{r}
pp_fullSet <- read.table("./03_tamRun3/03_run3GTscore/primerProbeFile_fullSet.txt", header = T)

pp_gtseq_seqTest <- pp_fullSet %>%
  select(c(Locus, Primer, Probe1, Probe2))

write.table(pp_gtseq_seqTest,"./gtscore_vs_gtseq/primerProbe_gtseq.txt", quote = F, sep = "\t", col.names = F, row.names = F)
```

#### Run script

```{bash}
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq

for i in ./hashSeqs/tamRun3*; do perl GTseq_SeqTest.pl primerProbe_gtseq.txt $i > ./seqTest/$(basename "${i/.hash/.seqtest.csv}"); done
```

## 4.3 GTseq_Genotyper_v3.pl

Genotype individuals using GTseq_Genotyper.pl script or GTseq_Genotyper_v2.pl script which allows for count corrections for loci which amplify

Supply Loci information file containing locus names, allele names, and in-silico probe sequences in .csv format.

```{r}
pp_gtseq_genotyper <- pp_fullSet %>%
  select(c(Locus, Allele1, Allele2, Probe1, Probe2, Primer))

write.table(pp_gtseq_genotyper, "./gtscore_vs_gtseq/locusInfo_gtseq.csv", col.names = F, row.names = F, sep = ",", quote = F)
```

Going to use GTseq_Genotyper_v3.pl

Output files contain a header line with summary information followed by locus specific data.  Locus specific output fields are LocusName, Allele1_counts, Allele2_counts, A1/A2-ratio, Genotype, Genotype_Class, A1_correction value, A2_correction value, On_target reads, Locus OT_percentage, Locus On-target reads as percentage of total on-target reads

```{r}
# original example from gtseq
perl GTseq_Genotyper.pl LocusInfo.csv i20_90_P0123_IndividualID.fastq > i20_90_P0123_IndividualID.genos &
  
# first get into proper directory
cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq

# Genotyping loop
for i in ./../03_tamRun3/02_run3Interleaved/tamRun3*; do perl GTseq_Genotyper_v3.pl locusInfo_gtseq.csv $i > ./genos/$(basename "${i/.fastq/.genos}"); done
```

## 4.4 GTseq_GenoCompile_v3.pl

Compile genotypes from individual genotype files from GTseq_Genotyper_v3 output ".genos" files

This version utilizes the expanded output from the GTseq_Genotyper_v3 script to gather summary data and does not require the individual fastq files.

Includes IFI score (Individual Fuzziness Index) as an indication of possible contamination of DNA from another individual.

For optional output formats use arguments.
N for numeric genotypes or C for allele counts.
Defaults to SNP genotypes.

Optional filtered output requires 2 argument values.
Genotype output type and a genotyping threshold [S,N, or C] [90]

-   example: \$ GTseq_GenoCompile_v3.pl S 90 (outputs SNP genotypes for individual .genos files with 90% or higher genotyping percentage)
-   genotypes for individuals with less than the threshold genotyping percentage are converted to "00".

```{r}
perl GTseq_GenoCompile.pl > Library_Genotypes.csv

cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq/genos

perl ./../GTseq_GenoCompile_v3.pl > compiledGenos_gtseq_tamRun3.csv
```

## 4.5 GTseq_GenoCompile_Counts.pl

note that the .fastq and .genos files need to be in the same directory for this to work; I also haven't gotten it working 100%, it puts out tamRun3_001 as a column for some reason, but have to check if it affects allele read counts or not

```{bash}
# original
perl GTseq_GenoCompile_Counts.pl > Library_Counts.csv

cd /home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq/test/

# test
perl ./../GTseq_GenoCompile_Counts.pl > Library_Counts.csv

```


# 5 GTscore vs. GTseq

## 5.1 Data

*GTscore*

Read in info for the full set (vs. those separated by species) to match with gtseq

```{r}
# Genotypes
genos_gtscore <- read.table("./03_tamRun3/03_run3GTscore/fullSet_polyGenResults_singleSNP.txt", header = T)

# GTscore individual summary
indivSum_gtscore <- read.delim("./03_tamRun3/03_run3GTscore/fullSet_GTscore_individualSummary.txt",header=TRUE,stringsAsFactors=FALSE)

# GTscore single-SNP summary
ssSum_gtscore <- read.csv("./03_tamRun3/03_run3GTscore/fullSet_singleSNP_sampleSummary.csv")
```

*GTseq*

NOTE - genocompiler added an extra blank column to the end; delete this or the csv won't read in correctly

```{r}
# Genotypes
genos_gtseq <- read.csv("./gtscore_vs_gtseq/genos/compiledGenos_gtseq_tamRun3.csv", header = T) %>%
  mutate(Sample = sub("\\_.*", "", Sample))
```

## 5.2 Total reads per sample

Total reads per sample match 100%

```{r}
totalReads <- indivSum_gtscore %>%
  select(c(Sample, Total.Reads)) %>%
  dplyr::rename("totalReads_gtscore" = "Total.Reads") %>%
  merge(., genos_gtseq[, c("Sample", "Raw.Reads")], by = "Sample") %>%
  dplyr::rename("totalReads_gtseq" = "Raw.Reads") %>%
  mutate(totalReads_diff = totalReads_gtscore - totalReads_gtseq)

summary(totalReads$totalReads_diff)
```

## 5.3 On-target reads per sample

Counts of on-target reads were usually LOWER from GTscore vs. GTseq, with a median difference of 39 reads.

```{r}
otReads <- indivSum_gtscore %>%
  select(c(Sample, Primer.Probe.Reads)) %>%
  dplyr::rename("otReads_gtscore" = "Primer.Probe.Reads") %>%
  select(c(Sample, otReads_gtscore)) %>%
  merge(., genos_gtseq[, c("Sample", "On.Target.Reads")], by = "Sample") %>%
  dplyr::rename("otReads_gtseq" = "On.Target.Reads") %>%
  mutate(otReads_diff = otReads_gtscore - otReads_gtseq)

summary(otReads$otReads_diff)
```

## 5.4 Genotype success per sample

Genotype success per sample was usually HIGHER with GTscore vs. GTseq, with a median difference of 2%

```{r}
genoRate <- ssSum_gtscore %>%
  select(c(sample, GenotypeRate)) %>%
  dplyr::rename("Sample" = "sample") %>%
  dplyr::rename("genoRate_gtscore" = "GenotypeRate") %>%
  select(c(Sample, genoRate_gtscore)) %>%
  merge(., genos_gtseq[, c("Sample", "X.GT")], by = "Sample") %>%
  dplyr::rename("genoRate_gtseq" = "X.GT") %>%
  mutate(genoRate_gtseq = genoRate_gtseq/100) %>%
  mutate_if(is.numeric, round, 2) %>%
  mutate(genoRate_diff = genoRate_gtscore - genoRate_gtseq)

summary(genoRate$genoRate_diff)
```

## 5.5 Genotypes

There were no sample/loci combinations where gtscore and gtseq made different genotype calls.

```{r}
genos_gtscore_l <- genos_gtscore %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub("_[^_]+$", "", locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("gtscore" = "genotype") %>%
  mutate(gtscore = sub(",", "", gtscore))
  
genos_gtseq_l <- genos_gtseq %>%
  select(!c(Raw.Reads, On.Target.Reads, X.On.Target, X.GT, IFI)) %>%
  pivot_longer(!Sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("gtseq" = "genotype") %>%
  mutate(Sample = sub("-", "\\.", Sample)) %>%
  rename("sample" = "Sample")

genos_seqScore <- merge(genos_gtscore_l, genos_gtseq_l, by = c("sample", "locus"), all = T) %>%
  mutate(genosCalled = rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  mutate(zerosCalled = rowSums(select(., -sample, -locus, -genosCalled) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos = ifelse(zerosCalled > 0,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T)) - 1,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T))
                              ))

max(genos_seqScore$uniqueGenos)
```

However, there were 3262 sample/loci combinations where only one pipeline called a genotype.

Of these, GTseq called 87 genotypes where GTscore did not, and GTscore called 3175 genotypes when GTseq did not.

Perhaps of note, of the instances where GTseq called genotypes when GTscore didn't, the vast majority were homozygous.

```{r}
genos_seqScore_zeros <- genos_seqScore %>%
  filter(genosCalled == 1)

length(genos_seqScore_zeros$sample)
length(which(genos_seqScore_zeros$gtscore == 0))
length(which(genos_seqScore_zeros$gtseq == 0))
```

## 5.6 Loci performance

### 5.6.1 Prep

##### GTscore

The GTscore pipeline includes scripts that provide locus-specific metrics; e.g., primer reads, probe reads, genotype rate, etc. per locus.

I compiled the locus-specific metrics from the GTscore pipeline into a "master summary", in which...

-   values for LWED-specific loci reflect performance with LWED individuals only
-   values for SIMP-specific loci reflect performance with SIMP individuals only
-   all other loci reflect performance with LWED and SIMP individuals combined

```{r}
locusSum_gtscore <- read.csv("/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/03_tamRun3/03_run3GTscore/summaryFiles/master_locusSummary.csv")
```

##### GTseq 

The GTseq pipeline, however, doesn't have a script that provides metrics per locus from what I could tell. Instead, we need to create this based on the results from GTseq_SeqTest.pl

The GTseq_SeqTest.pl script outputs files for each sample that contain three columns, which represent counts of each...
-   primer
-   probe
-   primer-probe combination

Large discrepancies between values can indicate artifact sequences, unaccounted for in-silico probe variations, or an in-silico probe sequence that's not specific to the target amplicon.

Example 1: Artifact sequences.
		LocusName1,325096,2609,2367
		
Example 2: Unaccounted for in-silico probe variation or possibly entirely off-target amplification.
		LocusName2,32252,0,0
		
Example 3: in-silico probe not specific enough.
		LocusName3,34435,3400826,33213

```{r}
setwd("./gtscore_vs_gtseq/seqTest/")

library(data.table)
library(magrittr)

seqTest_compiled <- lapply(list.files(pattern = "*.csv"), fread, select = c("V1", "V2", "V3", "V4")) %>%
  set_names(list.files(pattern = "*.csv")) %>%
  rbindlist(idcol = "sample") %>%
  as.data.frame() %>%
  dplyr::rename("locus" = "V1",
                "primerReads" = "V2",
                "probeReads" = "V3", 
                "primerProbeReads" = "V4") %>%
  dplyr::mutate(sample = sub("\\_.*", "", sample)) %>%
  dplyr::mutate(sample = sub("-", "\\.", sample))

readCounts_lwedSIMP <- seqTest_compiled %>%
  filter(sample %in% samples_noNeg$sample) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  dplyr::mutate(primerProbeProportion = primerProbeReads/primerReads)

readCounts_lwed <- seqTest_compiled %>%
  filter(sample %in% samplesLWED$sampleID) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  dplyr::mutate(primerProbeProportion = primerProbeReads/primerReads)

readCounts_simp <- seqTest_compiled %>%
  filter(sample %in% samplesSIMP$sampleID) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame() %>%
  dplyr::mutate(primerProbeProportion = primerProbeReads/primerReads)
```

Below is the summary of primer, probe, and primer-probe counts per locus:

```{r}
rc <- locusSum_lwedSIMP %>%
  filter(!str_detect(locus, "LWED|SIMP"))
lwed_rc <- locusSum_lwed %>%
  filter(str_detect(locus, "LWED"))
simp_rc <- locusSum_simp %>%
  filter(str_detect(locus, "SIMP"))

readCounts_combine <- rbind(ls, lwed_ls, simp_ls)
```

However, we also need allele frequency metrics per locus. To do this, we need to get the allele count data from the .genos files and compile.

Note that reading in all of the .genos files will give a warning; this is because these files are output with two extra (empty) columns
```{r}
# use full.names = T to list full sample path
genoFiles_gtseq <- list.files(path = "/home/rachelvoyt/Documents/UT-Grad/Development/repos/tamGenetics_primatesPeru/gtscore_vs_gtseq/genos", pattern = "*interleaved.genos", full.names = T)
head(genoFiles_gtseq)

# use skip = 1 to skip reading the header
alleleReads_gtseq <- lapply(genoFiles_gtseq, fread, skip = 1, sep = ",", header = F, select = c("V1", "V2", "V3")) %>%
  set_names(genoFiles_gtseq) %>%
  rbindlist(idcol = "sample") %>%
  as.data.frame() %>%
  dplyr::mutate(sample = sub(".*genos/", "", sample)) %>%
  dplyr::mutate(sample = sub("\\_.*", "", sample)) %>%
  dplyr::mutate(sample = gsub("-","\\.", sample)) %>%
  dplyr::rename("locus" = "V1") %>%
  dplyr::mutate(allele1 = substr(V2, 1, 1)) %>%
  dplyr::mutate(allele2 = substr(V3, 1, 1)) %>%
  dplyr::mutate(allele1Reads = sub(".*=", "", V2)) %>%
  dplyr::mutate(allele2Reads = sub(".*=", "", V3)) %>%
  dplyr::mutate_at(c("allele1Reads", "allele2Reads"), as.numeric) %>%
  select(c(sample, locus, allele1, allele2, allele1Reads, allele2Reads))

alleleReads_lwedSIMP <- alleleReads_gtseq %>%
  filter(sample %in% samples_noNeg$sampleID) %>%
  dplyr::group_by(locus, allele1, allele2) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame()

test <- alleleReads_lwedSIMP %>%
  distinct(sample)

alleleReads_lwed <- alleleReads_gtseq %>%
  filter(sample %in% samplesLWED$sampleID) %>%
  dplyr::group_by(locus, allele1, allele2) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame()

alleleReads_simp <- alleleReads_gtseq %>%
  filter(sample %in% samplesSIMP$sampleID) %>%
  dplyr::group_by(locus, allele1, allele2) %>%
  dplyr::summarise_if(is.numeric, sum) %>%
  as.data.frame()
```

Now combine allele read info...

```{r}
ar <- alleleReads_lwedSIMP %>%
  filter(!str_detect(locus, "LWED|SIMP"))
lwed_ar <- alleleReads_lwed %>%
  filter(str_detect(locus, "LWED"))
simp_ar <- alleleReads_simp %>%
  filter(str_detect(locus, "SIMP"))

alleleReads_combine <- rbind(ar, lwed_ar, simp_ar)
```

And finally create the full locus summary for GTseq:

```{r}
locusSum_gtseq <- merge(readCounts_combine, alleleReads_combine, by = "locus") %>%
  dplyr::mutate(alleleReadsTotal = allele1Reads + allele2Reads) %>%
  dplyr::mutate(maf_gtseq = pmax(allele1Reads, allele2Reads)/alleleReadsTotal)
```

In looking at the full summary, the total allele reads (adding up allele1Reads and allele2Reads) are often VERY different from the primerProbeReads value-- perhaps this is because the allele reads are based off the genotype files vs. raw data, so it's not counting all the reads from loci that were under 10x? 

### 5.6.2 Primer-probe proportion

The proportion of primer-probe reads to primer reads was overall LOWER with GTscore vs. GTseq, though the mean difference was less than 1%. 

```{r}
ppProp <- locusSum_gtscore %>%
  select(c(Locus, Primer.Probe.Proportion)) %>%
  dplyr::rename("locus" = "Locus") %>%
  dplyr::rename("ppProp_gtscore" = "Primer.Probe.Proportion") %>%
  merge(., locusSum_gtseq[, c("locus", "primerProbeProportion")], by = "locus") %>%
  dplyr::rename("ppProp_gtseq" = "primerProbeProportion") %>%
  mutate(ppProp_diff = ppProp_gtscore - ppProp_gtseq) %>%
  mutate_if(is.numeric, round, 2)

summary(ppProp$ppProp_diff)
t.test(ppProp$ppProp_gtscore, ppProp$ppProp_gtseq, paired = T)
```

### 5.6.3 Fixed alleles

GTscore gives two loci as fixed. GTseq, however, shows these same loci as *almost* fixed, but not 100%. Given that there aren't any mismatched genotypes, we can assume that the allele reads for those "almost-fixed" loci are from the cases where GTseq gave a genotype, but GTscore did not.

```{r}
maf <- locusSum_gtscore %>%
  dplyr::rename("locus" = "Locus") %>%
  dplyr::rename("maf_gtscore" = "majAF") %>%
  select(c(locus, maf_gtscore)) %>%
  merge(., locusSum_gtseq[, c("locus", "maf_gtseq")], by = "locus") %>%
  dplyr::mutate(mafDiff = maf_gtscore - maf_gtseq) %>%
  mutate_if(is.numeric, round, 2)
```

### 5.6.4 Overamplifiers

# X Compare genotypes without the 10x gtscore cutoff

There were still no sample/loci combinations where gtscore and gtseq made different genotype calls.

```{r}
genos_gtscore_no10x <- read.table("./03_tamRun3/03_run3GTscore/fullSet_polyGenResults_singleSNP_no10x.txt", header = T)

genos_gtscore_no10x_l <- genos_gtscore_no10x %>%
  rownames_to_column("locus") %>%
  mutate(locus = sub("_[^_]+$", "", locus)) %>%
  column_to_rownames("locus") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  pivot_longer(!sample,
               names_to = "locus",
               values_to = "genotype") %>%
  dplyr::rename("gtscore" = "genotype") %>%
  mutate(gtscore = sub(",", "", gtscore))

genos_seqScore_no10x <- merge(genos_gtscore_no10x_l, genos_gtseq_l, by = c("sample", "locus"), all = T) %>%
  mutate(genosCalled = rowSums(select(., -sample, -locus) != 0, na.rm = T)) %>%
  mutate(zerosCalled = rowSums(select(., -sample, -locus, -genosCalled) == 0, na.rm = T)) %>%
  rowwise() %>%
  mutate(uniqueGenos = ifelse(zerosCalled > 0,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T)) - 1,
                              (n_distinct(c_across(gtscore:gtseq), na.rm = T))
                              ))

max(genos_seqScore_no10x$uniqueGenos)
```

However, there were 14,700 (vs. 3262) sample/loci combinations where only one pipeline called a genotype.

Of these, GTseq called 57 (vs. 87) genotypes where GTscore did not, and GTscore called 14,643 (vs. 3175) genotypes when GTseq did not.

Similar to the 10x gtseq genotype comparison, of the instances where GTseq called genotypes when GTscore didn't, all calls were homozygous.

```{r}
genos_seqScore_zeros_no10x <- genos_seqScore_no10x %>%
  filter(genosCalled == 1)

length(genos_seqScore_zeros_no10x$sample)
length(which(genos_seqScore_zeros_no10x$gtscore == 0))
length(which(genos_seqScore_zeros_no10x$gtseq == 0))

genos_no10x_zeros_score <- genos_seqScore_zeros_no10x %>%
  filter(gtscore == 0)
table(genos_no10x_zeros_score$gtseq)

genos_no10x_zeros_seq <- genos_seqScore_zeros_no10x %>%
  filter(gtseq == 0)
```


